// @instructure/ui-checkbox/lib/Checkbox/v1/CheckboxFacade/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.CheckboxFacade = void 0;
var _react = require("react");
var _SVGIcon = require("@instructure/ui-svg-images/lib/SVGIcon");
var _IconCheckMarkSolid = require("@instructure/ui-icons/lib/generated/IconCheckMarkSolid.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: Checkbox
---
**/
let CheckboxFacade = exports.CheckboxFacade = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class CheckboxFacade extends _react.Component {
  static displayName = "CheckboxFacade";
  static componentId = 'CheckboxFacade';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    checked: false,
    focused: false,
    hovered: false,
    size: 'medium',
    indeterminate: false
  };
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  renderIcon() {
    if (this.props.indeterminate) {
      return (0, _jsxRuntime.jsx)(_SVGIcon.SVGIcon, {
        viewBox: "0 0 1920 1920",
        inline: false,
        children: (0, _jsxRuntime.jsx)("rect", {
          x: "140",
          y: "820",
          width: "1640",
          height: "280"
        })
      });
    } else if (this.props.checked) {
      return (0, _jsxRuntime.jsx)(_IconCheckMarkSolid.IconCheckMarkSolid, {
        inline: false
      });
    } else {
      return null;
    }
  }
  render() {
    const {
      children,
      styles
    } = this.props;
    return (0, _jsxRuntime.jsxs)("span", {
      css: styles?.checkboxFacade,
      ref: this.handleRef,
      children: [(0, _jsxRuntime.jsx)("span", {
        css: styles?.facade,
        "aria-hidden": "true",
        children: this.renderIcon()
      }), (0, _jsxRuntime.jsx)("span", {
        css: styles?.label,
        children: children
      })]
    });
  }
}) || _class);
var _default = exports.default = CheckboxFacade;
