// @instructure/ui-avatar/es/Avatar/v2/index.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { useStyleNew } from '@instructure/emotion';
import { useState, useEffect, forwardRef } from 'react';
import { passthroughProps } from '@instructure/ui-react-utils';
import { renderIconWithProps } from '@instructure/ui-icons';
import { avatarSizeToIconSize } from "./props.js";
import generateStyle from "./styles.js";
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
const ICON_COLOR_MAP = {
  accent1: 'accentBlueColor',
  accent2: 'accentGreenColor',
  accent3: 'accentRedColor',
  accent4: 'accentOrangeColor',
  accent5: 'accentGreyColor',
  accent6: 'accentAshColor',
  ai: 'onColor'
};

/**
---
category: components
---
**/
const Avatar = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    size = 'medium',
    color = 'accent1',
    hasInverseColor = false,
    showBorder = 'auto',
    shape = 'circle',
    display = 'inline',
    onImageLoaded = _event => {},
    src,
    name,
    renderIcon,
    alt,
    themeOverride,
    margin
  } = props;
  const [loaded, setLoaded] = useState(false);
  const styles = useStyleNew({
    generateStyle,
    themeOverride,
    params: {
      loaded,
      size,
      color,
      hasInverseColor,
      shape,
      src,
      showBorder,
      display,
      margin
    },
    componentId: 'Avatar',
    displayName: 'Avatar'
  });
  useEffect(() => {
    // in case the image is unset in an update, show icons/initials again
    if (loaded && !src) {
      setLoaded(false);
    }
    // Image already loaded (common in SSR)
    //TODO-rework make this work
    // if (src && !loaded && imgRef.current && imgRef.current.complete) {
    //   setLoaded(true)
    //   onImageLoaded?.()
    // }
  }, [loaded, src]);
  const makeInitialsFromName = () => {
    if (!name || typeof name !== 'string' || name.trim().length === 0) {
      return;
    }
    const currentName = name.trim();
    if (currentName.match(/\s+/)) {
      const names = currentName.split(/\s+/);
      return (names[0][0] + names[names.length - 1][0]).toUpperCase();
    } else {
      return currentName[0].toUpperCase();
    }
  };
  const renderInitials = () => {
    return _jsx("span", {
      "aria-hidden": "true",
      children: makeInitialsFromName()
    });
  };
  const renderImage = () => _jsx("img", {
    style: {
      width: '100%',
      height: '100%',
      objectFit: 'cover',
      objectPosition: 'center',
      ...(loaded ? {} : {
        display: 'none'
      })
    },
    src: src,
    alt: alt ? alt : '',
    onLoad: event => {
      setLoaded(true);
      onImageLoaded(event);
    }
  });
  const renderContent = () => {
    //image in avatar - prioritize image over icon
    if (src) {
      return _jsxs(_Fragment, {
        children: [renderImage(), loaded ? null : renderInitials()]
      });
    }

    //icon in avatar
    if (renderIcon) {
      const iconSize = avatarSizeToIconSize[size];
      const iconColor = hasInverseColor ? 'onColor' : ICON_COLOR_MAP[color];
      return renderIconWithProps(renderIcon, iconSize, iconColor);
    }

    //initials in avatar
    return renderInitials();
  };
  return _jsx("div", {
    ...passthroughProps({
      ...props
    }),
    ref: ref,
    "aria-label": alt ? alt : undefined,
    role: alt ? 'img' : undefined,
    css: styles?.avatar,
    children: renderContent()
  });
});
Avatar.displayName = 'Avatar';
export default Avatar;
export { Avatar };
