// @instructure/ui-spinner/lib/Spinner/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Spinner = void 0;
var _react = require("react");
var _v11_ = require("@instructure/ui-view/v11_6");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _console = require("@instructure/console");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let Spinner = exports.Spinner = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = _dec2(_class = class Spinner extends _react.Component {
  static displayName = "Spinner";
  static componentId = 'Spinner';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    as: 'div',
    size: 'medium',
    variant: 'default'
  };
  ref = null;
  titleId;
  delayTimeout;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  constructor(props) {
    super(props);
    this.titleId = props.deterministicId();
    this.state = {
      shouldRender: !props.delay
    };
  }
  componentDidMount() {
    this.props.makeStyles?.();
    const {
      delay
    } = this.props;
    if (delay) {
      this.delayTimeout = setTimeout(() => {
        this.setState({
          shouldRender: true
        });
      }, delay);
    }
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  componentWillUnmount() {
    clearTimeout(this.delayTimeout);
  }
  radius() {
    switch (this.props.size) {
      case 'x-small':
        return '0.5em';
      case 'small':
        return '1em';
      case 'large':
        return '2.25em';
      default:
        return '1.75em';
    }
  }
  renderSpinner() {
    const passthroughProps = _v11_.View.omitViewProps((0, _omitProps.omitProps)(this.props, Spinner.allowedProps), Spinner);
    const hasTitle = this.props.renderTitle;
    (0, _console.logError)(!!hasTitle, '[Spinner] The renderTitle prop is necessary for screen reader support.');
    return (0, _jsxRuntime.jsx)(_v11_.View, {
      ...passthroughProps,
      as: this.props.as,
      elementRef: this.handleRef,
      css: this.props.styles?.spinner,
      margin: this.props.margin,
      "data-cid": "Spinner",
      children: (0, _jsxRuntime.jsxs)("svg", {
        css: this.props.styles?.circle,
        role: "img",
        "aria-labelledby": this.titleId,
        focusable: "false",
        children: [(0, _jsxRuntime.jsx)("title", {
          id: this.titleId,
          children: (0, _callRenderProp.callRenderProp)(this.props.renderTitle)
        }), (0, _jsxRuntime.jsxs)("g", {
          role: "presentation",
          children: [this.props.variant !== 'inverse' && (0, _jsxRuntime.jsx)("circle", {
            css: this.props.styles?.circleTrack,
            cx: "50%",
            cy: "50%",
            r: this.radius()
          }), (0, _jsxRuntime.jsx)("circle", {
            css: this.props.styles?.circleSpin,
            cx: "50%",
            cy: "50%",
            r: this.radius()
          })]
        })]
      })
    });
  }
  render() {
    return this.state.shouldRender ? this.renderSpinner() : null;
  }
}) || _class) || _class);
var _default = exports.default = Spinner;
