// @instructure/ui-top-nav-bar/lib/TopNavBar/v2/TopNavBarMenuItems/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TopNavBarMenuItems = void 0;
var _react = require("react");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _console = require("@instructure/console");
var _emotion = require("@instructure/emotion");
var _latest = require("@instructure/ui-drilldown/latest");
var _TruncateList = require("@instructure/ui-truncate-list/lib/TruncateList");
var _TopNavBarContext = require("../TopNavBarContext");
var _TopNavBarItem = require("../TopNavBarItem");
var _mapItemsForDrilldown = require("../utils/mapItemsForDrilldown");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _react2 = require("@emotion/react");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: TopNavBar
id: TopNavBar.MenuItems
---
@module TopNavBarMenuItems
**/
let TopNavBarMenuItems = exports.TopNavBarMenuItems = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyleNew)(_styles.default, 'TopNavBarMenuItems'), _dec(_class = _dec2(_class = class TopNavBarMenuItems extends _react.Component {
  static displayName = "TopNavBarMenuItems";
  static componentId = 'TopNavBar.MenuItems';
  static allowedProps = _props.allowedProps;
  static defaultProps = {};
  static contextType = _TopNavBarContext.TopNavBarContext;
  _hiddenMenuItemsMenuTriggerId;
  _hiddenItemsMenuId;
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  constructor(props) {
    super(props);
    this._hiddenMenuItemsMenuTriggerId = props.deterministicId('TopNavBarMenuItems-hiddenMenuItemsMenuTrigger');
    this._hiddenItemsMenuId = props.deterministicId('TopNavBarSmallViewportLayout-drilldown');
    this.state = {
      key: 0,
      visibleItemsCount: undefined
    };
  }
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate(prevProps) {
    this.props.makeStyles?.();
    if (_react.Children.count(prevProps.children) !== this.childrenArray.length) {
      this.setState({
        key: this.state.key + 1
      });
    }
  }
  get childrenArray() {
    return _react.Children.toArray(this.props.children);
  }
  renderOptionContent = (children, itemProps) => {
    const {
      styles
    } = this.props;
    return (0, _jsxRuntime.jsx)("span", {
      css: itemProps?.status === 'active' ? styles?.submenuOptionActive : styles?.submenuOption,
      children: children
    });
  };
  renderHiddenItemsMenu(hiddenItems) {
    const {
      currentPageId,
      renderHiddenItemsMenuTriggerLabel,
      renderHiddenItemsMenuTriggerAriaLabel
    } = this.props;
    if (!hiddenItems.length) {
      return (0, _jsxRuntime.jsx)("span", {});
    }
    const mappedItems = (0, _mapItemsForDrilldown.mapItemsForDrilldown)(hiddenItems, {
      renderOptionContent: this.renderOptionContent,
      currentPageId
    });
    const subPages = (0, _mapItemsForDrilldown.renderMappedItemDrilldownSubpages)(mappedItems);
    const options = (0, _mapItemsForDrilldown.renderMappedItemsAsDrilldownOptions)(mappedItems);
    const hasActiveChild = hiddenItems.find(child => child.props.status === 'active');
    return (0, _jsxRuntime.jsx)(_TopNavBarItem.TopNavBarItem, {
      "aria-label": renderHiddenItemsMenuTriggerAriaLabel ? renderHiddenItemsMenuTriggerAriaLabel(hiddenItems.length) : undefined,
      id: this._hiddenMenuItemsMenuTriggerId,
      status: hasActiveChild ? 'active' : 'default',
      renderSubmenu: (0, _jsxRuntime.jsx)(_latest.Drilldown, {
        shouldSetAriaExpanded: false,
        rootPageId: this._hiddenItemsMenuId,
        children: [(0, _jsxRuntime.jsx)(_latest.Drilldown.Page, {
          id: this._hiddenItemsMenuId,
          children: options
        }, this._hiddenItemsMenuId), ...subPages]
      }),
      children: renderHiddenItemsMenuTriggerLabel(hiddenItems.length)
    });
  }
  renderChildren() {
    const {
      currentPageId
    } = this.props;
    return this.childrenArray.map(child => {
      if (!child) {
        return;
      }
      const {
        id,
        status,
        variant = 'default',
        // workaround because after react 19 defaultProps are not working as expected
        renderSubmenu,
        renderAvatar
      } = child.props;
      const isCurrentPage = currentPageId === id;
      if (renderAvatar) {
        (0, _console.error)(false, `Items in <TopNavBar.MenuItems> are not allowed to have avatars, but item with id: "${id}" has \`renderAvatar\` prop.`);
        return null;
      }
      if (!isCurrentPage) {
        return child;
      } else {
        if (variant !== 'default') {
          (0, _console.warn)(false, `Only \`variant="default"\` items can be set to current/active, but the item with id "${id}" is "${variant}" variant.`);
          return child;
        }
        if (status === 'disabled') {
          (0, _console.warn)(false, `Disabled items can not be set to current/active, but the item with id "${id}" is disabled.`);
          return child;
        }
        return (0, _safeCloneElement.safeCloneElement)(child, {
          status: 'active',
          // if it has submenu, the root item cannot be current,
          // just a link inside the submenu
          'aria-current': renderSubmenu ? undefined : 'page'
        });
      }
    });
  }
  render() {
    const {
      listLabel,
      styles,
      renderHiddenItemsMenuTriggerLabel,
      renderHiddenItemsMenuTriggerAriaLabel,
      ...restProps
    } = this.props;
    if (!this.childrenArray.length) {
      return null;
    }
    if (this.context.layout === 'smallViewport') {
      // in smallViewport mode it is rendered as a Drilldown
      return null;
    }
    return (0, _react2.createElement)(_TruncateList.TruncateList, {
      ...(0, _omitProps.omitProps)(restProps, _props.allowedProps),
      key: this.state.key // rerender if child count changes
      ,
      elementRef: this.handleRef,
      css: styles?.topNavBarMenuItems,
      visibleItemsCount: this.state.visibleItemsCount,
      onUpdate: ({
        visibleItemsCount
      }) => {
        this.setState({
          visibleItemsCount
        });
      },
      renderHiddenItemMenu: hiddenChildren => this.renderHiddenItemsMenu(hiddenChildren),
      ...(styles?.itemSpacing && {
        itemSpacing: styles.itemSpacing
      }),
      "aria-label": listLabel,
      "data-cid": "TopNavBarMenuItems"
    }, this.renderChildren());
  }
}) || _class) || _class);
var _default = exports.default = TopNavBarMenuItems;
