// @instructure/ui-radio-input/lib/RadioInput/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.RadioInput = void 0;
var _react = require("react");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _isActiveElement = require("@instructure/ui-dom-utils/lib/isActiveElement.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let RadioInput = exports.RadioInput = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = _dec2(_class = class RadioInput extends _react.Component {
  static displayName = "RadioInput";
  static componentId = 'RadioInput';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    variant: 'simple',
    size: 'medium',
    disabled: false,
    inline: false,
    context: 'success',
    readOnly: false
  };
  ref = null;
  _defaultId;
  _input = null;
  constructor(props) {
    super(props);
    if (typeof props.checked === 'undefined') {
      this.state = {
        checked: false
      };
    }
    this._defaultId = props.deterministicId();
  }
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  handleInputRef = el => {
    this._input = el;
    if (typeof this.props.inputRef === 'function') {
      this.props.inputRef(el);
    }
  };
  handleClick = e => {
    if (this.props.disabled || this.props.readOnly) {
      e.preventDefault();
      return;
    }
    if (typeof this.props.onClick === 'function') {
      this.props.onClick(e);
    }
  };
  handleChange = e => {
    if (this.props.disabled || this.props.readOnly) {
      e.preventDefault();
      return;
    }
    if (typeof this.props.checked === 'undefined') {
      this.setState({
        checked: !this.state.checked
      });
    }
    if (typeof this.props.onChange === 'function') {
      this.props.onChange(e);
    }
  };
  focus() {
    this._input?.focus();
  }
  get id() {
    return this.props.id || this._defaultId;
  }
  get focused() {
    return (0, _isActiveElement.isActiveElement)(this._input);
  }
  get checked() {
    return typeof this.props.checked === 'undefined' ? this.state.checked : this.props.checked;
  }
  render() {
    const {
      disabled,
      readOnly,
      label,
      value,
      name,
      styles
    } = this.props;
    const props = (0, _omitProps.omitProps)(this.props, RadioInput.allowedProps);
    return (0, _jsxRuntime.jsx)("div", {
      css: styles?.radioInput,
      "data-cid": "RadioInput",
      ref: el => {
        this.ref = el;
      },
      children: (0, _jsxRuntime.jsxs)("div", {
        css: styles?.container,
        children: [(0, _jsxRuntime.jsx)("input", {
          ...props,
          id: this.id,
          ref: this.handleInputRef,
          value: value,
          name: name,
          checked: this.checked,
          type: "radio",
          css: styles?.input,
          disabled: disabled || readOnly,
          onChange: this.handleChange,
          onClick: this.handleClick
        }), (0, _jsxRuntime.jsxs)("label", {
          css: styles?.control,
          htmlFor: this.id,
          children: [(0, _jsxRuntime.jsx)("span", {
            css: styles?.facade,
            "aria-hidden": "true"
          }), (0, _jsxRuntime.jsx)("span", {
            css: styles?.label,
            children: label
          })]
        })]
      })
    });
  }
}) || _class) || _class);
var _default = exports.default = RadioInput;
