// @instructure/ui-themes/src/themes/newThemeTokens/light/index.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import primitives, { type Primitives } from "./primitives"
import semantics, { type Semantics } from "./semantics"
import sharedTokens from "./sharedTokens"

import aiInformation from "./components/aiInformation"
import alert from "./components/alert"
import appNav from "./components/appNav"
import appNavItem from "./components/appNavItem"
import avatar from "./components/avatar"
import badge from "./components/badge"
import baseButton from "./components/baseButton"
import billboard from "./components/billboard"
import breadcrumb from "./components/breadcrumb"
import byline from "./components/byline"
import calendar from "./components/calendar"
import calendarDay from "./components/calendarDay"
import checkbox from "./components/checkbox"
import colorContrast from "./components/colorContrast"
import colorIndicator from "./components/colorIndicator"
import colorPicker from "./components/colorPicker"
import colorPreset from "./components/colorPreset"
import contextView from "./components/contextView"
import dataPermissionLevels from "./components/dataPermissionLevels"
import drawerLayoutContent from "./components/drawerLayoutContent"
import drawerLayoutTray from "./components/drawerLayoutTray"
import drilldown from "./components/drilldown"
import fileDrop from "./components/fileDrop"
import flex from "./components/flex"
import formFieldLayout from "./components/formFieldLayout"
import formFieldMessage from "./components/formFieldMessage"
import grid from "./components/grid"
import gridCol from "./components/gridCol"
import gridRow from "./components/gridRow"
import heading from "./components/heading"
import icon from "./components/icon"
import img from "./components/img"
import inlineListItem from "./components/inlineListItem"
import link from "./components/link"
import list from "./components/list"
import listItem from "./components/listItem"
import mask from "./components/mask"
import menu from "./components/menu"
import menuGroup from "./components/menuGroup"
import menuItem from "./components/menuItem"
import menuSeparator from "./components/menuSeparator"
import metric from "./components/metric"
import modal from "./components/modal"
import modalBody from "./components/modalBody"
import modalFooter from "./components/modalFooter"
import modalHeader from "./components/modalHeader"
import nutritionFacts from "./components/nutritionFacts"
import options from "./components/options"
import optionsItem from "./components/optionsItem"
import optionsSeparator from "./components/optionsSeparator"
import pagination from "./components/pagination"
import paginationPageInput from "./components/paginationPageInput"
import palette from "./components/palette"
import pill from "./components/pill"
import popover from "./components/popover"
import progressBar from "./components/progressBar"
import progressCircle from "./components/progressCircle"
import radioInput from "./components/radioInput"
import rangeInput from "./components/rangeInput"
import ratingIcon from "./components/ratingIcon"
import rgbaInput from "./components/rgbaInput"
import select from "./components/select"
import sideNavBar from "./components/sideNavBar"
import sideNavBarItem from "./components/sideNavBarItem"
import sideNavBarNew from "./components/sideNavBarNew"
import slider from "./components/slider"
import sourceCodeEditor from "./components/sourceCodeEditor"
import spinner from "./components/spinner"
import table from "./components/table"
import tableBody from "./components/tableBody"
import tableCell from "./components/tableCell"
import tableColHeader from "./components/tableColHeader"
import tableHead from "./components/tableHead"
import tableRow from "./components/tableRow"
import tableRowHeader from "./components/tableRowHeader"
import tabs from "./components/tabs"
import tabsPanel from "./components/tabsPanel"
import tabsTab from "./components/tabsTab"
import tag from "./components/tag"
import text from "./components/text"
import textArea from "./components/textArea"
import textInput from "./components/textInput"
import toggle from "./components/toggle"
import toggleDetails from "./components/toggleDetails"
import toggleGroup from "./components/toggleGroup"
import tooltip from "./components/tooltip"
import topNavBarActionItems from "./components/topNavBarActionItems"
import topNavBarBrand from "./components/topNavBarBrand"
import topNavBarItem from "./components/topNavBarItem"
import topNavBarLayout from "./components/topNavBarLayout"
import topNavBarMenuItems from "./components/topNavBarMenuItems"
import tray from "./components/tray"
import treeBrowser from "./components/treeBrowser"
import treeBrowserTreeButton from "./components/treeBrowserTreeButton"
import treeBrowserTreeCollection from "./components/treeBrowserTreeCollection"
import truncateText from "./components/truncateText"
import view from "./components/view"

import type { BaseTheme } from "../commonTypes"

export type Theme = BaseTheme<Primitives, Semantics>
// TODO-theme-types: fix any
const theme: any = {
  primitives,
  semantics,
  sharedTokens,
  components: {
    Metric: metric,
    Avatar: avatar,
    Breadcrumb: breadcrumb,
    Pill: pill,
    FormFieldMessage: formFieldMessage,
    Spinner: spinner,
    Link: link,
    TextInput: textInput,
    TextArea: textArea,
    Icon: icon,
    Checkbox: checkbox,
    BaseButton: baseButton,
    Text: text,
    RadioInput: radioInput,
    Toggle: toggle,
    Popover: popover,
    Tooltip: tooltip,
    Modal: modal,
    ModalHeader: modalHeader,
    ModalBody: modalBody,
    ModalFooter: modalFooter,
    Mask: mask,
    Badge: badge,
    FormFieldLayout: formFieldLayout,
    Heading: heading,
    Tray: tray,
    ContextView: contextView,
    View: view,
    ProgressBar: progressBar,
    Tabs: tabs,
    TabsPanel: tabsPanel,
    TabsTab: tabsTab,
    Drilldown: drilldown,
    Menu: menu,
    MenuGroup: menuGroup,
    MenuItem: menuItem,
    MenuSeparator: menuSeparator,
    Options: options,
    OptionsItem: optionsItem,
    OptionsSeparator: optionsSeparator,
    RangeInput: rangeInput,
    FileDrop: fileDrop,
    Tag: tag,
    Calendar: calendar,
    CalendarDay: calendarDay,
    Grid: grid,
    GridCol: gridCol,
    GridRow: gridRow,
    InlineListItem: inlineListItem,
    ListItem: listItem,
    List: list,
    Table: table,
    TableBody: tableBody,
    TableHead: tableHead,
    TableCell: tableCell,
    TableRowHeader: tableRowHeader,
    TableRow: tableRow,
    TableColHeader: tableColHeader,
    SourceCodeEditor: sourceCodeEditor,
    Flex: flex,
    TreeBrowser: treeBrowser,
    TreeBrowserTreeButton: treeBrowserTreeButton,
    TreeBrowserTreeCollection: treeBrowserTreeCollection,
    ToggleDetails: toggleDetails,
    Alert: alert,
    SideNavBar: sideNavBar,
    SideNavBarItem: sideNavBarItem,
    ToggleGroup: toggleGroup,
    Billboard: billboard,
    AiInformation: aiInformation,
    DataPermissionLevels: dataPermissionLevels,
    NutritionFacts: nutritionFacts,
    RatingIcon: ratingIcon,
    ProgressCircle: progressCircle,
    TopNavBarActionItems: topNavBarActionItems,
    TopNavBarBrand: topNavBarBrand,
    TopNavBarItem: topNavBarItem,
    TopNavBarLayout: topNavBarLayout,
    TopNavBarMenuItems: topNavBarMenuItems,
    Select: select,
    Pagination: pagination,
    PaginationPageInput: paginationPageInput,
    DrawerLayoutContent: drawerLayoutContent,
    DrawerLayoutTray: drawerLayoutTray,
    ColorPicker: colorPicker,
    ColorPreset: colorPreset,
    ColorIndicator: colorIndicator,
    ColorContrast: colorContrast,
    Palette: palette,
    Slider: slider,
    RgbaInput: rgbaInput,
    TruncateText: truncateText,
    AppNav: appNav,
    AppNavItem: appNavItem,
    Byline: byline,
    SideNavBarNew: sideNavBarNew,
    Img: img,
  },
}

export default theme
