// @instructure/ui-checkbox/lib/Checkbox/v1/ToggleFacade/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ToggleFacade = void 0;
var _react = require("react");
var _IconCheckSolid = require("@instructure/ui-icons/lib/generated/IconCheckSolid.js");
var _IconXSolid = require("@instructure/ui-icons/lib/generated/IconXSolid.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: Checkbox
---
**/
let ToggleFacade = exports.ToggleFacade = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class ToggleFacade extends _react.Component {
  static displayName = "ToggleFacade";
  static componentId = 'ToggleFacade';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    checked: false,
    focused: false,
    size: 'medium',
    disabled: false,
    readOnly: false,
    labelPlacement: 'end'
  };
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  renderIcon() {
    const {
      styles,
      checked
    } = this.props;
    if (checked) {
      return (0, _jsxRuntime.jsx)(_IconCheckSolid.IconCheckSolid, {
        css: styles?.iconSVG
      });
    } else {
      return (0, _jsxRuntime.jsx)(_IconXSolid.IconXSolid, {
        css: styles?.iconSVG
      });
    }
  }
  renderLabel() {
    const {
      children,
      styles
    } = this.props;
    return (0, _jsxRuntime.jsx)("span", {
      css: styles?.label,
      children: children
    });
  }
  render() {
    const {
      labelPlacement,
      styles
    } = this.props;
    return (0, _jsxRuntime.jsxs)("span", {
      css: styles?.toggleFacade,
      ref: this.handleRef,
      children: [(labelPlacement === 'top' || labelPlacement === 'start') && this.renderLabel(), (0, _jsxRuntime.jsx)("span", {
        css: styles?.facade,
        "aria-hidden": "true",
        children: (0, _jsxRuntime.jsx)("span", {
          css: styles?.icon,
          children: (0, _jsxRuntime.jsx)("span", {
            css: styles?.iconToggle,
            children: this.renderIcon()
          })
        })
      }), labelPlacement === 'end' && this.renderLabel()]
    });
  }
}) || _class);
var _default = exports.default = ToggleFacade;
