// @instructure/ui-view/lib/View/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.View = void 0;
var _react = require("react");
var _getCSSStyleDeclaration = require("@instructure/ui-dom-utils/lib/getCSSStyleDeclaration.js");
var _textDirectionContextConsumer = require("@instructure/ui-i18n/lib/textDirectionContextConsumer.js");
var _console = require("@instructure/console");
var _getElementType = require("@instructure/ui-react-utils/lib/getElementType.js");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _pickProps = require("@instructure/ui-react-utils/lib/pickProps.js");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
@module View
**/
let View = exports.View = (_dec = (0, _textDirectionContextConsumer.textDirectionContextConsumer)(), _dec2 = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = _dec2(_class = class View extends _react.Component {
  static displayName = "View";
  static componentId = 'View';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    display: 'auto',
    overflowX: 'visible',
    overflowY: 'visible',
    withVisualDebug: false,
    borderColor: 'primary',
    position: 'static',
    focusPosition: 'offset',
    focusColor: 'info',
    shouldAnimateFocus: true,
    overscrollBehavior: 'auto'
  };

  // TODO: Remove this code once all components are using passthroughProps in place
  // of omitProps and have removed this function

  // omitViewProps needs to be called on the composed View component so that the
  // View.allowedProps in the method matches the View.allowedProps that will be called in
  // the consumers. Otherwise the discrepancy could cause unexpected props being
  // allowed through.

  // Removes View's own props from the given object. Automatically excludes the
  // following props: 'theme', 'children', 'className', 'style', 'styles',
  // 'makeStyles', 'themeOverride'
  // @param props the object to process
  // @param Component The component whose props are processed.
  // Only needed for debug logging in non-production builds.
  static omitViewProps = (props, Component) => {
    // We don't want the theming and styling props to pass
    // (these are added and handled by the `@withStyleNew` decorator)
    const propsToOmit = [...View.allowedProps, 'styles', 'makeStyles', 'themeOverride'];
    if (typeof process !== 'undefined' && (process?.env?.GITHUB_PULL_REQUEST_PREVIEW === 'true' || process?.env?.NODE_ENV !== 'production')) {
      Object.keys((0, _pickProps.pickProps)(props, propsToOmit)).forEach(prop => {
        (0, _console.logError)(false, `[${Component.name}] prop '${prop}' is not allowed.`);
      });
    }
    return (0, _omitProps.omitProps)(props, propsToOmit);
  };
  spanMarginVerified;
  get _element() {
    return this.ref;
  }
  constructor(props) {
    super(props);
    this.spanMarginVerified = false;
  }
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();

    // Not calling getCSSStyleDeclaration can save hundreds of ms in tests and production
    if (typeof process !== 'undefined' && process?.env?.NODE_ENV === 'development' && !this.spanMarginVerified) {
      // We have to verify margins in the first 'componentDidUpdate',
      // because that is when all styles are calculated,
      // but we only want to check once, using a flag
      this.spanMarginVerified = true;
      (0, _console.logError)(!function verifySpanMargin(element, margin) {
        if (!element) {
          return;
        }
        const display = (0, _getCSSStyleDeclaration.getCSSStyleDeclaration)(element)?.display;
        if (display !== 'inline') {
          return;
        }
        const marginValues = margin ? margin.split(' ') : null;
        let verticalMargin = false;

        // either top or bottom margin are set
        if (margin) {
          if (marginValues && marginValues[0] && marginValues[0] !== 'none' && marginValues[0] !== '0') {
            verticalMargin = true;
          }
          if (marginValues && marginValues[2] && marginValues[2] !== 'none' && marginValues[2] !== '0') {
            verticalMargin = true;
          }
        }
        return verticalMargin;
      }(this.ref, this.props.margin), `[View] display style is set to 'inline' and will allow for horizontal margins only.`);
    }
  }
  ref = null;
  handleElementRef = el => {
    if (typeof this.props.elementRef === 'function') {
      this.props.elementRef(el);
    }
    this.ref = el;
  };
  render() {
    const {
      children,
      textAlign,
      background,
      display,
      withVisualDebug,
      width,
      height,
      minWidth,
      minHeight,
      maxWidth,
      maxHeight,
      overflowX,
      overflowY,
      stacking,
      shadow,
      position,
      focusPosition,
      focusColor,
      shouldAnimateFocus,
      borderColor,
      className,
      overscrollBehavior,
      styles,
      makeStyles,
      ...props
    } = this.props;
    const ElementType = (0, _getElementType.getElementType)(View, this.props);
    return (0, _jsxRuntime.jsx)(ElementType, {
      ...(0, _passthroughProps.passthroughProps)(props),
      className: className,
      css: [styles?.view, styles?.inlineStyles],
      ref: this.handleElementRef,
      children: children
    });
  }
}) || _class) || _class);
var _default = exports.default = View;
