// @instructure/ui-checkbox/es/Checkbox/v2/ToggleFacade/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { CheckInstUIIcon, XInstUIIcon, renderIconWithProps } from '@instructure/ui-icons';
import { withStyleNew } from '@instructure/emotion';
import generateStyle from "./styles.js";
import { allowedProps } from "./props.js";
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
/**
---
parent: Checkbox
---
**/
let ToggleFacade = (_dec = withStyleNew(generateStyle, 'Toggle'), _dec(_class = class ToggleFacade extends Component {
  static displayName = "ToggleFacade";
  static componentId = 'ToggleFacade';
  static allowedProps = allowedProps;
  static defaultProps = {
    checked: false,
    focused: false,
    hovered: false,
    size: 'medium',
    disabled: false,
    readOnly: false,
    labelPlacement: 'end'
  };
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  renderIcon() {
    const {
      disabled,
      readOnly,
      checked
    } = this.props;
    const getIconColor = () => {
      if (disabled) {
        return 'disabledBaseColor';
      }
      if (readOnly) {
        return 'mutedColor';
      }
      if (checked) {
        return 'successColor';
      }
      return 'baseColor';
    };
    const iconColor = getIconColor();
    if (checked) {
      return renderIconWithProps(CheckInstUIIcon, 'xs', iconColor);
    } else {
      return renderIconWithProps(XInstUIIcon, 'xs', iconColor);
    }
  }
  renderLabel() {
    const {
      children,
      styles
    } = this.props;
    return _jsx("span", {
      css: styles?.label,
      children: children
    });
  }
  render() {
    const {
      labelPlacement,
      styles
    } = this.props;
    return _jsxs("span", {
      css: styles?.toggleFacade,
      ref: this.handleRef,
      children: [(labelPlacement === 'top' || labelPlacement === 'start') && this.renderLabel(), _jsx("span", {
        css: styles?.facade,
        "aria-hidden": "true",
        children: _jsx("span", {
          css: styles?.icon,
          children: _jsx("span", {
            css: styles?.iconToggle,
            children: this.renderIcon()
          })
        })
      }), labelPlacement === 'end' && this.renderLabel()]
    });
  }
}) || _class);
export default ToggleFacade;
export { ToggleFacade };
