// @instructure/ui-color-picker/es/ColorContrast/v2/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { omitProps } from '@instructure/ui-react-utils';
import { error } from '@instructure/console';
import { contrastWithAlpha, validateContrast } from '@instructure/ui-color-utils';
import { withStyleNew } from '@instructure/emotion';
import { Text } from '@instructure/ui-text/latest';
import { Heading } from '@instructure/ui-heading/latest';
import { Pill } from '@instructure/ui-pill/latest';
import ColorIndicator from "../../ColorIndicator/v2/index.js";
import { allowedProps } from "./props.js";
import generateStyle from "./styles.js";
/**
---
category: components
---
**/
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "@emotion/react/jsx-runtime";
let ColorContrast = (_dec = withStyleNew(generateStyle), _dec(_class = class ColorContrast extends Component {
  static displayName = "ColorContrast";
  static allowedProps = allowedProps;
  static componentId = 'ColorContrast';
  static defaultProps = {
    withoutColorPreview: false,
    validationLevel: 'AA'
  };
  constructor(props) {
    super(props);
    this.state = {
      contrast: 1,
      isValidNormalText: false,
      isValidLargeText: false,
      isValidGraphicsText: false
    };
  }
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.();
    this.calcState();
  }
  componentDidUpdate(prevProps) {
    this.props.makeStyles?.();
    if (prevProps?.firstColor !== this.props?.firstColor || prevProps?.secondColor !== this.props?.secondColor || prevProps?.validationLevel !== this.props?.validationLevel) {
      const newState = this.calcState();
      this.props?.onContrastChange?.({
        contrast: newState.contrast,
        isValidNormalText: newState.isValidNormalText,
        isValidLargeText: newState.isValidLargeText,
        isValidGraphicsText: newState.isValidGraphicsText,
        firstColor: this.props.firstColor,
        secondColor: this.props.secondColor
      });
    }
  }
  renderStatus = (pass, description) => {
    const {
      successLabel,
      failureLabel,
      styles
    } = this.props;
    return _jsxs("div", {
      css: styles?.statusWrapper,
      children: [_jsx("div", {
        css: pass ? styles?.successDescription : styles?.failureDescription,
        children: description
      }), _jsx("div", {
        css: styles?.statusIndicatorWrapper,
        children: _jsx(Pill, {
          color: pass ? 'success' : 'error',
          children: pass ? successLabel : failureLabel
        })
      })]
    });
  };
  renderColorIndicator = (color, label) => _jsxs(_Fragment, {
    children: [_jsx("div", {
      children: _jsx("div", {
        css: this.props.styles?.colorIndicator,
        children: _jsx(ColorIndicator, {
          color: color
        })
      })
    }), _jsxs("div", {
      children: [_jsx("div", {
        css: this.props.styles?.colorIndicatorLabel,
        children: label
      }), _jsx("div", {
        css: this.props.styles?.pickedColorHex,
        children: color
      })]
    })]
  });
  renderPreview() {
    const {
      styles,
      withoutColorPreview,
      firstColor,
      secondColor,
      firstColorLabel,
      secondColorLabel
    } = this.props;
    if (withoutColorPreview) {
      return null;
    }
    if (!firstColorLabel || !secondColorLabel) {
      error(false, 'When `withoutColorPreview` is not set to true, the properties `firstColorLabel` and `secondColorLabel` are required!');
    }
    return !withoutColorPreview && _jsxs("div", {
      css: styles?.colorPreview,
      children: [_jsx("div", {
        css: styles?.firstColorPreview,
        children: this.renderColorIndicator(firstColor, firstColorLabel || '')
      }), _jsx("div", {
        css: styles?.secondColorPreview,
        children: this.renderColorIndicator(secondColor, secondColorLabel || '')
      })]
    });
  }
  calcState() {
    const contrast = contrastWithAlpha(this.props.firstColor, this.props.secondColor);
    const newState = {
      contrast,
      ...validateContrast(contrast, this.props.validationLevel)
    };
    this.setState(newState);
    return newState;
  }
  render() {
    const {
      styles,
      label,
      labelLevel,
      normalTextLabel,
      largeTextLabel,
      graphicsTextLabel
    } = this.props;
    const {
      contrast,
      isValidNormalText,
      isValidLargeText,
      isValidGraphicsText
    } = this.state;
    return _jsxs("div", {
      ...omitProps(this.props, ColorContrast.allowedProps),
      ref: this.handleRef,
      css: styles?.colorContrast,
      "data-cid": "ColorContrast",
      children: [_jsx("div", {
        css: styles?.label,
        children: labelLevel ? _jsx(Heading, {
          as: labelLevel,
          level: "reset",
          children: label
        }) : _jsx(Text, {
          weight: "bold",
          as: "div",
          children: label
        })
      }), _jsxs(Text, {
        size: "x-large",
        children: [contrast, ":1"]
      }), this.renderPreview(), this.renderStatus(isValidNormalText, normalTextLabel), this.renderStatus(isValidLargeText, largeTextLabel), this.renderStatus(isValidGraphicsText, graphicsTextLabel)]
    });
  }
}) || _class);
export { ColorContrast };
export default ColorContrast;
