// @instructure/ui-side-nav-bar/lib/SideNavBar/v2/SideNavBarItem/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.SideNavBarItem = void 0;
var _react = require("react");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _getElementType = require("@instructure/ui-react-utils/lib/getElementType.js");
var _latest = require("@instructure/ui-tooltip/latest");
var _hasVisibleChildren = require("@instructure/ui-a11y-utils/lib/hasVisibleChildren.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: SideNavBar
id: SideNavBar.Item
---
**/
let SideNavBarItem = exports.SideNavBarItem = (_dec = (0, _emotion.withStyleNew)(_styles.default, 'SideNavBarItem'), _dec(_class = class SideNavBarItem extends _react.Component {
  static displayName = "SideNavBarItem";
  static componentId = 'SideNavBar.Item';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    as: 'a',
    selected: false,
    minimized: false
  };
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.({
      minimized: this.props.minimized
    });
  }
  componentDidUpdate() {
    this.props.makeStyles?.({
      minimized: this.props.minimized
    });
  }
  renderLink(addRef) {
    const ElementType = (0, _getElementType.getElementType)(SideNavBarItem, this.props);
    const {
      href,
      onClick,
      icon,
      label
    } = this.props;
    const props = (0, _omitProps.omitProps)(this.props, SideNavBarItem.allowedProps);
    return (0, _jsxRuntime.jsxs)(ElementType, {
      ...props,
      href: href,
      onClick: onClick,
      ref: addRef ? this.handleRef : undefined,
      css: this.props.styles?.navigationItem,
      "aria-label": this.props.minimized ? label : undefined,
      "data-cid": "SideNavBarItem",
      children: [(0, _jsxRuntime.jsx)("div", {
        css: this.props.styles?.icon,
        children: icon
      }), !this.props.minimized ? (0, _jsxRuntime.jsx)("div", {
        css: this.props.styles?.label,
        children: label
      }) : null]
    });
  }
  render() {
    const {
      minimized,
      label
    } = this.props;
    const hasTooltip = minimized && (0, _hasVisibleChildren.hasVisibleChildren)(label);
    const link = this.renderLink(!hasTooltip);
    return hasTooltip ? (0, _jsxRuntime.jsx)(_latest.Tooltip, {
      renderTip: label,
      placement: "end",
      elementRef: this.handleRef,
      children: link
    }) : link;
  }
}) || _class);
var _default = exports.default = SideNavBarItem;
