// @instructure/ui-form-field/es/FormFieldMessage/v2/index.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { forwardRef } from 'react';
import { ScreenReaderContent } from '@instructure/ui-a11y-content';
import { AlertCircleInstUIIcon, CheckCircle2InstUIIcon } from '@instructure/ui-icons';
import { useStyleNew } from '@instructure/emotion';
import generateStyle from "./styles.js";
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
/**
---
parent: FormField
---

This is a helper component that is used by most of the custom form
components. In most cases it shouldn't be used directly.

```js
---
type: example
---
<FormFieldMessage variant="error">Invalid value</FormFieldMessage>
```
**/
const FormFieldMessage = /*#__PURE__*/forwardRef(({
  variant = 'hint',
  children,
  themeOverride
}, ref) => {
  const styles = useStyleNew({
    generateStyle,
    themeOverride,
    params: {
      variant
    },
    componentId: 'FormFieldMessage',
    displayName: 'FormFieldMessage'
  });
  const handleRef = el => {
    if (typeof ref === 'function') {
      ref(el);
    } else if (ref) {
      const refObject = ref;
      refObject.current = el;
    }
  };
  const isErrorVariant = variant === 'error' || variant === 'newError';
  const shouldShowIcon = (isErrorVariant || variant === 'success') && children;
  return variant !== 'screenreader-only' ? _jsxs("span", {
    css: {
      display: 'flex'
    },
    children: [shouldShowIcon && _jsx("span", {
      css: styles?.icon,
      children: isErrorVariant ? _jsx(AlertCircleInstUIIcon, {
        size: "sm",
        color: "errorColor"
      }) : _jsx(CheckCircle2InstUIIcon, {
        size: "sm",
        color: "successColor"
      })
    }), _jsx("span", {
      css: styles?.formFieldMessage,
      ref: handleRef,
      children: children
    })]
  }) : _jsx(ScreenReaderContent, {
    elementRef: handleRef,
    children: children
  });
});
FormFieldMessage.displayName = 'FormFieldMessage';
export default FormFieldMessage;
export { FormFieldMessage };
