// @instructure/ui-modal/lib/Modal/v1/ModalHeader/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ModalHeader = void 0;
var _react = require("react");
var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _emotion = require("@instructure/emotion");
var _v11_ = require("@instructure/ui-buttons/v11_6");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _ModalContext = _interopRequireDefault(require("../ModalContext"));
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: Modal
id: Modal.Header
---
**/
let ModalHeader = exports.ModalHeader = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class ModalHeader extends _react.Component {
  static displayName = "ModalHeader";
  static componentId = 'Modal.Header';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    variant: 'default',
    spacing: 'default'
  };
  ref = null;
  static contextType = _ModalContext.default;

  /**
   * Gets all text in a DOM subtree, text under <button> nodes is excluded
   */
  getTextExcludingButtons = root => {
    const walker = document.createTreeWalker(root, NodeFilter.SHOW_TEXT, {
      acceptNode(node) {
        return node.parentElement?.closest('button') ? NodeFilter.FILTER_REJECT : NodeFilter.FILTER_ACCEPT;
      }
    });
    let text = '';
    let current;
    while (current = walker.nextNode()) {
      text += current.nodeValue;
    }
    return text;
  };
  handleRef = el => {
    this.ref = el;
    if (el) {
      const txt = this.getTextExcludingButtons(el);
      this.context.setBodyScrollAriaLabel?.(txt);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.(this.makeStyleProps());
  }
  componentDidUpdate() {
    this.props.makeStyles?.(this.makeStyleProps());
  }
  makeStyleProps = () => {
    return {
      withCloseButton: this.usesCloseButton
    };
  };
  get usesCloseButton() {
    let hasCloseButton = false;
    _react.Children.forEach(this.props.children, child => {
      if (child && (0, _matchComponentTypes.matchComponentTypes)(child, [_v11_.CloseButton])) {
        hasCloseButton = true;
      }
    });
    return hasCloseButton;
  }
  render() {
    const {
      children,
      ...rest
    } = this.props;
    return (0, _jsxRuntime.jsx)("div", {
      css: this.props.styles?.modalHeader,
      ...(0, _passthroughProps.passthroughProps)(rest),
      ref: this.handleRef,
      "data-cid": "ModalHeader",
      children: children
    });
  }
}) || _class);
var _default = exports.default = ModalHeader;
