// @instructure/ui-side-nav-bar/es/SideNavBar/v1/SideNavBarItem/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { omitProps, getElementType } from '@instructure/ui-react-utils';
import { Tooltip } from '@instructure/ui-tooltip/v11_6';
import { hasVisibleChildren } from '@instructure/ui-a11y-utils';
import { withStyle } from '@instructure/emotion';
import generateStyle from "./styles.js";
import generateComponentTheme from "./theme.js";
import { allowedProps } from "./props.js";
/**
---
parent: SideNavBar
id: SideNavBar.Item
---
**/
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
let SideNavBarItem = (_dec = withStyle(generateStyle, generateComponentTheme), _dec(_class = class SideNavBarItem extends Component {
  static displayName = "SideNavBarItem";
  static componentId = 'SideNavBar.Item';
  static allowedProps = allowedProps;
  static defaultProps = {
    as: 'a',
    selected: false,
    minimized: false
  };
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.({
      minimized: this.props.minimized
    });
  }
  componentDidUpdate() {
    this.props.makeStyles?.({
      minimized: this.props.minimized
    });
  }
  renderLink(addRef) {
    const ElementType = getElementType(SideNavBarItem, this.props);
    const {
      href,
      onClick,
      icon,
      label
    } = this.props;
    const props = omitProps(this.props, SideNavBarItem.allowedProps);
    return _jsxs(ElementType, {
      ...props,
      href: href,
      onClick: onClick,
      ref: addRef ? this.handleRef : undefined,
      css: this.props.styles?.navigationItem,
      "aria-label": this.props.minimized ? label : undefined,
      "data-cid": "SideNavBarItem",
      children: [_jsx("div", {
        css: this.props.styles?.icon,
        children: icon
      }), !this.props.minimized ? _jsx("div", {
        css: this.props.styles?.label,
        children: label
      }) : null]
    });
  }
  render() {
    const {
      minimized,
      label
    } = this.props;
    const hasTooltip = minimized && hasVisibleChildren(label);
    const link = this.renderLink(!hasTooltip);
    return hasTooltip ? _jsx(Tooltip, {
      renderTip: label,
      placement: "end",
      elementRef: this.handleRef,
      children: link
    }) : link;
  }
}) || _class);
export default SideNavBarItem;
export { SideNavBarItem };
