// @instructure/ui-top-nav-bar/lib/TopNavBar/v2/TopNavBarBrand/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TopNavBarBrand = void 0;
var _react = require("react");
var _getElementType = require("@instructure/ui-react-utils/lib/getElementType.js");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _emotion = require("@instructure/emotion");
var _ScreenReaderContent = require("@instructure/ui-a11y-content/lib/ScreenReaderContent");
var _latest = require("@instructure/ui-view/latest");
var _TopNavBarContext = require("../TopNavBarContext");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: TopNavBar
id: TopNavBar.Brand
---
@module TopNavBarBrand
**/
let TopNavBarBrand = exports.TopNavBarBrand = (_dec = (0, _emotion.withStyleNew)(_styles.default, 'TopNavBarBrand'), _dec(_class = class TopNavBarBrand extends _react.Component {
  static displayName = "TopNavBarBrand";
  static componentId = 'TopNavBar.Brand';
  // TODO: add to the docs: making it static on parent and jsdocs parent/module settings, dont export child on its own

  static allowedProps = _props.allowedProps;
  static defaultProps = {};
  static contextType = _TopNavBarContext.TopNavBarContext;
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.(this.makeStylesVariables);
  }
  componentDidUpdate() {
    this.props.makeStyles?.(this.makeStylesVariables);
  }
  get makeStylesVariables() {
    return {
      layout: this.context.layout
    };
  }
  render() {
    const {
      screenReaderLabel,
      renderIcon,
      href,
      onClick,
      styles
    } = this.props;
    const ElementType = (0, _getElementType.getElementType)(TopNavBarBrand, this.props);
    return (0, _jsxRuntime.jsx)("div", {
      ref: this.handleRef,
      css: styles?.topNavBarBrand,
      "data-cid": "TopNavBarBrand",
      children: renderIcon && (0, _jsxRuntime.jsx)(_emotion.InstUISettingsProvider, {
        theme: currentTheme => ({
          newTheme: {
            sharedTokens: {
              focusOutline: {
                ...currentTheme?.newTheme?.sharedTokens?.focusOutline,
                inset: styles?.focusOutlineInset
              }
            }
          }
        }),
        children: (0, _jsxRuntime.jsxs)(_latest.View, {
          ...(0, _omitProps.omitProps)(this.props, _props.allowedProps),
          css: styles?.container,
          as: ElementType,
          href: href,
          onClick: onClick,
          position: "relative",
          focusColor: this.context.inverseColor ? 'info' : 'inverse',
          focusPosition: "inset",
          borderRadius: "medium",
          children: [(0, _jsxRuntime.jsx)(_ScreenReaderContent.ScreenReaderContent, {
            children: screenReaderLabel
          }), renderIcon && this.context.layout !== 'smallViewport' && (0, _jsxRuntime.jsx)("div", {
            css: styles?.iconContainer,
            role: "presentation",
            "aria-hidden": "true",
            children: renderIcon
          })]
        })
      })
    });
  }
}) || _class);
var _default = exports.default = TopNavBarBrand;
