// @instructure/ui-overlays/lib/Mask/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Mask = void 0;
var _react = require("react");
var _noScroll = _interopRequireDefault(require("no-scroll"));
var _emotion = require("@instructure/emotion");
var _ensureSingleChild = require("@instructure/ui-react-utils/lib/ensureSingleChild.js");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _MaskCounter = _interopRequireDefault(require("./MaskCounter"));
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components/utilities
---
**/
let Mask = exports.Mask = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class Mask extends _react.Component {
  static displayName = "Mask";
  static componentId = 'Mask';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    placement: 'center',
    fullscreen: false
  };
  componentDidMount() {
    this.props.makeStyles?.();
    if (this.props.fullscreen) {
      _noScroll.default.on();
      _MaskCounter.default.incrementCounter();
    }
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  componentWillUnmount() {
    if (this.props.fullscreen) {
      _MaskCounter.default.decrementCounter();
      if (_MaskCounter.default.getCounter() <= 0) {
        _noScroll.default.off();
      }
    }
  }
  ref = null;
  handleElementRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };

  // It can be a ref for any type of child
  _content;
  contentRef = el => {
    this._content = el;
  };
  render() {
    const content = (0, _ensureSingleChild.ensureSingleChild)(this.props.children, {
      ref: this.contentRef
    });
    const props = {
      ...(0, _omitProps.omitProps)(this.props, Mask.allowedProps),
      css: this.props.styles?.mask,
      ref: this.handleElementRef
    };
    if (typeof this.props.onClick === 'function') {
      props.onClick = this.props.onClick;
      props.tabIndex = -1;
    }
    return (0, _jsxRuntime.jsx)("span", {
      ...props,
      children: content
    });
  }
}) || _class);
var _default = exports.default = Mask;
