// @instructure/ui-modal/lib/Modal/v1/ModalBody/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ModalBody = void 0;
var _react = require("react");
var _v11_ = require("@instructure/ui-view/v11_6");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _getCSSStyleDeclaration = require("@instructure/ui-dom-utils/lib/getCSSStyleDeclaration.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _ModalContext = _interopRequireDefault(require("../ModalContext"));
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: Modal
id: Modal.Body
---
**/
let ModalBody = exports.ModalBody = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class ModalBody extends _react.Component {
  static displayName = "ModalBody";
  static componentId = 'Modal.Body';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    padding: 'medium',
    as: 'div',
    variant: 'default'
  };
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  constructor(props) {
    super(props);
  }
  componentDidMount() {
    this.props.makeStyles?.();

    // We detect if -moz- prefixed style is present to identify whether we are in Firefox browser
    const style = this.ref && (0, _getCSSStyleDeclaration.getCSSStyleDeclaration)(this.ref);
    const isFirefox = !!(style && Array.prototype.slice.call(style).join('').match(/(?:-moz-)/));
    if (isFirefox) {
      this.setState({
        isFirefox
      });
    }
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  getFinalRef(el) {
    if (!el) {
      return undefined;
    }
    if (el instanceof Element) {
      return el;
    }
    if (el.ref) {
      return this.getFinalRef(el.ref);
    }
    return undefined;
  }
  render() {
    const {
      as,
      elementRef,
      overflow,
      variant,
      padding,
      children,
      ...rest
    } = this.props;
    const passthroughProps = _v11_.View.omitViewProps((0, _omitProps.omitProps)(rest, ModalBody.allowedProps), ModalBody);
    const isFit = overflow === 'fit';
    // this recursive function is needed because `ref` can be a React component.
    // TODO rethink, the 'as' prop, likely its not a good idea to allow React
    // components. See INSTUI-4674
    const finalRef = this.getFinalRef(this.ref);
    const hasScrollbar = finalRef && Math.abs((finalRef.scrollHeight ?? 0) - (finalRef.getBoundingClientRect()?.height ?? 0)) > 1;
    return (0, _jsxRuntime.jsx)(_ModalContext.default.Consumer, {
      children: value => (0, _jsxRuntime.jsx)(_v11_.View, {
        ...passthroughProps,
        display: "block",
        "data-cid": "ModalBody",
        width: isFit ? '100%' : undefined,
        height: isFit ? '100%' : undefined,
        elementRef: this.handleRef,
        as: as,
        css: this.props.styles?.modalBody,
        padding: padding
        // check if there is a scrollbar, if so, the element has to be tabbable
        ,
        ...(hasScrollbar ? {
          tabIndex: 0,
          'aria-label': value.bodyScrollAriaLabel
        } : {}),
        children: children
      })
    });
  }
}) || _class);
var _default = exports.default = ModalBody;
