// @instructure/ui-avatar/lib/Avatar/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Avatar = void 0;
var _emotion = require("@instructure/emotion");
var _react = require("react");
var _v11_ = require("@instructure/ui-view/v11_6");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
category: components
---
**/const Avatar = exports.Avatar = /*#__PURE__*/(0, _react.forwardRef)(({
  size = 'medium',
  color = 'default',
  hasInverseColor = false,
  showBorder = 'auto',
  shape = 'circle',
  display = 'inline-block',
  onImageLoaded,
  src,
  name,
  renderIcon,
  alt,
  as,
  margin,
  themeOverride,
  elementRef,
  ...rest
}, ref) => {
  const imgRef = (0, _react.useRef)(null);
  const [loaded, setLoaded] = (0, _react.useState)(false);
  const styles = (0, _emotion.useStyle)({
    generateStyle: _styles.default,
    generateComponentTheme: _theme.default,
    params: {
      loaded,
      size,
      color,
      hasInverseColor,
      shape,
      src,
      showBorder,
      themeOverride
    },
    componentId: 'Avatar',
    displayName: 'Avatar'
  });
  (0, _react.useEffect)(() => {
    // in case the image is unset in an update, show icons/initials again
    if (loaded && !src) {
      setLoaded(false);
    }
    // Image already loaded (common in SSR)
    if (src && !loaded && imgRef.current && imgRef.current.complete) {
      setLoaded(true);
      onImageLoaded?.();
    }
  }, [loaded, src]);
  const makeInitialsFromName = () => {
    if (!name || typeof name !== 'string') {
      return;
    }
    const currentName = name.trim();
    if (currentName.length === 0) {
      return;
    }
    if (currentName.match(/\s+/)) {
      const names = currentName.split(/\s+/);
      return (names[0][0] + names[names.length - 1][0]).toUpperCase();
    } else {
      return currentName[0].toUpperCase();
    }
  };
  const handleImageLoaded = event => {
    setLoaded(true);
    onImageLoaded?.(event);
  };
  const renderInitials = () => {
    return (0, _jsxRuntime.jsx)("span", {
      css: styles?.initials,
      "aria-hidden": "true",
      children: makeInitialsFromName()
    });
  };
  const renderContent = () => {
    if (!renderIcon) {
      return renderInitials();
    }
    return (0, _jsxRuntime.jsx)("span", {
      css: styles?.iconSVG,
      children: (0, _callRenderProp.callRenderProp)(renderIcon)
    });
  };
  return (0, _jsxRuntime.jsxs)(_v11_.View, {
    ...(0, _passthroughProps.passthroughProps)({
      size,
      color,
      hasInverseColor,
      showBorder,
      shape,
      display,
      src,
      name,
      renderIcon,
      alt,
      as,
      margin,
      ...rest
    }),
    ref: ref,
    "aria-label": alt ? alt : undefined,
    role: alt ? 'img' : undefined,
    as: as,
    elementRef: elementRef,
    margin: margin,
    css: styles?.avatar,
    display: display,
    children: [(0, _jsxRuntime.jsx)("img", {
      // This is visually hidden and is here for loading purposes only
      src: src,
      ref: imgRef,
      css: styles?.loadImage,
      alt: alt,
      onLoad: handleImageLoaded,
      "aria-hidden": "true"
    }), !loaded && renderContent()]
  });
});
Avatar.displayName = 'Avatar';

// TODO - why is this needed?
Avatar.displayName = 'Avatar';
var _default = exports.default = Avatar;
