// @instructure/ui-buttons/lib/CondensedButton/v2/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.CondensedButton = void 0;
var _react = require("react");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _emotion = require("@instructure/emotion");
var _v = require("../../BaseButton/v2");
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
// needed for listing the available theme variables on docs page
let CondensedButton = exports.CondensedButton = (_dec = (0, _emotion.withStyleNew)(null, 'BaseButton'), _dec(_class = class CondensedButton extends _react.Component {
  static displayName = "CondensedButton";
  static componentId = 'CondensedButton';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    type: 'button',
    size: 'medium',
    as: 'button',
    // Leave interaction default undefined so that `disabled` and `readOnly` can also be supplied
    interaction: undefined,
    color: 'primary',
    margin: '0',
    cursor: 'pointer',
    display: 'inline-block'
  };
  _baseButton = null;
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  get focused() {
    return this._baseButton && this._baseButton.focused;
  }
  focus() {
    this._baseButton && this._baseButton.focus();
  }
  render() {
    const {
      children,
      type,
      size,
      elementRef,
      as,
      interaction,
      color,
      margin,
      cursor,
      href,
      renderIcon,
      display,
      ...props
    } = this.props;
    const themeOverride = this.props.themeOverride;
    return (0, _jsxRuntime.jsx)(_v.BaseButton, {
      ...(0, _passthroughProps.passthroughProps)(props),
      isCondensed: true,
      display: display,
      withBackground: false,
      withBorder: false,
      type: type,
      size: size,
      elementRef: this.handleRef,
      as: as,
      interaction: interaction,
      color: color,
      margin: margin,
      cursor: cursor,
      href: href,
      renderIcon: renderIcon,
      themeOverride: themeOverride,
      ref: component => {
        this._baseButton = component;
      },
      "data-cid": "CondensedButton",
      children: children
    });
  }
}) || _class);
var _default = exports.default = CondensedButton;
