// @instructure/ui-radio-input/lib/RadioInput/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.RadioInput = void 0;
var _react = require("react");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _useDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/useDeterministicId.js");
var _isActiveElement = require("@instructure/ui-dom-utils/lib/isActiveElement.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
category: components
---
**/
const RadioInput = exports.RadioInput = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    variant = 'simple',
    size = 'medium',
    disabled = false,
    inline = false,
    context = 'success',
    readOnly = false,
    id: idProp,
    label,
    value,
    name,
    checked: checkedProp,
    onClick,
    onChange,
    inputRef,
    themeOverride,
    ...rest
  } = props;
  const [hovered, setHovered] = (0, _react.useState)(false);
  // State for uncontrolled mode
  const [internalChecked, setInternalChecked] = (0, _react.useState)(false);

  // Refs
  const containerRef = (0, _react.useRef)(null);
  const inputElementRef = (0, _react.useRef)(null);

  // Deterministic ID generation
  const [deterministicId, setDeterministicId] = (0, _react.useState)();
  const getId = (0, _useDeterministicId.useDeterministicId)('RadioInput');
  (0, _react.useEffect)(() => {
    setDeterministicId(getId());
  }, []);
  const id = idProp || deterministicId;

  // Computed checked value
  const checked = typeof checkedProp === 'undefined' ? internalChecked : checkedProp;

  // Styles - pass props with defaults applied for generateStyle
  const styles = (0, _emotion.useStyleNew)({
    generateStyle: _styles.default,
    themeOverride,
    params: {
      disabled,
      context,
      inline,
      hovered,
      readOnly,
      size,
      variant
    },
    componentId: 'RadioInput',
    displayName: 'RadioInput'
  });

  // Event handlers
  const handleInputRef = (0, _react.useCallback)(el => {
    inputElementRef.current = el;
    if (typeof inputRef === 'function') {
      inputRef(el);
    }
  }, [inputRef]);
  const handleClick = (0, _react.useCallback)(e => {
    if (disabled || readOnly) {
      e.preventDefault();
      return;
    }
    if (typeof onClick === 'function') {
      onClick(e);
    }
  }, [disabled, readOnly, onClick]);
  const handleChange = (0, _react.useCallback)(e => {
    if (disabled || readOnly) {
      e.preventDefault();
      return;
    }
    if (typeof checkedProp === 'undefined') {
      setInternalChecked(!internalChecked);
    }
    if (typeof onChange === 'function') {
      onChange(e);
    }
  }, [disabled, readOnly, checkedProp, internalChecked, onChange]);

  // Expose imperative API via ref
  (0, _react.useImperativeHandle)(ref, () => ({
    focus: () => {
      inputElementRef.current?.focus();
    },
    get focused() {
      return (0, _isActiveElement.isActiveElement)(inputElementRef.current);
    },
    get checked() {
      return checked;
    },
    get id() {
      return id;
    },
    ref: containerRef
  }), [checked, id]);
  const handleMouseOver = () => {
    setHovered(true);
  };
  const handleMouseOut = () => {
    setHovered(false);
  };
  const passedProps = (0, _passthroughProps.passthroughProps)(rest);
  return (0, _jsxRuntime.jsxs)("div", {
    css: styles?.radioInput,
    "data-cid": "RadioInput",
    ref: containerRef
    /* eslint-disable-next-line jsx-a11y/mouse-events-have-key-events */,
    onMouseOver: handleMouseOver
    /* eslint-disable-next-line jsx-a11y/mouse-events-have-key-events */,
    onMouseOut: handleMouseOut,
    children: [(0, _jsxRuntime.jsx)("input", {
      ...passedProps,
      id: id,
      ref: handleInputRef,
      value: value,
      name: name,
      checked: checked,
      type: "radio",
      css: styles?.input,
      readOnly: readOnly,
      disabled: disabled,
      onChange: handleChange,
      onClick: handleClick
    }), (0, _jsxRuntime.jsx)("label", {
      htmlFor: id,
      css: styles?.label,
      children: label
    })]
  });
});
RadioInput.displayName = 'RadioInput';
var _default = exports.default = RadioInput;
