// @instructure/ui-avatar/lib/Avatar/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Avatar = void 0;
var _emotion = require("@instructure/emotion");
var _react = require("react");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _renderIconWithProps = require("@instructure/ui-icons/lib/IconPropsProvider/renderIconWithProps.js");
var _props = require("./props");
var _styles = _interopRequireDefault(require("./styles"));
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const ICON_COLOR_MAP = {
  accent1: 'accentBlueColor',
  accent2: 'accentGreenColor',
  accent3: 'accentRedColor',
  accent4: 'accentOrangeColor',
  accent5: 'accentGreyColor',
  accent6: 'accentAshColor',
  ai: 'onColor'
};

/**
---
category: components
---
**/
const Avatar = exports.Avatar = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    size = 'medium',
    color = 'accent1',
    hasInverseColor = false,
    showBorder = 'auto',
    shape = 'circle',
    display = 'inline',
    onImageLoaded = _event => {},
    src,
    name,
    renderIcon,
    alt,
    themeOverride,
    margin
  } = props;
  const [loaded, setLoaded] = (0, _react.useState)(false);
  const styles = (0, _emotion.useStyleNew)({
    generateStyle: _styles.default,
    themeOverride,
    params: {
      loaded,
      size,
      color,
      hasInverseColor,
      shape,
      src,
      showBorder,
      display,
      margin
    },
    componentId: 'Avatar',
    displayName: 'Avatar'
  });
  (0, _react.useEffect)(() => {
    // in case the image is unset in an update, show icons/initials again
    if (loaded && !src) {
      setLoaded(false);
    }
    // Image already loaded (common in SSR)
    //TODO-rework make this work
    // if (src && !loaded && imgRef.current && imgRef.current.complete) {
    //   setLoaded(true)
    //   onImageLoaded?.()
    // }
  }, [loaded, src]);
  const makeInitialsFromName = () => {
    if (!name || typeof name !== 'string' || name.trim().length === 0) {
      return;
    }
    const currentName = name.trim();
    if (currentName.match(/\s+/)) {
      const names = currentName.split(/\s+/);
      return (names[0][0] + names[names.length - 1][0]).toUpperCase();
    } else {
      return currentName[0].toUpperCase();
    }
  };
  const renderInitials = () => {
    return (0, _jsxRuntime.jsx)("span", {
      "aria-hidden": "true",
      children: makeInitialsFromName()
    });
  };
  const renderImage = () => (0, _jsxRuntime.jsx)("img", {
    style: {
      width: '100%',
      height: '100%',
      objectFit: 'cover',
      objectPosition: 'center',
      ...(loaded ? {} : {
        display: 'none'
      })
    },
    src: src,
    alt: alt ? alt : '',
    onLoad: event => {
      setLoaded(true);
      onImageLoaded(event);
    }
  });
  const renderContent = () => {
    //image in avatar - prioritize image over icon
    if (src) {
      return (0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
        children: [renderImage(), loaded ? null : renderInitials()]
      });
    }

    //icon in avatar
    if (renderIcon) {
      const iconSize = _props.avatarSizeToIconSize[size];
      const iconColor = hasInverseColor ? 'onColor' : ICON_COLOR_MAP[color];
      return (0, _renderIconWithProps.renderIconWithProps)(renderIcon, iconSize, iconColor);
    }

    //initials in avatar
    return renderInitials();
  };
  return (0, _jsxRuntime.jsx)("div", {
    ...(0, _passthroughProps.passthroughProps)({
      ...props
    }),
    ref: ref,
    "aria-label": alt ? alt : undefined,
    role: alt ? 'img' : undefined,
    css: styles?.avatar,
    children: renderContent()
  });
});
Avatar.displayName = 'Avatar';
var _default = exports.default = Avatar;
