// @instructure/ui-top-nav-bar/es/TopNavBar/v1/TopNavBarMenuItems/index.js
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Children, Component } from 'react';
import { omitProps, safeCloneElement, withDeterministicId } from '@instructure/ui-react-utils';
import { warn, error } from '@instructure/console';
import { withStyle } from '@instructure/emotion';
import { Drilldown } from '@instructure/ui-drilldown/v11_6';
import { TruncateList } from '@instructure/ui-truncate-list';
import { TopNavBarContext } from "../TopNavBarContext.js";
import { TopNavBarItem } from "../TopNavBarItem/index.js";
import { mapItemsForDrilldown, renderMappedItemDrilldownSubpages, renderMappedItemsAsDrilldownOptions } from "../utils/mapItemsForDrilldown.js";
import generateStyle from "./styles.js";
import generateComponentTheme from "./theme.js";
import { allowedProps } from "./props.js";
import { jsx as _jsx } from "@emotion/react/jsx-runtime";
import { createElement as _createElement } from "@emotion/react";
/**
---
parent: TopNavBar
id: TopNavBar.MenuItems
---
@module TopNavBarMenuItems
**/
let TopNavBarMenuItems = (_dec = withDeterministicId(), _dec2 = withStyle(generateStyle, generateComponentTheme), _dec(_class = _dec2(_class = class TopNavBarMenuItems extends Component {
  static displayName = "TopNavBarMenuItems";
  static componentId = 'TopNavBar.MenuItems';
  static allowedProps = allowedProps;
  static defaultProps = {};
  static contextType = TopNavBarContext;
  _hiddenMenuItemsMenuTriggerId;
  _hiddenItemsMenuId;
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  constructor(props) {
    super(props);
    this._hiddenMenuItemsMenuTriggerId = props.deterministicId('TopNavBarMenuItems-hiddenMenuItemsMenuTrigger');
    this._hiddenItemsMenuId = props.deterministicId('TopNavBarSmallViewportLayout-drilldown');
    this.state = {
      key: 0,
      visibleItemsCount: undefined
    };
  }
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate(prevProps) {
    this.props.makeStyles?.();
    if (Children.count(prevProps.children) !== this.childrenArray.length) {
      this.setState({
        key: this.state.key + 1
      });
    }
  }
  get childrenArray() {
    return Children.toArray(this.props.children);
  }
  renderOptionContent = (children, itemProps) => {
    const {
      styles
    } = this.props;
    return _jsx("span", {
      css: itemProps?.status === 'active' ? styles?.submenuOptionActive : styles?.submenuOption,
      children: children
    });
  };
  renderHiddenItemsMenu(hiddenItems) {
    const {
      currentPageId,
      renderHiddenItemsMenuTriggerLabel,
      renderHiddenItemsMenuTriggerAriaLabel
    } = this.props;
    if (!hiddenItems.length) {
      return _jsx("span", {});
    }
    const mappedItems = mapItemsForDrilldown(hiddenItems, {
      renderOptionContent: this.renderOptionContent,
      currentPageId
    });
    const subPages = renderMappedItemDrilldownSubpages(mappedItems);
    const options = renderMappedItemsAsDrilldownOptions(mappedItems);
    const hasActiveChild = hiddenItems.find(child => child.props.status === 'active');
    return _jsx(TopNavBarItem, {
      "aria-label": renderHiddenItemsMenuTriggerAriaLabel ? renderHiddenItemsMenuTriggerAriaLabel(hiddenItems.length) : undefined,
      id: this._hiddenMenuItemsMenuTriggerId,
      status: hasActiveChild ? 'active' : 'default',
      renderSubmenu: _jsx(Drilldown, {
        shouldSetAriaExpanded: false,
        rootPageId: this._hiddenItemsMenuId,
        children: [_jsx(Drilldown.Page, {
          id: this._hiddenItemsMenuId,
          children: options
        }, this._hiddenItemsMenuId), ...subPages]
      }),
      children: renderHiddenItemsMenuTriggerLabel(hiddenItems.length)
    });
  }
  renderChildren() {
    const {
      currentPageId
    } = this.props;
    return this.childrenArray.map(child => {
      if (!child) {
        return;
      }
      const {
        id,
        status,
        variant = 'default',
        // workaround because after react 19 defaultProps are not working as expected
        renderSubmenu,
        renderAvatar
      } = child.props;
      const isCurrentPage = currentPageId === id;
      if (renderAvatar) {
        error(false, `Items in <TopNavBar.MenuItems> are not allowed to have avatars, but item with id: "${id}" has \`renderAvatar\` prop.`);
        return null;
      }
      if (!isCurrentPage) {
        return child;
      } else {
        if (variant !== 'default') {
          warn(false, `Only \`variant="default"\` items can be set to current/active, but the item with id "${id}" is "${variant}" variant.`);
          return child;
        }
        if (status === 'disabled') {
          warn(false, `Disabled items can not be set to current/active, but the item with id "${id}" is disabled.`);
          return child;
        }
        return safeCloneElement(child, {
          status: 'active',
          // if it has submenu, the root item cannot be current,
          // just a link inside the submenu
          'aria-current': renderSubmenu ? undefined : 'page'
        });
      }
    });
  }
  render() {
    const {
      listLabel,
      styles,
      renderHiddenItemsMenuTriggerLabel,
      renderHiddenItemsMenuTriggerAriaLabel,
      ...restProps
    } = this.props;
    if (!this.childrenArray.length) {
      return null;
    }
    if (this.context.layout === 'smallViewport') {
      // in smallViewport mode it is rendered as a Drilldown
      return null;
    }
    return _createElement(TruncateList, {
      ...omitProps(restProps, allowedProps),
      key: this.state.key // rerender if child count changes
      ,
      elementRef: this.handleRef,
      css: styles?.topNavBarMenuItems,
      visibleItemsCount: this.state.visibleItemsCount,
      onUpdate: ({
        visibleItemsCount
      }) => {
        this.setState({
          visibleItemsCount
        });
      },
      renderHiddenItemMenu: hiddenChildren => this.renderHiddenItemsMenu(hiddenChildren),
      ...(styles?.itemSpacing && {
        itemSpacing: styles.itemSpacing
      }),
      "aria-label": listLabel,
      "data-cid": "TopNavBarMenuItems"
    }, this.renderChildren());
  }
}) || _class) || _class);
export { TopNavBarMenuItems };
export default TopNavBarMenuItems;
