// @instructure/ui-breadcrumb/es/Breadcrumb/v1/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { isValidElement, cloneElement, Children, Component } from 'react';
import { View } from '@instructure/ui-view/v11_6';
import { withStyle } from '@instructure/emotion';
import { IconArrowOpenEndSolid } from '@instructure/ui-icons';
import { BreadcrumbLink } from "./BreadcrumbLink/index.js";
import generateStyle from "./styles.js";
import generateComponentTheme from "./theme.js";
import { allowedProps } from "./props.js";
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
/**
---
category: components
---
**/
let Breadcrumb = (_dec = withStyle(generateStyle, generateComponentTheme), _dec(_class = class Breadcrumb extends Component {
  static displayName = "Breadcrumb";
  static componentId = 'Breadcrumb';
  static allowedProps = allowedProps;
  static defaultProps = {
    size: 'medium',
    children: null
  };
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  addAriaCurrent = child => {
    const updatedChild = /*#__PURE__*/cloneElement(child, {
      'aria-current': 'page'
    });
    return updatedChild;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  static Link = BreadcrumbLink;
  renderChildren() {
    const {
      styles,
      children
    } = this.props;
    const numChildren = Children.count(children);
    const inlineStyle = {
      maxWidth: `${Math.floor(100 / numChildren)}%`
    };
    let isAriaCurrentSet = false;
    return Children.map(children, (child, index) => {
      const isLastElement = index === numChildren - 1;
      if (/*#__PURE__*/isValidElement(child)) {
        const isCurrentPage = child.props.isCurrentPage || false;
        if (isAriaCurrentSet && isCurrentPage) {}
        if (isCurrentPage) {
          isAriaCurrentSet = true;
        }
      }
      return _jsxs("li", {
        css: styles?.crumb,
        style: inlineStyle,
        children: [!isAriaCurrentSet && isLastElement && child.props.isCurrentPage !== false ? this.addAriaCurrent(child) : child, index < numChildren - 1 && _jsx(IconArrowOpenEndSolid, {
          color: "auto",
          css: styles?.separator
        })]
      });
    });
  }
  render() {
    const {
      styles
    } = this.props;
    return _jsx(View, {
      role: "navigation",
      as: "div",
      margin: this.props.margin,
      "aria-label": this.props.label,
      elementRef: this.handleRef,
      "data-cid": "Breadcrumb",
      children: _jsx("ol", {
        css: styles?.breadcrumb,
        children: this.renderChildren()
      })
    });
  }
}) || _class);
export default Breadcrumb;
export { Breadcrumb, BreadcrumbLink };
