// @instructure/ui-top-nav-bar/lib/TopNavBar/v1/utils/mapItemsForDrilldown.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.renderMappedItemsAsDrilldownOptions = exports.renderMappedItemDrilldownSubpages = exports.mapItemsForDrilldown = exports.default = void 0;
require("core-js/modules/es.map.get-or-insert.js");
require("core-js/modules/es.map.get-or-insert-computed.js");
var _react = require("react");
var _console = require("@instructure/console");
var _generateId = require("@instructure/ui-utils/lib/generateId.js");
var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");
var _v11_ = require("@instructure/ui-drilldown/v11_6");
var _TopNavBarItem = require("../TopNavBarItem");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _react2 = require("@emotion/react");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const mapItemsForDrilldown = (itemList, options = {}) => {
  const submenus = [];
  const {
    currentPageId,
    renderOptionContent
  } = options;
  const customPopoverIdMap = new Map();
  _react.Children.forEach(itemList, item => {
    if (!item || !(0, _matchComponentTypes.matchComponentTypes)(item, [_TopNavBarItem.TopNavBarItem])) return;
    const {
      renderSubmenu,
      customPopoverConfig,
      id,
      children,
      status,
      variant = 'default',
      // workaround because defaultProps not working after react 19 upgrade
      href,
      onClick,
      shouldCloseOnClick
    } = item.props;
    let submenu = renderSubmenu;
    let customPopover = customPopoverConfig;
    let customPopoverId;
    let optionSubPageId;
    const submenuPages = [];
    if (submenu) {
      if (!(0, _matchComponentTypes.matchComponentTypes)(submenu, [_v11_.Drilldown])) {
        (0, _console.warn)(false, `The "renderSubmenu" prop accepts only Drilldown components, but the item with id "${id}" received: "${renderSubmenu}".`);
        submenu = undefined;
      }

      // if still has submenu...
      if (submenu) {
        optionSubPageId = submenu.props.rootPageId;
        submenuPages.push(..._react.Children.toArray(submenu.props.children));
      }
    }
    if (customPopover) {
      if (!customPopover.children) {
        (0, _console.warn)(false, `Pass the content of the custom Popover as "customPopoverConfig.children" for the item with id: "${id}".`);
        customPopover = undefined;
      }
      if (submenu) {
        (0, _console.warn)(false, `TopNavBar.Items are not allowed to have both the "renderSubmenu" and "customPopoverConfig" props. For submenus, pass a Drilldown component via the "renderSubmenu" prop, and only use "customPopoverConfig" for custom features. Item with id: "${id}" will ignore the "customPopoverConfig" prop.`);
        customPopover = undefined;
      }

      // if still has customPopover...
      if (customPopover) {
        customPopoverId = (0, _generateId.generateId)(`TopNavBarItem__customPopoverOption`, customPopoverIdMap);
        optionSubPageId = customPopoverId;
        submenuPages.push((0, _jsxRuntime.jsx)(_v11_.Drilldown.Page, {
          id: customPopoverId,
          children: (0, _jsxRuntime.jsx)(_v11_.Drilldown.Option, {
            id: `${customPopoverId}__option`,
            children: customPopover.children
          })
        }, customPopoverId));
      }
    }
    let ariaCurrent = item.props['aria-current'] || (id === currentPageId ? 'page' : undefined);
    if (ariaCurrent) {
      if (variant !== 'default') {
        (0, _console.warn)(false, `Only \`variant="default"\` items can be set to current/active, but the item with id "${id}" is "${variant}" variant.`);
        ariaCurrent = undefined;
      }
      if (status === 'disabled') {
        (0, _console.warn)(false, `Disabled items can not be set to current/active, but the item with id "${id}" is disabled.`);
        ariaCurrent = undefined;
      }
    }
    submenus.push({
      item,
      submenuPages,
      optionData: {
        id,
        disabled: status === 'disabled',
        href,
        onOptionClick: event => {
          onClick?.(event);
        },
        children: typeof renderOptionContent === 'function' ? renderOptionContent(children, {
          ...item.props,
          status: ariaCurrent === 'page' ? 'active' : item.props.status
        }) : children,
        subPageId: optionSubPageId,
        'aria-current': ariaCurrent,
        shouldCloseOnClick: shouldCloseOnClick
      }
    });
  });
  return submenus;
};
exports.mapItemsForDrilldown = mapItemsForDrilldown;
const renderMappedItemDrilldownSubpages = mappedItems => {
  const subPages = mappedItems.map(option => option.submenuPages).filter(submenu => !!submenu);
  return subPages.flat();
};
exports.renderMappedItemDrilldownSubpages = renderMappedItemDrilldownSubpages;
const renderMappedItemsAsDrilldownOptions = mappedItems => {
  return mappedItems.map(mappedItem => {
    const {
      optionData,
      item: originalItem
    } = mappedItem;
    const onHiddenClick = originalItem.props.onHiddenClick;
    if (onHiddenClick) {
      // if the `onHiddenClick` prop is provided for the MenuItem it should overwrite the `onClick` prop if there is any
      optionData.onOptionClick = onHiddenClick;
    }
    return (0, _react2.createElement)(_v11_.Drilldown.Option, {
      ...optionData,
      key: optionData.id,
      afterLabelContentVAlign: "center"
    }, optionData.children);
  });
};
exports.renderMappedItemsAsDrilldownOptions = renderMappedItemsAsDrilldownOptions;
var _default = exports.default = mapItemsForDrilldown;
