// @instructure/ui-flex/lib/Flex/v1/Item/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Item = void 0;
var _react = require("react");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _v11_ = require("@instructure/ui-view/v11_6");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: Flex
id: Flex.Item
---
**/
let Item = exports.Item = (_dec = (0, _emotion.withStyle)(_styles.default, null), _dec(_class = class Item extends _react.Component {
  static displayName = "Item";
  static componentId = 'Flex.Item';
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    as: 'span',
    shouldGrow: false,
    shouldShrink: false
  };
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  render() {
    const props = (0, _omitProps.omitProps)(this.props, Item.allowedProps);
    const {
      as,
      children,
      withVisualDebug,
      textAlign,
      size,
      direction,
      margin,
      padding,
      overflowX,
      overflowY,
      styles
    } = this.props;
    const dirColumn = direction === 'column';
    const dirRow = direction === 'row';
    return (0, _jsxRuntime.jsx)(_v11_.View, {
      ...props,
      css: styles?.flexItem,
      elementRef: this.handleRef,
      as: as,
      minHeight: dirColumn ? size : undefined,
      minWidth: dirRow ? size : undefined,
      textAlign: textAlign,
      margin: margin,
      padding: padding,
      overflowX: overflowX,
      overflowY: overflowY || (dirColumn ? 'auto' : 'visible'),
      withVisualDebug: withVisualDebug,
      children: children
    });
  }
}) || _class);
var _default = exports.default = Item;
