// @instructure/ui-form-field/lib/FormFieldMessage/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.FormFieldMessage = void 0;
var _react = require("react");
var _ScreenReaderContent = require("@instructure/ui-a11y-content/lib/ScreenReaderContent");
var _uiIcons = require("@instructure/ui-icons");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
parent: FormField
---

This is a helper component that is used by most of the custom form
components. In most cases it shouldn't be used directly.

```js
---
type: example
---
<FormFieldMessage variant="error">Invalid value</FormFieldMessage>
```
**/
const FormFieldMessage = exports.FormFieldMessage = /*#__PURE__*/(0, _react.forwardRef)(({
  variant = 'hint',
  children,
  themeOverride
}, ref) => {
  const styles = (0, _emotion.useStyleNew)({
    generateStyle: _styles.default,
    themeOverride,
    params: {
      variant
    },
    componentId: 'FormFieldMessage',
    displayName: 'FormFieldMessage'
  });
  const handleRef = el => {
    if (typeof ref === 'function') {
      ref(el);
    } else if (ref) {
      const refObject = ref;
      refObject.current = el;
    }
  };
  const isErrorVariant = variant === 'error' || variant === 'newError';
  const shouldShowIcon = (isErrorVariant || variant === 'success') && children;
  return variant !== 'screenreader-only' ? (0, _jsxRuntime.jsxs)("span", {
    css: {
      display: 'flex'
    },
    children: [shouldShowIcon && (0, _jsxRuntime.jsx)("span", {
      css: styles?.icon,
      children: isErrorVariant ? (0, _jsxRuntime.jsx)(_uiIcons.AlertCircleInstUIIcon, {
        size: "sm",
        color: "errorColor"
      }) : (0, _jsxRuntime.jsx)(_uiIcons.CheckCircle2InstUIIcon, {
        size: "sm",
        color: "successColor"
      })
    }), (0, _jsxRuntime.jsx)("span", {
      css: styles?.formFieldMessage,
      ref: handleRef,
      children: children
    })]
  }) : (0, _jsxRuntime.jsx)(_ScreenReaderContent.ScreenReaderContent, {
    elementRef: handleRef,
    children: children
  });
});
FormFieldMessage.displayName = 'FormFieldMessage';
var _default = exports.default = FormFieldMessage;
