// @instructure/ui-number-input/es/NumberInput/v1/index.js
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Fragment, Component } from 'react';
import keycode from 'keycode';
import { FormField } from '@instructure/ui-form-field/v11_6';
import { IconArrowOpenDownLine, IconArrowOpenUpLine } from '@instructure/ui-icons';
import { omitProps, pickProps, callRenderProp, getInteraction, withDeterministicId } from '@instructure/ui-react-utils';
import { hasVisibleChildren } from '@instructure/ui-a11y-utils';
import { withStyle } from '@instructure/emotion';
import generateStyle from "./styles.js";
import generateComponentTheme from "./theme.js";
import { allowedProps } from "./props.js";
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
/**
---
category: components
id: NumberInput
---
**/
let NumberInput = (_dec = withDeterministicId(), _dec2 = withStyle(generateStyle, generateComponentTheme), _dec(_class = _dec2(_class = class NumberInput extends Component {
  static displayName = "NumberInput";
  static componentId = 'NumberInput';
  static allowedProps = allowedProps;
  static defaultProps = {
    // Leave interaction default undefined so that `disabled` and `readOnly` can also be supplied
    interaction: undefined,
    messages: [],
    isRequired: false,
    showArrows: true,
    size: 'medium',
    display: 'block',
    textAlign: 'start',
    inputMode: 'numeric',
    allowStringValue: false
  };
  state = {
    hasFocus: false
  };
  ref = null;
  _input = null;
  _id;
  handleRef = el => {
    this.ref = el;
  };
  get id() {
    if (this.props.id) {
      return this.props.id;
    }
    if (!this._id) {
      this._id = this.props.deterministicId();
    }
    return this._id;
  }
  get invalid() {
    return !!this.props.messages && this.props.messages.some(message => message.type === 'error' || message.type === 'newError');
  }
  get interaction() {
    return getInteraction({
      props: this.props
    });
  }
  componentDidMount() {
    this.props.makeStyles?.(this.makeStyleVariables);
  }
  componentDidUpdate() {
    this.props.makeStyles?.(this.makeStyleVariables);
  }
  get makeStyleVariables() {
    return {
      interaction: this.interaction,
      hasFocus: this.state.hasFocus,
      invalid: this.invalid
    };
  }
  handleInputRef = element => {
    this._input = element;
    if (typeof this.props.inputRef === 'function') {
      this.props.inputRef(element);
    }
  };
  handleFocus = event => {
    this.setState({
      hasFocus: true
    });
    if (typeof this.props.onFocus === 'function') {
      this.props.onFocus(event);
    }
  };
  handleBlur = event => {
    this.setState({
      hasFocus: false
    });
    if (typeof this.props.onBlur === 'function') {
      this.props.onBlur(event);
    }
  };
  handleChange = event => {
    if (typeof this.props.onChange === 'function') {
      this.props.onChange(event, event.target.value);
    }
  };
  handleKeyDown = event => {
    const {
      onKeyDown,
      onDecrement,
      onIncrement
    } = this.props;
    if (typeof onKeyDown === 'function') {
      onKeyDown(event);
    }
    if (event.keyCode === keycode.codes.down) {
      event.preventDefault();
      if (typeof onDecrement === 'function') {
        onDecrement(event);
      }
    } else if (event.keyCode === keycode.codes.up) {
      event.preventDefault();
      if (typeof onIncrement === 'function') {
        onIncrement(event);
      }
    }
  };
  handleClickUpArrow = event => {
    this.arrowClicked(event, this.props.onIncrement);
  };
  handleClickDownArrow = event => {
    this.arrowClicked(event, this.props.onDecrement);
  };
  arrowClicked(event, callback) {
    const {
      interaction
    } = this;
    event.preventDefault();
    if (interaction === 'enabled') {
      this._input?.focus();
      if (typeof callback === 'function') {
        callback(event);
      }
    }
  }
  renderArrows(customIcons) {
    return _jsxs("span", {
      css: this.props.styles?.arrowContainer,
      children: [_jsx("button", {
        "aria-hidden": true,
        css: this.props.styles?.arrow,
        onMouseDown: this.handleClickUpArrow,
        tabIndex: -1,
        type: "button",
        children: customIcons?.increase ? callRenderProp(customIcons.increase) : _jsx(IconArrowOpenUpLine, {})
      }), _jsx("button", {
        "aria-hidden": true,
        css: this.props.styles?.arrow,
        onMouseDown: this.handleClickDownArrow,
        tabIndex: -1,
        type: "button",
        children: customIcons?.decrease ? callRenderProp(customIcons.decrease) : _jsx(IconArrowOpenDownLine, {})
      })]
    });
  }
  render() {
    const {
      renderLabel,
      display,
      placeholder,
      isRequired,
      showArrows,
      value,
      width,
      styles,
      allowStringValue,
      renderIcons,
      margin
    } = this.props;
    const {
      interaction
    } = this;
    const rawLabel = callRenderProp(renderLabel);
    const label = hasVisibleChildren(rawLabel) ? _jsxs(Fragment, {
      children: [rawLabel, isRequired && _jsxs("span", {
        css: this.invalid ? styles?.requiredInvalid : {},
        "aria-hidden": true,
        children: [' ', "*"]
      })]
    }) : rawLabel;
    return _jsx(FormField, {
      ...pickProps(this.props, FormField.allowedProps),
      label: label,
      inline: display === 'inline-block',
      id: this.id,
      elementRef: this.handleRef,
      margin: margin,
      "data-cid": "NumberInput",
      children: _jsx("span", {
        css: this.props.styles?.inputWidth,
        style: width ? {
          width
        } : undefined,
        children: _jsxs("span", {
          css: this.props.styles?.inputContainer,
          children: [_jsx("input", {
            ...omitProps(this.props, [...FormField.allowedProps, ...NumberInput.allowedProps]),
            css: this.props.styles?.input,
            "aria-invalid": this.invalid ? 'true' : undefined,
            id: this.id,
            type: allowStringValue ? 'text' : 'number',
            inputMode: this.props.inputMode,
            placeholder: placeholder,
            ref: this.handleInputRef,
            required: isRequired,
            value: value,
            disabled: interaction === 'disabled',
            readOnly: interaction === 'readonly',
            onFocus: this.handleFocus,
            onBlur: this.handleBlur,
            onChange: this.handleChange,
            onKeyDown: this.handleKeyDown
          }), showArrows ? this.renderArrows(renderIcons) : null]
        })
      })
    });
  }
}) || _class) || _class);
export default NumberInput;
export { NumberInput };
