// @instructure/ui-progress/es/ProgressBar/v1/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { View } from '@instructure/ui-view/v11_6';
import { callRenderProp, passthroughProps } from '@instructure/ui-react-utils';
import { withStyle } from '@instructure/emotion';
import generateStyle from "./styles.js";
import generateComponentTheme from "./theme.js";
import { allowedProps } from "./props.js";
/**
---
category: components
---
**/
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
let ProgressBar = (_dec = withStyle(generateStyle, generateComponentTheme), _dec(_class = class ProgressBar extends Component {
  static displayName = "ProgressBar";
  static componentId = 'ProgressBar';
  static allowedProps = allowedProps;
  static defaultProps = {
    formatScreenReaderValue: ({
      valueNow,
      valueMax
    }) => `${valueNow} / ${valueMax}`,
    size: 'medium',
    valueMax: 100,
    valueNow: 0,
    as: 'div',
    color: 'primary',
    shouldAnimate: false,
    // default to showing `success` color on completion
    meterColor: ({
      valueNow,
      valueMax
    }) => valueNow / valueMax >= 1 ? 'success' : 'brand'
  };
  ref = null;
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  render() {
    const {
      renderValue,
      formatScreenReaderValue,
      valueNow,
      valueMax,
      screenReaderLabel,
      size,
      color,
      meterColor,
      renderValueInside,
      styles,
      ...props
    } = this.props;
    const valueText = typeof formatScreenReaderValue === 'function' ? formatScreenReaderValue({
      valueNow: valueNow,
      valueMax: valueMax
    }) : undefined;
    const value = callRenderProp(renderValue, {
      valueNow: valueNow,
      valueMax: valueMax
    });
    return _jsxs(View, {
      ...passthroughProps(props),
      as: this.props.as,
      css: styles?.progressBar,
      margin: this.props.margin,
      elementRef: this.handleRef,
      "data-cid": "ProgressBar",
      children: [_jsxs("span", {
        css: styles?.trackLayout,
        children: [_jsx("progress", {
          css: styles?.htmlProgress,
          max: valueMax,
          value: valueNow,
          "aria-valuetext": valueText,
          "aria-label": screenReaderLabel
        }), _jsx("span", {
          css: styles?.track,
          role: "presentation",
          "aria-hidden": "true",
          children: _jsx("span", {
            css: styles?.trackValue,
            children: renderValueInside && value
          })
        })]
      }), value && !renderValueInside && _jsx("span", {
        css: styles?.value,
        "aria-hidden": "true",
        children: value
      })]
    });
  }
}) || _class);
export default ProgressBar;
export { ProgressBar };
