// @instructure/ui-navigation/lib/AppNav/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.AppNav = void 0;
var _react = require("react");
var _emotion = require("@instructure/emotion");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _latest = require("@instructure/ui-view/latest");
var _latest2 = require("@instructure/ui-menu/latest");
var _Item = require("./Item");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _TruncateList = require("@instructure/ui-truncate-list/lib/TruncateList");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let AppNav = exports.AppNav = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class AppNav extends _react.Component {
  static displayName = "AppNav";
  static componentId = 'AppNav';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    children: null,
    debounce: 300,
    margin: '0',
    renderTruncateLabel: () => 'More',
    visibleItemsCount: 0
  };
  static Item = _Item.Item;
  state = {
    isMeasuring: false
  };
  ref = null;
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  renderMenu(items) {
    return (0, _jsxRuntime.jsx)(_latest2.Menu, {
      trigger: (0, _jsxRuntime.jsx)(AppNav.Item, {
        renderLabel: (0, _callRenderProp.callRenderProp)(this.props.renderTruncateLabel)
      }),
      children: items.map((item, index) => {
        return (0, _jsxRuntime.jsx)(_latest2.Menu.Item, {
          href: item.props.href ? item.props.href : undefined,
          onClick: item.props.onClick && !item.props.href ? item.props.onClick : undefined,
          children: (0, _callRenderProp.callRenderProp)(item.props.renderLabel)
        }, index);
      })
    });
  }
  render() {
    const {
      visibleItemsCount,
      screenReaderLabel,
      margin,
      debounce,
      styles
    } = this.props;
    const passthroughProps = _latest.View.omitViewProps((0, _omitProps.omitProps)(this.props, AppNav.allowedProps), AppNav);
    const renderBeforeItems = (0, _callRenderProp.callRenderProp)(this.props.renderBeforeItems);
    const renderAfterItems = (0, _callRenderProp.callRenderProp)(this.props.renderAfterItems);
    const hasRenderedContent = renderBeforeItems || renderAfterItems;
    return (0, _jsxRuntime.jsxs)(_latest.View, {
      ...passthroughProps,
      as: "nav",
      css: [styles?.appNav, hasRenderedContent ? styles?.alignCenter : ''],
      margin: margin,
      display: hasRenderedContent ? 'flex' : 'block',
      elementRef: this.handleRef,
      "data-cid": "AppNav",
      children: [renderBeforeItems && (0, _jsxRuntime.jsx)("span", {
        children: renderBeforeItems
      }), (0, _jsxRuntime.jsx)(_TruncateList.TruncateList, {
        visibleItemsCount: visibleItemsCount,
        debounce: debounce,
        onUpdate: this.props.onUpdate,
        renderHiddenItemMenu: hiddenChildren => this.renderMenu(hiddenChildren),
        itemSpacing: styles?.horizontalMargin,
        fixMenuTriggerWidth: styles?.menuTriggerWidth,
        css: styles?.list,
        "aria-label": (0, _callRenderProp.callRenderProp)(screenReaderLabel),
        children: this.props.children
      }), renderAfterItems && (0, _jsxRuntime.jsx)("span", {
        children: renderAfterItems
      })]
    });
  }
}) || _class);
var _default = exports.default = AppNav;
