// @instructure/ui-color-picker/lib/ColorPreset/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ColorPreset = void 0;
var _react = require("react");
var _emotion = require("@instructure/emotion");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _uiColorUtils = _interopRequireDefault(require("@instructure/ui-color-utils"));
var _v11_ = require("@instructure/ui-buttons/v11_6");
var _v11_2 = require("@instructure/ui-view/v11_6");
var _v11_3 = require("@instructure/ui-tooltip/v11_6");
var _v11_4 = require("@instructure/ui-popover/v11_6");
var _v11_5 = require("@instructure/ui-text/v11_6");
var _v11_6 = require("@instructure/ui-drilldown/v11_6");
var _IconAddLine = require("@instructure/ui-icons/lib/generated/IconAddLine.js");
var _IconCheckDarkSolid = require("@instructure/ui-icons/lib/generated/IconCheckDarkSolid.js");
var _v = require("../../ColorIndicator/v1");
var _v2 = require("../../ColorMixer/v1");
var _v3 = require("../../ColorContrast/v1");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let ColorPreset = exports.ColorPreset = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class ColorPreset extends _react.Component {
  static displayName = "ColorPreset";
  static allowedProps = _props.allowedProps;
  static componentId = 'ColorPreset';
  static defaultProps = {
    disabled: false
  };
  constructor(props) {
    super(props);
    this.state = {
      openEditor: false,
      openAddNew: false,
      newColor: {
        r: 51,
        g: 99,
        b: 42,
        a: 1
      }
    };
  }
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  get isModifiable() {
    return typeof this.props.colorMixerSettings?.onPresetChange === 'function';
  }
  isSelectedColor(color) {
    const {
      selected
    } = this.props;
    return !!selected && _uiColorUtils.default.colorToHex8(selected) === _uiColorUtils.default.colorToHex8(color);
  }
  onMenuItemSelected = color => (_e, args) => {
    if (args.value === 'select') {
      this.props.onSelect(color);
    }
    if (args.value === 'remove') {
      this.props?.colorMixerSettings?.onPresetChange(this.props.colors.filter(clr => clr !== color));
    }
  };
  renderAddNewPresetButton = () => (0, _jsxRuntime.jsxs)(_v11_4.Popover, {
    renderTrigger: (0, _jsxRuntime.jsx)("div", {
      css: this.props?.styles?.addNewPresetButton,
      children: (0, _jsxRuntime.jsx)(_v11_.IconButton, {
        disabled: this.props.disabled,
        screenReaderLabel: this.props.colorMixerSettings.addNewPresetButtonScreenReaderLabel,
        children: (0, _jsxRuntime.jsx)(_IconAddLine.IconAddLine, {})
      })
    }),
    isShowingContent: this.state.openAddNew,
    onShowContent: () => {
      if (this.props.disabled) return;
      this.setState({
        openAddNew: true
      });
    },
    onHideContent: () => {
      this.setState({
        openAddNew: false
      });
    },
    on: "click",
    screenReaderLabel: this.props.popoverScreenReaderLabel,
    shouldContainFocus: true,
    shouldReturnFocus: true,
    shouldCloseOnDocumentClick: true,
    offsetY: 16,
    mountNode: () => document.getElementById('main'),
    children: [(0, _jsxRuntime.jsxs)("div", {
      css: this.props.styles?.popoverContent,
      children: [(0, _jsxRuntime.jsx)(_v2.ColorMixer, {
        value: _uiColorUtils.default.colorToHex8(this.state.newColor),
        onChange: newColor => this.setState({
          newColor: _uiColorUtils.default.colorToRGB(newColor)
        }),
        withAlpha: this.props?.colorMixerSettings?.colorMixer?.withAlpha,
        rgbRedInputScreenReaderLabel: this.props.colorMixerSettings.colorMixer.rgbRedInputScreenReaderLabel,
        rgbGreenInputScreenReaderLabel: this.props.colorMixerSettings.colorMixer.rgbGreenInputScreenReaderLabel,
        rgbBlueInputScreenReaderLabel: this.props.colorMixerSettings.colorMixer.rgbBlueInputScreenReaderLabel,
        rgbAlphaInputScreenReaderLabel: this.props.colorMixerSettings.colorMixer.rgbAlphaInputScreenReaderLabel,
        colorSliderNavigationExplanationScreenReaderLabel: this.props.colorMixerSettings.colorMixer.colorSliderNavigationExplanationScreenReaderLabel,
        alphaSliderNavigationExplanationScreenReaderLabel: this.props.colorMixerSettings.colorMixer.alphaSliderNavigationExplanationScreenReaderLabel,
        colorPaletteNavigationExplanationScreenReaderLabel: this.props.colorMixerSettings.colorMixer.colorPaletteNavigationExplanationScreenReaderLabel
      }), this.props?.colorMixerSettings?.colorContrast && (0, _jsxRuntime.jsx)("div", {
        css: this.props.styles?.popoverContrastBlock,
        children: (0, _jsxRuntime.jsx)(_v3.ColorContrast, {
          firstColor: this.props.colorMixerSettings.colorContrast.firstColor,
          secondColor: _uiColorUtils.default.colorToHex8(this.state.newColor),
          label: this.props.colorMixerSettings.colorContrast.label,
          successLabel: this.props.colorMixerSettings.colorContrast.successLabel,
          failureLabel: this.props.colorMixerSettings.colorContrast.failureLabel,
          normalTextLabel: this.props.colorMixerSettings.colorContrast.normalTextLabel,
          largeTextLabel: this.props.colorMixerSettings.colorContrast.largeTextLabel,
          graphicsTextLabel: this.props.colorMixerSettings.colorContrast.graphicsTextLabel,
          firstColorLabel: this.props.colorMixerSettings.colorContrast.firstColorLabel,
          secondColorLabel: this.props.colorMixerSettings.colorContrast.secondColorLabel
        })
      })]
    }), (0, _jsxRuntime.jsxs)("div", {
      css: this.props.styles?.popoverFooter,
      children: [(0, _jsxRuntime.jsx)(_v11_.Button, {
        onClick: () => {
          this.props?.colorMixerSettings?.onPresetChange([_uiColorUtils.default.colorToHex8(this.state.newColor), ...this.props.colors]);
          this.setState({
            openAddNew: false
          });
        },
        color: "primary",
        margin: "0 xx-small 0 xx-small",
        children: this.props.colorMixerSettings?.popoverAddButtonLabel
      }), (0, _jsxRuntime.jsx)(_v11_.Button, {
        onClick: () => this.setState({
          openAddNew: false
        }),
        color: "secondary",
        margin: "0 xx-small 0 xx-small",
        children: this.props.colorMixerSettings?.popoverCloseButtonLabel
      })]
    })]
  });
  renderColorIndicator = (color, selectOnClick) => {
    const indicatorBase = this.renderIndicatorBase(color, selectOnClick);
    return this.props.disabled ? indicatorBase : this.renderIndicatorTooltip(indicatorBase, color);
  };
  renderIndicatorBase = (color, selectOnClick) => {
    const hexCode = color;
    const isSelected = this.isSelectedColor(color);
    const screenReaderLabel = typeof this.props.colorScreenReaderLabel === 'function' ? this.props.colorScreenReaderLabel(hexCode, isSelected) : hexCode;
    return (0, _jsxRuntime.jsx)(_v11_2.View, {
      disabled: this.props.disabled,
      position: "relative",
      width: "2.375rem",
      height: "2.375rem",
      background: "transparent",
      margin: "xx-small",
      display: "inline-block",
      borderRadius: "medium",
      borderWidth: "0",
      padding: "0",
      cursor: this.props.disabled ? 'not-allowed' : 'auto',
      as: "button",
      type: "button",
      "aria-label": screenReaderLabel,
      ...(selectOnClick ? {
        onClick: () => this.props.onSelect(color)
      } : {}),
      children: (0, _jsxRuntime.jsxs)("div", {
        children: [(0, _jsxRuntime.jsx)(_v.ColorIndicator, {
          color: color,
          shape: "rectangle",
          role: "presentation"
        }), this.isSelectedColor(color) && (0, _jsxRuntime.jsx)("div", {
          css: this.props?.styles?.selectedIndicator,
          children: (0, _jsxRuntime.jsx)(_IconCheckDarkSolid.IconCheckDarkSolid, {
            themeOverride: {
              sizeXSmall: '0.8rem'
            },
            size: "x-small"
          })
        })]
      })
    });
  };
  renderIndicatorTooltip = (child, color) => {
    return (0, _jsxRuntime.jsx)(_v11_3.Tooltip, {
      renderTip: (0, _jsxRuntime.jsx)("div", {
        children: color
      }),
      elementRef: element => {
        if (element && element.firstElementChild instanceof HTMLButtonElement) {
          // The tooltip and the button's aria-label has the same text content. This is redundant for screenreaders.
          // Aria-describedby is removed to bypass reading the tooltip twice
          element.firstElementChild.removeAttribute('aria-describedby');
        }
      },
      children: child
    });
  };
  renderSettingsMenu = (color, index) => (0, _jsxRuntime.jsx)(_v11_6.Drilldown, {
    onSelect: this.onMenuItemSelected(color),
    trigger: this.renderColorIndicator(color),
    rootPageId: "root",
    width: "10rem",
    offsetY: "15rem",
    children: (0, _jsxRuntime.jsxs)(_v11_6.Drilldown.Page, {
      withoutHeaderSeparator: true,
      id: "root",
      renderTitle: color,
      children: [(0, _jsxRuntime.jsx)(_v11_6.Drilldown.Option, {
        value: "select",
        id: "select",
        children: this.props.colorMixerSettings.selectColorLabel
      }), (0, _jsxRuntime.jsx)(_v11_6.Drilldown.Option, {
        value: "remove",
        id: "remove",
        children: this.props.colorMixerSettings.removeColorLabel
      })]
    })
  }, `color-preset-color-${index}`);
  render() {
    const {
      styles,
      label,
      colors
    } = this.props;
    if (!this.isModifiable && colors.length === 0) {
      // if there is no preset and no ability to add new,
      // there is no point of rendering the component
      return null;
    }
    return (0, _jsxRuntime.jsxs)("div", {
      ...(0, _omitProps.omitProps)(this.props, ColorPreset.allowedProps),
      ref: this.handleRef,
      css: styles?.colorPreset,
      "data-cid": "ColorPreset",
      children: [label && (0, _jsxRuntime.jsx)("div", {
        css: styles?.label,
        children: (0, _jsxRuntime.jsx)(_v11_5.Text, {
          weight: "bold",
          children: label
        })
      }), this.isModifiable && this.renderAddNewPresetButton(), colors.map((color, index) => {
        return this.isModifiable ? this.renderSettingsMenu(color, index) : (0, _jsxRuntime.jsx)("div", {
          children: this.renderColorIndicator(color, true)
        }, `color-preset-color-${index}`);
      })]
    });
  }
}) || _class);
var _default = exports.default = ColorPreset;
