// @instructure/ui-grid/lib/Grid/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Grid = void 0;
Object.defineProperty(exports, "GridCol", {
  enumerable: true,
  get: function () {
    return _v2.GridCol;
  }
});
Object.defineProperty(exports, "GridRow", {
  enumerable: true,
  get: function () {
    return _v.GridRow;
  }
});
exports.default = void 0;
var _react = require("react");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _pickProps = require("@instructure/ui-react-utils/lib/pickProps.js");
var _v = require("../../GridRow/v1");
var _v2 = require("../../GridCol/v1");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let Grid = exports.Grid = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class Grid extends _react.Component {
  static displayName = "Grid";
  static componentId = 'Grid';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    colSpacing: 'medium',
    rowSpacing: 'medium',
    hAlign: 'start',
    startAt: 'small',
    vAlign: 'top',
    visualDebug: false,
    children: null
  };
  static Row = _v.GridRow;
  static Col = _v2.GridCol;
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  renderChildren() {
    const {
      styles,
      makeStyles,
      ...props
    } = this.props;
    const children = _react.Children.toArray(this.props.children);
    return children.map((child, index) => {
      if ((0, _matchComponentTypes.matchComponentTypes)(child, [_v.GridRow])) {
        return (0, _safeCloneElement.safeCloneElement)(child, {
          ...(0, _pickProps.pickProps)(props, Grid.allowedProps),
          ...child.props /* child props should override parent */,
          isLastRow: index + 1 === children.length
        });
      } else {
        return child; // PropType validation should handle errors
      }
    });
  }
  render() {
    const {
      styles,
      ...restProps
    } = this.props;
    const props = (0, _omitProps.omitProps)(restProps, Grid.allowedProps);
    return (0, _jsxRuntime.jsx)("span", {
      ...props,
      css: styles?.grid,
      ref: this.handleRef,
      children: this.renderChildren()
    });
  }
}) || _class);
var _default = exports.default = Grid;
