// @instructure/ui-truncate-text/lib/TruncateText/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TruncateText = void 0;
var _react = require("react");
var _debounce = require("@instructure/debounce");
var _canUseDOM = require("@instructure/ui-dom-utils/lib/canUseDOM.js");
var _getBoundingClientRect = require("@instructure/ui-dom-utils/lib/getBoundingClientRect.js");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _ensureSingleChild = require("@instructure/ui-react-utils/lib/ensureSingleChild.js");
var _console = require("@instructure/console");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _truncate = _interopRequireDefault(require("./utils/truncate"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let TruncateText = exports.TruncateText = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class TruncateText extends _react.Component {
  static displayName = "TruncateText";
  static componentId = 'TruncateText';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    maxLines: 1,
    ellipsis: '\u2026',
    truncate: 'character',
    position: 'end',
    ignore: [' ', '.', ','],
    debounce: 0
  };
  ref = null;
  _text;
  _debounced;
  _stage = null;
  _wasTruncated;
  _resizeListener;
  _prevWidth = null;
  constructor(props) {
    super(props);
    this.state = this.initialState;
  }
  get _ref() {
    return this.ref;
  }
  get initialState() {
    return {
      isTruncated: false,
      needsSecondRender: true,
      truncatedElement: undefined,
      truncatedText: undefined
    };
  }
  componentDidMount() {
    const {
      children,
      makeStyles
    } = this.props;
    makeStyles?.();
    if (children) {
      this.checkChildren();
      const txt = (0, _ensureSingleChild.ensureSingleChild)(children);
      this._text = txt ? txt : undefined;
      this.truncate();
      this._debounced = (0, _debounce.debounce)(this.update, this.props.debounce, {
        leading: true,
        trailing: true
      });
      this._prevWidth = (0, _getBoundingClientRect.getBoundingClientRect)(this.ref)?.width;
      this._resizeListener = new ResizeObserver(entries => {
        // requestAnimationFrame call is needed becuase some truncatetext test cases
        // failed due to ResizeObserver was not able to deliver all observations within a single animation frame
        // see: https://stackoverflow.com/questions/49384120/resizeobserver-loop-limit-exceeded
        requestAnimationFrame(() => {
          for (const entry of entries) {
            const {
              width
            } = entry.contentRect;
            if (this._prevWidth !== width) {
              this._prevWidth = width;
              this.props.debounce === 0 ? this.update() : this._debounced();
            }
          }
        });
      });
      this._resizeListener.observe(this.ref);
    }
  }
  componentWillUnmount() {
    if (this._resizeListener) {
      this._resizeListener.disconnect();
    }
    if (this._debounced) {
      this._debounced.cancel();
    }
  }
  shallowCompare(obj1, obj2) {
    const keys1 = Object.keys(obj1);
    const keys2 = Object.keys(obj2);
    if (keys1.length !== keys2.length) {
      return false;
    }
    for (const key of keys1) {
      if (obj1[key] !== obj2[key]) {
        return false;
      }
    }
    return true;
  }
  componentDidUpdate(prevProps) {
    const {
      children,
      onUpdate,
      makeStyles
    } = this.props;
    makeStyles?.();
    const {
      isTruncated,
      needsSecondRender,
      truncatedText
    } = this.state;
    if (children) {
      // for some reason in React 19 prevPros and this.props are a different
      // object even if their contents are the same, so we cannot use !==
      if (!this.shallowCompare(prevProps, this.props)) {
        if (prevProps.children !== this.props.children) {
          // reset internal text variable if children change
          this.checkChildren();
          const txt = (0, _ensureSingleChild.ensureSingleChild)(children);
          this._text = txt ? txt : undefined;
        }
        // require the double render whenever props change
        this.setState(this.initialState);
        return;
      }
      if (!needsSecondRender && (isTruncated || this._wasTruncated)) {
        onUpdate?.(isTruncated, truncatedText);
        this._wasTruncated = isTruncated;
      } else {
        this.truncate();
      }
    }
  }
  checkChildren() {
    (0, _console.logError)(!(() => {
      let isTooDeep = false;
      const text = (0, _ensureSingleChild.ensureSingleChild)(this.props.children);
      _react.Children.forEach(text.props.children, child => {
        if (child.props) {
          _react.Children.forEach(child.props.children, grandChild => {
            // currently we don't support node trees deeper than 2 levels
            // truncation will still occur on their text content, but their actual node structure will be lost
            if (grandChild.props) {
              isTooDeep = true;
            }
          });
        }
      });
      return isTooDeep;
    })(), `[TruncateText] Some children are too deep in the node tree and will not render.`);
  }
  update = () => {
    if (this.ref) {
      this.setState(this.initialState);
    }
  };
  truncate() {
    if (!this.state.needsSecondRender) {
      return;
    }
    if (_canUseDOM.canUseDOM) {
      const result = (0, _truncate.default)(this._stage, {
        ...this.props,
        parent: this.ref ? this.ref : undefined,
        lineHeight: this.props.styles?.lineHeight
      });
      if (result) {
        const element = this.renderChildren(result.isTruncated, result.data, result.constraints.width);
        this.setState({
          needsSecondRender: false,
          isTruncated: result.isTruncated,
          truncatedElement: element,
          truncatedText: result.text
        });
      }
    } else {
      const textContent = this.ref?.textContent ? this.ref?.textContent : undefined;
      // if dom isn't available, use original children
      this.setState({
        needsSecondRender: false,
        isTruncated: false,
        truncatedElement: this._text,
        truncatedText: textContent
      });
    }
  }
  renderChildren(truncated, data, width) {
    if (!truncated) {
      return this._text;
    }
    const childElements = [];
    // iterate over each node used in the truncated string
    for (let i = 0; i < data.length; i++) {
      const item = data[i];
      const element = this._text.props.children[i];
      const nodeText = item.join('');
      if (element && element.props) {
        // if node is an html element and not just a string
        childElements.push((0, _safeCloneElement.safeCloneElement)(element, element.props, nodeText));
      } else {
        childElements.push(nodeText);
      }
    }
    // this spacer element is set to the max width the full text could
    // potentially be without this, text in `width: auto` elements won't expand
    // to accommodate more text, once truncated
    // Breadcrumb is modifying this element's display to inline to prevent layout issues
    // TODO: find a better way to handle this
    childElements.push((0, _jsxRuntime.jsx)("span", {
      css: this.props.styles?.spacer,
      style: {
        width: width || undefined
      }
    }, "spacer"));
    const children = _react.Children.map(childElements, child => child);
    return this._text.props ? (0, _safeCloneElement.safeCloneElement)(this._text, this._text.props, children) : children;
  }
  render() {
    const {
      truncatedElement
    } = this.state;
    const {
      children
    } = this.props;
    return (0, _jsxRuntime.jsxs)("span", {
      "data-cid": "TruncateText",
      css: this.props.styles?.truncateText,
      ref: el => {
        this.ref = el;
      },
      children: [children && (truncatedElement ? null : (0, _jsxRuntime.jsx)("span", {
        ref: el => {
          this._stage = el;
        },
        children: (0, _ensureSingleChild.ensureSingleChild)(children)
      })), truncatedElement]
    });
  }
}) || _class);
var _default = exports.default = TruncateText;
