// @instructure/ui-form-field/es/FormFieldMessages/v2/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { omitProps } from '@instructure/ui-react-utils';
import { withStyleNew } from '@instructure/emotion';
import { FormFieldMessage } from "../../FormFieldMessage/v2/index.js";
import generateStyle from "./styles.js";
import { allowedProps } from "./props.js";
import { jsx as _jsx } from "@emotion/react/jsx-runtime";
/**
---
parent: FormField
---

A FormFieldMessages component

```js
---
type: example
---
<FormFieldMessages messages={[
  { text: 'Invalid name', type: 'error' },
  { text: 'Good job!', type: 'success' },
  { text: 'Full name, first and last', type: 'hint' },
]} />
```
**/
let FormFieldMessages = (_dec = withStyleNew(generateStyle), _dec(_class = class FormFieldMessages extends Component {
  static displayName = "FormFieldMessages";
  static componentId = 'FormFieldMessages';
  static allowedProps = allowedProps;
  static defaultProps = {};
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  render() {
    const {
      messages,
      styles
    } = this.props;
    return messages && messages.length > 0 ? _jsx("div", {
      css: styles?.formFieldMessages,
      ...omitProps(this.props, FormFieldMessages.allowedProps),
      ref: this.handleRef,
      children: messages.map((msg, i) => {
        return _jsx("span", {
          css: styles?.message,
          children: _jsx(FormFieldMessage, {
            variant: msg.type,
            children: msg.text
          })
        }, `error${i}`);
      })
    }) : null;
  }
}) || _class);
export default FormFieldMessages;
export { FormFieldMessages };
