// @instructure/ui-progress/lib/ProgressBar/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ProgressBar = void 0;
var _react = require("react");
var _v11_ = require("@instructure/ui-view/v11_6");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let ProgressBar = exports.ProgressBar = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class ProgressBar extends _react.Component {
  static displayName = "ProgressBar";
  static componentId = 'ProgressBar';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    formatScreenReaderValue: ({
      valueNow,
      valueMax
    }) => `${valueNow} / ${valueMax}`,
    size: 'medium',
    valueMax: 100,
    valueNow: 0,
    as: 'div',
    color: 'primary',
    shouldAnimate: false,
    // default to showing `success` color on completion
    meterColor: ({
      valueNow,
      valueMax
    }) => valueNow / valueMax >= 1 ? 'success' : 'brand'
  };
  ref = null;
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  render() {
    const {
      renderValue,
      formatScreenReaderValue,
      valueNow,
      valueMax,
      screenReaderLabel,
      size,
      color,
      meterColor,
      renderValueInside,
      styles,
      ...props
    } = this.props;
    const valueText = typeof formatScreenReaderValue === 'function' ? formatScreenReaderValue({
      valueNow: valueNow,
      valueMax: valueMax
    }) : undefined;
    const value = (0, _callRenderProp.callRenderProp)(renderValue, {
      valueNow: valueNow,
      valueMax: valueMax
    });
    return (0, _jsxRuntime.jsxs)(_v11_.View, {
      ...(0, _passthroughProps.passthroughProps)(props),
      as: this.props.as,
      css: styles?.progressBar,
      margin: this.props.margin,
      elementRef: this.handleRef,
      "data-cid": "ProgressBar",
      children: [(0, _jsxRuntime.jsxs)("span", {
        css: styles?.trackLayout,
        children: [(0, _jsxRuntime.jsx)("progress", {
          css: styles?.htmlProgress,
          max: valueMax,
          value: valueNow,
          "aria-valuetext": valueText,
          "aria-label": screenReaderLabel
        }), (0, _jsxRuntime.jsx)("span", {
          css: styles?.track,
          role: "presentation",
          "aria-hidden": "true",
          children: (0, _jsxRuntime.jsx)("span", {
            css: styles?.trackValue,
            children: renderValueInside && value
          })
        })]
      }), value && !renderValueInside && (0, _jsxRuntime.jsx)("span", {
        css: styles?.value,
        "aria-hidden": "true",
        children: value
      })]
    });
  }
}) || _class);
var _default = exports.default = ProgressBar;
