// @instructure/ui-byline/es/Byline/v2/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { omitProps } from '@instructure/ui-react-utils';
import { View } from '@instructure/ui-view/latest';
import { withStyleNew } from '@instructure/emotion';
import generateStyle from "./styles.js";
import { allowedProps } from "./props.js";
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
/**
---
category: components
---
**/
let Byline = (_dec = withStyleNew(generateStyle), _dec(_class = class Byline extends Component {
  static displayName = "Byline";
  static componentId = 'Byline';
  static allowedProps = allowedProps;
  static defaultProps = {
    alignContent: 'center'
  };
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  render() {
    const passthroughProps = View.omitViewProps(omitProps(this.props, Byline.allowedProps), Byline);
    return _jsxs(View, {
      ...passthroughProps,
      elementRef: this.handleRef,
      css: this.props.styles?.byline,
      margin: this.props.margin,
      maxWidth: this.props.styles?.maxWidth,
      children: [_jsx("div", {
        css: this.props.styles?.figure,
        children: this.props.children
      }), _jsxs("div", {
        css: this.props.styles?.caption,
        children: [this.props.title && _jsx("span", {
          css: this.props.styles?.title,
          children: this.props.title
        }), this.props.description && _jsx("div", {
          css: this.props.styles?.description,
          children: this.props.description
        })]
      })]
    });
  }
}) || _class);
export default Byline;
export { Byline };
