// @instructure/ui-themes/src/themes/newThemeTokens/componentTypes/avatar.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { TokenBoxshadowValueInst } from "../commonTypes"

export type Avatar = {
  backgroundColor: string
  borderColor: string
  borderWidthSm: string
  borderWidthMd: string
  /** Do not use this for other components */
  boxShadow: TokenBoxshadowValueInst
  fontWeight: string | number
  blueBackgroundColor: string
  blueTextColor: string
  greenBackgroundColor: string
  greenTextColor: string
  redBackgroundColor: string
  redTextColor: string
  orangeBackgroundColor: string
  orangeTextColor: string
  greyBackgroundColor: string
  greyTextColor: string
  ashBackgroundColor: string
  ashTextColor: string
  aiBottomGradientColor: string
  aiTopGradientColor: string
  textOnColor: string
  size2xs: string
  sizeXs: string
  sizeSm: string
  sizeMd: string
  sizeLg: string
  sizeXl: string
  size2xl: string
  fontSize2xs: string
  fontSizeXs: string
  fontSizeSm: string
  fontSizeMd: string
  fontSizeLg: string
  fontSizeXl: string
  fontSize2xl: string
  fontFamily: string
  rectangleRadius: string
}

export default Avatar
