// @instructure/ui-themes/src/themes/newThemeTokens/componentTypes/baseButton.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

export type BaseButton = {
  primaryActiveBackground: string
  primaryActiveBorderColor: string
  primaryActiveTextColor: string
  primaryBackground: string
  primaryBorderColor: string
  primaryColor: string
  primaryDisabledBackgroundColor: string
  primaryDisabledBorderColor: string
  primaryDisabledTextColor: string
  primaryGhostActiveBackground: string
  primaryGhostBorderColor: string
  primaryGhostColor: string
  primaryGhostHoverBackground: string
  primaryHoverBackground: string
  primaryHoverBorderColor: string
  primaryHoverTextColor: string
  primaryInverseActiveBackground: string
  primaryInverseBackground: string
  primaryInverseBorderColor: string
  primaryInverseColor: string
  primaryInverseGhostActiveBackground: string
  primaryInverseGhostBorderColor: string
  primaryInverseGhostColor: string
  primaryInverseGhostHoverBackground: string
  primaryInverseHoverBackground: string
  primaryOnColorActiveBorderColor: string
  primaryOnColorActiveTextColor: string
  primaryOnColorDisabledBackgroundColor: string
  primaryOnColorDisabledBorderColor: string
  primaryOnColorDisabledTextColor: string
  primaryOnColorHoverBorderColor: string
  primaryOnColorHoverTextColor: string
  secondaryActiveBackground: string
  secondaryActiveBorderColor: string
  secondaryActiveTextColor: string
  secondaryBackground: string
  secondaryBorderColor: string
  secondaryColor: string
  secondaryDisabledBackgroundColor: string
  secondaryDisabledBorderColor: string
  secondaryDisabledTextColor: string
  /** This token is not used in design, kept only for backward compatibility reasons. */
  secondaryGhostBorderColor: string
  /** This token is not used in design, kept only for backward compatibility reasons. */
  secondaryGhostColor: string
  /** This token is not used in design, kept only for backward compatibility reasons. */
  secondaryGhostHoverBackground: string
  secondaryHoverBackground: string
  secondaryHoverBorderColor: string
  secondaryHoverTextColor: string
  secondaryOnColorActiveBorderColor: string
  secondaryOnColorActiveTextColor: string
  secondaryOnColorDisabledBorderColor: string
  secondaryOnColorDisabledTextColor: string
  secondaryOnColorHoverBorderColor: string
  secondaryOnColorHoverTextColor: string
  successActiveBackground: string
  successActiveBorderColor: string
  successActiveTextColor: string
  successBackground: string
  successBorderColor: string
  successColor: string
  successDisabledBackgroundColor: string
  successDisabledBorderColor: string
  successDisabledTextColor: string
  successGhostActiveBackground: string
  successGhostBorderColor: string
  successGhostColor: string
  successGhostHoverBackground: string
  successHoverBackground: string
  successHoverBorderColor: string
  successHoverTextColor: string
  successSecondaryActiveBorderColor: string
  successSecondaryActiveTextColor: string
  successSecondaryDisabledBorderColor: string
  successSecondaryDisabledTextColor: string
  successSecondaryHoverBorderColor: string
  successSecondaryHoverTextColor: string
  dangerActiveBackground: string
  dangerBorderColor: string
  dangerGhostActiveBackground: string
  dangerGhostBorderColor: string
  dangerGhostColor: string
  dangerGhostHoverBackground: string
  dangerHoverBackground: string
  destructiveActiveBorderColor: string
  destructiveActiveTextColor: string
  dangerBackground: string
  dangerColor: string
  destructiveDisabledBackgroundColor: string
  destructiveDisabledBorderColor: string
  destructiveDisabledTextColor: string
  destructiveHoverBorderColor: string
  destructiveHoverTextColor: string
  destructiveSecondaryActiveBorderColor: string
  destructiveSecondaryActiveTextColor: string
  destructiveSecondaryDisabledBorderColor: string
  destructiveSecondaryDisabledTextColor: string
  destructiveSecondaryHoverBorderColor: string
  destructiveSecondaryHoverTextColor: string
  aiActiveBackgroundBottomGradientColor: string
  aiActiveBackgroundTopGradientColor: string
  aiActiveBorderBottomGradientColor: string
  aiActiveBorderTopGradientColor: string
  aiActiveTextColor: string
  aiBackgroundBottomGradientColor: string
  aiBackgroundTopGradientColor: string
  aiBaseTextColor: string
  aiBorderBottomGradientColor: string
  aiBorderTopGradientColor: string
  aiDisabledBackgroundColor: string
  aiDisabledBorderColor: string
  aiDisabledTextColor: string
  aiHoverBackgroundBottomGradientColor: string
  aiHoverBackgroundTopGradientColor: string
  aiHoverBorderBottomGradientColor: string
  aiHoverBorderTopGradientColor: string
  aiHoverTextColor: string
  aiSecondaryActiveBackgroundBottomGradientColor: string
  aiSecondaryActiveBackgroundTopGradientColor: string
  aiSecondaryDisabledBorderColor: string
  aiSecondaryDisabledTextColor: string
  aiSecondaryHoverBackgroundBottomGradientColor: string
  aiSecondaryHoverBackgroundTopGradientColor: string
  aiSecondaryTextBottomGradientColor: string
  aiSecondaryTextTopGradientColor: string
  smallFontSize: string
  mediumFontSize: string
  largeFontSize: string
  largeHeight: string
  largePaddingHorizontal: string
  mediumHeight: string
  mediumPaddingHorizontal: string
  smallHeight: string
  smallPaddingHorizontal: string
  borderRadius: string
  borderRadiusFull: string
  borderRadiusSm: string
  borderWidth: string
  fontFamily: string
  fontWeight: string | number
  gapButtonContentLg: string
  gapButtonContentMd: string
  gapButtonContentSm: string
  heightXs: string
  heightXxs: string
  opacityBase: string | number
  opacityDisabled: string | number
  tertiaryActiveBorderColor: string
  tertiaryActiveTextColor: string
  tertiaryDisabledBorderColor: string
  tertiaryDisabledTextColor: string
  tertiaryHoverBorderColor: string
  tertiaryHoverTextColor: string
  dangerBoxShadow: string
  dangerGhostBackground: string
  dangerGhostBoxShadow: string
  dangerGhostHoverBoxShadow: string
  dangerHoverBoxShadow: string
  hoverTransform: string
  letterSpacing: string
  primaryBoxShadow: string
  primaryGhostBackground: string
  primaryHoverBoxShadow: string
  primaryInverseBoxShadow: string
  primaryInverseGhostBackground: string
  primaryInverseGhostBoxShadow: string
  primaryInverseGhostHoverBoxShadow: string
  primaryInverseHoverBoxShadow: string
  secondaryBoxShadow: string
  secondaryGhostActiveBackground: string
  secondaryGhostBackground: string
  secondaryGhostBoxShadow: string
  secondaryGhostHoverBoxShadow: string
  secondaryHoverBoxShadow: string
  successBoxShadow: string
  successGhostBackground: string
  successGhostBoxShadow: string
  successGhostHoverBoxShadow: string
  successHoverBoxShadow: string
  textTransform: string
  transform: string
  borderStyle: string | number
  paddingVertical: string
  lineHeight: string | number
}

export default BaseButton
