// @instructure/ui-themes/src/themes/newThemeTokens/componentTypes/heading.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { TokenTypographyValueInst } from "../commonTypes"

export type Heading = {
  h1FontSize: string
  h2FontSize: string
  h3FontSize: string
  h4FontSize: string
  h5FontSize: string
  h6FontSize: string
  h1FontWeight: string | number
  h2FontWeight: string | number
  h3FontWeight: string | number
  h4FontWeight: string | number
  h5FontWeight: string | number
  h6FontWeight: string | number
  h1FontFamily: string
  h2FontFamily: string
  h3FontFamily: string
  h4FontFamily: string
  h5FontFamily: string
  h6FontFamily: string
  lineHeight: string | number
  borderWidth: string
  baseColor: string
  baseOnColor: string
  mutedColor: string
  mutedOnColor: string
  inverseColor: string
  aiTextTopGradientColor: string
  aiTextBottomGradientColor: string
  borderColor: string
  borderStyle: string
  borderPadding: string
  titlePageDesktop: TokenTypographyValueInst
  titlePageMobile: TokenTypographyValueInst
  titleSection: TokenTypographyValueInst
  titleCardSection: TokenTypographyValueInst
  titleModule: TokenTypographyValueInst
  titleCardLarge: TokenTypographyValueInst
  titleCardRegular: TokenTypographyValueInst
  titleCardMini: TokenTypographyValueInst
  label: TokenTypographyValueInst
  labelInline: TokenTypographyValueInst
  gapIconSm: string
  gapIconMd: string
  gapIconLg: string
}

export default Heading
