// @instructure/ui-themes/src/themes/newThemeTokens/componentTypes/index.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type AiInformation from "./aiInformation"
import type Alert from "./alert"
import type AppNav from "./appNav"
import type AppNavItem from "./appNavItem"
import type Avatar from "./avatar"
import type Badge from "./badge"
import type BaseButton from "./baseButton"
import type Billboard from "./billboard"
import type Breadcrumb from "./breadcrumb"
import type Byline from "./byline"
import type Calendar from "./calendar"
import type CalendarDay from "./calendarDay"
import type Checkbox from "./checkbox"
import type ColorContrast from "./colorContrast"
import type ColorIndicator from "./colorIndicator"
import type ColorPicker from "./colorPicker"
import type ColorPreset from "./colorPreset"
import type ContextView from "./contextView"
import type DataPermissionLevels from "./dataPermissionLevels"
import type DrawerLayoutContent from "./drawerLayoutContent"
import type DrawerLayoutTray from "./drawerLayoutTray"
import type Drilldown from "./drilldown"
import type FileDrop from "./fileDrop"
import type Flex from "./flex"
import type FormFieldLayout from "./formFieldLayout"
import type FormFieldMessage from "./formFieldMessage"
import type Grid from "./grid"
import type GridCol from "./gridCol"
import type GridRow from "./gridRow"
import type Heading from "./heading"
import type Icon from "./icon"
import type Img from "./img"
import type InlineListItem from "./inlineListItem"
import type Link from "./link"
import type List from "./list"
import type ListItem from "./listItem"
import type Mask from "./mask"
import type Menu from "./menu"
import type MenuGroup from "./menuGroup"
import type MenuItem from "./menuItem"
import type MenuSeparator from "./menuSeparator"
import type Metric from "./metric"
import type Modal from "./modal"
import type ModalBody from "./modalBody"
import type ModalFooter from "./modalFooter"
import type ModalHeader from "./modalHeader"
import type NutritionFacts from "./nutritionFacts"
import type Options from "./options"
import type OptionsItem from "./optionsItem"
import type OptionsSeparator from "./optionsSeparator"
import type Pagination from "./pagination"
import type PaginationPageInput from "./paginationPageInput"
import type Palette from "./palette"
import type Pill from "./pill"
import type Popover from "./popover"
import type ProgressBar from "./progressBar"
import type ProgressCircle from "./progressCircle"
import type RadioInput from "./radioInput"
import type RangeInput from "./rangeInput"
import type RatingIcon from "./ratingIcon"
import type RgbaInput from "./rgbaInput"
import type Select from "./select"
import type SideNavBar from "./sideNavBar"
import type SideNavBarItem from "./sideNavBarItem"
import type SideNavBarNew from "./sideNavBarNew"
import type Slider from "./slider"
import type SourceCodeEditor from "./sourceCodeEditor"
import type Spinner from "./spinner"
import type Table from "./table"
import type TableBody from "./tableBody"
import type TableCell from "./tableCell"
import type TableColHeader from "./tableColHeader"
import type TableHead from "./tableHead"
import type TableRow from "./tableRow"
import type TableRowHeader from "./tableRowHeader"
import type Tabs from "./tabs"
import type TabsPanel from "./tabsPanel"
import type TabsTab from "./tabsTab"
import type Tag from "./tag"
import type Text from "./text"
import type TextArea from "./textArea"
import type TextInput from "./textInput"
import type Toggle from "./toggle"
import type ToggleDetails from "./toggleDetails"
import type ToggleGroup from "./toggleGroup"
import type Tooltip from "./tooltip"
import type TopNavBarActionItems from "./topNavBarActionItems"
import type TopNavBarBrand from "./topNavBarBrand"
import type TopNavBarItem from "./topNavBarItem"
import type TopNavBarLayout from "./topNavBarLayout"
import type TopNavBarMenuItems from "./topNavBarMenuItems"
import type Tray from "./tray"
import type TreeBrowser from "./treeBrowser"
import type TreeBrowserTreeButton from "./treeBrowserTreeButton"
import type TreeBrowserTreeCollection from "./treeBrowserTreeCollection"
import type TruncateText from "./truncateText"
import type View from "./view"

type ComponentTypes = {
  AiInformation: (semantics: any) => AiInformation
  Alert: (semantics: any) => Alert
  AppNav: (semantics: any) => AppNav
  AppNavItem: (semantics: any) => AppNavItem
  Avatar: (semantics: any) => Avatar
  Badge: (semantics: any) => Badge
  BaseButton: (semantics: any) => BaseButton
  Billboard: (semantics: any) => Billboard
  Breadcrumb: (semantics: any) => Breadcrumb
  Byline: (semantics: any) => Byline
  Calendar: (semantics: any) => Calendar
  CalendarDay: (semantics: any) => CalendarDay
  Checkbox: (semantics: any) => Checkbox
  ColorContrast: (semantics: any) => ColorContrast
  ColorIndicator: (semantics: any) => ColorIndicator
  ColorPicker: (semantics: any) => ColorPicker
  ColorPreset: (semantics: any) => ColorPreset
  ContextView: (semantics: any) => ContextView
  DataPermissionLevels: (semantics: any) => DataPermissionLevels
  DrawerLayoutContent: (semantics: any) => DrawerLayoutContent
  DrawerLayoutTray: (semantics: any) => DrawerLayoutTray
  Drilldown: (semantics: any) => Drilldown
  FileDrop: (semantics: any) => FileDrop
  Flex: (semantics: any) => Flex
  FormFieldLayout: (semantics: any) => FormFieldLayout
  FormFieldMessage: (semantics: any) => FormFieldMessage
  Grid: (semantics: any) => Grid
  GridCol: (semantics: any) => GridCol
  GridRow: (semantics: any) => GridRow
  Heading: (semantics: any) => Heading
  Icon: (semantics: any) => Icon
  Img: (semantics: any) => Img
  InlineListItem: (semantics: any) => InlineListItem
  Link: (semantics: any) => Link
  List: (semantics: any) => List
  ListItem: (semantics: any) => ListItem
  Mask: (semantics: any) => Mask
  Menu: (semantics: any) => Menu
  MenuGroup: (semantics: any) => MenuGroup
  MenuItem: (semantics: any) => MenuItem
  MenuSeparator: (semantics: any) => MenuSeparator
  Metric: (semantics: any) => Metric
  Modal: (semantics: any) => Modal
  ModalBody: (semantics: any) => ModalBody
  ModalFooter: (semantics: any) => ModalFooter
  ModalHeader: (semantics: any) => ModalHeader
  NutritionFacts: (semantics: any) => NutritionFacts
  Options: (semantics: any) => Options
  OptionsItem: (semantics: any) => OptionsItem
  OptionsSeparator: (semantics: any) => OptionsSeparator
  Pagination: (semantics: any) => Pagination
  PaginationPageInput: (semantics: any) => PaginationPageInput
  Palette: (semantics: any) => Palette
  Pill: (semantics: any) => Pill
  Popover: (semantics: any) => Popover
  ProgressBar: (semantics: any) => ProgressBar
  ProgressCircle: (semantics: any) => ProgressCircle
  RadioInput: (semantics: any) => RadioInput
  RangeInput: (semantics: any) => RangeInput
  RatingIcon: (semantics: any) => RatingIcon
  RgbaInput: (semantics: any) => RgbaInput
  Select: (semantics: any) => Select
  SideNavBar: (semantics: any) => SideNavBar
  SideNavBarItem: (semantics: any) => SideNavBarItem
  SideNavBarNew: (semantics: any) => SideNavBarNew
  Slider: (semantics: any) => Slider
  SourceCodeEditor: (semantics: any) => SourceCodeEditor
  Spinner: (semantics: any) => Spinner
  Table: (semantics: any) => Table
  TableBody: (semantics: any) => TableBody
  TableCell: (semantics: any) => TableCell
  TableColHeader: (semantics: any) => TableColHeader
  TableHead: (semantics: any) => TableHead
  TableRow: (semantics: any) => TableRow
  TableRowHeader: (semantics: any) => TableRowHeader
  Tabs: (semantics: any) => Tabs
  TabsPanel: (semantics: any) => TabsPanel
  TabsTab: (semantics: any) => TabsTab
  Tag: (semantics: any) => Tag
  Text: (semantics: any) => Text
  TextArea: (semantics: any) => TextArea
  TextInput: (semantics: any) => TextInput
  Toggle: (semantics: any) => Toggle
  ToggleDetails: (semantics: any) => ToggleDetails
  ToggleGroup: (semantics: any) => ToggleGroup
  Tooltip: (semantics: any) => Tooltip
  TopNavBarActionItems: (semantics: any) => TopNavBarActionItems
  TopNavBarBrand: (semantics: any) => TopNavBarBrand
  TopNavBarItem: (semantics: any) => TopNavBarItem
  TopNavBarLayout: (semantics: any) => TopNavBarLayout
  TopNavBarMenuItems: (semantics: any) => TopNavBarMenuItems
  Tray: (semantics: any) => Tray
  TreeBrowser: (semantics: any) => TreeBrowser
  TreeBrowserTreeButton: (semantics: any) => TreeBrowserTreeButton
  TreeBrowserTreeCollection: (semantics: any) => TreeBrowserTreeCollection
  TruncateText: (semantics: any) => TruncateText
  View: (semantics: any) => View
}

export type { ComponentTypes }
export default ComponentTypes
