// @instructure/ui-themes/src/themes/newThemeTokens/componentTypes/radioInput.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { TokenBoxshadowValueInst } from "../commonTypes"

export type RadioInput = {
  backgroundColor: string
  backgroundHoverColor: string
  backgroundReadonlyColor: string
  backgroundDisabledColor: string
  borderColor: string
  borderHoverColor: string
  borderReadonlyColor: string
  borderDisabledColor: string
  borderSelectedColor: string
  fontSizeSm: string
  fontSizeMd: string
  fontSizeLg: string
  toggleSmallFontSize: string
  toggleMediumFontSize: string
  toggleLargeFontSize: string
  lineHeightSm: string | number
  lineHeightMd: string | number
  lineHeightLg: string | number
  fontFamily: string
  fontWeight: string | number
  gap: string
  borderWidth: string
  labelBaseColor: string
  labelHoverColor: string
  labelDisabledColor: string
  labelReadonlyColor: string
  controlSizeSm: string
  controlSizeMd: string
  controlSizeLg: string
  checkedInsetSm: string
  checkedInsetMd: string
  checkedInsetLg: string
  toggleSmallHeight: string
  toggleMediumHeight: string
  toggleLargeHeight: string
  controlVerticalMargin: string
  toggleBackgroundDanger: string
  toggleBackgroundOff: string
  toggleBackgroundSuccess: string
  toggleBackgroundWarning: string
  toggleHandleText: string
  toggleBorderRadius: string
  toggleBorderWidth: string
  /** Do not use this for other components */
  toggleShadow: { "0": TokenBoxshadowValueInst; "1": TokenBoxshadowValueInst }
}

export default RadioInput
