// @instructure/ui-themes/src/themes/newThemeTokens/componentTypes/rangeInput.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { TokenBoxshadowValueInst } from "../commonTypes"

export type RangeInput = {
  handleBackground: string
  handleHoverBackground: string
  handleFocusBackground: string
  /** Deprecated, remove with "deprecated" thumbVariant */
  handleFocusOutlineColor: string
  /** Deprecated, remove with "deprecated" thumbVariant
   */
  handleShadowColor: string
  handleBorderColor: string
  handleBorderSize: string
  handleFocusInset: string
  trackBackground: string
  valueFontFamily: string
  valueFontWeight: string | number
  valueSmallFontSize: string
  valueMediumFontSize: string
  valueLargeFontSize: string
  valueSmallPadding: string
  valueMediumPadding: string
  valueLargePadding: string
  valueSmallLineHeight: string | number
  valueMediumLineHeight: string | number
  valueLargeLineHeight: string | number
  trackBorderColor: string
  minWidth: string
  handleSize: string
  /** Do not use this for other components */
  boxShadow: { "0": TokenBoxshadowValueInst; "1": TokenBoxshadowValueInst }
  /** Deprecated, remove with "deprecated" thumbVariant */
  handleFocusOutlineWidth: string
}

export default RangeInput
