// @instructure/ui-themes/src/themes/newThemeTokens/componentTypes/sideNavBarNew.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { TokenBoxshadowValueInst } from "../commonTypes"

export type SideNavBarNew = {
  backgroundColor: string
  fill: string
  borderColor: string
  textColor: string
  minimizedminWidth: string
  toggleTransition: string | number
  minWidth: string
  margin: string
  minimizedMargin: string
  mainNavMargin: string
  mainNavMinimizedMargin: string
  mainNavGap: string
  navItemsGap: string
  /** Do not use this for other components */
  boxShadow: { "0": TokenBoxshadowValueInst; "1": TokenBoxshadowValueInst }
  borderRadius: string
  minimizedBorderRadius: string
  borderWidth: string
  item: {
    horizontalPadding: string
    verticalPadding: string
    iconWrapperHeight: string
    borderRadius: string
    backgroundColor: string
    fontColor: string
    fontFamily: string
    fontSize: string
    fontWeight: string | number
    innerFocusOutlineColor: string
    outerFocusOutlineColor: string
    hoverBackgroundColor: string
    lineHeight: string | number
    linkTextDecoration: string | number
    currentBackgroundColor: string
    selectedBackgroundColor: string
    currentTextColor: string
    selectedAccountBackgroundColor: string
    selectedAccountFontColor: string
    selectedInnerFocusOutlineColor: string
    selectedOuterFocusOutlineColor: string
  }
  drilldownOption: {
    backgroundColor: string
    hoverBackgroundColor: string
    activeBackgroundColor: string
    selectedBackgroundColor: string
    textColor: string
    selectedTextColor: string
    logoutTextColor: string
    borderRadius: string
  }
  button: { textColor: string }
  subNav: { strokeColor: string; textColor: string; padding: string; gap: string }
  institutionLogo: { textColor: string; strokeColor: string }
  insitutionLogo: { padding: string; gap: string }
}

export default SideNavBarNew
