// @instructure/ui-themes/src/themes/newThemeTokens/componentTypes/sideNavBar.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { TokenBoxshadowValueInst } from "../commonTypes"

export type SideNavBar = {
  backgroundColor: string
  fill: string
  fontColor: string
  minimizedWidth: string
  toggleTransition: string | number
  width: string
  margin: string
  minimizedMargin: string
  contentMargin: string
  contentGap: string
  /** Do not use this for other components */
  boxShadow: { "0": TokenBoxshadowValueInst; "1": TokenBoxshadowValueInst }
  borderRadius: string
  minimizedBorderRadius: string
}

export default SideNavBar
