// @instructure/ui-themes/src/themes/newThemeTokens/componentTypes/text.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { TokenTypographyValueInst } from "../commonTypes"

export type Text = {
  /** Used for introductory or overview text that explains the purpose or context of an entire page. */
  descriptionPage: TokenTypographyValueInst
  /** Used for text that introduces or summarizes a specific section within a page. */
  descriptionSection: TokenTypographyValueInst
  /** Used for standard descriptive body text that provides detailed information. */
  content: TokenTypographyValueInst
  /** Used to highlight key statements or essential information within descriptive text. */
  contentImportant: TokenTypographyValueInst
  /** Used to style quoted or referenced text distinct from the main content. */
  contentQuote: TokenTypographyValueInst
  /** Used for fine-print or secondary descriptions such as notes or disclaimers. */
  contentSmall: TokenTypographyValueInst
  /** Used for short explanatory text accompanying data visualizations, icons, or UI elements. */
  legend: TokenTypographyValueInst
  fontSizeXSmall: string
  fontSizeSmall: string
  fontSizeMedium: string
  fontSizeLarge: string
  fontSizeXLarge: string
  fontSizeXXLarge: string
  fontWeightLight: string | number
  fontWeightNormal: string | number
  fontWeightBold: string | number
  fontWeightImportant: string | number
  fontWeightRegular: string | number
  lineHeight: string | number
  lineHeight100: string | number
  lineHeight125: string | number
  lineHeight150: string | number
  lineHeightCondensed: string | number
  lineHeightDouble: string | number
  lineHeightFit: string | number
  paragraphMargin: string
  letterSpacingCondensed: string
  letterSpacingExpanded: string
  letterSpacingNormal: string
  aiBackgroundColor: string
  aiColor: string
  primaryColor: string
  errorColor: string
  baseColor: string
  baseOnColor: string
  inverseColor: string
  mutedColor: string
  mutedOnColor: string
  successColor: string
  warningColor: string
  fontFamily: string
  fontFamilyMonospace: string
}

export default Text
