// @instructure/ui-themes/src/themes/newThemeTokens/componentTypes/toggle.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { TokenBoxshadowValueInst } from "../commonTypes"

export type Toggle = {
  backgroundColor: string
  backgroundHoverColor: string
  backgroundDisabledColor: string
  backgroundReadonlyColor: string
  errorBackgroundColor: string
  errorBackgroundHoverColor: string
  borderColor: string
  borderHoverColor: string
  borderDisabledColor: string
  borderReadonlyColor: string
  borderWidth: string
  borderRadius: string
  errorBorderColor: string
  checkedBorderColor: string
  checkedBorderHoverColor: string
  checkedBorderDisabledColor: string
  checkedBorderReadonlyColor: string
  checkedBackgroundColor: string
  checkedBackgroundHoverColor: string
  checkedBackgroundDisabledColor: string
  checkedBackgroundReadonlyColor: string
  labelColor: string
  labelDisabledColor: string
  uncheckedIconBorderColor: string
  uncheckedIconErrorBorderColor: string
  uncheckedIconBorderHoverColor: string
  uncheckedIconBorderDisabledColor: string
  uncheckedIconBorderReadonlyColor: string
  checkedIconBorderColor: string
  checkedIconBorderHoverColor: string
  checkedIconBorderDisabledColor: string
  marginStart: string
  marginEnd: string
  marginVertical: string
  toggleSize: string
  labelFontFamily: string
  labelFontWeight: string | number
  labelLineHeightSm: string | number
  labelLineHeightMd: string | number
  labelLineHeightLg: string | number
  labelFontSizeMd: string
  labelFontSizeLg: string
  labelFontSizeSm: string
  checkedIconBorderReadonlyColor: string
  toggleBackground: string
  toggleWidth: string
  /** Do not use this for other components */
  toggleShadow: { "0": TokenBoxshadowValueInst; "1": TokenBoxshadowValueInst }
  disabledOpacity: string | number
}

export default Toggle
