// @instructure/ui-themes/src/themes/newThemeTokens/componentTypes/treeBrowserTreeButton.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

export type TreeBrowserTreeButton = {
  baseSpacingLarge: string
  baseSpacingMedium: string
  baseSpacingSmall: string
  borderColor: string
  borderRadius: string
  borderWidth: string
  descriptorFontSizeLarge: string
  descriptorFontSizeMedium: string
  descriptorFontSizeSmall: string
  descriptorMarginTop: string
  descriptorTextColor: string
  hoverBackgroundColor: string
  hoverTextColor: string
  nameFontSizeLarge: string
  nameFontSizeMedium: string
  nameFontSizeSmall: string
  nameTextColor: string
  selectedBackgroundColor: string
  selectedTextColor: string
  textLineHeight: string | number
  iconsMarginRightSmall: string
  iconsMarginRightMedium: string
  iconsMarginRightLarge: string
}

export default TreeBrowserTreeButton
