// @instructure/ui-grid/lib/GridCol/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.GridCol = void 0;
var _react = require("react");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _emotion = require("@instructure/emotion");
var _console = require("@instructure/console");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: Grid
id: Grid.Col
---
**/
let GridCol = exports.GridCol = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class GridCol extends _react.Component {
  static displayName = "GridCol";
  static componentId = 'Grid.Col';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    textAlign: 'inherit',
    children: null,
    isLastCol: false,
    isLastRow: false
  };
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  widthCheck() {
    const {
      width
    } = this.props;
    let shouldWarn = false;
    if (width) {
      if (typeof width === 'number' && width <= 0) {
        shouldWarn = true;
      }
      if (typeof width === 'object') {
        Object.keys(width).forEach(breakpoint => {
          //@ts-expect-error Ts doesn't understand Object.keys properly
          if (typeof width[breakpoint] === 'number' && width[breakpoint] <= 0) {
            shouldWarn = true;
          }
        });
      }
    }
    if (shouldWarn) {
      (0, _console.logWarn)(false, 'Col width must be positive!');
    }
  }
  componentDidMount() {
    this.props.makeStyles?.();
    this.widthCheck();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
    this.widthCheck();
  }
  render() {
    const {
      children,
      styles
    } = this.props;
    const props = (0, _omitProps.omitProps)(this.props, GridCol.allowedProps);
    return (0, _jsxRuntime.jsx)("span", {
      ...props,
      ref: this.handleRef,
      css: styles?.gridCol,
      children: children
    });
  }
}) || _class);
var _default = exports.default = GridCol;
