// @instructure/ui-tooltip/es/Tooltip/v1/index.js
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { getElementType, omitProps, ensureSingleChild, passthroughProps, callRenderProp, withDeterministicId } from '@instructure/ui-react-utils';
import { Popover } from '@instructure/ui-popover/v11_6';
import { withStyle } from '@instructure/emotion';
import generateStyle from "./styles.js";
import generateComponentTheme from "./theme.js";
import { allowedProps } from "./props.js";
/**
---
category: components
---
**/
import { jsx as _jsx } from "@emotion/react/jsx-runtime";
let Tooltip = (_dec = withDeterministicId(), _dec2 = withStyle(generateStyle, generateComponentTheme), _dec(_class = _dec2(_class = class Tooltip extends Component {
  static displayName = "Tooltip";
  static componentId = 'Tooltip';
  static allowedProps = allowedProps;
  static defaultProps = {
    defaultIsShowingContent: false,
    color: 'primary',
    placement: 'top',
    constrain: 'window',
    offsetX: 0,
    offsetY: 0,
    preventTooltip: false
  };
  _id;
  ref = null;
  handleRef = el => {
    this.ref = el;
    if (typeof this.props.elementRef === 'function') {
      this.props.elementRef(el);
    }
  };
  constructor(props) {
    super(props);
    this._id = props.deterministicId();
    this.state = {
      hasFocus: false
    };
  }
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  handleFocus = () => {
    this.setState({
      hasFocus: true
    });
  };
  handleBlur = () => {
    this.setState({
      hasFocus: false
    });
  };
  renderTrigger() {
    const {
      children,
      as
    } = this.props;
    const {
      hasFocus
    } = this.state;
    const triggerProps = {
      'aria-describedby': this._id
    };
    if (as) {
      const Trigger = getElementType(Tooltip, this.props);
      const props = omitProps(this.props, Tooltip.allowedProps);
      return _jsx(Trigger, {
        ...props,
        ...triggerProps,
        children: children /*TODO check if it can be TooltipRenderChildren, this might cause a crash*/
      });
    } else if (typeof children === 'function') {
      return children({
        focused: hasFocus,
        getTriggerProps: props => ({
          ...triggerProps,
          ...props
        })
      });
    } else {
      return ensureSingleChild(children, triggerProps);
    }
  }
  render() {
    const {
      renderTip,
      isShowingContent,
      defaultIsShowingContent,
      on,
      color,
      placement,
      mountNode,
      constrain,
      offsetX,
      offsetY,
      positionTarget,
      onShowContent,
      onHideContent,
      preventTooltip,
      styles,
      ...rest
    } = this.props;
    return _jsx(Popover, {
      ...passthroughProps(rest),
      isShowingContent: !preventTooltip && isShowingContent,
      defaultIsShowingContent: !preventTooltip && defaultIsShowingContent,
      on: on,
      shouldRenderOffscreen: true,
      shouldReturnFocus: false,
      placement: placement,
      color: color === 'primary' ? 'primary-inverse' : 'primary',
      mountNode: mountNode,
      constrain: constrain,
      shadow: "none",
      offsetX: offsetX,
      offsetY: offsetY,
      positionTarget: positionTarget,
      renderTrigger: () => this.renderTrigger(),
      onShowContent: onShowContent,
      onHideContent: onHideContent,
      onFocus: this.handleFocus,
      onBlur: this.handleBlur,
      elementRef: this.handleRef,
      shouldCloseOnDocumentClick: false,
      shouldCloseOnEscape: true,
      shouldSetAriaExpanded: false,
      "data-cid": "Tooltip",
      children: !preventTooltip ? _jsx("span", {
        id: this._id,
        css: styles?.tooltip,
        role: "tooltip",
        children: callRenderProp(renderTip)
      }) : null
    });
  }
}) || _class) || _class);
export default Tooltip;
export { Tooltip };
