// @instructure/emotion/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  useComputedTheme: true,
  InstUISettingsProvider: true,
  withStyle: true,
  getComponentThemeOverride: true,
  withStyleNew: true,
  ThemeablePropValues: true,
  makeThemeVars: true,
  getShorthandPropValue: true,
  mirrorShorthandCorners: true,
  mirrorShorthandEdges: true,
  calcSpacingFromShorthand: true,
  calcFocusOutlineStyles: true,
  useStyle: true,
  useStyleNew: true,
  useTheme: true
};
Object.defineProperty(exports, "InstUISettingsProvider", {
  enumerable: true,
  get: function () {
    return _InstUISettingsProvider.InstUISettingsProvider;
  }
});
Object.defineProperty(exports, "ThemeablePropValues", {
  enumerable: true,
  get: function () {
    return _styleUtils.ThemeablePropValues;
  }
});
Object.defineProperty(exports, "calcFocusOutlineStyles", {
  enumerable: true,
  get: function () {
    return _styleUtils.calcFocusOutlineStyles;
  }
});
Object.defineProperty(exports, "calcSpacingFromShorthand", {
  enumerable: true,
  get: function () {
    return _styleUtils.calcSpacingFromShorthand;
  }
});
Object.defineProperty(exports, "getComponentThemeOverride", {
  enumerable: true,
  get: function () {
    return _getComponentThemeOverride.getComponentThemeOverride;
  }
});
Object.defineProperty(exports, "getShorthandPropValue", {
  enumerable: true,
  get: function () {
    return _styleUtils.getShorthandPropValue;
  }
});
Object.defineProperty(exports, "makeThemeVars", {
  enumerable: true,
  get: function () {
    return _styleUtils.makeThemeVars;
  }
});
Object.defineProperty(exports, "mirrorShorthandCorners", {
  enumerable: true,
  get: function () {
    return _styleUtils.mirrorShorthandCorners;
  }
});
Object.defineProperty(exports, "mirrorShorthandEdges", {
  enumerable: true,
  get: function () {
    return _styleUtils.mirrorShorthandEdges;
  }
});
Object.defineProperty(exports, "useComputedTheme", {
  enumerable: true,
  get: function () {
    return _useComputedTheme.useComputedTheme;
  }
});
Object.defineProperty(exports, "useStyle", {
  enumerable: true,
  get: function () {
    return _useStyle.useStyle;
  }
});
Object.defineProperty(exports, "useStyleNew", {
  enumerable: true,
  get: function () {
    return _useStyleNew.useStyleNew;
  }
});
Object.defineProperty(exports, "useTheme", {
  enumerable: true,
  get: function () {
    return _useTheme.useTheme;
  }
});
Object.defineProperty(exports, "withStyle", {
  enumerable: true,
  get: function () {
    return _withStyle.withStyle;
  }
});
Object.defineProperty(exports, "withStyleNew", {
  enumerable: true,
  get: function () {
    return _withStyleNew.withStyleNew;
  }
});
var _react = require("@emotion/react");
Object.keys(_react).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _react[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _react[key];
    }
  });
});
var _useComputedTheme = require("./useComputedTheme");
var _InstUISettingsProvider = require("./InstUISettingsProvider");
var _withStyle = require("./withStyle");
var _getComponentThemeOverride = require("./getComponentThemeOverride");
var _withStyleNew = require("./withStyleNew");
var _styleUtils = require("./styleUtils");
var _useStyle = require("./useStyle");
var _useStyleNew = require("./useStyleNew");
var _useTheme = require("./useTheme");
