// @instructure/emotion/lib/withStyleNew.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withStyleNew = exports.default = void 0;
var _react = require("react");
var _hoistNonReactStatics = _interopRequireDefault(require("hoist-non-react-statics"));
var _deepEqual = require("@instructure/ui-utils/lib/deepEqual.js");
var _mergeDeep = require("@instructure/ui-utils/lib/mergeDeep.js");
var _console = require("@instructure/console");
var _decorator = require("@instructure/ui-decorator/lib/decorator.js");
var _useTheme = require("./useTheme");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

// Extract is needed because it would allow number otherwise
// https://stackoverflow.com/a/51808262/319473

// Unique name of an InstUI component

const defaultValues = {
  styles: {},
  makeStyles: () => {}
};

/**
 * ---
 * category: utilities/themes
 * ---
 *
 * A decorator or higher order component that makes a component themeable.
 *
 * It adds a `makeStyles` function and the generated `styles` object to the decorated Component's props. If it has an own theme, it also adds the `themeOverride` prop to the component.
 *
 * As a HOC:
 *
 * ```js-code
 * import { withStyleNew } from '@instructure/emotion'
 * import generateStyle from './styles'
 *
 * export default withStyleNew(generateStyle)(ExampleComponent)
 * ```
 *
 * Themeable components inject their themed styles into the document
 * when they are mounted.
 *
 * ### Applying themes
 *
 * A themeable component’s theme can be configured via wrapping it in an
 * [InstUISettingsProvider](InstUISettingsProvider) component, and/or set
 * explicitly via its `themeOverride` prop.
 *
 * InstUISettingsProvider provides a theme object (e.g. the [canvas theme](/#canvas)).
 * These variables are mapped to the component's own variables in `theme.js` (see [theming](theming-basics) for more info).
 *
 * With the `themeOverride` prop you can directly set/override the component theme variables declared in theme.js. It accepts an object or a function. The function has the component's theme and the currently active main theme as its parameter.
 *
 * See more about the overrides on the [Using theme overrides](/#using-theme-overrides) docs page.
 *
 * ```jsx-code
 * {// global theme override}
 * <InstUISettingsProvider theme={{
 *   colors: { backgroundMedium: '#888' }
 * }}>
 *  {// component theme override}
 *   <ExampleComponent themeOverride={{ hoverColor: '#eee' }} />
 *
 *  {// component theme override with function}
 *   <ExampleComponent themeOverride={(componentTheme, currentTheme) => ({
 *     hoverBackground: componentTheme.background,
 *     activeBackground: currentTheme.colors.backgroundBrand
 *   })} />
 * </InstUISettingsProvider>
 * ```
 *
 * @module withStyleNew
 *
 * @param {function} generateStyle - The function that returns the component's style object
 * @returns {ReactElement} The decorated WithStyle Component
 */
const withStyleNew = exports.withStyleNew = (0, _decorator.decorator)((ComposedComponent, generateStyle, useTokensFrom, frozenTheme, generateComponentTheme) => {
  const displayName = ComposedComponent.displayName || ComposedComponent.name;
  const componentId = useTokensFrom ?? ComposedComponent.componentId?.replace('.', '');
  const WithStyle = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
    const themeInContext = (0, _useTheme.useTheme)();
    const themeKey = themeInContext.key;

    // this is for the new overrides. theme.themeOverride has all the overrides
    const themeOverride = themeInContext.themeOverride;

    // if a new theme has been added to the lib since this component version has been frozen, it can't be used with this
    // theme, so we throw an error. Solution: upgrade to the latest version, it will support it
    if (frozenTheme && !frozenTheme[themeKey]) {}

    // if this component is an older component version but still uses the new theming system, it'll have a frozenTheme
    // object passed to it. We use that instead of the current theme so backward compatibility stays intact
    const theme = frozenTheme ? {
      newTheme: frozenTheme[themeKey]
    } : themeInContext;
    if (props.styles) {
      (0, _console.warn)(false, `Manually passing the "styles" property is not allowed on the ${displayName} component. Using the default styles calculated by the @withStyleNew decorator instead.\n`, props.styles);
    }
    if (props.makeStyles) {
      (0, _console.warn)(false, `Manually passing the "makeStyles" property is not allowed on the ${displayName} component. Styles are calculated by the @withStyleNew decorator.`);
    }
    const componentProps = {
      ...ComposedComponent.defaultProps,
      ...props,
      ...defaultValues
    };
    // resolving the theming functions and applying the overrides
    const primitiveOverrides = themeOverride?.primitives;
    const semanticsOverrides = themeOverride?.semantics;
    const sharedTokensOverrides = themeOverride?.sharedTokens;
    const componentOverridesFromSettingsProvider = themeOverride?.components?.[componentId];
    const componentOverridesFromThemeOverrideProp = componentProps.themeOverride;
    const primitives = (0, _mergeDeep.mergeDeep)(theme.newTheme.primitives, primitiveOverrides);
    const semantics = (0, _mergeDeep.mergeDeep)(theme.newTheme.semantics?.(primitives), semanticsOverrides);
    const sharedTokens = (0, _mergeDeep.mergeDeep)(theme.newTheme.sharedTokens?.(semantics), sharedTokensOverrides);
    // Note: Some components do not have a theme, e.g., FormFieldMessages
    const baseComponentTheme = generateComponentTheme ? generateComponentTheme({
      primitives,
      semantics,
      sharedTokens
    }) : theme.newTheme.components[componentId]?.(semantics);
    const componentThemeFromSettingsProvider = (0, _mergeDeep.mergeDeep)(baseComponentTheme, componentOverridesFromSettingsProvider);
    const componentTheme = (0, _mergeDeep.mergeDeep)(componentThemeFromSettingsProvider,
    // @ts-ignore TODO-theme-types: fix typing
    typeof componentOverridesFromThemeOverrideProp === 'function' ? componentOverridesFromThemeOverrideProp(componentThemeFromSettingsProvider, themeInContext) : componentOverridesFromThemeOverrideProp);

    // TODO do not call here generateStyle, it does not receive the extraArgs
    const [styles, setStyles] = (0, _react.useState)(generateStyle ?
    // @ts-ignore TODO-theme-types: fix typing
    generateStyle(componentTheme, componentProps, sharedTokens, {}) : {});
    const makeStyleHandler = extraArgs => {
      const calculatedStyles = generateStyle(componentTheme, componentProps,
      // @ts-ignore TODO-theme-types: fix typing
      sharedTokens, extraArgs);
      if (!(0, _deepEqual.deepEqual)(calculatedStyles, styles)) {
        setStyles(calculatedStyles);
      }
    };
    return (0, _jsxRuntime.jsx)(ComposedComponent, {
      ref: ref,
      ...props,
      makeStyles: makeStyleHandler,
      styles: styles
    });
  });
  (0, _hoistNonReactStatics.default)(WithStyle, ComposedComponent);

  // added so it can be tested with ReactTestUtils
  // more info: https://github.com/facebook/react/issues/13455
  WithStyle.originalType = ComposedComponent.originalType || ComposedComponent;
  WithStyle.defaultProps = ComposedComponent.defaultProps;
  // These static fields exist on InstUI components
  WithStyle.allowedProps = ComposedComponent.allowedProps;

  // we have to add defaults to makeStyles and styles added by this decorator
  // eslint-disable-next-line no-param-reassign
  ComposedComponent.defaultProps = {
    ...ComposedComponent.defaultProps,
    makeStyles: defaultValues.makeStyles,
    styles: defaultValues.styles
  };
  if (typeof process !== 'undefined' && process?.env?.NODE_ENV !== 'production') {
    WithStyle.displayName = `WithStyle(${displayName})`;
  }
  return WithStyle;
});
var _default = exports.default = withStyleNew;
