// @instructure/ui-progress/lib/ProgressCircle/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ProgressCircle = void 0;
var _react = require("react");
var _ScreenReaderContent = require("@instructure/ui-a11y-content/lib/ScreenReaderContent");
var _v11_ = require("@instructure/ui-view/v11_6");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let ProgressCircle = exports.ProgressCircle = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class ProgressCircle extends _react.Component {
  static displayName = "ProgressCircle";
  static componentId = 'ProgressCircle';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    formatScreenReaderValue: ({
      valueNow,
      valueMax
    }) => `${valueNow} / ${valueMax}`,
    size: 'medium',
    valueMax: 100,
    valueNow: 0,
    as: 'div',
    color: 'primary',
    shouldAnimateOnMount: false,
    // default to showing `success` color on completion
    meterColor: ({
      valueNow,
      valueMax
    }) => valueNow / valueMax >= 1 ? 'success' : 'brand'
  };
  _timeouts = [];
  ref = null;
  constructor(props) {
    super(props);
    this.state = {
      shouldAnimateOnMount: props.shouldAnimateOnMount
    };
  }
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  get makeStylesVariables() {
    return {
      shouldAnimateOnMount: this.state.shouldAnimateOnMount
    };
  }
  componentDidMount() {
    if (this.state.shouldAnimateOnMount) {
      this._timeouts.push(setTimeout(() => {
        this.setState({
          shouldAnimateOnMount: false
        });
      }, this.props.animationDelay || 500));
    }
    this.props.makeStyles?.(this.makeStylesVariables);
  }
  componentDidUpdate() {
    this.props.makeStyles?.(this.makeStylesVariables);
  }
  componentWillUnmount() {
    this._timeouts.forEach(timeout => clearTimeout(timeout));
  }
  render() {
    const {
      color,
      renderValue,
      formatScreenReaderValue,
      meterColor,
      valueNow,
      valueMax,
      screenReaderLabel,
      size,
      styles,
      ...props
    } = this.props;
    const valueText = typeof formatScreenReaderValue === 'function' ? formatScreenReaderValue({
      valueNow: valueNow,
      valueMax: valueMax
    }) : undefined;
    // consolidating the label and aria-valuetext to put in aria-label because
    // NVDA does not read aria-valuetext: https://github.com/nvaccess/nvda/issues/913
    // But leaving aria-valuetext because JAWS ignores aria-label
    const labelAndValueText = `${screenReaderLabel} ${valueText}`;
    const value = (0, _callRenderProp.callRenderProp)(renderValue, {
      valueNow: valueNow,
      valueMax: valueMax
    });
    const style = {
      strokeDashoffset: `${styles?.dashOffset}em`
    };
    return (0, _jsxRuntime.jsxs)(_v11_.View, {
      ...(0, _passthroughProps.passthroughProps)(props),
      as: this.props.as,
      elementRef: this.handleRef,
      css: styles?.progressCircle,
      margin: this.props.margin,
      "data-cid": "ProgressCircle",
      children: [(0, _jsxRuntime.jsx)(_ScreenReaderContent.ScreenReaderContent, {
        children: (0, _jsxRuntime.jsx)("progress", {
          max: valueMax,
          value: valueNow,
          "aria-valuetext": labelAndValueText
        })
      }), value && (0, _jsxRuntime.jsx)("span", {
        css: styles?.center,
        "aria-hidden": "true",
        children: (0, _jsxRuntime.jsx)("span", {
          css: styles?.value,
          children: value
        })
      }), (0, _jsxRuntime.jsxs)("svg", {
        css: styles?.circle,
        role: "presentation",
        focusable: "false",
        children: [(0, _jsxRuntime.jsx)("circle", {
          css: styles?.track,
          role: "presentation",
          cx: "50%",
          cy: "50%",
          r: styles?.radii?.radius
        }), (0, _jsxRuntime.jsx)("circle", {
          css: styles?.border,
          role: "presentation",
          cx: "50%",
          cy: "50%",
          r: styles?.radii?.borderOffsetRadius
        }), (0, _jsxRuntime.jsx)("circle", {
          css: styles?.meter,
          role: "presentation",
          style: style,
          cx: "50%",
          cy: "50%",
          r: styles?.radii?.radius
        })]
      })]
    });
  }
}) || _class);
var _default = exports.default = ProgressCircle;
