// @instructure/ui-themes/src/themes/newThemeTokens/light/components/nutritionFacts.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { NutritionFacts } from "../../componentTypes/nutritionFacts"
import type { Semantics } from "../semantics"

const nutritionFacts = (semantics: Semantics): NutritionFacts => ({
  borderRadius: semantics.borderRadius.lg,
  cardBorderColor: semantics.color.stroke.muted,
  cardBorderWidth: semantics.borderWidth.sm,
  cardPadding: semantics.spacing.gap.cards.sm,
  cardExplainerContainerBottomMargin: semantics.spacing.spaceSm,
  cardGap: semantics.spacing.spaceXl,
  bodyPadding: semantics.spacing.padding.container.md,
  blockGap: semantics.spacing.gap.cards.md,
})
export default nutritionFacts
