// @instructure/ui-themes/src/themes/newThemeTokens/light/components/radioInput.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { RadioInput } from "../../componentTypes/radioInput"
import type { Semantics } from "../semantics"

const radioInput = (semantics: Semantics): RadioInput => ({
  backgroundColor: semantics.color.background.interactive.input.base,
  backgroundHoverColor: semantics.color.background.interactive.input.hover,
  backgroundReadonlyColor: semantics.color.background.interactive.input.readonly,
  backgroundDisabledColor: semantics.color.background.interactive.input.disabled,
  borderColor: semantics.color.stroke.interactive.input.base,
  borderHoverColor: semantics.color.stroke.interactive.input.hover,
  borderReadonlyColor: semantics.color.stroke.interactive.input.readonly,
  borderDisabledColor: semantics.color.stroke.interactive.input.disabled,
  borderSelectedColor: semantics.color.stroke.interactive.input.selected,
  fontSizeSm: semantics.fontSize.textSm,
  fontSizeMd: semantics.fontSize.textBase,
  fontSizeLg: semantics.fontSize.textLg,
  toggleSmallFontSize: semantics.fontSize.textXs,
  toggleMediumFontSize: semantics.fontSize.textSm,
  toggleLargeFontSize: semantics.fontSize.textBase,
  lineHeightSm: semantics.lineHeight.paragraph.textSm,
  lineHeightMd: semantics.lineHeight.paragraph.textBase,
  lineHeightLg: semantics.lineHeight.paragraph.textBase,
  fontFamily: semantics.fontFamily.base,
  fontWeight: semantics.fontWeight.body.base,
  gap: semantics.spacing.spaceSm,
  borderWidth: semantics.borderWidth.interactive.base,
  labelBaseColor: semantics.color.text.base,
  labelHoverColor: semantics.color.text.base,
  labelDisabledColor: semantics.color.text.interactive.disabled.base,
  labelReadonlyColor: semantics.color.text.base,
  controlSizeSm: semantics.size.choiceControl.height.sm,
  controlSizeMd: semantics.size.choiceControl.height.md,
  controlSizeLg: semantics.size.choiceControl.height.lg,
  checkedInsetSm: "0.375rem",
  checkedInsetMd: "0.375rem",
  checkedInsetLg: "0.375rem",
  toggleSmallHeight: semantics.size.interactive.height.sm,
  toggleMediumHeight: semantics.size.interactive.height.md,
  toggleLargeHeight: semantics.size.interactive.height.lg,
  controlVerticalMargin: "0rem",
  toggleBackgroundDanger: semantics.color.background.warning,
  toggleBackgroundOff: semantics.color.background.success,
  toggleBackgroundSuccess: semantics.color.background.success,
  toggleBackgroundWarning: semantics.color.background.warning,
  toggleHandleText: semantics.color.text.onColor,
  toggleBorderRadius: semantics.borderRadius.sm,
  toggleBorderWidth: semantics.borderWidth.lg,
  toggleShadow: {
    0: {
      x: semantics.dropShadow.x.elevation1.dropshadow1,
      y: semantics.dropShadow.y.elevation1.dropshadow1,
      blur: semantics.dropShadow.blur.elevation1.dropshadow1,
      spread: semantics.dropShadow.spread.elevation1.dropshadow1,
      color: semantics.color.dropShadow.shadowColor1,
      type: "dropShadow",
    },
    1: {
      x: semantics.dropShadow.x.elevation1.dropshadow2,
      y: semantics.dropShadow.y.elevation1.dropshadow2,
      blur: semantics.dropShadow.blur.elevation1.dropshadow2,
      spread: semantics.dropShadow.spread.elevation1.dropshadow2,
      color: semantics.color.dropShadow.shadowColor2,
      type: "innerShadow",
    },
  },
})
export default radioInput
