// @instructure/ui-link/lib/Link/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Link = void 0;
var _react = require("react");
var _latest = require("@instructure/ui-view/latest");
var _hasVisibleChildren = require("@instructure/ui-a11y-utils/lib/hasVisibleChildren.js");
var _isActiveElement = require("@instructure/ui-dom-utils/lib/isActiveElement.js");
var _findFocusable = require("@instructure/ui-dom-utils/lib/findFocusable.js");
var _getElementType = require("@instructure/ui-react-utils/lib/getElementType.js");
var _getInteraction = require("@instructure/ui-react-utils/lib/getInteraction.js");
var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _combineDataCid = require("@instructure/ui-utils/lib/combineDataCid.js");
var _console = require("@instructure/console");
var _renderIconWithProps = require("@instructure/ui-icons/lib/IconPropsProvider/renderIconWithProps.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let Link = exports.Link = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class Link extends _react.Component {
  static displayName = "Link";
  static componentId = 'Link';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    // Leave interaction default undefined so that `disabled` can also be supplied
    interaction: undefined,
    color: 'link',
    iconPlacement: 'start',
    forceButtonRole: true
  };
  state = {
    hasFocus: false
  };
  get _link() {
    return this.ref;
  }
  ref = null;
  componentDidMount() {
    this.props.makeStyles?.(this.makeStyleProps());
  }
  componentDidUpdate() {
    this.props.makeStyles?.(this.makeStyleProps());
  }
  makeStyleProps = () => {
    const {
      variant: variantProp,
      size: sizeProp
    } = this.props;

    // Handle deprecated variant values by mapping them to new variant + size props
    let variant = variantProp;
    let size = sizeProp;
    if (variantProp === 'inline-small' || variantProp === 'standalone-small') {
      (0, _console.logWarn)(false, `[Link] The variant value "${variantProp}" is deprecated. Use variant="${variantProp.replace('-small', '')}" with size="small" instead.`);
      variant = variantProp.replace('-small', '');
      // Only set size from deprecated variant if size prop is not explicitly provided
      if (!sizeProp) {
        size = 'small';
      }
    } else if ((variantProp === 'inline' || variantProp === 'standalone') && !sizeProp) {
      // When using new variant values without explicit size, default to medium
      // This maintains the old behavior where 'inline' and 'standalone' were medium-sized
      size = 'medium';
    }
    return {
      containsTruncateText: this.containsTruncateText,
      hasVisibleChildren: this.hasVisibleChildren,
      variant,
      size
    };
  };
  handleElementRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  handleClick = event => {
    const {
      onClick
    } = this.props;
    const {
      interaction
    } = this;
    if (interaction === 'disabled') {
      event.preventDefault();
      event.stopPropagation();
    } else if (typeof onClick === 'function') {
      onClick(event);
    }
  };
  handleFocus = event => {
    this.setState({
      hasFocus: true
    });
    if (typeof this.props.onFocus === 'function') {
      this.props.onFocus(event);
    }
  };
  handleBlur = event => {
    this.setState({
      hasFocus: false
    });
    if (typeof this.props.onBlur === 'function') {
      this.props.onBlur(event);
    }
  };
  get containsTruncateText() {
    let truncateText = false;
    _react.Children.forEach(this.props.children, child => {
      if (child && (0, _matchComponentTypes.matchComponentTypes)(child, ['TruncateText'])) {
        truncateText = true;
      }
    });
    (0, _console.logWarn)(
    // if display prop is used, warn about icon or TruncateText
    !truncateText || this.props.display === undefined, '[Link] Using the display property with TruncateText may cause layout issues.');
    return truncateText;
  }
  get display() {
    if (this.props.display) {
      return this.props.display; // user-entered display property
    }
    const {
      containsTruncateText
    } = this;
    if (this.props.renderIcon) {
      return containsTruncateText ? 'inline-flex' : 'inline-block';
    } else {
      return containsTruncateText ? 'block' : 'auto';
    }
  }
  get interaction() {
    return (0, _getInteraction.getInteraction)({
      props: this.props,
      interactionTypes: ['disabled']
    });
  }
  get element() {
    return (0, _getElementType.getElementType)(Link, this.props);
  }
  get focused() {
    return (0, _isActiveElement.isActiveElement)(this.ref);
  }
  get focusable() {
    return (0, _findFocusable.findFocusable)(this.ref);
  }
  get hasVisibleChildren() {
    return (0, _hasVisibleChildren.hasVisibleChildren)(this.props.children);
  }
  get role() {
    const {
      role,
      forceButtonRole,
      onClick
    } = this.props;
    if (forceButtonRole) {
      return onClick && this.element !== 'button' ? 'button' : role;
    }
    return role;
  }
  focus() {
    this.ref && this.ref.focus();
  }
  renderIcon() {
    const {
      display,
      renderIcon,
      variant: variantProp,
      size: sizeProp
    } = this.props;
    (0, _console.logWarn)(
    // if display prop is used, warn about icon or TruncateText
    display === undefined, '[Link] Using the display property with an icon may cause layout issues.');

    // Determine the actual size being used (considering deprecated variants)
    let size = sizeProp;
    if (variantProp === 'inline-small' || variantProp === 'standalone-small') {
      size = sizeProp || 'small';
    } else if ((variantProp === 'inline' || variantProp === 'standalone') && !sizeProp) {
      size = 'medium';
    }

    // Map Link sizes to icon sizes
    const linkSizeToIconSize = {
      small: 'xs',
      medium: 'sm',
      large: 'lg'
    };
    const iconSize = linkSizeToIconSize[size || 'medium'];
    return (0, _jsxRuntime.jsx)("span", {
      css: this.props.styles?.icon,
      children: (0, _renderIconWithProps.renderIconWithProps)(renderIcon, iconSize, undefined)
    });
  }
  render() {
    const {
      children,
      onClick,
      onMouseEnter,
      color,
      href,
      margin,
      renderIcon,
      iconPlacement,
      ...props
    } = this.props;
    const {
      interaction
    } = this;
    const isDisabled = interaction === 'disabled';
    const type = this.element === 'button' || this.element === 'input' ? 'button' : undefined;
    const tabIndex = this.role === 'button' && !isDisabled ? 0 : undefined;
    return (0, _jsxRuntime.jsxs)(_latest.View, {
      ...(0, _passthroughProps.passthroughProps)(props),
      elementRef: this.handleElementRef,
      as: this.element,
      display: this.display,
      margin: margin,
      href: href,
      onMouseEnter: onMouseEnter,
      onClick: this.handleClick,
      onFocus: this.handleFocus,
      onBlur: this.handleBlur,
      "aria-disabled": isDisabled ? 'true' : undefined,
      role: this.role,
      type: type,
      tabIndex: tabIndex,
      css: this.props.styles?.link,
      "data-cid": (0, _combineDataCid.combineDataCid)('Link', this.props),
      children: [renderIcon && iconPlacement === 'start' ? this.renderIcon() : null, children, renderIcon && iconPlacement === 'end' ? this.renderIcon() : null]
    });
  }
}) || _class);
var _default = exports.default = Link;
