// @instructure/ui-top-nav-bar/es/TopNavBar/v1/TopNavBarBreadcrumb/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Children, Component } from 'react';
import { allowedProps } from "./props.js";
import TopNavBar from "../index.js";
import { withStyle } from '@instructure/emotion';
import generateStyle from "../TopNavBarBreadcrumb/styles.js";
import { IconArrowOpenStartLine, IconHamburgerLine } from '@instructure/ui-icons';
import TopNavBarContext from "../TopNavBarContext.js";
import { error } from '@instructure/console';
import { Link } from '@instructure/ui-link/v11_6';

/**
---
parent: TopNavBar
id: TopNavBar.Breadcrumb
---
@module TopNavBarBreadcrumb
 **/
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
let TopNavBarBreadcrumb = (_dec = withStyle(generateStyle, null), _dec(_class = class TopNavBarBreadcrumb extends Component {
  static displayName = "TopNavBarBreadcrumb";
  static allowedProps = allowedProps;
  static defaultProps = {};
  static contextType = TopNavBarContext;
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.({
      inverseColor: this.context.inverseColor
    });
  }
  componentDidUpdate() {
    this.props.makeStyles?.({
      inverseColor: this.context.inverseColor
    });
  }
  renderMenu() {
    const {
      onClick
    } = this.props;
    return _jsx("div", {
      css: this.props.styles?.icon,
      children: _jsx(TopNavBar.Item, {
        id: "iconItem",
        variant: "icon",
        renderIcon: _jsx(IconHamburgerLine, {}),
        onClick: onClick,
        children: "Hamburger menu"
      })
    });
  }
  render() {
    const {
      children,
      styles
    } = this.props;
    if (!this.context.inverseColor) {
      error(false, `[TopNavBarBreadcrumb] If the inverseColor prop is not set to true, TopNavBarBreadcrumb fails to render.`);
    }
    const breadcrumbElement = Children.toArray(children)[0];
    const breadCrumbLinks = breadcrumbElement.props.children;
    const lastButOneLink = Children.toArray(breadCrumbLinks)[Children.count(breadCrumbLinks) - 2];
    return this.context.inverseColor && (this.context.layout === 'desktop' ? _jsxs("div", {
      ref: this.handleRef,
      css: styles?.topNavBarBreadcrumb,
      "data-cid": "TopNavBarBreadcrumb",
      children: [_jsx("div", {
        css: styles?.iconContainer,
        children: this.renderMenu()
      }), _jsx("div", {
        css: styles?.breadcrumbContainer,
        children: children
      })]
    }) : _jsx("div", {
      ref: this.handleRef,
      css: styles?.topNavBarBreadcrumb,
      "data-cid": "TopNavBarBreadcrumb",
      children: _jsx("div", {
        css: styles?.linkContainer,
        children: lastButOneLink && _jsx(Link, {
          href: lastButOneLink.props.href,
          isWithinText: false,
          renderIcon: IconArrowOpenStartLine,
          children: lastButOneLink.props.children
        })
      })
    }));
  }
}) || _class);
export { TopNavBarBreadcrumb };
export default TopNavBarBreadcrumb;
