// @instructure/ui-pages/lib/Pages/v1/Page/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Page = void 0;
var _react = require("react");
var _findDOMNode = require("@instructure/ui-dom-utils/lib/findDOMNode.js");
var _findTabbable = require("@instructure/ui-dom-utils/lib/findTabbable.js");
var _console = require("@instructure/console");
var _v11_ = require("@instructure/ui-view/v11_6");
var _PagesContext = require("../PagesContext");
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
parent: Pages
id: Pages.Page
---
**/
class Page extends _react.Component {
  static displayName = "Page";
  static componentId = 'Pages.Page';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    defaultFocusElement: null,
    padding: 'small',
    textAlign: 'start',
    children: null
  };
  get _content() {
    return this.ref;
  }
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  get defaultFocusElement() {
    let {
      defaultFocusElement
    } = this.props;
    if (typeof defaultFocusElement === 'function') {
      defaultFocusElement = defaultFocusElement();
    }
    if (defaultFocusElement) {
      defaultFocusElement = (0, _findDOMNode.findDOMNode)(defaultFocusElement);
    }
    if (!defaultFocusElement) {
      const tabbable = (0, _findTabbable.findTabbable)(this.ref);
      defaultFocusElement = tabbable && tabbable[0];
    }
    (0, _console.logError)(defaultFocusElement && !!defaultFocusElement.focus, '[Page] A default focusable element is required or focus will be lost.');
    return defaultFocusElement;
  }
  get focusable() {
    const element = this.defaultFocusElement;
    return !!element && typeof element.focus === 'function';
  }
  focus() {
    if (this.focusable) {
      ;
      this.defaultFocusElement.focus();
    }
  }
  render() {
    const {
      children,
      ...props
    } = this.props;
    return (0, _jsxRuntime.jsx)(_PagesContext.PagesContext.Consumer, {
      children: ({
        history,
        navigateToPreviousPage
      }) => (0, _jsxRuntime.jsx)(_v11_.View, {
        as: "div",
        padding: props.padding,
        textAlign: props.textAlign,
        elementRef: this.handleRef,
        children: children && typeof children === 'function' ? children(history, navigateToPreviousPage) : children
      })
    });
  }
}
exports.Page = Page;
var _default = exports.default = Page;
