// @instructure/ui-top-nav-bar/lib/TopNavBar/v1/TopNavBarItem/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TopNavBarItem = void 0;
var _react = require("react");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _createChainedFunction = require("@instructure/ui-utils/lib/createChainedFunction.js");
var _console = require("@instructure/console");
var _emotion = require("@instructure/emotion");
var _IconArrowOpenDownSolid = require("@instructure/ui-icons/lib/generated/IconArrowOpenDownSolid.js");
var _IconArrowOpenUpSolid = require("@instructure/ui-icons/lib/generated/IconArrowOpenUpSolid.js");
var _v11_ = require("@instructure/ui-avatar/v11_6");
var _v11_2 = require("@instructure/ui-buttons/v11_6");
var _v11_3 = require("@instructure/ui-tooltip/v11_6");
var _v11_4 = require("@instructure/ui-popover/v11_6");
var _v11_5 = require("@instructure/ui-drilldown/v11_6");
var _ScreenReaderContent = require("@instructure/ui-a11y-content/lib/ScreenReaderContent");
var _TopNavBarContext = require("../TopNavBarContext");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
const navbarStackingDefault = 10000;
const submenuStacking = navbarStackingDefault + 1;
const tooltipStacking = navbarStackingDefault + 2;

/**
---
parent: TopNavBar
id: TopNavBar.Item
---
@module TopNavBarItem
**/
let TopNavBarItem = exports.TopNavBarItem = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = _dec2(_class = class TopNavBarItem extends _react.Component {
  static displayName = "TopNavBarItem";
  static componentId = 'TopNavBar.Item';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    status: 'default',
    variant: 'default',
    showSubmenuChevron: true,
    shouldCloseOnClick: 'auto'
  };
  static contextType = _TopNavBarContext.TopNavBarContext;
  ref = null;
  _itemRef = null;
  _tooltipContentId;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  handleItemRef = el => {
    const {
      itemRef
    } = this.props;
    this._itemRef = el;
    if (typeof itemRef === 'function') {
      itemRef(el);
    }
  };
  constructor(props) {
    super(props);
    this._tooltipContentId = props.deterministicId('TopNavBarItem-tooltip');
    this.state = {
      isSubmenuOpen: false,
      isPopoverOpen: false,
      isFocused: false
    };
  }
  componentDidMount() {
    const {
      makeStyles,
      customPopoverConfig
    } = this.props;
    makeStyles?.(this.makeStylesVariables);
    if (customPopoverConfig && this.isCustomPopoverControlled) {
      this.setState({
        isPopoverOpen: !!customPopoverConfig.isShowingContent
      });
    }
  }
  componentDidUpdate() {
    const {
      makeStyles,
      customPopoverConfig
    } = this.props;
    makeStyles?.(this.makeStylesVariables);
    if (customPopoverConfig && this.isCustomPopoverControlled && customPopoverConfig.isShowingContent !== this.state.isPopoverOpen) {
      this.setState({
        isPopoverOpen: !!customPopoverConfig.isShowingContent
      });
    }
  }
  get makeStylesVariables() {
    const {
      layout,
      inverseColor
    } = this.context;
    return {
      layout,
      inverseColor
    };
  }
  get shouldRenderSubmenu() {
    const {
      renderSubmenu
    } = this.props;
    return !!renderSubmenu && (0, _matchComponentTypes.matchComponentTypes)(renderSubmenu, [_v11_5.Drilldown]);
  }
  get shouldRenderPopover() {
    const {
      id,
      renderSubmenu,
      customPopoverConfig
    } = this.props;
    if (!customPopoverConfig) {
      return false;
    }
    if (!customPopoverConfig.children) {
      (0, _console.warn)(false, `Pass the content of the custom Popover as "customPopoverConfig.children" for the item with id: "${id}".`);
      return false;
    }
    if (customPopoverConfig && renderSubmenu) {
      (0, _console.warn)(false, `TopNavBar.Items are not allowed to have both the "renderSubmenu" and "customPopoverConfig" props. For submenus, pass a Drilldown component via the "renderSubmenu" prop, and only use "customPopoverConfig" for custom features. Item with id: "${id}" will ignore the "customPopoverConfig" prop.`);
      return false;
    }
    return true;
  }
  get isCustomPopoverControlled() {
    const {
      customPopoverConfig
    } = this.props;
    return customPopoverConfig && typeof customPopoverConfig.isShowingContent !== 'undefined';
  }
  get hasOpenPopover() {
    return this.state.isSubmenuOpen || this.state.isPopoverOpen;
  }
  get isDefaultVariant() {
    return this.props.variant === 'default' || this.props.variant === 'forceIconWithLabel';
  }
  get isButtonVariant() {
    return this.props.variant === 'button';
  }
  get isIconVariant() {
    return this.props.variant === 'icon';
  }
  get isAvatarOnlyVariant() {
    return this.props.variant === 'avatar';
  }
  get hasAvatar() {
    const {
      id,
      renderIcon,
      renderAvatar,
      variant
    } = this.props;
    const hasAvatar = !!renderAvatar;
    if (this.isAvatarOnlyVariant && !hasAvatar) {
      (0, _console.warn)(false, `The "renderAvatar" config is required for the 'variant="avatar"' type <TopNavBar.Item> components, but received none for the item with id "${id}".`);
      return false;
    }
    if (hasAvatar && (renderIcon || variant === 'icon')) {
      (0, _console.warn)(false, `<TopNavBar.Item> components with icons cannot display avatars, so the "renderAvatar" config prop will be ignored for item with id "${id}".`);
      return false;
    }
    return hasAvatar;
  }
  get submenuIcon() {
    if (!(this.shouldRenderSubmenu || this.shouldRenderPopover) || !this.props.showSubmenuChevron) {
      return null;
    }
    return (0, _jsxRuntime.jsx)("span", {
      css: this.props.styles?.submenuIcon,
      children: this.shouldRenderSubmenu && this.state.isSubmenuOpen || this.shouldRenderPopover && this.state.isPopoverOpen ? (0, _jsxRuntime.jsx)(_IconArrowOpenUpSolid.IconArrowOpenUpSolid, {}) : (0, _jsxRuntime.jsx)(_IconArrowOpenDownSolid.IconArrowOpenDownSolid, {})
    });
  }
  get buttonThemeOverride() {
    const {
      styles
    } = this.props;
    const themeOverride = {};
    if (this.isDefaultVariant || this.hasAvatar) {
      themeOverride.mediumPaddingHorizontal = styles?.itemInlinePadding;
    }
    if (this.hasAvatar) {
      themeOverride.mediumPaddingTop = '0.125rem';
      themeOverride.mediumPaddingBottom = '0.125rem';
    }
    if (this.isAvatarOnlyVariant && !this.submenuIcon) {
      themeOverride.mediumPaddingHorizontal = '0';
      themeOverride.mediumPaddingTop = '0';
      themeOverride.mediumPaddingBottom = '0';
    }
    return Object.keys(themeOverride).length > 0 ? themeOverride : undefined;
  }
  get colorProps() {
    let color = 'primary-inverse';
    let focusColor = 'inverse';
    if (this.context.inverseColor) {
      color = 'secondary';
      focusColor = 'info';
      if (this.isButtonVariant) {
        color = 'primary';
      }
    }
    return {
      color,
      focusColor
    };
  }
  get ariaProps() {
    let hasPopup;
    let expanded;
    if (this.shouldRenderSubmenu) {
      hasPopup = 'menu';
      expanded = this.state.isSubmenuOpen;
    }
    if (this.shouldRenderPopover) {
      hasPopup = 'true';
      expanded = this.state.isPopoverOpen;
    }
    return {
      'aria-haspopup': hasPopup,
      'aria-expanded': expanded
    };
  }
  get itemProps() {
    const {
      id,
      as,
      variant,
      href: hrefOriginal,
      onClick: onClickOriginal,
      onMouseOver,
      onMouseOut,
      onFocus,
      onBlur,
      onKeyDown,
      onKeyUp,
      renderSubmenu,
      status: statusOriginal,
      renderAvatar,
      renderIcon,
      withFocusOutline
    } = this.props;
    let href = hrefOriginal;
    let onClick = onClickOriginal;
    let status = statusOriginal;
    if (renderSubmenu) {
      if (href) {
        (0, _console.warn)(false, `TopNavBar.Items with submenus are not allowed to have 'href' property, but received href "${href}" for item with the id: "${id}".`);
        href = undefined;
      }
      if (onClick) {
        (0, _console.warn)(false, `TopNavBar.Items with submenus are not allowed to have 'onClick' property, but received onClick for item with the id: "${id}".Use the \`onSubmenuToggle\` prop instead. OnClick: ${onClick}`);
        onClick = undefined;
      }
    }
    if (status === 'active' && !this.isDefaultVariant) {
      (0, _console.warn)(false, `Only \`variant="default"\` <TopNavBar.Item> components can be set to active, but item with id "${id}" has variant: "${variant}".`);
      status = 'default';
    }
    if (this.hasAvatar) {
      if (!renderAvatar?.avatarName) {
        (0, _console.error)(false, `The "avatarName" prop is required for for <TopNavBar.Item> components with avatar, but the item with id "${id}" is missing it.`);
        return null;
      }
      if (status === 'active') {
        (0, _console.warn)(false, `<TopNavBar.Item> components with avatar cannot have "active" status, so the "active" status on the item with id "${id}" will be ignored.`);
        status = 'default';
      }
    }
    if (this.isIconVariant) {
      if (!renderIcon) {
        (0, _console.error)(false, `The "renderIcon" prop is required for the \`variant="icon"\` type <TopNavBar.Item> components, but the item with id "${id}" is missing it.`);
        return null;
      }
    }
    return {
      id,
      as,
      ...this.colorProps,
      ...this.ariaProps,
      size: 'medium',
      withBorder: this.isButtonVariant,
      withBackground: this.isButtonVariant,
      interaction: status === 'disabled' ? 'disabled' : undefined,
      'aria-disabled': status === 'disabled' ? 'true' : undefined,
      shape: this.isAvatarOnlyVariant && !this.submenuIcon ? 'circle' : 'rectangle',
      href,
      onClick,
      onMouseOver,
      onMouseOut,
      onFocus: (0, _createChainedFunction.createChainedFunction)(onFocus, this.onFocus),
      onBlur: (0, _createChainedFunction.createChainedFunction)(onBlur, this.onBlur),
      onKeyDown: (0, _createChainedFunction.createChainedFunction)(onKeyDown, this.handleKeyDown),
      onKeyUp,
      renderIcon,
      themeOverride: this.buttonThemeOverride,
      elementRef: e => {
        this.handleItemRef(e);
      },
      withFocusOutline: withFocusOutline || this.hasOpenPopover ? true : undefined
    };
  }
  onFocus = () => {
    this.setState({
      isFocused: true
    });
  };
  onBlur = () => {
    this.setState({
      isFocused: false
    });
  };
  handleKeyDown = e => {
    if (e.key === 'ArrowDown') {
      if (this.shouldRenderSubmenu && !this.state.isSubmenuOpen || this.shouldRenderPopover && !this.state.isPopoverOpen) {
        e.preventDefault();
        this._itemRef?.click();
      }
    }
    if (e.key === 'Tab') {
      if (this.state.isPopoverOpen && !this.isCustomPopoverControlled) {
        this.setState({
          isPopoverOpen: false
        });
      }
    }
  };
  handleMenuToggle = (event, args) => {
    this.setState({
      isSubmenuOpen: args.shown
    });
    if (typeof this.props.onSubmenuToggle === 'function') {
      this.props.onSubmenuToggle(event, args);
    }
  };
  renderContent() {
    const {
      children,
      styles
    } = this.props;
    const {
      itemProps
    } = this;
    if (!itemProps) {
      return null;
    }
    let content = children;
    if (this.isAvatarOnlyVariant) {
      content = this.renderAvatar();
    } else if (this.isIconVariant) {
      content = (0, _jsxRuntime.jsx)(_ScreenReaderContent.ScreenReaderContent, {
        children: content
      });
    }
    const button = (0, _jsxRuntime.jsx)(_v11_2.BaseButton, {
      ...itemProps,
      ...(0, _passthroughProps.passthroughProps)((0, _omitProps.omitProps)(this.props, _props.allowedProps)),
      children: this.isIconVariant && !this.submenuIcon ? content : (0, _jsxRuntime.jsxs)("div", {
        css: styles?.content,
        children: [this.hasAvatar && !this.isAvatarOnlyVariant && (0, _jsxRuntime.jsx)("span", {
          css: this.props.styles?.avatarContainer,
          children: this.renderAvatar()
        }), (0, _jsxRuntime.jsx)("span", {
          css: styles?.contentText,
          children: content
        }), this.submenuIcon]
      })
    });
    return (0, _jsxRuntime.jsx)("div", {
      css: styles?.container,
      children: this.props.tooltip ? this.wrapButtonInTooltip(button, children) : button
    });
  }
  wrapButtonInTooltip(button, content) {
    const {
      tooltip,
      styles
    } = this.props;
    if (!tooltip) {
      return button;
    }
    let tooltipProps = {
      placement: 'bottom',
      color: this.context.inverseColor ? 'primary' : 'primary-inverse'
    };
    if (typeof tooltip === 'string') {
      tooltipProps.renderTip = tooltip;
    } else {
      tooltipProps = {
        ...tooltipProps,
        ...tooltip
      };
    }
    const tipContent = (0, _callRenderProp.callRenderProp)(tooltipProps.renderTip);
    if (!tipContent) {
      return button;
    }
    const contentAndTooltipIdentical = tipContent === content;
    return (0, _jsxRuntime.jsx)(_emotion.InstUISettingsProvider, {
      theme: {
        componentOverrides: {
          View: {
            // moves tooltips above submenus/popovers
            stackingTopmost: tooltipStacking
          }
        }
      },
      children: (0, _jsxRuntime.jsx)(_v11_3.Tooltip, {
        ...tooltipProps,
        renderTip: contentAndTooltipIdentical ? tipContent : (0, _jsxRuntime.jsx)("div", {
          id: this._tooltipContentId,
          children: tipContent
        }),
        positionContainerDisplay: "block",
        children: (0, _jsxRuntime.jsx)("div", {
          css: styles?.submenuTriggerContainer,
          children: contentAndTooltipIdentical ? button : (0, _safeCloneElement.safeCloneElement)(button, {
            'aria-describedby': this._tooltipContentId
          })
        })
      })
    });
  }
  renderAvatar() {
    if (!this.hasAvatar) {
      return null;
    }
    const {
      id,
      children,
      renderAvatar
    } = this.props;
    const {
      avatarName,
      avatarSrc,
      avatarAlt
    } = renderAvatar;
    const label = avatarAlt || (typeof children === 'string' ? children : undefined);
    if (!label) {
      (0, _console.warn)(false, `Please supply a label for the avatar with either the "renderAvatar.avatarAlt" or the "children" (as string) prop. It is needed for screen reader support, but missing on the item with the id: "${id}".`);
    }
    return avatarName ? (0, _jsxRuntime.jsx)(_v11_.Avatar, {
      name: avatarName,
      src: avatarSrc,
      alt: label,
      size: "small"
    }) : null;
  }
  renderDropdownMenu() {
    const {
      id,
      renderSubmenu,
      status,
      styles
    } = this.props;
    if (!renderSubmenu || !this.shouldRenderSubmenu) {
      return null;
    }
    if (typeof renderSubmenu.props.show !== 'undefined') {
      (0, _console.warn)(false, `TopNavBar.Item Drilldown submenus are controlled by the component. The "show" prop will be ignored on the submenu of the item with id: "${id}".`);
    }
    if (renderSubmenu.props.trigger) {
      (0, _console.warn)(false, `TopNavBar.Item submenus have the item itself as their trigger. The "trigger" prop will be ignored on the Drilldown submenu of item with id: "${id}".`);
    }
    return (0, _safeCloneElement.safeCloneElement)(renderSubmenu, {
      trigger: (0, _jsxRuntime.jsx)("div", {
        css: styles?.submenuTriggerContainer,
        children: this.renderContent()
      }),
      show: this.state.isSubmenuOpen,
      onToggle: (0, _createChainedFunction.createChainedFunction)(renderSubmenu.props?.onToggle, this.handleMenuToggle),
      positionContainerDisplay: 'block',
      placement: 'bottom start',
      withArrow: false,
      minWidth: renderSubmenu.props?.minWidth || '18.5rem',
      maxHeight: renderSubmenu.props?.maxHeight || `calc(100vh - 10rem)`,
      // this is needed because the trigger is not a button. `aria-expanded`
      // is set on `get itemProps()`
      shouldSetAriaExpanded: false,
      ...(status === 'disabled' && {
        disabled: true,
        show: false,
        onToggle: undefined
      })
    });
  }
  renderPopover() {
    const {
      id,
      customPopoverConfig,
      styles
    } = this.props;
    if (!customPopoverConfig || !this.shouldRenderPopover) {
      return null;
    }

    // @ts-expect-error throw a warning just in case they pass trigger
    if (customPopoverConfig.renderTrigger) {
      (0, _console.warn)(false, `TopNavBar.Item popovers have the item itself as their trigger. The "renderTrigger" prop will be ignored on the popover of item with id: "${id}".`);
    }
    const popoverProps = {
      // defaults:
      placement: 'bottom start',
      withArrow: true,
      shouldReturnFocus: true,
      shouldContainFocus: false,
      // user input:
      ...customPopoverConfig,
      // fix overrides:
      positionContainerDisplay: 'block',
      renderTrigger: (0, _jsxRuntime.jsx)("div", {
        css: styles?.submenuTriggerContainer,
        children: this.renderContent()
      }),
      onShowContent: (0, _createChainedFunction.createChainedFunction)(customPopoverConfig.onShowContent, () => {
        if (!this.isCustomPopoverControlled) {
          this.setState({
            isPopoverOpen: true
          });
        }
      }),
      onHideContent: (0, _createChainedFunction.createChainedFunction)(customPopoverConfig.onHideContent, () => {
        if (!this.isCustomPopoverControlled) {
          this.setState({
            isPopoverOpen: false
          });
        }
      }),
      isShowingContent: this.state.isPopoverOpen,
      // this is needed because the trigger is not a button. `aria-expanded`
      // is set on `get itemProps()`
      shouldSetAriaExpanded: false
    };
    return (0, _jsxRuntime.jsx)(_v11_4.Popover, {
      ...popoverProps,
      children: customPopoverConfig.children
    });
  }
  render() {
    const {
      styles
    } = this.props;
    let content;
    if (this.shouldRenderPopover) {
      content = (0, _jsxRuntime.jsx)(_emotion.InstUISettingsProvider, {
        theme: {
          componentOverrides: {
            View: {
              // moves popovers above navbar, below tooltips
              stackingTopmost: submenuStacking
            }
          }
        },
        children: this.renderPopover()
      });
    } else if (this.shouldRenderSubmenu) {
      content = (0, _jsxRuntime.jsx)(_emotion.InstUISettingsProvider, {
        theme: {
          componentOverrides: {
            View: {
              // moves submenus above navbar, below tooltips
              stackingTopmost: submenuStacking
            }
          }
        },
        children: this.renderDropdownMenu()
      });
    } else {
      content = this.renderContent();
    }
    return (0, _jsxRuntime.jsx)("div", {
      ref: this.handleRef,
      css: styles?.topNavBarItem,
      "data-cid": "TopNavBarItem",
      children: content
    });
  }
}) || _class) || _class);
var _default = exports.default = TopNavBarItem;
