// @instructure/ui-utils/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "camelize", {
  enumerable: true,
  get: function () {
    return _camelize.camelize;
  }
});
Object.defineProperty(exports, "capitalizeFirstLetter", {
  enumerable: true,
  get: function () {
    return _capitalizeFirstLetter.capitalizeFirstLetter;
  }
});
Object.defineProperty(exports, "cloneArray", {
  enumerable: true,
  get: function () {
    return _cloneArray.cloneArray;
  }
});
Object.defineProperty(exports, "combineDataCid", {
  enumerable: true,
  get: function () {
    return _combineDataCid.combineDataCid;
  }
});
Object.defineProperty(exports, "createChainedFunction", {
  enumerable: true,
  get: function () {
    return _createChainedFunction.createChainedFunction;
  }
});
Object.defineProperty(exports, "deepEqual", {
  enumerable: true,
  get: function () {
    return _deepEqual.deepEqual;
  }
});
Object.defineProperty(exports, "generateId", {
  enumerable: true,
  get: function () {
    return _generateId.generateId;
  }
});
Object.defineProperty(exports, "getBrowser", {
  enumerable: true,
  get: function () {
    return _getBrowser.getBrowser;
  }
});
Object.defineProperty(exports, "hash", {
  enumerable: true,
  get: function () {
    return _hash.hash;
  }
});
Object.defineProperty(exports, "isAndroidOrIOS", {
  enumerable: true,
  get: function () {
    return _getBrowser.isAndroidOrIOS;
  }
});
Object.defineProperty(exports, "isBaseTheme", {
  enumerable: true,
  get: function () {
    return _isBaseTheme.isBaseTheme;
  }
});
Object.defineProperty(exports, "isChromium", {
  enumerable: true,
  get: function () {
    return _getBrowser.isChromium;
  }
});
Object.defineProperty(exports, "isEdge", {
  enumerable: true,
  get: function () {
    return _getBrowser.isEdge;
  }
});
Object.defineProperty(exports, "isEmpty", {
  enumerable: true,
  get: function () {
    return _isEmpty.isEmpty;
  }
});
Object.defineProperty(exports, "isFirefox", {
  enumerable: true,
  get: function () {
    return _getBrowser.isFirefox;
  }
});
Object.defineProperty(exports, "isIE", {
  enumerable: true,
  get: function () {
    return _getBrowser.isIE;
  }
});
Object.defineProperty(exports, "isMac", {
  enumerable: true,
  get: function () {
    return _getBrowser.isMac;
  }
});
Object.defineProperty(exports, "isSafari", {
  enumerable: true,
  get: function () {
    return _getBrowser.isSafari;
  }
});
Object.defineProperty(exports, "mergeDeep", {
  enumerable: true,
  get: function () {
    return _mergeDeep.mergeDeep;
  }
});
Object.defineProperty(exports, "ms", {
  enumerable: true,
  get: function () {
    return _ms.ms;
  }
});
Object.defineProperty(exports, "parseUnit", {
  enumerable: true,
  get: function () {
    return _parseUnit.parseUnit;
  }
});
Object.defineProperty(exports, "pascalize", {
  enumerable: true,
  get: function () {
    return _pascalize.pascalize;
  }
});
Object.defineProperty(exports, "px", {
  enumerable: true,
  get: function () {
    return _px.px;
  }
});
Object.defineProperty(exports, "shallowEqual", {
  enumerable: true,
  get: function () {
    return _shallowEqual.shallowEqual;
  }
});
Object.defineProperty(exports, "within", {
  enumerable: true,
  get: function () {
    return _within.within;
  }
});
var _capitalizeFirstLetter = require("./capitalizeFirstLetter");
var _cloneArray = require("./cloneArray");
var _createChainedFunction = require("./createChainedFunction");
var _deepEqual = require("./deepEqual");
var _hash = require("./hash");
var _generateId = require("./generateId");
var _isEmpty = require("./isEmpty");
var _mergeDeep = require("./mergeDeep");
var _ms = require("./ms");
var _parseUnit = require("./parseUnit");
var _px = require("./px");
var _shallowEqual = require("./shallowEqual");
var _within = require("./within");
var _camelize = require("./camelize");
var _pascalize = require("./pascalize");
var _isBaseTheme = require("./isBaseTheme");
var _getBrowser = require("./getBrowser");
var _combineDataCid = require("./combineDataCid");
