// @instructure/ui-tree-browser/es/TreeBrowser/v1/TreeNode/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2021 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { Img } from '@instructure/ui-img/v11_6';
import { callRenderProp } from '@instructure/ui-react-utils';
import { withStyle } from '@instructure/emotion';
import generateStyles from "../TreeButton/styles.js";
import generateComponentTheme from "../TreeButton/theme.js";
import { allowedProps } from "./props.js";
import TreeBrowserContext from "../TreeBrowserContext.js"; // Todo: merge TreeButton and TreeNode: TreeButton should be a special type of TreeNode
/**
---
parent: TreeBrowser
id: TreeBrowser.Node
---
A helper class used to render the :renderBeforeItems and :renderAfterItems
in the TreeBrowser.
**/
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
let TreeNode = (_dec = withStyle(generateStyles, generateComponentTheme), _dec(_class = class TreeNode extends Component {
  static displayName = "TreeNode";
  static componentId = 'TreeBrowser.Node';
  static allowedProps = allowedProps;
  static contextType = TreeBrowserContext;
  static defaultProps = {
    size: 'medium',
    variant: 'folderTree',
    selected: false,
    focused: false
  };
  ref = null;
  componentDidMount() {
    this.props.makeStyles?.({
      animation: this.context?.animation
    });
  }
  componentDidUpdate() {
    this.props.makeStyles?.({
      animation: this.context?.animation
    });
  }
  handleRef = el => {
    if (el && typeof this.props.containerRef === 'function') {
      this.props.containerRef(el.parentElement);
    }
    this.ref = el;
  };
  renderItemImage() {
    const {
      thumbnail,
      itemIcon,
      styles
    } = this.props;
    if (thumbnail) {
      return _jsx("div", {
        css: styles?.thumbnail,
        children: _jsx(Img, {
          src: thumbnail,
          constrain: "cover",
          alt: ""
        })
      });
    }
    if (itemIcon) {
      return _jsx("div", {
        css: styles?.icon,
        children: callRenderProp(itemIcon)
      });
    }
    return undefined;
  }
  render() {
    const {
      children,
      styles
    } = this.props;
    return _jsx("div", {
      ref: this.handleRef,
      tabIndex: -1,
      css: styles?.treeButton,
      "data-cid": "TreeNode",
      children: _jsxs("span", {
        css: styles?.layout,
        children: [this.renderItemImage(), _jsx("span", {
          css: styles?.node,
          children: children
        })]
      })
    });
  }
}) || _class);
export default TreeNode;
export { TreeNode };
