// @instructure/ui-themes/es/themes/newThemeTokens/legacyCanvasHighContrast/index.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import primitives from "./primitives.js";
import semantics from "./semantics.js";
import sharedTokens from "./sharedTokens.js";
import aiInformation from "./components/aiInformation.js";
import alert from "./components/alert.js";
import appNav from "./components/appNav.js";
import appNavItem from "./components/appNavItem.js";
import avatar from "./components/avatar.js";
import badge from "./components/badge.js";
import baseButton from "./components/baseButton.js";
import billboard from "./components/billboard.js";
import breadcrumb from "./components/breadcrumb.js";
import byline from "./components/byline.js";
import calendar from "./components/calendar.js";
import calendarDay from "./components/calendarDay.js";
import checkbox from "./components/checkbox.js";
import colorContrast from "./components/colorContrast.js";
import colorIndicator from "./components/colorIndicator.js";
import colorPicker from "./components/colorPicker.js";
import colorPreset from "./components/colorPreset.js";
import contextView from "./components/contextView.js";
import dataPermissionLevels from "./components/dataPermissionLevels.js";
import drawerLayoutContent from "./components/drawerLayoutContent.js";
import drawerLayoutTray from "./components/drawerLayoutTray.js";
import drilldown from "./components/drilldown.js";
import fileDrop from "./components/fileDrop.js";
import flex from "./components/flex.js";
import formFieldLayout from "./components/formFieldLayout.js";
import formFieldMessage from "./components/formFieldMessage.js";
import grid from "./components/grid.js";
import gridCol from "./components/gridCol.js";
import gridRow from "./components/gridRow.js";
import heading from "./components/heading.js";
import icon from "./components/icon.js";
import img from "./components/img.js";
import inlineListItem from "./components/inlineListItem.js";
import link from "./components/link.js";
import list from "./components/list.js";
import listItem from "./components/listItem.js";
import mask from "./components/mask.js";
import menu from "./components/menu.js";
import menuGroup from "./components/menuGroup.js";
import menuItem from "./components/menuItem.js";
import menuSeparator from "./components/menuSeparator.js";
import metric from "./components/metric.js";
import modal from "./components/modal.js";
import modalBody from "./components/modalBody.js";
import modalFooter from "./components/modalFooter.js";
import modalHeader from "./components/modalHeader.js";
import nutritionFacts from "./components/nutritionFacts.js";
import options from "./components/options.js";
import optionsItem from "./components/optionsItem.js";
import optionsSeparator from "./components/optionsSeparator.js";
import pagination from "./components/pagination.js";
import paginationPageInput from "./components/paginationPageInput.js";
import palette from "./components/palette.js";
import pill from "./components/pill.js";
import popover from "./components/popover.js";
import progressBar from "./components/progressBar.js";
import progressCircle from "./components/progressCircle.js";
import radioInput from "./components/radioInput.js";
import rangeInput from "./components/rangeInput.js";
import ratingIcon from "./components/ratingIcon.js";
import rgbaInput from "./components/rgbaInput.js";
import select from "./components/select.js";
import sideNavBar from "./components/sideNavBar.js";
import sideNavBarItem from "./components/sideNavBarItem.js";
import sideNavBarNew from "./components/sideNavBarNew.js";
import slider from "./components/slider.js";
import sourceCodeEditor from "./components/sourceCodeEditor.js";
import spinner from "./components/spinner.js";
import table from "./components/table.js";
import tableBody from "./components/tableBody.js";
import tableCell from "./components/tableCell.js";
import tableColHeader from "./components/tableColHeader.js";
import tableHead from "./components/tableHead.js";
import tableRow from "./components/tableRow.js";
import tableRowHeader from "./components/tableRowHeader.js";
import tabs from "./components/tabs.js";
import tabsPanel from "./components/tabsPanel.js";
import tabsTab from "./components/tabsTab.js";
import tag from "./components/tag.js";
import text from "./components/text.js";
import textArea from "./components/textArea.js";
import textInput from "./components/textInput.js";
import toggle from "./components/toggle.js";
import toggleDetails from "./components/toggleDetails.js";
import toggleGroup from "./components/toggleGroup.js";
import tooltip from "./components/tooltip.js";
import topNavBarActionItems from "./components/topNavBarActionItems.js";
import topNavBarBrand from "./components/topNavBarBrand.js";
import topNavBarItem from "./components/topNavBarItem.js";
import topNavBarLayout from "./components/topNavBarLayout.js";
import topNavBarMenuItems from "./components/topNavBarMenuItems.js";
import tray from "./components/tray.js";
import treeBrowser from "./components/treeBrowser.js";
import treeBrowserTreeButton from "./components/treeBrowserTreeButton.js";
import treeBrowserTreeCollection from "./components/treeBrowserTreeCollection.js";
import truncateText from "./components/truncateText.js";
import view from "./components/view.js";
// TODO-theme-types: fix any
const theme = {
  primitives,
  semantics,
  sharedTokens,
  components: {
    Avatar: avatar,
    Breadcrumb: breadcrumb,
    Metric: metric,
    Pill: pill,
    FormFieldMessage: formFieldMessage,
    Spinner: spinner,
    Link: link,
    TextInput: textInput,
    TextArea: textArea,
    Icon: icon,
    Checkbox: checkbox,
    Text: text,
    RadioInput: radioInput,
    Toggle: toggle,
    Tooltip: tooltip,
    Modal: modal,
    ModalHeader: modalHeader,
    ModalFooter: modalFooter,
    ModalBody: modalBody,
    Mask: mask,
    FormFieldLayout: formFieldLayout,
    Heading: heading,
    Tray: tray,
    ContextView: contextView,
    View: view,
    ProgressBar: progressBar,
    Tabs: tabs,
    TabsPanel: tabsPanel,
    TabsTab: tabsTab,
    RangeInput: rangeInput,
    FileDrop: fileDrop,
    Popover: popover,
    Badge: badge,
    Tag: tag,
    Calendar: calendar,
    CalendarDay: calendarDay,
    Grid: grid,
    GridCol: gridCol,
    GridRow: gridRow,
    InlineListItem: inlineListItem,
    Menu: menu,
    MenuGroup: menuGroup,
    MenuItem: menuItem,
    MenuSeparator: menuSeparator,
    Drilldown: drilldown,
    BaseButton: baseButton,
    ListItem: listItem,
    List: list,
    Table: table,
    TableBody: tableBody,
    TableHead: tableHead,
    TableCell: tableCell,
    TableRowHeader: tableRowHeader,
    TableRow: tableRow,
    TableColHeader: tableColHeader,
    Options: options,
    OptionsItem: optionsItem,
    OptionsSeparator: optionsSeparator,
    Flex: flex,
    SourceCodeEditor: sourceCodeEditor,
    TreeBrowser: treeBrowser,
    TreeBrowserTreeButton: treeBrowserTreeButton,
    TreeBrowserTreeCollection: treeBrowserTreeCollection,
    ToggleDetails: toggleDetails,
    Alert: alert,
    SideNavBar: sideNavBar,
    SideNavBarItem: sideNavBarItem,
    ToggleGroup: toggleGroup,
    Billboard: billboard,
    AiInformation: aiInformation,
    DataPermissionLevels: dataPermissionLevels,
    NutritionFacts: nutritionFacts,
    RatingIcon: ratingIcon,
    ProgressCircle: progressCircle,
    TopNavBarActionItems: topNavBarActionItems,
    TopNavBarLayout: topNavBarLayout,
    TopNavBarBrand: topNavBarBrand,
    TopNavBarItem: topNavBarItem,
    TopNavBarMenuItems: topNavBarMenuItems,
    Select: select,
    Pagination: pagination,
    PaginationPageInput: paginationPageInput,
    DrawerLayoutContent: drawerLayoutContent,
    DrawerLayoutTray: drawerLayoutTray,
    ColorPicker: colorPicker,
    ColorPreset: colorPreset,
    ColorIndicator: colorIndicator,
    ColorContrast: colorContrast,
    Palette: palette,
    Slider: slider,
    RgbaInput: rgbaInput,
    TruncateText: truncateText,
    AppNav: appNav,
    AppNavItem: appNavItem,
    Byline: byline,
    SideNavBarNew: sideNavBarNew,
    Img: img
  }
};
export default theme;
