// @instructure/ui-drilldown/lib/Drilldown/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Drilldown = void 0;
require("core-js/modules/es.array.includes.js");
require("core-js/modules/es.map.get-or-insert.js");
require("core-js/modules/es.map.get-or-insert-computed.js");
var _react = require("react");
var _console = require("@instructure/console");
var _contains = require("@instructure/ui-dom-utils/lib/contains.js");
var _containsActiveElement = require("@instructure/ui-dom-utils/lib/containsActiveElement.js");
var _deepEqual = require("@instructure/ui-utils/lib/deepEqual.js");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _v11_ = require("@instructure/ui-view/v11_6");
var _v11_2 = require("@instructure/ui-options/v11_6");
var _v11_3 = require("@instructure/ui-popover/v11_6");
var _Selectable = require("@instructure/ui-selectable/lib/Selectable");
var _IconArrowOpenStartSolid = require("@instructure/ui-icons/lib/generated/IconArrowOpenStartSolid.js");
var _IconArrowOpenEndSolid = require("@instructure/ui-icons/lib/generated/IconArrowOpenEndSolid.js");
var _IconCheckSolid = require("@instructure/ui-icons/lib/generated/IconCheckSolid.js");
var _emotion = require("@instructure/emotion");
var _DrilldownSeparator = require("./DrilldownSeparator");
var _DrilldownOption = require("./DrilldownOption");
var _DrilldownGroup = require("./DrilldownGroup");
var _DrilldownPage = require("./DrilldownPage");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _react2 = require("@emotion/react");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
// Additional data we need to track on the Options,
// but shouldn't be settable via props
// Contains the Option ComponentElement and the extra data we need track on it,
// but don't want to expose as props
// Contains the props object of the Page
// with the `children` transformed into an array
// An object with the mapped Pages with their id as keys
/**
---
category: components
---
**/
let Drilldown = exports.Drilldown = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = _dec2(_class = class Drilldown extends _react.Component {
  static displayName = "Drilldown";
  static componentId = 'Drilldown';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    disabled: false,
    rotateFocus: true,
    as: 'ul',
    role: 'menu',
    overflowX: 'auto',
    overflowY: 'auto',
    placement: 'bottom center',
    defaultShow: false,
    shouldHideOnSelect: true,
    shouldContainFocus: false,
    shouldReturnFocus: true,
    withArrow: true,
    offsetX: 0,
    offsetY: 0,
    shouldSetAriaExpanded: true
  };
  static Group = _DrilldownGroup.DrilldownGroup;
  static Option = _DrilldownOption.DrilldownOption;
  static Page = _DrilldownPage.DrilldownPage;
  static Separator = _DrilldownSeparator.DrilldownSeparator;
  _drilldownRef = null;
  _popover = null;
  _trigger = null;
  _containerElement = null;
  _id;
  _triggerId;
  _headerBackId;
  _headerTitleId;
  _headerTitleLabelId;
  _headerActionId;

  // Array of visited page ids in "breadcrumbs" fashion
  _pageHistory;

  // Map of the active options on the page (includes header options too)
  _activeOptionsMap = {};
  ref = null;
  handleRef = el => {
    // ref and elementRef have to be set together for the same element
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  handleDrilldownRef = el => {
    const {
      drilldownRef
    } = this.props;
    this._drilldownRef = el;
    if (typeof drilldownRef === 'function') {
      drilldownRef(el);
    }

    // setting ref for "non-popover" version, the drilldown itself
    // (if a trigger is provided, the Popover sets the ref)
    if (!this.props.trigger) {
      this.handleRef(el);
    }
  };
  constructor(props) {
    super(props);
    this.state = {
      isShowingPopover: props.trigger ? !!props.show : false,
      activePageId: props.rootPageId,
      highlightedOptionId: undefined,
      lastSelectedId: undefined,
      selectedGroupOptionsMap: this.getSelectedGroupOptionsMap()
    };
    this._pageHistory = [props.rootPageId];
    this._id = props.id || props.deterministicId();
    this._triggerId = props.deterministicId('Drilldown-Trigger');
    this._headerBackId = props.deterministicId('DrilldownHeader-Back');
    this._headerTitleId = props.deterministicId('DrilldownHeader-Title');
    this._headerTitleLabelId = props.deterministicId('DrilldownHeader-Title-Label');
    this._headerActionId = props.deterministicId('DrilldownHeader-Action');
  }
  componentDidMount() {
    this.props.makeStyles?.(this.makeStylesVariables);
  }
  componentDidUpdate(_prevProps, prevState) {
    this.props.makeStyles?.(this.makeStylesVariables);
    if (prevState.activePageId !== this.state.activePageId) {
      // on page change with mouse navigation, some option can get rendered
      // under the cursor and get focused, so we need to wait a tick to see
      // if anything gets focused, otherwise we focus the whole drilldown
      setTimeout(() => {
        if (!this.focused()) {
          this.focus();
        }
      }, 0);
    }

    // if the current page was removed
    if (!this.currentPage) {
      if (this.previousPage) {
        this.goToPreviousPage();
      } else {
        this.goToPage(this.props.rootPageId);
      }
    }
    if (this.state.highlightedOptionId && !this.getPageChildById(this.state.highlightedOptionId)) {
      this.setState({
        highlightedOptionId: undefined
      });
    }

    // This block syncs the internal state if the selectedOptions prop changes
    // selectedOptions prop values always win over internal state.
    let shouldUpdateState = false;
    const nextSelectionMap = {
      ...this.state.selectedGroupOptionsMap
    };
    this.pages.forEach(page => {
      this.getChildrenArray(page.props.children).forEach(child => {
        if ((0, _matchComponentTypes.matchComponentTypes)(child, [_DrilldownGroup.DrilldownGroup])) {
          const {
            id: groupId,
            selectableType,
            selectedOptions,
            children: groupChildren
          } = child.props;
          if (!selectableType || !Array.isArray(selectedOptions)) return;
          const newGroupMap = new Map();
          this.getChildrenArray(groupChildren)?.forEach(groupChild => {
            if ((0, _matchComponentTypes.matchComponentTypes)(groupChild, [_DrilldownOption.DrilldownOption])) {
              const {
                id: optionId,
                value: optionValue
              } = groupChild.props;
              if (selectedOptions.includes(optionValue)) {
                newGroupMap.set(optionId, optionValue);
              }
            }
          });
          const currentGroupMapInState = this.state.selectedGroupOptionsMap[groupId] || new Map();
          if (!(0, _deepEqual.deepEqual)(this.getNormalizedMap(newGroupMap), this.getNormalizedMap(currentGroupMapInState))) {
            nextSelectionMap[groupId] = newGroupMap;
            shouldUpdateState = true;
          }
        }
      });
    });
    if (shouldUpdateState) {
      this.setState({
        selectedGroupOptionsMap: nextSelectionMap
      });
    }
  }
  get makeStylesVariables() {
    return {
      hasHighlightedOption: !!this.state.highlightedOptionId
    };
  }
  get activeOptionIds() {
    const orderedKeys = Object.keys(this._activeOptionsMap).sort((a, b) => {
      return this._activeOptionsMap[a].index - this._activeOptionsMap[b].index;
    });
    return orderedKeys;
  }
  get activeOptions() {
    return Object.values(this._activeOptionsMap);
  }
  get pages() {
    const {
      children
    } = this.props;
    return _react.Children.toArray(children || []);
  }
  get pageMap() {
    const {
      children
    } = this.props;
    if (!children) {
      return undefined;
    }
    const map = {};
    this.pages.forEach(page => {
      const {
        props
      } = page;
      const {
        children
      } = props;
      map[props.id] = {
        ...props,
        children: _react.Children.toArray(children || [])
      };
    });
    return map;
  }
  get isOnRootPage() {
    return this.state.activePageId === this.props.rootPageId;
  }
  get currentPage() {
    return this.getPageById(this.state.activePageId);
  }
  get previousPage() {
    const previousPageId = this._pageHistory[this._pageHistory.length - 2];
    return this.getPageById(previousPageId);
  }

  // Returns the navigation methods and the page history.
  // These are used in some callbacks to expose the navigation logic.
  get exposedNavigationProps() {
    const {
      goToPage,
      goToPreviousPage
    } = this;

    // we make a copy of the array so the original history
    // cannot be modified from the outside
    const pageHistory = [...this._pageHistory];
    return {
      pageHistory,
      goToPage,
      goToPreviousPage
    };
  }
  get currentPageAriaLabel() {
    // return, if explicitly set
    if (this.props['aria-labelledby']) {
      return this.props['aria-labelledby'];
    }

    // if it has title, point to the title label content
    if (this.currentPage?.renderTitle) {
      return this._headerTitleLabelId;
    }

    // if root page has no title, try the trigger, if exists
    if (this.isOnRootPage && this.props.trigger) {
      return this._triggerId;
    }
    return undefined;
  }
  getChildrenArray(children) {
    if (!children) {
      return [];
    }
    return Array.isArray(children) ? children : [children];
  }
  getPageById(id) {
    return this.pageMap && id ? this.pageMap[id] : undefined;
  }
  getPageChildById(id) {
    return id ? this._activeOptionsMap[id] : undefined;
  }
  getNormalizedMap(map) {
    return Array.from(map.entries());
  }

  /**
   * Initializes the map of selected options for each group on initial render.
   *
   * This function establishes the selection state based on a clear precedence:
   * 1. **Controlled Mode**: If a `Drilldown.Group` has the `selectedOptions` prop
   * (as an array), it is treated as the absolute source of truth.
   *
   * 2. **Uncontrolled Mode**: If `selectedOptions` is not provided, the selection
   * is determined by `defaultSelected` props, with the following priority:
   * a. The `defaultSelected` boolean on an individual `Drilldown.Option`.
   * b. The `defaultSelected` array of values on the `Drilldown.Group`.
   *
   * It also validates that 'single' selection groups do not mistakenly receive
   * multiple selected or default values.
   *
   * @returns {SelectedGroupOptionsMap} An object where keys are group IDs and
   * values are Maps of the selected { optionId: optionValue } pairs for that group.
   */
  getSelectedGroupOptionsMap() {
    const selectedGroupOptionsMap = {};
    this.pages.forEach(page => {
      const {
        children
      } = page.props;
      this.getChildrenArray(children).forEach(child => {
        if ((0, _matchComponentTypes.matchComponentTypes)(child, [_DrilldownGroup.DrilldownGroup])) {
          const {
            id: groupId,
            selectableType,
            selectedOptions,
            defaultSelected = [],
            children: groupChildren
          } = child.props;
          if (!selectableType) return;
          selectedGroupOptionsMap[groupId] = new Map();
          if (selectableType === 'single') {
            if (Array.isArray(selectedOptions) && selectedOptions.length > 1) {
              (0, _console.error)(false, `Radio type selectable groups can have only one item selected! Group "${groupId}" has multiple values: [${selectedOptions.join(', ')}]!`);
              return;
            }
            if (defaultSelected.length > 1) {
              (0, _console.error)(false, `Radio type selectable groups can have only one default item selected! Group "${groupId}" has multiple values: [${defaultSelected.join(', ')}]!`);
              return;
            }
          }
          this.getChildrenArray(groupChildren)?.forEach(groupChild => {
            if ((0, _matchComponentTypes.matchComponentTypes)(groupChild, [_DrilldownOption.DrilldownOption])) {
              const {
                id: optionId,
                value: optionValue,
                defaultSelected: optionDefaultSelected
              } = groupChild.props;
              if (optionValue == null) return;
              let isSelected = false;
              // If the group is controlled via the `selectedOptions` prop, it is the source of truth.
              // `defaultSelected` values are ignored.
              if (Array.isArray(selectedOptions)) {
                isSelected = selectedOptions.includes(optionValue);
                // Second strongest is the `defaultSelected` value on the option itself.
              } else if (optionDefaultSelected === false) {
                isSelected = false;
              } else {
                // Last is the `defaultSelected` array on the group.
                const isGroupDefaultSelected = defaultSelected.includes(optionValue);
                isSelected = optionDefaultSelected || isGroupDefaultSelected;
              }
              if (isSelected) {
                selectedGroupOptionsMap[groupId].set(optionId, optionValue);
              }
            }
          });
        }
      });
    });
    return selectedGroupOptionsMap;
  }
  get selectedGroupOptionIdsArray() {
    return Object.values(this.state.selectedGroupOptionsMap).map(groupIdMap => Array.from(groupIdMap.keys())).flat();
  }
  get headerChildren() {
    const {
      currentPage
    } = this;
    const {
      styles,
      deterministicId
    } = this.props;
    const headerChildren = [];
    if (!currentPage) {
      return headerChildren;
    }
    const {
      children,
      renderBackButtonLabel,
      renderTitle,
      renderActionLabel,
      onHeaderActionClicked,
      withoutHeaderSeparator
    } = currentPage;

    // Back navigation option
    if (this.previousPage) {
      const prevPageTitle = (0, _callRenderProp.callRenderProp)(this.previousPage.renderTitle);
      const backButtonLabel = (0, _callRenderProp.callRenderProp)(renderBackButtonLabel, prevPageTitle);
      headerChildren.push((0, _jsxRuntime.jsx)(_DrilldownOption.DrilldownOption, {
        id: this._headerBackId,
        onOptionClick: this.handleBackButtonClick,
        children: (0, _jsxRuntime.jsx)("div", {
          css: styles?.headerBack,
          role: "presentation",
          children: backButtonLabel
        })
      }));
    }

    // Header title
    if (renderTitle) {
      const title = (0, _callRenderProp.callRenderProp)(renderTitle);
      if (title) {
        headerChildren.push((0, _jsxRuntime.jsx)(_DrilldownOption.DrilldownOption, {
          id: this._headerTitleId,
          role: "presentation",
          "aria-hidden": "true",
          children: (0, _jsxRuntime.jsx)("div", {
            css: styles?.headerTitle,
            id: this._headerTitleLabelId,
            children: title
          })
        }));
      }
    }

    // Action label
    if (renderActionLabel) {
      const actionLabel = (0, _callRenderProp.callRenderProp)(renderActionLabel);
      if (actionLabel) {
        headerChildren.push((0, _jsxRuntime.jsx)(_DrilldownOption.DrilldownOption, {
          id: this._headerActionId,
          themeOverride: {
            color: styles?.headerActionColor
          },
          onOptionClick: event => {
            if (typeof onHeaderActionClicked === 'function') {
              onHeaderActionClicked(event);
            }
          },
          children: actionLabel
        }));
      }
    }

    // header separator
    if (headerChildren.length > 0 && children.length > 0 && !withoutHeaderSeparator) {
      headerChildren.push((0, _jsxRuntime.jsx)(_DrilldownSeparator.DrilldownSeparator, {
        id: deterministicId('DrilldownHeader-Separator')
      }));
    }
    return headerChildren;
  }
  get shown() {
    return this.props.trigger ? this.state.isShowingPopover : true;
  }
  containsDuplicateChild(children) {
    let containsDuplicate = false;
    const childMap = new Map();
    for (const child of children) {
      if (child && typeof child === 'object' && child.props?.id) {
        if (!childMap.has(child.props.id)) {
          childMap.set(child.props.id, true);
        } else {
          (0, _console.warn)(false, `Duplicate id: "${child.props.id}"! Make sure all options have unique ids, otherwise they won't be rendered.`);
          return containsDuplicate = true;
        }
      }
    }
    return containsDuplicate;
  }
  show = event => {
    if (this._popover) {
      this._popover.show(event);
      this.setState({
        isShowingPopover: true
      });
    }
  };
  hide = event => {
    if (this._popover) {
      this._popover.hide(event);
      this.setState({
        isShowingPopover: false
      });
      this.reset();
    }
  };
  reset() {
    this._activeOptionsMap = {};
    this.setState({
      highlightedOptionId: undefined
    });
  }
  focus() {
    if (this.shown) {
      (0, _console.error)(!!this._drilldownRef?.focus, '[Drilldown] Could not focus the drilldown.');
      this._drilldownRef?.focus();
    } else {
      (0, _console.error)(!!this._trigger?.focus, '[Drilldown] Could not focus the trigger.');
      this._trigger.focus();
    }
  }
  focused() {
    if (this.shown) {
      return (0, _containsActiveElement.containsActiveElement)(this._drilldownRef);
    } else {
      return (0, _containsActiveElement.containsActiveElement)(this._trigger);
    }
  }
  focusOption(id) {
    const container = this._containerElement;
    const optionElement = container?.querySelector(`[id="${CSS.escape(id)}"]`);
    optionElement?.focus();
  }
  handleOptionHighlight = (_event, {
    id,
    direction
  }) => {
    const {
      rotateFocus
    } = this.props;
    const {
      highlightedOptionId
    } = this.state;

    // if id exists, use that, or calculate from direction
    let highlightId = this.getPageChildById(id) ? id : undefined;
    if (!highlightId) {
      if (!highlightedOptionId) {
        // nothing highlighted yet, highlight first option
        highlightId = this.activeOptionIds[0];
      } else if (direction) {
        // if it has direction, find next id based on it
        const index = this.activeOptionIds.indexOf(highlightedOptionId);
        const newIndex = index + direction;
        highlightId = index > -1 ? this.activeOptionIds[newIndex] : undefined;
        if (rotateFocus) {
          const lastOptionsIndex = this.activeOptionIds.length - 1;
          if (newIndex < 0) {
            highlightId = this.activeOptionIds[lastOptionsIndex];
          }
          if (newIndex > lastOptionsIndex) {
            highlightId = this.activeOptionIds[0];
          }
        }
      }
    }
    if (highlightId) {
      // only highlight if id exists as a valid option
      this.setState({
        highlightedOptionId: highlightId
      }, () => {
        this.focusOption(highlightId);
      });
    }
  };

  // Navigates to the page and also returns the new and old pageIds
  goToPage = newPageId => {
    if (!newPageId) {
      (0, _console.warn)(false, `Cannot go to page because there was no page id provided.`);
      return undefined;
    }

    // TS complains that it cannot be true, but since it is an exposed method,
    // it is better if we provide a warning for this case too
    if (typeof newPageId !== 'string') {
      (0, _console.warn)(false, `Cannot go to page because parameter newPageId has to be string (valid page id). Current newPageId is "${typeof newPageId}".`);
      return undefined;
    }
    if (!this.pageMap?.[newPageId]) {
      (0, _console.warn)(false, `Cannot go to page because page with id: "${newPageId}" doesn't exist.`);
      return undefined;
    }

    // the last page id in the history is the current one,
    // it will become the "prevPage"
    const prevPageId = this._pageHistory[this._pageHistory.length - 1];
    const idxInHistory = this._pageHistory.indexOf(newPageId);
    if (idxInHistory < 0) {
      // if it is not in the page history, we have to add it
      this._pageHistory.push(newPageId);
    } else {
      // if it was already in the history, we go back to that page,
      // and clear the rest from the history
      this._pageHistory.splice(idxInHistory + 1, this._pageHistory.length - 1);
    }
    this.setState({
      activePageId: newPageId,
      highlightedOptionId: undefined
    });
    return {
      prevPageId,
      newPageId
    };
  };
  goToPreviousPage = () => {
    if (!this.previousPage) {
      (0, _console.warn)(false, `There is no previous page to go to. The current page history is: [${this._pageHistory.join(', ')}].`);
      return undefined;
    }

    // If there is a previous page, goToPage will succeed and return the data
    const {
      newPageId,
      prevPageId
    } = this.goToPage(this.previousPage.id);
    return {
      newPageId,
      prevPageId
    };
  };
  handleBackButtonClick = () => {
    const {
      onBackButtonClicked
    } = this.currentPage;

    // we only display the back button when there is a page to go back to
    const {
      newPageId,
      prevPageId
    } = this.goToPreviousPage();
    if (typeof onBackButtonClicked === 'function') {
      onBackButtonClicked(newPageId, prevPageId);
    }
  };
  handleGroupOptionSelected(event, selectedOption) {
    this.setState(oldState => {
      const {
        id: optionId,
        value
      } = selectedOption.props;
      const {
        id: groupId,
        selectableType
      } = selectedOption.groupProps;
      let newState = new Map(oldState.selectedGroupOptionsMap[groupId]);
      if (selectableType === 'multiple' && Boolean(oldState.selectedGroupOptionsMap[groupId]?.has(optionId))) {
        // toggle off, if already selected
        newState.delete(optionId);
      } else {
        if (selectableType === 'multiple') {
          // "checkbox"
          newState.set(optionId, value);
        } else if (selectableType === 'single') {
          // "radio"
          newState = new Map();
          newState.set(optionId, value);
        }
      }
      return {
        ...oldState,
        selectedGroupOptionsMap: {
          ...oldState.selectedGroupOptionsMap,
          [groupId]: newState
        }
      };
    }, () => {
      const {
        value
      } = selectedOption.props;
      const {
        id: groupId,
        onSelect: groupOnSelect
      } = selectedOption.groupProps;
      const {
        onSelect
      } = this.props;
      const {
        groupProps,
        ...option
      } = selectedOption;
      const selectedOptionValuesInGroup = [...this.state.selectedGroupOptionsMap[groupId].values()];
      if (typeof groupOnSelect === 'function') {
        groupOnSelect(event, {
          value: selectedOptionValuesInGroup,
          isSelected: selectedOptionValuesInGroup.includes(value),
          selectedOption: option,
          drilldown: this
        });
      }
      if (typeof onSelect === 'function') {
        onSelect(event, {
          value: selectedOptionValuesInGroup,
          isSelected: selectedOptionValuesInGroup.includes(value),
          selectedOption: option,
          drilldown: this
        });
      }
    });
  }
  handleOptionSelect = (event, {
    id
  }) => {
    const selectedOption = this.getPageChildById(id);
    if (!id || !selectedOption) {
      event.preventDefault();
      event.stopPropagation();
      return;
    }
    const isOptionDisabled = id !== this._headerBackId && (this.props.disabled || this.currentPage?.disabled || selectedOption.groupProps?.disabled || selectedOption.props.disabled);
    if (isOptionDisabled) {
      event.preventDefault();
      event.stopPropagation();
      return;
    }
    const {
      shouldHideOnSelect,
      onSelect
    } = this.props;
    const {
      groupProps,
      ...selectedOptionChild
    } = selectedOption;
    const {
      subPageId,
      href,
      value,
      onOptionClick
    } = selectedOptionChild.props;
    if (typeof onOptionClick === 'function') {
      onOptionClick(event, {
        optionId: id,
        drilldown: this,
        ...this.exposedNavigationProps
      });
    }
    if (subPageId) {
      this.goToPage(subPageId);
    }
    if (event.type === 'keydown' && href) {
      const optionEl = this._drilldownRef?.querySelector(`#${CSS.escape(id)}`);
      const isLink = optionEl.tagName.toLowerCase() === 'a';

      // we need this check because in some cases
      // we ignore href prop in renderOption()
      if (isLink && optionEl?.href) {
        optionEl.click();
      }
    }
    if (groupProps?.selectableType) {
      this.handleGroupOptionSelected(event, selectedOption);
    } else {
      // TODO workaround for react 19 default props
      const optionWithDefaultProps = {
        ...selectedOptionChild,
        props: {
          ...selectedOptionChild.props,
          role: 'menuitem'
        }
      };
      if (typeof onSelect === 'function') {
        onSelect(event, {
          value,
          isSelected: true,
          selectedOption: optionWithDefaultProps,
          drilldown: this
        });
      }
    }

    // should prevent closing on page navigation
    if (shouldHideOnSelect && !subPageId && id !== this._headerBackId) {
      this.hide(event);
    }
  };

  // Setting extra logic for keyboard navigation.
  // Enter, Esc and up/down/home/end keys are handled by Selectable.
  handleKeyDown = event => {
    const id = event.target.id;
    const option = this.getPageChildById(id);

    // On Space...
    if ([' ', 'space', 'Space'].includes(event.key)) {
      // we need to preventDefault so the page doesn't scroll on Space
      event.preventDefault();
      event.stopPropagation();

      // for options, Space it has to work as Enter (get selected)
      if (option) {
        this.handleOptionSelect(event, {
          id
        });
      }
    }

    // On Right arrow...
    if (event.key === 'ArrowRight') {
      // if the option has subpage, we open it
      if (option?.props.subPageId) {
        this.handleOptionSelect(event, {
          id
        });
      }
    }

    // On Left arrow...
    if (event.key === 'ArrowLeft') {
      // if it is possible, we go a level up in the history
      if (this._pageHistory.length > 1) {
        this.handleBackButtonClick();
      }

      // if on root page and popover is open, we close it
      if (this.isOnRootPage && this._popover?.shown) {
        this._popover.hide(event);
      }
    }
  };
  handleToggle = (event, shown) => {
    const {
      onToggle,
      trigger
    } = this.props;
    if (trigger && shown && this.currentPage) {
      const actionLabel = (0, _callRenderProp.callRenderProp)(this.currentPage.renderActionLabel);
      // Use action ID if exists, otherwise first non-action option's ID
      const targetId = actionLabel ? this._headerActionId : this.getFirstOption()?.props.id;
      setTimeout(() => {
        this.setState({
          highlightedOptionId: targetId
        });
      }, 10);
    }
    this.setState({
      isShowingPopover: shown
    });
    if (typeof onToggle === 'function') {
      onToggle(event, {
        shown,
        drilldown: this,
        ...this.exposedNavigationProps
      });
    }
  };
  getFirstOption = () => {
    const children = _react.Children.toArray(this.currentPage?.children);
    const child = children[0];
    if (!child) return undefined;

    // If it's a regular option, return it
    if ((0, _matchComponentTypes.matchComponentTypes)(child, [_DrilldownOption.DrilldownOption])) {
      return child;
    }
    // If it's a group, get its options
    if ((0, _matchComponentTypes.matchComponentTypes)(child, [Drilldown.Group])) {
      const groupOptions = _react.Children.toArray(child.props.children);
      return groupOptions[0];
    }
    return undefined;
  };
  renderList(getOptionProps, getDisabledOptionProps) {
    const {
      currentPage,
      headerChildren
    } = this;
    if (!currentPage || this.containsDuplicateChild(currentPage.children)) {
      return null;
    }
    const pageChildren = [...headerChildren, ...currentPage.children];

    // for tracking if the last item was a Separator or not
    let lastItemWasSeparator = false;
    return pageChildren.map((child, index) => {
      /**
       * --- RENDER GROUP ---
       */
      if ((0, _matchComponentTypes.matchComponentTypes)(child, [_DrilldownGroup.DrilldownGroup])) {
        const isFirstChild = index === 0;
        const isLastChild = index === pageChildren.length - 1;
        const afterSeparator = lastItemWasSeparator;
        const needsFirstSeparator = !child.props.withoutSeparators && !isFirstChild && !afterSeparator;
        const needsLastSeparator = !child.props.withoutSeparators && !isLastChild;
        lastItemWasSeparator = needsLastSeparator;
        return this.renderGroup(child, getOptionProps, getDisabledOptionProps,
        // for rendering separators appropriately
        needsFirstSeparator, needsLastSeparator);

        /**
         * --- RENDER SEPARATOR ---
         */
      } else if ((0, _matchComponentTypes.matchComponentTypes)(child, [_DrilldownSeparator.DrilldownSeparator])) {
        // if the last item was separator, we don't want to duplicate it
        if (lastItemWasSeparator) {
          return null;
        }
        lastItemWasSeparator = true;
        return this.renderSeparator(child);

        /**
         * --- RENDER OPTION ---
         */
      } else if ((0, _matchComponentTypes.matchComponentTypes)(child, [_DrilldownOption.DrilldownOption])) {
        lastItemWasSeparator = false;
        return this.renderOption(child, getOptionProps, getDisabledOptionProps);
      } else {
        return null;
      }
    });
  }
  renderSeparator(separator) {
    const {
      id,
      themeOverride,
      ...props
    } = separator.props;
    return (0, _react2.createElement)(_v11_2.Options.Separator, {
      ...props,
      id: id,
      key: id,
      role: "separator"
      // we pass the themeOverride to Options.Separator
      ,
      themeOverride: themeOverride
    });
  }
  renderOption(option, getOptionProps, getDisabledOptionProps, groupProps) {
    const {
      styles
    } = this.props;
    let isSelected = false;
    const {
      id,
      children,
      href,
      as = 'li',
      // workaround after the react 19 upgrade, defaultProps doesn't work anymore
      role = 'menuitem',
      // workaround after the react 19 upgrade, defaultProps doesn't work anymore
      subPageId,
      disabled = false,
      // workaround after the react 19 upgrade, defaultProps doesn't work anymore
      renderLabelInfo,
      renderBeforeLabel,
      renderAfterLabel,
      beforeLabelContentVAlign = 'start',
      // workaround after the react 19 upgrade, defaultProps doesn't work anymore
      afterLabelContentVAlign = 'start',
      // workaround after the react 19 upgrade, defaultProps doesn't work anymore
      description,
      descriptionRole,
      elementRef,
      themeOverride
    } = option.props;
    if (!id) {
      (0, _console.warn)(false, `Drilldown.Option without id won't be rendered. It is needed to internally track the options.`);
      return null;
    }

    // Props passed to the Option.Item
    let optionProps = {
      // passthrough props
      ...(0, _omitProps.omitProps)(option.props, [..._DrilldownOption.DrilldownOption.allowedProps, ..._v11_2.Options.Item.allowedProps]),
      // props from selectable
      ...getOptionProps({
        id,
        // aria-selected is only valid for these roles, otherwise we need to unset it
        ...(role && !['gridcell', 'option', 'row', 'tab', 'columnheader', 'rowheader', 'treeitem'].includes(role) && {
          'aria-selected': undefined
        })
      }),
      // we pass the themeOverride to Options.Item
      themeOverride,
      // directly passed props
      renderBeforeLabel,
      renderAfterLabel,
      beforeLabelContentVAlign,
      afterLabelContentVAlign,
      description,
      descriptionRole,
      as,
      role,
      elementRef,
      variant: 'default',
      tabIndex: -1
    };

    // extra data we need track on the option,
    // but don't want to expose as props
    const optionData = {
      groupProps: groupProps
    };
    const isOptionDisabled = id !== this._headerBackId && (
    // Back nav is never disabled
    this.props.disabled || this.currentPage?.disabled || groupProps?.disabled || disabled);

    // display option as disabled
    if (isOptionDisabled) {
      optionProps.variant = 'disabled';
      optionProps = {
        ...optionProps,
        ...getDisabledOptionProps()
      };
    }

    // track as valid active option if not the title and the map doesn't already contain the id
    if (id !== this._headerTitleId && !this._activeOptionsMap[id]) {
      // store index to know the order of ids later; js objects doesn't preserve order
      this._activeOptionsMap[id] = {
        ...option,
        ...optionData,
        index: Object.keys(this._activeOptionsMap).length + 1
      };
    }
    const customRole = role !== _DrilldownOption.DrilldownOption.defaultProps.role ? role : undefined;

    // BEFORE/AFTER elements:
    // we set a few manually on Options.Item,
    // the rest are passed directly
    if (subPageId) {
      optionProps.renderAfterLabel = (0, _jsxRuntime.jsx)(_IconArrowOpenEndSolid.IconArrowOpenEndSolid, {});
      optionProps['aria-haspopup'] = true;
      optionProps.role = customRole || 'menuitem';
      (0, _console.warn)(!renderAfterLabel, `The prop "renderAfterLabel" is reserved on item with id: "${id}". When it has "subPageId" provided, a navigation arrow will render after the label.`);
    }
    if (id === this._headerBackId) {
      optionProps.renderBeforeLabel = (0, _jsxRuntime.jsx)(_IconArrowOpenStartSolid.IconArrowOpenStartSolid, {});
    }
    const isOptionControlled = typeof option.props.selected === 'boolean';
    if ((groupProps?.selectableType || isOptionControlled) && groupProps) {
      if (!isOptionControlled) {
        isSelected = Boolean(this.state.selectedGroupOptionsMap[groupProps.id]?.has(id));
      } else {
        isSelected = Boolean(option.props.selected);
      }
      optionProps['aria-checked'] = isSelected;
      optionProps.renderBeforeLabel = (0, _jsxRuntime.jsx)(_IconCheckSolid.IconCheckSolid, {
        style: {
          opacity: isSelected ? 1 : 0
        }
      });
      (0, _console.warn)(!renderBeforeLabel, `The prop "renderBeforeLabel" is reserved on item with id: "${id}". When this option is a selectable member of a Drilldown.Group, selection indicator will render before the label.`);

      // setting aria roles and attributes for selectable group items
      if (groupProps.selectableType === 'single') {
        optionProps.role = customRole || 'menuitemradio';
      }
      if (groupProps.selectableType === 'multiple') {
        optionProps.role = customRole || 'menuitemcheckbox';
      }
    }
    // display option as highlighted
    if (id === this.state.highlightedOptionId) {
      optionProps.variant = 'highlighted';
      if (isOptionDisabled) {
        optionProps.variant = 'highlighted-disabled';
      }
    }
    if (href) {
      if (subPageId) {
        (0, _console.warn)(false, `Drilldown.Option with id "${id}" has subPageId, so it will ignore the "href" property.`);
      } else if (groupProps?.selectableType) {
        (0, _console.warn)(false, `Drilldown.Option with id "${id}" is in a selectable group, so it will ignore the "href" property.`);
      } else {
        optionProps.href = href;
      }
    }
    const optionLabel = (0, _callRenderProp.callRenderProp)(children, {
      id,
      variant: optionProps.variant,
      isSelected
    });
    if (!optionLabel) {
      (0, _console.warn)(false, `There are no "children" prop provided for option with id: "${id}", so it won't be rendered.`);
      return null;
    }
    const renderLabelProps = {
      variant: optionProps.variant,
      vAlign: afterLabelContentVAlign,
      as,
      role: optionProps.role,
      isSelected
    };

    // we need to bind our own option props the render functions
    if (typeof optionProps.renderBeforeLabel === 'function' && !optionProps.renderBeforeLabel?.prototype?.isReactComponent) {
      optionProps.renderBeforeLabel = optionProps.renderBeforeLabel.bind(null, renderLabelProps);
    }
    if (typeof optionProps.renderAfterLabel === 'function' && !optionProps.renderAfterLabel?.prototype?.isReactComponent) {
      optionProps.renderAfterLabel = optionProps.renderAfterLabel.bind(null, renderLabelProps);
    }
    const labelInfo = renderLabelInfo && (0, _callRenderProp.callRenderProp)(renderLabelInfo, renderLabelProps);
    const vAlignMap = {
      start: 'flex-start',
      center: 'center',
      end: 'flex-end'
    };
    const labelAriaId = `${id}__label`;
    const infoAriaId = `${id}__info`;
    const labelledby = option.props['aria-labelledby'] || labelAriaId;
    const describedby = option.props['aria-describedby'] || (labelInfo ? infoAriaId : undefined);
    return (0, _react2.createElement)(_v11_2.Options.Item, {
      ...optionProps,
      key: id,
      "aria-labelledby": labelledby,
      "aria-describedby": describedby
    }, (0, _jsxRuntime.jsxs)("div", {
      css: styles?.optionContainer,
      role: "none",
      children: [(0, _jsxRuntime.jsx)("span", {
        css: styles?.optionContent,
        role: "none",
        id: labelAriaId,
        children: optionLabel
      }), labelInfo ? (0, _jsxRuntime.jsx)("span", {
        css: styles?.optionLabelInfo,
        role: "presentation",
        style: {
          alignSelf: vAlignMap[afterLabelContentVAlign]
        },
        children: (0, _jsxRuntime.jsx)("span", {
          id: infoAriaId,
          children: labelInfo
        })
      }) : null]
    }));
  }
  renderGroup(group, getOptionProps, getDisabledOptionProps, needsFirstSeparator, needsLastSeparator) {
    const {
      id,
      children,
      renderGroupTitle,
      themeOverride,
      role = 'group',
      // react 19 defaultProps workaround
      as,
      elementRef
    } = group.props;
    if (!children) {
      return null;
    }
    const groupChildren = [];

    // add a separator above
    if (needsFirstSeparator) {
      groupChildren.push((0, _jsxRuntime.jsx)(_v11_2.Options.Separator, {}, "drilldown-options-separator"));
    }

    // create a sublist as a group
    // (a wrapping list item will be created by Options)
    groupChildren.push((0, _jsxRuntime.jsx)(_v11_2.Options, {
      id: id,
      role: role,
      as: as || this.props.as,
      renderLabel: renderGroupTitle,
      elementRef: elementRef
      // we pass the themeOverride to Options
      ,
      themeOverride: themeOverride,
      children: this.getChildrenArray(children).map(child => {
        if ((0, _matchComponentTypes.matchComponentTypes)(child, [_DrilldownSeparator.DrilldownSeparator])) {
          return this.renderSeparator(child);
        } else if ((0, _matchComponentTypes.matchComponentTypes)(child, [_DrilldownOption.DrilldownOption])) {
          return this.renderOption(child, getOptionProps, getDisabledOptionProps, group.props);
        } else {
          return null;
        }
      })
    }, id));

    // add a separator below
    if (needsLastSeparator) {
      groupChildren.push((0, _jsxRuntime.jsx)(_v11_2.Options.Separator, {}, "drilldown-options-separator-2"));
    }
    return groupChildren;
  }
  renderPage() {
    const {
      styles,
      overflowY,
      overflowX,
      height,
      width,
      minHeight,
      minWidth,
      maxHeight,
      maxWidth,
      role,
      as,
      label,
      trigger
    } = this.props;
    if (!this.currentPage) {
      return null;
    }
    return (0, _jsxRuntime.jsx)(_Selectable.Selectable, {
      isShowingOptions: this.shown,
      highlightedOptionId: this.state.highlightedOptionId,
      selectedOptionId: this.selectedGroupOptionIdsArray,
      onRequestShowOptions: this.show,
      onRequestHideOptions: this.hide,
      onRequestSelectOption: this.handleOptionSelect,
      onRequestHighlightOption: this.handleOptionHighlight,
      onRequestHighlightFirstOption: event => {
        const firstOptionId = this.activeOptionIds[0];
        this.handleOptionHighlight(event, {
          id: firstOptionId
        });
      },
      onRequestHighlightLastOption: event => {
        const lastOptionId = this.activeOptionIds[this.activeOptionIds.length - 1];
        this.handleOptionHighlight(event, {
          id: lastOptionId
        });
      },
      children: ({
        // TODO: figure out what other Selectable props we need, if we want to add a Select version for drilldown:
        // getRootProps, - we probably don't need this
        // getLabelProps, - do we need label?
        // getDescriptionProps, - might be nice for assistiveText like in Select
        // getInputProps, - hidden input for a11y? role="combobox" might be needed
        getTriggerProps,
        getListProps,
        getOptionProps,
        getDisabledOptionProps
      }) => (0, _jsxRuntime.jsx)(_v11_.View
      // if there is a trigger the border needs to be rendered by the <Popover>
      // (because it has an arrow)
      , {
        ...(trigger ? {} : {
          borderWidth: 'small'
        }),
        as: "div",
        "data-cid": "Drilldown",
        elementRef: this.handleDrilldownRef,
        tabIndex: 0,
        css: styles?.drilldown,
        position: "relative",
        borderRadius: "small",
        width: width,
        minWidth: maxWidth,
        maxWidth: maxWidth,
        role: role,
        "aria-label": label,
        "aria-labelledby": this.currentPageAriaLabel,
        ...getTriggerProps({
          id: this._id,
          // We need to override these aria attributes added by Selectable,
          // since Drilldown is not a combobox and has no popup
          'aria-haspopup': false,
          'aria-expanded': undefined,
          onKeyDown: this.handleKeyDown,
          onBlur: event => {
            const target = event.currentTarget;
            const related = event.relatedTarget;
            const containsRelated = (0, _contains.contains)(target, related);
            if (!related || related === this._drilldownRef || related !== target && !containsRelated) {
              this.setState({
                highlightedOptionId: undefined
              });
            }
          },
          onMouseLeave: () => {
            this.setState({
              highlightedOptionId: undefined
            });
          }
        }),
        children: (0, _jsxRuntime.jsx)(_v11_.View, {
          as: "div",
          overflowY: overflowY,
          overflowX: overflowX,
          height: height,
          width: width,
          minHeight: minHeight,
          minWidth: minWidth,
          maxHeight: maxHeight,
          maxWidth: maxWidth,
          css: styles?.container,
          borderRadius: "small",
          role: "presentation",
          elementRef: element => {
            this._containerElement = element;
          },
          children: (0, _jsxRuntime.jsx)(_v11_2.Options, {
            ...getListProps(),
            role: "presentation",
            as: as,
            children: this.renderList(getOptionProps, getDisabledOptionProps)
          })
        })
      })
    });
  }
  render() {
    // clear temporary option store
    this._activeOptionsMap = {};
    const {
      show,
      defaultShow,
      placement,
      withArrow,
      shouldContainFocus,
      shouldReturnFocus,
      trigger,
      mountNode,
      constrain,
      positionTarget,
      positionContainerDisplay,
      popoverRef,
      disabled,
      onDismiss,
      onFocus,
      onMouseOver,
      offsetX,
      offsetY,
      shouldSetAriaExpanded,
      styles
    } = this.props;
    const borderColor = styles?.drilldown?.borderColor;
    return trigger ? (0, _jsxRuntime.jsx)(_v11_3.Popover, {
      "data-cid": "Drilldown",
      themeOverride: {
        // use here own theme's color
        borderColor: borderColor
      },
      shouldSetAriaExpanded: shouldSetAriaExpanded,
      isShowingContent: show,
      defaultIsShowingContent: defaultShow,
      shouldCloseOnDocumentClick: true,
      onHideContent: (event, {
        documentClick
      }) => {
        if (typeof onDismiss === 'function') {
          onDismiss(event, documentClick);
        }
        this.reset();
        this.handleToggle(event, false);
      },
      onShowContent: event => this.handleToggle(event, true),
      mountNode: mountNode || this.ref,
      placement: placement,
      withArrow: withArrow,
      positionTarget: positionTarget,
      positionContainerDisplay: positionContainerDisplay,
      constrain: constrain,
      shouldContainFocus: shouldContainFocus,
      shouldReturnFocus: shouldReturnFocus,
      id: this._id,
      on: ['click'],
      onFocus: onFocus,
      onMouseOver: onMouseOver,
      offsetX: offsetX,
      offsetY: offsetY,
      defaultFocusElement: () => {
        if (!this.currentPage) return null;
        const actionLabel = (0, _callRenderProp.callRenderProp)(this.currentPage.renderActionLabel);
        // Use action ID if exists, otherwise first non-action option's ID
        const targetId = actionLabel ? this._headerActionId : this.getFirstOption()?.props.id;
        if (!targetId) return null;
        return this._popover?._contentElement?.querySelector(`#${CSS.escape(targetId)}`);
      },
      elementRef: element => {
        // setting ref for "Popover" version, the popover root
        // (if there is no trigger, we set it in handleDrilldownRef)
        this.handleRef(element);
      },
      ref: el => {
        this._popover = el;
        if (typeof popoverRef === 'function') {
          popoverRef(el);
        }
      },
      renderTrigger: (0, _safeCloneElement.safeCloneElement)(trigger, {
        ref: el => {
          this._trigger = el;
        },
        'aria-haspopup': this.props.role,
        id: this._triggerId,
        disabled: !!(trigger.props.disabled || disabled),
        'aria-disabled': trigger.props.disabled || disabled ? 'true' : undefined
      }),
      children: this.renderPage()
    }) : this.renderPage();
  }
}) || _class) || _class);
var _default = exports.default = Drilldown;
