// @instructure/ui-tag/es/Tag/v1/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { IconXLine } from '@instructure/ui-icons';
import { View } from '@instructure/ui-view/v11_6';
import { omitProps } from '@instructure/ui-react-utils';
import { isActiveElement } from '@instructure/ui-dom-utils';
import { withStyle } from '@instructure/emotion';
import generateStyle from "./styles.js";
import generateComponentTheme from "./theme.js";
import { allowedProps } from "./props.js";
/**
---
category: components
---
**/
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
let Tag = (_dec = withStyle(generateStyle, generateComponentTheme), _dec(_class = class Tag extends Component {
  static displayName = "Tag";
  static componentId = 'Tag';
  static allowedProps = allowedProps;
  static defaultProps = {
    size: 'medium',
    dismissible: false,
    variant: 'default',
    disabled: false,
    readOnly: false
  };
  ref = null;
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  get focused() {
    return isActiveElement(this.ref);
  }
  focus = () => {
    this.ref && this.ref.focus();
  };
  handleClick = e => {
    const {
      disabled,
      readOnly,
      onClick
    } = this.props;
    if (disabled || readOnly) {
      e.preventDefault();
      e.stopPropagation();
    } else if (typeof onClick === 'function') {
      onClick(e);
    }
  };
  handleRef = element => {
    this.ref = element;
    if (typeof this.props.elementRef === 'function') {
      this.props.elementRef(element);
    }
  };
  render() {
    const {
      className,
      dismissible,
      disabled,
      readOnly,
      text,
      title,
      onClick,
      margin,
      styles
    } = this.props;
    const passthroughProps = View.omitViewProps(omitProps(this.props, Tag.allowedProps), Tag);
    return _jsxs(View, {
      ...passthroughProps,
      elementRef: this.handleRef,
      css: styles?.tag,
      className: className,
      as: onClick ? 'button' : 'span',
      margin: margin,
      type: onClick ? 'button' : undefined,
      ...(onClick && {
        onClick: this.handleClick
      }),
      disabled: disabled || readOnly,
      display: undefined,
      title: title || (typeof text === 'string' ? text : undefined),
      "data-cid": "Tag",
      children: [_jsx("span", {
        css: styles?.text,
        children: text
      }), onClick && dismissible ? _jsx(IconXLine, {
        css: styles?.icon
      }) : null]
    });
  }
}) || _class);
export default Tag;
export { Tag };
