// @instructure/ui-editable/lib/InPlaceEdit/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.InPlaceEdit = void 0;
var _react = require("react");
var _latest = require("@instructure/ui-flex/latest");
var _latest2 = require("@instructure/ui-buttons/latest");
var _uiIcons = require("@instructure/ui-icons");
var _console = require("@instructure/console");
var _createChainedFunction = require("@instructure/ui-utils/lib/createChainedFunction.js");
var _emotion = require("@instructure/emotion");
var _latest3 = require("@instructure/ui-view/latest");
var _v = require("../../Editable/v1");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2018 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let InPlaceEdit = exports.InPlaceEdit = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class InPlaceEdit extends _react.Component {
  static displayName = "InPlaceEdit";
  static componentId = 'InPlaceEdit';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    readOnly: false,
    showFocusRing: true,
    inline: true,
    editButtonPlacement: 'end'
  };
  ref = null;
  _editButtonRef = null;
  handleRef = el => {
    this.ref = el;
  };
  constructor(props) {
    super(props);
    (0, _console.logWarn)(props.readOnly ? props.mode === 'view' : true, '[InPlaceEdit] When readOnly is true, mode is forced to "view"');
  }
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  handleEditButtonRef = el => {
    this._editButtonRef = el;
  };
  renderEditor({
    mode,
    onBlur,
    editorRef,
    readOnly
  }) {
    const {
      showFocusRing,
      renderEditor
    } = this.props;
    const isEditMode = !readOnly && mode === 'edit';
    return isEditMode ? (0, _jsxRuntime.jsx)(_latest3.View, {
      as: "span",
      display: "block",
      withFocusOutline: showFocusRing,
      position: "relative",
      css: this.props.styles?.inPlaceEdit,
      borderRadius: "medium",
      margin: "auto",
      children: renderEditor({
        onBlur,
        editorRef
      })
    }) : null;
  }
  renderViewer({
    readOnly,
    mode
  }) {
    return readOnly || mode === 'view' ? this.props.renderViewer() : null;
  }
  renderEditButton({
    buttonRef,
    ...rest
  }) {
    return this.props.renderEditButton({
      elementRef: (0, _createChainedFunction.createChainedFunction)(this.handleEditButtonRef, buttonRef),
      ...rest
    });
  }

  // Render a default edit button, an icon button with the edit icon
  // the margin makes room for the focus ring
  static renderDefaultEditButton = ({
    isVisible,
    readOnly,
    label,
    ...buttonProps
  }) => {
    if (readOnly) {
      return null;
    }
    return (0, _jsxRuntime.jsx)(_latest2.IconButton, {
      size: "small",
      screenReaderLabel: label,
      withBackground: false,
      withBorder: false,
      ...buttonProps,
      children: isVisible ? _uiIcons.PencilInstUIIcon : null
    });
  };
  renderAll = ({
    getContainerProps,
    getViewerProps,
    getEditorProps,
    getEditButtonProps
  }) => {
    const flexDir = this.props.editButtonPlacement === 'start' ? 'row-reverse' : 'row';
    const justifyItems = flexDir === 'row-reverse' ? 'end' : 'start';
    const buttonMargin = this.props.editButtonPlacement === 'start' ? '0 xx-small 0 0' : '0 0 0 xx-small';
    return (0, _jsxRuntime.jsxs)(_latest.Flex, {
      display: this.props.inline ? 'inline-flex' : 'flex',
      direction: flexDir,
      justifyItems: justifyItems,
      ...getContainerProps(),
      children: [(0, _jsxRuntime.jsxs)(_latest.Flex.Item, {
        shouldGrow: true,
        shouldShrink: true,
        children: [this.renderEditor(getEditorProps()), this.renderViewer(getViewerProps())]
      }), (0, _jsxRuntime.jsx)(_latest.Flex.Item, {
        margin: buttonMargin,
        children: this.renderEditButton(getEditButtonProps())
      })]
    });
  };
  render() {
    const {
      mode,
      value,
      onChange,
      onChangeMode,
      readOnly
    } = this.props;
    return (0, _jsxRuntime.jsx)(_v.Editable, {
      mode: mode,
      onChangeMode: onChangeMode,
      render: this.renderAll,
      value: value,
      onChange: onChange,
      readOnly: readOnly,
      elementRef: this.handleRef
    });
  }
}) || _class);
var _default = exports.default = InPlaceEdit;
