// @instructure/ui-tooltip/lib/Tooltip/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Tooltip = void 0;
var _react = require("react");
var _getElementType = require("@instructure/ui-react-utils/lib/getElementType.js");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _ensureSingleChild = require("@instructure/ui-react-utils/lib/ensureSingleChild.js");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _latest = require("@instructure/ui-popover/latest");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let Tooltip = exports.Tooltip = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = _dec2(_class = class Tooltip extends _react.Component {
  static displayName = "Tooltip";
  static componentId = 'Tooltip';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    defaultIsShowingContent: false,
    color: 'primary',
    placement: 'top',
    constrain: 'window',
    offsetX: 0,
    offsetY: 0,
    preventTooltip: false
  };
  _id;
  ref = null;
  handleRef = el => {
    this.ref = el;
    if (typeof this.props.elementRef === 'function') {
      this.props.elementRef(el);
    }
  };
  constructor(props) {
    super(props);
    this._id = props.deterministicId();
    this.state = {
      hasFocus: false
    };
  }
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  handleFocus = () => {
    this.setState({
      hasFocus: true
    });
  };
  handleBlur = () => {
    this.setState({
      hasFocus: false
    });
  };
  renderTrigger() {
    const {
      children,
      as
    } = this.props;
    const {
      hasFocus
    } = this.state;
    const triggerProps = {
      'aria-describedby': this._id
    };
    if (as) {
      const Trigger = (0, _getElementType.getElementType)(Tooltip, this.props);
      const props = (0, _omitProps.omitProps)(this.props, Tooltip.allowedProps);
      return (0, _jsxRuntime.jsx)(Trigger, {
        ...props,
        ...triggerProps,
        children: children /*TODO check if it can be TooltipRenderChildren, this might cause a crash*/
      });
    } else if (typeof children === 'function') {
      return children({
        focused: hasFocus,
        getTriggerProps: props => ({
          ...triggerProps,
          ...props
        })
      });
    } else {
      return (0, _ensureSingleChild.ensureSingleChild)(children, triggerProps);
    }
  }
  render() {
    const {
      renderTip,
      isShowingContent,
      defaultIsShowingContent,
      on,
      color,
      placement,
      mountNode,
      constrain,
      offsetX,
      offsetY,
      positionTarget,
      onShowContent,
      onHideContent,
      preventTooltip,
      styles,
      ...rest
    } = this.props;
    return (0, _jsxRuntime.jsx)(_latest.Popover, {
      ...(0, _passthroughProps.passthroughProps)(rest),
      isShowingContent: !preventTooltip && isShowingContent,
      defaultIsShowingContent: !preventTooltip && defaultIsShowingContent,
      on: on,
      shouldRenderOffscreen: true,
      shouldReturnFocus: false,
      placement: placement,
      color: color === 'primary' ? 'primary-inverse' : 'primary',
      mountNode: mountNode,
      constrain: constrain,
      shadow: "resting",
      offsetX: offsetX,
      offsetY: offsetY,
      positionTarget: positionTarget,
      renderTrigger: () => this.renderTrigger(),
      onShowContent: onShowContent,
      onHideContent: onHideContent,
      onFocus: this.handleFocus,
      onBlur: this.handleBlur,
      elementRef: this.handleRef,
      shouldCloseOnDocumentClick: false,
      shouldCloseOnEscape: true,
      shouldSetAriaExpanded: false,
      "data-cid": "Tooltip",
      children: !preventTooltip ? (0, _jsxRuntime.jsx)("span", {
        id: this._id,
        css: styles?.tooltip,
        role: "tooltip",
        children: (0, _callRenderProp.callRenderProp)(renderTip)
      }) : null
    });
  }
}) || _class) || _class);
var _default = exports.default = Tooltip;
