// @instructure/ui-form-field/es/FormFieldMessage/v1/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { ScreenReaderContent } from '@instructure/ui-a11y-content';
import { IconWarningSolid } from '@instructure/ui-icons';
import { withStyle } from '@instructure/emotion';
import generateStyle from "./styles.js";
import generateComponentTheme from "./theme.js";
import { allowedProps } from "./props.js";
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
/**
---
parent: FormField
---

This is a helper component that is used by most of the custom form
components. In most cases it shouldn't be used directly.

```js
---
type: example
---
<FormFieldMessage variant="error">Invalid value</FormFieldMessage>
```
**/
let FormFieldMessage = (_dec = withStyle(generateStyle, generateComponentTheme), _dec(_class = class FormFieldMessage extends Component {
  static displayName = "FormFieldMessage";
  static componentId = 'FormFieldMessage';
  static allowedProps = allowedProps;
  static defaultProps = {
    variant: 'hint'
  };
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  render() {
    const {
      children,
      styles
    } = this.props;
    return this.props.variant !== 'screenreader-only' ? _jsxs("span", {
      css: {
        display: 'flex'
      },
      children: [this.props.variant === 'newError' && this.props.children && _jsx("span", {
        css: styles?.errorIcon,
        children: _jsx(IconWarningSolid, {
          color: "error"
        })
      }), _jsx("span", {
        css: styles?.formFieldMessage,
        ref: this.handleRef,
        children: children
      })]
    }) : _jsx(ScreenReaderContent, {
      elementRef: this.handleRef,
      children: children
    });
  }
}) || _class);
export default FormFieldMessage;
export { FormFieldMessage };
