// @instructure/ui-table/es/Table/v2/Head/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component, Children } from 'react';
import { omitProps, callRenderProp } from '@instructure/ui-react-utils';
import { SimpleSelect } from '@instructure/ui-simple-select/latest';
import { ScreenReaderContent } from '@instructure/ui-a11y-content';
import { CheckInstUIIcon } from '@instructure/ui-icons';
import { warn } from '@instructure/console';
import { withStyleNew } from '@instructure/emotion';
import generateStyle from "./styles.js";
import { allowedProps } from "./props.js";
import TableContext from "../TableContext.js";
/**
---
parent: Table
id: Table.Head
---
**/
import { jsx as _jsx } from "@emotion/react/jsx-runtime";
let Head = (_dec = withStyleNew(generateStyle), _dec(_class = class Head extends Component {
  static displayName = "Head";
  static componentId = 'Table.Head';
  static contextType = TableContext;
  static allowedProps = allowedProps;
  static defaultProps = {
    children: null
  };

  /**
   * Returns `true` if the first child's children have a `onRequestSort` prop
   */
  get isSortable() {
    const [firstRow] = Children.toArray(this.props.children);
    let sortable = false;
    if (firstRow && firstRow.props && firstRow.props.children) {
      Children.forEach(firstRow.props.children, grandchild => {
        if (grandchild?.props?.onRequestSort) {
          sortable = true;
          return;
        }
      });
    }
    return sortable;
  }
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    if (this.isSortable && typeof this.props.renderSortLabel === 'undefined') {
      warn(false, '[Table.Head] The `renderSortLabel` prop should be provided when Table is sortable.');
    }
    this.props.makeStyles?.();
  }

  /**
   * This `Select` is used in `stacked` layout. It's populated by iterating
   * through the first child's children (by default `ColHeader`) and reading
   * there the `id`, `stackedSortByLabel`, `sortDirection`, `onRequestSort` props
   */
  renderSelect() {
    const {
      children,
      renderSortLabel
    } = this.props;
    const [firstRow] = Children.toArray(children);
    if (!firstRow?.props?.children) {
      return null;
    }
    const options = [];
    const clickHandlers = {};
    let selectedOption;
    let count = 0;
    Children.forEach(firstRow.props.children, grandchild => {
      count += 1;
      if (!grandchild?.props) return; // grandchild can be false
      const {
        id,
        stackedSortByLabel,
        sortDirection,
        onRequestSort
      } = grandchild.props;
      if (id && onRequestSort) {
        const label = stackedSortByLabel || id;
        options.push({
          id,
          label
        });
        clickHandlers[id] = onRequestSort;
        if (sortDirection !== 'none') {
          selectedOption = id;
        }
      }
    });
    if (!options.length) {
      return null;
    }
    const handleSelect = (event, {
      value
    }) => {
      if (value && typeof clickHandlers[value] === 'function') {
        clickHandlers[value](event, {
          id: `${value}`
        });
      }
    };
    return _jsx("div", {
      role: "rowgroup",
      children: _jsx("div", {
        role: "row",
        children: _jsx("div", {
          role: "cell",
          "aria-colspan": count,
          children: _jsx(SimpleSelect, {
            renderLabel: renderSortLabel ? callRenderProp(renderSortLabel) : _jsx(ScreenReaderContent, {}),
            renderBeforeInput: selectedOption ? () => _jsx(CheckInstUIIcon, {}) : undefined,
            value: selectedOption,
            onChange: handleSelect,
            children: options.map(({
              id,
              label
            }) => _jsx(SimpleSelect.Option, {
              id: id,
              value: id,
              renderBeforeLabel: id === selectedOption ? () => _jsx(CheckInstUIIcon, {}) : () => _jsx("span", {
                style: {
                  visibility: 'hidden'
                },
                children: _jsx(CheckInstUIIcon, {})
              }),
              children: label
            }, id))
          })
        })
      })
    });
  }
  render() {
    const {
      children,
      styles
    } = this.props;
    return this.context.isStacked ? this.renderSelect() : _jsx("thead", {
      ...omitProps(this.props, Head.allowedProps),
      css: styles?.head,
      children: children
    });
  }
}) || _class);
export default Head;
export { Head };
