// @instructure/ui-text-area/lib/TextArea/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TextArea = void 0;
var _react = require("react");
var _latest = require("@instructure/ui-form-field/latest");
var _addEventListener = require("@instructure/ui-dom-utils/lib/addEventListener.js");
var _isActiveElement = require("@instructure/ui-dom-utils/lib/isActiveElement.js");
var _requestAnimationFrame = require("@instructure/ui-dom-utils/lib/requestAnimationFrame.js");
var _getBoundingClientRect = require("@instructure/ui-dom-utils/lib/getBoundingClientRect.js");
var _debounce = require("@instructure/debounce");
var _emotion = require("@instructure/emotion");
var _generateId = require("@instructure/ui-utils/lib/generateId.js");
var _px = require("@instructure/ui-utils/lib/px.js");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _pickProps = require("@instructure/ui-react-utils/lib/pickProps.js");
var _DeterministicIdContext = require("@instructure/ui-react-utils/lib/DeterministicIdContext/DeterministicIdContext.js");
var _styles = _interopRequireDefault(require("./styles"));
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
category: components
---
**/
const TextArea = exports.TextArea = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    size = 'medium',
    autoGrow = true,
    resize = 'none',
    inline = false,
    messages = [],
    disabled = false,
    readOnly = false,
    layout = 'stacked',
    required = false,
    placeholder,
    value,
    defaultValue,
    width,
    height,
    maxHeight,
    textareaRef,
    margin,
    label,
    onChange,
    id: propId,
    themeOverride,
    ...rest
  } = props;

  // Use deterministic ID
  const instanceCounterMap = (0, _react.useContext)(_DeterministicIdContext.DeterministicIdContext);
  const defaultId = (0, _react.useMemo)(() => (0, _generateId.generateId)('TextArea', instanceCounterMap), [instanceCounterMap]);
  const id = propId || defaultId;

  // Use refs for mutable values
  const _listener = (0, _react.useRef)();
  const _request = (0, _react.useRef)();
  const _textareaResizeListener = (0, _react.useRef)();
  const _debounced = (0, _react.useRef)();
  const _textarea = (0, _react.useRef)(null);
  const _container = (0, _react.useRef)(null);
  const _height = (0, _react.useRef)();
  const _manuallyResized = (0, _react.useRef)(false);
  const formFieldRef = (0, _react.useRef)(null);

  // Compute invalid state
  const {
    isInvalid,
    isSuccess
  } = (0, _react.useMemo)(() => {
    const isInvalid = messages && messages.findIndex(message => message.type === 'error' || message.type === 'newError') >= 0;
    const isSuccess = messages && messages.findIndex(message => message.type === 'success') >= 0;
    return {
      isInvalid,
      isSuccess
    };
  }, [messages]);

  // Use styles
  const styles = (0, _emotion.useStyleNew)({
    generateStyle: _styles.default,
    themeOverride,
    params: {
      size,
      disabled,
      readOnly,
      success: isSuccess,
      invalid: isInvalid
    },
    componentId: 'TextArea',
    displayName: 'TextArea'
  });

  // Mock ResizeObserver for SSR
  (0, _react.useEffect)(() => {
    if (typeof window === 'undefined') {
      global.ResizeObserver = class ResizeObserver {
        observe() {
          // do nothing
        }
        unobserve() {
          // do nothing
        }
        disconnect() {
          // do nothing
        }
      };
    }
  }, []);

  // Grow function
  const grow = (0, _react.useCallback)(() => {
    if (!_textarea.current || _manuallyResized.current) {
      return;
    }
    const offset = _textarea.current.offsetHeight - _textarea.current.clientHeight;
    let heightValue = '';

    // Notes:
    // 1. height has to be reset to `auto` every time this method runs, or scrollHeight will not reset
    // 2. `_textarea.current.scrollHeight` will not reset if assigned to a variable; it needs to be written out each time
    _textarea.current.style.height = 'auto';
    _textarea.current.style.overflowY = 'hidden'; // hide scrollbars for autoGrow textareas
    heightValue = _textarea.current.scrollHeight + offset + 'px';
    const maxHeightPx = maxHeight ? (0, _px.px)(maxHeight, _container.current) : undefined;
    if (maxHeight && maxHeightPx !== undefined && _textarea.current.scrollHeight > maxHeightPx) {
      _textarea.current.style.overflowY = 'auto'; // add scroll if scrollHeight exceeds maxHeight in pixels
    } else if (height) {
      if (_textarea.current.value === '') {
        heightValue = height;
      } else if ((0, _px.px)(height, _container.current) > _textarea.current.scrollHeight) {
        _textarea.current.style.overflowY = 'auto'; // add scroll if scrollHeight exceeds height in pixels
        heightValue = height;
      }
    }

    // preserve container height to prevent scroll jumping on long textareas,
    // but make sure container doesn't exceed maxHeight prop
    const heightExceedsMax = maxHeightPx !== undefined && (0, _px.px)(heightValue) > maxHeightPx;
    if (!heightExceedsMax && _container.current) {
      _container.current.style.minHeight = heightValue;
    }
    _height.current = heightValue;
    _textarea.current.style.height = heightValue;
  }, [height, maxHeight]);

  // Textarea resize handler
  const _textareaResize = (0, _react.useCallback)(() => {
    if (!_textarea.current) return;
    const textareaHeight = _textarea.current.style.height;
    if (textareaHeight !== '' && textareaHeight !== _height.current) {
      _manuallyResized.current = true;
      _textarea.current.style.overflowY = 'auto';

      // update container minHeight to ensure focus ring always wraps input
      if (_container.current) {
        _container.current.style.minHeight = textareaHeight;
      }
    }
  }, []);

  // AutoGrow effect
  (0, _react.useEffect)(() => {
    if (autoGrow) {
      if (!_debounced.current) {
        _debounced.current = (0, _debounce.debounce)(grow, 200, {
          leading: false,
          trailing: true
        });
      }
      if (!_listener.current) {
        _listener.current = (0, _addEventListener.addEventListener)(window, 'resize', _debounced.current);
      }
      if (_textarea.current && !_textareaResizeListener.current) {
        const {
          height: origHeight
        } = (0, _getBoundingClientRect.getBoundingClientRect)(_textarea.current);
        _textareaResizeListener.current = new ResizeObserver(entries => {
          for (const entry of entries) {
            const {
              height: entryHeight
            } = entry.contentRect;
            if (origHeight !== entryHeight) {
              _textareaResize();
            }
          }
        });
        _textareaResizeListener.current.observe(_textarea.current);
      }
      _request.current = (0, _requestAnimationFrame.requestAnimationFrame)(grow);
    }
  }, [autoGrow, grow, _textareaResize]);

  // Cleanup effect
  (0, _react.useEffect)(() => {
    return () => {
      if (_listener.current) {
        _listener.current.remove();
      }
      if (_textareaResizeListener.current) {
        _textareaResizeListener.current.disconnect();
      }
      if (_request.current) {
        _request.current.cancel();
      }
      if (_debounced.current) {
        _debounced.current.cancel();
      }
    };
  }, []);

  // Handle change
  const handleChange = (0, _react.useCallback)(event => {
    if (disabled || readOnly) {
      event.preventDefault();
      return;
    }
    if (typeof value === 'undefined') {
      // if uncontrolled
      if (autoGrow) {
        if (!_debounced.current) {
          _debounced.current = (0, _debounce.debounce)(grow, 200, {
            leading: false,
            trailing: true
          });
        }
        if (!_listener.current) {
          _listener.current = (0, _addEventListener.addEventListener)(window, 'resize', _debounced.current);
        }
        if (_textarea.current && !_textareaResizeListener.current) {
          const {
            height: origHeight
          } = (0, _getBoundingClientRect.getBoundingClientRect)(_textarea.current);
          _textareaResizeListener.current = new ResizeObserver(entries => {
            for (const entry of entries) {
              const {
                height: entryHeight
              } = entry.contentRect;
              if (origHeight !== entryHeight) {
                _textareaResize();
              }
            }
          });
          _textareaResizeListener.current.observe(_textarea.current);
        }
        _request.current = (0, _requestAnimationFrame.requestAnimationFrame)(grow);
      }
    }
    if (typeof onChange === 'function') {
      onChange(event);
    }
  }, [disabled, readOnly, value, onChange, autoGrow, grow, _textareaResize]);

  // Expose API through ref using useImperativeHandle
  (0, _react.useImperativeHandle)(ref, () => ({
    focus: () => {
      _textarea.current?.focus();
    },
    get value() {
      return _textarea.current?.value || '';
    },
    get minHeight() {
      return _textarea.current?.style.minHeight || '';
    },
    get focused() {
      return (0, _isActiveElement.isActiveElement)(_textarea.current);
    },
    get invalid() {
      return messages && messages.findIndex(message => message.type === 'error' || message.type === 'newError') >= 0;
    },
    get id() {
      return id;
    },
    ref: formFieldRef
  }), [id, messages]);
  const style = {
    width,
    resize,
    height: !autoGrow ? height : undefined,
    maxHeight
  };
  const textarea = (0, _jsxRuntime.jsx)("textarea", {
    ...(0, _passthroughProps.passthroughProps)(rest),
    value: value,
    defaultValue: defaultValue,
    placeholder: placeholder,
    ref: textarea => {
      _textarea.current = textarea;
      if (typeof textareaRef === 'function') {
        textareaRef(textarea);
      }
    },
    style: style,
    id: id,
    required: required,
    "aria-required": required,
    "aria-invalid": isInvalid ? 'true' : undefined,
    disabled: disabled,
    readOnly: readOnly,
    css: styles?.textArea,
    onChange: handleChange
  });
  return (0, _jsxRuntime.jsx)(_latest.FormField, {
    ...(0, _pickProps.pickProps)(props, _latest.FormField.allowedProps),
    label: label,
    vAlign: "top",
    id: id,
    elementRef: el => {
      formFieldRef.current = el;
    },
    margin: margin,
    isRequired: required,
    disabled: disabled,
    readOnly: readOnly,
    "data-cid": "TextArea",
    children: (0, _jsxRuntime.jsx)("div", {
      css: styles?.textAreaLayout,
      style: {
        width,
        maxHeight
      },
      ref: node => {
        _container.current = node;
      },
      children: textarea
    })
  });
});
TextArea.displayName = 'TextArea';
var _default = exports.default = TextArea;
