// @instructure/ui-table/lib/Table/v1/ColHeader/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ColHeader = void 0;
var _react = require("react");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _IconMiniArrowUpLine = require("@instructure/ui-icons/lib/generated/IconMiniArrowUpLine.js");
var _IconMiniArrowDownLine = require("@instructure/ui-icons/lib/generated/IconMiniArrowDownLine.js");
var _IconMiniArrowDoubleLine = require("@instructure/ui-icons/lib/generated/IconMiniArrowDoubleLine.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: Table
id: Table.ColHeader
---
**/
let ColHeader = exports.ColHeader = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class ColHeader extends _react.Component {
  static displayName = "ColHeader";
  static componentId = 'Table.ColHeader';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    textAlign: 'start',
    sortDirection: 'none',
    children: null,
    scope: 'col'
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  handleClick = event => {
    const {
      id,
      onRequestSort
    } = this.props;
    onRequestSort && onRequestSort(event, {
      id
    });
  };
  renderSortArrow() {
    const {
      sortDirection,
      onRequestSort,
      styles
    } = this.props;
    if (sortDirection === 'ascending') {
      return (0, _jsxRuntime.jsx)(_IconMiniArrowUpLine.IconMiniArrowUpLine, {
        css: styles?.sortedIconColor
      });
    }
    if (sortDirection === 'descending') {
      return (0, _jsxRuntime.jsx)(_IconMiniArrowDownLine.IconMiniArrowDownLine, {
        css: styles?.sortedIconColor
      });
    }
    if (onRequestSort) {
      return (0, _jsxRuntime.jsx)(_IconMiniArrowDoubleLine.IconMiniArrowDoubleLine, {
        css: styles?.unSortedIconColor
      });
    }
    return undefined;
  }
  render() {
    const {
      onRequestSort,
      width,
      children,
      sortDirection,
      scope,
      styles
    } = this.props;
    return (0, _jsxRuntime.jsxs)("th", {
      ...(0, _omitProps.omitProps)(this.props, ColHeader.allowedProps),
      css: styles?.colHeader,
      style: {
        width
      },
      scope: scope,
      "aria-sort": sortDirection,
      children: [onRequestSort && (0, _jsxRuntime.jsx)("button", {
        onClick: this.handleClick,
        css: styles?.button,
        children: (0, _jsxRuntime.jsxs)("div", {
          css: styles?.buttonContent,
          children: [(0, _callRenderProp.callRenderProp)(children), this.renderSortArrow()]
        })
      }), !onRequestSort && children, !onRequestSort && this.renderSortArrow()]
    });
  }
}) || _class);
var _default = exports.default = ColHeader;
