// @instructure/ui-drawer-layout/lib/DrawerLayout/v1/__fixtures__/DrawerLayout.fixture.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _index = require("../index");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

class DrawerLayoutFixture extends _react.Component {
  static displayName = "DrawerLayoutFixture";
  static defaultProps = {
    layoutWidth: '600px',
    trayWidth: '200px',
    open: false,
    placement: 'start',
    onOverlayTrayChange: () => {}
  };
  render() {
    const {
      layoutWidth,
      trayWidth,
      open,
      placement,
      onOverlayTrayChange
    } = this.props;
    return (0, _jsxRuntime.jsx)("div", {
      style: {
        width: layoutWidth
      },
      children: (0, _jsxRuntime.jsxs)(_index.DrawerLayout, {
        minWidth: "500px",
        onOverlayTrayChange: onOverlayTrayChange,
        children: [(0, _jsxRuntime.jsx)(_index.DrawerLayout.Tray, {
          open: open,
          placement: placement,
          label: "Test DrawerTray",
          children: (0, _jsxRuntime.jsx)("div", {
            style: {
              width: trayWidth
            },
            children: "Hello from tray"
          })
        }), (0, _jsxRuntime.jsx)(_index.DrawerLayout.Content, {
          label: "Test DrawerContent",
          children: (0, _jsxRuntime.jsxs)("div", {
            children: ["Hello from content", (0, _jsxRuntime.jsx)("button", {
              children: "Expand"
            })]
          })
        })]
      })
    });
  }
}
exports.default = DrawerLayoutFixture;
