// @instructure/ui-toggle-details/es/ToggleDetails/v2/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component, createRef } from 'react';
import { Button } from '@instructure/ui-buttons/latest';
import { ChevronRightInstUIIcon, ChevronDownInstUIIcon } from '@instructure/ui-icons';
import { Expandable } from '@instructure/ui-expandable';
import { omitProps, pickProps } from '@instructure/ui-react-utils';
import { isActiveElement } from '@instructure/ui-dom-utils';
import { withStyleNew } from '@instructure/emotion';
import generateStyle from "./styles.js";
import { allowedProps } from "./props.js";
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
/**
---
category: components
---
**/
let ToggleDetails = (_dec = withStyleNew(generateStyle), _dec(_class = class ToggleDetails extends Component {
  static displayName = "ToggleDetails";
  static componentId = 'ToggleDetails';
  static allowedProps = allowedProps;
  static defaultProps = {
    variant: 'default',
    size: 'medium',
    fluidWidth: false,
    icon: ChevronRightInstUIIcon,
    iconExpanded: ChevronDownInstUIIcon,
    iconPosition: 'start',
    defaultExpanded: false,
    children: null
  };
  _buttonNode = null;
  ref = /*#__PURE__*/createRef();

  // TODO this solution was generated with Gemini, should be tested if it works as fine as before
  setButtonRef = node => {
    this._buttonNode = node;
  };
  get focused() {
    return isActiveElement(this._buttonNode);
  }
  focus() {
    this._buttonNode?.focus();
  }
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  renderSummary(expanded) {
    const {
      summary,
      iconPosition
    } = this.props;
    return _jsxs("span", {
      css: this.props.styles?.summary,
      children: [iconPosition === 'start' && this.renderIcon(expanded), _jsx("span", {
        css: this.props.styles?.summaryText,
        children: summary
      }), iconPosition === 'end' && this.renderIcon(expanded)]
    });
  }
  renderToggle(toggleProps, expanded) {
    const {
      variant
    } = this.props;
    const tProps = this.props.children ? toggleProps : {
      onClick: toggleProps.onClick
    };
    const props = {
      ...omitProps(this.props, ToggleDetails.allowedProps),
      ...tProps,
      children: this.renderSummary(expanded)
    };
    const summary = this.renderSummary(expanded);
    if (variant === 'filled') {
      return _jsx(Button, {
        ...props,
        display: "block",
        textAlign: "start",
        elementRef: this.setButtonRef,
        children: summary
      });
    } else if (props.href) {
      return _jsx("a", {
        ...props,
        css: this.props.styles?.toggle,
        ref: this.setButtonRef,
        children: summary
      });
    } else {
      return _jsx("button", {
        ...props,
        type: "button",
        css: this.props.styles?.toggle,
        ref: this.setButtonRef,
        children: summary
      });
    }
  }
  renderIcon(expanded) {
    const Icon = expanded ? this.props.iconExpanded : this.props.icon;
    const iconSize = this.props.size === 'large' ? 'lg' : 'md';
    const iconColor = this.props.variant === 'filled' ? 'actionSecondaryBaseColor' : 'baseColor';
    return this.props.children && Icon ? _jsx("span", {
      css: this.props.styles?.icon,
      children: _jsx(Icon, {
        size: iconSize,
        color: iconColor
      })
    }) : null;
  }
  renderDetails(expanded, detailsProps) {
    if (!this.props.children) return null;
    const expandedStyles = expanded ? {
      display: 'block'
    } : {
      display: 'none'
    };
    return _jsx("div", {
      ...detailsProps,
      css: [this.props.styles?.details, expandedStyles],
      children: expanded && this.renderContent()
    });
  }
  renderContent() {
    return _jsx("div", {
      css: this.props.styles?.content,
      children: this.props.children
    });
  }
  handleToggle = (event, expanded) => {
    if (typeof this.props.onToggle === 'function') {
      this.props.onToggle(event, expanded);
    }
  };
  render() {
    return _jsx(Expandable, {
      ...pickProps(this.props, Expandable.allowedProps),
      onToggle: this.handleToggle,
      children: ({
        expanded,
        getToggleProps,
        getDetailsProps
      }) => {
        return _jsxs("div", {
          css: this.props.styles?.toggleDetails,
          ref: this.ref,
          "data-cid": "ToggleDetails",
          children: [this.renderToggle(getToggleProps(), expanded), this.renderDetails(expanded, getDetailsProps())]
        });
      }
    });
  }
}) || _class);
export default ToggleDetails;
export { ToggleDetails };
