// @instructure/ui-svg-images/lib/InlineSVG/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.InlineSVG = void 0;
var _react = require("react");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _combineDataCid = require("@instructure/ui-utils/lib/combineDataCid.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components/utilities
---
**/
let InlineSVG = exports.InlineSVG = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = _dec2(_class = class InlineSVG extends _react.Component {
  static displayName = "InlineSVG";
  static componentId = 'InlineSVG';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    focusable: false,
    src: '',
    title: '',
    description: '',
    inline: true,
    width: '1em',
    height: '1em',
    color: 'inherit'
  };
  titleId;
  descId;
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  constructor(props) {
    super(props);
    this.titleId = props.deterministicId('InlineSVG-title');
    this.descId = props.deterministicId('InlineSVG-desc');
  }
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  static prepareSrc = src => {
    const pattern = /<svg[^>]*>((.|[\n\r])*)<\/svg>/;
    const matches = pattern.exec(src);
    return matches ? matches[1] : src;
  };
  get role() {
    if (this.props.title) {
      return 'img';
    } else {
      return 'presentation';
    }
  }
  renderTitle() {
    const {
      title
    } = this.props;
    return title ? (0, _jsxRuntime.jsx)("title", {
      id: this.titleId,
      children: title
    }) : null;
  }
  renderDesc(desc) {
    return desc ? (0, _jsxRuntime.jsx)("desc", {
      id: this.descId,
      children: desc
    }) : null;
  }
  get labelledBy() {
    const ids = [];
    if (this.props.title) {
      ids.push(this.titleId);
    }
    if (this.props.description) {
      ids.push(this.descId);
    }
    return ids.length > 0 ? ids.join(' ') : undefined;
  }
  renderContent() {
    if (this.props.src) {
      const src = InlineSVG.prepareSrc(this.props.src);
      return (0, _jsxRuntime.jsx)("g", {
        role: "presentation"
        // eslint-disable-next-line react/no-danger
        ,
        dangerouslySetInnerHTML: {
          __html: src
        }
      });
    } else {
      return (0, _jsxRuntime.jsx)("g", {
        role: "presentation",
        children: this.props.children
      });
    }
  }

  // Firefox workaround: If width/height is a string that's a number (e.g. '100')
  // it needs to be converted to a number, otherwise it is not added to the
  // stylesheet
  numberToNumberType(n) {
    if (typeof n === 'string') {
      const parsed = parseFloat(n);
      // from https://stackoverflow.com/a/52986361/319473
      if (!isNaN(parsed) && isFinite(n)) {
        return parsed;
      }
    }
    return n;
  }
  render() {
    const {
      style,
      title,
      description,
      focusable,
      src,
      styles,
      ...props
    } = this.props;

    // if width or height are 'auto', don't supply anything to the SVG
    const width = this.props.width === 'auto' ? undefined : this.numberToNumberType(this.props.width);
    const height = this.props.height === 'auto' ? undefined : this.numberToNumberType(this.props.height);
    return (0, _jsxRuntime.jsxs)("svg", {
      ...parseAttributes(src),
      ...(0, _omitProps.omitProps)(this.props, InlineSVG.allowedProps, ['inline']),
      style: {
        ...style,
        width,
        height
      },
      width: width,
      height: height,
      "aria-hidden": title ? undefined : 'true',
      "aria-labelledby": this.labelledBy,
      role: this.role,
      focusable: focusable ? 'true' : 'false',
      css: styles?.inlineSVG,
      className: props.className,
      ref: this.handleRef,
      "data-cid": (0, _combineDataCid.combineDataCid)('InlineSVG', this.props),
      children: [this.renderTitle(), this.renderDesc(description), this.renderContent()]
    });
  }
}) || _class) || _class);
function parseAttributes(src) {
  const attributes = {};
  const SVGAttributesRegExp = /<svg\s+([^>]*)\s*>/;
  // Match either quoted or unquoted attribute values
  // Handles both standard format: attr="value" and non-standard: attr=value
  // (\S+?) = attribute name, (?:["']([^"']*)["']|(\S+)) = quoted (group 2) OR unquoted (group 3) value
  const namesAndValuesRegExp = /(\S+?)=(?:["']([^"']*)["']|(\S+))/g;
  if (typeof src === 'string') {
    const attributesMatches = SVGAttributesRegExp.exec(src);
    const attributesString = attributesMatches ? attributesMatches[1] : '';
    const excludes = ['xmlns', 'xmlns:xlink', 'version'];
    let match = namesAndValuesRegExp.exec(attributesString);
    while (match != null) {
      if (excludes.indexOf(match[1]) === -1) {
        // match[1] = attribute name, match[2] = quoted value, match[3] = unquoted value
        attributes[match[1]] = match[2] || match[3];
      }
      match = namesAndValuesRegExp.exec(attributesString);
    }
  }
  return attributes;
}
var _default = exports.default = InlineSVG;
