// @instructure/ui-top-nav-bar/lib/TopNavBar/v1/TopNavBarUser/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TopNavBarUser = void 0;
require("core-js/modules/es.array.includes.js");
var _react = require("react");
var _console = require("@instructure/console");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _emotion = require("@instructure/emotion");
var _TopNavBarContext = require("../TopNavBarContext");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: TopNavBar
id: TopNavBar.User
---
@module TopNavBarUser
**/
let TopNavBarUser = exports.TopNavBarUser = (_dec = (0, _emotion.withStyle)(_styles.default, null), _dec(_class = class TopNavBarUser extends _react.Component {
  static displayName = "TopNavBarUser";
  static componentId = 'TopNavBar.User';
  static allowedProps = _props.allowedProps;
  static defaultProps = {};
  static contextType = _TopNavBarContext.TopNavBarContext;
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  get content() {
    const children = _react.Children.toArray(this.props.children);
    const allowedVariants = ['default', 'button', 'avatar'];
    return _react.Children.map(children, child => {
      const {
        id,
        variant
      } = child.props;
      if (variant && !allowedVariants.includes(variant)) {
        (0, _console.error)(false, `Item with id "${id}" has "${variant}" variant, but only the following variants are allowed in <TopNavBarUser>: ${allowedVariants.join(', ')}.`);
        return null;
      }
      return child;
    });
  }
  render() {
    const {
      styles
    } = this.props;
    if (!this.content || this.content.length === 0) {
      return null;
    }
    if (this.context.layout === 'smallViewport') {
      // in smallViewport mode it is rendered as a Drilldown
      return null;
    }
    return (0, _jsxRuntime.jsx)("div", {
      ...(0, _omitProps.omitProps)(this.props, _props.allowedProps),
      ref: this.handleRef,
      css: styles?.topNavBarUser,
      "data-cid": "TopNavBarUser",
      children: this.content
    });
  }
}) || _class);
var _default = exports.default = TopNavBarUser;
