// @instructure/ui-table/lib/Table/v1/Row/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Row = void 0;
var _react = require("react");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _v11_ = require("@instructure/ui-view/v11_6");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _TableContext = _interopRequireDefault(require("../TableContext"));
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: Table
id: Table.Row
---
**/
let Row = exports.Row = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class Row extends _react.Component {
  static displayName = "Row";
  static componentId = 'Table.Row';
  static contextType = _TableContext.default;
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    children: null
  };
  componentDidMount() {
    this.props.makeStyles?.({
      isStacked: this.context.isStacked,
      hover: this.context.hover
    });
  }
  componentDidUpdate() {
    this.props.makeStyles?.({
      isStacked: this.context.isStacked,
      hover: this.context.hover
    });
  }
  render() {
    const {
      children,
      styles
    } = this.props;
    const isStacked = this.context.isStacked;
    const headers = this.context.headers;
    return (0, _jsxRuntime.jsx)(_v11_.View, {
      ..._v11_.View.omitViewProps((0, _omitProps.omitProps)(this.props, Row.allowedProps), Row),
      as: isStacked ? 'div' : 'tr',
      css: styles?.row,
      role: isStacked ? 'row' : undefined,
      children: _react.Children.toArray(children).filter(Boolean).map((child, index) => {
        if (/*#__PURE__*/(0, _react.isValidElement)(child)) {
          return (0, _safeCloneElement.safeCloneElement)(child, {
            key: child.props.name,
            // used by `Cell` to render its column title in `stacked` layout
            header: headers && headers[index]
          });
        }
        return child;
      })
    });
  }
}) || _class);
var _default = exports.default = Row;
