// @instructure/ui-drawer-layout/lib/DrawerLayout/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DrawerContent", {
  enumerable: true,
  get: function () {
    return _DrawerContent.DrawerContent;
  }
});
exports.DrawerLayoutContext = exports.DrawerLayout = void 0;
Object.defineProperty(exports, "DrawerTray", {
  enumerable: true,
  get: function () {
    return _DrawerTray.DrawerTray;
  }
});
exports.default = void 0;
var _react = require("react");
var _textDirectionContextConsumer = require("@instructure/ui-i18n/lib/textDirectionContextConsumer.js");
var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _getBoundingClientRect = require("@instructure/ui-dom-utils/lib/getBoundingClientRect.js");
var _createChainedFunction = require("@instructure/ui-utils/lib/createChainedFunction.js");
var _px = require("@instructure/ui-utils/lib/px.js");
var _console = require("@instructure/console");
var _mirrorHorizontalPlacement = require("@instructure/ui-position/lib/mirrorHorizontalPlacement.js");
var _DrawerContent = require("./DrawerContent");
var _DrawerTray = require("./DrawerTray");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _dec3, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let DrawerLayout = exports.DrawerLayout = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyleNew)(_styles.default), _dec3 = (0, _textDirectionContextConsumer.textDirectionContextConsumer)(), _dec(_class = _dec2(_class = _dec3(_class = class DrawerLayout extends _react.Component {
  static displayName = "DrawerLayout";
  static componentId = 'DrawerLayout';
  static locatorAttribute = 'data-drawer-layout';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    minWidth: '30rem'
  };
  static Content = _DrawerContent.DrawerContent;
  static Tray = _DrawerTray.DrawerTray;
  constructor(props) {
    super(props);
    this.state = {
      shouldOverlayTray: false,
      trayWidth: 0,
      contentWidth: 0
    };
    this._id = props.deterministicId();
  }
  _id;
  _content = null;
  _tray = null;
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  get trayProps() {
    const tray = _react.Children.toArray(this.props.children).filter(child => (0, _matchComponentTypes.matchComponentTypes)(child, [_DrawerTray.DrawerTray]))[0];
    return tray.props;
  }
  get trayPlacement() {
    const {
      placement,
      dir = this.props.dir
    } = this.trayProps;
    const isRtl = dir === _textDirectionContextConsumer.textDirectionContextConsumer.DIRECTION.rtl;
    return isRtl ? (0, _mirrorHorizontalPlacement.mirrorHorizontalPlacement)(placement, ' ') : placement;
  }
  get contentMargin() {
    const trayWidth = this.state.trayWidth || 0;
    return this.state.shouldOverlayTray ? 0 : trayWidth;
  }
  get contentStyle() {
    const shouldOverlayTray = this.shouldOverlayTray(this.props.minWidth, this.state.trayWidth, this.state.contentWidth, this.state.shouldOverlayTray);
    let marginLeft = 0;
    let marginRight = 0;
    if (!shouldOverlayTray) {
      if (this.trayPlacement === 'start') {
        marginLeft = this.contentMargin;
      }
      if (this.trayPlacement === 'end') {
        marginRight = this.contentMargin;
      }
    }
    return {
      marginLeft: `${marginLeft}px`,
      marginRight: `${marginRight}px`
    };
  }
  handleTrayContentRef = el => {
    this._tray = el;
  };
  shouldOverlayTray(minWidth, trayWidth, contentWidth, trayIsOverlayed) {
    if (!this._content) return false;
    const minWidthPx = (0, _px.px)(minWidth, this._content);
    if (trayIsOverlayed) {
      return contentWidth - trayWidth < minWidthPx;
    } else {
      return contentWidth < minWidthPx;
    }
  }
  getNextState(minWidth, trayWidth, contentWidth, trayIsOverlayed) {
    const shouldOverlayTray = this.shouldOverlayTray(minWidth, trayWidth, contentWidth, trayIsOverlayed);
    return {
      trayWidth,
      contentWidth,
      shouldOverlayTray
    };
  }
  notifyOverlayTrayChange(shouldOverlayTray) {
    const {
      onOverlayTrayChange
    } = this.props;
    if (typeof onOverlayTrayChange === 'function') {
      onOverlayTrayChange(shouldOverlayTray);
    }
  }
  handleContentSizeChange = ({
    width
  }) => {
    this.setState((state, props) => {
      const nextState = this.getNextState(props.minWidth, state.trayWidth, width, state.shouldOverlayTray);
      if (state.shouldOverlayTray !== nextState.shouldOverlayTray) {
        this.notifyOverlayTrayChange(nextState.shouldOverlayTray);
      }
      return nextState;
    });
  };
  handleTraySizeChange = ({
    width
  }) => {
    this.setState((state, props) => {
      const nextState = this.getNextState(props.minWidth, width, state.contentWidth, true);
      if (state.shouldOverlayTray !== nextState.shouldOverlayTray) {
        this.notifyOverlayTrayChange(nextState.shouldOverlayTray);
      }
      return nextState;
    });
  };
  handleTrayTransitionEnter = () => {
    let width = 0;
    if (this._tray) {
      width = (0, _getBoundingClientRect.getBoundingClientRect)(this._tray).width;
    }
    this.handleTraySizeChange({
      width
    });
  };
  handleTrayTransitionExit = () => {
    this.handleTraySizeChange({
      width: 0
    });
  };
  renderChildren() {
    let trayCount = 0;
    let contentCount = 0;
    const children = _react.Children.map(this.props.children, (child, _index) => {
      if ((0, _matchComponentTypes.matchComponentTypes)(child, [_DrawerTray.DrawerTray])) {
        trayCount++;
        return (0, _safeCloneElement.safeCloneElement)(child, {
          label: child.props.label,
          key: child.props.label,
          [_DrawerTray.DrawerTray.locatorAttribute]: this._id,
          contentRef: this.handleTrayContentRef,
          onEnter: this.handleTrayTransitionEnter,
          onExit: this.handleTrayTransitionExit
        });
      } else if ((0, _matchComponentTypes.matchComponentTypes)(child, [_DrawerContent.DrawerContent])) {
        contentCount++;
        const handleContentRef = el => {
          const {
            contentRef
          } = child.props;
          this._content = el;
          if (typeof contentRef === 'function') {
            contentRef(el);
          }
        };
        return this.state.trayWidth !== null ? (0, _safeCloneElement.safeCloneElement)(child, {
          label: child.props.label,
          key: child.props.label,
          [_DrawerContent.DrawerContent.locatorAttribute]: this._id,
          style: this.contentStyle,
          onSizeChange: (0, _createChainedFunction.createChainedFunction)(this.handleContentSizeChange, child.props.onSizeChange),
          contentRef: handleContentRef
        }) : null;
      } else {
        return child;
      }
    });
    (0, _console.logError)(trayCount <= 1, `[DrawerLayout] Only one 'DrawerTray' per 'DrawerLayout' is supported.`);
    (0, _console.logError)(contentCount <= 1, `[DrawerLayout] Only one 'DrawerContent' per 'DrawerLayout' is supported.`);
    return children;
  }
  render() {
    const props = {
      [DrawerLayout.locatorAttribute]: this._id
    };
    return (0, _jsxRuntime.jsx)(DrawerLayoutContext.Provider, {
      value: this.state.shouldOverlayTray,
      children: (0, _jsxRuntime.jsx)("div", {
        ...props,
        css: this.props.styles?.drawerLayout,
        ref: this.handleRef,
        "data-cid": "DrawerLayout",
        children: this.renderChildren()
      })
    });
  }
}) || _class) || _class) || _class);
var _default = exports.default = DrawerLayout;
const DrawerLayoutContext = exports.DrawerLayoutContext = /*#__PURE__*/(0, _react.createContext)(false);
