// @instructure/ui-themes/lib/themes/canvas/colors.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.colors = void 0;
var _primitives = require("../../legacySharedThemeTokens/colors/primitives");
var _dataVisualization = _interopRequireDefault(require("../../legacySharedThemeTokens/colors/dataVisualization"));
var _getUIColors = _interopRequireDefault(require("../../utils/getUIColors"));
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const contrasts = {
  white1010: _primitives.primitives.white,
  white1010op75: _primitives.primitives.white10op75,
  grey1111: _primitives.primitives.grey11,
  grey1214: _primitives.primitives.grey12,
  grey1424: _primitives.primitives.grey14,
  grey2424: _primitives.primitives.grey24,
  grey3045: _primitives.primitives.grey30,
  grey4570: _primitives.primitives.grey45,
  grey5782: _primitives.primitives.grey57,
  grey100100: _primitives.primitives.grey100,
  grey100100op75: _primitives.primitives.grey100op75,
  grey125125: _primitives.primitives.grey125,
  blue1212: _primitives.primitives.blue12,
  blue4570: _primitives.primitives.blue45,
  blue5782: _primitives.primitives.blue57,
  green1212: _primitives.primitives.green12,
  green4570: _primitives.primitives.green45,
  green5782: _primitives.primitives.green57,
  orange1212: _primitives.primitives.orange12,
  orange3045: _primitives.primitives.orange30,
  orange4570: _primitives.primitives.orange45,
  orange5782: _primitives.primitives.orange57,
  red1212: _primitives.primitives.red12,
  red4570: _primitives.primitives.red45,
  red5782: _primitives.primitives.red57,
  violet1212: _primitives.primitives.violet12,
  violet4570: _primitives.primitives.violet45,
  violet5790: _primitives.primitives.violet57,
  sea4570: _primitives.primitives.sea45,
  sea5790: _primitives.primitives.sea57
};
const ui = (0, _getUIColors.default)(contrasts);
const colors = exports.colors = {
  primitives: _primitives.primitives,
  additionalPrimitives: _primitives.additionalPrimitives,
  contrasts,
  ui,
  dataVisualization: _dataVisualization.default
};
var _default = exports.default = {
  primitives: _primitives.primitives,
  contrasts,
  ui
};
