// @instructure/ui-pagination/lib/Pagination/v1/PaginationArrowButton/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.PaginationArrowButton = void 0;
var _react = require("react");
var _v11_ = require("@instructure/ui-buttons/v11_6");
var _PresentationContent = require("@instructure/ui-a11y-content/lib/PresentationContent");
var _v11_2 = require("@instructure/ui-tooltip/v11_6");
var _IconArrowOpenStartSolid = require("@instructure/ui-icons/lib/generated/IconArrowOpenStartSolid.js");
var _IconArrowOpenEndSolid = require("@instructure/ui-icons/lib/generated/IconArrowOpenEndSolid.js");
var _IconArrowDoubleStartSolid = require("@instructure/ui-icons/lib/generated/IconArrowDoubleStartSolid.js");
var _IconArrowDoubleEndSolid = require("@instructure/ui-icons/lib/generated/IconArrowDoubleEndSolid.js");
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
parent: Pagination
id: Pagination.Navigation
---
**/class PaginationArrowButton extends _react.Component {
  static displayName = "PaginationArrowButton";
  static componentId = 'Pagination.Navigation';
  static allowedProps = _props.allowedProps;
  static defaultProps = {};
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  get margin() {
    switch (this.props.direction) {
      case 'first':
        return '0 xx-small 0 0';
      case 'last':
        return '0 0 0 xx-small';
      default:
        return undefined;
    }
  }
  get Icon() {
    switch (this.props.direction) {
      case 'first':
        return _IconArrowDoubleStartSolid.IconArrowDoubleStartSolid;
      case 'prev':
        return _IconArrowOpenStartSolid.IconArrowOpenStartSolid;
      case 'next':
        return _IconArrowOpenEndSolid.IconArrowOpenEndSolid;
      case 'last':
        return _IconArrowDoubleEndSolid.IconArrowDoubleEndSolid;
      default:
        return null;
    }
  }
  render() {
    const {
      label,
      direction,
      buttonRef,
      ...props
    } = this.props;
    return (0, _jsxRuntime.jsx)(_v11_2.Tooltip, {
      elementRef: this.handleRef,
      on: ['hover', 'focus'],
      renderTip: (0, _jsxRuntime.jsx)(_PresentationContent.PresentationContent, {
        children: label
      }),
      children: (0, _jsxRuntime.jsx)(_v11_.IconButton, {
        ...props,
        size: "small",
        withBackground: false,
        withBorder: false,
        screenReaderLabel: label,
        rel: props.href || props.to ? direction : undefined,
        elementRef: buttonRef,
        margin: this.margin,
        "data-cid": "PaginationArrowButton",
        children: this.Icon
      })
    });
  }
}
exports.PaginationArrowButton = PaginationArrowButton;
var _default = exports.default = PaginationArrowButton;
