// @instructure/ui-color-picker/es/ColorMixer/v2/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { withStyleNew } from '@instructure/emotion';
import { omitProps } from '@instructure/ui-react-utils';
import { isValid } from '@instructure/ui-color-utils';
import conversions from '@instructure/ui-color-utils';
import { logWarn as warn } from '@instructure/console';
import ColorPalette from "./ColorPalette/index.js";
import Slider from "./Slider/index.js";
import RGBAInput from "./RGBAInput/index.js";
import { allowedProps } from "./props.js";
import generateStyle from "./styles.js";
/**
---
category: components
---
**/
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
let ColorMixer = (_dec = withStyleNew(generateStyle), _dec(_class = class ColorMixer extends Component {
  static displayName = "ColorMixer";
  static allowedProps = allowedProps;
  static componentId = 'ColorMixer';
  static defaultProps = {
    value: '#000',
    withAlpha: false,
    disabled: false
  };
  constructor(props) {
    super(props);
    this.state = {
      h: 0,
      s: 0,
      v: 0,
      a: 1
    };
  }
  ref = null;
  _width = 272;
  _paletteHeight = 160;
  _sliderHeight = 8;
  _sliderIndicatorRadius = 6;
  _paletteIndicatorRadius = 6;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.();
    warn(isValid(this.props.value), `[ColorMixer] The passed color value is not valid.`);
    this.setState({
      ...conversions.colorToHsva(this.props.value)
    });
  }
  componentDidUpdate(prevProps, prevState) {
    this.props.makeStyles?.();
    const {
      h,
      s,
      v,
      a
    } = this.state;
    if (prevState.h !== h || prevState.s !== s || prevState.v !== v || prevState.a !== a) {
      this.props.onChange(conversions.colorToHex8({
        h,
        s,
        v,
        a
      }));
    }
    if (prevProps.value !== this.props.value && conversions.colorToHex8({
      h,
      s,
      v,
      a
    }) !== this.props.value) {
      this.setState({
        ...conversions.colorToHsva(this.props.value)
      });
    }
  }
  render() {
    const {
      disabled,
      styles,
      withAlpha,
      rgbRedInputScreenReaderLabel,
      rgbGreenInputScreenReaderLabel,
      rgbBlueInputScreenReaderLabel,
      rgbAlphaInputScreenReaderLabel,
      colorSliderNavigationExplanationScreenReaderLabel,
      alphaSliderNavigationExplanationScreenReaderLabel,
      colorPaletteNavigationExplanationScreenReaderLabel
    } = this.props;
    const {
      h,
      s,
      v,
      a
    } = this.state;
    return _jsxs("div", {
      ...omitProps(this.props, ColorMixer.allowedProps),
      "aria-disabled": disabled,
      ref: this.handleRef,
      css: styles?.colorMixer,
      "data-cid": "ColorMixer",
      children: [_jsxs("span", {
        css: styles?.sliderAndPaletteContainer,
        "aria-label": `${conversions.colorToHex8({
          h,
          s,
          v,
          a
        })}`,
        "aria-live": "polite",
        children: [_jsx(ColorPalette, {
          disabled: disabled,
          width: this._width,
          height: this._paletteHeight,
          indicatorRadius: this._paletteIndicatorRadius,
          hue: h,
          color: {
            h,
            s,
            v
          },
          onChange: color => {
            this.setState({
              s: color.s,
              v: color.v
            });
          },
          navigationExplanationScreenReaderLabel: colorPaletteNavigationExplanationScreenReaderLabel
        }), _jsx(Slider, {
          disabled: disabled,
          isColorSlider: true,
          width: this._width,
          height: this._sliderHeight,
          indicatorRadius: this._sliderIndicatorRadius,
          value: h,
          minValue: 0,
          maxValue: 359,
          color: conversions.colorToHex8({
            h,
            s,
            v,
            a
          }),
          onChange: hue => {
            this.setState({
              h: hue
            });
          },
          navigationExplanationScreenReaderLabel: colorSliderNavigationExplanationScreenReaderLabel
        }), withAlpha && _jsx(Slider, {
          disabled: disabled,
          width: this._width,
          height: this._sliderHeight,
          indicatorRadius: this._sliderIndicatorRadius,
          color: conversions.colorToHex8({
            h,
            s,
            v
          }),
          value: a,
          minValue: 0,
          maxValue: 1,
          onChange: opacity => this.setState({
            a: opacity / 100
          }),
          navigationExplanationScreenReaderLabel: alphaSliderNavigationExplanationScreenReaderLabel
        })]
      }), _jsx(RGBAInput, {
        disabled: disabled,
        label: withAlpha ? 'RGBA' : 'RGB',
        width: this._width,
        value: conversions.colorToRGB({
          h,
          s,
          v,
          a
        }),
        onChange: color => this.setState({
          ...conversions.colorToHsva(color)
        }),
        withAlpha: withAlpha,
        rgbRedInputScreenReaderLabel: rgbRedInputScreenReaderLabel,
        rgbGreenInputScreenReaderLabel: rgbGreenInputScreenReaderLabel,
        rgbBlueInputScreenReaderLabel: rgbBlueInputScreenReaderLabel,
        rgbAlphaInputScreenReaderLabel: rgbAlphaInputScreenReaderLabel
      })]
    });
  }
}) || _class);
export { ColorMixer };
export default ColorMixer;
