// @instructure/ui-radio-input/lib/RadioInputGroup/v2/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.RadioInputGroup = void 0;
var _react = require("react");
var _latest = require("@instructure/ui-form-field/latest");
var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _pickProps = require("@instructure/ui-react-utils/lib/pickProps.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _v = require("../../RadioInput/v2");
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let RadioInputGroup = exports.RadioInputGroup = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec(_class = class RadioInputGroup extends _react.Component {
  static displayName = "RadioInputGroup";
  static componentId = 'RadioInputGroup';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    disabled: false,
    variant: 'simple',
    size: 'medium',
    layout: 'stacked',
    readOnly: false
  };
  ref = null;
  _messagesId;
  handleRef = el => {
    this.ref = el;
  };
  constructor(props) {
    super(props);
    if (typeof props.value === 'undefined') {
      this.state = {
        value: props.defaultValue
      };
    }
    this._messagesId = props.deterministicId('RadioInputGroup-messages');
  }
  get hasMessages() {
    return !!this.props.messages && this.props.messages.length > 0;
  }
  get invalid() {
    return !!this.props.messages?.find(m => m.type === 'newError' || m.type === 'error');
  }
  handleChange = e => {
    const value = e.target.value;
    if (this.props.disabled || this.props.readOnly) {
      e.preventDefault();
      return;
    }
    if (typeof this.props.value === 'undefined') {
      this.setState({
        value
      });
    }
    if (typeof this.props.onChange === 'function') {
      this.props.onChange(e, value);
    }
  };
  get value() {
    return typeof this.props.value === 'undefined' ? this.state.value : this.props.value;
  }
  renderChildren() {
    const {
      children,
      name,
      variant,
      size,
      disabled,
      readOnly
    } = this.props;

    // This adds the passed in name property to each RadioInput component
    // and checks the input whose value matches the value property
    return _react.Children.map(children, (child, index) => {
      if ((0, _matchComponentTypes.matchComponentTypes)(child, [_v.RadioInput])) {
        const isChecked = this.value === child.props.value;
        const defaultFocus = !this.value && index === 0;
        return (0, _safeCloneElement.safeCloneElement)(child, {
          name,
          disabled: disabled || child.props.disabled,
          variant,
          size,
          checked: isChecked,
          onChange: this.handleChange,
          readOnly: readOnly || child.props.readOnly,
          width: child.props.width || 'auto',
          'aria-describedby': this.hasMessages ? this._messagesId : undefined,
          // only one radio in a group should be considered tabbable
          // if a radio is checked, it should be the input to receive focus when tabbed to
          // if none of the inputs are checked, the first should receive the focus
          tabIndex: isChecked || defaultFocus ? 0 : -1,
          label: child.props.label
        });
      } else {
        return child; // ignore (but preserve) children that aren't RadioInput
      }
    });
  }
  render() {
    const {
      variant,
      layout,
      description,
      isRequired
    } = this.props;
    return (0, _jsxRuntime.jsx)(_latest.FormFieldGroup, {
      ...(0, _omitProps.omitProps)(this.props, RadioInputGroup.allowedProps),
      ...(0, _pickProps.pickProps)(this.props, _latest.FormFieldGroup.allowedProps),
      description: description,
      layout: layout === 'columns' && variant === 'toggle' ? 'stacked' : layout // toggles already display in cols
      ,
      vAlign: variant === 'toggle' ? 'middle' : 'top',
      rowSpacing: "small",
      colSpacing: variant === 'toggle' ? 'none' : 'small' // keep toggles close together
      ,
      startAt: variant === 'toggle' ? 'small' : undefined,
      messagesId: this._messagesId,
      elementRef: this.handleRef,
      role: "radiogroup",
      isRequired: isRequired,
      "data-cid": "RadioInputGroup",
      children: this.renderChildren()
    });
  }
}) || _class);
var _default = exports.default = RadioInputGroup;
