// @instructure/ui-alerts/lib/Alert/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Alert = void 0;
var _react = require("react");
var _reactDom = _interopRequireDefault(require("react-dom"));
var _keycode = _interopRequireDefault(require("keycode"));
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _latest = require("@instructure/ui-buttons/latest");
var _latest2 = require("@instructure/ui-view/latest");
var _ScreenReaderContent = require("@instructure/ui-a11y-content/lib/ScreenReaderContent");
var _uiIcons = require("@instructure/ui-icons");
var _Transition = require("@instructure/ui-motion/lib/Transition");
var _console = require("@instructure/console");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let Alert = exports.Alert = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = _dec2(_class = class Alert extends _react.Component {
  static displayName = "Alert";
  static componentId = 'Alert';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    variant: 'info',
    margin: 'x-small 0',
    timeout: 0,
    transition: 'fade',
    open: true,
    screenReaderOnly: false,
    liveRegionPoliteness: 'assertive',
    isLiveRegionAtomic: false,
    children: null,
    hasShadow: true
  };
  constructor(props) {
    super(props);
    this.srid = this.props.deterministicId();
    this.state = {
      open: true
    };
  }
  _timeouts = [];
  srid;
  variantUI = {
    error: _uiIcons.XCircleInstUIIcon,
    info: _uiIcons.InfoInstUIIcon,
    success: _uiIcons.CircleCheckInstUIIcon,
    warning: _uiIcons.TriangleAlertInstUIIcon
  };
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  handleTimeout = () => {
    if (this.props.timeout > 0) {
      this._timeouts.push(setTimeout(() => {
        this.close();
      }, this.props.timeout));
    }
  };
  clearTimeouts() {
    this._timeouts.forEach(timeout => clearTimeout(timeout));
    this._timeouts = [];
  }
  onExitTransition = () => {
    if (this.props.onDismiss) {
      this.props.onDismiss();
    }
  };
  close = () => {
    this.clearTimeouts();
    this.removeScreenreaderAlert();
    this.setState({
      open: false
    }, () => {
      if (this.props.onDismiss && (this.props.transition === 'none' || this.props.screenReaderOnly)) {
        this.props.onDismiss();
      }
    });
  };

  // duck type for a dom node
  isDOMNode(n) {
    return n && typeof n === 'object' && n.nodeType === 1;
  }
  getLiveRegion() {
    const lr = typeof this.props.liveRegion === 'function' ? this.props.liveRegion() : this.props.liveRegion;
    return this.isDOMNode(lr) ? lr : null;
  }
  initLiveRegion(liveRegion) {
    (0, _console.logError)(liveRegion.getAttribute('role') === 'alert', `[Alert] live region must have role='alert' set on page load in order to announce content`);
    if (liveRegion) {
      liveRegion.setAttribute('aria-live', this.props.liveRegionPoliteness);
      // indicates what notifications the user agent will trigger when the
      // accessibility tree within a live region is modified.
      // additions: elements are added, text: Text content is added
      liveRegion.setAttribute('aria-relevant', 'additions text');
      liveRegion.setAttribute('aria-atomic', `${this.props.isLiveRegionAtomic}`);
    }
  }
  createScreenreaderContentNode() {
    return (0, _jsxRuntime.jsxs)(_ScreenReaderContent.ScreenReaderContent, {
      children: [this.props.variantScreenReaderLabel || '', " ", this.props.children]
    });
  }
  createScreenreaderAlert() {
    const liveRegion = this.getLiveRegion();
    if (liveRegion) {
      const div = document.createElement('div');
      div.setAttribute('id', this.srid);
      liveRegion.appendChild(div);
    }
  }
  removeScreenreaderAlert() {
    const liveRegion = this.getLiveRegion();
    if (liveRegion) {
      const div = document.getElementById(this.srid);
      if (div) {
        // Accessibility attributes must be removed for the deletion of the node
        // and then reapplied because JAWS/IE will not respect the
        // "aria-relevant" attribute and read when the node is deleted if
        // the attributes are in place
        liveRegion.removeAttribute('aria-live');
        liveRegion.removeAttribute('aria-relevant');
        liveRegion.removeAttribute('aria-atomic');
        this.initLiveRegion(liveRegion);
      }
    }
  }
  handleKeyUp = event => {
    if (this.props.renderCloseButtonLabel && event.keyCode === _keycode.default.codes.esc) {
      this.close();
    }
  };
  componentDidMount() {
    this.props.makeStyles?.();
    const liveRegion = this.getLiveRegion();
    if (liveRegion) {
      this.initLiveRegion(liveRegion);
    }
    this.handleTimeout();
  }
  componentWillUnmount() {
    this.clearTimeouts();
  }
  componentDidUpdate(prevProps) {
    this.props.makeStyles?.();
    if (!!this.props.open === false && !!this.props.open !== !!prevProps.open) {
      // this outside world is asking us to close the alert, which needs to
      // take place internally so the transition runs
      this.close();
    }
  }
  renderIcon() {
    const {
      renderCustomIcon,
      variant,
      styles
    } = this.props;
    const Icon = this.variantUI[variant];
    return (0, _jsxRuntime.jsx)("div", {
      css: styles?.icon,
      children: renderCustomIcon ? (0, _callRenderProp.callRenderProp)(renderCustomIcon) : (0, _jsxRuntime.jsx)(Icon, {})
    });
  }
  renderCloseButton() {
    const closeButtonLabel = this.props.renderCloseButtonLabel && (0, _callRenderProp.callRenderProp)(this.props.renderCloseButtonLabel);
    return closeButtonLabel ? (0, _jsxRuntime.jsx)("div", {
      css: this.props.styles?.closeButton,
      children: (0, _jsxRuntime.jsx)(_latest.CloseButton, {
        onClick: this.close,
        size: "small",
        screenReaderLabel: closeButtonLabel
      })
    }, "closeButton") : null;
  }
  renderAlert() {
    // prevent onDismiss from being passed to the View component
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const {
      margin,
      styles,
      children,
      onDismiss,
      variantScreenReaderLabel,
      ...props
    } = this.props;
    return (0, _jsxRuntime.jsxs)(_latest2.View, {
      ...(0, _passthroughProps.passthroughProps)({
        ...props
      }),
      as: "div",
      margin: margin,
      css: styles?.alert,
      onKeyUp: this.handleKeyUp,
      elementRef: this.handleRef,
      children: [this.renderIcon(), (0, _jsxRuntime.jsxs)("div", {
        css: styles?.content,
        children: [variantScreenReaderLabel && (0, _jsxRuntime.jsx)("span", {
          css: styles?.variantScreenReaderLabel,
          children: variantScreenReaderLabel
        }), children]
      }), this.renderCloseButton()]
    });
  }
  createScreenReaderPortal(liveRegion) {
    const {
      open
    } = this.state;
    return open ? /*#__PURE__*/_reactDom.default.createPortal((0, _jsxRuntime.jsx)("div", {
      id: this.srid,
      children: this.createScreenreaderContentNode()
    }), liveRegion) : null;
  }
  render() {
    const liveRegion = this.getLiveRegion();
    const screenReaderContent = liveRegion ? this.createScreenReaderPortal(liveRegion) : null;
    // Don't render anything if screen reader only
    if (this.props.screenReaderOnly) {
      (0, _console.logError)(!!this.getLiveRegion(), `[Alert] The 'screenReaderOnly' prop must be used in conjunction with 'liveRegion'.`);
      return screenReaderContent;
    }
    if (this.props.transition === 'none') {
      return this.state.open ? (0, _jsxRuntime.jsxs)(_react.Fragment, {
        children: [screenReaderContent, this.renderAlert()]
      }) : null;
    }
    return (0, _jsxRuntime.jsxs)(_react.Fragment, {
      children: [screenReaderContent, (0, _jsxRuntime.jsx)(_Transition.Transition, {
        type: this.props.transition,
        transitionOnMount: true,
        in: this.state.open,
        unmountOnExit: true,
        onExited: this.onExitTransition,
        children: this.renderAlert()
      })]
    });
  }
}) || _class) || _class);
var _default = exports.default = Alert;
