// @instructure/ui-metric/lib/Metric/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Metric = void 0;
var _react = require("react");
var _emotion = require("@instructure/emotion");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let Metric = exports.Metric = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class Metric extends _react.Component {
  static displayName = "Metric";
  static componentId = 'Metric';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    textAlign: 'center',
    isGroupChild: false
  };
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  render() {
    const {
      textAlign,
      renderLabel,
      renderValue,
      isGroupChild,
      ...rest
    } = this.props;
    return (0, _jsxRuntime.jsxs)("div", {
      ...(0, _passthroughProps.passthroughProps)(rest),
      role: isGroupChild ? 'row' : undefined,
      css: this.props.styles?.metric,
      ref: this.handleRef,
      "data-cid": "Metric",
      children: [(0, _jsxRuntime.jsx)("div", {
        role: isGroupChild ? 'rowheader' : undefined,
        css: this.props.styles?.label,
        children: (0, _callRenderProp.callRenderProp)(renderLabel)
      }), (0, _jsxRuntime.jsx)("div", {
        role: isGroupChild ? 'gridcell' : undefined,
        css: this.props.styles?.value,
        children: (0, _callRenderProp.callRenderProp)(renderValue)
      })]
    });
  }
}) || _class);
var _default = exports.default = Metric;
