// @instructure/ui-table/lib/Table/v2/Head/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Head = void 0;
var _react = require("react");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _latest = require("@instructure/ui-simple-select/latest");
var _ScreenReaderContent = require("@instructure/ui-a11y-content/lib/ScreenReaderContent");
var _uiIcons = require("@instructure/ui-icons");
var _console = require("@instructure/console");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _TableContext = _interopRequireDefault(require("../TableContext"));
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: Table
id: Table.Head
---
**/
let Head = exports.Head = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class Head extends _react.Component {
  static displayName = "Head";
  static componentId = 'Table.Head';
  static contextType = _TableContext.default;
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    children: null
  };

  /**
   * Returns `true` if the first child's children have a `onRequestSort` prop
   */
  get isSortable() {
    const [firstRow] = _react.Children.toArray(this.props.children);
    let sortable = false;
    if (firstRow && firstRow.props && firstRow.props.children) {
      _react.Children.forEach(firstRow.props.children, grandchild => {
        if (grandchild?.props?.onRequestSort) {
          sortable = true;
          return;
        }
      });
    }
    return sortable;
  }
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    if (this.isSortable && typeof this.props.renderSortLabel === 'undefined') {
      (0, _console.warn)(false, '[Table.Head] The `renderSortLabel` prop should be provided when Table is sortable.');
    }
    this.props.makeStyles?.();
  }

  /**
   * This `Select` is used in `stacked` layout. It's populated by iterating
   * through the first child's children (by default `ColHeader`) and reading
   * there the `id`, `stackedSortByLabel`, `sortDirection`, `onRequestSort` props
   */
  renderSelect() {
    const {
      children,
      renderSortLabel
    } = this.props;
    const [firstRow] = _react.Children.toArray(children);
    if (!firstRow?.props?.children) {
      return null;
    }
    const options = [];
    const clickHandlers = {};
    let selectedOption;
    let count = 0;
    _react.Children.forEach(firstRow.props.children, grandchild => {
      count += 1;
      if (!grandchild?.props) return; // grandchild can be false
      const {
        id,
        stackedSortByLabel,
        sortDirection,
        onRequestSort
      } = grandchild.props;
      if (id && onRequestSort) {
        const label = stackedSortByLabel || id;
        options.push({
          id,
          label
        });
        clickHandlers[id] = onRequestSort;
        if (sortDirection !== 'none') {
          selectedOption = id;
        }
      }
    });
    if (!options.length) {
      return null;
    }
    const handleSelect = (event, {
      value
    }) => {
      if (value && typeof clickHandlers[value] === 'function') {
        clickHandlers[value](event, {
          id: `${value}`
        });
      }
    };
    return (0, _jsxRuntime.jsx)("div", {
      role: "rowgroup",
      children: (0, _jsxRuntime.jsx)("div", {
        role: "row",
        children: (0, _jsxRuntime.jsx)("div", {
          role: "cell",
          "aria-colspan": count,
          children: (0, _jsxRuntime.jsx)(_latest.SimpleSelect, {
            renderLabel: renderSortLabel ? (0, _callRenderProp.callRenderProp)(renderSortLabel) : (0, _jsxRuntime.jsx)(_ScreenReaderContent.ScreenReaderContent, {}),
            renderBeforeInput: selectedOption ? () => (0, _jsxRuntime.jsx)(_uiIcons.CheckInstUIIcon, {}) : undefined,
            value: selectedOption,
            onChange: handleSelect,
            children: options.map(({
              id,
              label
            }) => (0, _jsxRuntime.jsx)(_latest.SimpleSelect.Option, {
              id: id,
              value: id,
              renderBeforeLabel: id === selectedOption ? () => (0, _jsxRuntime.jsx)(_uiIcons.CheckInstUIIcon, {}) : () => (0, _jsxRuntime.jsx)("span", {
                style: {
                  visibility: 'hidden'
                },
                children: (0, _jsxRuntime.jsx)(_uiIcons.CheckInstUIIcon, {})
              }),
              children: label
            }, id))
          })
        })
      })
    });
  }
  render() {
    const {
      children,
      styles
    } = this.props;
    return this.context.isStacked ? this.renderSelect() : (0, _jsxRuntime.jsx)("thead", {
      ...(0, _omitProps.omitProps)(this.props, Head.allowedProps),
      css: styles?.head,
      children: children
    });
  }
}) || _class);
var _default = exports.default = Head;
