// @instructure/ui-heading/lib/Heading/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Heading = void 0;
require("core-js/modules/es.array.includes.js");
var _react = require("react");
var _latest = require("@instructure/ui-view/latest");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _uiIcons = require("@instructure/ui-icons");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
const variantLevels = {
  titlePageDesktop: 'h1',
  titlePageMobile: 'h1',
  titleSection: 'h2',
  titleCardSection: 'h2',
  titleModule: 'h3',
  titleCardLarge: 'h4',
  titleCardRegular: 'h4',
  titleCardMini: 'h4',
  label: 'h5',
  labelInline: 'h5'
};
const variantToIconSize = {
  titlePageDesktop: '2xl',
  titlePageMobile: 'xl',
  titleSection: 'xl',
  titleCardSection: 'xl',
  titleModule: 'xl',
  titleCardLarge: 'xl',
  titleCardRegular: 'lg',
  titleCardMini: 'md',
  label: 'md',
  labelInline: 'md'
};
const levelToIconSize = {
  h1: '2xl',
  h2: 'xl',
  h3: 'xl',
  h4: 'lg',
  h5: 'md',
  h6: 'md'
};
const variantToAIHorizontalIconSize = {
  titlePageDesktop: '2xl',
  titlePageMobile: 'xl',
  titleSection: 'xl',
  titleCardSection: 'xl',
  titleModule: 'md',
  titleCardLarge: 'md',
  titleCardRegular: 'md',
  titleCardMini: 'xs',
  label: 'xs',
  labelInline: 'xs'
};
const levelToAIHorizontalIconSize = {
  h1: '2xl',
  h2: 'lg',
  h3: 'md',
  h4: 'sm',
  h5: 'xs',
  h6: 'xs'
};

/**
---
category: components
---
**/
let Heading = exports.Heading = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class Heading extends _react.Component {
  static displayName = "Heading";
  static componentId = 'Heading';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    children: null,
    border: 'none',
    color: 'primary'
  };
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  checkProps() {
    const {
      variant,
      as
    } = this.props;
    if (variant) {
      if (as) {}
    }
  }
  componentDidMount() {
    this.props.makeStyles?.();
    this.checkProps();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
    this.checkProps();
  }
  getIconSize(forAIHorizontal = false) {
    const {
      variant,
      level,
      as
    } = this.props;
    if (variant) {
      return forAIHorizontal ? variantToAIHorizontalIconSize[variant] : variantToIconSize[variant];
    }
    if (level && level !== 'reset') {
      return forAIHorizontal ? levelToAIHorizontalIconSize[level] : levelToIconSize[level];
    }
    if (as && typeof as === 'string' && ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'].includes(as)) {
      const asLevel = as;
      return forAIHorizontal ? levelToAIHorizontalIconSize[asLevel] : levelToIconSize[asLevel];
    }
    return 'md';
  }
  renderContent() {
    const {
      children,
      renderIcon,
      aiVariant
    } = this.props;
    if (renderIcon && !aiVariant) {
      return (0, _jsxRuntime.jsxs)("span", {
        css: [this.props.styles?.withIcon],
        "aria-hidden": "true",
        children: [(0, _uiIcons.renderIconWithProps)(renderIcon, this.getIconSize(), 'inherit'), children]
      });
    }
    if (aiVariant === 'stacked') {
      return (0, _jsxRuntime.jsxs)("span", {
        css: [this.props.styles?.withIcon],
        "aria-hidden": "true",
        children: [(0, _jsxRuntime.jsxs)("span", {
          css: this.props.styles?.igniteAIStacked,
          children: [(0, _jsxRuntime.jsx)(_uiIcons.IgniteaiLogoInstUIIcon, {
            color: "ai",
            size: "sm"
          }), (0, _jsxRuntime.jsx)("span", {
            css: this.props.styles?.igniteAI,
            children: "IgniteAI"
          })]
        }), children]
      });
    }
    if (aiVariant === 'horizontal') {
      return (0, _jsxRuntime.jsxs)("span", {
        css: this.props.styles?.withIcon,
        "aria-hidden": "true",
        children: [(0, _jsxRuntime.jsxs)("span", {
          css: this.props.styles?.igniteAIHorizontal,
          children: [(0, _jsxRuntime.jsx)(_uiIcons.IgniteaiLogoInstUIIcon, {
            color: "ai",
            size: this.getIconSize(true)
          }), (0, _jsxRuntime.jsx)("span", {
            css: this.props.styles?.igniteAI,
            children: "IgniteAI"
          })]
        }), children]
      });
    }
    if (aiVariant === 'iconOnly') {
      return (0, _jsxRuntime.jsxs)("span", {
        css: this.props.styles?.withIcon,
        "aria-hidden": "true",
        children: [(0, _jsxRuntime.jsx)(_uiIcons.IgniteaiLogoInstUIIcon, {
          color: "ai",
          size: this.getIconSize(true)
        }), children]
      });
    }
    return children;
  }

  //overriding default screen reader functionality is needed to read spans in h tags correctly
  getAriaLabel() {
    const {
      aiVariant,
      children,
      renderIcon
    } = this.props;
    if (aiVariant === 'stacked' || aiVariant === 'horizontal') {
      return `IgniteAI, ${children}`;
    }
    if (aiVariant === 'iconOnly' || renderIcon) {
      return `${children}`;
    }
    return undefined;
  }
  render() {
    const {
      border,
      children,
      color,
      level,
      margin,
      elementRef,
      makeStyles,
      variant,
      ...props
    } = this.props;
    let ElementType = 'h2';
    if (variant) {
      // TODO deprecated, remove. `variant` should not set DOM level
      if (level) {
        if (level === 'reset') {
          ElementType = 'span';
        } else {
          ElementType = level;
        }
      } else {
        ElementType = variantLevels[variant];
      }
    } else if (props.as) {
      ElementType = props.as;
    } else if (level) {
      if (level === 'reset') {
        ElementType = 'span';
      } else {
        ElementType = level;
      }
    }
    return (0, _jsxRuntime.jsx)(_latest.View, {
      "aria-label": this.getAriaLabel(),
      ...(0, _passthroughProps.passthroughProps)(props),
      css: this.props.styles?.heading,
      as: ElementType,
      elementRef: this.handleRef,
      margin: margin,
      "data-cid": "Heading",
      children: this.renderContent()
    });
  }
}) || _class);
var _default = exports.default = Heading;
