// @instructure/ui-form-field/lib/FormFieldGroup/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.FormFieldGroup = void 0;
require("core-js/modules/es.array.includes.js");
var _react = require("react");
var _v11_ = require("@instructure/ui-grid/v11_6");
var _pickProps = require("@instructure/ui-react-utils/lib/pickProps.js");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _emotion = require("@instructure/emotion");
var _props = require("../../FormFieldLayout/v1/props");
var _v = require("../../FormFieldLayout/v1");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props2 = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let FormFieldGroup = exports.FormFieldGroup = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class FormFieldGroup extends _react.Component {
  static displayName = "FormFieldGroup";
  static componentId = 'FormFieldGroup';
  static allowedProps = _props2.allowedProps;
  static defaultProps = {
    as: 'fieldset',
    disabled: false,
    rowSpacing: 'medium',
    colSpacing: 'small',
    vAlign: 'middle',
    isGroup: true
  };
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.(this.makeStylesVariables);
  }
  componentDidUpdate() {
    this.props.makeStyles?.(this.makeStylesVariables);
  }
  get makeStylesVariables() {
    // new form errors dont need borders
    const oldInvalid = !!this.props.messages && this.props.messages.findIndex(message => {
      return message.type === 'error';
    }) >= 0;
    return {
      invalid: oldInvalid
    };
  }
  get invalid() {
    return !!this.props.messages && this.props.messages.findIndex(message => {
      return message.type === 'error' || message.type === 'newError';
    }) >= 0;
  }
  renderColumns() {
    return _react.Children.map(this.props.children, (child, index) => {
      return child ? (0, _jsxRuntime.jsx)(_v11_.Grid.Col, {
        width: child.props && child.props.width ? 'auto' : undefined,
        children: child
      }, index) : null;
    });
  }
  renderChildren() {
    return (0, _jsxRuntime.jsx)(_v11_.Grid, {
      colSpacing: this.props.colSpacing,
      rowSpacing: this.props.rowSpacing,
      vAlign: this.props.vAlign,
      startAt: this.props.startAt || (this.props.layout === 'columns' ? 'medium' : null),
      children: (0, _jsxRuntime.jsx)(_v11_.Grid.Row, {
        children: this.renderColumns()
      })
    });
  }
  renderFields() {
    const {
      styles
    } = this.props;
    return (0, _jsxRuntime.jsx)("span", {
      css: styles?.formFieldGroup,
      children: this.renderChildren()
    }, "fields");
  }
  render() {
    const {
      styles,
      makeStyles,
      isGroup,
      ...props
    } = this.props;
    // This is quite ugly, but according to ARIA spec the `aria-invalid` prop
    // can only be used with certain roles see
    // https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-invalid#associated_roles
    // `aria-invalid` is put on in FormFieldLayout because the error message
    // DOM part gets there its ID.
    let ariaInvalid = undefined;
    if (this.props.role && this.invalid && ['application', 'checkbox', 'combobox', 'gridcell', 'listbox', 'radiogroup', 'slider', 'spinbutton', 'textbox', 'tree', 'columnheader', 'rowheader', 'searchbox', 'switch', 'treegrid'].includes(this.props.role)) {
      ariaInvalid = 'true';
    }
    return (0, _jsxRuntime.jsx)(_v.FormFieldLayout, {
      ...(0, _omitProps.omitProps)(props, FormFieldGroup.allowedProps),
      ...(0, _pickProps.pickProps)(props, _props.allowedProps),
      vAlign: props.vAlign,
      layout: props.layout === 'inline' ? 'inline' : 'stacked',
      label: props.description,
      "aria-disabled": props.disabled ? 'true' : undefined,
      "aria-invalid": ariaInvalid,
      elementRef: this.handleRef,
      isGroup: isGroup,
      children: this.renderFields()
    });
  }
}) || _class);
var _default = exports.default = FormFieldGroup;
