// @instructure/ui-pagination/lib/Pagination/v1/PaginationPageInput/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.PaginationPageInput = void 0;
var _react = require("react");
var _emotion = require("@instructure/emotion");
var _v11_ = require("@instructure/ui-number-input/v11_6");
var _ScreenReaderContent = require("@instructure/ui-a11y-content/lib/ScreenReaderContent");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: Pagination
id: Pagination.PageInput
---
**/
let PaginationPageInput = exports.PaginationPageInput = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class PaginationPageInput extends _react.Component {
  static displayName = "PaginationPageInput";
  static componentId = 'Pagination.PageInput';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    disabled: false
  };
  ref = null;
  constructor(props) {
    super(props);
    this.state = this.initialState;
  }
  get currentPage() {
    return this.props.currentPageIndex + 1;
  }
  get initialState() {
    return {
      number: this.currentPage,
      value: `${this.currentPage}`
    };
  }
  get MIN() {
    return 1;
  }
  get MAX() {
    return this.props.numberOfPages;
  }
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate(prevProps) {
    this.props.makeStyles?.();

    // when the current page changes from outside (e.g.: from arrow buttons),
    // we update the initial state
    if (this.props.currentPageIndex !== prevProps.currentPageIndex) {
      this.resetInitialState();
    }
  }
  resetInitialState() {
    this.setState(this.initialState);
  }
  handleRef = el => {
    this.ref = el;
  };
  handleInputRef = el => {
    if (typeof this.props.inputRef === 'function') {
      this.props.inputRef(el);
    }
  };
  handleChange = (_event, value) => {
    let newValue = value;

    // if the last character was not a number, don't accept it
    if (value.length > 0 && isNaN(value.slice(-1))) {
      newValue = value.slice(0, -1);
    }
    this.setState({
      number: newValue ? Number(newValue) : this.MIN,
      value: newValue
    });
  };
  handleKeyDown(event) {
    const {
      key
    } = event;
    if (key.toLowerCase() === 'enter') {
      const {
        value,
        number
      } = this.state;

      // If it is invalid, we reset the input to the current page
      if (value === '' || isNaN(value)) {
        this.resetInitialState();
        return;
      }
      this.setNewPage(event, number);
    }
  }
  handleDecrement = event => {
    this.setNewPage(event, Math.floor(this.currentPage - 1));
  };
  handleIncrement = event => {
    this.setNewPage(event, Math.floor(this.currentPage + 1));
  };
  handleBlur = event => {
    if (Math.round(this.state.number) !== this.currentPage) {
      this.setNewPage(event, Math.round(this.state.number));
    }
  };
  getNumberWithinRange(n) {
    if (n < this.MIN) return this.MIN;
    if (n > this.MAX) return this.MAX;
    return n;
  }
  setNewPage(event, n) {
    const number = this.getNumberWithinRange(n);
    this.setState({
      number,
      value: `${number}`
    });
    this.props.onChange(event, number - 1);
  }
  renderLabel() {
    const {
      label,
      numberOfPages,
      styles
    } = this.props;
    return typeof label === 'function' && label(numberOfPages) ? (0, _jsxRuntime.jsx)("span", {
      css: styles?.inputLabel,
      children: label(numberOfPages)
    }) : null;
  }
  renderScreenReaderLabel() {
    const {
      screenReaderLabel,
      numberOfPages
    } = this.props;
    return (0, _jsxRuntime.jsx)(_ScreenReaderContent.ScreenReaderContent, {
      children: screenReaderLabel(this.currentPage, numberOfPages)
    });
  }
  render() {
    const {
      styles,
      disabled
    } = this.props;
    return (0, _jsxRuntime.jsxs)("span", {
      css: styles?.paginationPageInput,
      ref: this.handleRef,
      "data-cid": "PaginationPageInput",
      children: [(0, _jsxRuntime.jsx)("span", {
        css: styles?.numberInput,
        children: (0, _jsxRuntime.jsx)(_v11_.NumberInput, {
          renderLabel: this.renderScreenReaderLabel(),
          onBlur: this.handleBlur,
          onChange: this.handleChange,
          onDecrement: this.handleDecrement,
          onIncrement: this.handleIncrement,
          onKeyDown: this.handleKeyDown.bind(this),
          interaction: disabled ? 'disabled' : 'enabled',
          showArrows: false,
          value: this.state.value,
          width: "100%",
          textAlign: "center",
          inputRef: this.handleInputRef
        })
      }), this.renderLabel()]
    });
  }
}) || _class);
var _default = exports.default = PaginationPageInput;
