// @instructure/ui-drawer-layout/lib/DrawerLayout/v1/DrawerTray/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.DrawerTray = void 0;
var _react = require("react");
var _emotion = require("@instructure/emotion");
var _textDirectionContextConsumer = require("@instructure/ui-i18n/lib/textDirectionContextConsumer.js");
var _Transition = require("@instructure/ui-motion/lib/Transition");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _createChainedFunction = require("@instructure/ui-utils/lib/createChainedFunction.js");
var _Dialog = require("@instructure/ui-dialog/lib/Dialog");
var _Portal = require("@instructure/ui-portal/lib/Portal");
var _mirrorHorizontalPlacement = require("@instructure/ui-position/lib/mirrorHorizontalPlacement.js");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _index = require("../index");
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: DrawerLayout
id: DrawerLayout.Tray
---
**/
let DrawerTray = exports.DrawerTray = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec2 = (0, _textDirectionContextConsumer.textDirectionContextConsumer)(), _dec(_class = _dec2(_class = class DrawerTray extends _react.Component {
  static displayName = "DrawerTray";
  static componentId = 'DrawerLayout.Tray';
  static locatorAttribute = 'data-drawer-tray';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    shouldContainFocus: true,
    shouldCloseOnEscape: true,
    shouldCloseOnDocumentClick: true,
    shouldReturnFocus: true,
    open: false,
    shadow: true,
    border: true,
    placement: 'start'
  };
  ref = null;
  _DOMNode = null;
  get _content() {
    return this.ref;
  }
  constructor(props) {
    super(props);
    this.state = {
      transitioning: false,
      portalOpen: false
    };
  }
  componentDidMount() {
    this.props.makeStyles?.(this.makeStyleProps());
  }
  componentDidUpdate(prevProps) {
    if (this.props.open !== prevProps.open) {
      this.setState({
        transitioning: true
      });
    }
    this.props.makeStyles?.(this.makeStyleProps());
  }
  makeStyleProps = () => {
    return {
      placement: this.placement
    };
  };
  get placement() {
    const {
      placement,
      dir
    } = this.props;
    const isRtl = dir === _textDirectionContextConsumer.textDirectionContextConsumer.DIRECTION.rtl;
    return isRtl ? (0, _mirrorHorizontalPlacement.mirrorHorizontalPlacement)(placement, ' ') : placement;
  }
  get transition() {
    return this.placement === 'end' ? 'fade' : 'slide-left';
  }
  handleContentRef = node => {
    this.ref = node;
    if (typeof this.props.contentRef === 'function') {
      this.props.contentRef(node);
    }
  };
  handleTransitionEntered = _transitionType => {
    this.setState({
      transitioning: false
    });
  };
  handleTransitionExited = _transitionType => {
    this.setState({
      transitioning: false
    });
  };
  handlePortalOpen = DOMNode => {
    this.DOMNode = DOMNode;
    this.setState({
      portalOpen: true
    });
  };
  get DOMNode() {
    return this._DOMNode;
  }
  set DOMNode(el) {
    this._DOMNode = el;
  }
  renderContent() {
    const {
      children,
      render
    } = this.props;
    if (typeof render === 'function') {
      return render();
    } else if (typeof children === 'function') {
      return children();
    } else {
      return children;
    }
  }
  render() {
    const {
      label,
      children,
      render,
      placement,
      open,
      onOpen,
      onClose,
      border,
      shadow,
      onEnter,
      onEntering,
      onEntered,
      onExit,
      onExiting,
      onExited,
      contentRef,
      mountNode,
      defaultFocusElement,
      liveRegion,
      onDismiss,
      onTransition,
      shouldReturnFocus,
      shouldCloseOnEscape,
      shouldCloseOnDocumentClick,
      shouldContainFocus,
      styles,
      ...props
    } = this.props;
    return (0, _jsxRuntime.jsx)(_index.DrawerLayoutContext.Consumer, {
      children: shouldOverlayTray => {
        const {
          portalOpen
        } = this.state;
        const needsPortal = shouldOverlayTray && !!mountNode;
        // we have to handle the shadow version here,
        // because passing and handling it in styles.js makes it
        // rerender the component during the transition, breaking it
        const trayStyles = shadow && shouldOverlayTray ? styles?.drawerTrayWithShadow : styles?.drawerTray;
        let transitionIn = open;
        if (needsPortal && !portalOpen) {
          transitionIn = false;
        }
        const content = (0, _jsxRuntime.jsx)(_Transition.Transition, {
          in: transitionIn,
          type: this.transition,
          onTransition: onTransition,
          onEnter: onEnter,
          onEntering: onEntering,
          onEntered: (0, _createChainedFunction.createChainedFunction)(this.handleTransitionEntered, onEntered, onOpen),
          onExit: onExit,
          onExiting: onExiting,
          onExited: (0, _createChainedFunction.createChainedFunction)(this.handleTransitionExited, onExited, onClose),
          unmountOnExit: true,
          children: (0, _jsxRuntime.jsx)("div", {
            ...(0, _omitProps.omitProps)(props, DrawerTray.allowedProps),
            ref: this.handleContentRef,
            css: trayStyles,
            "data-cid": "DrawerTray",
            children: (0, _jsxRuntime.jsx)(_Dialog.Dialog, {
              open: true,
              role: shouldOverlayTray ? 'dialog' : 'region',
              label: label,
              shouldReturnFocus: shouldReturnFocus,
              shouldContainFocus: !this.state.transitioning && shouldContainFocus && shouldOverlayTray,
              shouldCloseOnDocumentClick: shouldCloseOnDocumentClick && shouldOverlayTray,
              shouldCloseOnEscape: shouldCloseOnEscape && shouldOverlayTray,
              defaultFocusElement: defaultFocusElement,
              liveRegion: liveRegion,
              onDismiss: onDismiss,
              as: "div",
              css: this.props.styles?.drawerTrayContent,
              children: this.renderContent()
            })
          })
        });
        if (needsPortal) {
          return (0, _jsxRuntime.jsx)(_Portal.Portal, {
            mountNode: mountNode,
            open: true,
            onOpen: this.handlePortalOpen,
            children: content
          });
        } else {
          return content;
        }
      }
    });
  }
}) || _class) || _class);
var _default = exports.default = DrawerTray;
