// @instructure/ui-drawer-layout/es/DrawerLayout/v2/DrawerTray/index.js
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { withStyleNew } from '@instructure/emotion';
import { textDirectionContextConsumer } from '@instructure/ui-i18n';
import { Transition } from '@instructure/ui-motion';
import { omitProps } from '@instructure/ui-react-utils';
import { createChainedFunction } from '@instructure/ui-utils';
import { Dialog } from '@instructure/ui-dialog';
import { Portal } from '@instructure/ui-portal';
import { mirrorHorizontalPlacement } from '@instructure/ui-position';
import generateStyle from "./styles.js";
import { DrawerLayoutContext } from "../index.js";
import { allowedProps } from "./props.js";
import { jsx as _jsx } from "@emotion/react/jsx-runtime";
/**
---
parent: DrawerLayout
id: DrawerLayout.Tray
---
**/
let DrawerTray = (_dec = withStyleNew(generateStyle), _dec2 = textDirectionContextConsumer(), _dec(_class = _dec2(_class = class DrawerTray extends Component {
  static displayName = "DrawerTray";
  static componentId = 'DrawerLayout.Tray';
  static locatorAttribute = 'data-drawer-tray';
  static allowedProps = allowedProps;
  static defaultProps = {
    shouldContainFocus: true,
    shouldCloseOnEscape: true,
    shouldCloseOnDocumentClick: true,
    shouldReturnFocus: true,
    open: false,
    shadow: true,
    border: true,
    placement: 'start'
  };
  ref = null;
  _DOMNode = null;
  get _content() {
    return this.ref;
  }
  constructor(props) {
    super(props);
    this.state = {
      transitioning: false,
      portalOpen: false
    };
  }
  componentDidMount() {
    this.props.makeStyles?.(this.makeStyleProps());
  }
  componentDidUpdate(prevProps) {
    if (this.props.open !== prevProps.open) {
      this.setState({
        transitioning: true
      });
    }
    this.props.makeStyles?.(this.makeStyleProps());
  }
  makeStyleProps = () => {
    return {
      placement: this.placement
    };
  };
  get placement() {
    const {
      placement,
      dir
    } = this.props;
    const isRtl = dir === textDirectionContextConsumer.DIRECTION.rtl;
    return isRtl ? mirrorHorizontalPlacement(placement, ' ') : placement;
  }
  get transition() {
    return this.placement === 'end' ? 'fade' : 'slide-left';
  }
  handleContentRef = node => {
    this.ref = node;
    if (typeof this.props.contentRef === 'function') {
      this.props.contentRef(node);
    }
  };
  handleTransitionEntered = _transitionType => {
    this.setState({
      transitioning: false
    });
  };
  handleTransitionExited = _transitionType => {
    this.setState({
      transitioning: false
    });
  };
  handlePortalOpen = DOMNode => {
    this.DOMNode = DOMNode;
    this.setState({
      portalOpen: true
    });
  };
  get DOMNode() {
    return this._DOMNode;
  }
  set DOMNode(el) {
    this._DOMNode = el;
  }
  renderContent() {
    const {
      children,
      render
    } = this.props;
    if (typeof render === 'function') {
      return render();
    } else if (typeof children === 'function') {
      return children();
    } else {
      return children;
    }
  }
  render() {
    const {
      label,
      children,
      render,
      placement,
      open,
      onOpen,
      onClose,
      border,
      shadow,
      onEnter,
      onEntering,
      onEntered,
      onExit,
      onExiting,
      onExited,
      contentRef,
      mountNode,
      defaultFocusElement,
      liveRegion,
      onDismiss,
      onTransition,
      shouldReturnFocus,
      shouldCloseOnEscape,
      shouldCloseOnDocumentClick,
      shouldContainFocus,
      styles,
      ...props
    } = this.props;
    return _jsx(DrawerLayoutContext.Consumer, {
      children: shouldOverlayTray => {
        const {
          portalOpen
        } = this.state;
        const needsPortal = shouldOverlayTray && !!mountNode;
        // we have to handle the shadow version here,
        // because passing and handling it in styles.js makes it
        // rerender the component during the transition, breaking it
        const trayStyles = shadow && shouldOverlayTray ? styles?.drawerTrayWithShadow : styles?.drawerTray;
        let transitionIn = open;
        if (needsPortal && !portalOpen) {
          transitionIn = false;
        }
        const content = _jsx(Transition, {
          in: transitionIn,
          type: this.transition,
          onTransition: onTransition,
          onEnter: onEnter,
          onEntering: onEntering,
          onEntered: createChainedFunction(this.handleTransitionEntered, onEntered, onOpen),
          onExit: onExit,
          onExiting: onExiting,
          onExited: createChainedFunction(this.handleTransitionExited, onExited, onClose),
          unmountOnExit: true,
          children: _jsx("div", {
            ...omitProps(props, DrawerTray.allowedProps),
            ref: this.handleContentRef,
            css: trayStyles,
            "data-cid": "DrawerTray",
            children: _jsx(Dialog, {
              open: open,
              role: shouldOverlayTray ? 'dialog' : 'region',
              label: label,
              shouldReturnFocus: shouldReturnFocus,
              shouldContainFocus: !this.state.transitioning && shouldContainFocus && shouldOverlayTray,
              shouldCloseOnDocumentClick: shouldCloseOnDocumentClick && shouldOverlayTray,
              shouldCloseOnEscape: shouldCloseOnEscape && shouldOverlayTray,
              defaultFocusElement: defaultFocusElement,
              liveRegion: liveRegion,
              onDismiss: onDismiss,
              as: "div",
              css: this.props.styles?.drawerTrayContent,
              children: this.renderContent()
            })
          })
        });
        if (needsPortal) {
          return _jsx(Portal, {
            mountNode: mountNode,
            open: true,
            onOpen: this.handlePortalOpen,
            children: content
          });
        } else {
          return content;
        }
      }
    });
  }
}) || _class) || _class);
export default DrawerTray;
export { DrawerTray };
