// @instructure/ui-number-input/lib/NumberInput/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.NumberInput = void 0;
var _react = require("react");
var _keycode = _interopRequireDefault(require("keycode"));
var _latest = require("@instructure/ui-form-field/latest");
var _uiIcons = require("@instructure/ui-icons");
var _pickProps = require("@instructure/ui-react-utils/lib/pickProps.js");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _getInteraction = require("@instructure/ui-react-utils/lib/getInteraction.js");
var _useDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/useDeterministicId.js");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
category: components
id: NumberInput
---
**/
const NumberInput = exports.NumberInput = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    messages = [],
    isRequired = false,
    showArrows = true,
    size = 'medium',
    display = 'block',
    textAlign = 'start',
    inputMode = 'numeric',
    allowStringValue = false,
    renderLabel,
    placeholder,
    value,
    width,
    renderIcons,
    margin,
    inputRef: inputRefProp,
    onFocus,
    onBlur,
    onChange,
    onKeyDown,
    onDecrement,
    onIncrement,
    id: idProp,
    themeOverride,
    ...rest
  } = props;
  // these are icon tokens

  const [upButtonState, setUpButtonState] = (0, _react.useState)('actionSecondaryBaseColor');
  const [downButtonState, setDownButtonState] = (0, _react.useState)('actionSecondaryBaseColor');
  // Refs
  const containerRef = (0, _react.useRef)(null);
  const inputRef = (0, _react.useRef)(null);

  // Deterministic ID generation
  const [deterministicId, setDeterministicId] = (0, _react.useState)();
  const getId = (0, _useDeterministicId.useDeterministicId)('NumberInput');
  (0, _react.useEffect)(() => {
    setDeterministicId(getId());
  }, []); // Empty deps array - only run once on mount
  const id = idProp || deterministicId;

  // Computed values
  const invalid = !!messages && messages.some(message => message.type === 'error' || message.type === 'newError');
  const success = !!messages && messages.some(message => message.type === 'success');
  const interaction = (0, _getInteraction.getInteraction)({
    props
  });
  if (interaction === 'disabled' && upButtonState !== 'actionSecondaryDisabledColor') {
    setUpButtonState('actionSecondaryDisabledColor');
    setDownButtonState('actionSecondaryDisabledColor');
  } else if (interaction === 'enabled' && downButtonState !== 'actionSecondaryBaseColor') {
    setUpButtonState('actionSecondaryBaseColor');
    setDownButtonState('actionSecondaryBaseColor');
  }
  // Styles - useStyleNew will pass these to generateStyle(componentTheme, params as props, params as state)
  // We need to provide all values that generateStyle needs from both props and state
  const styles = (0, _emotion.useStyleNew)({
    generateStyle: _styles.default,
    themeOverride,
    params: {
      size,
      textAlign,
      interaction,
      invalid,
      success
    },
    componentId: 'NumberInput',
    displayName: 'NumberInput',
    useTokensFrom: 'TextInput'
  });

  // Event handlers
  const handleInputRef = (0, _react.useCallback)(element => {
    inputRef.current = element;
    if (typeof inputRefProp === 'function') {
      inputRefProp(element);
    }
  }, [inputRefProp]);
  const handleRef = (0, _react.useCallback)(el => {
    containerRef.current = el;
  }, []);
  const handleFocus = (0, _react.useCallback)(event => {
    if (typeof onFocus === 'function') {
      onFocus(event);
    }
  }, [onFocus]);
  const handleBlur = (0, _react.useCallback)(event => {
    if (typeof onBlur === 'function') {
      onBlur(event);
    }
  }, [onBlur]);
  const handleChange = (0, _react.useCallback)(event => {
    if (typeof onChange === 'function') {
      onChange(event, event.target.value);
    }
  }, [onChange]);
  const handleKeyDown = (0, _react.useCallback)(event => {
    if (typeof onKeyDown === 'function') {
      onKeyDown(event);
    }
    if (event.keyCode === _keycode.default.codes.down) {
      event.preventDefault();
      if (typeof onDecrement === 'function') {
        onDecrement(event);
      }
    } else if (event.keyCode === _keycode.default.codes.up) {
      event.preventDefault();
      if (typeof onIncrement === 'function') {
        onIncrement(event);
      }
    }
  }, [onKeyDown, onDecrement, onIncrement]);
  const arrowClicked = (0, _react.useCallback)((event, callback) => {
    event.preventDefault();
    if (interaction === 'enabled') {
      inputRef.current?.focus();
      if (typeof callback === 'function') {
        callback(event);
      }
    }
  }, [interaction]);
  const handleClickUpArrow = (0, _react.useCallback)(event => {
    setUpButtonState('actionSecondaryActiveColor');
    arrowClicked(event, onIncrement);
  }, [arrowClicked, onIncrement]);
  const handleClickDownArrow = (0, _react.useCallback)(event => {
    setDownButtonState('actionSecondaryActiveColor');
    arrowClicked(event, onDecrement);
  }, [arrowClicked, onDecrement]);

  // Expose imperative API via ref
  (0, _react.useImperativeHandle)(ref, () => ({
    focus: () => {
      inputRef.current?.focus();
    },
    get id() {
      return id;
    },
    get invalid() {
      return invalid;
    },
    get interaction() {
      return interaction;
    },
    get value() {
      return inputRef.current?.value;
    },
    ref: containerRef
  }), [id, invalid, interaction]);

  // Render methods
  const renderArrows = customIcons => {
    return (0, _jsxRuntime.jsxs)("span", {
      css: styles?.arrowContainer,
      children: [(0, _jsxRuntime.jsx)("button", {
        "aria-hidden": true,
        css: styles?.arrow,
        onMouseDown: handleClickUpArrow,
        onMouseOver: () => setUpButtonState('actionSecondaryHoverColor'),
        onMouseOut: () => setUpButtonState('actionSecondaryBaseColor'),
        tabIndex: -1,
        type: "button",
        children: customIcons?.increase ? (0, _callRenderProp.callRenderProp)(customIcons.increase) : (0, _jsxRuntime.jsx)(_uiIcons.ChevronUpInstUIIcon, {
          size: "sm",
          color: upButtonState
        })
      }), (0, _jsxRuntime.jsx)("button", {
        "aria-hidden": true,
        css: styles?.arrow,
        onMouseDown: handleClickDownArrow,
        onMouseOver: () => setDownButtonState('actionSecondaryHoverColor'),
        onMouseOut: () => setDownButtonState('actionSecondaryBaseColor'),
        tabIndex: -1,
        type: "button",
        children: customIcons?.decrease ? (0, _callRenderProp.callRenderProp)(customIcons.decrease) : (0, _jsxRuntime.jsx)(_uiIcons.ChevronDownInstUIIcon, {
          size: "sm",
          color: downButtonState
        })
      })]
    });
  };
  const label = (0, _callRenderProp.callRenderProp)(renderLabel);
  const passedProps = (0, _passthroughProps.passthroughProps)(rest);

  // Don't render until we have an ID
  if (!id) {
    return null;
  }
  return (0, _jsxRuntime.jsx)(_latest.FormField, {
    ...(0, _pickProps.pickProps)(props, _latest.FormField.allowedProps),
    label: label,
    inline: display === 'inline-block',
    id: id,
    elementRef: handleRef,
    margin: margin,
    isRequired: isRequired,
    disabled: interaction === 'disabled',
    readOnly: interaction === 'readonly',
    "data-cid": "NumberInput",
    children: (0, _jsxRuntime.jsx)("span", {
      css: styles?.inputWidth,
      style: width ? {
        width
      } : undefined,
      children: (0, _jsxRuntime.jsxs)("span", {
        css: styles?.inputContainer,
        children: [(0, _jsxRuntime.jsx)("input", {
          ...passedProps,
          css: styles?.input,
          "aria-invalid": invalid ? 'true' : undefined,
          id: id,
          type: allowStringValue ? 'text' : 'number',
          inputMode: inputMode,
          placeholder: interaction === 'enabled' ? placeholder : undefined,
          ref: handleInputRef,
          required: isRequired,
          value: value,
          disabled: interaction === 'disabled',
          readOnly: interaction === 'readonly',
          onFocus: handleFocus,
          onBlur: handleBlur,
          onChange: handleChange,
          onKeyDown: handleKeyDown
        }), showArrows && interaction !== 'readonly' ? renderArrows(renderIcons) : null]
      })
    })
  });
});
NumberInput.displayName = 'NumberInput';
var _default = exports.default = NumberInput;
