// @instructure/ui-pill/es/Pill/v2/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { View } from '@instructure/ui-view/latest';
import { passthroughProps } from '@instructure/ui-react-utils';
import { Tooltip } from '@instructure/ui-tooltip/latest';
import { withStyleNew } from '@instructure/emotion';
import generateStyle from "./styles.js";
import { allowedProps } from "./props.js";
/**
---
category: components
---
**/
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
let Pill = (_dec = withStyleNew(generateStyle), _dec(_class = class Pill extends Component {
  static displayName = "Pill";
  static componentId = 'Pill';
  static allowedProps = allowedProps;
  static defaultProps = {
    color: 'primary'
  };
  ref = null;
  ellipsisRef = null;
  constructor(props) {
    super(props);
    this.state = {
      truncated: false
    };
  }
  componentDidMount() {
    this.setTruncation();
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  setTruncation() {
    if (this.ellipsisRef) {
      this.setState({
        truncated: this.ellipsisRef.offsetWidth < this.ellipsisRef.scrollWidth
      });
    }
  }
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  renderPill(focused, getTriggerProps) {
    const {
      margin,
      children,
      color,
      as,
      elementRef,
      styles,
      makeStyles,
      statusLabel,
      renderIcon,
      ...props
    } = this.props;
    const filteredProps = passthroughProps(props);
    const containerProps = typeof getTriggerProps === 'function' ? getTriggerProps(filteredProps) : filteredProps;
    const refProp = this.state.truncated ? {} : {
      elementRef: this.handleRef
    };
    return _jsx(View, {
      ...containerProps,
      as: as,
      ...refProp,
      margin: margin,
      padding: "0",
      maxWidth: styles?.maxWidth,
      background: "transparent",
      borderRadius: "pill",
      borderWidth: "0",
      display: "inline-block",
      position: "relative",
      withFocusOutline: focused,
      focusColor: "info",
      "data-cid": "Pill",
      children: _jsxs("div", {
        css: styles?.pill,
        children: [renderIcon && _jsx("div", {
          css: styles?.icon,
          children: renderIcon
        }), _jsxs("div", {
          css: styles?.text,
          ref: el => {
            this.ellipsisRef = el;
          },
          children: [statusLabel && _jsx("span", {
            css: styles?.status,
            children: statusLabel && statusLabel.concat(':')
          }), children]
        })]
      })
    });
  }
  render() {
    if (this.state.truncated) {
      return _jsx(Tooltip, {
        renderTip: this.props.statusLabel ? this.props.statusLabel.concat(': ', this.props.children) : this.props.children,
        elementRef: this.handleRef,
        children: ({
          focused,
          getTriggerProps
        }) => {
          return this.renderPill(focused, getTriggerProps);
        }
      });
    } else {
      return this.renderPill();
    }
  }
}) || _class);
export default Pill;
export { Pill };
