// @instructure/ui-color-picker/lib/ColorMixer/v1/RGBAInput/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _emotion = require("@instructure/emotion");
var _shallowCompare = _interopRequireDefault(require("../utils/shallowCompare"));
var _ScreenReaderContent = require("@instructure/ui-a11y-content/lib/ScreenReaderContent");
var _v11_ = require("@instructure/ui-text-input/v11_6");
var _props = require("./props");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
private: true
---
**/
let RGBAInput = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class RGBAInput extends _react.Component {
  static displayName = "RGBAInput";
  static allowedProps = _props.allowedProps;
  static componentId = 'ColorMixer.RGBAInput';
  static defaultProps = {
    withAlpha: false
  };
  constructor(props) {
    super(props);
    this.state = {
      value: props.value
    };
  }
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate(prevProps) {
    this.props.makeStyles?.();
    if ((0, _shallowCompare.default)(prevProps.value, this.props.value) || (0, _shallowCompare.default)(this.state.value, this.props.value)) {
      this.setState({
        value: this.props.value
      });
    }
  }
  handleChange(type, e) {
    const upperLimit = type === 'a' ? 100 : 255;
    const newValue = type === 'a' ? Number(e.target.value) / 100 : Number(e.target.value);
    const newColor = {
      ...this.props.value,
      [type]: newValue
    };
    if (!isNaN(Number(newValue)) && newValue >= 0 && newValue <= upperLimit) {
      this.setState({
        value: newColor
      });
      this.props.onChange(newColor);
      return;
    }
  }
  render() {
    const {
      styles,
      disabled,
      label,
      withAlpha
    } = this.props;
    return (0, _jsxRuntime.jsxs)("div", {
      ref: this.handleRef,
      css: styles?.RGBAInput,
      children: [label && (0, _jsxRuntime.jsx)("div", {
        css: styles?.label,
        children: label
      }), (0, _jsxRuntime.jsxs)("div", {
        css: styles?.inputContainer,
        children: [(0, _jsxRuntime.jsx)("span", {
          css: styles?.rgbInput,
          children: (0, _jsxRuntime.jsx)(_v11_.TextInput, {
            disabled: disabled,
            value: `${this.state.value.r}`,
            onChange: e => this.handleChange('r', e),
            renderLabel: (0, _jsxRuntime.jsx)(_ScreenReaderContent.ScreenReaderContent, {
              children: this.props.rgbRedInputScreenReaderLabel
            })
          })
        }), (0, _jsxRuntime.jsx)("span", {
          css: styles?.rgbInput,
          children: (0, _jsxRuntime.jsx)(_v11_.TextInput, {
            disabled: disabled,
            value: `${this.state.value.g}`,
            onChange: e => this.handleChange('g', e),
            renderLabel: (0, _jsxRuntime.jsx)(_ScreenReaderContent.ScreenReaderContent, {
              children: this.props.rgbGreenInputScreenReaderLabel
            })
          })
        }), (0, _jsxRuntime.jsx)("span", {
          css: styles?.rgbInput,
          children: (0, _jsxRuntime.jsx)(_v11_.TextInput, {
            disabled: disabled,
            value: `${this.state.value.b}`,
            onChange: e => this.handleChange('b', e),
            renderLabel: (0, _jsxRuntime.jsx)(_ScreenReaderContent.ScreenReaderContent, {
              children: this.props.rgbBlueInputScreenReaderLabel
            })
          })
        }), withAlpha && (0, _jsxRuntime.jsx)("span", {
          css: styles?.aInput,
          children: (0, _jsxRuntime.jsx)(_v11_.TextInput, {
            "aria-label": "%",
            disabled: disabled,
            value: `${Math.round(this.state.value.a * 100)}`,
            onChange: e => this.handleChange('a', e),
            renderAfterInput: (0, _jsxRuntime.jsx)("span", {
              "aria-hidden": "true",
              role: "presentation",
              children: "%"
            }),
            renderLabel: (0, _jsxRuntime.jsx)(_ScreenReaderContent.ScreenReaderContent, {
              children: this.props.rgbAlphaInputScreenReaderLabel
            })
          })
        })]
      })]
    });
  }
}) || _class);
var _default = exports.default = RGBAInput;
