// @instructure/ui-tree-browser/lib/TreeBrowser/v2/TreeNode/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TreeNode = void 0;
var _react = require("react");
var _latest = require("@instructure/ui-img/latest");
var _emotion = require("@instructure/emotion");
var _renderIconWithProps = require("@instructure/ui-icons/lib/IconPropsProvider/renderIconWithProps.js");
var _styles = _interopRequireDefault(require("../TreeButton/styles"));
var _props = require("./props");
var _TreeBrowserContext = _interopRequireDefault(require("../TreeBrowserContext"));
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2021 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
// Map TreeBrowser size tokens to icon size tokens
const treeBrowserSizeToIconSize = {
  small: 'md',
  medium: 'lg',
  large: 'xl'
};

// Todo: merge TreeButton and TreeNode: TreeButton should be a special type of TreeNode

/**
---
parent: TreeBrowser
id: TreeBrowser.Node
---
A helper class used to render the :renderBeforeItems and :renderAfterItems
in the TreeBrowser.
**/
let TreeNode = exports.TreeNode = (_dec = (0, _emotion.withStyleNew)(_styles.default, 'TreeBrowserTreeButton'), _dec(_class = class TreeNode extends _react.Component {
  static displayName = "TreeNode";
  static componentId = 'TreeBrowser.Node';
  static allowedProps = _props.allowedProps;
  static contextType = _TreeBrowserContext.default;
  static defaultProps = {
    size: 'medium',
    variant: 'folderTree',
    selected: false,
    focused: false,
    hoverable: true
  };
  ref = null;
  state = {
    isHovered: false
  };
  componentDidMount() {
    this.props.makeStyles?.({
      animation: this.context?.animation,
      hoverable: this.props.hoverable
    });
  }
  componentDidUpdate() {
    this.props.makeStyles?.({
      animation: this.context?.animation,
      hoverable: this.props.hoverable
    });
  }
  handleRef = el => {
    if (el && typeof this.props.containerRef === 'function') {
      this.props.containerRef(el.parentElement);
    }
    this.ref = el;
  };
  handleMouseEnter = () => {
    if (this.props.hoverable) {
      this.setState({
        isHovered: true
      });
    }
  };
  handleMouseLeave = () => {
    if (this.props.hoverable) {
      this.setState({
        isHovered: false
      });
    }
  };
  getIconColor() {
    const {
      selected
    } = this.props;
    const {
      isHovered
    } = this.state;

    // Priority: selected > hover > default
    if (selected) return 'onColor';
    if (isHovered) return 'inverseColor';
    return 'baseColor';
  }
  renderItemImage() {
    const {
      thumbnail,
      itemIcon,
      styles,
      size
    } = this.props;
    if (thumbnail) {
      return (0, _jsxRuntime.jsx)("div", {
        css: styles?.thumbnail,
        children: (0, _jsxRuntime.jsx)(_latest.Img, {
          src: thumbnail,
          constrain: "cover",
          alt: ""
        })
      });
    }
    if (itemIcon) {
      const iconSize = treeBrowserSizeToIconSize[size];
      const iconColor = this.getIconColor();
      const iconElement = (0, _renderIconWithProps.renderIconWithProps)(itemIcon, iconSize, iconColor);
      return (0, _jsxRuntime.jsx)("div", {
        css: styles?.icon,
        children: iconElement
      });
    }
    return undefined;
  }
  render() {
    const {
      children,
      styles
    } = this.props;
    return (0, _jsxRuntime.jsx)("div", {
      ref: this.handleRef,
      tabIndex: -1,
      css: styles?.treeButton,
      "data-cid": "TreeNode",
      onMouseEnter: this.handleMouseEnter,
      onMouseLeave: this.handleMouseLeave,
      children: (0, _jsxRuntime.jsxs)("span", {
        css: styles?.layout,
        children: [this.renderItemImage(), (0, _jsxRuntime.jsx)("span", {
          css: styles?.node,
          children: children
        })]
      })
    });
  }
}) || _class);
var _default = exports.default = TreeNode;
