// @instructure/ui-toggle-details/es/ToggleGroup/v1/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { omitProps, pickProps, getElementType, callRenderProp } from '@instructure/ui-react-utils';
import { IconButton } from '@instructure/ui-buttons/v11_6';
import { Transition } from '@instructure/ui-motion';
import { Expandable } from '@instructure/ui-expandable';
import { isActiveElement } from '@instructure/ui-dom-utils';
import { Flex } from '@instructure/ui-flex/v11_6';
import { View } from '@instructure/ui-view/v11_6';
import { IconArrowOpenEndSolid, IconArrowOpenDownSolid } from '@instructure/ui-icons';
import { allowedProps } from "./props.js";
import { withStyle } from '@instructure/emotion';
import generateStyle from "./styles.js";
import generateComponentTheme from "./theme.js";
/**
---
category: components
---
**/
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
let ToggleGroup = (_dec = withStyle(generateStyle, generateComponentTheme), _dec(_class = class ToggleGroup extends Component {
  static displayName = "ToggleGroup";
  static componentId = 'ToggleGroup';
  static allowedProps = allowedProps;
  static defaultProps = {
    size: 'medium',
    icon: IconArrowOpenEndSolid,
    iconExpanded: IconArrowOpenDownSolid,
    defaultExpanded: false,
    transition: true,
    as: 'span',
    border: true
  };
  ref = null;
  _button = null;
  _shouldTransition = false;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  handleButtonRef = el => {
    this._button = el;
  };
  get focused() {
    return isActiveElement(this._button);
  }
  focus() {
    this._button?.focus();
  }
  componentDidMount() {
    this._shouldTransition = true;
  }
  componentDidUpdate() {
    this.props.makeStyles?.(this.state);
  }
  renderIcon(expanded) {
    const Icon = expanded ? this.props.iconExpanded : this.props.icon;
    return Icon ? callRenderProp(Icon) : null;
  }
  renderToggle(toggleProps, expanded) {
    const {
      toggleLabel,
      size
    } = this.props;
    let label;
    if (typeof toggleLabel === 'function') {
      label = toggleLabel(expanded);
    } else {
      label = toggleLabel;
    }
    const props = {
      ...toggleProps
    };
    return _jsx(IconButton, {
      ...props,
      withBackground: false,
      withBorder: false,
      size: size === 'large' ? 'medium' : 'small',
      elementRef: this.handleButtonRef,
      screenReaderLabel: label,
      children: this.renderIcon(expanded)
    });
  }
  renderDetails(detailsProps) {
    const {
      styles
    } = this.props;
    return _jsx(View, {
      ...detailsProps,
      display: "block",
      borderWidth: "small none none none",
      borderColor: styles?.borderColor,
      children: this.props.transition && this._shouldTransition ? _jsx(Transition, {
        transitionOnMount: true,
        in: true,
        type: "fade",
        children: this.props.children
      }) : this.props.children
    });
  }
  render() {
    const Element = getElementType(ToggleGroup, this.props);
    const {
      styles
    } = this.props;
    return _jsx(Expandable, {
      ...pickProps(this.props, Expandable.allowedProps),
      children: ({
        expanded,
        getToggleProps,
        getDetailsProps
      }) => {
        return _jsxs(View, {
          ...omitProps(this.props, ToggleGroup.allowedProps),
          borderWidth: this.props.border ? 'small' : 'none',
          as: Element,
          elementRef: this.handleRef,
          display: "block",
          borderRadius: "medium",
          background: "primary",
          borderColor: styles?.borderColor,
          "data-cid": "ToggleGroup",
          children: [_jsxs(Flex, {
            padding: this.props.size === 'small' ? 'x-small' : 'small small small x-small',
            children: [_jsx(Flex.Item, {
              children: this.renderToggle(getToggleProps(), expanded)
            }), _jsx(Flex.Item, {
              shouldGrow: true,
              shouldShrink: true,
              padding: "0 0 0 x-small",
              children: this.props.summary
            })]
          }), expanded ? this.renderDetails(getDetailsProps()) : _jsx("span", {
            ...getDetailsProps()
          })]
        });
      }
    });
  }
}) || _class);
export default ToggleGroup;
export { ToggleGroup };
