// @instructure/ui-themes/es/themes/newThemeTokens/light/components/colorPicker.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const colorPicker = semantics => ({
  checkerboardBackgroundColor: semantics.color.background.elevatedSurface.base,
  colorIndicatorBorderColor: semantics.color.stroke.base,
  colorMixerButtonContainerLeftMargin: "0.5rem",
  errorIconsRightPadding: semantics.spacing.padding.container.xs,
  hashMarkColor: semantics.color.text.muted,
  hashMarkContainerLeftPadding: "0.5rem",
  hashMarkContainerLineHeight: semantics.lineHeight.heading.textLg,
  hashMarkContainerRightPadding: semantics.spacing.space2xs,
  labelRightMargin: "0.375rem",
  popoverContentBlockBorderWidth: semantics.borderWidth.sm,
  popoverFooterTopBorderWidth: semantics.borderWidth.sm,
  popoverContentPadding: semantics.spacing.spaceMd,
  popoverFooterPadding: semantics.spacing.spaceSm,
  popoverSeparatorColor: semantics.color.stroke.muted,
  simpleColorContainerLeftPadding: semantics.spacing.spaceSm,
  successIconRightPadding: semantics.spacing.spaceMd,
  xLargeFontSize: semantics.fontSize.textXl,
  xSmallSpacing: semantics.spacing.spaceSm,
  xxxSmallSpacing: semantics.spacing.space2xs,
  popoverContentBlockBottomMargin: "0.5rem",
  popoverContentBlockTopMargin: "0.5rem",
  popoverContentBlockTopPadding: "0.5rem",
  popoverFooterColor: semantics.color.background.elevatedSurface.base
});
export default colorPicker;
