// @instructure/ui-themes/es/themes/newThemeTokens/light/components/progressBar.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const progressBar = semantics => ({
  borderRadius: semantics.borderRadius.md,
  fontFamily: semantics.fontFamily.base,
  fontSize: semantics.fontSize.textBase,
  fontWeight: semantics.fontWeight.body.base,
  lineHeight: semantics.lineHeight.heading.base,
  xSmallHeight: "0.5rem",
  smallHeight: "1rem",
  mediumHeight: "1.5rem",
  meterColorAlert: semantics.color.background.brand,
  meterColorAlertInverse: semantics.color.background.base,
  meterColorBrand: semantics.color.background.brand,
  meterColorBrandInverse: semantics.color.background.base,
  meterColorDanger: semantics.color.background.brand,
  meterColorDangerInverse: semantics.color.background.base,
  meterColorInfo: semantics.color.background.brand,
  meterColorInfoInverse: semantics.color.background.base,
  meterColorSuccess: semantics.color.background.brand,
  meterColorSuccessInverse: semantics.color.background.base,
  meterColorWarning: semantics.color.background.brand,
  meterColorWarningInverse: semantics.color.background.base,
  largeHeight: "2rem",
  trackBottomBorderColor: semantics.color.transparentColor,
  trackBottomBorderColorInverse: semantics.color.transparentColor,
  trackBottomBorderWidth: semantics.borderWidth.sm,
  trackColor: semantics.color.transparentColor,
  trackColorInverse: semantics.color.transparentColor,
  valuePadding: semantics.spacing.spaceSm,
  textColor: semantics.color.text.muted,
  textColorInverse: semantics.color.text.inverse,
  xSmallValueFontSize: semantics.fontSize.textBase,
  smallValueFontSize: semantics.fontSize.textBase,
  mediumValueFontSize: semantics.fontSize.textBase,
  largeValueFontSize: semantics.fontSize.textBase,
  borderColor: semantics.color.stroke.brand,
  borderColorInverse: semantics.color.stroke.inverse
});
export default progressBar;
