// @instructure/ui-top-nav-bar/lib/TopNavBar/v2/TopNavBarBreadcrumb/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TopNavBarBreadcrumb = void 0;
var _react = require("react");
var _props = require("./props");
var _index = _interopRequireDefault(require("../index"));
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("../TopNavBarBreadcrumb/styles"));
var _uiIcons = require("@instructure/ui-icons");
var _TopNavBarContext = _interopRequireDefault(require("../TopNavBarContext"));
var _console = require("@instructure/console");
var _latest = require("@instructure/ui-link/latest");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: TopNavBar
id: TopNavBar.Breadcrumb
---
@module TopNavBarBreadcrumb
 **/
let TopNavBarBreadcrumb = exports.TopNavBarBreadcrumb = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class TopNavBarBreadcrumb extends _react.Component {
  static displayName = "TopNavBarBreadcrumb";
  static allowedProps = _props.allowedProps;
  static defaultProps = {};
  static contextType = _TopNavBarContext.default;
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.({
      inverseColor: this.context.inverseColor
    });
  }
  componentDidUpdate() {
    this.props.makeStyles?.({
      inverseColor: this.context.inverseColor
    });
  }
  renderMenu() {
    const {
      onClick
    } = this.props;
    return (0, _jsxRuntime.jsx)("div", {
      css: this.props.styles?.icon,
      children: (0, _jsxRuntime.jsx)(_index.default.Item, {
        id: "iconItem",
        variant: "icon",
        renderIcon: (0, _jsxRuntime.jsx)(_uiIcons.MenuInstUIIcon, {}),
        onClick: onClick,
        children: "Hamburger menu"
      })
    });
  }
  render() {
    const {
      children,
      styles
    } = this.props;
    if (!this.context.inverseColor) {
      (0, _console.error)(false, `[TopNavBarBreadcrumb] If the inverseColor prop is not set to true, TopNavBarBreadcrumb fails to render.`);
    }
    const breadcrumbElement = _react.Children.toArray(children)[0];
    const breadCrumbLinks = breadcrumbElement.props.children;
    const lastButOneLink = _react.Children.toArray(breadCrumbLinks)[_react.Children.count(breadCrumbLinks) - 2];
    return this.context.inverseColor && (this.context.layout === 'desktop' ? (0, _jsxRuntime.jsxs)("div", {
      ref: this.handleRef,
      css: styles?.topNavBarBreadcrumb,
      "data-cid": "TopNavBarBreadcrumb",
      children: [(0, _jsxRuntime.jsx)("div", {
        css: styles?.iconContainer,
        children: this.renderMenu()
      }), (0, _jsxRuntime.jsx)("div", {
        css: styles?.breadcrumbContainer,
        children: children
      })]
    }) : (0, _jsxRuntime.jsx)("div", {
      ref: this.handleRef,
      css: styles?.topNavBarBreadcrumb,
      "data-cid": "TopNavBarBreadcrumb",
      children: (0, _jsxRuntime.jsx)("div", {
        css: styles?.linkContainer,
        children: lastButOneLink && (0, _jsxRuntime.jsx)(_latest.Link, {
          href: lastButOneLink.props.href,
          variant: "standalone",
          size: "medium",
          renderIcon: _uiIcons.ChevronLeftInstUIIcon,
          children: lastButOneLink.props.children
        })
      })
    }));
  }
}) || _class);
var _default = exports.default = TopNavBarBreadcrumb;
