// @instructure/ui-badge/lib/Badge/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Badge = void 0;
var _react = require("react");
var _latest = require("@instructure/ui-view/latest");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let Badge = exports.Badge = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = _dec2(_class = class Badge extends _react.Component {
  static displayName = "Badge";
  static componentId = 'Badge';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    standalone: false,
    type: 'count',
    variant: 'primary',
    display: 'inline-block',
    pulse: false,
    placement: 'top end',
    elementRef: () => {},
    formatOverflowText: (_count, countUntil) => `${countUntil - 1} +`
  };
  constructor(props) {
    super(props);
    this._defaultId = this.props.deterministicId();
  }
  _defaultId;
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  countOverflow() {
    const {
      count,
      countUntil
    } = this.props;
    return countUntil && count && countUntil > 1 && count >= countUntil;
  }
  renderOutput() {
    const {
      count,
      countUntil,
      formatOverflowText,
      formatOutput,
      type
    } = this.props;

    // If the badge count is >= than the countUntil limit, format the badge text
    // via the formatOverflowText function prop
    let formattedCount = (count || '').toString();
    if (count && countUntil && formatOverflowText && type === 'count' && this.countOverflow()) {
      formattedCount = formatOverflowText(count, countUntil);
    }
    if (typeof formatOutput === 'function') {
      return formatOutput(formattedCount);
    } else {
      return type === 'count' ? formattedCount : null;
    }
  }
  renderBadge() {
    const {
      count,
      margin,
      standalone,
      type,
      styles
    } = this.props;
    return (0, _jsxRuntime.jsx)(_latest.View, {
      margin: standalone ? margin : 'none',
      css: styles?.badge,
      title: type === 'count' && this.countOverflow() ? count?.toString() : '',
      id: !standalone ? this._defaultId : undefined,
      display: standalone ? 'inline-block' : 'block',
      ...(standalone && {
        elementRef: this.handleRef,
        'data-cid': 'Badge'
      }),
      children: this.renderOutput()
    });
  }
  renderChildren() {
    return _react.Children.map(this.props.children, child => {
      return (0, _safeCloneElement.safeCloneElement)(child, {
        'aria-describedby': this._defaultId
      });
    });
  }
  render() {
    const {
      margin,
      standalone,
      display,
      as,
      styles
    } = this.props;
    if (standalone) {
      return this.renderBadge();
    } else {
      return (0, _jsxRuntime.jsxs)(_latest.View, {
        as: as,
        margin: margin,
        elementRef: this.handleRef,
        css: styles?.wrapper,
        display: display,
        "data-cid": "Badge",
        children: [this.renderChildren(), this.renderBadge()]
      });
    }
  }
}) || _class) || _class);
var _default = exports.default = Badge;
