// @instructure/ui-rating/lib/Rating/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Rating = void 0;
var _react = require("react");
var _v11_ = require("@instructure/ui-view/v11_6");
var _ScreenReaderContent = require("@instructure/ui-a11y-content/lib/ScreenReaderContent");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _v = require("../../RatingIcon/v1");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let Rating = exports.Rating = (_dec = (0, _emotion.withStyle)(_styles.default, null), _dec(_class = class Rating extends _react.Component {
  static displayName = "Rating";
  static componentId = 'Rating';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    animateFill: false,
    formatValueText: (filled, iconCount) => `${filled} / ${iconCount}`,
    iconCount: 3,
    size: 'medium',
    valueNow: 0
  };
  static Icon = _v.RatingIcon;
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  get filled() {
    const {
      valueNow,
      iconCount,
      valueMax
    } = this.props;

    // prevent divide by zero errors
    const max = !!valueMax && valueMax > 0 ? valueMax : iconCount;
    const filledIcons = Math.round(valueNow * iconCount / max);

    // Handle edge case where valueNow is greater than valueMax
    if (filledIcons > iconCount) {
      return iconCount;
    } else {
      return filledIcons;
    }
  }
  get empty() {
    return this.props.iconCount - this.filled;
  }
  render() {
    const {
      iconCount,
      animateFill,
      size,
      margin,
      label,
      formatValueText
    } = this.props;
    const valueText = label + ' ' + formatValueText?.(this.filled, iconCount);
    const passthroughProps = _v11_.View.omitViewProps((0, _omitProps.omitProps)(this.props, Rating.allowedProps), Rating);
    return (0, _jsxRuntime.jsxs)(_v11_.View, {
      ...passthroughProps,
      css: this.props.styles?.rating,
      margin: margin,
      display: "inline-block",
      elementRef: this.handleRef,
      "data-cid": "Rating",
      children: [(0, _jsxRuntime.jsx)(_ScreenReaderContent.ScreenReaderContent, {
        children: valueText
      }), [...Array(this.filled)].map((_x, i) => (0, _jsxRuntime.jsx)(_v.RatingIcon, {
        filled: true,
        animateFill: animateFill,
        animationDelay: animateFill ? (i + 1) * 200 : undefined,
        size: size
      }, i + 1)), [...Array(this.empty)].map((_x, i) => (0, _jsxRuntime.jsx)(_v.RatingIcon, {
        size: size
      }, i + 1))]
    });
  }
}) || _class);
var _default = exports.default = Rating;
