// @instructure/ui-metric/es/MetricGroup/v1/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Children, Component } from 'react';
import { withStyle } from '@instructure/emotion';
import { passthroughProps, safeCloneElement } from '@instructure/ui-react-utils';
import generateStyle from "./styles.js";
import generateComponentTheme from "./theme.js";
import { allowedProps } from "./props.js";
import { jsx as _jsx } from "@emotion/react/jsx-runtime";
/**
---
category: components
---
**/
let MetricGroup = (_dec = withStyle(generateStyle, generateComponentTheme), _dec(_class = class MetricGroup extends Component {
  static displayName = "MetricGroup";
  static componentId = 'MetricGroup';
  static allowedProps = allowedProps;
  static defaultProps = {
    children: null
  };
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  renderChildren() {
    return Children.map(this.props.children, child => {
      return safeCloneElement(child, {
        isGroupChild: true
      });
    });
  }
  render() {
    return _jsx("div", {
      ...passthroughProps(this.props),
      css: this.props.styles?.metricGroup,
      role: "grid",
      "aria-readonly": "true",
      ref: this.handleRef,
      "data-cid": "MetricGroup",
      children: this.renderChildren()
    });
  }
}) || _class);
export default MetricGroup;
export { MetricGroup };
