// @instructure/ui-tabs/lib/Tabs/v2/Tab/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Tab = void 0;
var _react = require("react");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _latest = require("@instructure/ui-view/latest");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: Tabs
id: Tabs.Tab
---
**/
let Tab = exports.Tab = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class Tab extends _react.Component {
  static displayName = "Tab";
  static componentId = 'Tabs.Tab';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    variant: 'default',
    isDisabled: false,
    isSelected: false
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  handleClick = event => {
    const {
      onClick,
      index,
      id,
      isDisabled
    } = this.props;
    if (isDisabled) {
      return;
    }
    if (typeof onClick === 'function') {
      onClick(event, {
        index,
        id
      });
    }
  };
  handleKeyDown = event => {
    const {
      onKeyDown,
      index,
      id,
      isDisabled
    } = this.props;
    if (isDisabled) {
      return;
    }
    if (typeof onKeyDown === 'function') {
      onKeyDown(event, {
        index,
        id
      });
    }
  };
  render() {
    const {
      id,
      variant,
      isSelected,
      isDisabled,
      controls,
      children,
      styles,
      ...props
    } = this.props;
    return (0, _jsxRuntime.jsx)(_latest.View, {
      ...(0, _passthroughProps.passthroughProps)(props),
      as: "div",
      role: "tab",
      id: id,
      onClick: this.handleClick,
      onKeyDown: this.handleKeyDown,
      css: styles?.tab,
      "aria-selected": isSelected ? 'true' : undefined,
      "aria-disabled": isDisabled ? 'true' : undefined,
      "aria-controls": controls,
      tabIndex: isSelected && !isDisabled ? 0 : undefined,
      position: "relative",
      focusPosition: "inset",
      children: (0, _callRenderProp.callRenderProp)(children)
    });
  }
}) || _class);
var _default = exports.default = Tab;
