// @instructure/ui-pages/lib/Pages/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Page", {
  enumerable: true,
  get: function () {
    return _Page.Page;
  }
});
exports.default = exports.Pages = void 0;
var _react = require("react");
var _v11_ = require("@instructure/ui-view/v11_6");
var _containsActiveElement = require("@instructure/ui-dom-utils/lib/containsActiveElement.js");
var _findTabbable = require("@instructure/ui-dom-utils/lib/findTabbable.js");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _console = require("@instructure/console");
var _Page = require("./Page");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _PagesContext = require("./PagesContext");
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let Pages = exports.Pages = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = _dec2(_class = class Pages extends _react.Component {
  static displayName = "Pages";
  static componentId = 'Pages';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    activePageIndex: 0
  };
  static Page = _Page.Page;
  _timeouts = [];
  _activePage = null;
  _contentId;
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  get _contentElement() {
    return this.ref;
  }
  constructor(props) {
    super(props);
    this.state = {
      history: [typeof props.defaultPageIndex === 'number' ? props.defaultPageIndex : props.activePageIndex]
    };
    this._contentId = props.deterministicId();
  }
  componentDidMount() {
    this.props.makeStyles?.();
  }
  handleBackButtonClick = () => {
    const history = [...this.state.history];
    const oldPageIndex = history.pop();
    const newPageIndex = history[history.length - 1];
    this.setState({
      history
    });
    this.props.onPageIndexChange?.(newPageIndex || 0, oldPageIndex);
  };
  componentDidUpdate() {
    const {
      activePageIndex
    } = this.props;
    const {
      history
    } = this.state;
    if (typeof activePageIndex === 'number' && (history.length === 0 || activePageIndex !== history[history.length - 1])) {
      this.setState({
        history: [...history, activePageIndex]
      });
    }
    this._timeouts.push(setTimeout(() => {
      !this.focused && this.focus();
    }, 0));
    this.props.makeStyles?.();
  }
  componentWillUnmount() {
    this._timeouts.forEach(clearTimeout);
  }
  get focused() {
    return (0, _containsActiveElement.containsActiveElement)(this.ref);
  }
  focus() {
    this._timeouts.push(setTimeout(() => {
      const activePage = this._activePage;

      // Attempt to focus active page
      if (activePage && activePage.focusable) {
        activePage.focus();
      } else {
        // Use first tabbable as last ditch effort
        const tabbable = (0, _findTabbable.findTabbable)(this.ref);
        const element = tabbable && tabbable[0];
        element && element.focus();
      }
    }));
  }
  get activePage() {
    const {
      activePageIndex,
      children
    } = this.props;
    const pages = _react.Children.toArray(children);
    const activePage = activePageIndex < pages.length ? pages[activePageIndex] : null;
    (0, _console.logError)(!!activePage, '[Pages] Invalid `activePageIndex`.');
    return activePage ? (0, _safeCloneElement.safeCloneElement)(activePage, {
      ref: el => {
        this._activePage = el;
      }
    }) : null;
  }
  render() {
    const {
      activePage
    } = this;
    return activePage ? (0, _jsxRuntime.jsx)(_PagesContext.PagesContext.Provider, {
      value: {
        history: this.state.history,
        navigateToPreviousPage: () => {
          this.handleBackButtonClick();
        }
      },
      children: (0, _jsxRuntime.jsx)(_v11_.View, {
        as: "div",
        id: this._contentId,
        css: this.props.styles?.pages,
        margin: this.props.margin,
        role: "region",
        elementRef: this.handleRef,
        children: activePage
      })
    }) : null;
  }
}) || _class) || _class);
var _default = exports.default = Pages;
