// @instructure/ui-themes/es/themes/newThemeTokens/dark/components/modal.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const modal = semantics => ({
  autoMinWidth: "16em",
  backgroundColor: semantics.color.background.elevatedSurface.base,
  borderColor: semantics.color.stroke.container.base,
  inverseBackgroundColor: semantics.color.background.dark,
  inverseBorderColor: semantics.color.stroke.container.dark,
  borderRadius: semantics.borderRadius.xl,
  fontFamily: semantics.fontFamily.base,
  textColor: semantics.color.text.base,
  inverseTextColor: semantics.color.text.onColor,
  largeMaxWidth: "62em",
  mediumMaxWidth: "48em",
  smallMaxWidth: "30em",
  boxShadow: {
    0: {
      x: semantics.dropShadow.x.elevation4.dropshadow1,
      y: semantics.dropShadow.y.elevation4.dropshadow1,
      blur: semantics.dropShadow.blur.elevation4.dropshadow1,
      spread: semantics.dropShadow.spread.elevation4.dropshadow1,
      color: semantics.color.dropShadow.shadowColor2,
      type: "dropShadow"
    },
    1: {
      x: semantics.dropShadow.x.elevation4.dropshadow2,
      y: semantics.dropShadow.y.elevation4.dropshadow2,
      blur: semantics.dropShadow.blur.elevation4.dropshadow2,
      spread: semantics.dropShadow.spread.elevation4.dropshadow2,
      color: semantics.color.dropShadow.shadowColor1,
      type: "dropShadow"
    }
  },
  borderWidth: semantics.borderWidth.sm
});
export default modal;
