// @instructure/ui-themes/es/themes/newThemeTokens/dark/components/optionsItem.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const optionsItem = semantics => ({
  afterLabelContentVOffset: semantics.spacing.spaceMd,
  background: semantics.color.background.elevatedSurface.base,
  beforeLabelContentVOffset: semantics.spacing.spaceMd,
  color: semantics.color.text.base,
  descriptionColor: semantics.color.text.base,
  descriptionFontSize: semantics.fontSize.textSm,
  descriptionFontWeight: semantics.fontWeight.body.base,
  descriptionLineHeight: semantics.lineHeight.paragraph.base,
  fontFamily: semantics.fontFamily.base,
  fontSize: semantics.fontSize.textBase,
  fontWeight: semantics.fontWeight.body.base,
  fontWeightSelected: semantics.fontWeight.body.base,
  highlightedBackground: semantics.color.background.interactive.action.tertiary.hover,
  highlightedLabelColor: semantics.color.text.base,
  iconPadding: semantics.spacing.spaceMd,
  lineHeight: semantics.lineHeight.standalone.base,
  paddingHorizontal: semantics.spacing.spaceMd,
  paddingVertical: semantics.spacing.spaceMd,
  selectedBackground: semantics.color.background.interactive.action.primary.base,
  selectedHighlightedBackground: semantics.color.background.interactive.action.primary.hover,
  selectedLabelColor: semantics.color.text.interactive.action.primary.base,
  descriptionPaddingStart: "0rem",
  nestedPadding: semantics.spacing.spaceMd
});
export default optionsItem;
