// @instructure/ui-themes/es/themes/newThemeTokens/dark/components/progressCircle.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const progressCircle = semantics => ({
  fontFamily: semantics.fontFamily.base,
  fontWeight: semantics.fontWeight.heading.base,
  lineHeight: semantics.lineHeight.heading.base,
  xSmallSize: "3em",
  xSmallCircumference: "6.25em",
  xSmallTransform: "1.5em",
  xSmallStrokeWidth: "0.2em",
  smallSize: "5em",
  smallCircumference: "11em",
  smallTransform: "2.5em",
  smallStrokeWidth: "0.5em",
  mediumSize: "7em",
  mediumCircumference: "17.25em",
  mediumTransform: "3.5em",
  mediumStrokeWidth: "0.625em",
  largeSize: "9em",
  largeCircumference: "22em",
  largeTransform: "4.5em",
  largeStrokeWidth: "0.875em",
  color: semantics.color.text.base,
  colorInverse: semantics.color.text.inverse,
  trackColor: semantics.color.background.base,
  trackColorInverse: semantics.color.transparentColor,
  trackBorderColor: semantics.color.stroke.base,
  trackBorderColorInverse: semantics.color.stroke.inverse,
  meterColorAlert: semantics.color.stroke.info,
  meterColorAlertInverse: semantics.color.stroke.inverse,
  meterColorBrand: semantics.color.stroke.brand,
  meterColorBrandInverse: semantics.color.stroke.inverse,
  meterColorDanger: semantics.color.stroke.error,
  meterColorDangerInverse: semantics.color.stroke.inverse,
  meterColorInfo: semantics.color.stroke.info,
  meterColorInfoInverse: semantics.color.stroke.inverse,
  meterColorSuccess: semantics.color.stroke.success,
  meterColorSuccessInverse: semantics.color.stroke.inverse,
  meterColorWarning: semantics.color.stroke.warning,
  meterColorWarningInverse: semantics.color.stroke.inverse,
  xSmallRadius: "1em",
  smallRadius: "1.8em",
  mediumRadius: "2.75em",
  largeRadius: "3.5em",
  xSmallBorderOffset: "0.94em",
  smallBorderOffset: "1.6125em",
  mediumBorderOffset: "2.5em",
  largeBorderOffset: "3.125em"
});
export default progressCircle;
