// @instructure/ui-themes/es/themes/newThemeTokens/dark/components/view.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const view = semantics => ({
  fontFamily: semantics.fontFamily.base,
  color: semantics.color.text.base,
  colorPrimaryInverse: semantics.color.text.inverse,
  borderColorPrimary: semantics.color.stroke.muted,
  borderColorSecondary: semantics.color.stroke.base,
  borderColorSuccess: semantics.color.stroke.success,
  borderColorBrand: semantics.color.stroke.brand,
  borderColorInfo: semantics.color.stroke.info,
  borderColorAlert: semantics.color.stroke.info,
  borderColorWarning: semantics.color.stroke.warning,
  borderColorDanger: semantics.color.stroke.error,
  borderColorTransparent: "#00000000",
  debugOutlineColor: semantics.color.stroke.error,
  backgroundPrimary: semantics.color.background.elevatedSurface.base,
  backgroundSecondary: semantics.color.background.muted,
  backgroundPrimaryInverse: semantics.color.background.elevatedSurface.inverse,
  backgroundBrand: semantics.color.background.brand,
  backgroundInfo: semantics.color.background.info,
  backgroundAlert: semantics.color.background.info,
  backgroundSuccess: semantics.color.background.success,
  backgroundDanger: semantics.color.background.error,
  backgroundWarning: semantics.color.background.warning,
  stackingTopmost: 9999,
  stackingAbove: 1,
  stackingBelow: -1,
  stackingDeepest: -9999,
  borderStyle: "solid"
});
export default view;
