// @instructure/ui-metric/es/Metric/v2/index.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { forwardRef } from 'react';
import { useStyleNew } from '@instructure/emotion';
import { callRenderProp, passthroughProps } from '@instructure/ui-react-utils';
import generateStyle from "./styles.js";
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
/**
---
category: components
---
**/
const Metric = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    textAlign = 'center',
    renderLabel,
    renderValue,
    isGroupChild = false,
    themeOverride
  } = props;
  const styles = useStyleNew({
    generateStyle,
    themeOverride,
    params: {
      textAlign
    },
    componentId: 'Metric',
    displayName: 'Metric'
  });
  return _jsxs("div", {
    ...passthroughProps(props),
    role: isGroupChild ? 'row' : undefined,
    css: styles?.metric,
    ref: ref,
    "data-cid": "Metric",
    children: [_jsx("div", {
      role: isGroupChild ? 'rowheader' : undefined,
      css: styles?.label,
      children: callRenderProp(renderLabel)
    }), _jsx("div", {
      role: isGroupChild ? 'gridcell' : undefined,
      css: styles?.value,
      children: callRenderProp(renderValue)
    })]
  });
});
Metric.displayName = 'Metric';
export default Metric;
export { Metric };
