// @instructure/ui-themes/es/themes/newThemeTokens/legacyCanvasHighContrast/components/progressBar.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const progressBar = semantics => ({
  borderRadius: "0rem",
  fontFamily: semantics.fontFamily.base,
  fontSize: semantics.fontSize.textBase,
  xSmallValueFontSize: semantics.fontSize.textXs,
  smallValueFontSize: semantics.fontSize.textXs,
  mediumValueFontSize: semantics.fontSize.textSm,
  largeValueFontSize: semantics.fontSize.textBase,
  fontWeight: semantics.fontWeight.body.base,
  lineHeight: semantics.lineHeight.heading.base,
  xSmallHeight: "0.5rem",
  smallHeight: "1.125rem",
  mediumHeight: "1.5rem",
  meterColorAlert: semantics.color.background.info,
  meterColorAlertInverse: semantics.color.background.info,
  meterColorBrand: semantics.color.background.brand,
  meterColorBrandInverse: semantics.color.background.brand,
  meterColorDanger: semantics.color.background.error,
  meterColorDangerInverse: semantics.color.background.error,
  meterColorInfo: semantics.color.background.info,
  meterColorInfoInverse: semantics.color.background.info,
  meterColorSuccess: semantics.color.background.success,
  meterColorSuccessInverse: semantics.color.background.success,
  meterColorWarning: semantics.color.background.warning,
  meterColorWarningInverse: semantics.color.background.warning,
  largeHeight: "2.25rem",
  trackBottomBorderColor: semantics.color.stroke.strong,
  trackBottomBorderColorInverse: semantics.color.stroke.onColor,
  trackBottomBorderWidth: semantics.borderWidth.sm,
  trackColor: semantics.color.background.base,
  trackColorInverse: semantics.color.transparentColor,
  textColor: semantics.color.text.base,
  valuePadding: "0.375rem",
  textColorInverse: semantics.color.text.onColor,
  borderColor: semantics.color.transparentColor,
  borderColorInverse: semantics.color.transparentColor
});
export default progressBar;
