// @instructure/ui-themes/es/themes/newThemeTokens/legacyCanvasHighContrast/components/sideNavBar.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const sideNavBar = semantics => ({
  backgroundColor: semantics.color.institutional.brandGlobalNavBgd,
  fill: semantics.color.icon.onColor,
  fontColor: semantics.color.institutional.brandGlobalNavMenuItemTextColor,
  minimizedWidth: "3.375rem",
  toggleTransition: "300ms",
  width: "5.25rem",
  margin: "0rem",
  minimizedMargin: "0rem",
  contentMargin: "0rem",
  contentGap: "0rem",
  boxShadow: {
    0: {
      x: "0rem",
      y: "0rem",
      blur: "0rem",
      spread: "0rem",
      color: semantics.color.transparentColor,
      type: "dropShadow"
    },
    1: {
      x: "0rem",
      y: "0rem",
      blur: "0rem",
      spread: "0rem",
      color: semantics.color.transparentColor,
      type: "dropShadow"
    }
  },
  borderRadius: "0rem",
  minimizedBorderRadius: "0rem"
});
export default sideNavBar;
