// @instructure/ui-themes/es/themes/newThemeTokens/legacyCanvasHighContrast/components/tag.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const tag = semantics => ({
  defaultBackground: semantics.color.background.interactive.action.secondary.base,
  defaultBackgroundHover: semantics.color.background.interactive.action.secondary.hover,
  defaultBorderColor: semantics.color.stroke.interactive.action.secondary.base,
  defaultBorderRadius: semantics.borderRadius.full,
  defaultBorderStyle: "solid",
  inlineBorderStyle: "solid",
  defaultBorderWidth: semantics.borderWidth.sm,
  inlineBorderWidth: semantics.borderWidth.sm,
  defaultColor: semantics.color.institutional.brandFontColorDark,
  inlineColor: semantics.color.text.base,
  fontFamily: semantics.fontFamily.base,
  fontSizeLarge: semantics.fontSize.textBase,
  fontSizeMedium: semantics.fontSize.textSm,
  fontSizeSmall: semantics.fontSize.textXs,
  heightLarge: "2.375rem",
  heightMedium: "1.75rem",
  heightSmall: "1.3125rem",
  iconMargin: "0.5rem",
  inlineBackground: semantics.color.background.interactive.action.secondary.base,
  inlineBackgroundHover: semantics.color.background.interactive.action.secondary.hover,
  inlineBorderColor: semantics.color.stroke.interactive.action.secondary.base,
  inlineBorderRadius: semantics.borderRadius.sm,
  maxWidth: "10rem",
  paddingHorizontal: semantics.spacing.padding.container.xxs,
  paddingHorizontalSmall: semantics.spacing.padding.container.xxs,
  transitionTiming: "0.2s",
  inlineIconColor: semantics.color.background.interactive.action.primary.base,
  inlineIconHoverColor: semantics.color.background.interactive.action.primary.hover
});
export default tag;
