// @instructure/ui-themes/es/themes/newThemeTokens/legacyCanvasHighContrast/components/topNavBarLayout.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const topNavBarLayout = semantics => ({
  smallViewportDropdownMenuActiveOptionIndicatorSpacing: semantics.spacing.spaceXs,
  smallViewportDropdownMenuActiveOptionIndicatorWidth: semantics.borderWidth.lg,
  desktopHeight: "4rem",
  desktopInlinePadding: semantics.spacing.padding.container.xs,
  desktopBackground: semantics.color.background.inverse,
  desktopBackgroundInverse: semantics.color.background.muted,
  desktopFontFamily: semantics.fontFamily.base,
  smallViewportFontFamily: semantics.fontFamily.base,
  desktopFontWeight: semantics.fontWeight.body.base,
  smallViewportFontWeight: semantics.fontWeight.body.base,
  desktopFontSize: semantics.fontSize.textBase,
  desktopBottomBorder: "none",
  desktopMenuItemsContainerInlineMargin: "0 2.25rem",
  desktopUserContainerInlineMargin: "0.5rem 0",
  desktopUserSeparatorColor: semantics.color.stroke.muted,
  desktopUserSeparatorColorInverse: semantics.color.stroke.muted,
  desktopUserSeparatorGap: semantics.spacing.spaceSm,
  desktopUserSeparatorHeight: "1.5rem",
  desktopUserSeparatorWidth: semantics.borderWidth.sm,
  desktopZIndex: 9998,
  smallViewportAlternativeTitleBlockMargin: "0rem",
  smallViewportAlternativeTitleInlineMargin: "0.375rem",
  smallViewportBackground: semantics.color.background.dark,
  smallViewportBackgroundInverse: semantics.color.background.muted,
  smallViewportBottomBorder: "none",
  smallViewportBottomBorderInverse: "0.0625rem solid #E8EAEC",
  smallViewportDropdownMenuActiveOptionFontWeight: semantics.fontWeight.body.strong,
  smallViewportDropdownMenuActiveOptionIndicatorColor: "currentColor",
  smallViewportHeight: "3.5rem",
  smallViewportInlinePadding: "0.25rem",
  smallViewportTrayPosition: "fixed",
  smallViewportTrayZIndex: 9998,
  smallViewportZIndex: 10000,
  desktopBottomBorderInverse: semantics.color.stroke.inverse,
  smallViewportFontSize: semantics.fontSize.textSm,
  smallViewportTrayFixTopPosition: "undefined",
  desktopBrandContainerInlineMargin: "0 1.5rem",
  desktopActionItemsContainerInlineMargin: "0.25rem 0"
});
export default topNavBarLayout;
