// @instructure/ui-options/lib/Options/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Options = void 0;
var _react = require("react");
var _uiReactUtils = require("@instructure/ui-react-utils");
var _latest = require("@instructure/ui-view/latest");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _Item = require("./Item");
var _Separator = require("./Separator");
var _props = require("./props");
var _uiUtils = require("@instructure/ui-utils");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components/utilities
---
**/
let Options = exports.Options = (_dec = (0, _uiReactUtils.withDeterministicId)(), _dec2 = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = _dec2(_class = class Options extends _react.Component {
  static displayName = "Options";
  static componentId = 'Options';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    as: 'span',
    role: 'list',
    elementRef: () => {},
    renderLabel: null,
    children: null
  };
  static Item = _Item.Item;
  static Separator = _Separator.Separator;
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  _labelId = this.props.deterministicId('Options-label');
  get childAs() {
    const {
      as
    } = this.props;
    if (as === 'ul' || as === 'ol') {
      return 'li';
    }
    return undefined;
  }
  renderLabel() {
    const {
      renderLabel,
      styles
    } = this.props;
    let labelContent;
    if (typeof renderLabel === 'function') {
      labelContent = /*#__PURE__*/(0, _react.createElement)(renderLabel, {});
    } else {
      labelContent = renderLabel;
    }
    return (0, _jsxRuntime.jsx)("span", {
      id: this._labelId,
      role: "presentation"
      // because TalkBack and iOS VoiceOver don't announce sub-menu labels, aria-hidden needs to be false to achive that
      ,
      "aria-hidden": (0, _uiUtils.isAndroidOrIOS)() ? 'false' : 'true',
      css: styles?.label,
      children: labelContent
    });
  }
  renderSubList(subOptions) {
    const {
      styles
    } = this.props;
    return (0, _jsxRuntime.jsx)(_Item.Item, {
      as: this.childAs,
      role: "presentation",
      css: styles?.label,
      children: subOptions
    });
  }
  renderChildren() {
    const {
      children
    } = this.props;
    return _react.Children.map(children, child => {
      if ((0, _uiReactUtils.matchComponentTypes)(child, ['Options'])) {
        return this.renderSubList(child);
      }
      if ((0, _uiReactUtils.matchComponentTypes)(child, ['Item', 'Separator'])) {
        return (0, _uiReactUtils.safeCloneElement)(child, {
          as: this.childAs || child.props.as
        });
      }
      return undefined;
    });
  }
  get ariaLabelledby() {
    if (this.props.renderLabel) {
      return this._labelId;
    }
    if (this.props['aria-labelledby']) {
      return this.props['aria-labelledby'];
    }
    return undefined;
  }
  render() {
    const passthroughProps = _latest.View.omitViewProps((0, _uiReactUtils.omitProps)(this.props, Options.allowedProps), Options);
    const {
      as,
      role,
      renderLabel,
      elementRef,
      styles
    } = this.props;
    return (0, _jsxRuntime.jsxs)("div", {
      css: styles?.options,
      role: "presentation",
      ref: this.handleRef,
      "data-cid": "Options",
      children: [renderLabel && this.renderLabel(), (0, _jsxRuntime.jsx)(_latest.View, {
        ...passthroughProps,
        elementRef: elementRef,
        css: styles?.list,
        as: as,
        role: role,
        display: "block",
        margin: "none",
        padding: "none",
        background: "primary",
        "aria-labelledby": this.ariaLabelledby,
        children: this.renderChildren()
      })]
    });
  }
}) || _class) || _class);
var _default = exports.default = Options;
