// @instructure/ui-top-nav-bar/es/TopNavBar/v2/TopNavBarBrand/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { getElementType, omitProps } from '@instructure/ui-react-utils';
import { withStyleNew, InstUISettingsProvider } from '@instructure/emotion';
import { ScreenReaderContent } from '@instructure/ui-a11y-content';
import { View } from '@instructure/ui-view/latest';
import { TopNavBarContext } from "../TopNavBarContext.js";
import generateStyle from "./styles.js";
import { allowedProps } from "./props.js";
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
/**
---
parent: TopNavBar
id: TopNavBar.Brand
---
@module TopNavBarBrand
**/
let TopNavBarBrand = (_dec = withStyleNew(generateStyle, 'TopNavBarBrand'), _dec(_class = class TopNavBarBrand extends Component {
  static displayName = "TopNavBarBrand";
  static componentId = 'TopNavBar.Brand';
  // TODO: add to the docs: making it static on parent and jsdocs parent/module settings, dont export child on its own

  static allowedProps = allowedProps;
  static defaultProps = {};
  static contextType = TopNavBarContext;
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.(this.makeStylesVariables);
  }
  componentDidUpdate() {
    this.props.makeStyles?.(this.makeStylesVariables);
  }
  get makeStylesVariables() {
    return {
      layout: this.context.layout
    };
  }
  render() {
    const {
      screenReaderLabel,
      renderIcon,
      href,
      onClick,
      styles
    } = this.props;
    const ElementType = getElementType(TopNavBarBrand, this.props);
    return _jsx("div", {
      ref: this.handleRef,
      css: styles?.topNavBarBrand,
      "data-cid": "TopNavBarBrand",
      children: renderIcon && _jsx(InstUISettingsProvider, {
        theme: currentTheme => ({
          newTheme: {
            sharedTokens: {
              focusOutline: {
                ...currentTheme?.newTheme?.sharedTokens?.focusOutline,
                inset: styles?.focusOutlineInset
              }
            }
          }
        }),
        children: _jsxs(View, {
          ...omitProps(this.props, allowedProps),
          css: styles?.container,
          as: ElementType,
          href: href,
          onClick: onClick,
          position: "relative",
          focusColor: this.context.inverseColor ? 'info' : 'inverse',
          focusPosition: "inset",
          borderRadius: "medium",
          children: [_jsx(ScreenReaderContent, {
            children: screenReaderLabel
          }), renderIcon && this.context.layout !== 'smallViewport' && _jsx("div", {
            css: styles?.iconContainer,
            role: "presentation",
            "aria-hidden": "true",
            children: renderIcon
          })]
        })
      })
    });
  }
}) || _class);
export { TopNavBarBrand };
export default TopNavBarBrand;
