// @instructure/ui-byline/lib/Byline/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Byline = void 0;
var _react = require("react");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _latest = require("@instructure/ui-view/latest");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let Byline = exports.Byline = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class Byline extends _react.Component {
  static displayName = "Byline";
  static componentId = 'Byline';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    alignContent: 'center'
  };
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  render() {
    const passthroughProps = _latest.View.omitViewProps((0, _omitProps.omitProps)(this.props, Byline.allowedProps), Byline);
    return (0, _jsxRuntime.jsxs)(_latest.View, {
      ...passthroughProps,
      elementRef: this.handleRef,
      css: this.props.styles?.byline,
      margin: this.props.margin,
      maxWidth: this.props.styles?.maxWidth,
      children: [(0, _jsxRuntime.jsx)("div", {
        css: this.props.styles?.figure,
        children: this.props.children
      }), (0, _jsxRuntime.jsxs)("div", {
        css: this.props.styles?.caption,
        children: [this.props.title && (0, _jsxRuntime.jsx)("span", {
          css: this.props.styles?.title,
          children: this.props.title
        }), this.props.description && (0, _jsxRuntime.jsx)("div", {
          css: this.props.styles?.description,
          children: this.props.description
        })]
      })]
    });
  }
}) || _class);
var _default = exports.default = Byline;
