// @instructure/ui-top-nav-bar/lib/TopNavBar/v1/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TopNavBar = void 0;
var _react = require("react");
var _console = require("@instructure/console");
var _px = require("@instructure/ui-utils/lib/px.js");
var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");
var _v11_ = require("@instructure/ui-responsive/v11_6");
var _TopNavBarActionItems = require("./TopNavBarActionItems");
var _TopNavBarBrand = require("./TopNavBarBrand");
var _TopNavBarItem = require("./TopNavBarItem");
var _TopNavBarLayout = require("./TopNavBarLayout");
var _TopNavBarMenuItems = require("./TopNavBarMenuItems");
var _TopNavBarUser = require("./TopNavBarUser");
var _TopNavBarBreadcrumb = require("./TopNavBarBreadcrumb");
var _TopNavBarContext = require("./TopNavBarContext");
var _props2 = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
category: components
---
**/
class TopNavBar extends _react.Component {
  static displayName = "TopNavBar";
  static componentId = 'TopNavBar';

  // TODO: mention subcomponent registration in "new component" docs
  static ActionItems = _TopNavBarActionItems.TopNavBarActionItems;
  static Brand = _TopNavBarBrand.TopNavBarBrand;
  static Item = _TopNavBarItem.TopNavBarItem;
  static Layout = _TopNavBarLayout.TopNavBarLayout;
  static MenuItems = _TopNavBarMenuItems.TopNavBarMenuItems;
  static User = _TopNavBarUser.TopNavBarUser;
  static Breadcrumb = _TopNavBarBreadcrumb.TopNavBarBreadcrumb;
  static contextType = _TopNavBarContext.TopNavBarContext;
  static allowedProps = _props2.allowedProps;
  static defaultProps = {
    breakpoint: 1024,
    mediaQueryMatch: 'media',
    inverseColor: false
  };
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  get breakpoint() {
    return (0, _px.px)(this.props.breakpoint);
  }
  render() {
    const {
      children,
      mediaQueryMatch,
      inverseColor
    } = this.props;
    return (0, _jsxRuntime.jsx)(_v11_.Responsive, {
      "data-cid": "TopNavBar",
      elementRef: this.handleRef,
      match: mediaQueryMatch,
      query: {
        smallViewport: {
          maxWidth: this.breakpoint - 1
        },
        desktop: {
          minWidth: this.breakpoint
        }
      },
      render: (_props, matches) => {
        const layout = matches ? matches[0] : 'smallViewport';
        const isInverseColor = typeof inverseColor === 'function' ? inverseColor(layout) : !!inverseColor;
        const content = children({
          currentLayout: layout,
          inverseColor: isInverseColor
        });
        if (!(0, _matchComponentTypes.matchComponentTypes)(content, [_TopNavBarLayout.TopNavBarLayout])) {
          (0, _console.error)(false, 'The `children` function prop of TopNavBar has to return a child of type <TopNavBar.Layout>, but it returned:', content);
        }
        return (0, _jsxRuntime.jsx)(_TopNavBarContext.TopNavBarContext.Provider, {
          value: {
            layout,
            inverseColor: isInverseColor
          },
          children: content
        });
      }
    });
  }
}
exports.TopNavBar = TopNavBar;
var _default = exports.default = TopNavBar;
