// @instructure/ui-dialog/lib/Dialog/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Dialog = void 0;
var _react = require("react");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _getElementType = require("@instructure/ui-react-utils/lib/getElementType.js");
var _findDOMNode = require("@instructure/ui-dom-utils/lib/findDOMNode.js");
var _requestAnimationFrame = require("@instructure/ui-dom-utils/lib/requestAnimationFrame.js");
var _console = require("@instructure/console");
var _FocusRegionManager = require("@instructure/ui-a11y-utils/lib/FocusRegionManager.js");
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
category: components/utilities
---
@module Dialog
**/
class Dialog extends _react.Component {
  static displayName = "Dialog";
  static componentId = 'Dialog';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    open: false,
    shouldFocusOnOpen: true,
    shouldContainFocus: false,
    shouldReturnFocus: false,
    shouldCloseOnDocumentClick: true,
    shouldCloseOnEscape: true
  };
  _raf = [];
  _focusRegion = null;
  ref = null;
  get _root() {
    return this.ref;
  }
  componentDidUpdate(prevProps) {
    const {
      open,
      contentElement,
      ...options
    } = this.props;
    if (open && !prevProps.open) {
      this.open();
    } else if (!open && prevProps.open) {
      this.close();
    }
    if (this._focusRegion) {
      this._focusRegion.updateElement(this.contentElement, options);
    }
  }
  componentDidMount() {
    if (this.props.open) {
      this.open();
    }
  }
  componentWillUnmount() {
    if (this.props.open) {
      this.close();
    }
    this._raf.forEach(request => request.cancel());
    this._raf = [];
  }
  open() {
    const {
      open,
      contentElement,
      ...options
    } = this.props;
    this._raf.push((0, _requestAnimationFrame.requestAnimationFrame)(() => {
      this._focusRegion = _FocusRegionManager.FocusRegionManager.activateRegion(this.contentElement, {
        ...options
      });
    }));
  }
  close() {
    const {
      _focusRegion,
      contentElement
    } = this;
    // setTimeout is used here to delay the blur after the mousedown event in FocusRegion for correct execution order
    if (_focusRegion) {
      _FocusRegionManager.FocusRegionManager.blurRegion(contentElement, _focusRegion.id);
    }
  }
  focus() {
    if (!this.props.open || !this.contentElement) {
      (0, _console.logError)(false, "[Dialog] Can't focus a Dialog that isn't open.");
      return;
    }
    if (this._focusRegion) {
      _FocusRegionManager.FocusRegionManager.focusRegion(this.contentElement, this._focusRegion.id);
    }
  }
  blur() {
    if (!this.props.open || !this.contentElement) {
      (0, _console.logError)(false, "[Dialog] Can't blur a Dialog that isn't open.");
      return;
    }
    this.close();
  }
  getRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  get contentElement() {
    let contentElement = (0, _findDOMNode.findDOMNode)(this.props.contentElement);
    if (!contentElement) {
      contentElement = this.ref ?? undefined;
    }
    return contentElement;
  }
  get focused() {
    return this.contentElement && this._focusRegion && _FocusRegionManager.FocusRegionManager.isFocused(this.contentElement, this._focusRegion.id);
  }
  render() {
    const ElementType = (0, _getElementType.getElementType)(Dialog, this.props);

    // In case the HTML role attribute is explicitly passed, use props.role
    const role = this.props.role || (this.props.label ? 'dialog' : undefined);
    return this.props.open ? (0, _jsxRuntime.jsx)(ElementType, {
      ...(0, _omitProps.omitProps)(this.props, Dialog.allowedProps),
      role: role,
      "aria-label": this.props.label,
      className: this.props.className // TODO in V2 remove className, there is no usage of it.
      ,
      style: {
        borderRadius: 'inherit'
      } // ensure the dialog inherits border radius from View
      ,
      ref: this.getRef,
      children: this.props.children
    }) : null;
  }
}
exports.Dialog = Dialog;
var _default = exports.default = Dialog;
