// @instructure/ui-table/lib/exports/a.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Table", {
  enumerable: true,
  get: function () {
    return _v.Table;
  }
});
Object.defineProperty(exports, "TableBody", {
  enumerable: true,
  get: function () {
    return _Body.Body;
  }
});
Object.defineProperty(exports, "TableCell", {
  enumerable: true,
  get: function () {
    return _Cell.Cell;
  }
});
Object.defineProperty(exports, "TableColHeader", {
  enumerable: true,
  get: function () {
    return _ColHeader.ColHeader;
  }
});
Object.defineProperty(exports, "TableContext", {
  enumerable: true,
  get: function () {
    return _TableContext.TableContext;
  }
});
Object.defineProperty(exports, "TableHead", {
  enumerable: true,
  get: function () {
    return _Head.Head;
  }
});
Object.defineProperty(exports, "TableRow", {
  enumerable: true,
  get: function () {
    return _Row.Row;
  }
});
Object.defineProperty(exports, "TableRowHeader", {
  enumerable: true,
  get: function () {
    return _RowHeader.RowHeader;
  }
});
var _v = require("../Table/v1");
var _TableContext = require("../Table/v1/TableContext");
var _Body = require("../Table/v1/Body");
var _Cell = require("../Table/v1/Cell");
var _ColHeader = require("../Table/v1/ColHeader");
var _Head = require("../Table/v1/Head");
var _Row = require("../Table/v1/Row");
var _RowHeader = require("../Table/v1/RowHeader");
