// @instructure/ui-color-utils/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "alpha", {
  enumerable: true,
  get: function () {
    return _alpha.alpha;
  }
});
Object.defineProperty(exports, "color2hex", {
  enumerable: true,
  get: function () {
    return _conversions.color2hex;
  }
});
Object.defineProperty(exports, "colorToHex8", {
  enumerable: true,
  get: function () {
    return _conversions.colorToHex8;
  }
});
Object.defineProperty(exports, "colorToHsla", {
  enumerable: true,
  get: function () {
    return _conversions.colorToHsla;
  }
});
Object.defineProperty(exports, "colorToHsva", {
  enumerable: true,
  get: function () {
    return _conversions.colorToHsva;
  }
});
Object.defineProperty(exports, "colorToRGB", {
  enumerable: true,
  get: function () {
    return _conversions.colorToRGB;
  }
});
Object.defineProperty(exports, "contrast", {
  enumerable: true,
  get: function () {
    return _contrast.contrast;
  }
});
Object.defineProperty(exports, "contrastWithAlpha", {
  enumerable: true,
  get: function () {
    return _contrastWithAlpha.contrastWithAlpha;
  }
});
Object.defineProperty(exports, "darken", {
  enumerable: true,
  get: function () {
    return _darken.darken;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "isValid", {
  enumerable: true,
  get: function () {
    return _isValid.isValid;
  }
});
Object.defineProperty(exports, "lighten", {
  enumerable: true,
  get: function () {
    return _lighten.lighten;
  }
});
Object.defineProperty(exports, "overlayColors", {
  enumerable: true,
  get: function () {
    return _overlayColors.overlayColors;
  }
});
Object.defineProperty(exports, "validateContrast", {
  enumerable: true,
  get: function () {
    return _validateContrast.validateContrast;
  }
});
var _alpha = require("./alpha");
var _darken = require("./darken");
var _lighten = require("./lighten");
var _contrast = require("./contrast");
var _isValid = require("./isValid");
var _overlayColors = require("./overlayColors");
var _contrastWithAlpha = require("./contrastWithAlpha");
var _validateContrast = require("./validateContrast");
var _conversions = require("./conversions");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
// TODO remove when we get rid of babel-plugin-transform-imports
// This default export is needed because babel-plugin-transform-imports will
// fail if the exported name is not the same as the filename
var _default = exports.default = {
  color2hex: _conversions.color2hex,
  colorToHex8: _conversions.colorToHex8,
  colorToHsva: _conversions.colorToHsva,
  colorToHsla: _conversions.colorToHsla,
  colorToRGB: _conversions.colorToRGB
};
