// @instructure/ui-toggle-details/lib/ToggleDetails/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ToggleDetails = void 0;
var _react = require("react");
var _latest = require("@instructure/ui-buttons/latest");
var _uiIcons = require("@instructure/ui-icons");
var _uiExpandable = require("@instructure/ui-expandable");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _pickProps = require("@instructure/ui-react-utils/lib/pickProps.js");
var _isActiveElement = require("@instructure/ui-dom-utils/lib/isActiveElement.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let ToggleDetails = exports.ToggleDetails = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class ToggleDetails extends _react.Component {
  static displayName = "ToggleDetails";
  static componentId = 'ToggleDetails';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    variant: 'default',
    size: 'medium',
    fluidWidth: false,
    icon: _uiIcons.ChevronRightInstUIIcon,
    iconExpanded: _uiIcons.ChevronDownInstUIIcon,
    iconPosition: 'start',
    defaultExpanded: false,
    children: null
  };
  _buttonNode = null;
  ref = /*#__PURE__*/(0, _react.createRef)();

  // TODO this solution was generated with Gemini, should be tested if it works as fine as before
  setButtonRef = node => {
    this._buttonNode = node;
  };
  get focused() {
    return (0, _isActiveElement.isActiveElement)(this._buttonNode);
  }
  focus() {
    this._buttonNode?.focus();
  }
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  renderSummary(expanded) {
    const {
      summary,
      iconPosition
    } = this.props;
    return (0, _jsxRuntime.jsxs)("span", {
      css: this.props.styles?.summary,
      children: [iconPosition === 'start' && this.renderIcon(expanded), (0, _jsxRuntime.jsx)("span", {
        css: this.props.styles?.summaryText,
        children: summary
      }), iconPosition === 'end' && this.renderIcon(expanded)]
    });
  }
  renderToggle(toggleProps, expanded) {
    const {
      variant
    } = this.props;
    const tProps = this.props.children ? toggleProps : {
      onClick: toggleProps.onClick
    };
    const props = {
      ...(0, _omitProps.omitProps)(this.props, ToggleDetails.allowedProps),
      ...tProps,
      children: this.renderSummary(expanded)
    };
    const summary = this.renderSummary(expanded);
    if (variant === 'filled') {
      return (0, _jsxRuntime.jsx)(_latest.Button, {
        ...props,
        display: "block",
        textAlign: "start",
        elementRef: this.setButtonRef,
        children: summary
      });
    } else if (props.href) {
      return (0, _jsxRuntime.jsx)("a", {
        ...props,
        css: this.props.styles?.toggle,
        ref: this.setButtonRef,
        children: summary
      });
    } else {
      return (0, _jsxRuntime.jsx)("button", {
        ...props,
        type: "button",
        css: this.props.styles?.toggle,
        ref: this.setButtonRef,
        children: summary
      });
    }
  }
  renderIcon(expanded) {
    const Icon = expanded ? this.props.iconExpanded : this.props.icon;
    const iconSize = this.props.size === 'large' ? 'lg' : 'md';
    const iconColor = this.props.variant === 'filled' ? 'actionSecondaryBaseColor' : 'baseColor';
    return this.props.children && Icon ? (0, _jsxRuntime.jsx)("span", {
      css: this.props.styles?.icon,
      children: (0, _jsxRuntime.jsx)(Icon, {
        size: iconSize,
        color: iconColor
      })
    }) : null;
  }
  renderDetails(expanded, detailsProps) {
    if (!this.props.children) return null;
    const expandedStyles = expanded ? {
      display: 'block'
    } : {
      display: 'none'
    };
    return (0, _jsxRuntime.jsx)("div", {
      ...detailsProps,
      css: [this.props.styles?.details, expandedStyles],
      children: expanded && this.renderContent()
    });
  }
  renderContent() {
    return (0, _jsxRuntime.jsx)("div", {
      css: this.props.styles?.content,
      children: this.props.children
    });
  }
  handleToggle = (event, expanded) => {
    if (typeof this.props.onToggle === 'function') {
      this.props.onToggle(event, expanded);
    }
  };
  render() {
    return (0, _jsxRuntime.jsx)(_uiExpandable.Expandable, {
      ...(0, _pickProps.pickProps)(this.props, _uiExpandable.Expandable.allowedProps),
      onToggle: this.handleToggle,
      children: ({
        expanded,
        getToggleProps,
        getDetailsProps
      }) => {
        return (0, _jsxRuntime.jsxs)("div", {
          css: this.props.styles?.toggleDetails,
          ref: this.ref,
          "data-cid": "ToggleDetails",
          children: [this.renderToggle(getToggleProps(), expanded), this.renderDetails(expanded, getDetailsProps())]
        });
      }
    });
  }
}) || _class);
var _default = exports.default = ToggleDetails;
