// @instructure/ui-table/lib/Table/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Table = void 0;
var _react = require("react");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _v11_ = require("@instructure/ui-view/v11_6");
var _ScreenReaderContent = require("@instructure/ui-a11y-content/lib/ScreenReaderContent");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _Head = require("./Head");
var _Body = require("./Body");
var _Row = require("./Row");
var _ColHeader = require("./ColHeader");
var _RowHeader = require("./RowHeader");
var _Cell = require("./Cell");
var _props = require("./props");
var _TableContext = _interopRequireDefault(require("./TableContext"));
var _console = require("@instructure/console");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let Table = exports.Table = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class Table extends _react.Component {
  static displayName = "Table";
  static componentId = 'Table';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    children: null,
    hover: false,
    layout: 'auto'
  };
  static Head = _Head.Head;
  static Body = _Body.Body;
  static Row = _Row.Row;
  static ColHeader = _ColHeader.ColHeader;
  static RowHeader = _RowHeader.RowHeader;
  static Cell = _Cell.Cell;
  ref = null;
  // Reference to hidden aria-live region for announcing caption changes to screen readers
  _liveRegionRef = null;
  // Timeout for delayed announcement (workaround for Safari/VoiceOver caption update bug)
  _announcementTimeout;
  handleRef = el => {
    this.ref = el;
    this.props.elementRef?.(el);
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate(prevProps) {
    this.props.makeStyles?.();
    // Announce caption changes for screen readers (especially VoiceOver)
    // Safari/VoiceOver has a known bug where dynamic <caption> updates aren't announced,
    // so we use an aria-live region as a workaround
    const prevSortInfo = this.getSortedHeaderInfo(prevProps);
    const currentSortInfo = this.getSortedHeaderInfo(this.props);
    // Only announce if sorting actually changed
    const sortingChanged = prevSortInfo?.header !== currentSortInfo?.header || prevSortInfo?.direction !== currentSortInfo?.direction;
    if (sortingChanged && currentSortInfo && this._liveRegionRef) {
      // Clear any pending announcement
      clearTimeout(this._announcementTimeout);
      // Clear the live region first (part of the clear-then-set pattern)
      this._liveRegionRef.textContent = '';
      // Wait 100ms before setting new content to ensure screen readers detect the change
      this._announcementTimeout = setTimeout(() => {
        if (this._liveRegionRef) {
          const currentCaption = this.getCaptionText(this.props);
          // Append non-breaking space (\u00A0) to force Safari/VoiceOver to treat
          // repeated captions as different announcements
          this._liveRegionRef.textContent = currentCaption + '\u00A0';
        }
      }, 100);
    }
  }
  componentWillUnmount() {
    // Clean up pending announcement timeout
    clearTimeout(this._announcementTimeout);
  }
  getHeaders() {
    const [headChild] = _react.Children.toArray(this.props.children);
    if (!headChild || ! /*#__PURE__*/(0, _react.isValidElement)(headChild)) return undefined;
    const [firstRow] = _react.Children.toArray(headChild.props.children);
    if (!firstRow || ! /*#__PURE__*/(0, _react.isValidElement)(firstRow)) return undefined;
    return _react.Children.map(firstRow.props.children, colHeader => {
      if (! /*#__PURE__*/(0, _react.isValidElement)(colHeader)) return undefined;
      return colHeader.props.children;
    });
  }
  getSortedHeaderInfo(props) {
    const [headChild] = _react.Children.toArray(props.children);
    const [firstRow] = _react.Children.toArray(/*#__PURE__*/(0, _react.isValidElement)(headChild) ? headChild.props.children : []);
    const colHeaders = _react.Children.toArray(/*#__PURE__*/(0, _react.isValidElement)(firstRow) ? firstRow.props.children : []);
    // Find the column with an active sort direction
    for (const colHeader of colHeaders) {
      if (/*#__PURE__*/(0, _react.isValidElement)(colHeader) && colHeader.props.sortDirection && colHeader.props.sortDirection !== 'none') {
        // Extract header text (may be nested in child components)
        const headerText = typeof colHeader.props.children === 'string' ? colHeader.props.children : colHeader.props.children?.props?.children ?? '';
        return {
          header: headerText,
          direction: colHeader.props.sortDirection
        };
      }
    }
    return null;
  }
  getCaptionText(props) {
    const sortInfo = this.getSortedHeaderInfo(props);
    const caption = props.caption;
    if (!sortInfo) return caption;
    const sortText = ` Sorted by ${sortInfo.header} (${sortInfo.direction})`;
    return caption ? caption + sortText : sortText.trim();
  }
  render() {
    const {
      margin,
      layout,
      caption,
      children,
      hover,
      styles,
      minWidth
    } = this.props;
    const isStacked = layout === 'stacked';
    const captionText = this.getCaptionText(this.props);
    if (!caption) {
      (0, _console.error)(false, `[Table] required prop caption is not set.`);
    }
    return (0, _jsxRuntime.jsxs)(_TableContext.default.Provider, {
      value: {
        isStacked,
        hover: hover,
        headers: isStacked ? this.getHeaders() : undefined
      },
      children: [(0, _jsxRuntime.jsx)("div", {
        ref: el => {
          this._liveRegionRef = el;
        },
        "aria-live": "polite",
        "aria-atomic": "true",
        role: "status",
        css: styles?.liveRegion
      }), (0, _jsxRuntime.jsxs)(_v11_.View
      // All HTML props, except the ones accepted by `View` and `Table`
      , {
        ..._v11_.View.omitViewProps((0, _omitProps.omitProps)(this.props, Table.allowedProps), Table),
        minWidth: minWidth,
        as: isStacked ? 'div' : 'table',
        margin: margin,
        elementRef: this.handleRef,
        css: styles?.table,
        role: isStacked ? 'table' : undefined,
        "aria-label": captionText,
        children: [!isStacked && caption && (0, _jsxRuntime.jsx)("caption", {
          children: (0, _jsxRuntime.jsx)(_ScreenReaderContent.ScreenReaderContent, {
            children: captionText
          })
        }), _react.Children.map(children, child => /*#__PURE__*/(0, _react.isValidElement)(child) ? (0, _safeCloneElement.safeCloneElement)(child, {
          key: child.props.name
        }) : child)]
      })]
    });
  }
}) || _class);
var _default = exports.default = Table;
