// @instructure/ui-themes/src/themes/newThemeTokens/dark/components/badge.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { Badge } from "../../componentTypes/badge"
import type { Semantics } from "../semantics"

const badge = (semantics: Semantics): Badge => ({
  borderRadius: semantics.borderRadius.full,
  fontFamily: semantics.fontFamily.base,
  color: semantics.color.text.onColor,
  colorInverse: semantics.color.text.dark,
  colorDanger: semantics.color.background.error,
  colorSuccess: semantics.color.background.success,
  colorPrimary: semantics.color.background.info,
  fontSize: semantics.fontSize.textSm,
  fontWeight: semantics.fontWeight.body.strong,
  sizeNotification: "0.75rem",
  size: "1.25rem",
  padding: semantics.spacing.spaceXs,
  notificationOffset: "0.125rem",
  notificationZIndex: 1,
  pulseBorderThickness: "0.125rem",
  countOffset: "0.5rem",
})
export default badge
