// @instructure/ui-themes/src/themes/newThemeTokens/dark/components/baseButton.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { BaseButton } from "../../componentTypes/baseButton"
import type { Semantics } from "../semantics"

const baseButton = (semantics: Semantics): BaseButton => ({
  primaryActiveBackground: semantics.color.background.interactive.action.primary.active,
  primaryActiveBorderColor: semantics.color.stroke.interactive.action.primary.active,
  primaryActiveTextColor: semantics.color.text.interactive.action.primary.active,
  primaryBackground: semantics.color.background.interactive.action.primary.base,
  primaryBorderColor: semantics.color.stroke.interactive.action.primary.base,
  primaryColor: semantics.color.text.interactive.action.primary.base,
  primaryDisabledBackgroundColor: semantics.color.background.interactive.action.primary.disabled,
  primaryDisabledBorderColor: semantics.color.stroke.interactive.action.primary.disabled,
  primaryDisabledTextColor: semantics.color.text.interactive.action.primary.disabled,
  primaryGhostActiveBackground: semantics.color.background.interactive.action.tertiary.active,
  primaryGhostBorderColor: semantics.color.stroke.interactive.action.tertiary.base,
  primaryGhostColor: semantics.color.text.interactive.action.tertiary.base,
  primaryGhostHoverBackground: semantics.color.background.interactive.action.tertiary.hover,
  primaryHoverBackground: semantics.color.background.interactive.action.primary.hover,
  primaryHoverBorderColor: semantics.color.stroke.interactive.action.primary.hover,
  primaryHoverTextColor: semantics.color.text.interactive.action.primary.hover,
  primaryInverseActiveBackground: semantics.color.background.interactive.action.primaryOnColor.active,
  primaryInverseBackground: semantics.color.background.interactive.action.primaryOnColor.base,
  primaryInverseBorderColor: semantics.color.stroke.interactive.action.primaryOnColor.base,
  primaryInverseColor: semantics.color.text.interactive.action.primaryOnColor.base,
  primaryInverseGhostActiveBackground: semantics.color.background.interactive.action.ghost.onColor.hover,
  primaryInverseGhostBorderColor: semantics.color.stroke.interactive.action.secondaryOnColor.base,
  primaryInverseGhostColor: semantics.color.text.interactive.action.secondaryOnColor.base,
  primaryInverseGhostHoverBackground: semantics.color.background.interactive.action.ghost.onColor.hover,
  primaryInverseHoverBackground: semantics.color.background.interactive.action.primaryOnColor.hover,
  primaryOnColorActiveBorderColor: semantics.color.stroke.interactive.action.primaryOnColor.active,
  primaryOnColorActiveTextColor: semantics.color.text.interactive.action.primaryOnColor.active,
  primaryOnColorDisabledBackgroundColor: semantics.color.background.interactive.action.primaryOnColor.disabled,
  primaryOnColorDisabledBorderColor: semantics.color.stroke.interactive.action.primaryOnColor.disabled,
  primaryOnColorDisabledTextColor: semantics.color.text.interactive.action.primaryOnColor.disabled,
  primaryOnColorHoverBorderColor: semantics.color.stroke.interactive.action.primaryOnColor.hover,
  primaryOnColorHoverTextColor: semantics.color.text.interactive.action.primaryOnColor.hover,
  secondaryActiveBackground: semantics.color.background.interactive.action.secondary.active,
  secondaryActiveBorderColor: semantics.color.stroke.interactive.action.secondary.active,
  secondaryActiveTextColor: semantics.color.text.interactive.action.secondary.active,
  secondaryBackground: semantics.color.background.interactive.action.secondary.base,
  secondaryBorderColor: semantics.color.stroke.interactive.action.secondary.base,
  secondaryColor: semantics.color.text.interactive.action.secondary.base,
  secondaryDisabledBackgroundColor: semantics.color.background.interactive.action.secondary.disabled,
  secondaryDisabledBorderColor: semantics.color.stroke.interactive.action.secondary.disabled,
  secondaryDisabledTextColor: semantics.color.text.interactive.action.secondary.disabled,
  secondaryGhostBorderColor: semantics.color.stroke.interactive.action.secondary.base,
  secondaryGhostColor: semantics.color.text.interactive.action.secondary.base,
  secondaryGhostHoverBackground: semantics.color.background.interactive.action.secondary.hover,
  secondaryHoverBackground: semantics.color.background.interactive.action.secondary.hover,
  secondaryHoverBorderColor: semantics.color.stroke.interactive.action.secondary.hover,
  secondaryHoverTextColor: semantics.color.text.interactive.action.secondary.hover,
  secondaryOnColorActiveBorderColor: semantics.color.stroke.interactive.action.secondaryOnColor.active,
  secondaryOnColorActiveTextColor: semantics.color.text.interactive.action.secondaryOnColor.active,
  secondaryOnColorDisabledBorderColor: semantics.color.stroke.interactive.action.secondaryOnColor.disabled,
  secondaryOnColorDisabledTextColor: semantics.color.text.interactive.action.secondaryOnColor.disabled,
  secondaryOnColorHoverBorderColor: semantics.color.stroke.interactive.action.secondaryOnColor.hover,
  secondaryOnColorHoverTextColor: semantics.color.text.interactive.action.secondaryOnColor.hover,
  successActiveBackground: semantics.color.background.interactive.action.success.active,
  successActiveBorderColor: semantics.color.stroke.interactive.action.success.active,
  successActiveTextColor: semantics.color.text.interactive.action.status.active,
  successBackground: semantics.color.background.interactive.action.success.base,
  successBorderColor: semantics.color.stroke.interactive.action.success.base,
  successColor: semantics.color.text.interactive.action.status.base,
  successDisabledBackgroundColor: semantics.color.background.interactive.action.success.disabled,
  successDisabledBorderColor: semantics.color.stroke.interactive.action.success.disabled,
  successDisabledTextColor: semantics.color.text.interactive.action.status.disabled,
  successGhostActiveBackground: semantics.color.background.interactive.action.success.secondary.active,
  successGhostBorderColor: semantics.color.stroke.interactive.action.success.secondary.base,
  successGhostColor: semantics.color.text.interactive.action.successSecondary.base,
  successGhostHoverBackground: semantics.color.background.interactive.action.success.secondary.hover,
  successHoverBackground: semantics.color.background.interactive.action.success.hover,
  successHoverBorderColor: semantics.color.stroke.interactive.action.success.hover,
  successHoverTextColor: semantics.color.text.interactive.action.status.hover,
  successSecondaryActiveBorderColor: semantics.color.stroke.interactive.action.success.secondary.active,
  successSecondaryActiveTextColor: semantics.color.text.interactive.action.successSecondary.active,
  successSecondaryDisabledBorderColor: semantics.color.stroke.interactive.action.success.secondary.disabled,
  successSecondaryDisabledTextColor: semantics.color.text.interactive.action.successSecondary.disabled,
  successSecondaryHoverBorderColor: semantics.color.stroke.interactive.action.success.secondary.hover,
  successSecondaryHoverTextColor: semantics.color.text.interactive.action.successSecondary.hover,
  dangerActiveBackground: semantics.color.background.interactive.action.destructive.active,
  dangerBorderColor: semantics.color.stroke.interactive.action.destructive.base,
  dangerGhostActiveBackground: semantics.color.background.interactive.action.destructive.secondary.active,
  dangerGhostBorderColor: semantics.color.stroke.interactive.action.destructive.secondary.base,
  dangerGhostColor: semantics.color.text.interactive.action.destructiveSecondary.base,
  dangerGhostHoverBackground: semantics.color.background.interactive.action.destructive.secondary.hover,
  dangerHoverBackground: semantics.color.background.interactive.action.destructive.hover,
  destructiveActiveBorderColor: semantics.color.stroke.interactive.action.destructive.active,
  destructiveActiveTextColor: semantics.color.text.interactive.action.status.active,
  dangerBackground: semantics.color.background.interactive.action.destructive.base,
  dangerColor: semantics.color.text.interactive.action.status.base,
  destructiveDisabledBackgroundColor: semantics.color.background.interactive.action.destructive.disabled,
  destructiveDisabledBorderColor: semantics.color.stroke.interactive.action.destructive.disabled,
  destructiveDisabledTextColor: semantics.color.text.interactive.action.status.disabled,
  destructiveHoverBorderColor: semantics.color.stroke.interactive.action.destructive.hover,
  destructiveHoverTextColor: semantics.color.text.interactive.action.status.hover,
  destructiveSecondaryActiveBorderColor: semantics.color.stroke.interactive.action.destructive.secondary.active,
  destructiveSecondaryActiveTextColor: semantics.color.text.interactive.action.destructiveSecondary.active,
  destructiveSecondaryDisabledBorderColor: semantics.color.stroke.interactive.action.destructive.secondary.disabled,
  destructiveSecondaryDisabledTextColor: semantics.color.text.interactive.action.destructiveSecondary.disabled,
  destructiveSecondaryHoverBorderColor: semantics.color.stroke.interactive.action.destructive.secondary.hover,
  destructiveSecondaryHoverTextColor: semantics.color.text.interactive.action.destructiveSecondary.hover,
  aiActiveBackgroundBottomGradientColor: semantics.color.background.interactive.action.ai.bottomGradient.active,
  aiActiveBackgroundTopGradientColor: semantics.color.background.interactive.action.ai.topGradient.active,
  aiActiveBorderBottomGradientColor: semantics.color.stroke.interactive.action.ai.bottomGradient.active,
  aiActiveBorderTopGradientColor: semantics.color.stroke.interactive.action.ai.topGradient.active,
  aiActiveTextColor: semantics.color.text.interactive.action.ai.active,
  aiBackgroundBottomGradientColor: semantics.color.background.interactive.action.ai.bottomGradient.base,
  aiBackgroundTopGradientColor: semantics.color.background.interactive.action.ai.topGradient.base,
  aiBaseTextColor: semantics.color.text.interactive.action.ai.base,
  aiBorderBottomGradientColor: semantics.color.stroke.interactive.action.ai.bottomGradient.base,
  aiBorderTopGradientColor: semantics.color.stroke.interactive.action.ai.topGradient.base,
  aiDisabledBackgroundColor: semantics.color.background.interactive.action.ai.disabled,
  aiDisabledBorderColor: semantics.color.stroke.interactive.action.ai.disabled,
  aiDisabledTextColor: semantics.color.text.interactive.action.ai.disabled,
  aiHoverBackgroundBottomGradientColor: semantics.color.background.interactive.action.ai.bottomGradient.hover,
  aiHoverBackgroundTopGradientColor: semantics.color.background.interactive.action.ai.topGradient.hover,
  aiHoverBorderBottomGradientColor: semantics.color.stroke.interactive.action.ai.bottomGradient.hover,
  aiHoverBorderTopGradientColor: semantics.color.stroke.interactive.action.ai.topGradient.hover,
  aiHoverTextColor: semantics.color.text.interactive.action.ai.hover,
  aiSecondaryActiveBackgroundBottomGradientColor:
    semantics.color.background.interactive.action.aiSecondary.active.bottomGradient,
  aiSecondaryActiveBackgroundTopGradientColor:
    semantics.color.background.interactive.action.aiSecondary.active.topGradient,
  aiSecondaryDisabledBorderColor: semantics.color.stroke.interactive.action.aiSecondary.disabled,
  aiSecondaryDisabledTextColor: semantics.color.text.interactive.action.aiSecondary.disabled,
  aiSecondaryHoverBackgroundBottomGradientColor:
    semantics.color.background.interactive.action.aiSecondary.hover.bottomGradient,
  aiSecondaryHoverBackgroundTopGradientColor:
    semantics.color.background.interactive.action.aiSecondary.hover.topGradient,
  aiSecondaryTextBottomGradientColor: semantics.color.text.interactive.action.aiSecondary.bottomGradient.base,
  aiSecondaryTextTopGradientColor: semantics.color.text.interactive.action.aiSecondary.topGradient.base,
  smallFontSize: semantics.fontSize.textSm,
  mediumFontSize: semantics.fontSize.textBase,
  largeFontSize: "1.125rem",
  largeHeight: semantics.size.interactive.height.lg,
  largePaddingHorizontal: semantics.spacing.padding.interactive.horizontal.xl,
  mediumHeight: semantics.size.interactive.height.md,
  mediumPaddingHorizontal: semantics.spacing.padding.interactive.horizontal.lg,
  smallHeight: semantics.size.interactive.height.sm,
  smallPaddingHorizontal: semantics.spacing.padding.interactive.horizontal.md,
  borderRadius: semantics.borderRadius.interactive.base,
  borderRadiusFull: semantics.borderRadius.full,
  borderRadiusSm: semantics.borderRadius.sm,
  borderWidth: semantics.borderWidth.interactive.base,
  fontFamily: semantics.fontFamily.base,
  fontWeight: semantics.fontWeight.body.strong,
  gapButtonContentLg: semantics.spacing.spaceMd,
  gapButtonContentMd: semantics.spacing.spaceSm,
  gapButtonContentSm: semantics.spacing.spaceXs,
  heightXs: semantics.size.interactive.height.xs,
  heightXxs: semantics.size.interactive.height.xxs,
  opacityBase: semantics.opacity.base,
  opacityDisabled: semantics.opacity.base,
  tertiaryActiveBorderColor: semantics.color.stroke.interactive.action.tertiary.active,
  tertiaryActiveTextColor: semantics.color.text.interactive.action.tertiary.active,
  tertiaryDisabledBorderColor: semantics.color.stroke.interactive.action.tertiary.disabled,
  tertiaryDisabledTextColor: semantics.color.text.interactive.action.tertiary.disabled,
  tertiaryHoverBorderColor: semantics.color.stroke.interactive.action.tertiary.hover,
  tertiaryHoverTextColor: semantics.color.text.interactive.action.tertiary.hover,
  dangerBoxShadow: "none",
  dangerGhostBackground: "transparent",
  dangerGhostBoxShadow: "none",
  dangerGhostHoverBoxShadow: "none",
  dangerHoverBoxShadow: "none",
  hoverTransform: "none",
  letterSpacing: "normal",
  primaryBoxShadow: "none",
  primaryGhostBackground: "transparent",
  primaryHoverBoxShadow: "none",
  primaryInverseBoxShadow: "none",
  primaryInverseGhostBackground: "transparent",
  primaryInverseGhostBoxShadow: "none",
  primaryInverseGhostHoverBoxShadow: "none",
  primaryInverseHoverBoxShadow: "none",
  secondaryBoxShadow: "none",
  secondaryGhostActiveBackground: "transparent",
  secondaryGhostBackground: "transparent",
  secondaryGhostBoxShadow: "none",
  secondaryGhostHoverBoxShadow: "none",
  secondaryHoverBoxShadow: "none",
  successBoxShadow: "none",
  successGhostBackground: "transparent",
  successGhostBoxShadow: "none",
  successGhostHoverBoxShadow: "none",
  successHoverBoxShadow: "none",
  textTransform: "none",
  transform: "none",
  borderStyle: "solid",
  paddingVertical: "0.25rem",
  lineHeight: semantics.lineHeight.standalone.base,
})
export default baseButton
