// @instructure/ui-themes/src/themes/newThemeTokens/dark/components/sideNavBarItem.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { SideNavBarItem } from "../../componentTypes/sideNavBarItem"
import type { Semantics } from "../semantics"

const sideNavBarItem = (semantics: Semantics): SideNavBarItem => ({
  iconWrapperHeight: "2.25rem",
  borderRadius: semantics.borderRadius.lg,
  backgroundColor: "transparent",
  contentPadding: "0.375rem",
  fontColor: semantics.color.text.base,
  fontFamily: semantics.fontFamily.base,
  fontSize: semantics.fontSize.textSm,
  fontWeight: semantics.fontWeight.body.base,
  hoverBackgroundColor: semantics.color.background.interactive.action.tertiary.hover,
  lineHeight: semantics.lineHeight.paragraph.base,
  linkTextDecoration: "none",
  selectedBackgroundColor: semantics.color.background.brand,
  selectedFontColor: semantics.color.text.inverse,
  innerFocusOutlineColor: semantics.color.stroke.interactive.focusRing.base,
  selectedInnerFocusOutlineColor: semantics.color.icon.inverse,
  outerFocusOutlineColor: semantics.color.transparentColor,
  selectedOuterFocusOutlineColor: semantics.color.transparentColor,
})
export default sideNavBarItem
