// @instructure/ui-themes/src/themes/newThemeTokens/dark/components/sideNavBar.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { SideNavBar } from "../../componentTypes/sideNavBar"
import type { Semantics } from "../semantics"

const sideNavBar = (semantics: Semantics): SideNavBar => ({
  backgroundColor: semantics.color.background.elevatedSurface.base,
  fill: semantics.color.icon.interactive.action.primaryOnColor.base,
  fontColor: semantics.color.text.base,
  minimizedWidth: "3.75rem",
  toggleTransition: "300ms",
  width: "5.875rem",
  margin: semantics.spacing.spaceMd,
  minimizedMargin: "0.375rem",
  contentMargin: "0.375rem",
  contentGap: semantics.spacing.spaceXs,
  boxShadow: {
    0: {
      x: "0rem",
      y: "0.0625rem",
      blur: "0.125rem",
      spread: "0rem",
      color: semantics.color.dropShadow.shadowColor1,
      type: "dropShadow",
    },
    1: {
      x: "0rem",
      y: "0.125rem",
      blur: "0.25rem",
      spread: "0.125rem",
      color: semantics.color.dropShadow.shadowColor2,
      type: "dropShadow",
    },
  },
  borderRadius: semantics.borderRadius.container.md,
  minimizedBorderRadius: semantics.borderRadius.container.sm,
})
export default sideNavBar
