// @instructure/ui-themes/src/themes/newThemeTokens/dark/components/text.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { Text } from "../../componentTypes/text"
import type { Semantics } from "../semantics"

const text = (semantics: Semantics): Text => ({
  descriptionPage: {
    fontFamily: semantics.fontFamily.base,
    fontWeight: semantics.fontWeight.body.base,
    fontSize: semantics.fontSize.textLg,
    lineHeight: semantics.lineHeight.paragraph.base,
  },
  descriptionSection: {
    fontFamily: semantics.fontFamily.base,
    fontWeight: semantics.fontWeight.body.base,
    fontSize: semantics.fontSize.textBase,
    lineHeight: semantics.lineHeight.paragraph.base,
  },
  content: {
    fontFamily: semantics.fontFamily.base,
    fontWeight: semantics.fontWeight.body.base,
    fontSize: semantics.fontSize.textBase,
    lineHeight: semantics.lineHeight.paragraph.base,
  },
  contentImportant: {
    fontFamily: semantics.fontFamily.base,
    fontWeight: semantics.fontWeight.body.strong,
    fontSize: semantics.fontSize.textBase,
    lineHeight: semantics.lineHeight.paragraph.base,
  },
  contentQuote: {
    fontFamily: semantics.fontFamily.base,
    fontWeight: "Medium Italic",
    fontSize: semantics.fontSize.textBase,
    lineHeight: semantics.lineHeight.paragraph.base,
  },
  contentSmall: {
    fontFamily: semantics.fontFamily.base,
    fontWeight: semantics.fontWeight.body.base,
    fontSize: semantics.fontSize.textSm,
    lineHeight: semantics.lineHeight.paragraph.base,
  },
  legend: {
    fontFamily: semantics.fontFamily.base,
    fontWeight: semantics.fontWeight.body.base,
    fontSize: semantics.fontSize.textXs,
    lineHeight: semantics.lineHeight.paragraph.base,
  },
  fontSizeXSmall: semantics.fontSize.textXs,
  fontSizeSmall: semantics.fontSize.textSm,
  fontSizeMedium: semantics.fontSize.textBase,
  fontSizeLarge: semantics.fontSize.textXl,
  fontSizeXLarge: semantics.fontSize.text2xl,
  fontSizeXXLarge: "2.375rem",
  fontWeightLight: 300,
  fontWeightNormal: semantics.fontWeight.body.base,
  fontWeightBold: semantics.fontWeight.body.strong,
  fontWeightImportant: semantics.fontWeight.body.strong,
  fontWeightRegular: semantics.fontWeight.body.base,
  lineHeight: "150%",
  lineHeight100: "100%",
  lineHeight125: "125%",
  lineHeight150: "150%",
  lineHeightCondensed: "125%",
  lineHeightDouble: "200%",
  lineHeightFit: "112.5%",
  paragraphMargin: "1.5rem 0rem",
  letterSpacingCondensed: "-0.0625rem",
  letterSpacingExpanded: "0.0625rem",
  letterSpacingNormal: "0rem",
  aiBackgroundColor: semantics.color.background.aiText,
  aiColor: semantics.color.text.aiColor,
  primaryColor: semantics.color.text.interactive.navigation.primary.base,
  errorColor: semantics.color.text.error,
  baseColor: semantics.color.text.base,
  baseOnColor: semantics.color.text.onColor,
  inverseColor: semantics.color.text.inverse,
  mutedColor: semantics.color.text.muted,
  mutedOnColor: semantics.color.text.onColor,
  successColor: semantics.color.text.success,
  warningColor: semantics.color.text.warning,
  fontFamily: semantics.fontFamily.base,
  fontFamilyMonospace: semantics.fontFamily.code,
})
export default text
