// @instructure/ui-top-nav-bar/lib/TopNavBar/v2/utils/exampleHelpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.itemSubmenuExample = exports.getUser = exports.getMenuItems = exports.getLayoutProps = exports.getInPlaceDialogConfig = exports.getCustomPopoverContent = exports.getCustomPopoverConfig = exports.getBreadcrumb = exports.getBrand = exports.getActionItems = exports.avatarExample = exports.SmallViewportModeWrapper = void 0;
var _latest = require("@instructure/ui-drilldown/latest");
var _uiIcons = require("@instructure/ui-icons");
var _TopNavBarContext = require("../TopNavBarContext");
var _index = require("../index");
var _exampleSvgFiles = require("./exampleSvgFiles");
var _latest2 = require("@instructure/ui-breadcrumb/latest");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const avatarExample = exports.avatarExample = {
  avatarName: 'User Name',
  avatarSrc: 'https://images.pexels.com/photos/220453/pexels-photo-220453.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1'
};
const itemSubmenuExample = exports.itemSubmenuExample = (0, _jsxRuntime.jsx)(_latest.Drilldown, {
  shouldSetAriaExpanded: false,
  rootPageId: "root",
  children: (0, _jsxRuntime.jsxs)(_latest.Drilldown.Page, {
    id: "root",
    children: [(0, _jsxRuntime.jsx)(_latest.Drilldown.Option, {
      id: "linkExampleOption1",
      href: "/#One",
      children: "Link One"
    }), (0, _jsxRuntime.jsx)(_latest.Drilldown.Option, {
      id: "linkExampleOption2",
      href: "/#Two",
      children: "Link Two"
    }), (0, _jsxRuntime.jsx)(_latest.Drilldown.Option, {
      id: "linkExampleOption3",
      href: "/#Three",
      children: "Link Three"
    })]
  })
});
const getCustomPopoverConfig = (inverseColor, extraProps = {}) => ({
  children: getCustomPopoverContent(!inverseColor),
  color: inverseColor ? 'primary-inverse' : 'primary',
  isShowingContent: true,
  placement: 'bottom center',
  ...extraProps
});
exports.getCustomPopoverConfig = getCustomPopoverConfig;
const getCustomPopoverContent = inverseColor => (0, _jsxRuntime.jsx)("div", {
  style: {
    color: inverseColor ? 'black' : 'white',
    padding: '0 0.5rem'
  },
  children: "dialog content"
});
exports.getCustomPopoverContent = getCustomPopoverContent;
const getInPlaceDialogConfig = (inverseColor, extraProps = {}) => ({
  open: true,
  onClose: () => {},
  closeButtonLabel: 'Close Search and return to Navigation',
  returnFocusElement: () => document.getElementById('Search'),
  shouldCloseOnEscape: true,
  shouldCloseOnDocumentClick: false,
  shouldContainFocus: false,
  content: () => getCustomPopoverContent(inverseColor),
  ...extraProps
});
exports.getInPlaceDialogConfig = getInPlaceDialogConfig;
const SmallViewportModeWrapper = props => {
  return (0, _jsxRuntime.jsx)(_TopNavBarContext.TopNavBarContext.Provider, {
    value: {
      layout: props.layout || 'smallViewport',
      inverseColor: typeof props.inverseColor === 'undefined' ? false : props.inverseColor
    },
    children: props.children
  });
};
exports.SmallViewportModeWrapper = SmallViewportModeWrapper;
SmallViewportModeWrapper.defaultProps = {
  layout: 'smallViewport',
  inverseColor: false
};
const getBrand = (config = {}) => {
  return (0, _jsxRuntime.jsx)(_index.TopNavBar.Brand, {
    screenReaderLabel: "Brand name",
    renderIcon: _exampleSvgFiles.elevateIcon,
    iconBackground: "#0097D3",
    href: "/#TopNavBar",
    ...config.brandProps
  });
};
exports.getBrand = getBrand;
const getMenuItems = (config = {}) => {
  const items = config.menuItemsCustomIdList || ['Overview', 'Admin', 'Settings', 'Maps', 'Assessments', 'Community'];
  const count = typeof config.menuItemsCount !== 'undefined' ? config.menuItemsCount : 6;
  const itemsSubset = items.slice(0, count);
  return (0, _jsxRuntime.jsx)(_index.TopNavBar.MenuItems, {
    listLabel: "Page navigation",
    currentPageId: config.currentPageId,
    renderHiddenItemsMenuTriggerLabel: hiddenChildrenCount => `${hiddenChildrenCount} More`,
    ...config.menuItemsProps,
    children: itemsSubset.map((item, idx) => {
      const itemProps = typeof config.menuItemsItemProps === 'function' ? config.menuItemsItemProps(item, idx) : config.menuItemsItemProps;
      return (0, _jsxRuntime.jsx)(_index.TopNavBar.Item, {
        id: item,
        href: config.menuItemsWithSubmenu ? undefined : '/#TopNavBar',
        renderSubmenu: config.menuItemsWithSubmenu ? itemSubmenuExample : undefined,
        ...itemProps,
        children: itemProps?.children || item
      }, item);
    })
  });
};
exports.getMenuItems = getMenuItems;
const getActionItems = (config = {}) => {
  const items = [{
    label: 'Search',
    icon: _uiIcons.SearchInstUIIcon
  }, {
    label: 'Info',
    icon: _uiIcons.CircleQuestionMarkInstUIIcon
  }, {
    label: 'Alerts',
    icon: _uiIcons.BellInstUIIcon
  }, {
    label: 'Discussions',
    icon: _uiIcons.MessagesSquareInstUIIcon
  }, {
    label: 'Dashboard',
    icon: _uiIcons.GaugeInstUIIcon
  }, {
    label: 'Settings',
    icon: _uiIcons.Settings2InstUIIcon
  }];
  const count = typeof config.actionItemsCount !== 'undefined' ? config.actionItemsCount : 3;
  const itemsSubset = items.slice(0, count);
  return (0, _jsxRuntime.jsx)(_index.TopNavBar.ActionItems, {
    ...config.actionItemsProps,
    listLabel: config.actionItemsProps?.listLabel || 'Actions',
    renderHiddenItemsMenuTriggerLabel: config.actionItemsProps?.renderHiddenItemsMenuTriggerLabel || (hiddenChildrenCount => `${hiddenChildrenCount} more actions`),
    children: itemsSubset.map((item, idx) => {
      const itemProps = typeof config.actionItemsItemProps === 'function' ? config.actionItemsItemProps(item.label, idx) : config.actionItemsItemProps;
      return (0, _jsxRuntime.jsx)(_index.TopNavBar.Item, {
        id: item.label,
        renderIcon: item.icon,
        variant: 'icon',
        href: "/#TopNavBar",
        customPopoverConfig: config.hasRenderInPlaceDialogConfig && item.label === 'Search' ? getCustomPopoverConfig(config.inverseColor) : undefined,
        ...itemProps,
        children: itemProps?.children || item.label
      }, item.label);
    })
  });
};
exports.getActionItems = getActionItems;
const getUser = (config = {}) => {
  return (0, _jsxRuntime.jsx)(_index.TopNavBar.User, {
    ...config.userProps,
    children: config.userProps && 'children' in config.userProps ? config.userProps?.children : (0, _jsxRuntime.jsx)(_index.TopNavBar.Item, {
      id: config.userId || 'User',
      href: config.userWithSubmenu ? undefined : '/#TopNavBar',
      variant: config.userVariant || 'default',
      renderAvatar: config.userWithAvatar ? avatarExample : undefined,
      renderSubmenu: config.userWithSubmenu ? itemSubmenuExample : undefined,
      ...config.userItemProps,
      children: config.userItemProps?.children || (config.userVariant === 'button' ? 'Log In/Register' : 'User Name')
    })
  });
};
exports.getUser = getUser;
const getBreadcrumb = (_config = {}) => {
  return (0, _jsxRuntime.jsx)(_index.TopNavBar.Breadcrumb, {
    children: (0, _jsxRuntime.jsxs)(_latest2.Breadcrumb, {
      label: "You are here",
      children: [(0, _jsxRuntime.jsx)(_latest2.Breadcrumb.Link, {
        children: "Course page 1"
      }), (0, _jsxRuntime.jsx)(_latest2.Breadcrumb.Link, {
        children: "Course page 2"
      }), (0, _jsxRuntime.jsx)(_latest2.Breadcrumb.Link, {
        children: "Course page 3"
      })]
    })
  });
};
exports.getBreadcrumb = getBreadcrumb;
const getLayoutProps = (props, config = {}) => {
  return {
    smallViewportConfig: {
      dropdownMenuToggleButtonLabel: 'Toggle Menu',
      dropdownMenuLabel: 'Main Menu',
      alternativeTitle: config.hasAlternativeTitle ? 'Page title' : undefined,
      renderInPlaceDialogConfig: config.hasRenderInPlaceDialogConfig ? getInPlaceDialogConfig(props.inverseColor) : undefined
    },
    renderBrand: config.hasRenderBreadcrumb ? undefined : getBrand({
      ...props,
      ...config
    }),
    renderMenuItems: config.hasRenderBreadcrumb ? undefined : getMenuItems({
      ...props,
      ...config
    }),
    renderActionItems: getActionItems({
      ...props,
      ...config
    }),
    renderUser: getUser({
      ...props,
      ...config
    }),
    renderBreadcrumb: config.hasRenderBreadcrumb ? getBreadcrumb({
      ...props,
      ...config
    }) : undefined
  };
};
exports.getLayoutProps = getLayoutProps;
