// @instructure/ui-modal/lib/Modal/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Modal = void 0;
Object.defineProperty(exports, "ModalBody", {
  enumerable: true,
  get: function () {
    return _ModalBody.ModalBody;
  }
});
Object.defineProperty(exports, "ModalFooter", {
  enumerable: true,
  get: function () {
    return _ModalFooter.ModalFooter;
  }
});
Object.defineProperty(exports, "ModalHeader", {
  enumerable: true,
  get: function () {
    return _ModalHeader.ModalHeader;
  }
});
exports.default = void 0;
var _react = require("react");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _createChainedFunction = require("@instructure/ui-utils/lib/createChainedFunction.js");
var _Transition = require("@instructure/ui-motion/lib/Transition");
var _Portal = require("@instructure/ui-portal/lib/Portal");
var _Dialog = require("@instructure/ui-dialog/lib/Dialog");
var _latest = require("@instructure/ui-overlays/latest");
var _ModalHeader = require("./ModalHeader");
var _ModalBody = require("./ModalBody");
var _ModalFooter = require("./ModalFooter");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _ModalContext = _interopRequireDefault(require("./ModalContext"));
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
tags: overlay, portal, dialog
---
**/
let Modal = exports.Modal = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class Modal extends _react.Component {
  static displayName = "Modal";
  static componentId = 'Modal';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    open: false,
    size: 'auto',
    variant: 'default',
    transition: 'fade',
    insertAt: 'bottom',
    shouldCloseOnDocumentClick: true,
    shouldReturnFocus: true,
    constrain: 'window',
    overflow: 'scroll'
  };
  static Header = _ModalHeader.ModalHeader;
  static Body = _ModalBody.ModalBody;
  static Footer = _ModalFooter.ModalFooter;
  constructor(props) {
    super(props);
    this.state = {
      transitioning: false,
      open: props.open ?? false,
      windowHeight: 99999,
      bodyScrollAriaLabel: undefined
    };
  }
  _DOMNode = null;
  _content = null;
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this.props.makeStyles?.();
    window.addEventListener('resize', this.updateHeight);
  }
  componentDidUpdate(prevProps) {
    if (this.props.open !== prevProps.open) {
      this.setState({
        transitioning: true,
        open: !!this.props.open
      });
    }
    this.props.makeStyles?.();
  }
  componentWillUnmount() {
    window.removeEventListener('resize', this.updateHeight);
  }
  updateHeight = () => {
    this.setState({
      windowHeight: window.innerHeight
    });
  };
  get defaultFocusElement() {
    return this.props.defaultFocusElement;
  }
  get DOMNode() {
    return this._DOMNode;
  }
  set DOMNode(el) {
    this._DOMNode = el;
  }
  get maskPlacement() {
    if (this.props.overflow === 'fit') {
      return 'stretch';
    } else {
      return 'center';
    }
  }
  handlePortalOpen = DOMNode => {
    this.DOMNode = DOMNode;
  };
  handleTransitionComplete = () => {
    this.setState({
      transitioning: false
    });
  };
  contentRef = el => {
    this._content = el;
    if (typeof this.props.contentRef === 'function') {
      this.props.contentRef(el);
    }
  };
  renderChildren() {
    const {
      children,
      variant,
      overflow,
      styles
    } = this.props;
    const childrenArray = _react.Children.toArray(children);
    const headerAndBody = [];
    const others = [];
    childrenArray.forEach(child => {
      if (/*#__PURE__*/(0, _react.isValidElement)(child)) {
        if (child.type === Modal.Header || child.type === Modal.Body) {
          headerAndBody.push(this.cloneChildWithProps(child, variant, overflow));
        } else {
          others.push(this.cloneChildWithProps(child, variant, overflow));
        }
      }
    });

    // Putting ModalHeader and ModalBody into the same container, so they can be styled together;
    // this is needed to apply a media query breakpoint
    const wrappedHeaderAndBody = headerAndBody.length > 0 ? (0, _jsxRuntime.jsx)("div", {
      css: styles?.joinedHeaderAndBody,
      children: headerAndBody
    }, "header-and-body") : null;
    return [...(wrappedHeaderAndBody ? [wrappedHeaderAndBody] : []), ...others];
  }
  cloneChildWithProps(child, variant, overflow) {
    if (/*#__PURE__*/(0, _react.isValidElement)(child)) {
      return (0, _safeCloneElement.safeCloneElement)(child, {
        variant: variant,
        overflow: child?.props?.overflow || overflow
      });
    } else {
      return child;
    }
  }
  renderDialog(props, open) {
    const {
      onDismiss,
      label,
      shouldCloseOnDocumentClick,
      shouldReturnFocus,
      liveRegion,
      size,
      constrain,
      as,
      styles
    } = this.props;
    const isFullScreen = size === 'fullscreen';
    const dialog = (0, _jsxRuntime.jsx)(_Dialog.Dialog, {
      ...(0, _passthroughProps.passthroughProps)(props),
      as: as,
      open: open && !this.state.transitioning,
      label: label,
      defaultFocusElement: this.defaultFocusElement,
      shouldCloseOnDocumentClick: shouldCloseOnDocumentClick,
      shouldCloseOnEscape: true,
      shouldContainFocus: true,
      shouldReturnFocus: shouldReturnFocus,
      liveRegion: liveRegion,
      onDismiss: onDismiss,
      css: styles?.modal,
      ref: this.contentRef
      // aria-modal="true" see VO bug https://bugs.webkit.org/show_bug.cgi?id=174667
      ,
      children: this.renderChildren()
    });
    return (0, _jsxRuntime.jsx)(_latest.Mask, {
      placement: this.maskPlacement,
      fullscreen: constrain === 'window',
      themeOverride: isFullScreen ? {
        borderRadius: '0em',
        borderWidth: '0em'
      } : {},
      children: dialog
    });
  }
  render() {
    const {
      open,
      onOpen,
      onClose,
      mountNode,
      insertAt,
      transition,
      onEnter,
      onEntering,
      onEntered,
      onExit,
      onExiting,
      onExited,
      constrain,
      overflow,
      // TODO this is not used currently
      ...passthroughProps
    } = this.props;
    const portalIsOpen = this.state.open || this.state.transitioning;
    return (0, _jsxRuntime.jsx)(_Portal.Portal, {
      mountNode: mountNode,
      insertAt: insertAt,
      open: portalIsOpen,
      onOpen: this.handlePortalOpen,
      "data-cid": "Modal",
      children: (0, _jsxRuntime.jsx)(_Transition.Transition, {
        in: open,
        transitionOnMount: true,
        type: transition,
        onEnter: onEnter,
        onEntering: onEntering,
        onEntered: (0, _createChainedFunction.createChainedFunction)(this.handleTransitionComplete, onEntered, onOpen),
        onExit: onExit,
        onExiting: onExiting,
        onExited: (0, _createChainedFunction.createChainedFunction)(this.handleTransitionComplete, onExited, onClose),
        children: (0, _jsxRuntime.jsx)(_ModalContext.default.Provider, {
          value: {
            bodyScrollAriaLabel: this.state.bodyScrollAriaLabel,
            setBodyScrollAriaLabel: txt => this.setState({
              bodyScrollAriaLabel: txt
            })
          },
          children: constrain === 'parent' ? (0, _jsxRuntime.jsx)("span", {
            css: this.props.styles?.constrainContext,
            children: this.renderDialog(passthroughProps, open)
          }) : this.renderDialog(passthroughProps, open)
        })
      })
    });
  }
}) || _class);
var _default = exports.default = Modal;
