// @instructure/ui-menu/lib/Menu/v1/MenuItemGroup/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.MenuItemGroup = void 0;
var _react = require("react");
var _emotion = require("@instructure/emotion");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _hasVisibleChildren = require("@instructure/ui-a11y-utils/lib/hasVisibleChildren.js");
var _MenuItem = require("../MenuItem");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: Menu
id: Menu.Group
---
**/
let MenuItemGroup = exports.MenuItemGroup = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = _dec2(_class = class MenuItemGroup extends _react.Component {
  static displayName = "MenuItemGroup";
  static componentId = 'Menu.Group';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    disabled: false,
    children: null,
    isTabbable: false,
    allowMultiple: false,
    defaultSelected: []
  };
  constructor(props) {
    super(props);
    if (typeof props.selected === 'undefined') {
      this.state = {
        selected: this.selectedFromChildren(props) || props.defaultSelected
      };
    }
  }
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  handleSelect = (e, value, selected, item) => {
    if (this.props.disabled) {
      e.preventDefault();
      return;
    }
    if (this.props.selected) {
      this.updateSelected(e, value, this.props.selected, selected, item);
    } else {
      this.setState(state => {
        return {
          selected: this.updateSelected(e, value, state.selected, selected, item)
        };
      });
    }
  };
  updateSelected = (e, value, items, selected, item) => {
    const {
      allowMultiple
    } = this.props;
    let updated = allowMultiple ? [...items] : [];
    const location = updated.indexOf(value);
    if (selected === true && location < 0) {
      updated.push(value);
    } else if (selected === false && location !== -1) {
      updated.splice(location, 1);
    } else if (!allowMultiple && updated.length < 1) {
      // don't allow nothing selected if it's not allowMultiple/checkbox
      updated = [...items];
    }
    if (typeof this.props.onSelect === 'function') {
      this.props.onSelect(e, updated, selected, item);
    }
    return updated;
  };
  selectedFromChildren(props) {
    const {
      children,
      allowMultiple
    } = props;
    const selected = [];
    const items = _react.Children.toArray(children).filter(child => {
      return (0, _matchComponentTypes.matchComponentTypes)(child, [_MenuItem.MenuItem]);
    });
    items.forEach((item, index) => {
      if ((selected.length === 0 || allowMultiple) && (item.props.selected || item.props.defaultSelected)) {
        selected.push(item.props.value || index);
      }
    });
    return selected.length > 0 ? selected : null;
  }
  get selected() {
    if (typeof this.props.selected === 'undefined' && typeof this.state.selected === 'undefined') {
      return [];
    } else {
      return typeof this.props.selected === 'undefined' ? [...this.state.selected] : [...this.props.selected];
    }
  }
  renderLabel() {
    const {
      label
    } = this.props;
    return (0, _hasVisibleChildren.hasVisibleChildren)(label) ? (0, _jsxRuntime.jsx)("span", {
      css: this.props.styles?.label,
      children: label
    }) : label;
  }
  renderChildren() {
    const {
      disabled,
      controls,
      allowMultiple,
      isTabbable,
      onMouseOver
    } = this.props;
    const children = this.props.children;
    let index = -1;
    return _react.Children.map(children, child => {
      if ((0, _matchComponentTypes.matchComponentTypes)(child, [_MenuItem.MenuItem])) {
        ++index;
        const value = child.props.value || index;
        return (0, _safeCloneElement.safeCloneElement)(child, {
          tabIndex: isTabbable && index === 0 ? 0 : -1,
          controls,
          value,
          children: child.props.children,
          type: allowMultiple ? 'checkbox' : 'radio',
          ref: this.props.itemRef,
          disabled: disabled || child.props.disabled,
          selected: this.selected.indexOf(value) > -1,
          onSelect: this.handleSelect,
          onMouseOver
        });
      } else {
        return child;
      }
    });
  }
  render() {
    const props = (0, _omitProps.omitProps)(this.props, MenuItemGroup.allowedProps);
    return (0, _jsxRuntime.jsxs)("span", {
      ...props,
      css: this.props.styles?.menuItemGroup,
      ref: this.handleRef,
      "data-cid": "MenuItemGroup",
      children: [this.renderLabel(), (0, _jsxRuntime.jsx)("div", {
        css: this.props.styles?.items,
        "aria-disabled": this.props.disabled ? 'true' : undefined,
        children: this.renderChildren()
      })]
    });
  }
}) || _class) || _class);
var _default = exports.default = MenuItemGroup;
