// @instructure/ui-breadcrumb/es/Breadcrumb/v2/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { isValidElement, Children, Component } from 'react';
import { View } from '@instructure/ui-view/latest';
import { safeCloneElement } from '@instructure/ui-react-utils';
import { withStyleNew } from '@instructure/emotion';
import { ChevronRightInstUIIcon } from '@instructure/ui-icons';
import { BreadcrumbLink } from "./BreadcrumbLink/index.js";
import generateStyle from "./styles.js";
import { allowedProps } from "./props.js";
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
/**
---
category: components
---
**/
let Breadcrumb = (_dec = withStyleNew(generateStyle), _dec(_class = class Breadcrumb extends Component {
  static displayName = "Breadcrumb";
  static componentId = 'Breadcrumb';
  static allowedProps = allowedProps;
  static defaultProps = {
    size: 'medium',
    children: null
  };
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  addAriaCurrent = child => {
    const updatedChild = safeCloneElement(child, {
      'aria-current': 'page'
    });
    return updatedChild;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  static Link = BreadcrumbLink;
  renderChildren() {
    const {
      styles,
      children,
      size
    } = this.props;
    const numChildren = Children.count(children);
    const inlineStyle = {
      maxWidth: `${Math.floor(100 / numChildren)}%`
    };
    let isAriaCurrentSet = false;
    return Children.map(children, (child, index) => {
      const isLastElement = index === numChildren - 1;
      if (/*#__PURE__*/isValidElement(child)) {
        const isCurrentPage = child.props.isCurrentPage || false;
        if (isAriaCurrentSet && isCurrentPage) {}
        if (isCurrentPage) {
          isAriaCurrentSet = true;
        }
      }

      // Clone child (BreadcrumbLink) and pass size prop
      const childWithSize = /*#__PURE__*/isValidElement(child) ? safeCloneElement(child, {
        size
      }) : child;

      // Map Breadcrumb sizes to icon size tokens
      const breadcrumbSizeToIconSize = {
        small: 'sm',
        medium: 'md',
        large: 'lg'
      };
      const iconSize = breadcrumbSizeToIconSize[size || 'small'];
      return _jsxs(_Fragment, {
        children: [_jsx("li", {
          css: styles?.crumb,
          style: inlineStyle,
          children: !isAriaCurrentSet && isLastElement && child.props.isCurrentPage !== false ? this.addAriaCurrent(childWithSize) : childWithSize
        }), index < numChildren - 1 && _jsx("li", {
          "aria-hidden": "true",
          css: styles?.separator,
          children: _jsx(ChevronRightInstUIIcon, {
            size: iconSize,
            color: 'mutedColor'
          })
        })]
      });
    });
  }
  render() {
    const {
      styles
    } = this.props;
    return _jsx(View, {
      role: "navigation",
      as: "div",
      margin: this.props.margin,
      "aria-label": this.props.label,
      elementRef: this.handleRef,
      "data-cid": "Breadcrumb",
      children: _jsx("ol", {
        css: styles?.breadcrumb,
        children: this.renderChildren()
      })
    });
  }
}) || _class);
export default Breadcrumb;
export { Breadcrumb, BreadcrumbLink };
