// @instructure/ui-text/lib/Text/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Text = void 0;
var _react = require("react");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _getElementType = require("@instructure/ui-react-utils/lib/getElementType.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let Text = exports.Text = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class Text extends _react.Component {
  static displayName = "Text";
  static componentId = 'Text';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    as: 'span',
    wrap: 'normal',
    size: 'medium',
    letterSpacing: 'normal',
    children: null
  };
  ref = null;
  handleRef = el => {
    this.ref = el;
    const {
      elementRef
    } = this.props;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  checkProps() {
    const {
      variant,
      lineHeight,
      weight,
      fontStyle
    } = this.props;
    if (variant) {
      if (lineHeight) {}
      if (weight) {}
      if (fontStyle) {}
    }
  }
  componentDidMount() {
    this.props.makeStyles?.();
    this.checkProps();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
    this.checkProps();
  }
  render() {
    const {
      children
    } = this.props;
    const ElementType = (0, _getElementType.getElementType)(Text, this.props);
    return (0, _jsxRuntime.jsx)(ElementType, {
      ...(0, _passthroughProps.passthroughProps)(this.props),
      css: this.props.styles?.text,
      ref: this.handleRef,
      children: children
    });
  }
}) || _class);
var _default = exports.default = Text;
