// @instructure/ui-navigation/es/AppNav/v1/Item/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { logError as error } from '@instructure/console';
import { callRenderProp, getElementType, matchComponentTypes, passthroughProps } from '@instructure/ui-react-utils';
import { View } from '@instructure/ui-view/v11_6';
import { ScreenReaderContent } from '@instructure/ui-a11y-content';
import { withStyle } from '@instructure/emotion';
import generateStyle from "./styles.js";
import generateComponentTheme from "./theme.js";
import { allowedProps } from "./props.js";
/**
---
parent: AppNav
id: AppNav.Item
---
@module Item
**/
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
let Item = (_dec = withStyle(generateStyle, generateComponentTheme), _dec(_class = class Item extends Component {
  static displayName = "Item";
  static componentId = 'AppNav.Item';
  static allowedProps = allowedProps;
  static defaultProps = {
    children: null,
    isSelected: false,
    cursor: 'pointer',
    isDisabled: false
  };
  ref = null;
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  handleClick = e => {
    const {
      isDisabled,
      onClick
    } = this.props;
    if (isDisabled) {
      e.preventDefault();
      e.stopPropagation();
    } else if (typeof onClick === 'function') {
      onClick(e);
    }
  };
  render() {
    const ElementType = getElementType(Item, this.props);
    const {
      renderIcon,
      renderLabel,
      href,
      renderAfter,
      cursor,
      isDisabled
    } = this.props;
    const icon = callRenderProp(renderIcon);
    const label = callRenderProp(renderLabel);
    const labelIsForScreenReaders = matchComponentTypes(label, [ScreenReaderContent]);
    if (icon) {
      error(labelIsForScreenReaders, '[AppNav] If an icon is used, the label text should be wrapped in <ScreenReaderContent />.');
    }
    return _jsxs(View, {
      ...passthroughProps(this.props),
      as: ElementType,
      href: href,
      onClick: this.handleClick,
      disabled: isDisabled,
      elementRef: this.handleRef,
      display: "flex",
      position: "relative",
      borderRadius: "medium",
      cursor: isDisabled ? 'not-allowed' : cursor,
      css: this.props.styles?.item,
      "data-cid": "AppNav.Item",
      children: [icon, labelIsForScreenReaders ? label : _jsx("span", {
        css: this.props.styles?.label,
        children: label
      }), renderAfter && callRenderProp(renderAfter)]
    });
  }
}) || _class);
export default Item;
export { Item };
