// @instructure/ui-form-field/lib/FormFieldMessage/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.FormFieldMessage = void 0;
var _react = require("react");
var _ScreenReaderContent = require("@instructure/ui-a11y-content/lib/ScreenReaderContent");
var _IconWarningSolid = require("@instructure/ui-icons/lib/generated/IconWarningSolid.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: FormField
---

This is a helper component that is used by most of the custom form
components. In most cases it shouldn't be used directly.

```js
---
type: example
---
<FormFieldMessage variant="error">Invalid value</FormFieldMessage>
```
**/
let FormFieldMessage = exports.FormFieldMessage = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class FormFieldMessage extends _react.Component {
  static displayName = "FormFieldMessage";
  static componentId = 'FormFieldMessage';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    variant: 'hint'
  };
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  render() {
    const {
      children,
      styles
    } = this.props;
    return this.props.variant !== 'screenreader-only' ? (0, _jsxRuntime.jsxs)("span", {
      css: {
        display: 'flex'
      },
      children: [this.props.variant === 'newError' && this.props.children && (0, _jsxRuntime.jsx)("span", {
        css: styles?.errorIcon,
        children: (0, _jsxRuntime.jsx)(_IconWarningSolid.IconWarningSolid, {
          color: "error"
        })
      }), (0, _jsxRuntime.jsx)("span", {
        css: styles?.formFieldMessage,
        ref: this.handleRef,
        children: children
      })]
    }) : (0, _jsxRuntime.jsx)(_ScreenReaderContent.ScreenReaderContent, {
      elementRef: this.handleRef,
      children: children
    });
  }
}) || _class);
var _default = exports.default = FormFieldMessage;
