// @instructure/ui/lib/v11_7.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  InstUISettingsProvider: true,
  withStyleNew: true,
  AccessibleContent: true,
  PresentationContent: true,
  ScreenReaderContent: true,
  Alert: true,
  Avatar: true,
  Badge: true,
  Billboard: true,
  Breadcrumb: true,
  BreadcrumbLink: true,
  BaseButton: true,
  Button: true,
  CloseButton: true,
  CondensedButton: true,
  IconButton: true,
  ToggleButton: true,
  Byline: true,
  Calendar: true,
  CalendarDay: true,
  Checkbox: true,
  CheckboxGroup: true,
  CheckboxFacade: true,
  ToggleFacade: true,
  ColorPicker: true,
  ColorMixer: true,
  ColorPreset: true,
  ColorContrast: true,
  ColorIndicator: true,
  DateInput: true,
  DateInput2: true,
  DateTimeInput: true,
  Dialog: true,
  DrawerLayout: true,
  DrawerContent: true,
  DrawerTray: true,
  Drilldown: true,
  DrilldownGroup: true,
  DrilldownOption: true,
  DrilldownPage: true,
  DrilldownSeparator: true,
  Editable: true,
  InPlaceEdit: true,
  Expandable: true,
  FileDrop: true,
  Flex: true,
  FlexItem: true,
  Focusable: true,
  FormField: true,
  FormFieldLabel: true,
  FormFieldMessage: true,
  FormFieldMessages: true,
  FormFieldLayout: true,
  FormFieldGroup: true,
  Grid: true,
  GridRow: true,
  GridCol: true,
  Heading: true,
  ApplyLocale: true,
  TextDirectionContext: true,
  ApplyLocaleContext: true,
  textDirectionContextConsumer: true,
  DateTime: true,
  getTextDirection: true,
  Locale: true,
  DIRECTION: true,
  Img: true,
  NutritionFacts: true,
  DataPermissionLevels: true,
  AiInformation: true,
  Link: true,
  List: true,
  ListItem: true,
  InlineList: true,
  InlineListItem: true,
  Menu: true,
  MenuItem: true,
  MenuItemGroup: true,
  MenuItemSeparator: true,
  MetricGroup: true,
  Metric: true,
  Modal: true,
  ModalBody: true,
  ModalFooter: true,
  ModalHeader: true,
  Transition: true,
  BaseTransition: true,
  AppNav: true,
  AppNavItem: true,
  NumberInput: true,
  Options: true,
  OptionSeparator: true,
  OptionItem: true,
  Mask: true,
  Overlay: true,
  Pages: true,
  PagesPage: true,
  Pagination: true,
  PaginationButton: true,
  Pill: true,
  Popover: true,
  Portal: true,
  Position: true,
  calculateElementPosition: true,
  executeMirrorFunction: true,
  mirrorHorizontalPlacement: true,
  mirrorPlacement: true,
  parsePlacement: true,
  ProgressBar: true,
  ProgressCircle: true,
  RadioInput: true,
  RadioInputGroup: true,
  RangeInput: true,
  Rating: true,
  RatingIcon: true,
  Responsive: true,
  ResponsivePropTypes: true,
  addElementQueryMatchListener: true,
  addMediaQueryMatchListener: true,
  parseQuery: true,
  jsonToMediaQuery: true,
  Select: true,
  SelectGroup: true,
  SelectOption: true,
  Selectable: true,
  SideNavBar: true,
  SideNavBarItem: true,
  SimpleSelect: true,
  SimpleSelectOption: true,
  SimpleSelectGroup: true,
  SourceCodeEditor: true,
  Spinner: true,
  InlineSVG: true,
  SVGIcon: true,
  Table: true,
  TableContext: true,
  TableBody: true,
  TableHead: true,
  TableCell: true,
  TableRow: true,
  TableColHeader: true,
  TableRowHeader: true,
  Tabs: true,
  TabsPanel: true,
  TabsTab: true,
  Tag: true,
  Text: true,
  TextArea: true,
  TextInput: true,
  canvas: true,
  canvasHighContrast: true,
  dark: true,
  light: true,
  primitives: true,
  additionalPrimitives: true,
  dataVisualization: true,
  TimeSelect: true,
  ToggleDetails: true,
  ToggleGroup: true,
  Tooltip: true,
  TopNavBar: true,
  TopNavBarActionItems: true,
  TopNavBarBrand: true,
  TopNavBarBreadcrumb: true,
  TopNavBarItem: true,
  TopNavBarLayout: true,
  TopNavBarMenuItems: true,
  TopNavBarUser: true,
  Tray: true,
  TreeBrowser: true,
  TreeNode: true,
  TreeButton: true,
  TreeCollection: true,
  TruncateList: true,
  TruncateText: true,
  View: true,
  ContextView: true
};
Object.defineProperty(exports, "AccessibleContent", {
  enumerable: true,
  get: function () {
    return _uiA11yContent.AccessibleContent;
  }
});
Object.defineProperty(exports, "AiInformation", {
  enumerable: true,
  get: function () {
    return _v11_20.AiInformation;
  }
});
Object.defineProperty(exports, "Alert", {
  enumerable: true,
  get: function () {
    return _v11_.Alert;
  }
});
Object.defineProperty(exports, "AppNav", {
  enumerable: true,
  get: function () {
    return _v11_26.AppNav;
  }
});
Object.defineProperty(exports, "AppNavItem", {
  enumerable: true,
  get: function () {
    return _v11_26.AppNavItem;
  }
});
Object.defineProperty(exports, "ApplyLocale", {
  enumerable: true,
  get: function () {
    return _uiI18n.ApplyLocale;
  }
});
Object.defineProperty(exports, "ApplyLocaleContext", {
  enumerable: true,
  get: function () {
    return _uiI18n.ApplyLocaleContext;
  }
});
Object.defineProperty(exports, "Avatar", {
  enumerable: true,
  get: function () {
    return _v11_2.Avatar;
  }
});
Object.defineProperty(exports, "Badge", {
  enumerable: true,
  get: function () {
    return _v11_3.Badge;
  }
});
Object.defineProperty(exports, "BaseButton", {
  enumerable: true,
  get: function () {
    return _v11_6.BaseButton;
  }
});
Object.defineProperty(exports, "BaseTransition", {
  enumerable: true,
  get: function () {
    return _uiMotion.BaseTransition;
  }
});
Object.defineProperty(exports, "Billboard", {
  enumerable: true,
  get: function () {
    return _v11_4.Billboard;
  }
});
Object.defineProperty(exports, "Breadcrumb", {
  enumerable: true,
  get: function () {
    return _v11_5.Breadcrumb;
  }
});
Object.defineProperty(exports, "BreadcrumbLink", {
  enumerable: true,
  get: function () {
    return _v11_5.BreadcrumbLink;
  }
});
Object.defineProperty(exports, "Button", {
  enumerable: true,
  get: function () {
    return _v11_6.Button;
  }
});
Object.defineProperty(exports, "Byline", {
  enumerable: true,
  get: function () {
    return _v11_7.Byline;
  }
});
Object.defineProperty(exports, "Calendar", {
  enumerable: true,
  get: function () {
    return _v11_8.Calendar;
  }
});
Object.defineProperty(exports, "CalendarDay", {
  enumerable: true,
  get: function () {
    return _v11_8.CalendarDay;
  }
});
Object.defineProperty(exports, "Checkbox", {
  enumerable: true,
  get: function () {
    return _v11_9.Checkbox;
  }
});
Object.defineProperty(exports, "CheckboxFacade", {
  enumerable: true,
  get: function () {
    return _v11_9.CheckboxFacade;
  }
});
Object.defineProperty(exports, "CheckboxGroup", {
  enumerable: true,
  get: function () {
    return _v11_9.CheckboxGroup;
  }
});
Object.defineProperty(exports, "CloseButton", {
  enumerable: true,
  get: function () {
    return _v11_6.CloseButton;
  }
});
Object.defineProperty(exports, "ColorContrast", {
  enumerable: true,
  get: function () {
    return _v11_0.ColorContrast;
  }
});
Object.defineProperty(exports, "ColorIndicator", {
  enumerable: true,
  get: function () {
    return _v11_0.ColorIndicator;
  }
});
Object.defineProperty(exports, "ColorMixer", {
  enumerable: true,
  get: function () {
    return _v11_0.ColorMixer;
  }
});
Object.defineProperty(exports, "ColorPicker", {
  enumerable: true,
  get: function () {
    return _v11_0.ColorPicker;
  }
});
Object.defineProperty(exports, "ColorPreset", {
  enumerable: true,
  get: function () {
    return _v11_0.ColorPreset;
  }
});
Object.defineProperty(exports, "CondensedButton", {
  enumerable: true,
  get: function () {
    return _v11_6.CondensedButton;
  }
});
Object.defineProperty(exports, "ContextView", {
  enumerable: true,
  get: function () {
    return _v11_57.ContextView;
  }
});
Object.defineProperty(exports, "DIRECTION", {
  enumerable: true,
  get: function () {
    return _uiI18n.DIRECTION;
  }
});
Object.defineProperty(exports, "DataPermissionLevels", {
  enumerable: true,
  get: function () {
    return _v11_20.DataPermissionLevels;
  }
});
Object.defineProperty(exports, "DateInput", {
  enumerable: true,
  get: function () {
    return _v11_1.DateInput;
  }
});
Object.defineProperty(exports, "DateInput2", {
  enumerable: true,
  get: function () {
    return _v11_1.DateInput2;
  }
});
Object.defineProperty(exports, "DateTime", {
  enumerable: true,
  get: function () {
    return _uiI18n.DateTime;
  }
});
Object.defineProperty(exports, "DateTimeInput", {
  enumerable: true,
  get: function () {
    return _v11_10.DateTimeInput;
  }
});
Object.defineProperty(exports, "Dialog", {
  enumerable: true,
  get: function () {
    return _uiDialog.Dialog;
  }
});
Object.defineProperty(exports, "DrawerContent", {
  enumerable: true,
  get: function () {
    return _v11_11.DrawerContent;
  }
});
Object.defineProperty(exports, "DrawerLayout", {
  enumerable: true,
  get: function () {
    return _v11_11.DrawerLayout;
  }
});
Object.defineProperty(exports, "DrawerTray", {
  enumerable: true,
  get: function () {
    return _v11_11.DrawerTray;
  }
});
Object.defineProperty(exports, "Drilldown", {
  enumerable: true,
  get: function () {
    return _v11_12.Drilldown;
  }
});
Object.defineProperty(exports, "DrilldownGroup", {
  enumerable: true,
  get: function () {
    return _v11_12.DrilldownGroup;
  }
});
Object.defineProperty(exports, "DrilldownOption", {
  enumerable: true,
  get: function () {
    return _v11_12.DrilldownOption;
  }
});
Object.defineProperty(exports, "DrilldownPage", {
  enumerable: true,
  get: function () {
    return _v11_12.DrilldownPage;
  }
});
Object.defineProperty(exports, "DrilldownSeparator", {
  enumerable: true,
  get: function () {
    return _v11_12.DrilldownSeparator;
  }
});
Object.defineProperty(exports, "Editable", {
  enumerable: true,
  get: function () {
    return _v11_13.Editable;
  }
});
Object.defineProperty(exports, "Expandable", {
  enumerable: true,
  get: function () {
    return _uiExpandable.Expandable;
  }
});
Object.defineProperty(exports, "FileDrop", {
  enumerable: true,
  get: function () {
    return _v11_14.FileDrop;
  }
});
Object.defineProperty(exports, "Flex", {
  enumerable: true,
  get: function () {
    return _v11_15.Flex;
  }
});
Object.defineProperty(exports, "FlexItem", {
  enumerable: true,
  get: function () {
    return _v11_15.FlexItem;
  }
});
Object.defineProperty(exports, "Focusable", {
  enumerable: true,
  get: function () {
    return _uiFocusable.Focusable;
  }
});
Object.defineProperty(exports, "FormField", {
  enumerable: true,
  get: function () {
    return _v11_16.FormField;
  }
});
Object.defineProperty(exports, "FormFieldGroup", {
  enumerable: true,
  get: function () {
    return _v11_16.FormFieldGroup;
  }
});
Object.defineProperty(exports, "FormFieldLabel", {
  enumerable: true,
  get: function () {
    return _v11_16.FormFieldLabel;
  }
});
Object.defineProperty(exports, "FormFieldLayout", {
  enumerable: true,
  get: function () {
    return _v11_16.FormFieldLayout;
  }
});
Object.defineProperty(exports, "FormFieldMessage", {
  enumerable: true,
  get: function () {
    return _v11_16.FormFieldMessage;
  }
});
Object.defineProperty(exports, "FormFieldMessages", {
  enumerable: true,
  get: function () {
    return _v11_16.FormFieldMessages;
  }
});
Object.defineProperty(exports, "Grid", {
  enumerable: true,
  get: function () {
    return _v11_17.Grid;
  }
});
Object.defineProperty(exports, "GridCol", {
  enumerable: true,
  get: function () {
    return _v11_17.GridCol;
  }
});
Object.defineProperty(exports, "GridRow", {
  enumerable: true,
  get: function () {
    return _v11_17.GridRow;
  }
});
Object.defineProperty(exports, "Heading", {
  enumerable: true,
  get: function () {
    return _v11_18.Heading;
  }
});
Object.defineProperty(exports, "IconButton", {
  enumerable: true,
  get: function () {
    return _v11_6.IconButton;
  }
});
Object.defineProperty(exports, "Img", {
  enumerable: true,
  get: function () {
    return _v11_19.Img;
  }
});
Object.defineProperty(exports, "InPlaceEdit", {
  enumerable: true,
  get: function () {
    return _v11_13.InPlaceEdit;
  }
});
Object.defineProperty(exports, "InlineList", {
  enumerable: true,
  get: function () {
    return _v11_22.InlineList;
  }
});
Object.defineProperty(exports, "InlineListItem", {
  enumerable: true,
  get: function () {
    return _v11_22.InlineListItem;
  }
});
Object.defineProperty(exports, "InlineSVG", {
  enumerable: true,
  get: function () {
    return _uiSvgImages.InlineSVG;
  }
});
Object.defineProperty(exports, "InstUISettingsProvider", {
  enumerable: true,
  get: function () {
    return _emotion.InstUISettingsProvider;
  }
});
Object.defineProperty(exports, "Link", {
  enumerable: true,
  get: function () {
    return _v11_21.Link;
  }
});
Object.defineProperty(exports, "List", {
  enumerable: true,
  get: function () {
    return _v11_22.List;
  }
});
Object.defineProperty(exports, "ListItem", {
  enumerable: true,
  get: function () {
    return _v11_22.ListItem;
  }
});
Object.defineProperty(exports, "Locale", {
  enumerable: true,
  get: function () {
    return _uiI18n.Locale;
  }
});
Object.defineProperty(exports, "Mask", {
  enumerable: true,
  get: function () {
    return _v11_29.Mask;
  }
});
Object.defineProperty(exports, "Menu", {
  enumerable: true,
  get: function () {
    return _v11_23.Menu;
  }
});
Object.defineProperty(exports, "MenuItem", {
  enumerable: true,
  get: function () {
    return _v11_23.MenuItem;
  }
});
Object.defineProperty(exports, "MenuItemGroup", {
  enumerable: true,
  get: function () {
    return _v11_23.MenuItemGroup;
  }
});
Object.defineProperty(exports, "MenuItemSeparator", {
  enumerable: true,
  get: function () {
    return _v11_23.MenuItemSeparator;
  }
});
Object.defineProperty(exports, "Metric", {
  enumerable: true,
  get: function () {
    return _v11_24.Metric;
  }
});
Object.defineProperty(exports, "MetricGroup", {
  enumerable: true,
  get: function () {
    return _v11_24.MetricGroup;
  }
});
Object.defineProperty(exports, "Modal", {
  enumerable: true,
  get: function () {
    return _v11_25.Modal;
  }
});
Object.defineProperty(exports, "ModalBody", {
  enumerable: true,
  get: function () {
    return _v11_25.ModalBody;
  }
});
Object.defineProperty(exports, "ModalFooter", {
  enumerable: true,
  get: function () {
    return _v11_25.ModalFooter;
  }
});
Object.defineProperty(exports, "ModalHeader", {
  enumerable: true,
  get: function () {
    return _v11_25.ModalHeader;
  }
});
Object.defineProperty(exports, "NumberInput", {
  enumerable: true,
  get: function () {
    return _v11_27.NumberInput;
  }
});
Object.defineProperty(exports, "NutritionFacts", {
  enumerable: true,
  get: function () {
    return _v11_20.NutritionFacts;
  }
});
Object.defineProperty(exports, "OptionItem", {
  enumerable: true,
  get: function () {
    return _v11_28.OptionItem;
  }
});
Object.defineProperty(exports, "OptionSeparator", {
  enumerable: true,
  get: function () {
    return _v11_28.OptionSeparator;
  }
});
Object.defineProperty(exports, "Options", {
  enumerable: true,
  get: function () {
    return _v11_28.Options;
  }
});
Object.defineProperty(exports, "Overlay", {
  enumerable: true,
  get: function () {
    return _v11_29.Overlay;
  }
});
Object.defineProperty(exports, "Pages", {
  enumerable: true,
  get: function () {
    return _v11_30.Pages;
  }
});
Object.defineProperty(exports, "PagesPage", {
  enumerable: true,
  get: function () {
    return _v11_30.PagesPage;
  }
});
Object.defineProperty(exports, "Pagination", {
  enumerable: true,
  get: function () {
    return _v11_31.Pagination;
  }
});
Object.defineProperty(exports, "PaginationButton", {
  enumerable: true,
  get: function () {
    return _v11_31.PaginationButton;
  }
});
Object.defineProperty(exports, "Pill", {
  enumerable: true,
  get: function () {
    return _v11_32.Pill;
  }
});
Object.defineProperty(exports, "Popover", {
  enumerable: true,
  get: function () {
    return _v11_33.Popover;
  }
});
Object.defineProperty(exports, "Portal", {
  enumerable: true,
  get: function () {
    return _uiPortal.Portal;
  }
});
Object.defineProperty(exports, "Position", {
  enumerable: true,
  get: function () {
    return _uiPosition.Position;
  }
});
Object.defineProperty(exports, "PresentationContent", {
  enumerable: true,
  get: function () {
    return _uiA11yContent.PresentationContent;
  }
});
Object.defineProperty(exports, "ProgressBar", {
  enumerable: true,
  get: function () {
    return _v11_34.ProgressBar;
  }
});
Object.defineProperty(exports, "ProgressCircle", {
  enumerable: true,
  get: function () {
    return _v11_34.ProgressCircle;
  }
});
Object.defineProperty(exports, "RadioInput", {
  enumerable: true,
  get: function () {
    return _v11_35.RadioInput;
  }
});
Object.defineProperty(exports, "RadioInputGroup", {
  enumerable: true,
  get: function () {
    return _v11_35.RadioInputGroup;
  }
});
Object.defineProperty(exports, "RangeInput", {
  enumerable: true,
  get: function () {
    return _v11_36.RangeInput;
  }
});
Object.defineProperty(exports, "Rating", {
  enumerable: true,
  get: function () {
    return _v11_37.Rating;
  }
});
Object.defineProperty(exports, "RatingIcon", {
  enumerable: true,
  get: function () {
    return _v11_37.RatingIcon;
  }
});
Object.defineProperty(exports, "Responsive", {
  enumerable: true,
  get: function () {
    return _v11_38.Responsive;
  }
});
Object.defineProperty(exports, "ResponsivePropTypes", {
  enumerable: true,
  get: function () {
    return _v11_38.ResponsivePropTypes;
  }
});
Object.defineProperty(exports, "SVGIcon", {
  enumerable: true,
  get: function () {
    return _uiSvgImages.SVGIcon;
  }
});
Object.defineProperty(exports, "ScreenReaderContent", {
  enumerable: true,
  get: function () {
    return _uiA11yContent.ScreenReaderContent;
  }
});
Object.defineProperty(exports, "Select", {
  enumerable: true,
  get: function () {
    return _v11_39.Select;
  }
});
Object.defineProperty(exports, "SelectGroup", {
  enumerable: true,
  get: function () {
    return _v11_39.SelectGroup;
  }
});
Object.defineProperty(exports, "SelectOption", {
  enumerable: true,
  get: function () {
    return _v11_39.SelectOption;
  }
});
Object.defineProperty(exports, "Selectable", {
  enumerable: true,
  get: function () {
    return _uiSelectable.Selectable;
  }
});
Object.defineProperty(exports, "SideNavBar", {
  enumerable: true,
  get: function () {
    return _v11_40.SideNavBar;
  }
});
Object.defineProperty(exports, "SideNavBarItem", {
  enumerable: true,
  get: function () {
    return _v11_40.SideNavBarItem;
  }
});
Object.defineProperty(exports, "SimpleSelect", {
  enumerable: true,
  get: function () {
    return _v11_41.SimpleSelect;
  }
});
Object.defineProperty(exports, "SimpleSelectGroup", {
  enumerable: true,
  get: function () {
    return _v11_41.SimpleSelectGroup;
  }
});
Object.defineProperty(exports, "SimpleSelectOption", {
  enumerable: true,
  get: function () {
    return _v11_41.SimpleSelectOption;
  }
});
Object.defineProperty(exports, "SourceCodeEditor", {
  enumerable: true,
  get: function () {
    return _v11_42.SourceCodeEditor;
  }
});
Object.defineProperty(exports, "Spinner", {
  enumerable: true,
  get: function () {
    return _v11_43.Spinner;
  }
});
Object.defineProperty(exports, "Table", {
  enumerable: true,
  get: function () {
    return _v11_44.Table;
  }
});
Object.defineProperty(exports, "TableBody", {
  enumerable: true,
  get: function () {
    return _v11_44.TableBody;
  }
});
Object.defineProperty(exports, "TableCell", {
  enumerable: true,
  get: function () {
    return _v11_44.TableCell;
  }
});
Object.defineProperty(exports, "TableColHeader", {
  enumerable: true,
  get: function () {
    return _v11_44.TableColHeader;
  }
});
Object.defineProperty(exports, "TableContext", {
  enumerable: true,
  get: function () {
    return _v11_44.TableContext;
  }
});
Object.defineProperty(exports, "TableHead", {
  enumerable: true,
  get: function () {
    return _v11_44.TableHead;
  }
});
Object.defineProperty(exports, "TableRow", {
  enumerable: true,
  get: function () {
    return _v11_44.TableRow;
  }
});
Object.defineProperty(exports, "TableRowHeader", {
  enumerable: true,
  get: function () {
    return _v11_44.TableRowHeader;
  }
});
Object.defineProperty(exports, "Tabs", {
  enumerable: true,
  get: function () {
    return _v11_45.Tabs;
  }
});
Object.defineProperty(exports, "TabsPanel", {
  enumerable: true,
  get: function () {
    return _v11_45.TabsPanel;
  }
});
Object.defineProperty(exports, "TabsTab", {
  enumerable: true,
  get: function () {
    return _v11_45.TabsTab;
  }
});
Object.defineProperty(exports, "Tag", {
  enumerable: true,
  get: function () {
    return _v11_46.Tag;
  }
});
Object.defineProperty(exports, "Text", {
  enumerable: true,
  get: function () {
    return _v11_47.Text;
  }
});
Object.defineProperty(exports, "TextArea", {
  enumerable: true,
  get: function () {
    return _v11_48.TextArea;
  }
});
Object.defineProperty(exports, "TextDirectionContext", {
  enumerable: true,
  get: function () {
    return _uiI18n.TextDirectionContext;
  }
});
Object.defineProperty(exports, "TextInput", {
  enumerable: true,
  get: function () {
    return _v11_49.TextInput;
  }
});
Object.defineProperty(exports, "TimeSelect", {
  enumerable: true,
  get: function () {
    return _v11_50.TimeSelect;
  }
});
Object.defineProperty(exports, "ToggleButton", {
  enumerable: true,
  get: function () {
    return _v11_6.ToggleButton;
  }
});
Object.defineProperty(exports, "ToggleDetails", {
  enumerable: true,
  get: function () {
    return _v11_51.ToggleDetails;
  }
});
Object.defineProperty(exports, "ToggleFacade", {
  enumerable: true,
  get: function () {
    return _v11_9.ToggleFacade;
  }
});
Object.defineProperty(exports, "ToggleGroup", {
  enumerable: true,
  get: function () {
    return _v11_51.ToggleGroup;
  }
});
Object.defineProperty(exports, "Tooltip", {
  enumerable: true,
  get: function () {
    return _v11_52.Tooltip;
  }
});
Object.defineProperty(exports, "TopNavBar", {
  enumerable: true,
  get: function () {
    return _v11_53.TopNavBar;
  }
});
Object.defineProperty(exports, "TopNavBarActionItems", {
  enumerable: true,
  get: function () {
    return _v11_53.TopNavBarActionItems;
  }
});
Object.defineProperty(exports, "TopNavBarBrand", {
  enumerable: true,
  get: function () {
    return _v11_53.TopNavBarBrand;
  }
});
Object.defineProperty(exports, "TopNavBarBreadcrumb", {
  enumerable: true,
  get: function () {
    return _v11_53.TopNavBarBreadcrumb;
  }
});
Object.defineProperty(exports, "TopNavBarItem", {
  enumerable: true,
  get: function () {
    return _v11_53.TopNavBarItem;
  }
});
Object.defineProperty(exports, "TopNavBarLayout", {
  enumerable: true,
  get: function () {
    return _v11_53.TopNavBarLayout;
  }
});
Object.defineProperty(exports, "TopNavBarMenuItems", {
  enumerable: true,
  get: function () {
    return _v11_53.TopNavBarMenuItems;
  }
});
Object.defineProperty(exports, "TopNavBarUser", {
  enumerable: true,
  get: function () {
    return _v11_53.TopNavBarUser;
  }
});
Object.defineProperty(exports, "Transition", {
  enumerable: true,
  get: function () {
    return _uiMotion.Transition;
  }
});
Object.defineProperty(exports, "Tray", {
  enumerable: true,
  get: function () {
    return _v11_54.Tray;
  }
});
Object.defineProperty(exports, "TreeBrowser", {
  enumerable: true,
  get: function () {
    return _v11_55.TreeBrowser;
  }
});
Object.defineProperty(exports, "TreeButton", {
  enumerable: true,
  get: function () {
    return _v11_55.TreeButton;
  }
});
Object.defineProperty(exports, "TreeCollection", {
  enumerable: true,
  get: function () {
    return _v11_55.TreeCollection;
  }
});
Object.defineProperty(exports, "TreeNode", {
  enumerable: true,
  get: function () {
    return _v11_55.TreeNode;
  }
});
Object.defineProperty(exports, "TruncateList", {
  enumerable: true,
  get: function () {
    return _uiTruncateList.TruncateList;
  }
});
Object.defineProperty(exports, "TruncateText", {
  enumerable: true,
  get: function () {
    return _v11_56.TruncateText;
  }
});
Object.defineProperty(exports, "View", {
  enumerable: true,
  get: function () {
    return _v11_57.View;
  }
});
Object.defineProperty(exports, "addElementQueryMatchListener", {
  enumerable: true,
  get: function () {
    return _v11_38.addElementQueryMatchListener;
  }
});
Object.defineProperty(exports, "addMediaQueryMatchListener", {
  enumerable: true,
  get: function () {
    return _v11_38.addMediaQueryMatchListener;
  }
});
Object.defineProperty(exports, "additionalPrimitives", {
  enumerable: true,
  get: function () {
    return _uiThemes.additionalPrimitives;
  }
});
Object.defineProperty(exports, "calculateElementPosition", {
  enumerable: true,
  get: function () {
    return _uiPosition.calculateElementPosition;
  }
});
Object.defineProperty(exports, "canvas", {
  enumerable: true,
  get: function () {
    return _uiThemes.canvas;
  }
});
Object.defineProperty(exports, "canvasHighContrast", {
  enumerable: true,
  get: function () {
    return _uiThemes.canvasHighContrast;
  }
});
Object.defineProperty(exports, "dark", {
  enumerable: true,
  get: function () {
    return _uiThemes.dark;
  }
});
Object.defineProperty(exports, "dataVisualization", {
  enumerable: true,
  get: function () {
    return _uiThemes.dataVisualization;
  }
});
Object.defineProperty(exports, "executeMirrorFunction", {
  enumerable: true,
  get: function () {
    return _uiPosition.executeMirrorFunction;
  }
});
Object.defineProperty(exports, "getTextDirection", {
  enumerable: true,
  get: function () {
    return _uiI18n.getTextDirection;
  }
});
Object.defineProperty(exports, "jsonToMediaQuery", {
  enumerable: true,
  get: function () {
    return _v11_38.jsonToMediaQuery;
  }
});
Object.defineProperty(exports, "light", {
  enumerable: true,
  get: function () {
    return _uiThemes.light;
  }
});
Object.defineProperty(exports, "mirrorHorizontalPlacement", {
  enumerable: true,
  get: function () {
    return _uiPosition.mirrorHorizontalPlacement;
  }
});
Object.defineProperty(exports, "mirrorPlacement", {
  enumerable: true,
  get: function () {
    return _uiPosition.mirrorPlacement;
  }
});
Object.defineProperty(exports, "parsePlacement", {
  enumerable: true,
  get: function () {
    return _uiPosition.parsePlacement;
  }
});
Object.defineProperty(exports, "parseQuery", {
  enumerable: true,
  get: function () {
    return _v11_38.parseQuery;
  }
});
Object.defineProperty(exports, "primitives", {
  enumerable: true,
  get: function () {
    return _uiThemes.primitives;
  }
});
Object.defineProperty(exports, "textDirectionContextConsumer", {
  enumerable: true,
  get: function () {
    return _uiI18n.textDirectionContextConsumer;
  }
});
Object.defineProperty(exports, "withStyleNew", {
  enumerable: true,
  get: function () {
    return _emotion.withStyleNew;
  }
});
var _emotion = require("@instructure/emotion");
var _uiA11yContent = require("@instructure/ui-a11y-content");
var _v11_ = require("@instructure/ui-alerts/v11_7");
var _v11_2 = require("@instructure/ui-avatar/v11_7");
var _v11_3 = require("@instructure/ui-badge/v11_7");
var _v11_4 = require("@instructure/ui-billboard/v11_7");
var _v11_5 = require("@instructure/ui-breadcrumb/v11_7");
var _v11_6 = require("@instructure/ui-buttons/v11_7");
var _v11_7 = require("@instructure/ui-byline/v11_7");
var _v11_8 = require("@instructure/ui-calendar/v11_7");
var _v11_9 = require("@instructure/ui-checkbox/v11_7");
var _v11_0 = require("@instructure/ui-color-picker/v11_7");
var _v11_1 = require("@instructure/ui-date-input/v11_7");
var _v11_10 = require("@instructure/ui-date-time-input/v11_7");
var _uiDialog = require("@instructure/ui-dialog");
var _v11_11 = require("@instructure/ui-drawer-layout/v11_7");
var _v11_12 = require("@instructure/ui-drilldown/v11_7");
var _v11_13 = require("@instructure/ui-editable/v11_7");
var _uiExpandable = require("@instructure/ui-expandable");
var _v11_14 = require("@instructure/ui-file-drop/v11_7");
var _v11_15 = require("@instructure/ui-flex/v11_7");
var _uiFocusable = require("@instructure/ui-focusable");
var _v11_16 = require("@instructure/ui-form-field/v11_7");
var _v11_17 = require("@instructure/ui-grid/v11_7");
var _v11_18 = require("@instructure/ui-heading/v11_7");
var _uiI18n = require("@instructure/ui-i18n");
var _uiIcons = require("@instructure/ui-icons");
Object.keys(_uiIcons).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _uiIcons[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _uiIcons[key];
    }
  });
});
var _v11_19 = require("@instructure/ui-img/v11_7");
var _v11_20 = require("@instructure/ui-instructure/v11_7");
var _v11_21 = require("@instructure/ui-link/v11_7");
var _v11_22 = require("@instructure/ui-list/v11_7");
var _v11_23 = require("@instructure/ui-menu/v11_7");
var _v11_24 = require("@instructure/ui-metric/v11_7");
var _v11_25 = require("@instructure/ui-modal/v11_7");
var _uiMotion = require("@instructure/ui-motion");
var _v11_26 = require("@instructure/ui-navigation/v11_7");
var _v11_27 = require("@instructure/ui-number-input/v11_7");
var _v11_28 = require("@instructure/ui-options/v11_7");
var _v11_29 = require("@instructure/ui-overlays/v11_7");
var _v11_30 = require("@instructure/ui-pages/v11_7");
var _v11_31 = require("@instructure/ui-pagination/v11_7");
var _v11_32 = require("@instructure/ui-pill/v11_7");
var _v11_33 = require("@instructure/ui-popover/v11_7");
var _uiPortal = require("@instructure/ui-portal");
var _uiPosition = require("@instructure/ui-position");
var _v11_34 = require("@instructure/ui-progress/v11_7");
var _v11_35 = require("@instructure/ui-radio-input/v11_7");
var _v11_36 = require("@instructure/ui-range-input/v11_7");
var _v11_37 = require("@instructure/ui-rating/v11_7");
var _v11_38 = require("@instructure/ui-responsive/v11_7");
var _v11_39 = require("@instructure/ui-select/v11_7");
var _uiSelectable = require("@instructure/ui-selectable");
var _v11_40 = require("@instructure/ui-side-nav-bar/v11_7");
var _v11_41 = require("@instructure/ui-simple-select/v11_7");
var _v11_42 = require("@instructure/ui-source-code-editor/v11_7");
var _v11_43 = require("@instructure/ui-spinner/v11_7");
var _uiSvgImages = require("@instructure/ui-svg-images");
var _v11_44 = require("@instructure/ui-table/v11_7");
var _v11_45 = require("@instructure/ui-tabs/v11_7");
var _v11_46 = require("@instructure/ui-tag/v11_7");
var _v11_47 = require("@instructure/ui-text/v11_7");
var _v11_48 = require("@instructure/ui-text-area/v11_7");
var _v11_49 = require("@instructure/ui-text-input/v11_7");
var _uiThemes = require("@instructure/ui-themes");
var _v11_50 = require("@instructure/ui-time-select/v11_7");
var _v11_51 = require("@instructure/ui-toggle-details/v11_7");
var _v11_52 = require("@instructure/ui-tooltip/v11_7");
var _v11_53 = require("@instructure/ui-top-nav-bar/v11_7");
var _v11_54 = require("@instructure/ui-tray/v11_7");
var _v11_55 = require("@instructure/ui-tree-browser/v11_7");
var _uiTruncateList = require("@instructure/ui-truncate-list");
var _v11_56 = require("@instructure/ui-truncate-text/v11_7");
var _v11_57 = require("@instructure/ui-view/v11_7");
