// @instructure/ui-navigation/es/AppNav/v2/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { withStyleNew } from '@instructure/emotion';
import { callRenderProp, omitProps } from '@instructure/ui-react-utils';
import { View } from '@instructure/ui-view/latest';
import { Menu } from '@instructure/ui-menu/latest';
import { Item } from "./Item/index.js";
import generateStyle from "./styles.js";
import { allowedProps } from "./props.js";
import { TruncateList } from '@instructure/ui-truncate-list';

/**
---
category: components
---
**/
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
let AppNav = (_dec = withStyleNew(generateStyle), _dec(_class = class AppNav extends Component {
  static displayName = "AppNav";
  static componentId = 'AppNav';
  static allowedProps = allowedProps;
  static defaultProps = {
    children: null,
    debounce: 300,
    margin: '0',
    renderTruncateLabel: () => 'More',
    visibleItemsCount: 0
  };
  static Item = Item;
  state = {
    isMeasuring: false
  };
  ref = null;
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  renderMenu(items) {
    return _jsx(Menu, {
      trigger: _jsx(AppNav.Item, {
        renderLabel: callRenderProp(this.props.renderTruncateLabel)
      }),
      children: items.map((item, index) => {
        return _jsx(Menu.Item, {
          href: item.props.href ? item.props.href : undefined,
          onClick: item.props.onClick && !item.props.href ? item.props.onClick : undefined,
          children: callRenderProp(item.props.renderLabel)
        }, index);
      })
    });
  }
  render() {
    const {
      visibleItemsCount,
      screenReaderLabel,
      margin,
      debounce,
      styles
    } = this.props;
    const passthroughProps = View.omitViewProps(omitProps(this.props, AppNav.allowedProps), AppNav);
    const renderBeforeItems = callRenderProp(this.props.renderBeforeItems);
    const renderAfterItems = callRenderProp(this.props.renderAfterItems);
    const hasRenderedContent = renderBeforeItems || renderAfterItems;
    return _jsxs(View, {
      ...passthroughProps,
      as: "nav",
      css: [styles?.appNav, hasRenderedContent ? styles?.alignCenter : ''],
      margin: margin,
      display: hasRenderedContent ? 'flex' : 'block',
      elementRef: this.handleRef,
      "data-cid": "AppNav",
      children: [renderBeforeItems && _jsx("span", {
        children: renderBeforeItems
      }), _jsx(TruncateList, {
        visibleItemsCount: visibleItemsCount,
        debounce: debounce,
        onUpdate: this.props.onUpdate,
        renderHiddenItemMenu: hiddenChildren => this.renderMenu(hiddenChildren),
        itemSpacing: styles?.horizontalMargin,
        fixMenuTriggerWidth: styles?.menuTriggerWidth,
        css: styles?.list,
        "aria-label": callRenderProp(screenReaderLabel),
        children: this.props.children
      }), renderAfterItems && _jsx("span", {
        children: renderAfterItems
      })]
    });
  }
}) || _class);
export { AppNav };
export default AppNav;
