// @instructure/ui-themes/lib/themes/newThemeTokens/legacyCanvasHighContrast/components/toggle.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const toggle = semantics => ({
  backgroundColor: semantics.color.background.muted,
  backgroundHoverColor: semantics.color.background.muted,
  backgroundDisabledColor: semantics.color.background.muted,
  backgroundReadonlyColor: semantics.color.background.muted,
  errorBackgroundColor: semantics.color.background.muted,
  errorBackgroundHoverColor: semantics.color.background.muted,
  borderColor: semantics.color.stroke.interactive.input.base,
  borderHoverColor: semantics.color.stroke.interactive.input.hover,
  borderDisabledColor: semantics.color.stroke.interactive.input.base,
  borderReadonlyColor: semantics.color.stroke.interactive.input.readonly,
  errorBorderColor: semantics.color.stroke.error,
  checkedBorderColor: semantics.color.stroke.success,
  checkedBorderHoverColor: semantics.color.stroke.success,
  checkedBorderDisabledColor: semantics.color.stroke.success,
  checkedBorderReadonlyColor: semantics.color.stroke.success,
  borderWidth: semantics.borderWidth.sm,
  borderRadius: semantics.borderRadius.full,
  checkedBackgroundColor: semantics.color.background.success,
  checkedBackgroundHoverColor: semantics.color.background.success,
  checkedBackgroundDisabledColor: semantics.color.background.success,
  checkedBackgroundReadonlyColor: semantics.color.background.success,
  labelColor: semantics.color.institutional.brandFontColorDark,
  labelDisabledColor: semantics.color.institutional.brandFontColorDark,
  uncheckedIconBorderColor: semantics.color.stroke.strong,
  uncheckedIconErrorBorderColor: semantics.color.stroke.strong,
  uncheckedIconBorderHoverColor: semantics.color.stroke.strong,
  uncheckedIconBorderDisabledColor: semantics.color.stroke.strong,
  uncheckedIconBorderReadonlyColor: semantics.color.stroke.strong,
  checkedIconBorderColor: semantics.color.stroke.interactive.action.success.base,
  checkedIconBorderHoverColor: semantics.color.stroke.interactive.action.success.hover,
  checkedIconBorderDisabledColor: semantics.color.stroke.interactive.action.success.base,
  checkedIconBorderReadonlyColor: semantics.color.stroke.interactive.action.success.base,
  marginStart: semantics.spacing.spaceMd,
  marginEnd: semantics.spacing.spaceMd,
  marginVertical: semantics.spacing.spaceSm,
  toggleShadow: {
    0: {
      x: semantics.dropShadow.x.elevation1.dropshadow1,
      y: semantics.dropShadow.y.elevation1.dropshadow1,
      blur: semantics.dropShadow.blur.elevation1.dropshadow1,
      spread: semantics.dropShadow.spread.elevation1.dropshadow1,
      color: semantics.color.dropShadow.shadowColor1,
      type: "dropShadow"
    },
    1: {
      x: semantics.dropShadow.x.elevation1.dropshadow2,
      y: semantics.dropShadow.y.elevation1.dropshadow2,
      blur: semantics.dropShadow.blur.elevation1.dropshadow2,
      spread: semantics.dropShadow.spread.elevation1.dropshadow2,
      color: semantics.color.dropShadow.shadowColor2,
      type: "dropShadow"
    }
  },
  toggleSize: "1.625rem",
  toggleWidth: "2.5rem",
  labelFontFamily: semantics.fontFamily.base,
  labelFontWeight: semantics.fontWeight.body.base,
  labelLineHeightMd: semantics.lineHeight.paragraph.textBase,
  labelLineHeightLg: semantics.lineHeight.paragraph.textBase,
  labelLineHeightSm: semantics.lineHeight.paragraph.textSm,
  labelFontSizeSm: semantics.fontSize.textSm,
  labelFontSizeMd: semantics.fontSize.textBase,
  labelFontSizeLg: semantics.fontSize.textLg,
  toggleBackground: semantics.color.background.base,
  disabledOpacity: semantics.opacity.disabled
});
var _default = exports.default = toggle;
