// @instructure/ui-range-input/lib/RangeInput/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.RangeInput = void 0;
var _react = require("react");
var _console = require("@instructure/console");
var _v11_ = require("@instructure/ui-view/v11_6");
var _v11_2 = require("@instructure/ui-form-field/v11_6");
var _addEventListener = require("@instructure/ui-dom-utils/lib/addEventListener.js");
var _emotion = require("@instructure/emotion");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _pickProps = require("@instructure/ui-react-utils/lib/pickProps.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let RangeInput = exports.RangeInput = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = _dec2(_class = class RangeInput extends _react.Component {
  static displayName = "RangeInput";
  static componentId = 'RangeInput';
  static outputLocatorAttribute = 'data-range-output';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    step: 1,
    formatValue: val => val,
    // If min and max has default value, they don't give a warning if not set, even if they are required props.
    // TODO: figure out if they don't need to be required or remove defaults in V9.
    max: 0,
    min: 0,
    inline: false,
    size: 'medium',
    layout: 'stacked',
    displayValue: true,
    disabled: false,
    readOnly: false,
    thumbVariant: 'deprecated'
  };
  ref = null;
  _input = null;
  _inputListener = null;
  _changeListener = null;
  defaultId;
  handleRef = el => {
    this.ref = el;
  };
  handleInputRef = el => {
    this._input = el;
    if (typeof this.props.inputRef === 'function') {
      this.props.inputRef(el);
    }
  };
  constructor(props) {
    super(props);
    if (typeof props.value === 'undefined') {
      this.state = {
        value: props.defaultValue
      };
    }
    this.defaultId = props.deterministicId();
  }

  /* workaround for https://github.com/facebook/react/issues/554 */
  componentDidMount() {
    this.props.makeStyles?.();
    if (this._input !== null) {
      // https://connect.microsoft.com/IE/Feedback/Details/856998
      this._inputListener = (0, _addEventListener.addEventListener)(this._input, 'input', this.handleChange);
      this._changeListener = (0, _addEventListener.addEventListener)(this._input, 'change', this.handleChange);
    }
  }
  componentWillUnmount() {
    if (!this._input) {
      return;
    }
    this._inputListener?.remove();
    this._changeListener?.remove();
  }
  /* end workaround */

  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  handleChange = event => {
    const {
      onChange,
      value
    } = this.props;
    if (typeof value === 'undefined') {
      this.setState({
        value: event.target.value
      });
    }
    if (typeof onChange === 'function') {
      onChange(event.target.value);
    }
  };

  // controlled input must have an onChange, but we're handling it with native events
  noopChange = () => {};
  get value() {
    const value = typeof this.props.value === 'undefined' ? this.state.value : this.props.value;
    return typeof value === 'string' ? parseInt(value) : value;
  }
  get id() {
    return this.props.id || this.defaultId;
  }
  renderValue() {
    if (this.props.displayValue) {
      if (!this.value) {
        (0, _console.warn)(false, 'RangeInput should have a `value` or `defaultValue` set for the value to be displayed. If no value has to be displayed, set `displayValue={false}`.');
        return null;
      }
      const props = {
        [RangeInput.outputLocatorAttribute]: this.id
      };
      return (0, _jsxRuntime.jsx)(_v11_.ContextView, {
        background: "inverse",
        placement: "end center",
        children: (0, _jsxRuntime.jsx)("div", {
          ...props,
          css: this.props.styles?.rangeInputInputValue,
          children: this.props.formatValue(this.value, this.props.max)
        })
      });
    }
    return null;
  }
  render() {
    const {
      formatValue,
      disabled,
      readOnly
    } = this.props;
    const props = (0, _omitProps.omitProps)(this.props, RangeInput.allowedProps);
    return (0, _jsxRuntime.jsx)(_v11_2.FormField, {
      ...(0, _pickProps.pickProps)(this.props, _v11_2.FormField.allowedProps),
      label: this.props.label,
      id: this.id,
      elementRef: this.handleRef,
      "data-cid": "RangeInput",
      children: (0, _jsxRuntime.jsxs)("div", {
        css: this.props.styles?.rangeInput,
        children: [(0, _jsxRuntime.jsx)("input", {
          css: this.props.styles?.rangeInputInput,
          ref: this.handleInputRef,
          type: "range",
          id: this.id,
          min: this.props.min,
          max: this.props.max,
          step: this.props.step,
          value: this.value,
          onChange: this.noopChange,
          "aria-valuetext": formatValue(this.value, this.props.max),
          ...props,
          disabled: disabled || readOnly,
          "aria-disabled": disabled || readOnly ? 'true' : undefined
        }), this.renderValue()]
      })
    });
  }
}) || _class) || _class);
var _default = exports.default = RangeInput;
