// @instructure/ui-selectable/lib/Selectable/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Selectable = void 0;
var _react = require("react");
var _keycode = _interopRequireDefault(require("keycode"));
var _isActiveElement = require("@instructure/ui-dom-utils/lib/isActiveElement.js");
var _createChainedFunction = require("@instructure/ui-utils/lib/createChainedFunction.js");
var _console = require("@instructure/console");
var _props = require("./props");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components/utilities
tags: autocomplete, typeahead, combobox, dropdown, search
---
**/
let Selectable = exports.Selectable = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec(_class = class Selectable extends _react.Component {
  static displayName = "Selectable";
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    isShowingOptions: false
  };
  _id = this.props.id || this.props.deterministicId();
  _listId = `${this._id}-list`;
  _descriptionId = `${this._id}-description`;
  _trigger = null;
  isSelectedOption = id => {
    const {
      selectedOptionId
    } = this.props;
    if (Array.isArray(selectedOptionId)) {
      return selectedOptionId.indexOf(id) > -1;
    }
    return selectedOptionId === id;
  };
  handleOpenClose = event => {
    const {
      isShowingOptions,
      onRequestShowOptions,
      onRequestHideOptions
    } = this.props;
    event.preventDefault();
    if (isShowingOptions) {
      onRequestHideOptions?.(event);
    } else {
      if (!(0, _isActiveElement.isActiveElement)(this._trigger)) {
        ;
        this._trigger.focus();
      }
      onRequestShowOptions?.(event);
    }
  };
  handleKeyDown = event => {
    const {
      isShowingOptions,
      highlightedOptionId,
      onRequestHighlightOption,
      onRequestHighlightFirstOption,
      onRequestHighlightLastOption,
      onRequestSelectOption
    } = this.props;
    const key = _keycode.default.names[event.keyCode];
    switch (key) {
      case 'space':
        if (!isShowingOptions) {
          // if options hidden, show them
          this.handleOpenClose(event);
        }
        break;
      case 'enter':
        if (highlightedOptionId) {
          // select highlighted option
          event.preventDefault();
          onRequestSelectOption?.(event, {
            id: highlightedOptionId
          });
        }
        break;
      case 'down':
        event.preventDefault();
        if (isShowingOptions) {
          // if options showing, change highlight
          onRequestHighlightOption?.(event, {
            direction: 1
          });
        } else {
          // otherwise, show options
          this.handleOpenClose(event);
        }
        break;
      case 'up':
        event.preventDefault();
        if (isShowingOptions) {
          // if options showing, change highlight
          onRequestHighlightOption?.(event, {
            direction: -1
          });
        } else {
          // otherwise, show options
          this.handleOpenClose(event);
        }
        break;
      case 'home':
        if (isShowingOptions) {
          // if options showing, highlight first option
          event.preventDefault();
          onRequestHighlightFirstOption?.(event);
        }
        break;
      case 'end':
        if (isShowingOptions) {
          // if options showing, highlight last option
          event.preventDefault();
          onRequestHighlightLastOption?.(event);
        }
        break;
    }
  };
  handleKeyUp = event => {
    const {
      isShowingOptions
    } = this.props;
    const key = _keycode.default.names[event.keyCode];
    if (key === 'esc') {
      if (isShowingOptions) {
        // if options showing, hide them
        this.handleOpenClose(event);
      }
    }
  };
  render() {
    const {
      isShowingOptions,
      highlightedOptionId,
      onRequestHighlightOption,
      onRequestSelectOption,
      children,
      render = children
    } = this.props;
    if (typeof render === 'function') {
      return render({
        getRootProps: ({
          onMouseDown,
          ...rest
        } = {}) => {
          return {
            onMouseDown: (0, _createChainedFunction.createChainedFunction)(event => {
              // if we call preventDefault, label can't be selected and copied, so we only call it when the options are shown
              if (event.target !== this._trigger && isShowingOptions) {
                event.preventDefault(); // prevent trigger from losing focus
              }
            }, onMouseDown),
            ...rest
          };
        },
        getLabelProps: props => {
          return {
            htmlFor: this._id,
            ...props
          };
        },
        getTriggerProps: ({
          ref,
          onKeyDown,
          onKeyUp,
          onClick,
          ...rest
        } = {}) => {
          return {
            id: this._id,
            ref: (0, _createChainedFunction.createChainedFunction)(ref, el => this._trigger = el),
            'aria-haspopup': 'listbox',
            'aria-expanded': isShowingOptions,
            'aria-controls': isShowingOptions ? this._listId : undefined,
            'aria-describedby': this._descriptionId,
            'aria-activedescendant': isShowingOptions ? highlightedOptionId : undefined,
            onKeyDown: (0, _createChainedFunction.createChainedFunction)(this.handleKeyDown, onKeyDown),
            onKeyUp: (0, _createChainedFunction.createChainedFunction)(this.handleKeyUp, onKeyUp),
            onClick: (0, _createChainedFunction.createChainedFunction)(this.handleOpenClose, onClick),
            ...rest
          };
        },
        getInputProps: ({
          readOnly,
          ...rest
        } = {}) => {
          return {
            role: 'combobox',
            'aria-autocomplete': readOnly ? 'none' : 'both',
            autoComplete: 'off',
            readOnly,
            ...rest
          };
        },
        getListProps: ({
          onMouseDown,
          onClick,
          ...rest
        } = {}) => {
          return {
            id: this._listId,
            role: 'listbox',
            onMouseDown: (0, _createChainedFunction.createChainedFunction)(event => {
              event.preventDefault(); // prevent trigger from losing focus
            }, onMouseDown),
            onClick: (0, _createChainedFunction.createChainedFunction)(event => {
              // prevent synthetic event from firing on the document
              // this event could inadvertently close a parent dialog
              event.stopPropagation();
              event.nativeEvent.stopImmediatePropagation();
            }, onClick),
            ...rest
          };
        },
        getOptionProps: ({
          id,
          onMouseOver,
          onClick,
          ...rest
        } = {}) => {
          (0, _console.logError)(Boolean(id), `[Selectable] Must provide id for each option via \`getOptionProps\`.`);
          return {
            id,
            role: 'option',
            'aria-selected': this.isSelectedOption(id) ? 'true' : 'false',
            onClick: (0, _createChainedFunction.createChainedFunction)(event => {
              onRequestSelectOption?.(event, {
                id
              });
            }, onClick),
            onMouseOver: (0, _createChainedFunction.createChainedFunction)(event => {
              onRequestHighlightOption?.(event, {
                id
              });
            }, onMouseOver),
            ...rest
          };
        },
        getDisabledOptionProps: props => {
          return {
            'aria-disabled': 'true',
            ...props
          };
        },
        getDescriptionProps: props => {
          return {
            id: this._descriptionId,
            ...props
          };
        }
      });
    } else {
      return null;
    }
  }
}) || _class);
var _default = exports.default = Selectable;
