// @instructure/ui-toggle-details/lib/ToggleGroup/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ToggleGroup = void 0;
var _react = require("react");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _pickProps = require("@instructure/ui-react-utils/lib/pickProps.js");
var _getElementType = require("@instructure/ui-react-utils/lib/getElementType.js");
var _latest = require("@instructure/ui-buttons/latest");
var _Transition = require("@instructure/ui-motion/lib/Transition");
var _uiExpandable = require("@instructure/ui-expandable");
var _isActiveElement = require("@instructure/ui-dom-utils/lib/isActiveElement.js");
var _latest2 = require("@instructure/ui-flex/latest");
var _latest3 = require("@instructure/ui-view/latest");
var _uiIcons = require("@instructure/ui-icons");
var _props = require("./props");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
const toggleGroupSizeToIconSize = {
  small: 'md',
  medium: 'md',
  large: 'lg'
};

/**
---
category: components
---
**/
let ToggleGroup = exports.ToggleGroup = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class ToggleGroup extends _react.Component {
  static displayName = "ToggleGroup";
  static componentId = 'ToggleGroup';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    size: 'medium',
    icon: _uiIcons.ChevronRightInstUIIcon,
    iconExpanded: _uiIcons.ChevronDownInstUIIcon,
    defaultExpanded: false,
    transition: true,
    as: 'span',
    border: true
  };
  ref = null;
  _button = null;
  _shouldTransition = false;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  handleButtonRef = el => {
    this._button = el;
  };
  get focused() {
    return (0, _isActiveElement.isActiveElement)(this._button);
  }
  focus() {
    this._button?.focus();
  }
  componentDidMount() {
    this._shouldTransition = true;
  }
  componentDidUpdate() {
    this.props.makeStyles?.(this.state);
  }
  renderIcon(expanded) {
    const Icon = expanded ? this.props.iconExpanded : this.props.icon;
    if (!Icon) return null;
    const iconSize = toggleGroupSizeToIconSize[this.props.size];
    return (0, _uiIcons.renderIconWithProps)(Icon, iconSize, 'baseColor');
  }
  renderToggle(toggleProps, expanded) {
    const {
      toggleLabel,
      size
    } = this.props;
    let label;
    if (typeof toggleLabel === 'function') {
      label = toggleLabel(expanded);
    } else {
      label = toggleLabel;
    }
    const props = {
      ...toggleProps
    };
    return (0, _jsxRuntime.jsx)(_latest.IconButton, {
      ...props,
      withBackground: false,
      withBorder: false,
      size: size === 'large' ? 'medium' : 'small',
      elementRef: this.handleButtonRef,
      screenReaderLabel: label,
      children: this.renderIcon(expanded)
    });
  }
  renderDetails(detailsProps) {
    const {
      styles
    } = this.props;
    return (0, _jsxRuntime.jsx)(_latest3.View, {
      ...detailsProps,
      display: "block",
      borderWidth: "small none none none",
      borderColor: styles?.borderColor,
      children: this.props.transition && this._shouldTransition ? (0, _jsxRuntime.jsx)(_Transition.Transition, {
        transitionOnMount: true,
        in: true,
        type: "fade",
        children: this.props.children
      }) : this.props.children
    });
  }
  render() {
    const Element = (0, _getElementType.getElementType)(ToggleGroup, this.props);
    const {
      styles
    } = this.props;
    return (0, _jsxRuntime.jsx)(_uiExpandable.Expandable, {
      ...(0, _pickProps.pickProps)(this.props, _uiExpandable.Expandable.allowedProps),
      children: ({
        expanded,
        getToggleProps,
        getDetailsProps
      }) => {
        return (0, _jsxRuntime.jsxs)(_latest3.View, {
          ...(0, _omitProps.omitProps)(this.props, ToggleGroup.allowedProps),
          borderWidth: this.props.border ? 'small' : 'none',
          as: Element,
          elementRef: this.handleRef,
          display: "block",
          borderRadius: "medium",
          background: "primary",
          borderColor: styles?.borderColor,
          "data-cid": "ToggleGroup",
          children: [(0, _jsxRuntime.jsxs)(_latest2.Flex, {
            padding: this.props.size === 'small' ? 'x-small' : 'small small small x-small',
            children: [(0, _jsxRuntime.jsx)(_latest2.Flex.Item, {
              children: this.renderToggle(getToggleProps(), expanded)
            }), (0, _jsxRuntime.jsx)(_latest2.Flex.Item, {
              shouldGrow: true,
              shouldShrink: true,
              padding: "0 0 0 x-small",
              children: this.props.summary
            })]
          }), expanded ? this.renderDetails(getDetailsProps()) : (0, _jsxRuntime.jsx)("span", {
            ...getDetailsProps()
          })]
        });
      }
    });
  }
}) || _class);
var _default = exports.default = ToggleGroup;
