// @instructure/ui-grid/lib/GridRow/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.GridRow = void 0;
var _react = require("react");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _pickProps = require("@instructure/ui-react-utils/lib/pickProps.js");
var _v = require("../../GridCol/v2/");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: Grid
id: Grid.Row
---
**/
let GridRow = exports.GridRow = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class GridRow extends _react.Component {
  static displayName = "GridRow";
  static componentId = 'Grid.Row';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    children: null,
    isLastRow: false
  };
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  renderChildren() {
    const {
      styles,
      makeStyles,
      ...props
    } = this.props;
    return _react.Children.map(this.props.children, (child, index) => {
      if ((0, _matchComponentTypes.matchComponentTypes)(child, [_v.GridCol])) {
        return (0, _safeCloneElement.safeCloneElement)(child, {
          ...(0, _pickProps.pickProps)(props, GridRow.allowedProps),
          ...child.props /* child props should override parent */,
          isLastRow: props.isLastRow,
          isLastCol: index + 1 === _react.Children.count(this.props.children)
        });
      } else {
        return child; // PropType validation should handle errors
      }
    });
  }
  render() {
    const {
      styles,
      ...restProps
    } = this.props;
    const props = (0, _omitProps.omitProps)(restProps, GridRow.allowedProps);
    return (0, _jsxRuntime.jsx)("span", {
      ...props,
      css: styles?.gridRow,
      ref: this.handleRef,
      children: this.renderChildren()
    });
  }
}) || _class);
var _default = exports.default = GridRow;
