// @instructure/ui-expandable/lib/Expandable/v1/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Expandable = void 0;
var _react = require("react");
var _createChainedFunction = require("@instructure/ui-utils/lib/createChainedFunction.js");
var _props = require("./props");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
const toggleExpanded = ({
  expanded
}) => ({
  expanded: !expanded
});

/**
---
category: components/utilities
---
**/
let Expandable = exports.Expandable = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec(_class = class Expandable extends _react.Component {
  static displayName = "Expandable";
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    defaultExpanded: false
  };
  _contentId;
  constructor(props) {
    super(props);
    this.state = {
      expanded: this.isControlled(props) ? props.expanded : props.defaultExpanded
    };
    this._contentId = props.deterministicId();
  }
  get expanded() {
    return this.isControlled() ? this.props.expanded : this.state.expanded;
  }
  isControlled(props = this.props) {
    return typeof props.expanded === 'boolean';
  }
  static getDerivedStateFromProps(nextProps, state) {
    if (
    // if component is controlled, keep internal state up to date
    // with the `expanded` prop value
    typeof nextProps.expanded === 'boolean' && nextProps.expanded !== state.expanded) {
      return {
        expanded: nextProps.expanded
      };
    } else {
      return null;
    }
  }
  handleToggle = event => {
    if (!this.isControlled()) {
      this.setState(toggleExpanded);
    }
    this.props.onToggle?.(event, !this.expanded);
  };
  render() {
    const {
      children,
      render = children
    } = this.props;
    if (typeof render === 'function') {
      return render({
        expanded: this.expanded,
        getToggleProps: (props = {}) => {
          return {
            ...props,
            'aria-controls': this._contentId,
            'aria-expanded': this.expanded,
            onClick: (0, _createChainedFunction.createChainedFunction)(this.handleToggle, props.onClick)
          };
        },
        getDetailsProps: () => {
          return {
            id: this._contentId
          };
        }
      });
    } else {
      return null;
    }
  }
}) || _class);
var _default = exports.default = Expandable;
