// @instructure/ui-top-nav-bar/lib/TopNavBar/v2/TopNavBarLayout/SmallViewportLayout/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TopNavBarSmallViewportLayout = void 0;
var _react = require("react");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _px = require("@instructure/ui-utils/lib/px.js");
var _console = require("@instructure/console");
var _getBoundingClientRect = require("@instructure/ui-dom-utils/lib/getBoundingClientRect.js");
var _requestAnimationFrame = require("@instructure/ui-dom-utils/lib/requestAnimationFrame.js");
var _emotion = require("@instructure/emotion");
var _latest = require("@instructure/ui-tray/latest");
var _uiIcons = require("@instructure/ui-icons");
var _latest2 = require("@instructure/ui-avatar/latest");
var _Dialog = require("@instructure/ui-dialog/lib/Dialog");
var _latest3 = require("@instructure/ui-drilldown/latest");
var _TopNavBarItem = require("../../TopNavBarItem");
var _mapItemsForDrilldown = require("../../utils/mapItemsForDrilldown");
var _TopNavBarContext = require("../../TopNavBarContext");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
private: true
---
**/
let TopNavBarSmallViewportLayout = exports.TopNavBarSmallViewportLayout = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyleNew)(_styles.default, 'TopNavBarLayout'), _dec(_class = _dec2(_class = class TopNavBarSmallViewportLayout extends _react.Component {
  static displayName = "TopNavBarSmallViewportLayout";
  static componentId = 'TopNavBar.SmallViewportLayout';
  static allowedProps = _props.allowedProps;
  static defaultProps = {};
  static contextType = _TopNavBarContext.TopNavBarContext;
  ref = null;
  _trayContainerId;
  _trayId;
  _drilldownId;
  _menuTriggerId;
  _menuId;
  _inPlaceDialogId;
  _inPlaceDialogCloseButtonId;
  _separatorId;
  _drilldownRef = null;
  _raf = [];
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  constructor(props) {
    super(props);
    this._trayContainerId = props.deterministicId('TopNavBarSmallViewportLayout-trayContainer');
    this._trayId = props.deterministicId('TopNavBarSmallViewportLayout-tray');
    this._menuId = props.deterministicId('TopNavBarSmallViewportLayout-menu');
    this._menuTriggerId = props.deterministicId('TopNavBarSmallViewportLayout-menuTrigger');
    this._drilldownId = props.deterministicId('TopNavBarSmallViewportLayout-drilldown');
    this._inPlaceDialogId = props.deterministicId('TopNavBarSmallViewportLayout-inPlaceDialog');
    this._inPlaceDialogCloseButtonId = props.deterministicId('TopNavBarSmallViewportLayout-inPlaceDialogCloseButton');
    this._separatorId = props.deterministicId('TopNavBarSmallViewportLayout-separator');
    this.state = {
      isDropdownMenuOpen: false,
      isDropdownMenuVisible: false,
      menuBottomPosition: (0, _px.px)(props.styles?.navbarHeight || 0)
    };
  }
  componentDidMount() {
    this.props.makeStyles?.(this.makeStylesVariables);
    this.updateMenuBottomPosition();
  }
  componentDidUpdate() {
    this.props.makeStyles?.(this.makeStylesVariables);
  }
  componentWillUnmount() {
    this._raf.forEach(request => request.cancel());
    if (this.state.isDropdownMenuOpen) {
      this.toggleDropdownMenu();
    }
  }
  get makeStylesVariables() {
    return {
      isDropdownMenuVisible: this.state.isDropdownMenuVisible,
      drilldownId: this._drilldownId,
      trayId: this._trayId,
      menuBottomPosition: this.state.menuBottomPosition,
      inverseColor: this.context.inverseColor
    };
  }
  hasBrandBlock(renderBrand) {
    return !!renderBrand && !!renderBrand.props.renderIcon;
  }
  hasMenuItemsBlock(renderMenuItems) {
    return !!renderMenuItems && _react.Children.count(renderMenuItems.props.children) > 0;
  }
  hasActionItemsBlock(renderActionItems) {
    return !!renderActionItems && _react.Children.count(renderActionItems.props.children) > 0;
  }
  hasUserBlock(renderUser) {
    return !!renderUser && _react.Children.count(renderUser.props.children) > 0;
  }
  get hasSubmenu() {
    return (this.dropdownMenuContent || []).length > 0;
  }
  get hasBreadcrumbBlock() {
    const {
      renderBreadcrumb
    } = this.props;
    return !!renderBreadcrumb && _react.Children.count(renderBreadcrumb.props.children) > 0;
  }
  get isInPlaceDialogOpen() {
    return this.props.renderInPlaceDialogConfig?.open;
  }
  get mappedUserOptions() {
    const {
      renderUser
    } = this.props;
    if (!this.hasUserBlock(renderUser)) {
      return [];
    }
    const userChildren = _react.Children.toArray(renderUser.props.children);
    return (0, _mapItemsForDrilldown.mapItemsForDrilldown)(userChildren, {
      renderOptionContent: this.renderOptionContent
    });
  }
  get mappedMenuItemsOptions() {
    const {
      renderMenuItems
    } = this.props;
    if (!this.hasMenuItemsBlock(renderMenuItems)) {
      return [];
    }
    const menuItemsChildren = _react.Children.toArray(renderMenuItems.props.children);
    return (0, _mapItemsForDrilldown.mapItemsForDrilldown)(menuItemsChildren, {
      renderOptionContent: this.renderOptionContent,
      currentPageId: renderMenuItems.props.currentPageId
    });
  }
  get extractDrilldownSubpages() {
    return (0, _mapItemsForDrilldown.renderMappedItemDrilldownSubpages)([...this.mappedUserOptions, ...this.mappedMenuItemsOptions]);
  }
  updateMenuBottomPosition() {
    const boundingRect = (0, _getBoundingClientRect.getBoundingClientRect)(this.ref);
    this.setState({
      menuBottomPosition: boundingRect.top + boundingRect.height
    });
  }
  toggleDropdownMenu() {
    const {
      onDropdownMenuToggle
    } = this.props;
    const {
      isDropdownMenuOpen
    } = this.state;
    if (!isDropdownMenuOpen) {
      this.updateMenuBottomPosition();
    }
    if (typeof onDropdownMenuToggle === 'function') {
      onDropdownMenuToggle(!isDropdownMenuOpen);
    }
    this.setState({
      isDropdownMenuOpen: !this.state.isDropdownMenuOpen
    }, () => {
      if (this.state.isDropdownMenuOpen) {
        this.focusFirstAvailableItem();
      }
    });
  }
  focusFirstAvailableItem() {
    const userChildren = _react.Children.toArray(this.props.renderUser?.props.children);

    // If option is a User, it preceeds other item, so focus that first
    const targetId = userChildren[0] ? userChildren[0].props.id : this.mappedMenuItemsOptions[0].optionData.id;
    setTimeout(() => {
      const container = document.getElementById(this._trayContainerId);
      const firstOption = container?.querySelector(`[id="${CSS.escape(targetId)}"]`);
      firstOption?.focus();
      if (this._drilldownRef) {
        const drilldownRef = this._drilldownRef;
        setTimeout(() => {
          if (drilldownRef) {
            // highlight the option using Drilldown's handleOptionHighlight function
            drilldownRef.handleOptionHighlight({}, {
              id: targetId
            });
          }
        }, 10);
      }
    }, 10);
  }
  renderOptionContent = (children, itemProps) => {
    const {
      styles
    } = this.props;
    const {
      status,
      renderAvatar
    } = itemProps;
    let content = children;
    let optionStyle = status === 'active' ? styles?.dropdownMenuOptionActive : styles?.dropdownMenuOption;
    if (renderAvatar) {
      const {
        avatarName,
        avatarSrc,
        avatarAlt
      } = renderAvatar;
      const label = avatarAlt || (typeof children === 'string' ? children : undefined);
      optionStyle = styles?.dropdownMenuOptionWithAvatar;
      content = (0, _jsxRuntime.jsxs)(_react.Fragment, {
        children: [(0, _jsxRuntime.jsx)(_latest2.Avatar, {
          name: avatarName,
          src: avatarSrc,
          alt: label,
          size: "small",
          margin: "0 small 0 0",
          role: "presentation",
          "aria-hidden": "true"
        }), children]
      });
    }
    return (0, _jsxRuntime.jsx)("span", {
      css: optionStyle,
      children: content
    });
  };
  renderMenuTrigger() {
    const {
      dropdownMenuToggleButtonLabel,
      dropdownMenuToggleButtonTooltip,
      renderBrand,
      alternativeTitle,
      styles
    } = this.props;
    const {
      isDropdownMenuOpen
    } = this.state;
    let menuTrigger;
    const menuTriggerStyle = [styles?.menuTrigger, ...(alternativeTitle ? [styles?.alternativeTitleContainer] : [])];
    if (!this.hasSubmenu) {
      menuTrigger = null;
      if (alternativeTitle) {
        (0, _console.warn)(false, 'There are no menu items or user menu to display in the <TopNavBar> dropdown menu! The menu trigger and the alternative title will not display.');
      }
    } else {
      const itemProps = {
        id: this._menuTriggerId,
        onClick: () => {
          this.toggleDropdownMenu();
        },
        tooltip: dropdownMenuToggleButtonTooltip,
        themeOverride: {
          itemSpacing: '0.375rem'
        },
        'aria-haspopup': 'menu',
        'aria-expanded': isDropdownMenuOpen,
        withFocusOutline: isDropdownMenuOpen ? true : undefined
      };
      const alternativeTitleIconProps = {
        size: 'x-small',
        style: {
          marginInlineEnd: '0.5em'
        }
      };
      menuTrigger = (0, _jsxRuntime.jsx)("div", {
        css: menuTriggerStyle,
        children: alternativeTitle ? (0, _jsxRuntime.jsxs)(_TopNavBarItem.TopNavBarItem, {
          ...itemProps,
          "aria-label": dropdownMenuToggleButtonLabel,
          children: [isDropdownMenuOpen ? (0, _jsxRuntime.jsx)(_uiIcons.ChevronUpInstUIIcon, {
            ...alternativeTitleIconProps
          }) : (0, _jsxRuntime.jsx)(_uiIcons.ChevronDownInstUIIcon, {
            ...alternativeTitleIconProps
          }), alternativeTitle]
        }) : (0, _jsxRuntime.jsx)(_TopNavBarItem.TopNavBarItem, {
          ...itemProps,
          variant: "icon",
          renderIcon: isDropdownMenuOpen ? (0, _jsxRuntime.jsx)(_uiIcons.XInstUIIcon, {}) : (0, _jsxRuntime.jsx)(_uiIcons.MenuInstUIIcon, {}),
          children: dropdownMenuToggleButtonLabel
        })
      });
    }
    return (0, _jsxRuntime.jsxs)("div", {
      css: styles?.menuTriggerContainer,
      children: [menuTrigger, !this.hasBreadcrumbBlock && !this.props.trayMountNode && (0, _jsxRuntime.jsx)("div", {
        css: styles?.trayContainer,
        id: this._trayContainerId
      }), this.hasBrandBlock(renderBrand) && !alternativeTitle && (0, _jsxRuntime.jsx)("div", {
        css: styles?.brandContainer,
        children: renderBrand
      })]
    });
  }
  get dropdownMenuContent() {
    const {
      renderUser,
      renderMenuItems
    } = this.props;
    const hasMenuItems = this.hasMenuItemsBlock(renderMenuItems);
    const hasUser = this.hasUserBlock(renderUser);
    let pageChildren = [];
    if (hasUser) {
      pageChildren = (0, _mapItemsForDrilldown.renderMappedItemsAsDrilldownOptions)(this.mappedUserOptions);
      if (hasMenuItems) {
        pageChildren.push((0, _jsxRuntime.jsx)(_latest3.Drilldown.Separator, {
          id: this._separatorId
        }, this._separatorId));
      }
    }
    if (hasMenuItems) {
      pageChildren = [...pageChildren, ...(0, _mapItemsForDrilldown.renderMappedItemsAsDrilldownOptions)(this.mappedMenuItemsOptions)];
    }
    return pageChildren.length ? pageChildren : undefined;
  }
  renderDropdownMenu() {
    const {
      dropdownMenuLabel,
      onDropdownMenuSelect
    } = this.props;
    return (0, _jsxRuntime.jsx)(_latest3.Drilldown, {
      ref: el => {
        this._drilldownRef = el;
      },
      id: this._drilldownId,
      rootPageId: this._menuId,
      label: dropdownMenuLabel,
      height: "100%",
      width: "100%",
      onSelect: (e, args) => {
        if (typeof onDropdownMenuSelect === 'function') {
          onDropdownMenuSelect(e, args);
        }
        if (args.selectedOption.props.shouldCloseOnClick === 'auto' && !!args.selectedOption.props.href || args.selectedOption.props.shouldCloseOnClick === 'always') {
          this.toggleDropdownMenu();
        }
      },
      children: [(0, _jsxRuntime.jsx)(_latest3.Drilldown.Page, {
        id: this._menuId,
        children: this.dropdownMenuContent
      }, this._menuId), ...this.extractDrilldownSubpages]
    });
  }
  renderDropdownMenuTray() {
    const {
      trayMountNode
    } = this.props;
    if (!this.hasSubmenu) {
      return null;
    }
    return (0, _jsxRuntime.jsx)(_latest.Tray, {
      id: this._trayId,
      label: '',
      role: "none",
      open: this.state.isDropdownMenuOpen,
      transitionExit: false,
      onDismiss: () => {
        this.toggleDropdownMenu();
      },
      onKeyUp: e => {
        if (e.key === 'Escape') {
          this.toggleDropdownMenu();
        }
      },
      onEnter: () => {
        this.setState({
          isDropdownMenuVisible: true
        });
      },
      onExited: () => {
        this.setState({
          isDropdownMenuVisible: false
        });
      },
      shouldCloseOnDocumentClick: true,
      placement: "top",
      shadow: false,
      mountNode: trayMountNode || document.getElementById(this._trayContainerId),
      defaultFocusElement: () => document.getElementById(this._drilldownId),
      children: this.renderDropdownMenu()
    });
  }
  renderInPlaceDialog() {
    const {
      renderInPlaceDialogConfig,
      styles
    } = this.props;
    if (!renderInPlaceDialogConfig) {
      return null;
    }
    const {
      content,
      open,
      onClose,
      closeButtonLabel,
      shouldContainFocus,
      shouldCloseOnEscape,
      shouldCloseOnDocumentClick,
      returnFocusElement
    } = renderInPlaceDialogConfig;
    const handleClose = () => {
      if (typeof onClose === 'function') {
        onClose();
      }
      if (returnFocusElement) {
        this._raf.push((0, _requestAnimationFrame.requestAnimationFrame)(() => {
          returnFocusElement()?.focus();
        }));
      }
    };
    const dialog = (0, _jsxRuntime.jsx)(_Dialog.Dialog, {
      display: "block",
      open: open,
      shouldContainFocus: shouldContainFocus,
      shouldCloseOnEscape: shouldCloseOnEscape,
      shouldCloseOnDocumentClick: shouldCloseOnDocumentClick,
      onDismiss: handleClose,
      children: (0, _jsxRuntime.jsxs)("div", {
        id: this._inPlaceDialogId,
        css: styles?.inPlaceDialogContainer,
        children: [(0, _jsxRuntime.jsx)("div", {
          css: styles?.inPlaceDialogContainerContent,
          children: (0, _callRenderProp.callRenderProp)(content, {
            closeInPlaceDialog: handleClose
          })
        }), (0, _jsxRuntime.jsx)("div", {
          css: styles?.inPlaceDialogContainerButton,
          children: (0, _jsxRuntime.jsx)(_TopNavBarItem.TopNavBarItem, {
            id: this._inPlaceDialogCloseButtonId,
            renderIcon: _uiIcons.XInstUIIcon,
            variant: "icon",
            onClick: handleClose,
            children: closeButtonLabel
          })
        })]
      })
    });
    return dialog;
  }
  render() {
    const {
      navLabel,
      renderActionItems,
      renderBreadcrumb,
      styles
    } = this.props;
    return (0, _jsxRuntime.jsxs)("nav", {
      ...(0, _omitProps.omitProps)(this.props, _props.allowedProps),
      ref: this.handleRef,
      "aria-label": navLabel,
      "data-cid": "TopNavBarSmallViewportLayout",
      children: [(0, _jsxRuntime.jsx)(_emotion.Global, {
        styles: styles?.globalStyles
      }), this.hasBreadcrumbBlock && (0, _jsxRuntime.jsx)("div", {
        css: styles?.navbar,
        children: renderBreadcrumb
      }), !this.hasBreadcrumbBlock && !this.isInPlaceDialogOpen && (0, _jsxRuntime.jsxs)("div", {
        css: styles?.navbar,
        children: [this.props.renderNavbarStartDangerousHack && this.props.renderNavbarStartDangerousHack, this.renderMenuTrigger(), this.hasActionItemsBlock(renderActionItems) && renderActionItems]
      }), !this.hasBreadcrumbBlock && this.renderInPlaceDialog(), !this.hasBreadcrumbBlock && this.renderDropdownMenuTray()]
    });
  }
}) || _class) || _class);
var _default = exports.default = TopNavBarSmallViewportLayout;
