// @instructure/ui-number-input/lib/NumberInput/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.NumberInput = void 0;
var _react = require("react");
var _keycode = _interopRequireDefault(require("keycode"));
var _v11_ = require("@instructure/ui-form-field/v11_6");
var _IconArrowOpenDownLine = require("@instructure/ui-icons/lib/generated/IconArrowOpenDownLine.js");
var _IconArrowOpenUpLine = require("@instructure/ui-icons/lib/generated/IconArrowOpenUpLine.js");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _pickProps = require("@instructure/ui-react-utils/lib/pickProps.js");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _getInteraction = require("@instructure/ui-react-utils/lib/getInteraction.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _hasVisibleChildren = require("@instructure/ui-a11y-utils/lib/hasVisibleChildren.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
id: NumberInput
---
**/
let NumberInput = exports.NumberInput = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = _dec2(_class = class NumberInput extends _react.Component {
  static displayName = "NumberInput";
  static componentId = 'NumberInput';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    // Leave interaction default undefined so that `disabled` and `readOnly` can also be supplied
    interaction: undefined,
    messages: [],
    isRequired: false,
    showArrows: true,
    size: 'medium',
    display: 'block',
    textAlign: 'start',
    inputMode: 'numeric',
    allowStringValue: false
  };
  state = {
    hasFocus: false
  };
  ref = null;
  _input = null;
  _id;
  handleRef = el => {
    this.ref = el;
  };
  get id() {
    if (this.props.id) {
      return this.props.id;
    }
    if (!this._id) {
      this._id = this.props.deterministicId();
    }
    return this._id;
  }
  get invalid() {
    return !!this.props.messages && this.props.messages.some(message => message.type === 'error' || message.type === 'newError');
  }
  get interaction() {
    return (0, _getInteraction.getInteraction)({
      props: this.props
    });
  }
  componentDidMount() {
    this.props.makeStyles?.(this.makeStyleVariables);
  }
  componentDidUpdate() {
    this.props.makeStyles?.(this.makeStyleVariables);
  }
  get makeStyleVariables() {
    return {
      interaction: this.interaction,
      hasFocus: this.state.hasFocus,
      invalid: this.invalid
    };
  }
  handleInputRef = element => {
    this._input = element;
    if (typeof this.props.inputRef === 'function') {
      this.props.inputRef(element);
    }
  };
  handleFocus = event => {
    this.setState({
      hasFocus: true
    });
    if (typeof this.props.onFocus === 'function') {
      this.props.onFocus(event);
    }
  };
  handleBlur = event => {
    this.setState({
      hasFocus: false
    });
    if (typeof this.props.onBlur === 'function') {
      this.props.onBlur(event);
    }
  };
  handleChange = event => {
    if (typeof this.props.onChange === 'function') {
      this.props.onChange(event, event.target.value);
    }
  };
  handleKeyDown = event => {
    const {
      onKeyDown,
      onDecrement,
      onIncrement
    } = this.props;
    if (typeof onKeyDown === 'function') {
      onKeyDown(event);
    }
    if (event.keyCode === _keycode.default.codes.down) {
      event.preventDefault();
      if (typeof onDecrement === 'function') {
        onDecrement(event);
      }
    } else if (event.keyCode === _keycode.default.codes.up) {
      event.preventDefault();
      if (typeof onIncrement === 'function') {
        onIncrement(event);
      }
    }
  };
  handleClickUpArrow = event => {
    this.arrowClicked(event, this.props.onIncrement);
  };
  handleClickDownArrow = event => {
    this.arrowClicked(event, this.props.onDecrement);
  };
  arrowClicked(event, callback) {
    const {
      interaction
    } = this;
    event.preventDefault();
    if (interaction === 'enabled') {
      this._input?.focus();
      if (typeof callback === 'function') {
        callback(event);
      }
    }
  }
  renderArrows(customIcons) {
    return (0, _jsxRuntime.jsxs)("span", {
      css: this.props.styles?.arrowContainer,
      children: [(0, _jsxRuntime.jsx)("button", {
        "aria-hidden": true,
        css: this.props.styles?.arrow,
        onMouseDown: this.handleClickUpArrow,
        tabIndex: -1,
        type: "button",
        children: customIcons?.increase ? (0, _callRenderProp.callRenderProp)(customIcons.increase) : (0, _jsxRuntime.jsx)(_IconArrowOpenUpLine.IconArrowOpenUpLine, {})
      }), (0, _jsxRuntime.jsx)("button", {
        "aria-hidden": true,
        css: this.props.styles?.arrow,
        onMouseDown: this.handleClickDownArrow,
        tabIndex: -1,
        type: "button",
        children: customIcons?.decrease ? (0, _callRenderProp.callRenderProp)(customIcons.decrease) : (0, _jsxRuntime.jsx)(_IconArrowOpenDownLine.IconArrowOpenDownLine, {})
      })]
    });
  }
  render() {
    const {
      renderLabel,
      display,
      placeholder,
      isRequired,
      showArrows,
      value,
      width,
      styles,
      allowStringValue,
      renderIcons,
      margin
    } = this.props;
    const {
      interaction
    } = this;
    const rawLabel = (0, _callRenderProp.callRenderProp)(renderLabel);
    const label = (0, _hasVisibleChildren.hasVisibleChildren)(rawLabel) ? (0, _jsxRuntime.jsxs)(_react.Fragment, {
      children: [rawLabel, isRequired && (0, _jsxRuntime.jsxs)("span", {
        css: this.invalid ? styles?.requiredInvalid : {},
        "aria-hidden": true,
        children: [' ', "*"]
      })]
    }) : rawLabel;
    return (0, _jsxRuntime.jsx)(_v11_.FormField, {
      ...(0, _pickProps.pickProps)(this.props, _v11_.FormField.allowedProps),
      label: label,
      inline: display === 'inline-block',
      id: this.id,
      elementRef: this.handleRef,
      margin: margin,
      "data-cid": "NumberInput",
      children: (0, _jsxRuntime.jsx)("span", {
        css: this.props.styles?.inputWidth,
        style: width ? {
          width
        } : undefined,
        children: (0, _jsxRuntime.jsxs)("span", {
          css: this.props.styles?.inputContainer,
          children: [(0, _jsxRuntime.jsx)("input", {
            ...(0, _omitProps.omitProps)(this.props, [..._v11_.FormField.allowedProps, ...NumberInput.allowedProps]),
            css: this.props.styles?.input,
            "aria-invalid": this.invalid ? 'true' : undefined,
            id: this.id,
            type: allowStringValue ? 'text' : 'number',
            inputMode: this.props.inputMode,
            placeholder: placeholder,
            ref: this.handleInputRef,
            required: isRequired,
            value: value,
            disabled: interaction === 'disabled',
            readOnly: interaction === 'readonly',
            onFocus: this.handleFocus,
            onBlur: this.handleBlur,
            onChange: this.handleChange,
            onKeyDown: this.handleKeyDown
          }), showArrows ? this.renderArrows(renderIcons) : null]
        })
      })
    });
  }
}) || _class) || _class);
var _default = exports.default = NumberInput;
