// @instructure/ui-themes/lib/legacySharedThemeTokens/colors/dataVisualization.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _primitives = require("./primitives");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const dataVisualization = {
  rose12Primary: _primitives.additionalPrimitives.rose12,
  rose30Primary: _primitives.additionalPrimitives.rose30,
  rose45Primary: _primitives.additionalPrimitives.rose45,
  rose70Primary: _primitives.additionalPrimitives.rose70,
  rose110Primary: _primitives.additionalPrimitives.rose110,
  rose35Secondary: _primitives.additionalPrimitives.rose35,
  rose40Secondary: _primitives.additionalPrimitives.rose40,
  rose50Secondary: _primitives.additionalPrimitives.rose50,
  rose57Secondary: _primitives.additionalPrimitives.rose57,
  rose90Secondary: _primitives.additionalPrimitives.rose90,
  copper12Primary: _primitives.additionalPrimitives.copper12,
  copper30Primary: _primitives.additionalPrimitives.copper30,
  copper45Primary: _primitives.additionalPrimitives.copper45,
  copper70Primary: _primitives.additionalPrimitives.copper70,
  copper110Primary: _primitives.additionalPrimitives.copper110,
  copper35Secondary: _primitives.additionalPrimitives.copper35,
  copper40Secondary: _primitives.additionalPrimitives.copper40,
  copper50Secondary: _primitives.additionalPrimitives.copper50,
  copper57Secondary: _primitives.additionalPrimitives.copper57,
  copper90Secondary: _primitives.additionalPrimitives.copper90,
  honey12Primary: _primitives.additionalPrimitives.honey12,
  honey30Primary: _primitives.additionalPrimitives.honey30,
  honey45Primary: _primitives.additionalPrimitives.honey45,
  honey70Primary: _primitives.additionalPrimitives.honey70,
  honey110Primary: _primitives.additionalPrimitives.honey110,
  honey35Secondary: _primitives.additionalPrimitives.honey35,
  honey40Secondary: _primitives.additionalPrimitives.honey40,
  honey50Secondary: _primitives.additionalPrimitives.honey50,
  honey57Secondary: _primitives.additionalPrimitives.honey57,
  honey90Secondary: _primitives.additionalPrimitives.honey90,
  forest12Primary: _primitives.additionalPrimitives.forest12,
  forest30Primary: _primitives.additionalPrimitives.forest30,
  forest45Primary: _primitives.additionalPrimitives.forest45,
  forest70Primary: _primitives.additionalPrimitives.forest70,
  forest110Primary: _primitives.additionalPrimitives.forest110,
  forest35Secondary: _primitives.additionalPrimitives.forest35,
  forest40Secondary: _primitives.additionalPrimitives.forest40,
  forest50Secondary: _primitives.additionalPrimitives.forest50,
  forest57Secondary: _primitives.additionalPrimitives.forest57,
  forest90Secondary: _primitives.additionalPrimitives.forest90,
  aurora12Primary: _primitives.additionalPrimitives.aurora12,
  aurora30Primary: _primitives.additionalPrimitives.aurora30,
  aurora45Primary: _primitives.additionalPrimitives.aurora45,
  aurora70Primary: _primitives.additionalPrimitives.aurora70,
  aurora110Primary: _primitives.additionalPrimitives.aurora110,
  aurora35Secondary: _primitives.additionalPrimitives.aurora35,
  aurora40Secondary: _primitives.additionalPrimitives.aurora40,
  aurora50Secondary: _primitives.additionalPrimitives.aurora50,
  aurora57Secondary: _primitives.additionalPrimitives.aurora57,
  aurora90Secondary: _primitives.additionalPrimitives.aurora90,
  sea12Primary: _primitives.additionalPrimitives.sea12,
  sea30Primary: _primitives.additionalPrimitives.sea30,
  sea45Primary: _primitives.additionalPrimitives.sea45,
  sea70Primary: _primitives.additionalPrimitives.sea70,
  sea110Primary: _primitives.additionalPrimitives.sea110,
  sea35Secondary: _primitives.additionalPrimitives.sea35,
  sea40Secondary: _primitives.additionalPrimitives.sea40,
  sea50Secondary: _primitives.additionalPrimitives.sea50,
  sea57Secondary: _primitives.additionalPrimitives.sea57,
  sea90Secondary: _primitives.additionalPrimitives.sea90,
  sky12Primary: _primitives.additionalPrimitives.sky12,
  sky30Primary: _primitives.additionalPrimitives.sky30,
  sky45Primary: _primitives.additionalPrimitives.sky45,
  sky70Primary: _primitives.additionalPrimitives.sky70,
  sky110Primary: _primitives.additionalPrimitives.sky110,
  sky35Secondary: _primitives.additionalPrimitives.sky35,
  sky40Secondary: _primitives.additionalPrimitives.sky40,
  sky50Secondary: _primitives.additionalPrimitives.sky50,
  sky57Secondary: _primitives.additionalPrimitives.sky57,
  sky90Secondary: _primitives.additionalPrimitives.sky90,
  ocean12Primary: _primitives.additionalPrimitives.ocean12,
  ocean30Primary: _primitives.additionalPrimitives.ocean30,
  ocean45Primary: _primitives.additionalPrimitives.ocean45,
  ocean70Primary: _primitives.additionalPrimitives.ocean70,
  ocean110Primary: _primitives.additionalPrimitives.ocean110,
  ocean35Secondary: _primitives.additionalPrimitives.ocean35,
  ocean40Secondary: _primitives.additionalPrimitives.ocean40,
  ocean50Secondary: _primitives.additionalPrimitives.ocean50,
  ocean57Secondary: _primitives.additionalPrimitives.ocean57,
  ocean90Secondary: _primitives.additionalPrimitives.ocean90,
  violet12Primary: _primitives.additionalPrimitives.violet12,
  violet30Primary: _primitives.additionalPrimitives.violet30,
  violet45Primary: _primitives.additionalPrimitives.violet45,
  violet70Primary: _primitives.additionalPrimitives.violet70,
  violet110Primary: _primitives.additionalPrimitives.violet110,
  violet35Secondary: _primitives.additionalPrimitives.violet35,
  violet40Secondary: _primitives.additionalPrimitives.violet40,
  violet50Secondary: _primitives.additionalPrimitives.violet50,
  violet57Secondary: _primitives.additionalPrimitives.violet57,
  violet90Secondary: _primitives.additionalPrimitives.violet90,
  plum12Primary: _primitives.additionalPrimitives.plum12,
  plum30Primary: _primitives.additionalPrimitives.plum30,
  plum45Primary: _primitives.additionalPrimitives.plum45,
  plum70Primary: _primitives.additionalPrimitives.plum70,
  plum110Primary: _primitives.additionalPrimitives.plum110,
  plum35Secondary: _primitives.additionalPrimitives.plum35,
  plum40Secondary: _primitives.additionalPrimitives.plum40,
  plum50Secondary: _primitives.additionalPrimitives.plum50,
  plum57Secondary: _primitives.additionalPrimitives.plum57,
  plum90Secondary: _primitives.additionalPrimitives.plum90,
  stone12Primary: _primitives.additionalPrimitives.stone12,
  stone30Primary: _primitives.additionalPrimitives.stone30,
  stone45Primary: _primitives.additionalPrimitives.stone45,
  stone70Primary: _primitives.additionalPrimitives.stone70,
  stone110Primary: _primitives.additionalPrimitives.stone110,
  stone35Secondary: _primitives.additionalPrimitives.stone35,
  stone40Secondary: _primitives.additionalPrimitives.stone40,
  stone50Secondary: _primitives.additionalPrimitives.stone50,
  stone57Secondary: _primitives.additionalPrimitives.stone57,
  stone90Secondary: _primitives.additionalPrimitives.stone90
};
var _default = exports.default = dataVisualization;
