// @instructure/ui-instructure/lib/AiInformation/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.AiInformation = void 0;
var _react = require("react");
var _v11_ = require("@instructure/ui-popover/v11_6");
var _v11_2 = require("@instructure/ui-buttons/v11_6");
var _v11_3 = require("@instructure/ui-heading/v11_6");
var _v11_4 = require("@instructure/ui-text/v11_6");
var _v11_5 = require("@instructure/ui-link/v11_6");
var _IconExternalLinkLine = require("@instructure/ui-icons/lib/generated/IconExternalLinkLine.js");
var _emotion = require("@instructure/emotion");
var _v = require("../../NutritionFacts/v1");
var _v2 = require("../../DataPermissionLevels/v1");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
category: components/AI Components
---
**/const AiInformation = ({
  title,
  data,
  trigger,
  fullscreenModals = false,
  dataPermissionLevelsModalLabel,
  dataPermissionLevelsTitle,
  dataPermissionLevelsData,
  dataPermissionLevelsCloseButtonText,
  dataPermissionLevelsCloseIconButtonScreenReaderLabel,
  dataPermissionLevelsCurrentFeature,
  dataPermissionLevelsCurrentFeatureText,
  nutritionFactsModalLabel,
  nutritionFactsTitle,
  nutritionFactsFeatureName,
  nutritionFactsData,
  nutritionFactsCloseButtonText,
  nutritionFactsCloseIconButtonScreenReaderLabel
}) => {
  const [open, setOpen] = (0, _react.useState)(false);
  const styles = (0, _emotion.useStyle)({
    generateStyle: _styles.default,
    generateComponentTheme: _theme.default,
    componentId: 'AiInformation',
    displayName: 'AiInformation'
  });
  return (0, _jsxRuntime.jsx)("div", {
    children: (0, _jsxRuntime.jsx)(_v11_.Popover, {
      renderTrigger: trigger,
      isShowingContent: open,
      on: "click",
      screenReaderLabel: "Popover Dialog Example",
      shouldContainFocus: true,
      shouldReturnFocus: true,
      shouldCloseOnDocumentClick: true,
      offsetY: "1rem",
      onShowContent: () => setOpen(true),
      onHideContent: () => setOpen(false),
      children: (0, _jsxRuntime.jsxs)("div", {
        css: styles?.aiInformation,
        children: [(0, _jsxRuntime.jsxs)("div", {
          css: styles?.header,
          children: [(0, _jsxRuntime.jsx)(_v11_3.Heading, {
            level: "h2",
            aiVariant: "stacked",
            variant: "titleModule",
            children: title
          }), (0, _jsxRuntime.jsx)(_v11_2.CloseButton, {
            placement: "end",
            offset: "small",
            onClick: () => setOpen(false),
            screenReaderLabel: "Close"
          })]
        }), (0, _jsxRuntime.jsx)("div", {
          children: data.map(({
            featureName,
            privacyNoticeText,
            privacyNoticeUrl,
            permissionLevelText,
            permissionLevel,
            description,
            permissionLevelsModalTriggerText,
            modelNameText,
            modelName,
            nutritionFactsModalTriggerText
          }, index) => (0, _jsxRuntime.jsxs)("div", {
            children: [(0, _jsxRuntime.jsx)("div", {
              css: styles?.privacyNotice,
              children: (0, _jsxRuntime.jsx)(_v11_5.Link, {
                href: privacyNoticeUrl,
                renderIcon: (0, _jsxRuntime.jsx)(_IconExternalLinkLine.IconExternalLinkLine, {}),
                iconPlacement: "end",
                variant: "standalone",
                target: "_blank",
                children: privacyNoticeText
              })
            }), (0, _jsxRuntime.jsx)("div", {
              css: styles?.featureName,
              children: (0, _jsxRuntime.jsx)(_v11_3.Heading, {
                level: "h3",
                variant: "titleCardRegular",
                children: featureName
              })
            }), (0, _jsxRuntime.jsx)("div", {
              css: styles?.permissionLevelText,
              children: (0, _jsxRuntime.jsx)(_v11_3.Heading, {
                level: "h4",
                variant: "label",
                children: permissionLevelText
              })
            }), (0, _jsxRuntime.jsx)("div", {
              css: styles?.permissionLevel,
              children: (0, _jsxRuntime.jsxs)(_v11_4.Text, {
                variant: "legend",
                children: [" ", permissionLevel, " "]
              })
            }), (0, _jsxRuntime.jsx)("div", {
              css: styles?.description,
              children: (0, _jsxRuntime.jsxs)(_v11_4.Text, {
                variant: "contentSmall",
                children: [" ", description, " "]
              })
            }), (0, _jsxRuntime.jsx)("div", {
              css: styles?.permissionLevelsModalTriggerText,
              children: (0, _jsxRuntime.jsx)(_v2.DataPermissionLevels, {
                fullscreen: fullscreenModals,
                title: dataPermissionLevelsTitle,
                currentFeatureText: dataPermissionLevelsCurrentFeatureText,
                currentFeature: dataPermissionLevelsCurrentFeature,
                closeIconButtonScreenReaderLabel: dataPermissionLevelsCloseIconButtonScreenReaderLabel,
                closeButtonText: dataPermissionLevelsCloseButtonText,
                modalLabel: dataPermissionLevelsModalLabel,
                triggerText: permissionLevelsModalTriggerText,
                data: dataPermissionLevelsData
              })
            }), (0, _jsxRuntime.jsx)("div", {
              css: styles?.modelNameText,
              children: (0, _jsxRuntime.jsxs)(_v11_3.Heading, {
                level: "h3",
                variant: "label",
                children: [' ', modelNameText, ' ']
              })
            }), (0, _jsxRuntime.jsx)("div", {
              css: styles?.modelName,
              children: (0, _jsxRuntime.jsxs)(_v11_4.Text, {
                variant: "contentSmall",
                children: [" ", modelName, " "]
              })
            }), (0, _jsxRuntime.jsx)(_v.NutritionFacts, {
              fullscreen: fullscreenModals,
              modalLabel: nutritionFactsModalLabel,
              title: nutritionFactsTitle,
              featureName: nutritionFactsFeatureName,
              closeButtonText: nutritionFactsCloseButtonText,
              closeIconButtonScreenReaderLabel: nutritionFactsCloseIconButtonScreenReaderLabel,
              triggerText: nutritionFactsModalTriggerText,
              data: nutritionFactsData
            }), data.length !== index + 1 ? (0, _jsxRuntime.jsx)("div", {
              css: styles?.divider
            }) : null]
          }, index))
        })]
      })
    })
  });
};
exports.AiInformation = AiInformation;
var _default = exports.default = AiInformation;
