// @instructure/ui-color-picker/lib/ColorMixer/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ColorMixer = void 0;
var _react = require("react");
var _emotion = require("@instructure/emotion");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _isValid = require("@instructure/ui-color-utils/lib/isValid.js");
var _uiColorUtils = _interopRequireDefault(require("@instructure/ui-color-utils"));
var _console = require("@instructure/console");
var _ColorPalette = _interopRequireDefault(require("./ColorPalette"));
var _Slider = _interopRequireDefault(require("./Slider"));
var _RGBAInput = _interopRequireDefault(require("./RGBAInput"));
var _props = require("./props");
var _styles = _interopRequireDefault(require("./styles"));
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let ColorMixer = exports.ColorMixer = (_dec = (0, _emotion.withStyle)(_styles.default, null), _dec(_class = class ColorMixer extends _react.Component {
  static displayName = "ColorMixer";
  static allowedProps = _props.allowedProps;
  static componentId = 'ColorMixer';
  static defaultProps = {
    value: '#000',
    withAlpha: false,
    disabled: false
  };
  constructor(props) {
    super(props);
    this.state = {
      h: 0,
      s: 0,
      v: 0,
      a: 1
    };
  }
  ref = null;
  _width = 272;
  _paletteHeight = 160;
  _sliderHeight = 8;
  _sliderIndicatorRadius = 6;
  _paletteIndicatorRadius = 6;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.();
    (0, _console.logWarn)((0, _isValid.isValid)(this.props.value), `[ColorMixer] The passed color value is not valid.`);
    this.setState({
      ..._uiColorUtils.default.colorToHsva(this.props.value)
    });
  }
  componentDidUpdate(prevProps, prevState) {
    this.props.makeStyles?.();
    const {
      h,
      s,
      v,
      a
    } = this.state;
    if (prevState.h !== h || prevState.s !== s || prevState.v !== v || prevState.a !== a) {
      this.props.onChange(_uiColorUtils.default.colorToHex8({
        h,
        s,
        v,
        a
      }));
    }
    if (prevProps.value !== this.props.value && _uiColorUtils.default.colorToHex8({
      h,
      s,
      v,
      a
    }) !== this.props.value) {
      this.setState({
        ..._uiColorUtils.default.colorToHsva(this.props.value)
      });
    }
  }
  render() {
    const {
      disabled,
      styles,
      withAlpha,
      rgbRedInputScreenReaderLabel,
      rgbGreenInputScreenReaderLabel,
      rgbBlueInputScreenReaderLabel,
      rgbAlphaInputScreenReaderLabel,
      colorSliderNavigationExplanationScreenReaderLabel,
      alphaSliderNavigationExplanationScreenReaderLabel,
      colorPaletteNavigationExplanationScreenReaderLabel
    } = this.props;
    const {
      h,
      s,
      v,
      a
    } = this.state;
    return (0, _jsxRuntime.jsxs)("div", {
      ...(0, _omitProps.omitProps)(this.props, ColorMixer.allowedProps),
      "aria-disabled": disabled,
      ref: this.handleRef,
      css: styles?.colorMixer,
      "data-cid": "ColorMixer",
      children: [(0, _jsxRuntime.jsxs)("span", {
        css: styles?.sliderAndPaletteContainer,
        "aria-label": `${_uiColorUtils.default.colorToHex8({
          h,
          s,
          v,
          a
        })}`,
        "aria-live": "polite",
        children: [(0, _jsxRuntime.jsx)(_ColorPalette.default, {
          disabled: disabled,
          width: this._width,
          height: this._paletteHeight,
          indicatorRadius: this._paletteIndicatorRadius,
          hue: h,
          color: {
            h,
            s,
            v
          },
          onChange: color => {
            this.setState({
              s: color.s,
              v: color.v
            });
          },
          navigationExplanationScreenReaderLabel: colorPaletteNavigationExplanationScreenReaderLabel
        }), (0, _jsxRuntime.jsx)(_Slider.default, {
          disabled: disabled,
          isColorSlider: true,
          width: this._width,
          height: this._sliderHeight,
          indicatorRadius: this._sliderIndicatorRadius,
          value: h,
          minValue: 0,
          maxValue: 359,
          color: _uiColorUtils.default.colorToHex8({
            h,
            s,
            v,
            a
          }),
          onChange: hue => {
            this.setState({
              h: hue
            });
          },
          navigationExplanationScreenReaderLabel: colorSliderNavigationExplanationScreenReaderLabel
        }), withAlpha && (0, _jsxRuntime.jsx)(_Slider.default, {
          disabled: disabled,
          width: this._width,
          height: this._sliderHeight,
          indicatorRadius: this._sliderIndicatorRadius,
          color: _uiColorUtils.default.colorToHex8({
            h,
            s,
            v
          }),
          value: a,
          minValue: 0,
          maxValue: 1,
          onChange: opacity => this.setState({
            a: opacity / 100
          }),
          navigationExplanationScreenReaderLabel: alphaSliderNavigationExplanationScreenReaderLabel
        })]
      }), (0, _jsxRuntime.jsx)(_RGBAInput.default, {
        disabled: disabled,
        label: withAlpha ? 'RGBA' : 'RGB',
        width: this._width,
        value: _uiColorUtils.default.colorToRGB({
          h,
          s,
          v,
          a
        }),
        onChange: color => this.setState({
          ..._uiColorUtils.default.colorToHsva(color)
        }),
        withAlpha: withAlpha,
        rgbRedInputScreenReaderLabel: rgbRedInputScreenReaderLabel,
        rgbGreenInputScreenReaderLabel: rgbGreenInputScreenReaderLabel,
        rgbBlueInputScreenReaderLabel: rgbBlueInputScreenReaderLabel,
        rgbAlphaInputScreenReaderLabel: rgbAlphaInputScreenReaderLabel
      })]
    });
  }
}) || _class);
var _default = exports.default = ColorMixer;
