// @instructure/ui-color-picker/lib/ColorContrast/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ColorContrast = void 0;
var _react = require("react");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _console = require("@instructure/console");
var _contrastWithAlpha = require("@instructure/ui-color-utils/lib/contrastWithAlpha.js");
var _validateContrast = require("@instructure/ui-color-utils/lib/validateContrast.js");
var _emotion = require("@instructure/emotion");
var _v11_ = require("@instructure/ui-text/v11_6");
var _v11_2 = require("@instructure/ui-heading/v11_6");
var _v11_3 = require("@instructure/ui-pill/v11_6");
var _v = _interopRequireDefault(require("../../ColorIndicator/v1"));
var _props = require("./props");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let ColorContrast = exports.ColorContrast = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class ColorContrast extends _react.Component {
  static displayName = "ColorContrast";
  static allowedProps = _props.allowedProps;
  static componentId = 'ColorContrast';
  static defaultProps = {
    withoutColorPreview: false,
    validationLevel: 'AA'
  };
  constructor(props) {
    super(props);
    this.state = {
      contrast: 1,
      isValidNormalText: false,
      isValidLargeText: false,
      isValidGraphicsText: false
    };
  }
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.();
    this.calcState();
  }
  componentDidUpdate(prevProps) {
    this.props.makeStyles?.();
    if (prevProps?.firstColor !== this.props?.firstColor || prevProps?.secondColor !== this.props?.secondColor || prevProps?.validationLevel !== this.props?.validationLevel) {
      const newState = this.calcState();
      this.props?.onContrastChange?.({
        contrast: newState.contrast,
        isValidNormalText: newState.isValidNormalText,
        isValidLargeText: newState.isValidLargeText,
        isValidGraphicsText: newState.isValidGraphicsText,
        firstColor: this.props.firstColor,
        secondColor: this.props.secondColor
      });
    }
  }
  renderStatus = (pass, description) => {
    const {
      successLabel,
      failureLabel,
      styles
    } = this.props;
    return (0, _jsxRuntime.jsxs)("div", {
      css: styles?.statusWrapper,
      children: [(0, _jsxRuntime.jsx)("div", {
        css: pass ? styles?.successDescription : styles?.failureDescription,
        children: description
      }), (0, _jsxRuntime.jsx)("div", {
        css: styles?.statusIndicatorWrapper,
        children: (0, _jsxRuntime.jsx)(_v11_3.Pill, {
          color: pass ? 'success' : 'danger',
          children: pass ? successLabel : failureLabel
        })
      })]
    });
  };
  renderColorIndicator = (color, label) => (0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [(0, _jsxRuntime.jsx)("div", {
      children: (0, _jsxRuntime.jsx)("div", {
        css: this.props.styles?.colorIndicator,
        children: (0, _jsxRuntime.jsx)(_v.default, {
          color: color
        })
      })
    }), (0, _jsxRuntime.jsxs)("div", {
      children: [(0, _jsxRuntime.jsx)("div", {
        css: this.props.styles?.colorIndicatorLabel,
        children: label
      }), (0, _jsxRuntime.jsx)("div", {
        css: this.props.styles?.pickedColorHex,
        children: color
      })]
    })]
  });
  renderPreview() {
    const {
      styles,
      withoutColorPreview,
      firstColor,
      secondColor,
      firstColorLabel,
      secondColorLabel
    } = this.props;
    if (withoutColorPreview) {
      return null;
    }
    if (!firstColorLabel || !secondColorLabel) {
      (0, _console.error)(false, 'When `withoutColorPreview` is not set to true, the properties `firstColorLabel` and `secondColorLabel` are required!');
    }
    return !withoutColorPreview && (0, _jsxRuntime.jsxs)("div", {
      css: styles?.colorPreview,
      children: [(0, _jsxRuntime.jsx)("div", {
        css: styles?.firstColorPreview,
        children: this.renderColorIndicator(firstColor, firstColorLabel || '')
      }), (0, _jsxRuntime.jsx)("div", {
        css: styles?.secondColorPreview,
        children: this.renderColorIndicator(secondColor, secondColorLabel || '')
      })]
    });
  }
  calcState() {
    const contrast = (0, _contrastWithAlpha.contrastWithAlpha)(this.props.firstColor, this.props.secondColor);
    const newState = {
      contrast,
      ...(0, _validateContrast.validateContrast)(contrast, this.props.validationLevel)
    };
    this.setState(newState);
    return newState;
  }
  render() {
    const {
      styles,
      label,
      labelLevel,
      normalTextLabel,
      largeTextLabel,
      graphicsTextLabel
    } = this.props;
    const {
      contrast,
      isValidNormalText,
      isValidLargeText,
      isValidGraphicsText
    } = this.state;
    return (0, _jsxRuntime.jsxs)("div", {
      ...(0, _omitProps.omitProps)(this.props, ColorContrast.allowedProps),
      ref: this.handleRef,
      css: styles?.colorContrast,
      "data-cid": "ColorContrast",
      children: [(0, _jsxRuntime.jsx)("div", {
        css: styles?.label,
        children: labelLevel ? (0, _jsxRuntime.jsx)(_v11_2.Heading, {
          as: labelLevel,
          level: "reset",
          children: label
        }) : (0, _jsxRuntime.jsx)(_v11_.Text, {
          weight: "bold",
          as: "div",
          children: label
        })
      }), (0, _jsxRuntime.jsxs)(_v11_.Text, {
        size: "x-large",
        children: [contrast, ":1"]
      }), this.renderPreview(), this.renderStatus(isValidNormalText, normalTextLabel), this.renderStatus(isValidLargeText, largeTextLabel), this.renderStatus(isValidGraphicsText, graphicsTextLabel)]
    });
  }
}) || _class);
var _default = exports.default = ColorContrast;
