// @instructure/ui-flex/lib/Flex/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Flex = void 0;
Object.defineProperty(exports, "FlexItem", {
  enumerable: true,
  get: function () {
    return _Item.Item;
  }
});
exports.default = void 0;
var _react = require("react");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _v11_ = require("@instructure/ui-view/v11_6");
var _emotion = require("@instructure/emotion");
var _Item = require("./Item");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
@module Flex
**/
let Flex = exports.Flex = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class Flex extends _react.Component {
  static displayName = "Flex";
  static componentId = 'Flex';
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  static Item = _Item.Item;
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    as: 'span',
    direction: 'row',
    justifyItems: 'start',
    display: 'flex',
    withVisualDebug: false,
    wrap: 'no-wrap',
    gap: 'none'
  };
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  renderChildren(children) {
    // TODO this might fail if children are () => ReactNode
    return _react.Children.map(children, child => {
      if (!child) {
        return null;
      }
      return (0, _matchComponentTypes.matchComponentTypes)(child, ['Item']) ? (0, _safeCloneElement.safeCloneElement)(child, {
        // child withVisualDebug prop should override parent
        withVisualDebug: this.props.withVisualDebug,
        ...child.props,
        direction: this.props.direction.replace(/-reverse/, '')
      }) : child;
    });
  }
  render() {
    const {
      as,
      withVisualDebug,
      height,
      display,
      margin,
      padding,
      textAlign,
      width,
      styles,
      gap,
      wrap
    } = this.props;
    const children = (0, _callRenderProp.callRenderProp)(this.props.children);
    const gapValuesArray = gap?.split(' ');
    if (gapValuesArray.length > 1 && wrap === 'no-wrap') {}
    if (children && _react.Children.count(children) > 0) {
      return (0, _jsxRuntime.jsx)(_v11_.View, {
        ...(0, _passthroughProps.passthroughProps)(this.props),
        css: styles?.flex,
        elementRef: this.handleRef,
        as: as,
        display: display,
        width: width,
        height: height,
        margin: margin,
        padding: padding,
        textAlign: textAlign,
        withVisualDebug: withVisualDebug,
        children: this.renderChildren(children)
      });
    } else {
      return null;
    }
  }
}) || _class);
var _default = exports.default = Flex;
