// @instructure/ui-checkbox/lib/Checkbox/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Checkbox = void 0;
Object.defineProperty(exports, "CheckboxFacade", {
  enumerable: true,
  get: function () {
    return _CheckboxFacade.CheckboxFacade;
  }
});
Object.defineProperty(exports, "ToggleFacade", {
  enumerable: true,
  get: function () {
    return _ToggleFacade.ToggleFacade;
  }
});
exports.default = void 0;
var _react = require("react");
var _keycode = _interopRequireDefault(require("keycode"));
var _v11_ = require("@instructure/ui-form-field/v11_6");
var _createChainedFunction = require("@instructure/ui-utils/lib/createChainedFunction.js");
var _console = require("@instructure/console");
var _isActiveElement = require("@instructure/ui-dom-utils/lib/isActiveElement.js");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _v11_2 = require("@instructure/ui-view/v11_6");
var _emotion = require("@instructure/emotion");
var _CheckboxFacade = require("./CheckboxFacade");
var _ToggleFacade = require("./ToggleFacade");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
tags: toggle, switch
---
**/
let Checkbox = exports.Checkbox = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = _dec2(_class = class Checkbox extends _react.Component {
  static displayName = "Checkbox";
  static componentId = 'Checkbox';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    size: 'medium',
    variant: 'simple',
    disabled: false,
    inline: false,
    indeterminate: false,
    readOnly: false,
    labelPlacement: 'end'
  };
  constructor(props) {
    super(props);
    this.state = {
      focused: false,
      hovered: false,
      checked: typeof props.checked === 'undefined' ? !!props.defaultChecked : undefined
    };
    if (typeof props.checked === 'undefined') {
      // @ts-expect-error ts-migrate(2339) FIXME: Property 'checked' does not exist on type 'Readonl... Remove this comment to see the full error message
      this.state.checked = !!props.defaultChecked;
    }
    this._defaultId = props.deterministicId();
  }
  _defaultId;
  _input = null;
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  handleInputRef = el => {
    this._input = el;
    if (typeof this.props.inputRef === 'function') {
      this.props.inputRef(el);
    }
  };
  componentDidMount() {
    if (this._input) {
      this._input.indeterminate = this.props.indeterminate;
    }
    this.props.makeStyles?.();
  }
  componentDidUpdate(prevProps) {
    if (prevProps.indeterminate !== this.props.indeterminate && this._input) {
      this._input.indeterminate = this.props.indeterminate || false;
    }
    this.props.makeStyles?.();
  }
  handleChange = e => {
    const {
      onChange,
      disabled,
      checked,
      readOnly
    } = this.props;
    if (disabled || readOnly) {
      e.preventDefault();
      return;
    }
    if (typeof checked === 'undefined') {
      this.setState({
        checked: !this.state.checked
      });
    }
    if (typeof onChange === 'function') {
      onChange(e);
    }
  };
  handleKeyDown = e => {
    if (this.props.variant === 'toggle' && e.keyCode === _keycode.default.codes.enter) {
      this._input && this._input.click();
      e.preventDefault();
    }
  };
  handleFocus = () => {
    this.setState({
      focused: true
    });
  };
  handleBlur = () => {
    this.setState({
      focused: false
    });
  };
  handleMouseOver = () => {
    !this.props.disabled && this.setState({
      hovered: true
    });
  };
  handleMouseOut = () => {
    this.setState({
      hovered: false
    });
  };
  get id() {
    return this.props.id || this._defaultId;
  }
  get checked() {
    return typeof this.props.checked === 'undefined' ? this.state.checked : this.props.checked;
  }
  get focused() {
    return (0, _isActiveElement.isActiveElement)(this._input);
  }
  get isNewError() {
    return !!this.props.messages?.find(m => m.type === 'newError');
  }
  get invalid() {
    return !!this.props.messages?.find(m => m.type === 'newError' || m.type === 'error');
  }
  focus() {
    this._input && this._input.focus();
  }
  renderFacade() {
    const {
      size,
      disabled,
      variant,
      label,
      readOnly,
      indeterminate,
      labelPlacement,
      themeOverride,
      isRequired,
      styles
    } = this.props;
    const {
      hovered,
      focused
    } = this.state;
    (0, _console.logError)(!(variant === 'simple' && labelPlacement !== 'end'), `[Checkbox] The \`simple\` variant does not support the \`labelPlacement\` property.  Use the \`toggle\` variant instead.`);
    if (variant === 'toggle') {
      return (0, _jsxRuntime.jsxs)(_ToggleFacade.ToggleFacade, {
        disabled: disabled,
        size: size,
        focused: focused,
        checked: this.checked,
        readOnly: readOnly,
        labelPlacement: labelPlacement,
        themeOverride: themeOverride,
        invalid: this.invalid,
        children: [label, isRequired && label && (0, _jsxRuntime.jsxs)("span", {
          css: this.invalid ? styles?.requiredInvalid : {},
          "aria-hidden": true,
          children: [' ', "*"]
        })]
      });
    } else {
      return (0, _jsxRuntime.jsxs)(_CheckboxFacade.CheckboxFacade, {
        size: size,
        hovered: hovered,
        focused: focused,
        checked: this.checked,
        indeterminate: indeterminate,
        themeOverride: themeOverride,
        invalid: this.invalid,
        children: [label, isRequired && label && (0, _jsxRuntime.jsxs)("span", {
          css: this.invalid ? styles?.requiredInvalid : {},
          "aria-hidden": true,
          children: [' ', "*"]
        })]
      });
    }
  }
  renderMessages() {
    const {
      messages,
      styles,
      variant
    } = this.props;
    return messages && messages.length > 0 ? (0, _jsxRuntime.jsx)(_v11_2.View, {
      display: "block",
      margin: "small 0 0",
      css: this.isNewError && (variant === 'toggle' ? styles?.indentedToggleError : styles?.indentedError),
      children: (0, _jsxRuntime.jsx)(_v11_.FormFieldMessages, {
        messages: messages
      })
    }) : null;
  }
  render() {
    const {
      disabled,
      readOnly,
      value,
      onKeyDown,
      onFocus,
      onBlur,
      onMouseOver,
      onMouseOut,
      indeterminate,
      variant,
      styles,
      isRequired
    } = this.props;
    const props = (0, _omitProps.omitProps)(this.props, Checkbox.allowedProps);
    (0, _console.logError)(!(variant === 'toggle' && indeterminate), `[Checkbox] The \`toggle\` variant does not support the \`indeterminate\` property. Use the \`simple\` variant instead.`);
    return (0, _jsxRuntime.jsx)("div", {
      css: styles?.checkbox
      /* eslint-disable-next-line jsx-a11y/mouse-events-have-key-events */,
      onMouseOver: (0, _createChainedFunction.createChainedFunction)(onMouseOver, this.handleMouseOver)
      /* eslint-disable-next-line jsx-a11y/mouse-events-have-key-events */,
      onMouseOut: (0, _createChainedFunction.createChainedFunction)(onMouseOut, this.handleMouseOut),
      ref: this.handleRef,
      "data-cid": "Checkbox",
      children: (0, _jsxRuntime.jsxs)("div", {
        css: styles?.container,
        children: [(0, _jsxRuntime.jsx)("input", {
          ...props,
          id: this.id,
          value: value,
          type: "checkbox",
          ref: this.handleInputRef,
          required: isRequired,
          disabled: disabled || readOnly,
          "aria-checked": indeterminate ? 'mixed' : undefined,
          css: styles?.input,
          onChange: this.handleChange,
          onKeyDown: (0, _createChainedFunction.createChainedFunction)(onKeyDown, this.handleKeyDown),
          onFocus: (0, _createChainedFunction.createChainedFunction)(onFocus, this.handleFocus),
          onBlur: (0, _createChainedFunction.createChainedFunction)(onBlur, this.handleBlur),
          checked: this.checked
        }), (0, _jsxRuntime.jsxs)("label", {
          htmlFor: this.id,
          css: styles?.control,
          children: [this.renderFacade(), this.renderMessages()]
        })]
      })
    });
  }
}) || _class) || _class);
var _default = exports.default = Checkbox;
