// @instructure/ui-file-drop/lib/FileDrop/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.FileDrop = void 0;
var _react = require("react");
var _keycode = _interopRequireDefault(require("keycode"));
var _latest = require("@instructure/ui-form-field/latest");
var _latest2 = require("@instructure/ui-view/latest");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _getInteraction = require("@instructure/ui-react-utils/lib/getInteraction.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _accepts = require("./utils/accepts");
var _getEventFiles = require("./utils/getEventFiles");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
function keyEventIsClickButton(e) {
  return e.keyCode === _keycode.default.codes.space || e.keyCode === _keycode.default.codes.enter;
}

/**
---
category: components
---
**/
let FileDrop = exports.FileDrop = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = _dec2(_class = class FileDrop extends _react.Component {
  static displayName = "FileDrop";
  static componentId = 'FileDrop';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    onClick: function (_e) {},
    shouldEnablePreview: false,
    shouldAllowMultiple: false,
    shouldAllowRepeats: true,
    maxSize: Infinity,
    minSize: 0,
    // Leave interaction default undefined so that `disabled` and `readOnly` can also be supplied
    messages: [],
    display: 'block'
  };
  constructor(props) {
    super(props);
    this.defaultId = props.deterministicId();
    this.messagesId = props.deterministicId('FileDrop-messages');
  }
  componentDidMount() {
    this.props.makeStyles?.(this.makeStyleProps());
  }
  componentDidUpdate() {
    this.props.makeStyles?.(this.makeStyleProps());
  }
  convertToFile(fileLikeItem) {
    if (fileLikeItem instanceof DataTransferItem) return fileLikeItem.getAsFile();else return fileLikeItem;
  }
  makeStyleProps = () => {
    return {
      ...this.props,
      functionallyDisabled: this.functionallyDisabled,
      visuallyDisabled: this.interaction === 'disabled',
      dragRejected: this.state.isDragRejected || this.invalid,
      dragAccepted: this.state.isDragAccepted
    };
  };
  state = {
    isDragAccepted: false,
    isDragRejected: false,
    isFocused: false,
    isFileBrowserDisplayed: false
  };
  enterCounter = 0;
  fileInputEl;
  defaultId = null;
  messagesId = '';
  ref = null;
  handleElementRef = el => {
    this.ref = el;
  };
  get functionallyDisabled() {
    return this.interaction === 'disabled' || this.interaction === 'readonly';
  }
  get interaction() {
    return (0, _getInteraction.getInteraction)({
      props: this.props
    });
  }
  get hasMessages() {
    return this.props.messages ? this.props.messages.length > 0 : false;
  }
  get invalid() {
    if (this.hasMessages) {
      return this.props.messages.findIndex(message => {
        return message.type === 'error' || message.type === 'newError';
      }) >= 0;
    }
    return false;
  }
  getDataTransferItems(e, shouldEnablePreview) {
    let list = [];
    Array.from((0, _getEventFiles.getEventFiles)(e, this.fileInputEl)).forEach(file => {
      const fileObj = this.convertToFile(file);
      if (fileObj) list.push(fileObj);
    });
    if (list.length > 1) {
      list = this.props.shouldAllowMultiple ? list : [list[0]];
    }
    if (shouldEnablePreview) {
      return list.map(file => Object.assign(file, {
        preview: window.URL.createObjectURL(file)
      }));
    }
    return list;
  }
  handleDragEnter = e => {
    e.preventDefault();
    // Count the dropzone and any children that are entered.
    this.enterCounter += 1;
    // Don't trigger onDragEnter for each children after the first one
    if (this.enterCounter > 1) {
      return;
    }
    const allFilesAccepted = this.allFilesAccepted(this.getDataTransferItems(e));
    this.setState({
      isDragAccepted: allFilesAccepted,
      isDragRejected: !allFilesAccepted
    });
    this.props.onDragEnter?.(e);
  };
  handleDragOver = e => {
    e.preventDefault();
    e.stopPropagation();
    try {
      const event = e;
      event.dataTransfer.dropEffect = 'copy';
    } catch (err) {
      // continue regardless of error
    }
    this.props.onDragOver?.(e);
    return false;
  };
  handleDragLeave = e => {
    e.preventDefault();
    this.enterCounter -= 1;
    // Only deactivate once the dropzone and all children was left
    if (this.enterCounter > 0) {
      return;
    }
    this.setState({
      isDragAccepted: false,
      isDragRejected: false
    });
    this.props.onDragLeave?.(e);
  };
  parseFiles(fileList) {
    const accepted = [];
    const rejected = [];
    fileList.forEach(file => {
      const fileObj = this.convertToFile(file);
      if (fileObj && this.fileAccepted(fileObj) && this.fileMatchSize(fileObj)) {
        accepted.push(file);
      } else {
        rejected.push(file);
      }
    });
    return [accepted, rejected];
  }
  handleChange = e => {
    const {
      onDrop,
      onDropAccepted,
      onDropRejected,
      shouldEnablePreview
    } = this.props;
    const fileList = this.getDataTransferItems(e, shouldEnablePreview);
    const [accepted, rejected] = this.parseFiles(fileList);
    e.preventDefault();
    this.enterCounter = 0;
    // When dropping a file the browser does not populate the input's
    // 'files' property, so we need to do it manually. This will also populate
    // the input's 'value' prop, which contains the first file's path
    if (accepted.length > 0 && 'dataTransfer' in e) {
      this.fileInputEl.files = e.dataTransfer.files;
    }
    onDrop && onDrop(accepted, rejected, e);
    if (rejected.length > 0 && onDropRejected) {
      onDropRejected(rejected, e);
    }
    if (accepted.length > 0 && onDropAccepted) {
      onDropAccepted(accepted, e);
    }
    this.setState({
      isDragAccepted: false,
      isDragRejected: false,
      isFileBrowserDisplayed: false
    });
  };
  fileAccepted = file => {
    if (this.props.accept) return (0, _accepts.accepts)(file, this.props.accept);else return true;
  };
  fileMatchSize(file) {
    return file.size <= this.props.maxSize && file.size >= this.props.minSize;
  }
  allFilesAccepted(files) {
    return files.every(this.fileAccepted);
  }
  acceptStr() {
    const {
      accept
    } = this.props;
    return accept ? (0, _accepts.getAcceptList)(accept).join(',') : undefined;
  }
  renderLabel() {
    const {
      renderLabel
    } = this.props;
    return (0, _callRenderProp.callRenderProp)(renderLabel, {
      isDragAccepted: this.state.isDragAccepted,
      isDragRejected: this.state.isDragRejected,
      interaction: this.interaction
    });
  }
  handleRef = el => {
    this.fileInputEl = el;
    if (typeof this.props.inputRef === 'function') {
      this.props.inputRef(el);
    }
  };
  handleBlur = () => {
    this.setState({
      isFocused: false,
      isFileBrowserDisplayed: false
    });
  };
  handleFocus = () => {
    this.setState({
      isFocused: true,
      isFileBrowserDisplayed: false
    });
  };
  handleClick = e => {
    if (this.fileInputEl.value && this.props.shouldAllowRepeats) {
      this.fileInputEl.value = null;
    }
    // focus the input (because FF won't)
    this.fileInputEl?.focus();
    this.props.onClick(e);
    this.setState({
      isFileBrowserDisplayed: true
    });
  };
  handleKeyDown = e => {
    if (this.state.isFocused && keyEventIsClickButton(e)) {
      if (this.props.shouldAllowRepeats) {
        this.fileInputEl.value = null;
      }
    }
  };
  handleKeyUp = e => {
    // This is to handle the case where ESC is pressed inside a Dialog so that
    // closing the file browser dialog doesn't also close the Dialog.
    if (e.keyCode === _keycode.default.codes.esc && this.state.isFileBrowserDisplayed) {
      e.stopPropagation();
      e.nativeEvent.stopImmediatePropagation();
      this.setState({
        isFileBrowserDisplayed: false
      });
    }
  };
  render() {
    const {
      display,
      height,
      width,
      minWidth,
      maxWidth,
      margin,
      onDropAccepted,
      onDropRejected,
      styles,
      ...props
    } = this.props;
    const id = this.props.id || this.defaultId;
    const focusColor = this.state.isDragRejected || this.invalid ? 'danger' : undefined;
    return (0, _jsxRuntime.jsxs)(_latest2.View, {
      display: display,
      position: "relative" // contain visually hidden file input element
      ,
      width: width,
      minWidth: minWidth,
      maxWidth: maxWidth,
      margin: margin,
      height: height,
      elementRef: this.handleElementRef,
      "data-cid": "FileDrop",
      children: [(0, _jsxRuntime.jsx)("label", {
        css: this.props.styles?.fileDropLabel,
        htmlFor: id,
        onDragEnter: this.handleDragEnter,
        onDragOver: this.handleDragOver,
        onDragLeave: this.handleDragLeave,
        onDrop: this.handleChange,
        children: (0, _jsxRuntime.jsx)(_latest2.View, {
          borderRadius: String(styles?.fileDropLayout?.borderRadius),
          display: "block",
          position: "relative",
          withFocusOutline: this.state.isFocused,
          focusColor: focusColor,
          height: height,
          children: (0, _jsxRuntime.jsx)("span", {
            css: this.props.styles?.fileDropLabelContent,
            children: (0, _jsxRuntime.jsx)("span", {
              css: this.props.styles?.fileDropLayout,
              children: (0, _jsxRuntime.jsx)(_latest2.View, {
                height: height,
                children: this.renderLabel()
              })
            })
          })
        })
      }), (0, _jsxRuntime.jsx)("input", {
        ...(0, _passthroughProps.passthroughProps)(props),
        onClick: this.handleClick,
        type: "file",
        css: this.props.styles?.fileDropInput,
        id: id,
        ref: this.handleRef,
        onFocus: this.handleFocus,
        onBlur: this.handleBlur,
        onKeyDown: this.handleKeyDown,
        onKeyUp: this.handleKeyUp,
        multiple: this.props.shouldAllowMultiple,
        accept: this.acceptStr(),
        onChange: this.handleChange,
        "aria-describedby": this.hasMessages ? this.messagesId : undefined,
        "aria-invalid": this.invalid,
        disabled: this.functionallyDisabled
      }), this.hasMessages ? (0, _jsxRuntime.jsx)(_latest2.View, {
        display: "block",
        margin: "small 0 0",
        children: (0, _jsxRuntime.jsx)(_latest.FormFieldMessages, {
          id: this.messagesId,
          messages: this.props.messages
        })
      }) : null]
    });
  }
}) || _class) || _class);
var _default = exports.default = FileDrop;
