// @instructure/ui-form-field/lib/FormField/v2/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.FormField = void 0;
var _react = require("react");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _pickProps = require("@instructure/ui-react-utils/lib/pickProps.js");
var _props = require("../../FormFieldLayout/v2/props");
var _v = require("../../FormFieldLayout/v2");
var _props2 = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
category: components
---
**/
class FormField extends _react.Component {
  static displayName = "FormField";
  static componentId = 'FormField';
  static allowedProps = _props2.allowedProps;
  static defaultProps = {
    inline: false,
    layout: 'stacked',
    labelAlign: 'end',
    vAlign: 'middle'
  };
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  render() {
    return (0, _jsxRuntime.jsx)(_v.FormFieldLayout, {
      ...(0, _omitProps.omitProps)(this.props, FormField.allowedProps),
      ...(0, _pickProps.pickProps)(this.props, _props.allowedProps),
      label: this.props.label,
      vAlign: this.props.vAlign,
      as: "label"
      // This makes the control in focus when the label is clicked
      // This is needed to prevent the wrong element to be focused, e.g.
      // multi selects Tag-s
      ,
      htmlFor: this.props.id,
      elementRef: this.handleRef,
      margin: this.props.margin,
      isRequired: this.props.isRequired,
      disabled: this.props.disabled,
      readOnly: this.props.readOnly
    });
  }
}
exports.FormField = FormField;
var _default = exports.default = FormField;
