// @instructure/ui-buttons/lib/CloseButton/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.CloseButton = void 0;
var _react = require("react");
var _IconXSolid = require("@instructure/ui-icons/lib/generated/IconXSolid.js");
var _ScreenReaderContent = require("@instructure/ui-a11y-content/lib/ScreenReaderContent");
var _getInteraction = require("@instructure/ui-react-utils/lib/getInteraction.js");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _v = require("../../BaseButton/v1");
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let CloseButton = exports.CloseButton = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class CloseButton extends _react.Component {
  static displayName = "CloseButton";
  static componentId = 'CloseButton';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    // Leave interaction default undefined so that `disabled` and `readOnly` can also be supplied
    interaction: undefined,
    type: 'button',
    placement: 'static',
    offset: 'x-small',
    size: 'small',
    margin: '0',
    as: 'button',
    cursor: 'pointer'
  };
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  get interaction() {
    return (0, _getInteraction.getInteraction)({
      props: this.props
    });
  }
  get color() {
    const {
      color
    } = this.props;
    return color === 'primary' ? 'secondary' : color;
  }
  render() {
    const {
      screenReaderLabel,
      elementRef,
      size,
      onClick,
      margin,
      placement,
      offset,
      type,
      as,
      href,
      cursor,
      tabIndex,
      styles,
      ...props
    } = this.props;
    return (0, _jsxRuntime.jsx)("span", {
      ...(0, _passthroughProps.passthroughProps)(props),
      css: styles?.closeButton,
      ref: el => {
        this.ref = el;
      },
      children: (0, _jsxRuntime.jsx)(_v.BaseButton, {
        renderIcon: _IconXSolid.IconXSolid,
        elementRef: this.handleRef,
        interaction: this.interaction,
        type: type,
        ...(this.color ? {
          color: this.color
        } : {}),
        size: size,
        onClick: onClick,
        margin: margin,
        withBorder: false,
        withBackground: false,
        as: as,
        href: href,
        cursor: cursor,
        tabIndex: tabIndex,
        "data-cid": "CloseButton",
        children: (0, _jsxRuntime.jsx)(_ScreenReaderContent.ScreenReaderContent, {
          children: screenReaderLabel
        })
      })
    });
  }
}) || _class);
var _default = exports.default = CloseButton;
