// @instructure/ui-side-nav-bar/es/SideNavBar/v2/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import { Component, Children, isValidElement } from 'react';
import { omitProps, safeCloneElement } from '@instructure/ui-react-utils';
import { PanelLeftCloseInstUIIcon, PanelLeftOpenInstUIIcon, renderIconWithProps } from '@instructure/ui-icons';
import { ScreenReaderContent } from '@instructure/ui-a11y-content';
import { withStyleNew } from '@instructure/emotion';
import { SideNavBarItem } from "./SideNavBarItem/index.js";
import generateStyle from "./styles.js";
import { allowedProps } from "./props.js";
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
const navMinimized = ({
  minimized
}) => ({
  minimized: !minimized
});

/**
---
category: components
---
**/
let SideNavBar = (_dec = withStyleNew(generateStyle), _dec(_class = class SideNavBar extends Component {
  static displayName = "SideNavBar";
  static componentId = 'SideNavBar';
  static allowedProps = allowedProps;
  static defaultProps = {
    children: null,
    defaultMinimized: false,
    // TODO we should investigate later if it used or not
    onClick: function (_e) {}
  };
  static Item = SideNavBarItem;
  ref = null;
  constructor(props) {
    super(props);
    this.state = {
      minimized: this.isControlled(props) ? !!props.minimized : !!props.defaultMinimized
    };
  }
  componentDidMount() {
    this.props.makeStyles?.({
      minimized: this.minimized
    });
  }
  componentDidUpdate() {
    this.props.makeStyles?.({
      minimized: this.minimized
    });
  }
  get minimized() {
    if (this.isControlled()) {
      return !!this.props.minimized;
    }
    return !!this.state.minimized;
  }
  isControlled(props = this.props) {
    return typeof props.minimized === 'boolean';
  }
  handleNavToggle = event => {
    if (!this.isControlled()) {
      this.setState(navMinimized);
    }
    if (typeof this.props.onMinimized === 'function') {
      this.props.onMinimized(event, !this.minimized);
    }
  };
  renderChildren() {
    return Children.map(this.props.children, child => {
      if (! /*#__PURE__*/isValidElement(child)) return null;
      const navItem = safeCloneElement(child, {
        minimized: this.state.minimized
      });
      return _jsx("li", {
        css: this.props.styles?.list,
        children: navItem
      });
    });
  }
  renderIcon() {
    const icon = this.minimized ? PanelLeftOpenInstUIIcon : PanelLeftCloseInstUIIcon;
    const iconColor = this.props.selected ? 'sideNavActiveColor' : 'sideNavColor';
    return renderIconWithProps(icon, 'lg', iconColor); //TODO-rework remove after merging into v12
  }
  toggleMessage() {
    return this.state.minimized ? this.props.toggleLabel.minimizedLabel : this.props.toggleLabel.expandedLabel;
  }
  render() {
    const {
      label
    } = this.props;
    const props = omitProps(this.props, SideNavBar.allowedProps, ['minimized']);
    return _jsxs("nav", {
      ...props,
      css: this.props.styles?.navigation,
      "data-cid": "SideNavBar",
      "aria-label": label,
      ref: element => {
        this.ref = element;
      },
      children: [_jsx("ul", {
        css: this.props.styles?.content,
        children: this.renderChildren()
      }), _jsx(SideNavBarItem, {
        "aria-expanded": !this.minimized,
        onClick: this.handleNavToggle,
        icon: this.renderIcon(),
        label: _jsx(ScreenReaderContent, {
          children: this.toggleMessage()
        })
      })]
    });
  }
}) || _class);
export default SideNavBar;
export { SideNavBar, SideNavBarItem };
