// @instructure/ui-form-field/lib/FormFieldMessages/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.FormFieldMessages = void 0;
var _react = require("react");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _emotion = require("@instructure/emotion");
var _v = require("../../FormFieldMessage/v1");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: FormField
---

A FormFieldMessages component

```js
---
type: example
---
<FormFieldMessages messages={[
  { text: 'Invalid name', type: 'error' },
  { text: 'Good job!', type: 'success' },
  { text: 'Full name, first and last', type: 'hint' },
]} />
```
**/
let FormFieldMessages = exports.FormFieldMessages = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class FormFieldMessages extends _react.Component {
  static displayName = "FormFieldMessages";
  static componentId = 'FormFieldMessages';
  static allowedProps = _props.allowedProps;
  static defaultProps = {};
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  render() {
    const {
      messages,
      styles
    } = this.props;
    return messages && messages.length > 0 ? (0, _jsxRuntime.jsx)("div", {
      css: styles?.formFieldMessages,
      ...(0, _omitProps.omitProps)(this.props, FormFieldMessages.allowedProps),
      ref: this.handleRef,
      children: messages.map((msg, i) => {
        return (0, _jsxRuntime.jsx)("span", {
          css: styles?.message,
          children: (0, _jsxRuntime.jsx)(_v.FormFieldMessage, {
            variant: msg.type,
            children: msg.text
          })
        }, `error${i}`);
      })
    }) : null;
  }
}) || _class);
var _default = exports.default = FormFieldMessages;
