// @instructure/ui-drawer-layout/es/DrawerLayout/v2/DrawerContent/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { debounce } from '@instructure/debounce';
import { getBoundingClientRect } from '@instructure/ui-dom-utils';
import { omitProps } from '@instructure/ui-react-utils';
import { withStyleNew } from '@instructure/emotion';
import generateStyle from "./styles.js";
import { allowedProps } from "./props.js";
import { jsx as _jsx } from "@emotion/react/jsx-runtime";
/**
---
parent: DrawerLayout
id: DrawerLayout.Content
---
**/
let DrawerContent = (_dec = withStyleNew(generateStyle), _dec(_class = class DrawerContent extends Component {
  static displayName = "DrawerContent";
  static componentId = 'DrawerLayout.Content';
  static locatorAttribute = 'data-drawer-content';
  static allowedProps = allowedProps;
  static defaultProps = {
    role: 'region'
  };
  ref = null;
  get _content() {
    return this.ref;
  }
  _resizeListener;
  _debounced;
  componentDidMount() {
    const rect = getBoundingClientRect(this.ref);
    const origSize = {
      width: rect.width
    };
    // set initial size
    if (typeof this.props.onSizeChange === 'function') {
      this.props.onSizeChange({
        width: rect.width,
        height: rect.height
      });

      // listen for changes to size
      this._debounced = debounce(this.props.onSizeChange, 100, {
        leading: false,
        trailing: true
      });
    }
    this._resizeListener = new ResizeObserver(entries => {
      for (const entry of entries) {
        const size = {
          width: entry.contentRect.width
        };
        if (size.width !== origSize.width) {
          this._debounced?.(size);
        }
      }
    });
    this._resizeListener.observe(this.ref);
    this.props.makeStyles?.(this.makeStyleProps(false));
  }
  componentDidUpdate() {
    this.props.makeStyles?.(this.makeStyleProps(true));
  }
  componentWillUnmount() {
    this._resizeListener?.disconnect();
    this._debounced?.cancel();
  }
  makeStyleProps = shouldTransition => {
    return {
      shouldTransition
    };
  };
  handleContentRef = node => {
    this.ref = node;
    if (typeof this.props.contentRef === 'function') {
      this.props.contentRef(node);
    }
  };
  render() {
    const {
      style,
      label,
      role
    } = this.props;
    return _jsx("div", {
      ...omitProps(this.props, DrawerContent.allowedProps),
      role: role,
      style: style,
      ref: this.handleContentRef,
      "aria-label": label,
      css: this.props.styles?.drawerContent,
      "data-cid": "DrawerContent",
      children: this.props.children
    });
  }
}) || _class);
export default DrawerContent;
export { DrawerContent };
