// @instructure/ui-color-picker/es/ColorMixer/v1/RGBAInput/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { withStyle } from '@instructure/emotion';
import shallowCompare from "../utils/shallowCompare.js";
import { ScreenReaderContent } from '@instructure/ui-a11y-content';
import { TextInput } from '@instructure/ui-text-input/v11_6';
import { allowedProps } from "./props.js";
import generateStyle from "./styles.js";
import generateComponentTheme from "./theme.js";
/**
---
private: true
---
**/
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
let RGBAInput = (_dec = withStyle(generateStyle, generateComponentTheme), _dec(_class = class RGBAInput extends Component {
  static displayName = "RGBAInput";
  static allowedProps = allowedProps;
  static componentId = 'ColorMixer.RGBAInput';
  static defaultProps = {
    withAlpha: false
  };
  constructor(props) {
    super(props);
    this.state = {
      value: props.value
    };
  }
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate(prevProps) {
    this.props.makeStyles?.();
    if (shallowCompare(prevProps.value, this.props.value) || shallowCompare(this.state.value, this.props.value)) {
      this.setState({
        value: this.props.value
      });
    }
  }
  handleChange(type, e) {
    const upperLimit = type === 'a' ? 100 : 255;
    const newValue = type === 'a' ? Number(e.target.value) / 100 : Number(e.target.value);
    const newColor = {
      ...this.props.value,
      [type]: newValue
    };
    if (!isNaN(Number(newValue)) && newValue >= 0 && newValue <= upperLimit) {
      this.setState({
        value: newColor
      });
      this.props.onChange(newColor);
      return;
    }
  }
  render() {
    const {
      styles,
      disabled,
      label,
      withAlpha
    } = this.props;
    return _jsxs("div", {
      ref: this.handleRef,
      css: styles?.RGBAInput,
      children: [label && _jsx("div", {
        css: styles?.label,
        children: label
      }), _jsxs("div", {
        css: styles?.inputContainer,
        children: [_jsx("span", {
          css: styles?.rgbInput,
          children: _jsx(TextInput, {
            disabled: disabled,
            value: `${this.state.value.r}`,
            onChange: e => this.handleChange('r', e),
            renderLabel: _jsx(ScreenReaderContent, {
              children: this.props.rgbRedInputScreenReaderLabel
            })
          })
        }), _jsx("span", {
          css: styles?.rgbInput,
          children: _jsx(TextInput, {
            disabled: disabled,
            value: `${this.state.value.g}`,
            onChange: e => this.handleChange('g', e),
            renderLabel: _jsx(ScreenReaderContent, {
              children: this.props.rgbGreenInputScreenReaderLabel
            })
          })
        }), _jsx("span", {
          css: styles?.rgbInput,
          children: _jsx(TextInput, {
            disabled: disabled,
            value: `${this.state.value.b}`,
            onChange: e => this.handleChange('b', e),
            renderLabel: _jsx(ScreenReaderContent, {
              children: this.props.rgbBlueInputScreenReaderLabel
            })
          })
        }), withAlpha && _jsx("span", {
          css: styles?.aInput,
          children: _jsx(TextInput, {
            "aria-label": "%",
            disabled: disabled,
            value: `${Math.round(this.state.value.a * 100)}`,
            onChange: e => this.handleChange('a', e),
            renderAfterInput: _jsx("span", {
              "aria-hidden": "true",
              role: "presentation",
              children: "%"
            }),
            renderLabel: _jsx(ScreenReaderContent, {
              children: this.props.rgbAlphaInputScreenReaderLabel
            })
          })
        })]
      })]
    });
  }
}) || _class);
export default RGBAInput;
