// @instructure/ui-themes/lib/legacySharedThemeTokens/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.canvas = void 0;
var _borders = require("./borders");
var _transitions = require("./transitions");
var _typography = require("./typography");
var _spacing = require("./spacing");
var _forms = require("./forms");
var _media = require("./media");
var _breakpoints = require("./breakpoints");
var _shadows = require("./shadows");
var _stacking = require("./stacking");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const canvas = exports.canvas = {
  borders: _borders.borders,
  transitions: _transitions.transitions,
  typography: _typography.typography,
  spacing: _spacing.spacing,
  forms: _forms.forms,
  media: _media.media,
  breakpoints: _breakpoints.breakpoints,
  shadows: _shadows.shadows,
  stacking: _stacking.stacking
};
var _default = exports.default = canvas;
