// @instructure/ui-table/es/Table/v2/Body/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component, Children, isValidElement } from 'react';
import { safeCloneElement, omitProps } from '@instructure/ui-react-utils';
import { View } from '@instructure/ui-view/latest';
import { withStyleNew } from '@instructure/emotion';
import generateStyle from "./styles.js";
import { allowedProps } from "./props.js";
import TableContext from "../TableContext.js";
/**
---
parent: Table
id: Table.Body
---
**/
import { jsx as _jsx } from "@emotion/react/jsx-runtime";
let Body = (_dec = withStyleNew(generateStyle), _dec(_class = class Body extends Component {
  static displayName = "Body";
  static componentId = 'Table.Body';
  static contextType = TableContext;
  static allowedProps = allowedProps;
  static defaultProps = {
    children: null
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  render() {
    const {
      children,
      styles
    } = this.props;
    const {
      isStacked
    } = this.context;
    return _jsx(View, {
      ...View.omitViewProps(omitProps(this.props, Body.allowedProps), Body),
      as: isStacked ? 'div' : 'tbody',
      css: styles?.body,
      role: isStacked ? 'rowgroup' : undefined,
      children: Children.map(children, child => {
        if (/*#__PURE__*/isValidElement(child)) {
          return safeCloneElement(child, {
            key: child.props.name
          });
        }
        return child;
      })
    });
  }
}) || _class);
export default Body;
export { Body };
