// @instructure/ui-metric/lib/MetricGroup/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.MetricGroup = void 0;
var _react = require("react");
var _emotion = require("@instructure/emotion");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let MetricGroup = exports.MetricGroup = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class MetricGroup extends _react.Component {
  static displayName = "MetricGroup";
  static componentId = 'MetricGroup';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    children: null
  };
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  renderChildren() {
    return _react.Children.map(this.props.children, child => {
      return (0, _safeCloneElement.safeCloneElement)(child, {
        isGroupChild: true
      });
    });
  }
  render() {
    return (0, _jsxRuntime.jsx)("div", {
      ...(0, _passthroughProps.passthroughProps)(this.props),
      css: this.props.styles?.metricGroup,
      role: "grid",
      "aria-readonly": "true",
      ref: this.handleRef,
      "data-cid": "MetricGroup",
      children: this.renderChildren()
    });
  }
}) || _class);
var _default = exports.default = MetricGroup;
