// @instructure/ui-navigation/lib/AppNav/v2/Item/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Item = void 0;
var _react = require("react");
var _console = require("@instructure/console");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _getElementType = require("@instructure/ui-react-utils/lib/getElementType.js");
var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _latest = require("@instructure/ui-view/latest");
var _ScreenReaderContent = require("@instructure/ui-a11y-content/lib/ScreenReaderContent");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: AppNav
id: AppNav.Item
---
@module Item
**/
let Item = exports.Item = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class Item extends _react.Component {
  static displayName = "Item";
  static componentId = 'AppNav.Item';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    children: null,
    isSelected: false,
    cursor: 'pointer',
    isDisabled: false
  };
  ref = null;
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  handleClick = e => {
    const {
      isDisabled,
      onClick
    } = this.props;
    if (isDisabled) {
      e.preventDefault();
      e.stopPropagation();
    } else if (typeof onClick === 'function') {
      onClick(e);
    }
  };
  render() {
    const ElementType = (0, _getElementType.getElementType)(Item, this.props);
    const {
      renderIcon,
      renderLabel,
      href,
      renderAfter,
      cursor,
      isDisabled
    } = this.props;
    const icon = (0, _callRenderProp.callRenderProp)(renderIcon);
    const label = (0, _callRenderProp.callRenderProp)(renderLabel);
    const labelIsForScreenReaders = (0, _matchComponentTypes.matchComponentTypes)(label, [_ScreenReaderContent.ScreenReaderContent]);
    if (icon) {
      (0, _console.logError)(labelIsForScreenReaders, '[AppNav] If an icon is used, the label text should be wrapped in <ScreenReaderContent />.');
    }
    return (0, _jsxRuntime.jsxs)(_latest.View, {
      ...(0, _passthroughProps.passthroughProps)(this.props),
      as: ElementType,
      href: href,
      onClick: this.handleClick,
      disabled: isDisabled,
      elementRef: this.handleRef,
      display: "flex",
      position: "relative",
      borderRadius: "medium",
      cursor: isDisabled ? 'not-allowed' : cursor,
      css: this.props.styles?.item,
      "data-cid": "AppNav.Item",
      children: [icon, labelIsForScreenReaders ? label : (0, _jsxRuntime.jsx)("span", {
        css: this.props.styles?.label,
        children: label
      }), renderAfter && (0, _callRenderProp.callRenderProp)(renderAfter)]
    });
  }
}) || _class);
var _default = exports.default = Item;
