// @instructure/ui-billboard/lib/Billboard/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Billboard = void 0;
var _react = require("react");
var _v11_ = require("@instructure/ui-heading/v11_6");
var _v11_2 = require("@instructure/ui-view/v11_6");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _getElementType = require("@instructure/ui-react-utils/lib/getElementType.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let Billboard = exports.Billboard = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class Billboard extends _react.Component {
  static displayName = "Billboard";
  static componentId = 'Billboard';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    disabled: false,
    readOnly: false,
    size: 'medium',
    headingAs: 'span',
    headingLevel: 'h1',
    as: 'span',
    elementRef: () => {}
  };
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  renderHeading() {
    const {
      headingLevel,
      headingAs,
      heading,
      styles
    } = this.props;
    return (0, _jsxRuntime.jsx)("span", {
      css: styles?.heading,
      children: (0, _jsxRuntime.jsx)(_v11_.Heading, {
        level: headingLevel,
        as: headingAs,
        color: "primary",
        children: heading
      })
    });
  }
  get SVGIconSize() {
    const size = this.props.size;

    // serve up appropriate SVGIcon size for each Billboard size
    if (size === 'small') {
      return 'medium';
    } else if (size === 'large') {
      return 'x-large';
    } else {
      return 'large';
    }
  }
  renderHero() {
    if (typeof this.props.hero === 'function') {
      return this.props.hero(this.SVGIconSize);
    } else {
      return this.props.hero;
    }
  }
  renderContent() {
    const {
      heading,
      message,
      hero,
      styles
    } = this.props;
    return (0, _jsxRuntime.jsxs)("span", {
      css: styles?.content,
      children: [hero && (0, _jsxRuntime.jsx)("span", {
        css: styles?.hero,
        children: this.renderHero()
      }), heading && this.renderHeading(), message && (0, _jsxRuntime.jsx)("span", {
        css: styles?.message,
        children: (0, _callRenderProp.callRenderProp)(message)
      })]
    });
  }
  handleClick = e => {
    const {
      readOnly,
      onClick
    } = this.props;
    if (readOnly) {
      e.preventDefault();
      e.stopPropagation();
    } else if (typeof onClick === 'function') {
      onClick(e);
    }
  };
  render() {
    const {
      href,
      disabled,
      readOnly,
      margin,
      styles
    } = this.props;
    const Element = (0, _getElementType.getElementType)(Billboard, this.props);
    return (0, _jsxRuntime.jsx)(_v11_2.View, {
      as: "div",
      margin: margin,
      children: (0, _jsxRuntime.jsx)(_v11_2.View, {
        ...(0, _omitProps.omitProps)(this.props, [...Billboard.allowedProps, ..._v11_2.View.allowedProps]),
        type: Element === 'button' ? 'button' : undefined,
        as: Element,
        elementRef: this.handleRef,
        css: styles?.billboard,
        href: href,
        onClick: this.handleClick,
        disabled: disabled,
        "aria-disabled": disabled || readOnly ? 'true' : undefined,
        children: this.renderContent()
      })
    });
  }
}) || _class);
var _default = exports.default = Billboard;
