// @instructure/ui-options/es/Options/v1/index.js
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component, Children, createElement } from 'react';
import { omitProps, matchComponentTypes, safeCloneElement, withDeterministicId } from '@instructure/ui-react-utils';
import { View } from '@instructure/ui-view/v11_6';
import { withStyle } from '@instructure/emotion';
import generateStyles from "./styles.js";
import generateComponentTheme from "./theme.js";
import { Item } from "./Item/index.js";
import { Separator } from "./Separator/index.js";
import { allowedProps } from "./props.js";
import { isAndroidOrIOS } from '@instructure/ui-utils';
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
/**
---
category: components/utilities
---
**/
let Options = (_dec = withDeterministicId(), _dec2 = withStyle(generateStyles, generateComponentTheme), _dec(_class = _dec2(_class = class Options extends Component {
  static displayName = "Options";
  static componentId = 'Options';
  static allowedProps = allowedProps;
  static defaultProps = {
    as: 'span',
    role: 'list',
    elementRef: () => {},
    renderLabel: null,
    children: null
  };
  static Item = Item;
  static Separator = Separator;
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  _labelId = this.props.deterministicId('Options-label');
  get childAs() {
    const {
      as
    } = this.props;
    if (as === 'ul' || as === 'ol') {
      return 'li';
    }
    return undefined;
  }
  renderLabel() {
    const {
      renderLabel,
      styles
    } = this.props;
    let labelContent;
    if (typeof renderLabel === 'function') {
      labelContent = /*#__PURE__*/createElement(renderLabel, {});
    } else {
      labelContent = renderLabel;
    }
    return _jsx("span", {
      id: this._labelId,
      role: "presentation"
      // because TalkBack and iOS VoiceOver don't announce sub-menu labels, aria-hidden needs to be false to achive that
      ,
      "aria-hidden": isAndroidOrIOS() ? 'false' : 'true',
      css: styles?.label,
      children: labelContent
    });
  }
  renderSubList(subOptions) {
    const {
      styles
    } = this.props;
    return _jsx(Item, {
      as: this.childAs,
      role: "presentation",
      css: styles?.label,
      children: subOptions
    });
  }
  renderChildren() {
    const {
      children
    } = this.props;
    return Children.map(children, child => {
      if (matchComponentTypes(child, ['Options'])) {
        return this.renderSubList(child);
      }
      if (matchComponentTypes(child, ['Item', 'Separator'])) {
        return safeCloneElement(child, {
          as: this.childAs || child.props.as
        });
      }
      return undefined;
    });
  }
  get ariaLabelledby() {
    if (this.props.renderLabel) {
      return this._labelId;
    }
    if (this.props['aria-labelledby']) {
      return this.props['aria-labelledby'];
    }
    return undefined;
  }
  render() {
    const passthroughProps = View.omitViewProps(omitProps(this.props, Options.allowedProps), Options);
    const {
      as,
      role,
      renderLabel,
      elementRef,
      styles
    } = this.props;
    return _jsxs("div", {
      css: styles?.options,
      role: "presentation",
      ref: this.handleRef,
      "data-cid": "Options",
      children: [renderLabel && this.renderLabel(), _jsx(View, {
        ...passthroughProps,
        elementRef: elementRef,
        css: styles?.list,
        as: as,
        role: role,
        display: "block",
        margin: "none",
        padding: "none",
        background: "primary",
        "aria-labelledby": this.ariaLabelledby,
        children: this.renderChildren()
      })]
    });
  }
}) || _class) || _class);
export default Options;
export { Options };
