// @instructure/ui-top-nav-bar/es/TopNavBar/v1/utils/exampleHelpers.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Drilldown } from '@instructure/ui-drilldown/v11_6';
import { IconQuestionLine, IconAlertsLine, IconConfigureLine, IconSearchLine, IconDiscussionLine, IconDashboardLine } from '@instructure/ui-icons';
import { TopNavBarContext } from "../TopNavBarContext.js";
import { TopNavBar } from "../index.js";
import { elevateIcon } from "./exampleSvgFiles.js";
import { Breadcrumb } from '@instructure/ui-breadcrumb/v11_6';
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
const avatarExample = {
  avatarName: 'User Name',
  avatarSrc: 'https://images.pexels.com/photos/220453/pexels-photo-220453.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=1'
};
const itemSubmenuExample = _jsx(Drilldown, {
  shouldSetAriaExpanded: false,
  rootPageId: "root",
  children: _jsxs(Drilldown.Page, {
    id: "root",
    children: [_jsx(Drilldown.Option, {
      id: "linkExampleOption1",
      href: "/#One",
      children: "Link One"
    }), _jsx(Drilldown.Option, {
      id: "linkExampleOption2",
      href: "/#Two",
      children: "Link Two"
    }), _jsx(Drilldown.Option, {
      id: "linkExampleOption3",
      href: "/#Three",
      children: "Link Three"
    })]
  })
});
const getCustomPopoverConfig = (inverseColor, extraProps = {}) => ({
  children: getCustomPopoverContent(!inverseColor),
  color: inverseColor ? 'primary-inverse' : 'primary',
  isShowingContent: true,
  placement: 'bottom center',
  ...extraProps
});
const getCustomPopoverContent = inverseColor => _jsx("div", {
  style: {
    color: inverseColor ? 'black' : 'white',
    padding: '0 0.5rem'
  },
  children: "dialog content"
});
const getInPlaceDialogConfig = (inverseColor, extraProps = {}) => ({
  open: true,
  onClose: () => {},
  closeButtonLabel: 'Close Search and return to Navigation',
  returnFocusElement: () => document.getElementById('Search'),
  shouldCloseOnEscape: true,
  shouldCloseOnDocumentClick: false,
  shouldContainFocus: false,
  content: () => getCustomPopoverContent(inverseColor),
  ...extraProps
});
const SmallViewportModeWrapper = props => {
  return _jsx(TopNavBarContext.Provider, {
    value: {
      layout: props.layout || 'smallViewport',
      inverseColor: typeof props.inverseColor === 'undefined' ? false : props.inverseColor
    },
    children: props.children
  });
};
SmallViewportModeWrapper.defaultProps = {
  layout: 'smallViewport',
  inverseColor: false
};
const getBrand = (config = {}) => {
  return _jsx(TopNavBar.Brand, {
    screenReaderLabel: "Brand name",
    renderIcon: elevateIcon,
    iconBackground: "#0097D3",
    href: "/#TopNavBar",
    ...config.brandProps
  });
};
const getMenuItems = (config = {}) => {
  const items = config.menuItemsCustomIdList || ['Overview', 'Admin', 'Settings', 'Maps', 'Assessments', 'Community'];
  const count = typeof config.menuItemsCount !== 'undefined' ? config.menuItemsCount : 6;
  const itemsSubset = items.slice(0, count);
  return _jsx(TopNavBar.MenuItems, {
    listLabel: "Page navigation",
    currentPageId: config.currentPageId,
    renderHiddenItemsMenuTriggerLabel: hiddenChildrenCount => `${hiddenChildrenCount} More`,
    ...config.menuItemsProps,
    children: itemsSubset.map((item, idx) => {
      const itemProps = typeof config.menuItemsItemProps === 'function' ? config.menuItemsItemProps(item, idx) : config.menuItemsItemProps;
      return _jsx(TopNavBar.Item, {
        id: item,
        href: config.menuItemsWithSubmenu ? undefined : '/#TopNavBar',
        renderSubmenu: config.menuItemsWithSubmenu ? itemSubmenuExample : undefined,
        ...itemProps,
        children: itemProps?.children || item
      }, item);
    })
  });
};
const getActionItems = (config = {}) => {
  const items = [{
    label: 'Search',
    icon: IconSearchLine
  }, {
    label: 'Info',
    icon: IconQuestionLine
  }, {
    label: 'Alerts',
    icon: IconAlertsLine
  }, {
    label: 'Discussions',
    icon: IconDiscussionLine
  }, {
    label: 'Dashboard',
    icon: IconDashboardLine
  }, {
    label: 'Settings',
    icon: IconConfigureLine
  }];
  const count = typeof config.actionItemsCount !== 'undefined' ? config.actionItemsCount : 3;
  const itemsSubset = items.slice(0, count);
  return _jsx(TopNavBar.ActionItems, {
    ...config.actionItemsProps,
    listLabel: config.actionItemsProps?.listLabel || 'Actions',
    renderHiddenItemsMenuTriggerLabel: config.actionItemsProps?.renderHiddenItemsMenuTriggerLabel || (hiddenChildrenCount => `${hiddenChildrenCount} more actions`),
    children: itemsSubset.map((item, idx) => {
      const itemProps = typeof config.actionItemsItemProps === 'function' ? config.actionItemsItemProps(item.label, idx) : config.actionItemsItemProps;
      return _jsx(TopNavBar.Item, {
        id: item.label,
        renderIcon: item.icon,
        variant: 'icon',
        href: "/#TopNavBar",
        customPopoverConfig: config.hasRenderInPlaceDialogConfig && item.label === 'Search' ? getCustomPopoverConfig(config.inverseColor) : undefined,
        ...itemProps,
        children: itemProps?.children || item.label
      }, item.label);
    })
  });
};
const getUser = (config = {}) => {
  return _jsx(TopNavBar.User, {
    ...config.userProps,
    children: config.userProps && 'children' in config.userProps ? config.userProps?.children : _jsx(TopNavBar.Item, {
      id: config.userId || 'User',
      href: config.userWithSubmenu ? undefined : '/#TopNavBar',
      variant: config.userVariant || 'default',
      renderAvatar: config.userWithAvatar ? avatarExample : undefined,
      renderSubmenu: config.userWithSubmenu ? itemSubmenuExample : undefined,
      ...config.userItemProps,
      children: config.userItemProps?.children || (config.userVariant === 'button' ? 'Log In/Register' : 'User Name')
    })
  });
};
const getBreadcrumb = (_config = {}) => {
  return _jsx(TopNavBar.Breadcrumb, {
    children: _jsxs(Breadcrumb, {
      label: "You are here",
      children: [_jsx(Breadcrumb.Link, {
        children: "Course page 1"
      }), _jsx(Breadcrumb.Link, {
        children: "Course page 2"
      }), _jsx(Breadcrumb.Link, {
        children: "Course page 3"
      })]
    })
  });
};
const getLayoutProps = (props, config = {}) => {
  return {
    smallViewportConfig: {
      dropdownMenuToggleButtonLabel: 'Toggle Menu',
      dropdownMenuLabel: 'Main Menu',
      alternativeTitle: config.hasAlternativeTitle ? 'Page title' : undefined,
      renderInPlaceDialogConfig: config.hasRenderInPlaceDialogConfig ? getInPlaceDialogConfig(props.inverseColor) : undefined
    },
    renderBrand: config.hasRenderBreadcrumb ? undefined : getBrand({
      ...props,
      ...config
    }),
    renderMenuItems: config.hasRenderBreadcrumb ? undefined : getMenuItems({
      ...props,
      ...config
    }),
    renderActionItems: getActionItems({
      ...props,
      ...config
    }),
    renderUser: getUser({
      ...props,
      ...config
    }),
    renderBreadcrumb: config.hasRenderBreadcrumb ? getBreadcrumb({
      ...props,
      ...config
    }) : undefined
  };
};
export { avatarExample, itemSubmenuExample, SmallViewportModeWrapper, getCustomPopoverConfig, getCustomPopoverContent, getInPlaceDialogConfig, getBrand, getMenuItems, getActionItems, getUser, getLayoutProps, getBreadcrumb };
