// @instructure/ui-themes/lib/themes/newThemeTokens/dark/components/menuItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const menuItem = semantics => ({
  activeBackground: semantics.color.background.interactive.action.primary.base,
  activeLabelColor: semantics.color.text.interactive.action.primary.base,
  background: semantics.color.background.elevatedSurface.base,
  fontFamily: semantics.fontFamily.base,
  fontSize: semantics.fontSize.textBase,
  fontWeight: semantics.fontWeight.body.base,
  iconPadding: semantics.spacing.spaceMd,
  labelColor: semantics.color.text.base,
  labelInfoColor: semantics.color.text.muted,
  lineHeight: semantics.lineHeight.paragraph.textSm,
  labelPadding: "0rem",
  paddingVertical: semantics.spacing.spaceSm,
  paddingHorizontal: semantics.spacing.spaceMd,
  highlightedBackground: semantics.color.background.interactive.action.tertiary.hover,
  selectedHighlightedBackground: semantics.color.background.interactive.action.primary.hover,
  highlightedLabelColor: semantics.color.text.base,
  highlightedLabelInfoColor: semantics.color.text.base
});
var _default = exports.default = menuItem;
