// @instructure/ui-date-input/lib/DateInput/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.DateInput = void 0;
var _react = require("react");
var _v11_ = require("@instructure/ui-calendar/v11_6");
var _IconCalendarMonthLine = require("@instructure/ui-icons/lib/generated/IconCalendarMonthLine.js");
var _v11_2 = require("@instructure/ui-popover/v11_6");
var _Selectable = require("@instructure/ui-selectable/lib/Selectable");
var _v11_3 = require("@instructure/ui-text-input/v11_6");
var _createChainedFunction = require("@instructure/ui-utils/lib/createChainedFunction.js");
var _getInteraction = require("@instructure/ui-react-utils/lib/getInteraction.js");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _DateTime = require("@instructure/ui-i18n/lib/DateTime.js");
var _Locale = require("@instructure/ui-i18n/lib/Locale.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let DateInput = exports.DateInput = (_dec = (0, _emotion.withStyle)(_styles.default, null), _dec(_class = class DateInput extends _react.Component {
  static displayName = "DateInput";
  static componentId = 'DateInput';
  static Day = _v11_.Calendar.Day;
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    value: '',
    size: 'medium',
    onBlur: () => {},
    // must have a default so createChainedFunction works
    isRequired: false,
    isInline: false,
    layout: 'stacked',
    display: 'inline-block',
    placement: 'bottom center',
    isShowingCalendar: false
  };
  state = {
    hasInputRef: false,
    isShowingCalendar: false,
    validatedDate: undefined,
    messages: []
  };
  _input = undefined;
  ref = null;
  locale() {
    if (this.props.locale) {
      return this.props.locale;
    } else if (this.context && this.context.locale) {
      return this.context.locale;
    }
    return _Locale.Locale.browserLocale();
  }
  timezone() {
    if (this.props.timezone) {
      return this.props.timezone;
    } else if (this.context && this.context.timezone) {
      return this.context.timezone;
    }
    return _DateTime.DateTime.browserTimeZone();
  }
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  get selectedDateId() {
    let selectedDateId;
    _react.Children.toArray(this.props.children).forEach(day => {
      const {
        date,
        isSelected
      } = day.props;
      if (isSelected) {
        selectedDateId = this.formatDateId(date);
      }
    });
    return selectedDateId;
  }
  get interaction() {
    return (0, _getInteraction.getInteraction)({
      props: this.props
    });
  }
  formatDateId = date => {
    // ISO8601 strings may contain a space. Remove any spaces before using the
    // date as the id.
    return date.replace(/\s/g, '');
  };
  handleInputRef = el => {
    // Ensures that we position the Calendar with respect to the input correctly
    // if the Calendar is open on mount
    if (!this.state.hasInputRef) {
      this.setState({
        hasInputRef: true
      });
    }
    this._input = el;
    this.props.inputRef?.(el);
  };
  handleInputChange = (event, value) => {
    this.props.onChange?.(event, {
      value
    });
    this.handleShowCalendar(event);
  };
  handleShowCalendar = event => {
    if (!this.props.children) {
      this.setState({
        isShowingCalendar: true
      });
    } else if (this.interaction === 'enabled' && this.props.children) {
      this.props.onRequestShowCalendar?.(event);
    }
  };
  validateDate = date => {
    const {
      invalidDateErrorMessage
    } = this.props;
    const disabledDateErrorMessage = this.props.disabledDateErrorMessage || invalidDateErrorMessage;
    const messages = [];
    // check if date is enabled
    const {
      disabledDates
    } = this.props;
    if (typeof disabledDates === 'function' && disabledDates(date) || Array.isArray(disabledDates) && disabledDates.find(dateString => _DateTime.DateTime.parse(dateString, this.locale(), this.timezone()).isSame(_DateTime.DateTime.parse(date, this.locale(), this.timezone()), 'day'))) {
      messages.push(typeof disabledDateErrorMessage === 'function' ? disabledDateErrorMessage(date) : {
        type: 'error',
        text: disabledDateErrorMessage
      });
    }

    // check if date is valid
    if (!_DateTime.DateTime.parse(date, this.locale(), this.timezone(), [_DateTime.DateTime.momentISOFormat, 'llll', 'LLLL', 'lll', 'LLL', 'll', 'LL', 'l', 'L'], true).isValid()) {
      messages.push(typeof invalidDateErrorMessage === 'function' ? invalidDateErrorMessage(date) : {
        type: 'error',
        text: invalidDateErrorMessage
      });
    }
    return messages;
  };
  handleHideCalendar = (event, setectedDate) => {
    if (!this.props.children) {
      const dateString = setectedDate || this.props.value;
      const messages = [];
      if (this.props.onRequestValidateDate) {
        const userValidatedDate = this.props.onRequestValidateDate?.(event, dateString || '', this.validateDate(dateString || ''));
        messages.push(...(userValidatedDate || []));
      } else {
        if (dateString) {
          messages.push(...this.validateDate(dateString));
        }
      }
      this.setState({
        messages,
        isShowingCalendar: false
      });
    } else {
      this.props.onRequestValidateDate?.(event);
      this.props.onRequestHideCalendar?.(event);
    }
  };
  handleHighlightOption = (event, {
    direction
  }) => {
    const {
      onRequestSelectNextDay,
      onRequestSelectPrevDay,
      onChange,
      value,
      currentDate
    } = this.props;
    const isValueValid = value && _DateTime.DateTime.parse(value, this.locale(), this.timezone()).isValid();
    if (direction === -1) {
      if (onRequestSelectPrevDay) {
        onRequestSelectPrevDay?.(event);
      } else {
        // @ts-expect-error TODO
        onChange(event, {
          value: _DateTime.DateTime.parse(isValueValid ? value : currentDate, this.locale(), this.timezone()).subtract(1, 'day').format('MMMM D, YYYY')
        });
        this.setState({
          messages: []
        });
      }
    }
    if (direction === 1) {
      if (onRequestSelectNextDay) {
        onRequestSelectNextDay?.(event);
      } else {
        // @ts-expect-error TODO
        onChange(event, {
          value: _DateTime.DateTime.parse(isValueValid ? value : currentDate, this.locale(), this.timezone()).add(1, 'day').format('MMMM D, YYYY')
        });
        this.setState({
          messages: []
        });
      }
    }
  };
  renderMonthNavigationButton(type = 'prev') {
    const {
      renderPrevMonthButton,
      renderNextMonthButton
    } = this.props;
    const button = type === 'prev' ? renderPrevMonthButton : renderNextMonthButton;
    return button && (0, _safeCloneElement.safeCloneElement)((0, _callRenderProp.callRenderProp)(button), {
      tabIndex: -1
    });
  }
  renderDays(getOptionProps) {
    const children = this.props.children;
    if (!children) return;
    return _react.Children.map(children, day => {
      const {
        date,
        isOutsideMonth
      } = day.props;
      const props = {
        tabIndex: -1,
        id: this.formatDateId(date)
      };
      const optionProps = getOptionProps(props);
      const propsAdded = isOutsideMonth ? {
        ...props,
        onClick: optionProps.onClick,
        role: 'presentation'
      } : optionProps;
      return (0, _safeCloneElement.safeCloneElement)(day, propsAdded);
    });
  }
  renderCalendar({
    getListProps,
    getOptionProps
  }) {
    const {
      onRequestRenderNextMonth,
      onRequestRenderPrevMonth,
      renderNavigationLabel,
      renderWeekdayLabels,
      value,
      onChange,
      disabledDates,
      currentDate,
      selectedLabel
    } = this.props;
    const isValidDate = value ? _DateTime.DateTime.parse(value, this.locale(), this.timezone()).isValid() : false;
    const noChildrenProps = this.props.children ? {} : {
      disabledDates,
      currentDate,
      selectedDate: isValidDate ? value : undefined,
      visibleMonth: isValidDate ? value : undefined,
      onDateSelected: (dateString, momentDate, e) => {
        // @ts-expect-error TODO
        onChange?.(e, {
          value: `${momentDate.format('MMMM')} ${momentDate.format('D')}, ${momentDate.format('YYYY')}`
        });
        this.handleHideCalendar(e, dateString);
      }
    };
    return (0, _jsxRuntime.jsx)(_v11_.Calendar, {
      ...getListProps({
        onRequestRenderNextMonth,
        onRequestRenderPrevMonth,
        renderNavigationLabel,
        renderWeekdayLabels,
        renderNextMonthButton: this.renderMonthNavigationButton('next'),
        renderPrevMonthButton: this.renderMonthNavigationButton('prev'),
        selectedLabel
      }),
      ...noChildrenProps,
      children: this.renderDays(getOptionProps)
    });
  }
  renderInput({
    getInputProps,
    getTriggerProps
  }) {
    const {
      renderLabel,
      value,
      placeholder,
      onBlur,
      isRequired,
      size,
      isInline,
      layout,
      width,
      onRequestValidateDate,
      onRequestShowCalendar,
      onRequestHideCalendar,
      onRequestSelectNextDay,
      onRequestSelectPrevDay,
      onRequestRenderNextMonth,
      onRequestRenderPrevMonth,
      ...rest
    } = this.props;
    const {
      interaction
    } = this;
    const {
      ref,
      // Apply this to the actual inputRef
      ...triggerProps
    } = getTriggerProps();
    const messages = this.props.messages || this.state.messages;
    return (0, _jsxRuntime.jsx)(_v11_3.TextInput, {
      ...triggerProps,
      ...(0, _passthroughProps.passthroughProps)(rest),
      ...getInputProps({
        renderLabel: (0, _callRenderProp.callRenderProp)(renderLabel),
        value,
        placeholder,
        size,
        layout,
        width,
        messages,
        onChange: this.handleInputChange,
        onBlur: (0, _createChainedFunction.createChainedFunction)(onBlur, this.handleHideCalendar),
        inputRef: (0, _createChainedFunction.createChainedFunction)(ref, this.handleInputRef),
        interaction,
        isRequired,
        display: isInline ? 'inline-block' : 'block',
        renderAfterInput: (0, _jsxRuntime.jsx)(_IconCalendarMonthLine.IconCalendarMonthLine, {
          inline: false
        })
      }),
      onKeyDown: e => {
        if (!this.props.children) {
          if (e.key === 'Enter') {
            // @ts-expect-error TODO
            this.handleHideCalendar(e);
          }
        }
        triggerProps.onKeyDown?.(e);
      }
    });
  }
  shouldShowCalendar = () => this.props.children ? this.props.isShowingCalendar : this.state.isShowingCalendar;
  render() {
    const {
      placement,
      assistiveText,
      styles
    } = this.props;
    const isShowingCalendar = this.shouldShowCalendar();
    return (0, _jsxRuntime.jsx)(_Selectable.Selectable, {
      isShowingOptions: isShowingCalendar,
      onRequestShowOptions: this.handleShowCalendar,
      onRequestHideOptions: this.handleHideCalendar,
      onRequestHighlightOption: this.handleHighlightOption,
      onRequestSelectOption: e => this.handleHideCalendar(e),
      selectedOptionId: this.selectedDateId,
      highlightedOptionId: this.selectedDateId,
      children: ({
        getRootProps,
        getInputProps,
        getTriggerProps,
        getListProps,
        getOptionProps,
        getDescriptionProps
      }) => (0, _jsxRuntime.jsxs)("span", {
        ...getRootProps({
          css: styles?.dateInput
        }),
        ref: el => {
          this.ref = el;
        },
        "data-cid": "DateInput",
        children: [this.renderInput({
          getInputProps,
          getTriggerProps
        }), (0, _jsxRuntime.jsx)("span", {
          ...getDescriptionProps(),
          css: styles?.assistiveText,
          children: assistiveText
        }), (0, _jsxRuntime.jsx)(_v11_2.Popover, {
          placement: placement,
          isShowingContent: isShowingCalendar,
          positionTarget: this._input,
          shouldReturnFocus: false,
          shouldFocusContentOnTriggerBlur: true,
          children: this.renderCalendar({
            getListProps,
            getOptionProps
          })
        })]
      })
    });
  }
}) || _class);
var _default = exports.default = DateInput;
