// @instructure/ui-menu/lib/Menu/v2/MenuItem/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.MenuItem = void 0;
var _react = require("react");
var _keycode = _interopRequireDefault(require("keycode"));
var _uiIcons = require("@instructure/ui-icons");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _getElementType = require("@instructure/ui-react-utils/lib/getElementType.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _createChainedFunction = require("@instructure/ui-utils/lib/createChainedFunction.js");
var _isActiveElement = require("@instructure/ui-dom-utils/lib/isActiveElement.js");
var _findDOMNode = require("@instructure/ui-dom-utils/lib/findDOMNode.js");
var _emotion = require("@instructure/emotion");
var _MenuContext = require("../../../utils/v1/MenuContext");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: Menu
id: Menu.Item
---
**/
let MenuItem = exports.MenuItem = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = _dec2(_class = class MenuItem extends _react.Component {
  static displayName = "MenuItem";
  static componentId = 'Menu.Item';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    type: 'button',
    disabled: false
  };
  static contextType = _MenuContext.MenuContext;
  constructor(props) {
    super(props);
    const state = {
      isHovered: false,
      isFocused: false
    };
    if (typeof props.selected === 'undefined') {
      state.selected = !!props.defaultSelected;
    }
    this.state = state;
    this.labelId = props.deterministicId('MenuItem__label');
  }
  get _node() {
    return this.ref;
  }
  labelId;
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this.props.makeStyles?.();
    const context = this.context;
    if (context && context.registerMenuItem) {
      context.registerMenuItem(this);
    }
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  componentWillUnmount() {
    const context = this.context;
    if (context && context.removeMenuItem) {
      context.removeMenuItem(this);
    }
  }
  handleClick = e => {
    const {
      onSelect,
      onClick,
      disabled,
      value
    } = this.props;
    const selected = !this.selected;
    if (disabled) {
      e.preventDefault();
      return;
    }
    if (typeof this.props.selected === 'undefined') {
      this.setState({
        selected
      });
    }
    if (typeof onSelect === 'function') {
      e.persist();
      onSelect(e, value, selected, this);
    }
    if (typeof onClick === 'function') {
      onClick(e);
    }
  };
  handleKeyDown = e => {
    const spaceKey = e.keyCode === _keycode.default.codes.space;
    const enterKey = e.keyCode === _keycode.default.codes.enter;
    if (spaceKey || enterKey) {
      e.preventDefault();
      e.stopPropagation();
      if (enterKey) {
        // handle space key on keyUp for FF
        const refNode = (0, _findDOMNode.findDOMNode)(this.ref);
        refNode.click();
      }
    }
  };
  handleKeyUp = e => {
    const spaceKey = e.keyCode === _keycode.default.codes.space;
    const enterKey = e.keyCode === _keycode.default.codes.enter;
    if (spaceKey || enterKey) {
      e.preventDefault();
      e.stopPropagation();
      if (spaceKey) {
        const refNode = (0, _findDOMNode.findDOMNode)(this.ref);
        refNode.click();
      }
    }
  };
  handleMouseOver = event => {
    if (!this.focused) {
      this.focus();
    }
    if (typeof this.props.onMouseOver === 'function') {
      this.props.onMouseOver(event, this);
    }
  };
  handleMouseEnter = () => {
    this.setState({
      isHovered: true
    });
  };
  handleMouseLeave = () => {
    this.setState({
      isHovered: false
    });
  };
  handleFocusEvent = () => {
    this.setState({
      isFocused: true
    });
  };
  handleBlurEvent = () => {
    this.setState({
      isFocused: false
    });
  };
  getIconColor = () => {
    const {
      type
    } = this.props;
    if (type === 'flyout') {
      return 'baseColor';
    }
    return this.selected ? 'inverseColor' : 'baseColor';
  };
  get elementType() {
    return (0, _getElementType.getElementType)(MenuItem, this.props);
  }
  get role() {
    switch (this.props.type) {
      case 'checkbox':
        return 'menuitemcheckbox';
      case 'radio':
        return 'menuitemradio';
      default:
        return 'menuitem';
    }
  }
  get selected() {
    return typeof this.props.selected === 'undefined' ? this.state.selected : this.props.selected;
  }
  get focused() {
    return (0, _isActiveElement.isActiveElement)(this.ref);
  }
  focus() {
    const refNode = (0, _findDOMNode.findDOMNode)(this.ref);
    refNode.focus();
  }
  renderContent() {
    const {
      children,
      type,
      renderLabelInfo
    } = this.props;
    return (0, _jsxRuntime.jsxs)("span", {
      children: [(0, _jsxRuntime.jsx)("span", {
        css: this.props.styles?.label,
        id: this.labelId,
        children: children
      }), (type === 'checkbox' || type === 'radio') && (0, _jsxRuntime.jsx)("span", {
        css: this.props.styles?.icon,
        children: this.selected && (0, _jsxRuntime.jsx)(_uiIcons.CheckInstUIIcon, {
          size: "md",
          color: this.getIconColor()
        })
      }), type === 'flyout' && (0, _jsxRuntime.jsx)("span", {
        css: this.props.styles?.icon,
        children: (0, _jsxRuntime.jsx)(_uiIcons.ChevronRightInstUIIcon, {
          size: "md",
          color: this.getIconColor()
        })
      }), renderLabelInfo && (0, _jsxRuntime.jsx)("span", {
        css: this.props.styles?.labelInfo,
        children: (0, _callRenderProp.callRenderProp)(renderLabelInfo)
      })]
    });
  }
  render() {
    const {
      disabled,
      controls,
      onKeyDown,
      onKeyUp,
      onFocus,
      onBlur,
      type,
      href,
      target
    } = this.props;
    const props = (0, _omitProps.omitProps)(this.props, MenuItem.allowedProps);
    const ElementType = this.elementType;
    return (0, _jsxRuntime.jsx)(ElementType, {
      tabIndex: -1 // note: tabIndex can be overridden by Menu or MenuItemGroup components
      ,
      ...props,
      href: href,
      target: target,
      role: this.role,
      "aria-labelledby": this.labelId,
      "aria-disabled": disabled ? 'true' : undefined,
      "aria-controls": controls,
      "aria-checked": type === 'checkbox' || type === 'radio' ? this.selected ? 'true' : 'false' : undefined,
      onClick: this.handleClick,
      onKeyUp: (0, _createChainedFunction.createChainedFunction)(onKeyUp, this.handleKeyUp),
      onKeyDown: (0, _createChainedFunction.createChainedFunction)(onKeyDown, this.handleKeyDown),
      ref: this.handleRef,
      css: this.props.styles?.menuItem,
      onMouseOver: this.handleMouseOver,
      onMouseEnter: this.handleMouseEnter,
      onMouseLeave: this.handleMouseLeave,
      onFocus: (0, _createChainedFunction.createChainedFunction)(onFocus, this.handleFocusEvent),
      onBlur: (0, _createChainedFunction.createChainedFunction)(onBlur, this.handleBlurEvent),
      "data-cid": "MenuItem",
      children: this.renderContent()
    });
  }
}) || _class) || _class);
var _default = exports.default = MenuItem;
