// @instructure/ui-tag/lib/Tag/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Tag = void 0;
var _react = require("react");
var _uiIcons = require("@instructure/ui-icons");
var _latest = require("@instructure/ui-view/latest");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _isActiveElement = require("@instructure/ui-dom-utils/lib/isActiveElement.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let Tag = exports.Tag = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class Tag extends _react.Component {
  static displayName = "Tag";
  static componentId = 'Tag';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    size: 'medium',
    dismissible: false,
    variant: 'default',
    disabled: false,
    readOnly: false
  };
  state = {
    iconHovered: false
  };
  ref = null;
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  get focused() {
    return (0, _isActiveElement.isActiveElement)(this.ref);
  }
  focus = () => {
    this.ref && this.ref.focus();
  };
  handleIconMouseEnter = () => {
    this.setState({
      iconHovered: true
    });
  };
  handleIconMouseLeave = () => {
    this.setState({
      iconHovered: false
    });
  };
  handleClick = e => {
    const {
      disabled,
      readOnly,
      onClick
    } = this.props;
    if (disabled || readOnly) {
      e.preventDefault();
      e.stopPropagation();
    } else if (typeof onClick === 'function') {
      onClick(e);
    }
  };
  handleRef = element => {
    this.ref = element;
    if (typeof this.props.elementRef === 'function') {
      this.props.elementRef(element);
    }
  };
  getIconSize = () => {
    const {
      size,
      variant
    } = this.props;
    if (variant === 'inline') {
      return 'xs';
    }
    const sizeMap = {
      small: 'xs',
      medium: 'sm',
      large: 'md'
    };
    return sizeMap[size];
  };
  render() {
    const {
      className,
      dismissible,
      disabled,
      readOnly,
      text,
      title,
      onClick,
      margin,
      styles,
      variant
    } = this.props;
    const passthroughProps = _latest.View.omitViewProps((0, _omitProps.omitProps)(this.props, Tag.allowedProps), Tag);
    const getIconColor = () => {
      if (disabled) {
        return 'mutedColor';
      }
      if (variant === 'inline') {
        return 'inverseColor';
      }
      return this.state.iconHovered ? 'actionSecondaryHoverColor' : 'baseColor';
    };
    return (0, _jsxRuntime.jsxs)(_latest.View, {
      ...passthroughProps,
      elementRef: this.handleRef,
      css: styles?.tag,
      className: className,
      as: onClick ? 'button' : 'span',
      margin: margin,
      type: onClick ? 'button' : undefined,
      ...(onClick && {
        onClick: this.handleClick
      }),
      disabled: disabled || readOnly,
      display: undefined,
      title: title || (typeof text === 'string' ? text : undefined),
      "data-cid": "Tag",
      onMouseEnter: this.handleIconMouseEnter,
      onMouseLeave: this.handleIconMouseLeave,
      children: [(0, _jsxRuntime.jsx)("span", {
        css: styles?.text,
        children: text
      }), onClick && dismissible ? (0, _jsxRuntime.jsx)("span", {
        css: styles?.icon,
        children: (0, _jsxRuntime.jsx)(_uiIcons.XInstUIIcon, {
          size: this.getIconSize(),
          color: getIconColor()
        })
      }) : null]
    });
  }
}) || _class);
var _default = exports.default = Tag;
