// @instructure/ui-top-nav-bar/lib/TopNavBar/v2/TopNavBarLayout/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TopNavBarLayout = void 0;
var _react = require("react");
var _emotion = require("@instructure/emotion");
var _TopNavBarContext = require("../TopNavBarContext");
var _SmallViewportLayout = require("./SmallViewportLayout");
var _DesktopLayout = require("./DesktopLayout");
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: TopNavBar
id: TopNavBar.Layout
---
@module TopNavBarLayout
**/
let TopNavBarLayout = exports.TopNavBarLayout = (_dec = (0, _emotion.withStyleNew)(null), _dec(_class = class TopNavBarLayout extends _react.Component {
  static displayName = "TopNavBarLayout";
  static componentId = 'TopNavBar.Layout';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    smallViewportConfig: {}
  };
  static contextType = _TopNavBarContext.TopNavBarContext;
  ref = null;
  componentDidMount() {
    // TODO: Remove this in v10!
    if ('desktopConfig' in this.props) {}
  }
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  get sortedThemeOverride() {
    const {
      themeOverride
    } = this.props;

    // TODO: try to type better, the themeOverride types are not prepared to be inherited and sorted

    if (!themeOverride || typeof themeOverride === 'function') {
      return {
        desktopThemeOverride: themeOverride,
        smallViewportThemeOverride: themeOverride
      };
    }
    const desktopThemeOverride = {};
    const smallViewportThemeOverride = {};
    for (const key in themeOverride) {
      if (key.startsWith('smallViewport')) {
        const variable = key;
        smallViewportThemeOverride[variable] = themeOverride[variable];
      } else if (key.startsWith('desktop')) {
        const variable = key;
        desktopThemeOverride[variable] = themeOverride[variable];
      }
    }
    return {
      desktopThemeOverride,
      smallViewportThemeOverride
    };
  }
  render() {
    const {
      smallViewportConfig,
      // @ts-expect-error prevents it to be passed
      styles,
      // @ts-expect-error prevents it to be passed
      makeStyles,
      ...restProps
    } = this.props;
    if (this.context.layout === 'smallViewport') {
      return (0, _jsxRuntime.jsx)(_SmallViewportLayout.TopNavBarSmallViewportLayout, {
        ...smallViewportConfig,
        ...restProps,
        themeOverride: this.sortedThemeOverride.smallViewportThemeOverride,
        elementRef: this.handleRef
      });
    }
    return (0, _jsxRuntime.jsx)(_DesktopLayout.TopNavBarDesktopLayout, {
      ...restProps,
      themeOverride: this.sortedThemeOverride.desktopThemeOverride,
      elementRef: this.handleRef
    });
  }
}) || _class);
var _default = exports.default = TopNavBarLayout;
