// @instructure/ui-breadcrumb/lib/Breadcrumb/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Breadcrumb = void 0;
Object.defineProperty(exports, "BreadcrumbLink", {
  enumerable: true,
  get: function () {
    return _BreadcrumbLink.BreadcrumbLink;
  }
});
exports.default = void 0;
var _react = require("react");
var _latest = require("@instructure/ui-view/latest");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _emotion = require("@instructure/emotion");
var _uiIcons = require("@instructure/ui-icons");
var _BreadcrumbLink = require("./BreadcrumbLink");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let Breadcrumb = exports.Breadcrumb = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class Breadcrumb extends _react.Component {
  static displayName = "Breadcrumb";
  static componentId = 'Breadcrumb';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    size: 'medium',
    children: null
  };
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  addAriaCurrent = child => {
    const updatedChild = (0, _safeCloneElement.safeCloneElement)(child, {
      'aria-current': 'page'
    });
    return updatedChild;
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  static Link = _BreadcrumbLink.BreadcrumbLink;
  renderChildren() {
    const {
      styles,
      children,
      size
    } = this.props;
    const numChildren = _react.Children.count(children);
    const inlineStyle = {
      maxWidth: `${Math.floor(100 / numChildren)}%`
    };
    let isAriaCurrentSet = false;
    return _react.Children.map(children, (child, index) => {
      const isLastElement = index === numChildren - 1;
      if (/*#__PURE__*/(0, _react.isValidElement)(child)) {
        const isCurrentPage = child.props.isCurrentPage || false;
        if (isAriaCurrentSet && isCurrentPage) {}
        if (isCurrentPage) {
          isAriaCurrentSet = true;
        }
      }

      // Clone child (BreadcrumbLink) and pass size prop
      const childWithSize = /*#__PURE__*/(0, _react.isValidElement)(child) ? (0, _safeCloneElement.safeCloneElement)(child, {
        size
      }) : child;

      // Map Breadcrumb sizes to icon size tokens
      const breadcrumbSizeToIconSize = {
        small: 'sm',
        medium: 'md',
        large: 'lg'
      };
      const iconSize = breadcrumbSizeToIconSize[size || 'small'];
      return (0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
        children: [(0, _jsxRuntime.jsx)("li", {
          css: styles?.crumb,
          style: inlineStyle,
          children: !isAriaCurrentSet && isLastElement && child.props.isCurrentPage !== false ? this.addAriaCurrent(childWithSize) : childWithSize
        }), index < numChildren - 1 && (0, _jsxRuntime.jsx)("li", {
          "aria-hidden": "true",
          css: styles?.separator,
          children: (0, _jsxRuntime.jsx)(_uiIcons.ChevronRightInstUIIcon, {
            size: iconSize,
            color: 'mutedColor'
          })
        })]
      });
    });
  }
  render() {
    const {
      styles
    } = this.props;
    return (0, _jsxRuntime.jsx)(_latest.View, {
      role: "navigation",
      as: "div",
      margin: this.props.margin,
      "aria-label": this.props.label,
      elementRef: this.handleRef,
      "data-cid": "Breadcrumb",
      children: (0, _jsxRuntime.jsx)("ol", {
        css: styles?.breadcrumb,
        children: this.renderChildren()
      })
    });
  }
}) || _class);
var _default = exports.default = Breadcrumb;
