// @instructure/ui-popover/lib/Popover/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Popover = void 0;
var _react = require("react");
var _keycode = _interopRequireDefault(require("keycode"));
var _Position = require("@instructure/ui-position/lib/Position");
var _parsePlacement = require("@instructure/ui-position/lib/parsePlacement.js");
var _mirrorHorizontalPlacement = require("@instructure/ui-position/lib/mirrorHorizontalPlacement.js");
var _latest = require("@instructure/ui-view/latest");
var _Dialog = require("@instructure/ui-dialog/lib/Dialog");
var _textDirectionContextConsumer = require("@instructure/ui-i18n/lib/textDirectionContextConsumer.js");
var _findDOMNode = require("@instructure/ui-dom-utils/lib/findDOMNode.js");
var _containsActiveElement = require("@instructure/ui-dom-utils/lib/containsActiveElement.js");
var _requestAnimationFrame = require("@instructure/ui-dom-utils/lib/requestAnimationFrame.js");
var _handleMouseOverOut = require("@instructure/ui-dom-utils/lib/handleMouseOverOut.js");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _createChainedFunction = require("@instructure/ui-utils/lib/createChainedFunction.js");
var _shallowEqual = require("@instructure/ui-utils/lib/shallowEqual.js");
var _px = require("@instructure/ui-utils/lib/px.js");
var _combineDataCid = require("@instructure/ui-utils/lib/combineDataCid.js");
var _console = require("@instructure/console");
var _FocusRegion = require("@instructure/ui-a11y-utils/lib/FocusRegion.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _dec3, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
tags: overlay, portal, dialog
---
**/
let Popover = exports.Popover = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _textDirectionContextConsumer.textDirectionContextConsumer)(), _dec3 = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = _dec2(_class = _dec3(_class = class Popover extends _react.Component {
  static displayName = "Popover";
  static componentId = 'Popover';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    defaultIsShowingContent: false,
    placement: 'bottom center',
    stacking: 'topmost',
    shadow: 'above',
    offsetX: 0,
    offsetY: 0,
    color: 'primary',
    on: ['hover', 'focus'],
    withArrow: true,
    constrain: 'window',
    insertAt: 'bottom',
    shouldAlignArrow: false,
    shouldTrackPosition: true,
    shouldRenderOffscreen: false,
    shouldContainFocus: false,
    shouldReturnFocus: true,
    shouldCloseOnDocumentClick: true,
    shouldFocusContentOnTriggerBlur: false,
    shouldCloseOnEscape: true,
    shouldSetAriaExpanded: true
  };
  constructor(props) {
    super(props);
    this._renderTriggerProp = this.props.renderTrigger;
    this._renderTrigger = (0, _callRenderProp.callRenderProp)(this.props.renderTrigger);
    this.state = {
      placement: props.placement,
      offsetX: props.offsetX,
      offsetY: props.offsetY,
      isShowingContent: typeof props.isShowingContent === 'undefined' ? props.defaultIsShowingContent : undefined
    };
    this._id = this.props.id || props.deterministicId();
    this._raf = [];
    this._handleMouseOver = _handleMouseOverOut.handleMouseOverOut.bind(null, event => {
      this.show(event);
      clearTimeout(this.mouseOutTimeout);
    });
    this._handleMouseOut = _handleMouseOverOut.handleMouseOverOut.bind(null, event => {
      // this is needed bc the trigger mouseOut fires before tooltip mouseOver
      this.mouseOutTimeout = setTimeout(() => {
        this.hide(event);
      }, 1);
    });
  }
  _handleMouseOver;
  _handleMouseOut;
  _id;
  _raf = [];
  _trigger = null;
  _view = null;
  _dialog = null;
  _contentElement = null;
  _focusRegion;
  // renderTrigger needs to be a variable because if it's a function it will
  // recreate the trigger on each render which will trigger MouseOver events
  // that will make the tooltip reappear and the trigger cannot accept
  // onClick events (since the state change caused by MouseDown recreates it)
  _renderTrigger;
  _renderTriggerProp;
  mouseOutTimeout;
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  get isTooltip() {
    return this.props.shouldRenderOffscreen && !this.props.shouldReturnFocus && !this.props.shouldContainFocus && !this.props.shouldFocusContentOnTriggerBlur;
  }
  componentDidMount() {
    this.props.makeStyles?.();
    if (this.isTooltip) {
      // if popover is being used as a tooltip with no focusable content
      // manage its FocusRegion internally rather than registering it with
      // the FocusRegionManager via Dialog
      this._focusRegion = new _FocusRegion.FocusRegion(this._contentElement, {
        shouldCloseOnEscape: this.props.shouldCloseOnEscape,
        shouldCloseOnDocumentClick: false,
        onDismiss: this.hide,
        isTooltip: true
      });
      if (this.shown) {
        this._focusRegion.activate();
      }
    }
  }
  componentWillUnmount() {
    this._raf.forEach(request => request.cancel());
    this._raf = [];
    if (this._focusRegion) {
      this._focusRegion.deactivate();
      this._focusRegion.blur();
    }
  }
  shouldComponentUpdate(nextProps, nextState) {
    return !(0, _shallowEqual.shallowEqual)(this.props, nextProps) || !(0, _shallowEqual.shallowEqual)(this.state, nextState);
  }
  componentDidUpdate(prevProps, prevState) {
    this.props.makeStyles?.();
    if (this._focusRegion && this.isTooltip) {
      // if focus region exists, popover is acting as a tooltip
      // so we manually activate and deactivate the region when showing/hiding
      if (!prevProps.isShowingContent && this.props.isShowingContent || !prevState.isShowingContent && this.state.isShowingContent) {
        // changed from hiding to showing
        this._focusRegion.activate();
        this._focusRegion.focus();
      }
      if (prevProps.isShowingContent && !this.props.isShowingContent || prevState.isShowingContent && !this.state.isShowingContent) {
        // changed from showing to hiding
        this._focusRegion.deactivate();
      }
    }

    // since `offsetX`, `offsetY` and `placement` are saved into the state
    // in the constructor and used from the state later,
    // we need to update the state if these props change
    if (this.props.offsetX !== prevProps.offsetX || this.props.offsetY !== prevProps.offsetY || this.props.placement !== prevProps.placement || this.props.shouldAlignArrow !== prevProps.shouldAlignArrow || this.props.withArrow !== prevProps.withArrow) {
      this.setState({
        ...this.computeOffsets(this.placement)
      });
    }
  }

  /**
   * Offsets the popover by the arrow size
   */
  computeOffsets(placement) {
    let {
      offsetX,
      offsetY
    } = this.props;
    if (this.props.shouldAlignArrow && this._view) {
      const secondaryPlacement = (0, _parsePlacement.parsePlacement)(placement)[1];

      // arrowSize and arrowBorderWidth are component theme variables
      // declared in ContextView's styles.js
      const {
        arrowSize = 0,
        arrowBorderWidth = 0
      } = this._view.props.styles;
      const offsetAmount = ((0, _px.px)(arrowSize) + (0, _px.px)(arrowBorderWidth)) * 2;
      if (secondaryPlacement === 'start') {
        offsetX = offsetAmount;
      } else if (secondaryPlacement === 'end') {
        offsetX = -offsetAmount;
      } else if (secondaryPlacement === 'top') {
        offsetY = offsetAmount;
      } else if (secondaryPlacement === 'bottom') {
        offsetY = -offsetAmount;
      }
    }
    return {
      offsetX,
      offsetY
    };
  }
  get placement() {
    let {
      placement
    } = this.props;
    const {
      dir
    } = this.props;
    const isRtl = dir === _textDirectionContextConsumer.textDirectionContextConsumer.DIRECTION.rtl;
    if (isRtl) {
      placement = (0, _mirrorHorizontalPlacement.mirrorHorizontalPlacement)(placement, ' ');
    }
    return !this.shown && this.props.shouldRenderOffscreen ? 'offscreen' : placement;
  }
  get positionProps() {
    return {
      offsetX: this.state.offsetX,
      offsetY: this.state.offsetY,
      shouldTrackPosition: this.props.shouldTrackPosition && this.shown,
      insertAt: this.props.insertAt,
      placement: this.placement,
      constrain: this.props.constrain,
      onPositioned: this.handlePositioned,
      onPositionChanged: this.handlePositionChanged,
      target: this.props.positionTarget,
      containerDisplay: this.props.positionContainerDisplay,
      mountNode: this.props.mountNode,
      id: this._id
    };
  }
  get shown() {
    return typeof this.props.isShowingContent === 'undefined' ? this.state.isShowingContent : this.props.isShowingContent;
  }
  get defaultFocusElement() {
    return this.props.defaultFocusElement;
  }
  show = event => {
    if (typeof this.props.isShowingContent === 'undefined') {
      this.setState({
        isShowingContent: true
      });
    }
    this.props.onShowContent?.(event);
  };
  hide = (event, documentClick = false) => {
    const {
      onHideContent,
      isShowingContent
    } = this.props;
    if (typeof isShowingContent === 'undefined') {
      // uncontrolled, set state, fire callbacks
      this.setState(({
        isShowingContent
      }) => {
        if (isShowingContent) {
          onHideContent?.(event, {
            documentClick
          });
        }
        return {
          isShowingContent: false
        };
      });
    } else if (isShowingContent) {
      // controlled, fire callback
      onHideContent?.(event, {
        documentClick
      });
    }
  };
  toggle = event => {
    if (this.shown) {
      this.hide(event);
    } else {
      this.show(event);
    }
  };
  handleDialogDismiss = (event, documentClick) => {
    if (!this.props.shouldReturnFocus && this.props.shouldFocusContentOnTriggerBlur) {
      const trigger = (0, _findDOMNode.findDOMNode)(this._trigger);
      if (trigger && typeof trigger.focus === 'function') {
        ;
        trigger.focus();
      }
    }
    this.hide(event, documentClick);
  };
  handleDialogBlur = event => {
    if (event.keyCode === _keycode.default.codes.tab && event.shiftKey && this.props.shouldFocusContentOnTriggerBlur) {
      return;
    }
    this.hide(event);
  };
  handleTriggerKeyDown = event => {
    if (!this.props.shouldFocusContentOnTriggerBlur) {
      return;
    }
    if (event.keyCode === _keycode.default.codes.tab && !event.shiftKey) {
      event.preventDefault();
      this._raf.push((0, _requestAnimationFrame.requestAnimationFrame)(() => {
        this._dialog && this._dialog.focus();
      }));
    }
  };
  handleTriggerKeyUp = event => {
    if (event.keyCode === _keycode.default.codes.esc && this.shown && this.isTooltip) {
      // if popover is tooltip, it is managing its own focus region so we need
      // to prevent esc keyup event from reaching FocusRegionManager
      event.preventDefault();
      this.hide(event);
    }
  };
  handleTriggerBlur = event => {
    const {
      on
    } = this.props;
    if (on && on.indexOf('focus') > -1) {
      this._raf.push((0, _requestAnimationFrame.requestAnimationFrame)(() => {
        if (!(0, _containsActiveElement.containsActiveElement)(this._view)) {
          this.hide(event);
        }
      }));
    }
  };
  handlePositioned = position => {
    const placement = position.placement;
    this.setState({
      placement,
      ...this.computeOffsets(placement)
    });
    this.props.onPositioned?.(position);
  };
  handlePositionChanged = position => {
    const placement = position.placement;
    this.setState({
      placement,
      ...this.computeOffsets(placement)
    });
    this.props.onPositionChanged?.(position);
  };
  renderTrigger() {
    if (this._renderTriggerProp != this.props.renderTrigger) {
      this._renderTriggerProp = this.props.renderTrigger;
      this._renderTrigger = (0, _callRenderProp.callRenderProp)(this.props.renderTrigger);
    }
    let trigger = this._renderTrigger;
    if (trigger) {
      const {
        on
      } = this.props;
      let onClick = undefined;
      let onFocus = undefined;
      let onMouseOut = undefined;
      let onMouseOver = undefined;
      let expanded;
      if (on && on.indexOf('click') > -1) {
        onClick = event => {
          this.toggle(event);
        };
      }
      if (on && on.indexOf('hover') > -1) {
        (0, _console.logError)(!(on === 'hover'), '[Popover] Specifying only the `"hover"` trigger limits the visibility' + ' of the Popover to just mouse users. Consider also including the `"focus"` trigger ' + 'so that touch and keyboard only users can see the Popover content as well.');
        onMouseOver = this._handleMouseOver;
        onMouseOut = this._handleMouseOut;
      }
      if (on && on.indexOf('focus') > -1) {
        onFocus = event => {
          this.show(event);
        };
      }
      if (this.props.shouldSetAriaExpanded) {
        // only set aria-expanded if popover can contain focus
        expanded = this.shown ? 'true' : 'false';
        if ('aria-expanded' in this.props) {
          // @ts-expect-error It is an escape hatch, in case someone
          // wants to remove/override aria-expanded even when shouldSetAriaExpanded is set
          expanded = this.props['aria-expanded'];
        }
      } else {
        expanded = undefined;
      }
      trigger = (0, _safeCloneElement.safeCloneElement)(trigger, {
        ref: el => {
          this._trigger = el;
        },
        'aria-expanded': expanded,
        'data-popover-trigger': true,
        onKeyDown: (0, _createChainedFunction.createChainedFunction)(this.handleTriggerKeyDown, this.props.onKeyDown),
        onKeyUp: (0, _createChainedFunction.createChainedFunction)(this.handleTriggerKeyUp, this.props.onKeyUp),
        onClick: (0, _createChainedFunction.createChainedFunction)(onClick, this.props.onClick),
        onBlur: (0, _createChainedFunction.createChainedFunction)(this.handleTriggerBlur, this.props.onBlur),
        onFocus: (0, _createChainedFunction.createChainedFunction)(onFocus, this.props.onFocus),
        onMouseOut: (0, _createChainedFunction.createChainedFunction)(onMouseOut, this.props.onMouseOut),
        onMouseOver: (0, _createChainedFunction.createChainedFunction)(onMouseOver, this.props.onMouseOver)
      });
    }
    return trigger;
  }
  renderContent() {
    let content = (0, _callRenderProp.callRenderProp)(this.props.children);
    if (this.shown && !this.isTooltip) {
      // if popover is NOT being used as a tooltip, create a Dialog
      // to manage the content FocusRegion, when showing
      content = (0, _jsxRuntime.jsx)(_Dialog.Dialog, {
        open: this.shown,
        label: this.props.screenReaderLabel,
        ref: el => {
          this._dialog = el;
        },
        display: "block",
        onBlur: this.handleDialogBlur,
        onDismiss: this.handleDialogDismiss,
        liveRegion: this.props.liveRegion,
        defaultFocusElement: this.props.defaultFocusElement,
        shouldContainFocus: this.props.shouldContainFocus,
        shouldReturnFocus: this.props.shouldReturnFocus,
        shouldFocusOnOpen: !this.props.shouldFocusContentOnTriggerBlur,
        shouldCloseOnDocumentClick: this.props.shouldCloseOnDocumentClick,
        shouldCloseOnEscape: this.props.shouldCloseOnEscape,
        children: content
      });
    }
    if (this.shown || this.props.shouldRenderOffscreen) {
      const color = this.props.color;
      let viewProps = {
        // TODO: try to type `ref` better, LegacyRef<T> was not compatible
        ref: c => this._view = c,
        elementRef: el => {
          this._contentElement = el;
          this.props.contentRef?.(el);
        },
        background: color,
        stacking: this.props.stacking,
        shadow: this.props.shadow,
        display: 'block'
      };
      if (this.isTooltip) {
        viewProps = {
          ...viewProps,
          // Because of a11y reasons popovers should not be hidden when hovered over
          onMouseOver: this._handleMouseOver,
          onMouseOut: this._handleMouseOut
        };
      }
      const {
        placement
      } = this.state;
      const {
        styles
      } = this.props;
      if (this.props.withArrow) {
        viewProps = {
          ...viewProps,
          // TODO: remove background override after contextview is updated
          background: color === 'primary' ? 'default' : 'inverse',
          placement: this.props.dir === _textDirectionContextConsumer.textDirectionContextConsumer.DIRECTION.rtl ? (0, _mirrorHorizontalPlacement.mirrorHorizontalPlacement)(placement, ' ') : placement
        };
        return (0, _jsxRuntime.jsx)(_latest.ContextView, {
          ...viewProps,
          borderColor: styles?.borderColor,
          children: content
        });
      } else {
        viewProps = {
          ...viewProps,
          borderWidth: this.props.borderWidth || 'small',
          borderRadius: 'medium',
          ...(color === 'primary-inverse' && {
            borderColor: 'transparent'
          })
        };
        return (0, _jsxRuntime.jsx)(_latest.View, {
          ...viewProps,
          borderColor: styles?.borderColor,
          borderRadius: styles?.borderRadius,
          children: content
        });
      }
    } else {
      return null;
    }
  }
  render() {
    const positionProps = this.positionProps;
    if (this.props.positionTarget) {
      return (0, _jsxRuntime.jsxs)("span", {
        ref: this.handleRef,
        children: [this.renderTrigger(), (0, _jsxRuntime.jsx)(_Position.Position, {
          ...positionProps,
          "data-cid": (0, _combineDataCid.combineDataCid)('Popover', this.props),
          children: this.renderContent()
        })]
      });
    } else {
      return (0, _jsxRuntime.jsx)(_Position.Position, {
        "data-cid": (0, _combineDataCid.combineDataCid)('Popover', this.props),
        ...positionProps,
        renderTarget: this.renderTrigger(),
        elementRef: this.handleRef,
        children: this.renderContent()
      });
    }
  }
}) || _class) || _class) || _class);
var _default = exports.default = Popover;
