// @instructure/ui-color-picker/lib/ColorPicker/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ColorPicker = void 0;
require("core-js/modules/es.array.includes.js");
var _react = require("react");
var _emotion = require("@instructure/emotion");
var _console = require("@instructure/console");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _isValid = require("@instructure/ui-color-utils/lib/isValid.js");
var _contrast = require("@instructure/ui-color-utils/lib/contrast.js");
var _uiColorUtils = _interopRequireDefault(require("@instructure/ui-color-utils"));
var _v11_ = require("@instructure/ui-text-input/v11_6");
var _v11_2 = require("@instructure/ui-tooltip/v11_6");
var _v11_3 = require("@instructure/ui-buttons/v11_6");
var _v11_4 = require("@instructure/ui-popover/v11_6");
var _IconCheckDarkLine = require("@instructure/ui-icons/lib/generated/IconCheckDarkLine.js");
var _IconWarningLine = require("@instructure/ui-icons/lib/generated/IconWarningLine.js");
var _IconTroubleLine = require("@instructure/ui-icons/lib/generated/IconTroubleLine.js");
var _IconInfoLine = require("@instructure/ui-icons/lib/generated/IconInfoLine.js");
var _v = _interopRequireDefault(require("../../ColorIndicator/v1"));
var _v2 = _interopRequireDefault(require("../../ColorMixer/v1"));
var _v3 = _interopRequireDefault(require("../../ColorContrast/v1"));
var _v4 = _interopRequireDefault(require("../../ColorPreset/v1"));
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
const acceptedCharactersForHEX = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'A', 'b', 'B', 'c', 'C', 'd', 'D', 'e', 'E', 'f', 'F', null];

/**
---
category: components
---
**/
let ColorPicker = exports.ColorPicker = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class ColorPicker extends _react.Component {
  static displayName = "ColorPicker";
  static allowedProps = _props.allowedProps;
  static componentId = 'ColorPicker';
  static defaultProps = {
    disabled: false,
    withAlpha: false,
    width: '22.5rem',
    popoverMaxHeight: '100vh'
  };
  constructor(props) {
    super(props);
    this.state = {
      hexCode: '',
      showHelperErrorMessages: false,
      openColorPicker: false,
      mixedColor: '',
      labelHeight: 0,
      calculatedPopoverMaxHeight: undefined,
      isHeightCalculated: false
    };
  }
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  inputContainerRef = null;
  handleInputContainerRef = el => {
    this.inputContainerRef = el;
    if (el) {
      // Defer measurement until after layout is complete and CSS-in-JS styles are applied
      requestAnimationFrame(() => {
        this.setLabelHeight();
      });
    }
  };
  popoverContentRef = null;
  handlePopoverContentRef = el => {
    this.popoverContentRef = el;
  };
  setLabelHeight = () => {
    if (this.inputContainerRef) {
      this.setState({
        labelHeight: this.inputContainerRef.getBoundingClientRect().y - (this.inputContainerRef.parentElement?.getBoundingClientRect().y || 0)
      });
    }
  };

  // Calculate the maximum height the popover can have without extending beyond
  // the viewport. This enables scrolling when the ColorPicker's content (all
  // color mixing controls, presets, and contrast checker) would otherwise exceed
  // the available viewport space. Without this calculation, the popover would
  // render off-screen on smaller viewports.
  handlePopoverPositioned = position => {
    if (this.popoverContentRef) {
      // Double requestAnimationFrame ensures measurements happen after all child components
      // (ColorMixer, ColorPreset, ColorContrast) complete their mount lifecycle and Emotion
      // finishes injecting CSS-in-JS styles. A single rAF was insufficient as styles are
      // injected dynamically in componentDidMount(). This timing issue only manifested when
      // StrictMode was disabled, since StrictMode's double-rendering provided an accidental
      // second measurement pass.
      requestAnimationFrame(() => {
        // First frame: DOM structure is laid out
        requestAnimationFrame(() => {
          // Second frame: styles injected, child components mounted, dimensions stable
          if (!this.popoverContentRef) return;
          const rect = this.popoverContentRef.getBoundingClientRect();
          const viewportHeight = window.innerHeight;

          // Detect if popover is positioned above (top) or below (bottom) the trigger.
          // The Position component provides placement strings like "top center" or "bottom center".
          const placement = position?.placement || '';
          const isPositionedAbove = placement.startsWith('top');
          let availableHeight;
          if (isPositionedAbove) {
            // When opening upward: available space is from viewport top to popover bottom.
            // This is the space where the popover can expand within the viewport.
            availableHeight = rect.top + rect.height - 16;
          } else {
            // When opening downward: available space is from popover top to viewport bottom.
            // Subtract a small buffer (16px) for padding/margin.
            availableHeight = viewportHeight - rect.top - 16;
          }
          const propMaxHeight = this.props.popoverMaxHeight;
          let calculatedMaxHeight = `${Math.max(100, availableHeight)}px`;

          // If prop specifies a maxHeight, respect it as an additional constraint
          if (propMaxHeight && propMaxHeight !== '100vh') {
            calculatedMaxHeight = propMaxHeight;
          }
          this.setState({
            calculatedPopoverMaxHeight: calculatedMaxHeight,
            isHeightCalculated: true
          });
        });
      });
    }
  };
  componentDidMount() {
    this.props.makeStyles?.({
      ...this.state,
      isSimple: this.isSimple
    });
    this.checkSettings();
    this.props.value && this.setState({
      hexCode: this.props.value?.slice(1)
    });
  }
  componentDidUpdate(prevProps) {
    this.props.makeStyles?.({
      ...this.state,
      isSimple: this.isSimple
    });
    if (prevProps.value !== this.props.value) {
      this.setState({
        showHelperErrorMessages: false,
        hexCode: this.props.value?.slice(1) || ''
      });
    }
    this.checkSettings();
  }
  checkSettings = () => {
    if (this.props.children && this.props.colorMixerSettings) {
      (0, _console.warn)(false, 'You should either use children, colorMixerSettings or neither, not both. In this case, the colorMixerSettings will be ignored.', '');
    }
    if (this.props.value && typeof this.props.onChange !== 'function') {
      (0, _console.error)(false, 'You provided a `value` prop on ColorPicker, which will render a controlled component. Please provide an `onChange` handler.');
    }
  };
  get renderMode() {
    if (this.props.children) {
      return 'customPopover';
    }
    if (this.props.colorMixerSettings) {
      return 'defaultPopover';
    }
    return 'simple';
  }
  get isSimple() {
    return this.renderMode === 'simple';
  }
  get isDefaultPopover() {
    return this.renderMode === 'defaultPopover';
  }
  get isCustomPopover() {
    return this.renderMode === 'customPopover';
  }
  get mixedColorWithStrippedAlpha() {
    const {
      mixedColor
    } = this.state;
    return mixedColor.length === 8 && mixedColor.slice(-2) === 'FF' ? mixedColor.slice(0, -2) : mixedColor;
  }
  getMinContrast(strength) {
    return {
      min: 3,
      mid: 4.5,
      max: 7
    }[strength];
  }
  renderMessages() {
    const {
      hexCode,
      showHelperErrorMessages
    } = this.state;
    const isValidHex = (0, _isValid.isValid)(hexCode);
    const {
      checkContrast,
      renderMessages,
      renderInvalidColorMessage,
      renderIsRequiredMessage,
      isRequired
    } = this.props;
    const contrast = isValidHex ? (0, _contrast.contrast)(this.props.checkContrast?.contrastAgainst || '#fff', hexCode, 2) : undefined;
    const contrastStrength = checkContrast?.contrastStrength ? checkContrast.contrastStrength : 'mid';
    const minContrast = this.getMinContrast(contrastStrength);
    let invalidColorMessages = [];
    let isRequiredMessages = [];
    let generalMessages = [];
    let contrastMessages = [];
    if (checkContrast && contrast) {
      const {
        renderContrastSuccessMessage,
        renderContrastErrorMessage
      } = checkContrast;
      if (contrast < minContrast) {
        contrastMessages = typeof renderContrastErrorMessage === 'function' ? renderContrastErrorMessage(contrast, minContrast) : checkContrast.isStrict ? [{
          type: 'error',
          text: ''
        }] : [];
      } else if (typeof renderContrastSuccessMessage === 'function') {
        contrastMessages = renderContrastSuccessMessage(contrast, minContrast);
      }
    }
    if (showHelperErrorMessages && hexCode !== '' && !isValidHex && typeof renderInvalidColorMessage === 'function') {
      invalidColorMessages = renderInvalidColorMessage(hexCode);
    }
    if (isRequired && showHelperErrorMessages && hexCode === '') {
      isRequiredMessages = typeof renderIsRequiredMessage === 'function' ? renderIsRequiredMessage() : [{
        type: 'error',
        text: ''
      }];
    }
    if (typeof renderMessages === 'function') {
      generalMessages = renderMessages(hexCode, isValidHex, minContrast, contrast);
    }
    return [...invalidColorMessages, ...isRequiredMessages, ...generalMessages, ...contrastMessages];
  }
  renderBeforeInput() {
    const {
      styles
    } = this.props;
    return (0, _jsxRuntime.jsxs)("div", {
      css: styles?.simpleColorContainer,
      children: [this.isSimple && (0, _jsxRuntime.jsx)(_v.default, {
        color: `#${this.state.hexCode}`
      }), (0, _jsxRuntime.jsx)("div", {
        css: styles?.hashMarkContainer,
        children: "#"
      })]
    });
  }
  renderAfterInput() {
    const {
      checkContrast,
      styles
    } = this.props;
    const {
      hexCode
    } = this.state;
    if (checkContrast && (0, _isValid.isValid)(hexCode)) {
      const contrast = (0, _contrast.contrast)(checkContrast.contrastAgainst || '#fff', hexCode, 2);
      const minContrast = this.getMinContrast(checkContrast.contrastStrength ? checkContrast.contrastStrength : 'mid');
      if (minContrast >= contrast) {
        return (0, _jsxRuntime.jsx)("div", {
          css: styles?.errorIcons,
          children: checkContrast.isStrict ? (0, _jsxRuntime.jsx)(_IconTroubleLine.IconTroubleLine, {}) : (0, _jsxRuntime.jsx)(_IconWarningLine.IconWarningLine, {})
        });
      }
      return (0, _jsxRuntime.jsx)("div", {
        css: styles?.successIcon,
        children: (0, _jsxRuntime.jsx)(_IconCheckDarkLine.IconCheckDarkLine, {})
      });
    }
    return null;
  }
  handleOnChange(_event, value) {
    const {
      onChange
    } = this.props;
    if (value.length > (this.props.withAlpha ? 8 : 6) || value.split('').find(char => !acceptedCharactersForHEX.includes(char)) !== undefined) {
      return;
    }
    if (typeof onChange === 'function') {
      onChange(`#${value}`);
    } else {
      this.setState({
        showHelperErrorMessages: false,
        hexCode: value
      });
    }
  }
  handleOnPaste(event) {
    const pasted = event.clipboardData.getData('Text');
    let toPaste = pasted;
    if (pasted[0] && pasted[0] === '#') {
      toPaste = pasted.slice(1);
    }
    if (this.props.withAlpha && toPaste.length > 8 || !this.props.withAlpha && toPaste.length > 6) {
      return event.preventDefault();
    }
    const input = event.target;
    const currentHex = this.state.hexCode;
    const selectionStart = input.selectionStart ?? currentHex.length;
    const selectionEnd = input.selectionEnd ?? currentHex.length;
    const newHex = currentHex.slice(0, selectionStart) + toPaste + currentHex.slice(selectionEnd);
    if ((0, _isValid.isValid)(newHex)) {
      if (typeof this.props.onChange === 'function') {
        this.props.onChange(`#${newHex}`);
      } else {
        this.setState({
          showHelperErrorMessages: false,
          hexCode: newHex
        });
      }
    }
    event.preventDefault();
  }
  handleOnBlur() {
    this.setState({
      showHelperErrorMessages: true
    });
  }
  renderLabel() {
    const {
      label,
      tooltip,
      styles
    } = this.props;
    return tooltip ? (0, _jsxRuntime.jsxs)("span", {
      children: [(0, _jsxRuntime.jsx)("span", {
        css: styles?.label,
        children: label
      }), (0, _jsxRuntime.jsx)("span", {
        children: (0, _jsxRuntime.jsx)(_emotion.InstUISettingsProvider, {
          theme: {
            componentOverrides: {
              BaseButton: {
                smallHeight: 'auto'
              }
            }
          },
          children: (0, _jsxRuntime.jsx)(_v11_2.Tooltip, {
            renderTip: (0, _jsxRuntime.jsx)("span", {
              "aria-hidden": true,
              children: tooltip
            }),
            children: (0, _jsxRuntime.jsx)(_v11_3.IconButton, {
              withBackground: false,
              withBorder: false,
              screenReaderLabel: tooltip,
              size: "small",
              shape: "circle",
              width: "auto",
              renderIcon: _IconInfoLine.IconInfoLine
            })
          })
        })
      })]
    }) : label;
  }
  renderPopover = () => (0, _jsxRuntime.jsx)(_v11_4.Popover, {
    renderTrigger: (0, _jsxRuntime.jsx)(_v11_3.IconButton, {
      disabled: this.props.disabled,
      screenReaderLabel: this.props.popoverButtonScreenReaderLabel || '',
      children: (0, _jsxRuntime.jsx)(_v.default, {
        color: `#${this.state.hexCode}`
      })
    }),
    isShowingContent: this.state.openColorPicker,
    onShowContent: () => {
      this.setState({
        openColorPicker: true,
        mixedColor: this.state.hexCode,
        calculatedPopoverMaxHeight: undefined,
        isHeightCalculated: false
      });
    },
    onHideContent: () => {
      this.setState({
        openColorPicker: false
      });
    },
    on: "click",
    screenReaderLabel: this.props.popoverScreenReaderLabel,
    shouldContainFocus: true,
    shouldReturnFocus: true,
    shouldCloseOnDocumentClick: true,
    offsetY: "10rem",
    onPositioned: this.handlePopoverPositioned,
    onPositionChanged: this.handlePopoverPositioned,
    children: (0, _jsxRuntime.jsx)("div", {
      css: this.props.styles?.popoverContentContainer,
      ref: this.handlePopoverContentRef,
      children: this.isDefaultPopover ? this.renderDefaultPopoverContent() : this.renderCustomPopoverContent()
    })
  });
  renderCustomPopoverContent = () => {
    const {
      children,
      onChange
    } = this.props;
    return (0, _jsxRuntime.jsx)("div", {
      children: typeof children === 'function' && children(`#${this.mixedColorWithStrippedAlpha}`, newColor => {
        this.setState({
          mixedColor: _uiColorUtils.default.colorToHex8(newColor).slice(1)
        });
      }, () => {
        this.setState({
          openColorPicker: false,
          hexCode: this.mixedColorWithStrippedAlpha
        });
        onChange?.(`#${this.mixedColorWithStrippedAlpha}`);
      }, () => this.setState({
        openColorPicker: false,
        mixedColor: this.state.hexCode,
        calculatedPopoverMaxHeight: undefined,
        isHeightCalculated: false
      }))
    });
  };
  renderDefaultPopoverContent = () => (0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
    children: [(0, _jsxRuntime.jsxs)("div", {
      css: this.props.styles?.popoverContent,
      children: [this.props?.colorMixerSettings?.colorMixer && (0, _jsxRuntime.jsx)(_v2.default, {
        value: `#${this.state.mixedColor}`,
        onChange: newColor => this.setState({
          mixedColor: _uiColorUtils.default.colorToHex8(newColor).slice(1)
        }),
        withAlpha: this.props.colorMixerSettings.colorMixer.withAlpha,
        rgbRedInputScreenReaderLabel: this.props.colorMixerSettings.colorMixer.rgbRedInputScreenReaderLabel,
        rgbGreenInputScreenReaderLabel: this.props.colorMixerSettings.colorMixer.rgbGreenInputScreenReaderLabel,
        rgbBlueInputScreenReaderLabel: this.props.colorMixerSettings.colorMixer.rgbBlueInputScreenReaderLabel,
        rgbAlphaInputScreenReaderLabel: this.props.colorMixerSettings.colorMixer.rgbAlphaInputScreenReaderLabel,
        colorSliderNavigationExplanationScreenReaderLabel: this.props.colorMixerSettings.colorMixer.colorSliderNavigationExplanationScreenReaderLabel,
        alphaSliderNavigationExplanationScreenReaderLabel: this.props.colorMixerSettings.colorMixer.alphaSliderNavigationExplanationScreenReaderLabel,
        colorPaletteNavigationExplanationScreenReaderLabel: this.props.colorMixerSettings.colorMixer.colorPaletteNavigationExplanationScreenReaderLabel
      }), this.props?.colorMixerSettings?.colorPreset && (0, _jsxRuntime.jsx)("div", {
        css: this.props.styles?.popoverContentBlock,
        children: (0, _jsxRuntime.jsx)(_v4.default, {
          label: this.props.colorMixerSettings.colorPreset.label,
          colors: this.props.colorMixerSettings.colorPreset.colors,
          selected: this.mixedColorWithStrippedAlpha ? `#${this.mixedColorWithStrippedAlpha}` : undefined,
          onSelect: color => this.setState({
            mixedColor: color.slice(1)
          })
        })
      }), this.props?.colorMixerSettings?.colorContrast && (0, _jsxRuntime.jsx)("div", {
        css: this.props.styles?.popoverContentBlock,
        children: (0, _jsxRuntime.jsx)(_v3.default, {
          firstColor: this.props.colorMixerSettings.colorContrast.firstColor,
          secondColor: `#${this.mixedColorWithStrippedAlpha}`,
          label: this.props.colorMixerSettings.colorContrast.label,
          successLabel: this.props.colorMixerSettings.colorContrast.successLabel,
          failureLabel: this.props.colorMixerSettings.colorContrast.failureLabel,
          normalTextLabel: this.props.colorMixerSettings.colorContrast.normalTextLabel,
          largeTextLabel: this.props.colorMixerSettings.colorContrast.largeTextLabel,
          graphicsTextLabel: this.props.colorMixerSettings.colorContrast.graphicsTextLabel,
          firstColorLabel: this.props.colorMixerSettings.colorContrast.firstColorLabel,
          secondColorLabel: this.props.colorMixerSettings.colorContrast.secondColorLabel,
          onContrastChange: this.props.colorMixerSettings.colorContrast.onContrastChange
        })
      })]
    }), (0, _jsxRuntime.jsxs)("div", {
      css: this.props.styles?.popoverFooter,
      children: [(0, _jsxRuntime.jsx)(_v11_3.Button, {
        onClick: () => this.setState({
          openColorPicker: false,
          mixedColor: this.state.hexCode,
          calculatedPopoverMaxHeight: undefined,
          isHeightCalculated: false
        }),
        color: "secondary",
        margin: "xx-small",
        children: this.props.colorMixerSettings?.popoverCloseButtonLabel
      }), (0, _jsxRuntime.jsx)(_v11_3.Button, {
        onClick: () => {
          if (typeof this.props.onChange === 'function') {
            this.props.onChange(`#${this.mixedColorWithStrippedAlpha}`);
            this.setState({
              openColorPicker: false
            });
          } else {
            this.setState({
              openColorPicker: false,
              hexCode: `${this.mixedColorWithStrippedAlpha}`
            });
          }
        },
        color: "primary",
        margin: "xx-small",
        children: this.props.colorMixerSettings?.popoverAddButtonLabel
      })]
    })]
  });
  render() {
    const {
      disabled,
      isRequired,
      placeholderText,
      width,
      id,
      inputRef
    } = this.props;
    return (0, _jsxRuntime.jsxs)("div", {
      ...(0, _omitProps.omitProps)(this.props, ColorPicker.allowedProps),
      css: this.props.styles?.colorPicker,
      ref: this.handleRef,
      "data-cid": "ColorPicker",
      children: [(0, _jsxRuntime.jsx)(_v11_.TextInput, {
        id: id,
        isRequired: isRequired,
        disabled: disabled,
        renderLabel: () => this.renderLabel(),
        display: "inline-block",
        width: width,
        placeholder: placeholderText,
        themeOverride: {
          padding: ''
        },
        renderAfterInput: this.renderAfterInput(),
        renderBeforeInput: this.renderBeforeInput(),
        inputContainerRef: this.handleInputContainerRef,
        value: this.state.hexCode,
        onChange: (event, value) => this.handleOnChange(event, value),
        onPaste: event => this.handleOnPaste(event),
        onBlur: () => this.handleOnBlur(),
        messages: this.renderMessages(),
        inputRef: inputRef
      }), !this.isSimple && (0, _jsxRuntime.jsx)("div", {
        css: this.props.styles?.colorMixerButtonContainer,
        style: {
          alignSelf: this.state.labelHeight > 0 ? 'flex-start' : 'flex-end',
          paddingTop: this.state.labelHeight
        },
        children: (0, _jsxRuntime.jsx)("div", {
          css: this.props.styles?.colorMixerButtonWrapper,
          children: this.renderPopover()
        })
      })]
    });
  }
}) || _class);
var _default = exports.default = ColorPicker;
