// @instructure/ui-themes/es/themes/newThemeTokens/legacyCanvas/components/aiInformation.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const aiInformation = semantics => ({
  bodyPadding: semantics.spacing.spaceXl,
  headingBottomMargin: semantics.spacing.spaceXl,
  featureNameBottomMargin: semantics.spacing.spaceLg,
  permissionLevelTextBottomMargin: semantics.spacing.spaceSm,
  permissionLevelBottomMargin: semantics.spacing.spaceSm,
  descriptionBottomMargin: semantics.spacing.spaceSm,
  modelNameTextBottomMargin: semantics.spacing.spaceXs,
  modelNameBottomMargin: semantics.spacing.spaceXs,
  dividerMargin: semantics.spacing.spaceLg,
  dividerColor: semantics.color.stroke.muted,
  levelColor: semantics.color.text.aiColor,
  permissionLevelsModalTriggerBottomMargin: semantics.spacing.spaceXl
});
export default aiInformation;
