// @instructure/ui-themes/es/themes/newThemeTokens/legacyCanvas/components/sideNavBarNew.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const sideNavBarNew = semantics => ({
  backgroundColor: semantics.color.background.dark,
  fill: semantics.color.icon.onColor,
  textColor: semantics.color.text.onColor,
  minimizedminWidth: "5.75rem",
  toggleTransition: "300ms",
  minWidth: "15.75rem",
  margin: "0rem",
  minimizedMargin: "0rem",
  mainNavMargin: "0rem",
  mainNavMinimizedMargin: "0rem",
  mainNavGap: "0rem",
  navItemsGap: "0rem",
  boxShadow: {
    0: {
      x: "0rem",
      y: "0rem",
      blur: "0rem",
      spread: "0rem",
      color: semantics.color.transparentColor,
      type: "dropShadow"
    },
    1: {
      x: "0rem",
      y: "0rem",
      blur: "0rem",
      spread: "0rem",
      color: semantics.color.transparentColor,
      type: "dropShadow"
    }
  },
  borderRadius: "0rem",
  minimizedBorderRadius: "0rem",
  borderColor: semantics.color.stroke.base,
  borderWidth: semantics.borderWidth.sm,
  item: {
    iconWrapperHeight: "2.25rem",
    borderRadius: "0rem",
    backgroundColor: "transparent",
    horizontalPadding: "0.5rem",
    verticalPadding: "0.375rem",
    fontColor: semantics.color.text.onColor,
    fontFamily: semantics.fontFamily.base,
    fontSize: semantics.fontSize.textBase,
    fontWeight: semantics.fontWeight.body.base,
    hoverBackgroundColor: semantics.color.background.interactive.action.primary.hover,
    lineHeight: semantics.lineHeight.paragraph.base,
    linkTextDecoration: "none",
    currentBackgroundColor: semantics.color.background.base,
    selectedBackgroundColor: semantics.color.background.interactive.action.tertiary.active,
    selectedAccountBackgroundColor: semantics.color.background.interactive.action.tertiary.active,
    currentTextColor: semantics.color.text.interactive.navigation.primary.base,
    selectedAccountFontColor: semantics.color.text.interactive.navigation.primary.base,
    innerFocusOutlineColor: semantics.color.stroke.interactive.focusRing.onColor,
    selectedInnerFocusOutlineColor: semantics.color.stroke.interactive.focusRing.base,
    outerFocusOutlineColor: semantics.color.stroke.base,
    selectedOuterFocusOutlineColor: semantics.color.stroke.interactive.focusRing.onColor
  },
  drilldownOption: {
    backgroundColor: semantics.color.transparentColor,
    hoverBackgroundColor: semantics.color.background.interactive.action.primary.hover,
    activeBackgroundColor: semantics.color.background.interactive.action.primary.active,
    selectedBackgroundColor: semantics.color.background.brand,
    textColor: semantics.color.text.onColor,
    selectedTextColor: semantics.color.text.interactive.navigation.primary.base,
    logoutTextColor: semantics.color.text.onColor,
    borderRadius: "0rem"
  },
  button: {
    textColor: semantics.color.text.onColor
  },
  subNav: {
    strokeColor: semantics.color.stroke.base,
    textColor: semantics.color.text.onColor,
    padding: semantics.spacing.spaceMd,
    gap: semantics.spacing.spaceSm
  },
  institutionLogo: {
    textColor: semantics.color.text.onColor,
    strokeColor: semantics.color.stroke.base
  },
  insitutionLogo: {
    padding: semantics.spacing.spaceSm,
    gap: semantics.spacing.spaceMd
  }
});
export default sideNavBarNew;
