// @instructure/ui-themes/es/themes/newThemeTokens/legacyCanvas/components/sourceCodeEditor.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const sourceCodeEditor = semantics => ({
  background: semantics.color.background.base,
  borderColor: semantics.color.stroke.base,
  borderRadius: semantics.borderRadius.md,
  borderWidth: semantics.borderWidth.sm,
  color: semantics.color.text.base,
  fontFamily: semantics.fontFamily.code,
  fontSize: semantics.fontSize.textSm,
  gutterBackground: semantics.color.background.muted,
  horizontalPadding: semantics.spacing.spaceSm,
  verticalPadding: "0.375rem",
  focusedSelectionBackgroundColor: "#d7d4f0",
  placeholderBackgroundColor: "#707070",
  tagKeywordColor: "#708",
  tagLiteralColor: "#164",
  tagStringColor: "#a11",
  tagDefinitionVariableNameColor: "#00f",
  tagLocalVariableNameColor: "#30a",
  tagClassNameColor: "#167",
  tagSpecialVariableNameColor: "#256",
  tagDefinitionPropertyNameColor: "#00c",
  tagCommentColor: "#940",
  tagMetaColor: "#757075",
  tagRegExpColor: "#d13b00",
  tagTypeNameColor: "#008051",
  tagInvalidColor: "#e60000",
  tagAtomColor: "#219",
  activeLineColor: "#99999912"
});
export default sourceCodeEditor;
