// @instructure/ui-themes/es/themes/newThemeTokens/legacyCanvas/components/toggleDetails.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const toggleDetails = semantics => ({
  fontFamily: semantics.fontFamily.base,
  fontSizeLarge: semantics.fontSize.textLg,
  fontSizeMedium: semantics.fontSize.textBase,
  fontSizeSmall: semantics.fontSize.textSm,
  fontWeight: semantics.fontWeight.body.base,
  iconMargin: "0.375rem",
  lineHeight: semantics.lineHeight.paragraph.base,
  textColor: semantics.color.institutional.brandFontColorDark,
  toggleBorderRadius: semantics.borderRadius.interactive.base,
  toggleBorderStyle: "solid",
  toggleBorderWidth: semantics.borderWidth.md,
  togglePadding: "0.375rem",
  contentPaddingSmall: "1.125rem",
  contentPaddingMedium: "1.125rem",
  contentPaddingLarge: "1.375rem"
});
export default toggleDetails;
