// @instructure/ui-position/lib/Position/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Position = void 0;
var _react = require("react");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _ensureSingleChild = require("@instructure/ui-react-utils/lib/ensureSingleChild.js");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _addPositionChangeListener = require("@instructure/ui-dom-utils/lib/addPositionChangeListener.js");
var _findDOMNode = require("@instructure/ui-dom-utils/lib/findDOMNode.js");
var _deepEqual = require("@instructure/ui-utils/lib/deepEqual.js");
var _shallowEqual = require("@instructure/ui-utils/lib/shallowEqual.js");
var _combineDataCid = require("@instructure/ui-utils/lib/combineDataCid.js");
var _debounce = require("@instructure/debounce");
var _Portal = require("@instructure/ui-portal/lib/Portal");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _calculateElementPosition = require("../calculateElementPosition");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components/utilities
---
**/
let Position = exports.Position = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = _dec2(_class = class Position extends _react.Component {
  static displayName = "Position";
  static componentId = 'Position';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    placement: 'bottom center',
    mountNode: null,
    insertAt: 'bottom',
    constrain: 'window',
    offsetX: 0,
    offsetY: 0,
    shouldTrackPosition: true,
    shouldPositionOverTarget: false,
    children: null
  };
  static locatorAttribute = 'data-position';
  static targetLocatorAttribute = 'data-position-target';
  static contentLocatorAttribute = 'data-position-content';
  constructor(props) {
    super(props);
    this.state = {
      positioned: false,
      ...this.calculatePosition(props)
    };
    this.position = (0, _debounce.debounce)(this.position, 0, {
      leading: false,
      trailing: true
    });
    this._id = this.props.id || props.deterministicId();
  }
  ref = null;
  _id;
  _timeouts = [];
  _listener = null;
  _content;
  _target;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  shouldComponentUpdate(nextProps, nextState, nextContext) {
    return !(0, _deepEqual.deepEqual)(this.state, nextState) || !(0, _shallowEqual.shallowEqual)(this.props, nextProps) || !(0, _shallowEqual.shallowEqual)(this.context, nextContext);
  }
  componentDidMount() {
    this.toggleLocatorAttributes(true);
    this.props.makeStyles?.();
  }
  componentDidUpdate(prevProps, prevState) {
    this.position();
    this.toggleLocatorAttributes(true);
    if (this.props.shouldTrackPosition !== prevProps.shouldTrackPosition) {
      this.props.shouldTrackPosition ? this.startTracking() : this.stopTracking();
    }
    const {
      style,
      placement
    } = this.state;
    if (style && prevState.style && (placement !== prevState.placement || style.top !== prevState.style.top || style.left !== prevState.style.left) && typeof this.props.onPositionChanged === 'function') {
      this.props.onPositionChanged({
        top: style.top,
        left: style.left,
        placement
      });
    }
    this.props.makeStyles?.();
  }
  componentWillUnmount() {
    ;
    this.position.cancel();
    this.stopTracking();
    this._timeouts.forEach(timeout => clearTimeout(timeout));
    this.toggleLocatorAttributes(false);
  }
  toggleLocatorAttributes(set) {
    // We have to find the actual DOM nodes and append the attributes
    // directly, as we can't be sure when safe cloning the child that
    // it will accept the data attribute as a prop
    this.toggleLocatorAttribute((0, _findDOMNode.findDOMNode)(this._content), Position.contentLocatorAttribute, set);
    this.toggleLocatorAttribute((0, _findDOMNode.findDOMNode)(this._target), Position.targetLocatorAttribute, set);
  }
  toggleLocatorAttribute(node, locator, set) {
    if (node && node.hasAttribute) {
      if (set && !node.hasAttribute(locator)) {
        ;
        node.setAttribute(locator, this._id);
      }
      if (!set && node.hasAttribute(locator)) {
        ;
        node.removeAttribute(locator);
      }
    }
  }
  handlePortalOpen = () => {
    this.position();
    if (this.props.shouldTrackPosition) {
      this.startTracking();
    }
    this._timeouts.push(setTimeout(() => {
      if (this.state.positioned && typeof this.props.onPositioned === 'function') {
        this.props.onPositioned({
          top: this.state.style.top,
          left: this.state.style.left,
          placement: this.state.placement
        });
      }
    }, 0));
  };
  calculatePosition(props) {
    return (0, _calculateElementPosition.calculateElementPosition)(this._content, this._target, {
      placement: props.placement,
      offsetX: props.offsetX,
      offsetY: props.offsetY,
      constrain: props.constrain,
      container: props.mountNode,
      over: props.shouldPositionOverTarget
    });
  }
  position = () => {
    this.setState({
      positioned: true,
      ...this.calculatePosition(this.props)
    });
  };
  startTracking() {
    this._listener = this._listener || (0, _addPositionChangeListener.addPositionChangeListener)(this._target, this.position);
  }
  stopTracking() {
    if (this._listener) {
      this._listener.remove();
      this._listener = null;
    }
  }

  // content that needs to be positioned relative to the target
  renderContent() {
    let content = (0, _ensureSingleChild.ensureSingleChild)(this.props.children);
    if (content) {
      content = (0, _safeCloneElement.safeCloneElement)(content, {
        ref: el => {
          this._content = el;
        },
        style: {
          boxSizing: 'border-box',
          zIndex: this.props.styles?.zIndex,
          ...content.props.style,
          ...this.state.style
        },
        ...(content.props.className && {
          className: content.props.className
        }),
        [Position.contentLocatorAttribute]: this._id
      });
      content = (0, _jsxRuntime.jsx)(_Portal.Portal, {
        open: true,
        onOpen: this.handlePortalOpen,
        mountNode: this.props.mountNode,
        insertAt: this.props.insertAt,
        children: content
      });
    }
    return content;
  }
  renderTarget() {
    const target = (0, _callRenderProp.callRenderProp)(this.props.renderTarget);
    if (target) {
      return (0, _safeCloneElement.safeCloneElement)(target, {
        ref: el => {
          this._target = el;
        },
        [Position.targetLocatorAttribute]: this._id
      });
    } else if (this.props.target) {
      this._target = (0, _callRenderProp.callRenderProp)(this.props.target);
    }
    return null;
  }
  render() {
    const props = {
      [Position.locatorAttribute]: this._id
    };
    return (0, _jsxRuntime.jsxs)("span", {
      ...props,
      css: this.props.styles?.position,
      ref: this.handleRef,
      "data-cid": (0, _combineDataCid.combineDataCid)('Position', this.props),
      children: [this.renderTarget(), this.renderContent()]
    });
  }
}) || _class) || _class);
var _default = exports.default = Position;
