// @instructure/ui-tray/lib/Tray/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Tray = void 0;
var _react = require("react");
var _Dialog = require("@instructure/ui-dialog/lib/Dialog");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _createChainedFunction = require("@instructure/ui-utils/lib/createChainedFunction.js");
var _textDirectionContextConsumer = require("@instructure/ui-i18n/lib/textDirectionContextConsumer.js");
var _Portal = require("@instructure/ui-portal/lib/Portal");
var _mirrorHorizontalPlacement = require("@instructure/ui-position/lib/mirrorHorizontalPlacement.js");
var _Transition = require("@instructure/ui-motion/lib/Transition");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _v11_ = require("@instructure/ui-overlays/v11_6");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let Tray = exports.Tray = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec2 = (0, _textDirectionContextConsumer.textDirectionContextConsumer)(), _dec(_class = _dec2(_class = class Tray extends _react.Component {
  static displayName = "Tray";
  static componentId = 'Tray';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    defaultFocusElement: null,
    open: false,
    insertAt: 'bottom',
    shouldCloseOnDocumentClick: false,
    shouldContainFocus: true,
    shouldReturnFocus: true,
    size: 'small',
    placement: 'start',
    transitionOnMount: true,
    transitionEnter: true,
    transitionExit: true,
    shadow: true,
    border: false,
    enableMask: false
  };
  ref = null;
  dialogRef = /*#__PURE__*/(0, _react.createRef)();
  state;
  constructor(props) {
    super(props);
    this.state = {
      transitioning: false,
      open: props.open ?? false
    };
  }
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate(prevProps, _prevState) {
    if (this.props.open !== prevProps.open) {
      if (!this.props.open) {
        // calling Dialog.close() to remove focusregion immediately when transition starts
        // so if a new Tray is opened (during transition) the new focusregion won't get treated as a child of this one
        this.dialogRef.current?.close();
      }
      this.setState({
        transitioning: true,
        open: this.props.open
      });
    }
    this.props.makeStyles?.();
  }
  get placement() {
    const {
      placement,
      dir
    } = this.props;
    const isRtl = dir === _textDirectionContextConsumer.textDirectionContextConsumer.DIRECTION.rtl;
    return isRtl ? (0, _mirrorHorizontalPlacement.mirrorHorizontalPlacement)(placement, ' ') : placement;
  }
  get direction() {
    switch (this.placement) {
      case 'top':
        return 'up';
      case 'bottom':
        return 'down';
      case 'end':
        return 'right';
      default:
        // start
        return 'left';
    }
  }
  get transition() {
    return `slide-${this.direction}`;
  }
  handleTransitionComplete = _type => {
    this.setState({
      transitioning: false
    });
  };
  get DOMNode() {
    // The DOMNode property is needed for Portal type components
    return this.ref;
  }
  set DOMNode(el) {
    this.ref = el ? el : null;
  }
  handlePortalOpen = DOMNode => {
    if (DOMNode) {
      this.DOMNode = DOMNode;
    }
  };
  render() {
    const {
      label,
      children,
      size,
      placement,
      open,
      defaultFocusElement,
      contentRef,
      shouldContainFocus,
      shouldReturnFocus,
      shouldCloseOnDocumentClick,
      onOpen,
      onClose,
      onDismiss,
      mountNode,
      insertAt,
      liveRegion,
      onEnter,
      onEntering,
      onEntered,
      onExit,
      onExiting,
      onExited,
      onTransition,
      transitionOnMount,
      transitionEnter,
      transitionExit,
      border,
      shadow,
      role,
      enableMask,
      ...props
    } = this.props;
    const portalIsOpen = this.state.open || this.state.transitioning;
    const content = (0, _jsxRuntime.jsx)(_Transition.Transition, {
      in: open,
      type: this.transition,
      onTransition: onTransition,
      onEnter: onEnter,
      onEntering: onEntering,
      onEntered: (0, _createChainedFunction.createChainedFunction)(this.handleTransitionComplete, onEntered, onOpen),
      onExit: onExit,
      onExiting: onExiting,
      onExited: (0, _createChainedFunction.createChainedFunction)(this.handleTransitionComplete, onExited, onClose),
      transitionOnMount: transitionOnMount,
      transitionEnter: transitionEnter,
      transitionExit: transitionExit,
      children: (0, _jsxRuntime.jsx)("span", {
        ...(0, _omitProps.omitProps)(props, Tray.allowedProps),
        css: this.props.styles?.tray,
        ref: contentRef,
        children: (0, _jsxRuntime.jsx)(_Dialog.Dialog, {
          ref: this.dialogRef,
          as: "div",
          label: label,
          defaultFocusElement: defaultFocusElement,
          open: true,
          shouldContainFocus: shouldContainFocus,
          shouldReturnFocus: shouldReturnFocus,
          shouldCloseOnDocumentClick: shouldCloseOnDocumentClick,
          shouldCloseOnEscape: true,
          liveRegion: liveRegion,
          onDismiss: onDismiss,
          role: role,
          children: (0, _jsxRuntime.jsx)("div", {
            css: this.props.styles?.content,
            children: children
          })
        })
      })
    });
    return (0, _jsxRuntime.jsx)(_Portal.Portal, {
      open: portalIsOpen,
      onOpen: this.handlePortalOpen,
      insertAt: insertAt,
      mountNode: mountNode,
      "data-cid": "Tray",
      children: enableMask ? (0, _jsxRuntime.jsx)(_v11_.Mask, {
        placement: 'center',
        fullscreen: true,
        children: content
      }) : content
    });
  }
}) || _class) || _class);
var _default = exports.default = Tray;
