// @instructure/ui-themes/lib/themes/newThemeTokens/light/components/sideNavBarNew.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const sideNavBarNew = semantics => ({
  backgroundColor: semantics.color.background.elevatedSurface.base,
  fill: semantics.color.icon.interactive.action.primaryOnColor.base,
  borderColor: semantics.color.stroke.muted,
  textColor: semantics.color.text.base,
  minimizedminWidth: "6.5rem",
  toggleTransition: "300ms",
  minWidth: "17.25rem",
  margin: semantics.spacing.spaceMd,
  minimizedMargin: "0.375rem",
  mainNavMargin: "0.75rem",
  mainNavMinimizedMargin: "0rem",
  mainNavGap: semantics.spacing.padding.container.xxs,
  navItemsGap: semantics.spacing.spaceXs,
  boxShadow: {
    0: {
      x: "0rem",
      y: "0.0625rem",
      blur: "0.125rem",
      spread: "0rem",
      color: semantics.color.dropShadow.shadowColor1,
      type: "dropShadow"
    },
    1: {
      x: "0rem",
      y: "0.125rem",
      blur: "0.25rem",
      spread: "0.125rem",
      color: semantics.color.dropShadow.shadowColor2,
      type: "dropShadow"
    }
  },
  borderRadius: semantics.borderRadius.container.md,
  minimizedBorderRadius: semantics.borderRadius.container.sm,
  borderWidth: semantics.borderWidth.sm,
  item: {
    horizontalPadding: semantics.spacing.spaceSm,
    verticalPadding: "0.25rem",
    iconWrapperHeight: "2.25rem",
    borderRadius: semantics.borderRadius.lg,
    backgroundColor: "transparent",
    fontColor: semantics.color.text.base,
    fontFamily: semantics.fontFamily.base,
    fontSize: semantics.fontSize.textBase,
    fontWeight: semantics.fontWeight.body.base,
    innerFocusOutlineColor: semantics.color.stroke.interactive.focusRing.base,
    outerFocusOutlineColor: semantics.color.transparentColor,
    hoverBackgroundColor: semantics.color.background.interactive.action.tertiary.hover,
    lineHeight: semantics.lineHeight.paragraph.base,
    linkTextDecoration: "none",
    currentBackgroundColor: semantics.color.background.brand,
    selectedBackgroundColor: semantics.color.background.interactive.action.tertiary.active,
    currentTextColor: semantics.color.text.inverse,
    selectedAccountBackgroundColor: semantics.color.background.interactive.action.tertiary.active,
    selectedAccountFontColor: semantics.color.text.base,
    selectedInnerFocusOutlineColor: semantics.color.icon.inverse,
    selectedOuterFocusOutlineColor: semantics.color.transparentColor
  },
  drilldownOption: {
    backgroundColor: semantics.color.transparentColor,
    hoverBackgroundColor: semantics.color.background.interactive.action.tertiary.hover,
    activeBackgroundColor: semantics.color.background.interactive.action.tertiary.active,
    selectedBackgroundColor: semantics.color.background.brand,
    textColor: semantics.color.text.base,
    selectedTextColor: semantics.color.text.inverse,
    logoutTextColor: semantics.color.text.error,
    borderRadius: semantics.borderRadius.lg
  },
  button: {
    textColor: semantics.color.text.base
  },
  subNav: {
    strokeColor: semantics.color.stroke.muted,
    textColor: semantics.color.text.base,
    padding: semantics.spacing.spaceMd,
    gap: semantics.spacing.spaceSm
  },
  institutionLogo: {
    textColor: semantics.color.text.base,
    strokeColor: semantics.color.stroke.muted
  },
  insitutionLogo: {
    padding: semantics.spacing.spaceSm,
    gap: semantics.spacing.spaceMd
  }
});
var _default = exports.default = sideNavBarNew;
