// @instructure/ui-themes/lib/themes/newThemeTokens/light/components/sourceCodeEditor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const sourceCodeEditor = semantics => ({
  background: semantics.color.background.base,
  borderColor: semantics.color.stroke.base,
  borderRadius: semantics.borderRadius.md,
  borderWidth: semantics.borderWidth.sm,
  color: semantics.color.text.base,
  fontFamily: semantics.fontFamily.code,
  fontSize: semantics.fontSize.textSm,
  gutterBackground: semantics.color.background.muted,
  horizontalPadding: semantics.spacing.spaceSm,
  verticalPadding: "0.375rem",
  focusedSelectionBackgroundColor: "rgba(90,75,227,0.2)",
  placeholderBackgroundColor: semantics.color.text.accent.stone,
  tagKeywordColor: semantics.color.text.accent.violet,
  tagLiteralColor: semantics.color.text.accent.green,
  tagStringColor: semantics.color.text.accent.red,
  tagDefinitionVariableNameColor: semantics.color.text.accent.blue,
  tagLocalVariableNameColor: semantics.color.text.accent.sky,
  tagClassNameColor: semantics.color.text.accent.sea,
  tagSpecialVariableNameColor: semantics.color.text.accent.sea,
  tagDefinitionPropertyNameColor: semantics.color.text.accent.blue,
  tagCommentColor: semantics.color.text.accent.orange,
  tagMetaColor: semantics.color.text.muted,
  tagRegExpColor: semantics.color.text.accent.orange,
  tagTypeNameColor: semantics.color.text.accent.aurora,
  tagInvalidColor: semantics.color.text.error,
  tagAtomColor: semantics.color.text.accent.blue,
  activeLineColor: "#99999912"
});
var _default = exports.default = sourceCodeEditor;
