// @instructure/ui-top-nav-bar/lib/TopNavBar/v1/TopNavBarActionItems/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TopNavBarActionItems = void 0;
var _react = require("react");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _console = require("@instructure/console");
var _emotion = require("@instructure/emotion");
var _v11_ = require("@instructure/ui-drilldown/v11_6");
var _IconMoreLine = require("@instructure/ui-icons/lib/generated/IconMoreLine.js");
var _TruncateList = require("@instructure/ui-truncate-list/lib/TruncateList");
var _TopNavBarItem = require("../TopNavBarItem");
var _mapItemsForDrilldown = require("../utils/mapItemsForDrilldown");
var _TopNavBarContext = require("../TopNavBarContext");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _react2 = require("@emotion/react");
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: TopNavBar
id: TopNavBar.ActionItems
---
@module TopNavBarActionItems
**/
let TopNavBarActionItems = exports.TopNavBarActionItems = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec2 = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = _dec2(_class = class TopNavBarActionItems extends _react.Component {
  static displayName = "TopNavBarActionItems";
  static componentId = 'TopNavBar.ActionItems';
  static allowedProps = _props.allowedProps;
  static defaultProps = {};
  static contextType = _TopNavBarContext.TopNavBarContext;
  ref = null;
  _hiddenActionItemsMenuId;
  _hiddenActionItemsMenuTriggerId;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  constructor(props) {
    super(props);
    this._hiddenActionItemsMenuId = props.deterministicId('TopNavBarActionItems-hiddenActionItemsMenu');
    this._hiddenActionItemsMenuTriggerId = props.deterministicId('TopNavBarActionItems-hiddenActionItemsMenuTrigger');
    this.state = {
      key: 0,
      visibleActionItemsCount: undefined
    };
  }
  componentDidMount() {
    this.props.makeStyles?.(this.makeStylesVariables);
  }
  componentDidUpdate(prevProps) {
    this.props.makeStyles?.(this.makeStylesVariables);
    if (this.context.layout === 'smallViewport' && _react.Children.count(prevProps.children) !== _react.Children.count(this.props.children)) {
      this.setState({
        key: this.state.key + 1
      });
    }
  }
  get makeStylesVariables() {
    return {
      layout: this.context.layout
    };
  }
  get childrenArray() {
    const children = _react.Children.toArray(this.props.children);
    return children.map(child => {
      if (!child?.props) {
        return null;
      }
      const {
        id,
        renderAvatar,
        renderIcon,
        variant
      } = child.props;
      if (renderAvatar) {
        (0, _console.warn)(false, `Items in <TopNavBar.ActionItems> are not allowed to have avatars, please remove it from item with the id "${id}".`);
        return (0, _safeCloneElement.safeCloneElement)(child, {
          renderAvatar: undefined,
          variant: 'default'
        });
      }
      if (this.context.layout === 'smallViewport' && variant !== 'icon') {
        if (!renderIcon) {
          (0, _console.error)(false, `Items in <TopNavBar.ActionItems> are required to have the \`renderIcon\` prop, because only the icons are displayed due to the lack of space. Please add an icon to the item with the id "${id}".`);
          return null;
        }
        return (0, _safeCloneElement.safeCloneElement)(child, {
          variant: variant === 'forceIconWithLabel' ? 'default' : 'icon'
        });
      }
      return child;
    });
  }
  renderOptionContent = (children, itemProps) => {
    const {
      styles
    } = this.props;
    const {
      status
    } = itemProps;
    return (0, _jsxRuntime.jsx)("span", {
      css: status === 'active' ? styles?.dropdownMenuOptionActive : styles?.dropdownMenuOption,
      children: children
    });
  };
  renderHiddenActionItemsMenu(hiddenItems) {
    const {
      renderHiddenItemsMenuTriggerLabel,
      renderHiddenItemsMenuTriggerTooltip
    } = this.props;
    const mappedItems = (0, _mapItemsForDrilldown.mapItemsForDrilldown)(hiddenItems, {
      renderOptionContent: this.renderOptionContent
    });
    const subPages = (0, _mapItemsForDrilldown.renderMappedItemDrilldownSubpages)(mappedItems);
    const options = (0, _mapItemsForDrilldown.renderMappedItemsAsDrilldownOptions)(mappedItems);
    const label = typeof renderHiddenItemsMenuTriggerLabel === 'function' ? renderHiddenItemsMenuTriggerLabel(hiddenItems.length) : renderHiddenItemsMenuTriggerLabel;
    const tooltip = typeof renderHiddenItemsMenuTriggerTooltip === 'function' ? renderHiddenItemsMenuTriggerTooltip(hiddenItems.length) : renderHiddenItemsMenuTriggerTooltip;
    return (0, _jsxRuntime.jsx)(_TopNavBarItem.TopNavBarItem, {
      id: this._hiddenActionItemsMenuTriggerId,
      renderIcon: _IconMoreLine.IconMoreLine,
      variant: "icon",
      tooltip: tooltip,
      showSubmenuChevron: false,
      renderSubmenu: (0, _jsxRuntime.jsx)(_v11_.Drilldown, {
        shouldSetAriaExpanded: false,
        rootPageId: this._hiddenActionItemsMenuId,
        children: [(0, _jsxRuntime.jsx)(_v11_.Drilldown.Page, {
          id: this._hiddenActionItemsMenuId,
          children: options
        }, this._hiddenActionItemsMenuId), ...subPages]
      }),
      children: label
    });
  }
  renderTruncatedActionItemList() {
    const {
      listLabel,
      styles,
      renderHiddenItemsMenuTriggerLabel,
      renderHiddenItemsMenuTriggerTooltip,
      ...restProps
    } = this.props;
    return (0, _react2.createElement)(_TruncateList.TruncateList, {
      ...(0, _passthroughProps.passthroughProps)((0, _omitProps.omitProps)(restProps, _props.allowedProps)),
      key: this.state.key // rerender if child count changes
      ,
      elementRef: this.handleRef,
      css: styles?.topNavBarActionItems,
      visibleItemsCount: this.state.visibleActionItemsCount,
      onUpdate: ({
        visibleItemsCount
      }) => {
        this.setState({
          visibleActionItemsCount: visibleItemsCount
        });
      },
      renderHiddenItemMenu: hiddenChildren => this.renderHiddenActionItemsMenu(hiddenChildren),
      "aria-label": listLabel
    }, this.childrenArray);
  }
  render() {
    const {
      listLabel,
      styles,
      renderHiddenItemsMenuTriggerLabel,
      renderHiddenItemsMenuTriggerTooltip,
      ...restProps
    } = this.props;
    if (!this.childrenArray.length) {
      return null;
    }
    if (this.context.layout === 'smallViewport') {
      return this.renderTruncatedActionItemList();
    }
    return (0, _jsxRuntime.jsx)("ul", {
      ...(0, _omitProps.omitProps)(restProps, _props.allowedProps),
      ref: this.handleRef,
      css: styles?.topNavBarActionItems,
      "aria-label": listLabel,
      "data-cid": "TopNavBarActionItems",
      children: this.childrenArray.map(item => (0, _jsxRuntime.jsx)("li", {
        css: styles?.listItem,
        children: item
      }, item.props.id))
    });
  }
}) || _class) || _class);
var _default = exports.default = TopNavBarActionItems;
