// @instructure/ui-themes/src/themes/newThemeTokens/legacyCanvasHighContrast/components/checkbox.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { Checkbox } from "../../componentTypes/checkbox"
import type { Semantics } from "../semantics"

const checkbox = (semantics: Semantics): Checkbox => ({
  backgroundColor: semantics.color.background.interactive.input.base,
  backgroundHoverColor: semantics.color.background.interactive.input.hover,
  backgroundReadonlyColor: semantics.color.background.interactive.input.readonly,
  backgroundDisabledColor: semantics.color.background.interactive.input.disabled,
  backgroundCheckedColor: semantics.color.institutional.brandFontColorDark,
  borderColor: semantics.color.stroke.interactive.input.base,
  borderHoverColor: semantics.color.stroke.interactive.input.hover,
  borderReadonlyColor: semantics.color.stroke.interactive.input.readonly,
  borderDisabledColor: semantics.color.stroke.interactive.input.disabled,
  errorBorderColor: semantics.color.stroke.interactive.action.destructive.secondary.base,
  errorBorderHoverColor: semantics.color.stroke.interactive.action.destructive.secondary.hover,
  asteriskColor: semantics.color.text.error,
  borderCheckedColor: semantics.color.institutional.brandFontColorDark,
  fontSizeSm: semantics.fontSize.textSm,
  fontSizeMd: semantics.fontSize.textBase,
  fontSizeLg: "1.375rem",
  lineHeight: semantics.lineHeight.standalone.base,
  fontFamily: semantics.fontFamily.base,
  fontWeight: semantics.fontWeight.body.base,
  gap: semantics.spacing.spaceSm,
  controlVerticalMargin: "0.0625rem",
  borderRadius: semantics.borderRadius.sm,
  controlSizeSm: semantics.size.choiceControl.height.sm,
  controlSizeMd: semantics.size.choiceControl.height.md,
  controlSizeLg: semantics.size.choiceControl.height.lg,
  borderWidth: semantics.borderWidth.interactive.base,
  labelBaseColor: semantics.color.institutional.brandFontColorDark,
  labelHoverColor: semantics.color.institutional.brandFontColorDark,
  labelDisabledColor: semantics.color.text.interactive.disabled.base,
  labelReadonlyColor: semantics.color.institutional.brandFontColorDark,
})
export default checkbox
