// @instructure/ui-themes/src/themes/newThemeTokens/legacyCanvasHighContrast/components/link.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { Link } from "../../componentTypes/link"
import type { Semantics } from "../semantics"

const link = (semantics: Semantics): Link => ({
  textColor: semantics.color.institutional.linkColor,
  textHoverColor: semantics.color.text.interactive.navigation.primary.hover,
  textDisabledColor: semantics.color.text.interactive.disabled.base,
  onColorTextColor: semantics.color.text.interactive.navigation.primaryOnColor.base,
  onColorTextHoverColor: semantics.color.text.interactive.navigation.primaryOnColor.hover,
  onColorTextDisabledColor: semantics.color.text.interactive.disabled.onColor,
  fontSizeSm: semantics.fontSize.textSm,
  fontSizeMd: semantics.fontSize.textBase,
  fontSizeLg: "22px",
  fontWeight: semantics.fontWeight.interactive,
  gapSm: semantics.spacing.space2xs,
  gapMd: semantics.spacing.spaceXs,
  gapLg: semantics.spacing.spaceXs,
  lineHeightSm: semantics.lineHeight.standalone.textSm,
  lineHeightMd: semantics.lineHeight.standalone.textBase,
  lineHeightLg: "33px",
  fontFamily: semantics.fontFamily.base,
  inlineLink: {
    small: {
      fontFamily: semantics.fontFamily.base,
      fontWeight: semantics.fontWeight.body.base,
      fontSize: semantics.fontSize.textSm,
      lineHeight: semantics.lineHeight.paragraph.textSm,
      textDecoration: "underline",
    },
    medium: {
      fontFamily: semantics.fontFamily.base,
      fontWeight: semantics.fontWeight.body.base,
      fontSize: semantics.fontSize.textBase,
      lineHeight: semantics.lineHeight.paragraph.textBase,
      textDecoration: "underline",
    },
    large: {
      fontFamily: semantics.fontFamily.base,
      fontWeight: semantics.fontWeight.body.base,
      fontSize: "22px",
      lineHeight: "33px",
      textDecoration: "underline",
    },
  },
  textDecorationWithinText: "underline",
  textDecorationOutsideText: "none",
  unstyledTextColor: semantics.color.text.base,
})
export default link
