// @instructure/ui-themes/src/themes/newThemeTokens/legacyCanvasHighContrast/components/textInput.ts
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import type { TextInput } from "../../componentTypes/textInput"
import type { Semantics } from "../semantics"

const textInput = (semantics: Semantics): TextInput => ({
  backgroundColor: semantics.color.background.interactive.input.base,
  backgroundHoverColor: semantics.color.background.interactive.input.hover,
  backgroundReadonlyColor: semantics.color.background.interactive.input.readonly,
  backgroundDisabledColor: semantics.color.background.interactive.input.disabled,
  borderColor: semantics.color.stroke.interactive.input.base,
  borderHoverColor: semantics.color.stroke.interactive.input.hover,
  borderReadonlyColor: semantics.color.stroke.interactive.input.readonly,
  borderDisabledColor: semantics.color.stroke.interactive.input.disabled,
  errorBorderColor: semantics.color.stroke.interactive.action.destructive.secondary.base,
  successBorderColor: semantics.color.stroke.interactive.input.base,
  borderRadius: semantics.borderRadius.interactive.base,
  borderWidth: semantics.borderWidth.interactive.base,
  textColor: semantics.color.institutional.brandFontColorDark,
  textReadonlyColor: semantics.color.institutional.brandFontColorDark,
  textDisabledColor: semantics.color.text.interactive.input.disabled,
  placeholderColor: semantics.color.text.interactive.input.placeholder,
  placeholderHoverColor: semantics.color.text.interactive.input.placeholder,
  arrowsContainerWidth: "2rem",
  arrowsBackgroundColor: semantics.color.background.interactive.action.secondary.base,
  arrowsBackgroundHoverColor: semantics.color.background.interactive.action.secondary.hover,
  arrowsBackgroundActiveColor: semantics.color.background.interactive.action.secondary.active,
  arrowsBackgroundDisabledColor: semantics.color.background.interactive.action.secondary.disabled,
  arrowsBorderColor: semantics.color.stroke.interactive.action.secondary.base,
  arrowsBorderHoverColor: semantics.color.stroke.interactive.action.secondary.hover,
  arrowsBorderActiveColor: semantics.color.stroke.interactive.action.secondary.active,
  arrowsBorderDisabledColor: semantics.color.stroke.interactive.action.secondary.disabled,
  fontFamily: semantics.fontFamily.base,
  fontWeight: semantics.fontWeight.body.base,
  fontSizeSm: semantics.fontSize.textSm,
  fontSizeMd: semantics.fontSize.textBase,
  fontSizeLg: "1.375rem",
  heightSm: semantics.size.interactive.height.sm,
  heightMd: semantics.size.interactive.height.md,
  heightLg: semantics.size.interactive.height.lg,
  gapContent: semantics.spacing.spaceMd,
  paddingHorizontalSm: semantics.spacing.spaceMd,
  paddingHorizontalMd: semantics.spacing.spaceMd,
  paddingHorizontalLg: semantics.spacing.spaceMd,
})
export default textInput
