// @instructure/ui-top-nav-bar/lib/TopNavBar/v1/TopNavBarLayout/DesktopLayout/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TopNavBarDesktopLayout = void 0;
var _react = require("react");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _emotion = require("@instructure/emotion");
var _TopNavBarContext = require("../../TopNavBarContext");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
private: true
---
**/
let TopNavBarDesktopLayout = exports.TopNavBarDesktopLayout = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class TopNavBarDesktopLayout extends _react.Component {
  static displayName = "TopNavBarDesktopLayout";
  static componentId = 'TopNavBar.DesktopLayout';
  static allowedProps = _props.allowedProps;
  static defaultProps = {};
  static contextType = _TopNavBarContext.TopNavBarContext;
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.(this.makeStylesVariables);
  }
  componentDidUpdate() {
    this.props.makeStyles?.(this.makeStylesVariables);
  }
  get makeStylesVariables() {
    return {
      inverseColor: this.context.inverseColor,
      hasBrandBlock: this.hasBrandBlock,
      hasActionItemsBlock: this.hasActionItemsBlock,
      hasUserBlock: this.hasUserBlock
    };
  }
  get hasBrandBlock() {
    const {
      renderBrand
    } = this.props;
    return !!renderBrand && !!renderBrand.props.renderIcon;
  }
  get hasActionItemsBlock() {
    const {
      renderActionItems
    } = this.props;
    return !!renderActionItems && _react.Children.count(renderActionItems.props.children) > 0;
  }
  get hasUserBlock() {
    const {
      renderUser
    } = this.props;
    return !!renderUser && _react.Children.count(renderUser.props.children) > 0;
  }
  get hasMenuItemsBlock() {
    const {
      renderMenuItems
    } = this.props;
    return !!renderMenuItems && _react.Children.count(renderMenuItems.props.children) > 0;
  }
  get hasBreadcrumbBlock() {
    const {
      renderBreadcrumb
    } = this.props;
    return !!renderBreadcrumb && _react.Children.count(renderBreadcrumb.props.children) > 0;
  }
  render() {
    const {
      renderBrand,
      renderMenuItems,
      renderActionItems,
      renderUser,
      renderBreadcrumb,
      navLabel,
      styles
    } = this.props;

    // only render breadcrumb if there is no brand or menu items
    const shouldRenderBreadcrumbBlock = !(this.hasBrandBlock || this.hasMenuItemsBlock) && this.hasBreadcrumbBlock;
    return (0, _jsxRuntime.jsxs)("nav", {
      ...(0, _omitProps.omitProps)(this.props, _props.allowedProps),
      ref: this.handleRef,
      css: styles?.topNavBarDesktopLayout,
      "aria-label": navLabel,
      "data-cid": "TopNavBarDesktopLayout",
      children: [this.hasBrandBlock && (0, _jsxRuntime.jsx)("div", {
        css: styles?.brandContainer,
        children: renderBrand
      }), this.hasMenuItemsBlock && (0, _jsxRuntime.jsx)("div", {
        css: styles?.menuItemsContainer,
        children: renderMenuItems
      }), shouldRenderBreadcrumbBlock && renderBreadcrumb, (0, _jsxRuntime.jsx)("span", {
        css: styles?.spacer
      }), this.hasActionItemsBlock && (0, _jsxRuntime.jsx)("div", {
        css: styles?.actionItemsContainer,
        children: renderActionItems
      }), this.hasUserBlock && (0, _jsxRuntime.jsx)("div", {
        css: styles?.userContainer,
        children: renderUser
      })]
    });
  }
}) || _class);
var _default = exports.default = TopNavBarDesktopLayout;
