// @instructure/ui-instructure/lib/DataPermissionLevels/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.DataPermissionLevels = void 0;
var _react = require("react");
var _v11_ = require("@instructure/ui-modal/v11_6");
var _v11_2 = require("@instructure/ui-buttons/v11_6");
var _v11_3 = require("@instructure/ui-heading/v11_6");
var _v11_4 = require("@instructure/ui-text/v11_6");
var _v11_5 = require("@instructure/ui-link/v11_6");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
category: components/AI Components
---
**/const DataPermissionLevels = ({
  modalLabel,
  title,
  data,
  closeButtonText,
  closeIconButtonScreenReaderLabel,
  currentFeature,
  currentFeatureText,
  triggerText,
  fullscreen = false
}) => {
  const [open, setOpen] = (0, _react.useState)(false);
  const styles = (0, _emotion.useStyle)({
    generateStyle: _styles.default,
    generateComponentTheme: _theme.default,
    componentId: 'DataPermissionLevels',
    displayName: 'DataPermissionLevels'
  });
  return (0, _jsxRuntime.jsxs)("div", {
    children: [(0, _jsxRuntime.jsx)(_v11_5.Link, {
      variant: "standalone",
      onClick: e => {
        e.preventDefault();
        setOpen(true);
      },
      forceButtonRole: false,
      href: "#",
      children: triggerText
    }), (0, _jsxRuntime.jsxs)(_v11_.Modal, {
      size: fullscreen ? 'fullscreen' : 'medium',
      open: open,
      onDismiss: () => {
        setOpen(false);
      },
      label: modalLabel,
      shouldCloseOnDocumentClick: true,
      children: [(0, _jsxRuntime.jsxs)(_v11_.Modal.Header, {
        spacing: "compact",
        children: [(0, _jsxRuntime.jsx)(_v11_3.Heading, {
          aiVariant: "stacked",
          children: title
        }), (0, _jsxRuntime.jsx)(_v11_2.CloseButton, {
          size: "medium",
          placement: "end",
          offset: "small",
          onClick: () => setOpen(false),
          screenReaderLabel: closeIconButtonScreenReaderLabel
        })]
      }), (0, _jsxRuntime.jsx)(_v11_.Modal.Body, {
        children: (0, _jsxRuntime.jsx)("div", {
          css: styles?.body,
          children: data.map(({
            level,
            title,
            description,
            highlighted
          }, index) => (0, _jsxRuntime.jsxs)("div", {
            css: highlighted ? styles?.highlightedCard : styles?.card,
            children: [highlighted ? (0, _jsxRuntime.jsxs)("div", {
              css: styles?.currentFeature,
              children: [(0, _jsxRuntime.jsxs)(_v11_3.Heading, {
                color: "primary-inverse",
                level: "reset",
                variant: "labelInline",
                children: [currentFeatureText, ' ']
              }), (0, _jsxRuntime.jsx)(_v11_4.Text, {
                color: "primary-inverse",
                variant: "content",
                children: currentFeature
              })]
            }) : null, (0, _jsxRuntime.jsxs)("div", {
              css: styles?.contentContainer,
              children: [(0, _jsxRuntime.jsx)("div", {
                css: styles?.level,
                children: level
              }), (0, _jsxRuntime.jsx)("div", {
                css: styles?.permissionTitle,
                children: (0, _jsxRuntime.jsxs)(_v11_4.Text, {
                  variant: "descriptionPage",
                  children: [title, " "]
                })
              }), (0, _jsxRuntime.jsx)(_v11_4.Text, {
                variant: "content",
                color: "secondary",
                children: description
              })]
            })]
          }, index))
        })
      }), (0, _jsxRuntime.jsx)(_v11_.Modal.Footer, {
        children: (0, _jsxRuntime.jsx)(_v11_2.Button, {
          onClick: () => setOpen(false),
          children: closeButtonText
        })
      })]
    })]
  });
};
exports.DataPermissionLevels = DataPermissionLevels;
var _default = exports.default = DataPermissionLevels;
