// @instructure/ui-table/es/Table/v2/ColHeader/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { omitProps, callRenderProp } from '@instructure/ui-react-utils';
import { withStyleNew } from '@instructure/emotion';
import generateStyle from "./styles.js";
import { allowedProps } from "./props.js";
import { ChevronsUpDownInstUIIcon, ChevronUpInstUIIcon, ChevronDownInstUIIcon } from '@instructure/ui-icons';

/**
---
parent: Table
id: Table.ColHeader
---
**/
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
let ColHeader = (_dec = withStyleNew(generateStyle), _dec(_class = class ColHeader extends Component {
  static displayName = "ColHeader";
  static componentId = 'Table.ColHeader';
  static allowedProps = allowedProps;
  static defaultProps = {
    textAlign: 'start',
    sortDirection: 'none',
    children: null,
    scope: 'col'
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  handleClick = event => {
    const {
      id,
      onRequestSort
    } = this.props;
    onRequestSort && onRequestSort(event, {
      id
    });
  };
  renderSortArrow() {
    const {
      sortDirection,
      onRequestSort
    } = this.props;
    if (sortDirection === 'ascending') {
      return _jsx(ChevronUpInstUIIcon, {
        color: "infoColor",
        size: "sm"
      });
    }
    if (sortDirection === 'descending') {
      return _jsx(ChevronDownInstUIIcon, {
        color: "infoColor",
        size: "sm"
      });
    }
    if (onRequestSort) {
      return _jsx(ChevronsUpDownInstUIIcon, {
        color: "baseColor",
        size: "sm"
      });
    }
    return undefined;
  }
  render() {
    const {
      onRequestSort,
      width,
      children,
      sortDirection,
      scope,
      styles
    } = this.props;
    return _jsxs("th", {
      ...omitProps(this.props, ColHeader.allowedProps),
      css: styles?.colHeader,
      style: {
        width
      },
      scope: scope,
      "aria-sort": sortDirection,
      children: [onRequestSort && _jsx("button", {
        onClick: this.handleClick,
        css: styles?.button,
        children: _jsxs("div", {
          css: styles?.buttonContent,
          children: [callRenderProp(children), this.renderSortArrow()]
        })
      }), !onRequestSort && children, !onRequestSort && this.renderSortArrow()]
    });
  }
}) || _class);
export default ColHeader;
export { ColHeader };
