// @instructure/ui-list/lib/List/v2/ListItem/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ListItem = void 0;
var _react = require("react");
var _latest = require("@instructure/ui-view/latest");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
parent: List
id: List.Item
---
**/
let ListItem = exports.ListItem = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class ListItem extends _react.Component {
  static displayName = "ListItem";
  static componentId = 'List.Item';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    padding: 'none',
    spacing: 'none',
    delimiter: 'none',
    size: 'medium'
  };
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  render() {
    const {
      delimiter,
      spacing,
      size,
      margin,
      padding,
      elementRef,
      children,
      styles,
      ...rest
    } = this.props;
    return (0, _jsxRuntime.jsx)(_latest.View, {
      ...(0, _passthroughProps.passthroughProps)(rest),
      css: styles?.listItem,
      as: "li",
      margin: margin,
      padding: padding,
      maxWidth: "100%",
      elementRef: this.handleRef,
      "data-cid": "ListItem",
      children: children
    });
  }
}) || _class);
var _default = exports.default = ListItem;
