// @instructure/ui-overlays/lib/Overlay/v1/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Overlay = void 0;
var _react = require("react");
var _createChainedFunction = require("@instructure/ui-utils/lib/createChainedFunction.js");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _pickProps = require("@instructure/ui-react-utils/lib/pickProps.js");
var _Dialog = require("@instructure/ui-dialog/lib/Dialog");
var _Portal = require("@instructure/ui-portal/lib/Portal");
var _Transition = require("@instructure/ui-motion/lib/Transition");
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
category: components
---
**/class Overlay extends _react.Component {
  static displayName = "Overlay";
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    open: false,
    insertAt: 'bottom',
    shouldContainFocus: false,
    shouldReturnFocus: false,
    shouldCloseOnDocumentClick: false,
    shouldCloseOnEscape: true,
    in: false,
    unmountOnExit: false,
    transitionOnMount: false,
    transitionEnter: true,
    transitionExit: true
  };
  constructor(props) {
    super(props);
    this.state = {
      open: props.open,
      transitioning: false
    };
  }
  _timeouts = [];
  _DOMNode = null;
  _isMounted = false;
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  componentDidMount() {
    this._isMounted = true;
  }
  componentDidUpdate(prevProps) {
    if (prevProps.open && !this.props.open) {
      // closing
      this.setState({
        transitioning: !!prevProps.transition
      });
    }
  }
  componentWillUnmount() {
    this._isMounted = false;
    this._timeouts.forEach(timeout => clearTimeout(timeout));
  }
  get DOMNode() {
    return this._DOMNode;
  }
  set DOMNode(el) {
    this._DOMNode = el;
  }
  handlePortalOpen = DOMNode => {
    this.DOMNode = DOMNode;
    this._timeouts.push(setTimeout(() => {
      if (this._isMounted) {
        this.setState({
          open: true
        });
      }
    }));
  };
  handleTransitionExited = _type => {
    this.setState({
      open: false,
      transitioning: false
    });
  };
  renderTransition(content) {
    return (0, _jsxRuntime.jsx)(_Transition.Transition, {
      ...(0, _pickProps.pickProps)(this.props, _Transition.Transition.allowedProps),
      in: this.props.open,
      transitionOnMount: true,
      unmountOnExit: true,
      type: this.props.transition,
      onExited: (0, _createChainedFunction.createChainedFunction)(this.handleTransitionExited, this.props.onExited),
      children: (0, _jsxRuntime.jsx)("div", {
        children: content
      })
    });
  }
  render() {
    let content = (0, _jsxRuntime.jsx)(_Dialog.Dialog, {
      ...(0, _omitProps.omitProps)(this.props, Overlay.allowedProps),
      ...(0, _pickProps.pickProps)(this.props, _Dialog.Dialog.allowedProps),
      defaultFocusElement: this.props.defaultFocusElement,
      open: this.state.open,
      elementRef: this.handleRef,
      children: this.props.children
    });
    if (this.props.transition) {
      content = this.renderTransition(content);
    }
    return (0, _jsxRuntime.jsx)(_Portal.Portal, {
      ...(0, _pickProps.pickProps)(this.props, _Portal.Portal.allowedProps),
      open: this.props.open || this.state.transitioning,
      onOpen: (0, _createChainedFunction.createChainedFunction)(this.handlePortalOpen, this.props.onOpen),
      "data-cid": "Overlay",
      children: content
    });
  }
}
exports.Overlay = Overlay;
var _default = exports.default = Overlay;
