// @instructure/ui-checkbox/es/Checkbox/v2/index.js
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import keycode from 'keycode';
import { FormFieldMessages } from '@instructure/ui-form-field/latest';
import { createChainedFunction } from '@instructure/ui-utils';
import { logError as error } from '@instructure/console';
import { isActiveElement } from '@instructure/ui-dom-utils';
import { omitProps, withDeterministicId } from '@instructure/ui-react-utils';
import { View } from '@instructure/ui-view/latest';
import { withStyleNew } from '@instructure/emotion';
import { CheckboxFacade } from "./CheckboxFacade/index.js";
import { ToggleFacade } from "./ToggleFacade/index.js";
import generateStyle from "./styles.js";
import { allowedProps } from "./props.js";
import { jsxs as _jsxs, jsx as _jsx } from "@emotion/react/jsx-runtime";
/**
---
category: components
tags: toggle, switch
---
**/
let Checkbox = (_dec = withDeterministicId(), _dec2 = withStyleNew(generateStyle), _dec(_class = _dec2(_class = class Checkbox extends Component {
  static displayName = "Checkbox";
  static componentId = 'Checkbox';
  static allowedProps = allowedProps;
  static defaultProps = {
    size: 'medium',
    variant: 'simple',
    disabled: false,
    inline: false,
    indeterminate: false,
    readOnly: false,
    labelPlacement: 'end'
  };
  constructor(props) {
    super(props);
    this.state = {
      focused: false,
      hovered: false,
      checked: typeof props.checked === 'undefined' ? !!props.defaultChecked : undefined
    };
    if (typeof props.checked === 'undefined') {
      // @ts-expect-error ts-migrate(2339) FIXME: Property 'checked' does not exist on type 'Readonl... Remove this comment to see the full error message
      this.state.checked = !!props.defaultChecked;
    }
    this._defaultId = props.deterministicId();
  }
  _defaultId;
  _input = null;
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  handleInputRef = el => {
    this._input = el;
    if (typeof this.props.inputRef === 'function') {
      this.props.inputRef(el);
    }
  };
  componentDidMount() {
    if (this._input) {
      this._input.indeterminate = this.props.indeterminate;
    }
    this.props.makeStyles?.();
  }
  componentDidUpdate(prevProps) {
    if (prevProps.indeterminate !== this.props.indeterminate && this._input) {
      this._input.indeterminate = this.props.indeterminate || false;
    }
    this.props.makeStyles?.();
  }
  handleChange = e => {
    const {
      onChange,
      disabled,
      checked,
      readOnly
    } = this.props;
    if (disabled || readOnly) {
      e.preventDefault();
      return;
    }
    if (typeof checked === 'undefined') {
      this.setState({
        checked: !this.state.checked
      });
    }
    if (typeof onChange === 'function') {
      onChange(e);
    }
  };
  handleKeyDown = e => {
    if (this.props.variant === 'toggle' && e.keyCode === keycode.codes.enter) {
      this._input && this._input.click();
      e.preventDefault();
    }
  };
  handleFocus = () => {
    this.setState({
      focused: true
    });
  };
  handleBlur = () => {
    this.setState({
      focused: false
    });
  };
  handleMouseOver = () => {
    !this.props.disabled && this.setState({
      hovered: true
    });
  };
  handleMouseOut = () => {
    this.setState({
      hovered: false
    });
  };
  get id() {
    return this.props.id || this._defaultId;
  }
  get checked() {
    return typeof this.props.checked === 'undefined' ? this.state.checked : this.props.checked;
  }
  get focused() {
    return isActiveElement(this._input);
  }
  get isError() {
    return !!this.props.messages?.find(m => m.type === 'error' || m.type === 'newError');
  }
  get invalid() {
    return !!this.props.messages?.find(m => m.type === 'newError' || m.type === 'error');
  }
  focus() {
    this._input && this._input.focus();
  }
  renderFacade() {
    const {
      size,
      disabled,
      variant,
      label,
      readOnly,
      indeterminate,
      labelPlacement,
      themeOverride,
      isRequired,
      styles
    } = this.props;
    const {
      hovered,
      focused
    } = this.state;
    error(!(variant === 'simple' && labelPlacement !== 'end'), `[Checkbox] The \`simple\` variant does not support the \`labelPlacement\` property.  Use the \`toggle\` variant instead.`);
    if (variant === 'toggle') {
      return _jsxs(ToggleFacade, {
        disabled: disabled,
        size: size,
        hovered: hovered,
        focused: focused,
        checked: this.checked,
        readOnly: readOnly,
        labelPlacement: labelPlacement
        // @ts-ignore TODO: This will lead to buggy overrides, `Toggle`'s
        //  styles should be a superset of Checkbox's styles
        ,
        themeOverride: themeOverride,
        invalid: this.invalid,
        children: [label, isRequired && label && _jsxs("span", {
          css: this.invalid ? styles?.requiredInvalid : {},
          "aria-hidden": true,
          children: [' ', "*"]
        })]
      });
    } else {
      return _jsxs(CheckboxFacade, {
        disabled: disabled,
        size: size,
        hovered: hovered,
        focused: focused,
        checked: this.checked,
        readOnly: readOnly,
        indeterminate: indeterminate,
        themeOverride: themeOverride,
        invalid: this.invalid,
        children: [label, isRequired && label && _jsxs("span", {
          css: this.invalid ? styles?.requiredInvalid : {},
          "aria-hidden": true,
          children: [' ', "*"]
        })]
      });
    }
  }
  renderMessages() {
    const {
      messages,
      styles,
      variant
    } = this.props;
    return messages && messages.length > 0 ? _jsx(View, {
      display: "block",
      margin: "small 0 0",
      css: this.isError && (variant === 'toggle' ? styles?.indentedToggleError : styles?.indentedError),
      children: _jsx(FormFieldMessages, {
        messages: messages
      })
    }) : null;
  }
  render() {
    const {
      disabled,
      readOnly,
      value,
      onKeyDown,
      onFocus,
      onBlur,
      onMouseOver,
      onMouseOut,
      indeterminate,
      variant,
      styles,
      isRequired
    } = this.props;
    const props = omitProps(this.props, Checkbox.allowedProps);
    error(!(variant === 'toggle' && indeterminate), `[Checkbox] The \`toggle\` variant does not support the \`indeterminate\` property. Use the \`simple\` variant instead.`);
    return _jsx("div", {
      css: styles?.checkbox
      /* eslint-disable-next-line jsx-a11y/mouse-events-have-key-events */,
      onMouseOver: createChainedFunction(onMouseOver, this.handleMouseOver)
      /* eslint-disable-next-line jsx-a11y/mouse-events-have-key-events */,
      onMouseOut: createChainedFunction(onMouseOut, this.handleMouseOut),
      ref: this.handleRef,
      "data-cid": "Checkbox",
      children: _jsxs("div", {
        css: styles?.container,
        children: [_jsx("input", {
          "data-checked": indeterminate ? 'mixed' : this.checked,
          ...props,
          id: this.id,
          value: value,
          type: "checkbox",
          ref: this.handleInputRef,
          required: isRequired,
          disabled: disabled,
          readOnly: readOnly,
          "aria-readonly": readOnly ? true : undefined,
          "aria-checked": indeterminate ? 'mixed' : undefined,
          css: styles?.input,
          onClickCapture: e => {
            if (readOnly) {
              e.stopPropagation();
              e.preventDefault();
            }
          },
          onChange: this.handleChange,
          onKeyDown: createChainedFunction(onKeyDown, this.handleKeyDown),
          onFocus: createChainedFunction(onFocus, this.handleFocus),
          onBlur: createChainedFunction(onBlur, this.handleBlur),
          checked: this.checked
        }), _jsxs("label", {
          htmlFor: this.id,
          css: styles?.control,
          onMouseDown: e => {
            e.preventDefault();
            this._input?.focus();
          },
          children: [this.renderFacade(), this.renderMessages()]
        })]
      })
    });
  }
}) || _class) || _class);
export default Checkbox;
export { Checkbox, CheckboxFacade, ToggleFacade };
