// @instructure/ui-img/lib/Img/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Img = void 0;
var _react = require("react");
var _latest = require("@instructure/ui-view/latest");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _emotion = require("@instructure/emotion");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let Img = exports.Img = (_dec = (0, _emotion.withStyleNew)(_styles.default), _dec(_class = class Img extends _react.Component {
  static displayName = "Img";
  static componentId = 'Img';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    alt: '',
    display: 'inline-block',
    withGrayscale: false,
    withBlur: false
  };
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  render() {
    const {
      src,
      alt,
      margin,
      display,
      overlay,
      withGrayscale,
      withBlur,
      constrain,
      width,
      height,
      elementRef,
      styles,
      loading,
      ...props
    } = this.props;
    const a11yProps = {
      alt: alt || ''
    };
    const imageProps = {
      css: styles?.img,
      src,
      loading
    };
    const containerProps = {
      ...(0, _passthroughProps.passthroughProps)(props),
      width,
      height,
      margin,
      display,
      elementRef: this.handleRef
    };
    if (overlay) {
      // if a background image is rendered we add the a11y props on the container element
      const rootProps = {
        ...containerProps
      };
      return (0, _jsxRuntime.jsxs)(_latest.View, {
        ...rootProps,
        as: "span",
        css: styles?.container,
        "data-cid": "Img",
        children: [(0, _jsxRuntime.jsx)("img", {
          ...imageProps,
          ...a11yProps
        }), overlay && (0, _jsxRuntime.jsx)("span", {
          css: styles?.overlay
        })]
      });
    } else {
      return (0, _jsxRuntime.jsx)(_latest.View, {
        ...containerProps,
        ...imageProps,
        ...a11yProps,
        as: "img",
        "data-cid": "Img"
      });
    }
  }
}) || _class);
var _default = exports.default = Img;
