// @instructure/ui-list/lib/InlineList/v1/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.InlineList = void 0;
var _react = require("react");
var _v11_ = require("@instructure/ui-view/v11_6");
var _passthroughProps = require("@instructure/ui-react-utils/lib/passthroughProps.js");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _InlineListItem = require("./InlineListItem");
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
category: components
---

**/
class InlineList extends _react.Component {
  static displayName = "InlineList";
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    itemSpacing: 'none',
    as: 'ul',
    margin: 'none',
    delimiter: 'none',
    size: 'medium'
  };
  static Item = _InlineListItem.InlineListItem;
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  renderChildren() {
    const childrenArray = _react.Children.toArray(this.props.children);
    const lastIndex = childrenArray.length - 1;
    return childrenArray.map((child, index) => {
      if (!child) return null;
      const shouldRenderPlaceholder = index === lastIndex;
      return (0, _safeCloneElement.safeCloneElement)(child, {
        delimiter: this.props.delimiter,
        size: this.props.size,
        spacing: this.props.itemSpacing,
        lastPlaceholder: shouldRenderPlaceholder
      });
    });
  }
  render() {
    const {
      as,
      margin,
      elementRef,
      ...rest
    } = this.props;
    return (0, _jsxRuntime.jsx)(_v11_.View, {
      ...(0, _passthroughProps.passthroughProps)(rest),
      as: as,
      margin: margin,
      padding: "0",
      elementRef: this.handleRef,
      display: "block",
      "data-cid": "InlineList",
      children: this.renderChildren()
    });
  }
}
exports.InlineList = InlineList;
var _default = exports.default = InlineList;
