// @instructure/ui-themes/lib/themes/newThemeTokens/legacyCanvas/components/heading.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const heading = semantics => ({
  titlePageDesktop: {
    fontFamily: semantics.fontFamily.heading,
    fontWeight: semantics.fontWeight.heading.strong,
    fontSize: semantics.fontSize.text2xl,
    lineHeight: semantics.lineHeight.heading.base
  },
  titlePageMobile: {
    fontFamily: semantics.fontFamily.heading,
    fontWeight: semantics.fontWeight.heading.strong,
    fontSize: "2rem",
    lineHeight: semantics.lineHeight.heading.base
  },
  titleSection: {
    fontFamily: semantics.fontFamily.heading,
    fontWeight: semantics.fontWeight.heading.strong,
    fontSize: semantics.fontSize.textXl,
    lineHeight: semantics.lineHeight.heading.base
  },
  titleCardSection: {
    fontFamily: semantics.fontFamily.heading,
    fontWeight: semantics.fontWeight.heading.strong,
    fontSize: semantics.fontSize.textXl,
    lineHeight: semantics.lineHeight.heading.base
  },
  titleModule: {
    fontFamily: semantics.fontFamily.heading,
    fontWeight: semantics.fontWeight.heading.strong,
    fontSize: "1.5rem",
    lineHeight: semantics.lineHeight.heading.base
  },
  titleCardLarge: {
    fontFamily: semantics.fontFamily.heading,
    fontWeight: semantics.fontWeight.heading.strong,
    fontSize: "1.5rem",
    lineHeight: semantics.lineHeight.heading.base
  },
  titleCardRegular: {
    fontFamily: semantics.fontFamily.heading,
    fontWeight: semantics.fontWeight.heading.strong,
    fontSize: semantics.fontSize.textLg,
    lineHeight: semantics.lineHeight.heading.base
  },
  titleCardMini: {
    fontFamily: semantics.fontFamily.heading,
    fontWeight: semantics.fontWeight.heading.strong,
    fontSize: semantics.fontSize.textBase,
    lineHeight: semantics.lineHeight.heading.base
  },
  label: {
    fontFamily: semantics.fontFamily.heading,
    fontWeight: semantics.fontWeight.heading.strong,
    fontSize: semantics.fontSize.textBase,
    lineHeight: semantics.lineHeight.heading.base
  },
  labelInline: {
    fontFamily: semantics.fontFamily.heading,
    fontWeight: semantics.fontWeight.heading.strong,
    fontSize: semantics.fontSize.textBase,
    lineHeight: semantics.lineHeight.heading.base
  },
  h1FontSize: semantics.fontSize.text2xl,
  h2FontSize: semantics.fontSize.textXl,
  h3FontSize: semantics.fontSize.textLg,
  h4FontSize: semantics.fontSize.textBase,
  h5FontSize: semantics.fontSize.textSm,
  h6FontSize: semantics.fontSize.textXs,
  h1FontWeight: semantics.fontWeight.heading.strong,
  h2FontWeight: semantics.fontWeight.heading.base,
  h3FontWeight: semantics.fontWeight.heading.strong,
  h4FontWeight: semantics.fontWeight.heading.strong,
  h5FontWeight: semantics.fontWeight.heading.base,
  h6FontWeight: semantics.fontWeight.heading.base,
  h1FontFamily: semantics.fontFamily.heading,
  h2FontFamily: semantics.fontFamily.heading,
  h3FontFamily: semantics.fontFamily.heading,
  h4FontFamily: semantics.fontFamily.heading,
  h5FontFamily: semantics.fontFamily.heading,
  h6FontFamily: semantics.fontFamily.heading,
  lineHeight: semantics.lineHeight.heading.base,
  baseColor: semantics.color.institutional.brandFontColorDark,
  baseOnColor: semantics.color.text.onColor,
  mutedColor: semantics.color.text.muted,
  mutedOnColor: semantics.color.text.onColor,
  inverseColor: semantics.color.text.inverse,
  aiTextTopGradientColor: semantics.color.text.accent.violet,
  aiTextBottomGradientColor: semantics.color.text.accent.sea,
  borderColor: semantics.color.stroke.base,
  borderWidth: semantics.borderWidth.sm,
  borderStyle: "solid",
  borderPadding: semantics.spacing.space2xs,
  gapIconSm: semantics.spacing.space2xs,
  gapIconMd: semantics.spacing.spaceXs,
  gapIconLg: semantics.spacing.spaceSm
});
var _default = exports.default = heading;
