// @instructure/ui-themes/lib/themes/newThemeTokens/legacyCanvas/components/rangeInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const rangeInput = semantics => ({
  handleBackground: semantics.color.institutional.brandPrimary,
  handleHoverBackground: semantics.color.institutional.brandPrimary,
  handleFocusBackground: semantics.color.institutional.brandPrimary,
  handleFocusOutlineColor: semantics.color.institutional.brandPrimary,
  handleShadowColor: semantics.color.institutional.brandPrimary,
  handleBorderColor: semantics.color.stroke.onColor,
  handleBorderSize: semantics.borderWidth.md,
  handleFocusOutlineWidth: "0.75em",
  handleFocusInset: semantics.spacing.space2xs,
  trackBackground: semantics.color.background.accent.grey,
  valueFontFamily: semantics.fontFamily.base,
  valueFontWeight: semantics.fontWeight.body.base,
  valueSmallFontSize: semantics.fontSize.textSm,
  valueMediumFontSize: semantics.fontSize.textBase,
  valueLargeFontSize: semantics.fontSize.textLg,
  valueSmallPadding: semantics.spacing.spaceSm,
  valueMediumPadding: semantics.spacing.spaceMd,
  valueLargePadding: semantics.spacing.spaceXl,
  valueSmallLineHeight: "1.75rem",
  valueMediumLineHeight: "2.375rem",
  valueLargeLineHeight: "3rem",
  trackBorderColor: semantics.color.stroke.strong,
  minWidth: "12.5rem",
  handleSize: "1.5rem",
  boxShadow: {
    0: {
      x: semantics.dropShadow.x.elevation1.dropshadow1,
      y: semantics.dropShadow.y.elevation1.dropshadow1,
      blur: semantics.dropShadow.blur.elevation1.dropshadow1,
      spread: semantics.dropShadow.spread.elevation1.dropshadow1,
      color: semantics.color.dropShadow.shadowColor1,
      type: "dropShadow"
    },
    1: {
      x: semantics.dropShadow.x.elevation1.dropshadow2,
      y: semantics.dropShadow.y.elevation1.dropshadow2,
      blur: semantics.dropShadow.blur.elevation1.dropshadow2,
      spread: semantics.dropShadow.spread.elevation1.dropshadow2,
      color: semantics.color.dropShadow.shadowColor2,
      type: "dropShadow"
    }
  }
});
var _default = exports.default = rangeInput;
