// @instructure/ui-themes/lib/themes/newThemeTokens/legacyCanvas/components/treeBrowserTreeButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

const treeBrowserTreeButton = semantics => ({
  baseSpacingLarge: semantics.spacing.spaceLg,
  baseSpacingMedium: semantics.spacing.spaceMd,
  baseSpacingSmall: semantics.spacing.spaceSm,
  borderColor: semantics.color.stroke.base,
  borderRadius: semantics.borderRadius.md,
  borderWidth: semantics.borderWidth.sm,
  descriptorFontSizeLarge: semantics.fontSize.textSm,
  descriptorFontSizeMedium: semantics.fontSize.textXs,
  descriptorFontSizeSmall: semantics.fontSize.textXs,
  descriptorMarginTop: semantics.spacing.space2xs,
  descriptorTextColor: semantics.color.institutional.brandFontColorDark,
  hoverBackgroundColor: semantics.color.background.brand,
  hoverTextColor: semantics.color.text.onColor,
  nameFontSizeLarge: semantics.fontSize.textBase,
  nameFontSizeMedium: semantics.fontSize.textSm,
  nameFontSizeSmall: semantics.fontSize.textXs,
  nameTextColor: semantics.color.text.interactive.navigation.primary.base,
  selectedBackgroundColor: semantics.color.background.accent.grey,
  selectedTextColor: semantics.color.text.onColor,
  textLineHeight: semantics.lineHeight.standalone.base,
  iconsMarginRightSmall: semantics.spacing.spaceXs,
  iconsMarginRightMedium: semantics.spacing.spaceSm,
  iconsMarginRightLarge: semantics.spacing.spaceLg
});
var _default = exports.default = treeBrowserTreeButton;
