// @instructure/ui-side-nav-bar/lib/SideNavBar/v1/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SideNavBar = void 0;
Object.defineProperty(exports, "SideNavBarItem", {
  enumerable: true,
  get: function () {
    return _SideNavBarItem.SideNavBarItem;
  }
});
exports.default = void 0;
var _react = require("react");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _IconMoveStartLine = require("@instructure/ui-icons/lib/generated/IconMoveStartLine.js");
var _ScreenReaderContent = require("@instructure/ui-a11y-content/lib/ScreenReaderContent");
var _emotion = require("@instructure/emotion");
var _SideNavBarItem = require("./SideNavBarItem");
var _styles = _interopRequireDefault(require("./styles"));
var _theme = _interopRequireDefault(require("./theme"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
const navMinimized = ({
  minimized
}) => ({
  minimized: !minimized
});

/**
---
category: components
---
**/
let SideNavBar = exports.SideNavBar = (_dec = (0, _emotion.withStyle)(_styles.default, _theme.default), _dec(_class = class SideNavBar extends _react.Component {
  static displayName = "SideNavBar";
  static componentId = 'SideNavBar';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    children: null,
    defaultMinimized: false,
    // TODO we should investigate later if it used or not
    onClick: function (_e) {}
  };
  static Item = _SideNavBarItem.SideNavBarItem;
  ref = null;
  constructor(props) {
    super(props);
    this.state = {
      minimized: this.isControlled(props) ? !!props.minimized : !!props.defaultMinimized
    };
  }
  componentDidMount() {
    this.props.makeStyles?.({
      minimized: this.minimized
    });
  }
  componentDidUpdate() {
    this.props.makeStyles?.({
      minimized: this.minimized
    });
  }
  get minimized() {
    if (this.isControlled()) {
      return !!this.props.minimized;
    }
    return !!this.state.minimized;
  }
  isControlled(props = this.props) {
    return typeof props.minimized === 'boolean';
  }
  handleNavToggle = event => {
    if (!this.isControlled()) {
      this.setState(navMinimized);
    }
    if (typeof this.props.onMinimized === 'function') {
      this.props.onMinimized(event, !this.minimized);
    }
  };
  renderChildren() {
    return _react.Children.map(this.props.children, child => {
      if (! /*#__PURE__*/(0, _react.isValidElement)(child)) return null;
      const navItem = (0, _safeCloneElement.safeCloneElement)(child, {
        minimized: this.state.minimized
      });
      return (0, _jsxRuntime.jsx)("li", {
        css: this.props.styles?.list,
        children: navItem
      });
    });
  }
  toggleMessage() {
    return this.state.minimized ? this.props.toggleLabel.minimizedLabel : this.props.toggleLabel.expandedLabel;
  }
  render() {
    const {
      label
    } = this.props;
    const props = (0, _omitProps.omitProps)(this.props, SideNavBar.allowedProps, ['minimized']);
    return (0, _jsxRuntime.jsxs)("nav", {
      ...props,
      css: this.props.styles?.navigation,
      "data-cid": "SideNavBar",
      "aria-label": label,
      ref: element => {
        this.ref = element;
      },
      children: [(0, _jsxRuntime.jsx)("ul", {
        css: this.props.styles?.content,
        children: this.renderChildren()
      }), (0, _jsxRuntime.jsx)(_SideNavBarItem.SideNavBarItem, {
        "aria-expanded": !this.minimized,
        onClick: this.handleNavToggle,
        icon: (0, _jsxRuntime.jsx)(_IconMoveStartLine.IconMoveStartLine, {
          css: this.props.styles?.toggleIcon,
          inline: false
        }),
        label: (0, _jsxRuntime.jsx)(_ScreenReaderContent.ScreenReaderContent, {
          children: this.toggleMessage()
        })
      })]
    });
  }
}) || _class);
var _default = exports.default = SideNavBar;
