// @instructure/ui-view/es/ContextView/v2/index.js
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component } from 'react';
import { withStyleNew } from '@instructure/emotion';
import { omitProps } from '@instructure/ui-react-utils';
import { View } from "../../View/v2/index.js";
import generateStyle from "./styles.js";
import { allowedProps } from "./props.js";
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
/**
---
category: components
---
**/
let ContextView = (_dec = withStyleNew(generateStyle), _dec(_class = class ContextView extends Component {
  static displayName = "ContextView";
  static componentId = 'ContextView';
  static allowedProps = allowedProps;
  static defaultProps = {
    as: 'span',
    elementRef: () => {},
    debug: false,
    width: 'auto',
    height: 'auto',
    children: null,
    textAlign: 'start',
    background: 'default',
    shadow: 'above',
    placement: 'center end'
  };
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  ref = null;
  handleRef = el => {
    const {
      elementRef
    } = this.props;
    this.ref = el;
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  };
  render() {
    const {
      as,
      background,
      children,
      debug,
      height,
      width,
      maxHeight,
      maxWidth,
      minHeight,
      minWidth,
      margin,
      padding,
      shadow,
      stacking,
      style,
      textAlign,
      styles,
      borderColor
    } = this.props;
    return _jsx(View, {
      ...omitProps(this.props, ContextView.allowedProps),
      css: styles?.contextView,
      style: style,
      borderWidth: "none",
      display: "inline-block",
      as: as,
      withVisualDebug: debug,
      elementRef: this.handleRef,
      margin: margin,
      stacking: stacking,
      children: _jsxs(View, {
        css: styles?.contextView__content,
        display: "block",
        borderRadius: styles?.borderRadius,
        borderWidth: "small",
        borderColor: borderColor || (background === 'default' ? 'primary' : 'transparent'),
        background: background === 'default' ? 'primary' : 'primary-inverse',
        withVisualDebug: debug,
        height: height,
        width: width,
        maxHeight: maxHeight,
        maxWidth: maxWidth,
        minHeight: minHeight,
        minWidth: minWidth,
        padding: padding,
        shadow: shadow,
        textAlign: textAlign,
        children: [_jsx("span", {
          css: styles?.contextView__arrow
        }), children]
      })
    });
  }
}) || _class);
export default ContextView;
export { ContextView };
