// @instructure/ui-form-field/lib/FormFieldLayout/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormFieldLayout = void 0;
Object.defineProperty(exports, "allowedProps", {
  enumerable: true,
  get: function () {
    return _props.allowedProps;
  }
});
exports.default = void 0;
var _react = require("react");
var _hasVisibleChildren = require("@instructure/ui-a11y-utils/lib/hasVisibleChildren.js");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _useDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/useDeterministicId.js");
var _emotion = require("@instructure/emotion");
var _v = require("../../FormFieldMessages/v2");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
parent: FormField
---
**/
const FormFieldLayout = exports.FormFieldLayout = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    inline = false,
    layout = 'stacked',
    as = 'label',
    labelAlign = 'end',
    vAlign,
    label,
    messages,
    messagesId: messagesIdProp,
    children,
    width,
    elementRef,
    inputContainerRef,
    isGroup,
    isRequired = false,
    margin,
    disabled = false,
    readOnly = false,
    themeOverride,
    ...rest
  } = props;

  // Deterministic ID generation
  const [deterministicId, setDeterministicId] = (0, _react.useState)();
  const getId = (0, _useDeterministicId.useDeterministicId)('FormFieldLayout');
  (0, _react.useEffect)(() => {
    setDeterministicId(getId());
  }, []);
  const messagesId = messagesIdProp || deterministicId;
  const labelId = deterministicId ? `${deterministicId}-Label` : undefined;

  // Filter out error and success messages when disabled or readOnly
  const filteredMessages = disabled || readOnly ? messages?.filter(msg => msg.type !== 'error' && msg.type !== 'newError' && msg.type !== 'success') : messages;

  // Compute style props
  const hasMessages = filteredMessages && filteredMessages.length > 0 ? filteredMessages.some(msg => {
    if (msg.text) {
      if (typeof msg.text === 'string') {
        return msg.text.length > 0;
      }
      return true;
    }
    return false;
  }) : false;
  const hasVisibleLabel = label ? (0, _hasVisibleChildren.hasVisibleChildren)(label) : false;
  const hasErrorMsgAndIsGroup = !!filteredMessages?.find(m => m.type === 'error' || m.type === 'newError') && !!isGroup;
  const invalid = !!filteredMessages?.find(m => m.type === 'error' || m.type === 'newError');

  // Styles
  const styles = (0, _emotion.useStyleNew)({
    generateStyle: _styles.default,
    themeOverride,
    params: {
      hasMessages,
      hasVisibleLabel,
      hasErrorMsgAndIsGroup,
      inline,
      layout,
      vAlign,
      labelAlign,
      margin,
      messages: filteredMessages,
      isRequired,
      invalid
    },
    componentId: 'FormFieldLayout',
    displayName: 'FormFieldLayout'
  });
  const ElementType = as;
  const handleRef = (0, _react.useCallback)(el => {
    if (typeof ref === 'function') {
      ref(el);
    } else if (ref) {
      const refObject = ref;
      refObject.current = el;
    }
    if (typeof elementRef === 'function') {
      elementRef(el);
    }
  }, [ref, elementRef]);
  const handleInputContainerRef = (0, _react.useCallback)(node => {
    if (typeof inputContainerRef === 'function') {
      inputContainerRef(node);
    }
  }, [inputContainerRef]);
  const renderLabel = () => {
    const labelContent = hasVisibleLabel ? (0, _jsxRuntime.jsxs)(_jsxRuntime.Fragment, {
      children: [label, isRequired && (0, _jsxRuntime.jsxs)("span", {
        css: invalid ? styles?.requiredAsterisk : {},
        "aria-hidden": true,
        children: [' ', "*"]
      })]
    }) : label;
    if (hasVisibleLabel) {
      if (ElementType === 'fieldset') {
        // `legend` has some special built in CSS, this can only be reset
        // this way https://stackoverflow.com/a/65866981/319473
        return (0, _jsxRuntime.jsx)("legend", {
          style: {
            display: 'contents'
          },
          children: (0, _jsxRuntime.jsx)("span", {
            css: styles?.formFieldLabel,
            children: labelContent
          })
        });
      }
      return (0, _jsxRuntime.jsx)("span", {
        css: styles?.formFieldLabel,
        children: labelContent
      });
    } else if (label) {
      if (ElementType === 'fieldset') {
        return (0, _jsxRuntime.jsx)("legend", {
          id: labelId,
          style: {
            display: 'contents'
          },
          children: label
        });
      }
      // needs to be wrapped because it needs an `id`
      return (0, _jsxRuntime.jsx)("div", {
        id: labelId,
        style: {
          display: 'contents'
        },
        children: label
      });
    } else return null;
  };
  const renderVisibleMessages = () => {
    return hasMessages ? (0, _jsxRuntime.jsx)(_v.FormFieldMessages, {
      id: messagesId,
      messages: filteredMessages,
      gridArea: "messages"
    }) : null;
  };
  return (0, _jsxRuntime.jsxs)(ElementType, {
    ...(0, _omitProps.omitProps)(rest, [..._props.allowedProps]),
    css: styles?.formFieldLayout,
    "aria-describedby": hasMessages ? messagesId : undefined,
    "aria-errormessage": rest['aria-invalid'] ? messagesId : undefined,
    style: {
      width
    },
    ref: handleRef,
    children: [renderLabel(), hasErrorMsgAndIsGroup && renderVisibleMessages(), (0, _jsxRuntime.jsx)("span", {
      css: styles?.formFieldChildren,
      ref: handleInputContainerRef,
      children: children
    }), !hasErrorMsgAndIsGroup && renderVisibleMessages()]
  });
});
FormFieldLayout.displayName = 'FormFieldLayout';
var _default = exports.default = FormFieldLayout;
