// @instructure/ui-options/es/Options/v2/Item/index.js
var _dec, _dec2, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import { Component, createElement } from 'react';
import { omitProps, getElementType, callRenderProp, withDeterministicId } from '@instructure/ui-react-utils';
import { withStyleNew } from '@instructure/emotion';
import generateStyles from "./styles.js";
import { allowedProps } from "./props.js";
/**
---
parent: Options
id: Options.Item
---
**/
import { jsx as _jsx, jsxs as _jsxs } from "@emotion/react/jsx-runtime";
let Item = (_dec = withDeterministicId(), _dec2 = withStyleNew(generateStyles), _dec(_class = _dec2(_class = class Item extends Component {
  static displayName = "Item";
  static componentId = 'Options.Item';
  static allowedProps = allowedProps;
  static defaultProps = {
    as: 'span',
    variant: 'default',
    role: 'listitem',
    voiceoverRoleBugWorkaround: false,
    beforeLabelContentVAlign: 'center',
    afterLabelContentVAlign: 'center',
    isSelected: false
  };
  ref = null;
  _descriptionId;
  constructor(props) {
    super(props);
    this._descriptionId = props.deterministicId('OptionsItem-description');
  }
  componentDidMount() {
    this.props.makeStyles?.();
  }
  componentDidUpdate() {
    this.props.makeStyles?.();
  }
  renderContent(renderLabel, contentVariant) {
    const {
      styles,
      variant,
      as,
      role,
      children
    } = this.props;
    let labelContent;
    if (typeof renderLabel === 'function') {
      labelContent = /*#__PURE__*/createElement(renderLabel, {
        variant,
        as,
        role
      }, children);
    } else {
      labelContent = renderLabel;
    }
    return _jsx("span", {
      css: [styles?.content, contentVariant],
      role: "presentation",
      "aria-hidden": "true",
      children: labelContent
    });
  }
  render() {
    const {
      as,
      href,
      role,
      styles,
      description,
      descriptionRole,
      renderBeforeLabel,
      renderAfterLabel,
      elementRef,
      children,
      voiceoverRoleBugWorkaround
    } = this.props;
    const ElementType = getElementType(Item, this.props, () => as);
    const InnerElementType = href ? 'a' : 'span';
    const passthroughProps = omitProps(this.props, Item.allowedProps);
    const childrenContent = callRenderProp(children);
    const descriptionContent = callRenderProp(description);
    const ariaDescribedBy = this.props['aria-describedby'] || (descriptionContent ? this._descriptionId : undefined);
    return _jsxs(ElementType, {
      role: voiceoverRoleBugWorkaround ? role : 'none',
      "data-cid": "Options.Item",
      css: styles?.item,
      ref: element => {
        this.ref = element;
        if (typeof elementRef === 'function') {
          elementRef(element);
        }
      },
      "aria-describedby": voiceoverRoleBugWorkaround ? ariaDescribedBy : undefined,
      children: [_jsxs(InnerElementType, {
        ...passthroughProps,
        css: styles?.container,
        role: href || voiceoverRoleBugWorkaround ? undefined : role,
        href: href,
        "aria-describedby": voiceoverRoleBugWorkaround ? undefined : ariaDescribedBy,
        children: [childrenContent, descriptionContent && _jsx("span", {
          css: styles?.description,
          role: descriptionRole,
          id: this._descriptionId,
          children: descriptionContent
        })]
      }), renderBeforeLabel && this.renderContent(renderBeforeLabel, styles?.contentBefore), renderAfterLabel && this.renderContent(renderAfterLabel, styles?.contentAfter)]
    });
  }
}) || _class) || _class);
export default Item;
export { Item };
