// @instructure/ui-breadcrumb/lib/Breadcrumb/v1/BreadcrumbLink/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.BreadcrumbLink = void 0;
var _react = require("react");
var _v11_ = require("@instructure/ui-truncate-text/v11_6");
var _v11_2 = require("@instructure/ui-link/v11_6");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _v11_3 = require("@instructure/ui-tooltip/v11_6");
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
parent: Breadcrumb
id: Breadcrumb.Link
---
**/

class BreadcrumbLink extends _react.Component {
  static displayName = "BreadcrumbLink";
  static componentId = 'Breadcrumb.Link';
  static allowedProps = _props.allowedProps;
  static defaultProps = {};
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  constructor(props) {
    super(props);
    this.state = {
      isTruncated: false
    };
  }
  handleTruncation(isTruncated) {
    if (isTruncated !== this.state.isTruncated) {
      this.setState({
        isTruncated
      });
    }
  }
  render() {
    const {
      children,
      href,
      renderIcon,
      iconPlacement,
      onClick,
      onMouseEnter,
      isCurrentPage
    } = this.props;
    const {
      isTruncated
    } = this.state;
    const props = (0, _omitProps.omitProps)(this.props, BreadcrumbLink.allowedProps);
    const isInteractive = onClick || href;
    return (0, _jsxRuntime.jsx)(_v11_3.Tooltip, {
      renderTip: children,
      preventTooltip: !isTruncated
      // this wraps the achor/button tag in a span and puts the aria-describedby on that instead of the anchor/button tag
      // to avoid SRs reading the text twice when there is already an aria-label
      ,
      ...(isInteractive && {
        as: 'span'
      }),
      children: (0, _jsxRuntime.jsx)(_v11_2.Link, {
        ...props,
        href: href,
        renderIcon: renderIcon,
        iconPlacement: iconPlacement,
        onClick: onClick,
        onMouseEnter: onMouseEnter,
        isWithinText: false,
        elementRef: this.handleRef,
        forceButtonRole: false,
        ...(isCurrentPage && {
          'aria-current': 'page'
        }),
        ...(isTruncated && {
          ...(typeof children === 'string' && {
            'aria-label': children
          }),
          ...(!isInteractive && {
            role: 'text'
          })
        }),
        "data-cid": "BreadcrumbLink",
        children: (0, _jsxRuntime.jsx)(_v11_.TruncateText, {
          onUpdate: isTruncated => this.handleTruncation(isTruncated),
          children: (0, _jsxRuntime.jsx)("span", {
            "aria-hidden": isTruncated,
            children: children
          })
        })
      })
    });
  }
}
exports.BreadcrumbLink = BreadcrumbLink;
var _default = exports.default = BreadcrumbLink;
