// @instructure/ui-spinner/lib/Spinner/v2/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Spinner = void 0;
var _react = require("react");
var _emotion = require("@instructure/emotion");
var _callRenderProp = require("@instructure/ui-react-utils/lib/callRenderProp.js");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _useDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/useDeterministicId.js");
var _console = require("@instructure/console");
var _styles = _interopRequireDefault(require("./styles"));
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
---
category: components
---
**/const Spinner = exports.Spinner = /*#__PURE__*/(0, _react.forwardRef)((props, ref) => {
  const {
    size = 'medium',
    variant = 'default',
    delay,
    renderTitle,
    margin,
    themeOverride
  } = props;
  const [shouldRender, setShouldRender] = (0, _react.useState)(!delay);
  // Deterministic ID generation
  const [titleId, setTitleId] = (0, _react.useState)();
  const getId = (0, _useDeterministicId.useDeterministicId)('Spinner');
  (0, _react.useEffect)(() => {
    setTitleId(getId());
  }, []);
  const styles = (0, _emotion.useStyleNew)({
    generateStyle: _styles.default,
    themeOverride,
    params: {
      size,
      variant,
      margin
    },
    componentId: 'Spinner',
    displayName: 'Spinner'
  });
  (0, _react.useEffect)(() => {
    if (delay) {
      const delayTimeout = setTimeout(() => {
        setShouldRender(true);
      }, delay);
      return () => clearTimeout(delayTimeout);
    }
    return undefined;
  }, [delay]);
  const renderSpinner = () => {
    (0, _console.logError)(!!renderTitle, '[Spinner] The `renderTitle` prop is necessary for screen reader support.');
    const passthroughProps = (0, _omitProps.omitProps)(props, _props.allowedProps);
    return (0, _jsxRuntime.jsx)("div", {
      ...passthroughProps,
      css: styles?.spinner,
      ref: ref,
      "data-cid": "Spinner",
      children: (0, _jsxRuntime.jsxs)("svg", {
        css: styles?.circle,
        role: "img",
        "aria-labelledby": titleId,
        focusable: "false",
        children: [(0, _jsxRuntime.jsx)("title", {
          id: titleId,
          children: (0, _callRenderProp.callRenderProp)(renderTitle)
        }), (0, _jsxRuntime.jsxs)("g", {
          role: "presentation",
          children: [variant !== 'inverse' && (0, _jsxRuntime.jsx)("circle", {
            css: styles?.circleTrack,
            cx: "50%",
            cy: "50%",
            r: styles?.radius
          }), (0, _jsxRuntime.jsx)("circle", {
            css: styles?.circleSpin,
            cx: "50%",
            cy: "50%",
            r: styles?.radius
          })]
        })]
      })
    });
  };
  return shouldRender ? renderSpinner() : null;
});
Spinner.displayName = 'Spinner';
var _default = exports.default = Spinner;
