// @instructure/ui-checkbox/lib/CheckboxGroup/v1/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.CheckboxGroup = void 0;
var _react = require("react");
var _v11_ = require("@instructure/ui-form-field/v11_6");
var _matchComponentTypes = require("@instructure/ui-react-utils/lib/matchComponentTypes.js");
var _safeCloneElement = require("@instructure/ui-react-utils/lib/safeCloneElement.js");
var _pickProps = require("@instructure/ui-react-utils/lib/pickProps.js");
var _omitProps = require("@instructure/ui-react-utils/lib/omitProps.js");
var _withDeterministicId = require("@instructure/ui-react-utils/lib/DeterministicIdContext/withDeterministicId.js");
var _v = require("../../Checkbox/v1");
var _props = require("./props");
var _jsxRuntime = require("@emotion/react/jsx-runtime");
var _dec, _class;
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2015 - present Instructure, Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
/**
---
category: components
---
**/
let CheckboxGroup = exports.CheckboxGroup = (_dec = (0, _withDeterministicId.withDeterministicId)(), _dec(_class = class CheckboxGroup extends _react.Component {
  static displayName = "CheckboxGroup";
  static componentId = 'CheckboxGroup';
  static allowedProps = _props.allowedProps;
  static defaultProps = {
    disabled: false,
    readOnly: false,
    size: 'medium',
    layout: 'stacked',
    children: null
  };
  constructor(props) {
    super(props);
    if (typeof props.value === 'undefined') {
      this.state = {
        value: props.defaultValue || []
      };
    }
    this._messagesId = props.deterministicId();
  }
  _messagesId;
  ref = null;
  handleRef = el => {
    this.ref = el;
  };
  get hasMessages() {
    return this.props.messages && this.props.messages.length > 0;
  }
  handleChange = e => {
    const newValue = this.value;
    if (this.props.disabled || this.props.readOnly) {
      e.preventDefault();
      return;
    }
    if (e.target.checked) {
      newValue.push(e.target.value);
    } else {
      newValue.splice(newValue.indexOf(e.target.value), 1);
    }
    if (typeof this.props.value === 'undefined') {
      this.setState({
        value: newValue
      });
    }
    if (typeof this.props.onChange === 'function') {
      this.props.onChange(newValue);
    }
  };
  get value() {
    return typeof this.props.value === 'undefined' ? [...this.state.value] : [...this.props.value];
  }
  renderChildren() {
    const {
      children,
      name,
      size,
      disabled,
      readOnly
    } = this.props;
    return _react.Children.map(children, child => {
      if ((0, _matchComponentTypes.matchComponentTypes)(child, [_v.Checkbox])) {
        return (0, _safeCloneElement.safeCloneElement)(child, {
          key: `${child.props.name}`,
          name,
          label: child.props.label,
          disabled: disabled || child.props.disabled,
          readOnly: readOnly || child.props.readOnly,
          size,
          checked: this.value.indexOf(child.props.value) > -1,
          onChange: this.handleChange,
          width: child.props.width || 'auto',
          'aria-describedby': this.hasMessages ? this._messagesId : undefined
        });
      } else {
        return child;
      }
    });
  }
  render() {
    return (0, _jsxRuntime.jsx)(_v11_.FormFieldGroup, {
      ...(0, _omitProps.omitProps)(this.props, CheckboxGroup.allowedProps),
      ...(0, _pickProps.pickProps)(this.props, _v11_.FormFieldGroup.allowedProps),
      description: this.props.description,
      rowSpacing: "small",
      vAlign: "top",
      messagesId: this._messagesId,
      elementRef: this.handleRef,
      "data-cid": "CheckboxGroup",
      children: this.renderChildren()
    });
  }
}) || _class);
var _default = exports.default = CheckboxGroup;
